# Author: Marcos Ramos Gonzalez

## app.R ##
library(shinydashboard)
library(bslib)
library(gridlayout)
#library(shinyWidgets)
#library(semantic.dashboard)
library(DT)
library(ape)
#library(seqinr)
library(msaR)
library(shinyjs)
library(phylowidget)
library(plotly)

# Functions

## PFAM link
## https://www.ebi.ac.uk/interpro/entry/pfam/PF02728
pfam.link <- function(pfam.term)
{
  pfam.red.term <- strsplit(pfam.term, split = "[.]")[[1]][1]
  link <- paste0("https://www.ebi.ac.uk/interpro/entry/pfam/", pfam.red.term)
  complete.link <- paste(c("<a href=\"",
                           link,
                           "\" target=\"_blank\">",
                           pfam.term, "</a>"),
                         collapse = "")
  return(complete.link)
}

## Remove gaps from MAFFT Alignment
remove_gaps <- function(multseq)
{
  matrix1 <- do.call(rbind, multseq)
  bool <- apply(matrix1, MARGIN = 2, function(x) sum(x == "-") == nrow(matrix1))
  indexes <- which(bool)
  if(length(indexes) != 0)
  {
    multseq <- lapply(multseq, function(x) x[-indexes])
  }
  return(multseq)
}

## Gene Ontology term link
# http://amigo.geneontology.org/amigo/term/GO:0015979
go.link <- function(go.term)
{
  link <- paste0("http://amigo.geneontology.org/amigo/term/", go.term)
  complete.link <- paste(c("<a href=\"",
                           link,
                           "\" target=\"_blank\">",
                           go.term, "</a>"),
                         collapse = "")
  return(complete.link)
}

## KO term link
# https://www.genome.jp/dbget-bin/www_bget?ko:K12133
ko.link <- function(ko.term)
{
  link <- paste0("https://www.genome.jp/dbget-bin/www_bget?ko:", ko.term)
  complete.link <- paste(c("<a href=\"",
                           link,
                           "\" target=\"_blank\">",
                           ko.term, "</a>"),
                         collapse = "")
  return(complete.link)
}

## KEGG term link
# https://www.genome.jp/dbget-bin/www_bget?ko04712
kegg.link <- function(kegg.term)
{
  link <- paste0("https://www.genome.jp/dbget-bin/www_bget?", kegg.term)
  complete.link <- paste(c("<a href=\"",
                           link,
                           "\" target=\"_blank\">",
                           kegg.term, "</a>"),
                         collapse = "")
  return(complete.link)
}

# Coloring STRING Table
color_string <- function(int_type)
{
  color <- ifelse(int_type == "Interolog", "#CA931B",  "#196F3D")
  comp_chain <- paste0("<i style='color:{", color, "}'> {", int_type, "} </i>")
  return(comp_chain)

}

# JS Functions
# STRING network retrieval
jsCode <- "
    shinyjs.loadStringData = function(gene) {
        getSTRING('https://string-db.org', {
            'identifiers': gene,
            'add_color_nodes':25,
            'network_flavor':'confidence'})
    }"

# Create data frame

column1 <- c("Prasinoderma coloniale") # 1

column2 <- c("Ostreococcus tauri", "Ostreococcus lucimarinus", "Ostreococcus mediterraneus",
             "Bathycoccus prasinos", "Micromonas pusilla", "Micromonas commoda", 
             "Cymbomonas tetramitiformis") # 7

column3 <- c("Chlamydomonas reinhardtii", "Chlamydomonas incerta", "Haematococcus lacustris",
             "Volvox carteri", "Dunaliella salina", "Gonium pectorale", "Tetrabaena socialis",
             "Chromochloris zofingiensis", "Raphidocelis subcapitata", "Scenedesmus obliquus",
             "Desmodesmus armatus", "Monoraphidium minutum", "Caulerpa lentillifera",
             "Ostreobium quekettii", "Bryopsis sp.", "Coccomyxa subellipsoidea", 
             "Elliptochloris bilobata", "Myrmecia bisecta", "Apatococcus lobatus",
             "Asterochloris glomerata", "Trebouxia sp.", "Picochlorum soloecismus",
             "Chlorella sorokiniana", "Auxenochlorella protothecoides", "Helicosporidium sp.",
             "Ulva mutabilis", "Tetraselmis striata", "Pedinophyceae sp.", "Chloropicon primus",
             "Picocystis sp.") # 30

column4 <- c("Chlorokybus atmophyticus", "Mesostigma viride CCAC 1140", "Mesostigma viride NIES 296",
             "Klebsormidium nitens", "Chara braunii", "Mesotaenium endlicherianum",
             "Mesotaenium kramstae", "Spirogloea muscicola", "Zygnema circumcarinatum SAG",
             "Zygnema circumcarinatum UTEX 1559", "Zygnema cylindricum") # 11

column5 <- c("Ceratodon purpureus", "Physcomitrium patens", "Pleurozium schreberi", 
             "Funaria hygrometrica", "Sphagnum magellanicum", "Sphagnum fallax",
             "Takakia lepidozioides", "Marchantia polymorpha", "Marchantia paleacea",
             "Riccia fluitans", "Riccia sorocarpa", "Anthoceros agrestis", "Anthoceros punctatus",
             "Anthoceros fusiformis", "Megaceros flagellaris", "Phaeomegaceros chiloensis",
             "Phymatoceros phymatodes", "Paraphymatoceros pearsonii", "Notothylas orbicularis",
             "Phaeoceros carolinianus", "Phaeoceros sp.", "Leiosporoceros dussii") # 22

column6 <- c("Selaginella moellendorffii", "Isoetes sinensis", "Diphasiastrum complanatum",
             "Huperzia asiatica", "Lycopodium clavatum") # 5

column7 <- c("Ceratopteris richardii", "Adiantum capillus", "Alsophila spinulosa",
             "Salvinia cucullata", "Azolla filiculoides", "Marsilea vestita") # 6

column8 <- c("Thuja plicata", "Sequoia sempervirens", "Sequoiadendron giganteum",
             "Taxus chinensis", "Abies alba", "Picea abies", "Gnetum montanum", 
             "Welwitschia mirabilis", "Cycas panzhihuaensis", "Ginkgo biloba") # 10

column9 <- c("Amborella trichopoda", "Nymphaea colorata", "Cinnamomum kanehirae") # 3

column10 <- c("Aegilops tauschii", "Triticum aestivum", "Hordeum vulgare",
             "Brachypodium distachyon", "Oryza sativa", "Setaria italica",
             "Sorghum bicolor", "Zea mays", "Panicum hallii", "Miscanthus sinensis",
             "Oropetium thomaeum", "Eleusine coracana", "Ananas comosus", "Musa acuminata",
             "Dioscorea alata", "Spirodela polyrhiza", "Asparagus officinalis",
             "Zostera marina") # 18

column11 <- c("Arabidopsis thaliana", "Arabidopsis lyrata",
             "Capsella grandiflora", "Brassica rapa", "Brassica oleracea", 
             "Sisymbrium irio", "Schrenkiella parvula") # 47

column12 <- c("Eutrema salsugineum",
             "Thlaspi arvense", "Boechera stricta", "Carica papaya",
             "Theobroma cacao", "Gossypium hirsutum", "Gossypium raimondii",
             "Citrus sinensis", "Poncirus trifoliata", "Eucalyptus grandis",
             "Malus domestica", "Prunus persica", "Fragaria vesca", "Corylus avellana",
             "Carya illinoinensis", "Quercus rubra", "Cucumis sativus",
             "Glycine max", "Medicago truncatula", "Trifolium pratense", "Cicer arietinum",
             "Lotus japonicus", "Phaseolus acutifolius", "Vigna unguiculata",
             "Lupinus albus", "Lens culinaris", "Arachis hypogaea", 
             "Populus trichocarpa", "Manihot esculenta", "Salix purpurea",
             "Linum usitatissimum", "Corymbia citriodora", "Vitis vinifera",
             "Kalanchoe fedtschenkoi", "Vaccinium darrowii", "Olea europaea",
             "Mimulus guttatus", "Solanum lycopersicum", "Solanum tuberosum",
             "Coffea arabica", "Lactuca sativa", "Daucus carota", 
             "Beta vulgaris", "Amaranthus hypochondriacus", "Chenopodium quinoa",
             "Saponaria officinalis", "Aquilegia coerulea") # 47

         
organisms_values <- c("praco", "ot", "ostlu", "ostme", "bp",
                      "mi", "micco", "cymte", "cr", "chlin",
                      "haema", "vc", "ds", "gonpe", "tetso",
                      "cz", "rs", "sceobli", "desar", "monmi",
                      "caule", "ostqu", "brysp", "cocco",
                      "ellbi", "myrbi", "apalo", "astgl",
                      "tresp", "picso", "chlso", "auxpr",
                      "helsp", "um", "tetst", "pedsp",
                      "chlpr", "picsp", "ca", "mv", "mesvb",
                      "kn", "chara", "me", "meskr", "sp",
                      "zygci", "zygcb", "zygcy", 
                      "cp", "pp" , "plesc", "funhy",
                      "smag", "sphfa" , "takle", "mp",
                      "marpa", "ricfl" , "ricso", "aa",
                      "antpu", "antfu" , "megfl", "phach",
                      "phyph", "parpe" , "notor", "phaca",
                      "phasp", "leidu", 
                      "sm", "isosi", "dipco", "hupas", "lyccl",
                      "cri", "adica", "alssp", "sc", "af", "marve", 
                      "tp" , "seqse", "seqgi", "taxch", "abial",
                      "picab" , "gnemo", "welmi", "cyc", "ginbi",
                      "ambtr", "nymco", "cinka", "aegi", 
                      "ta", "horvu", "bradi", "os",
                      "setit", "sb", "zm", "panha",
                      "missi", "oroth", "eleco", "anaco",
                      "musac", "dioal", "spipo", "aspof", 
                      "zosma", "at", "araly", "capgr",
                      "brara", "braol", "sisir", "schpa", 
                      "eutsa", "thlar", "boest", "carpa", 
                      "theca", "goshi", "gosra", "citsi",
                      "pontr", "eucgr", "maldo", "prupe",
                      "frave", "corav", "caril", "queru",
                      "cucsa", "glyma", "medtr", "tripr",
                      "cicar", "lotja", "phaac", "vigun", 
                      "lupal", "lencu", "arahy", "poptr",
                      "manes", "salpu", "linus", "corci",
                      "vitvi", "kalfe", "vacda", "oleeu",
                      "mimgu", "sl", "soltu", "cofar",
                      "lacsa", "dauca", "betvu", "amahy", 
                      "chequ", "sapof", "aquco")


names(organisms_values) <- c(column1, column2, column3, column4, column5, column6,
                             column7, column8, column9, column10, column11, column12)

ui <- dashboardPage(
  
  dashboardHeader( #disable = TRUE
    title = "PharaohFUN",
     tags$li(a(href = 'https://www.ibvf.us-csic.es/', tags$img(src = 'logoibvf.png',
                                                               title = "IBVF", height = "20px"), style = "margin-top: 0px;"),
             class = "dropdown"),
     tags$li(a(href = 'https://www.us.es/', tags$img(src = 'Logo_US.png',
                                                     title = "US", height = "20px"), style = "margin-top: 0px;"),
             class = "dropdown"),
     tags$li(a(href = 'https://www.csic.es/es', tags$img(src = '2560px-Logotipo_del_CSIC.svg.png',
                                                         title = "CSIC", height = "20px"), style = "margin-top: 0px;"),
             class = "dropdown")
  ),
  
  dashboardSidebar(
    ## Sidebar content
    dashboardSidebar( 
      sidebarMenu(
        menuItem("Home", tabName = "home", icon = icon("house")),
        menuItem("Gene ID from organisms list", icon = icon("dna"), tabName = "gene_search"),
        menuItem("Sequence from organisms list", icon = icon("a"), tabName = "seq_search"),
        menuItem("Existing Orthogroup ID", icon = icon("dashboard"), tabName = "og_id_search"),
        menuItem("Batch mode", icon = icon("layer-group"), tabName = "batch_search"),
        menuItem("Sequence from new organism", icon = icon("leaf", lib = "glyphicon"), tabName = "shoot_search"),
        menuItem("Complete Datasets", icon = icon("th-list", lib = "glyphicon"), tabName = "whole_data"),
        menuItem("Download genomes", icon = icon("download") , tabName = "down_genomes"),
        menuItem("Source code", icon = icon("code"), 
                 href = "https://github.com/ramosgonzmarc/PharaohFUN"),
        menuItem("Contact and Info", icon = icon("envelope"), tabName = "contact_tutorial")
      )
    )
  ),
  dashboardBody(
    ## Body content
    
    tabItems(
      
      # First tab content
      tabItem(tabName = "home",
              #tags$head(tags$style(HTML("a {color: white}"))),
              #h1("PharaohFUN: Phylogenomic Analysis for Plant Protein History and Function Elucidation"),
              #h1(span(HTML(tab2), style = 'color:green; font-weight: bold;')),
              tags$br(),
              
              fluidRow(column(9,
                              tags$div(span("PharaohFUN: Phylogenomic Analysis for Plant Protein History and Function Elucidation",
                                            style = 'color: #1f618d; font-weight: 540; font-size: 42px; font-family: "Alatsi"", Verdana, sans-serif;')),
                              style = "margin-top: 45px; margin-left: 35px;"
                              
              ),
              column(2, 
                     img(
                       src = "pharaohlogo-removebg-preview.png",
                       alt = "logo",
                       width = 220,
                       height = 220, style="display: block; margin-left: 0px``; margin-right: auto;"
                     )
              ),
              column(1)
              ),
              # Put a margin on the right
              fluidRow(
                column(11,
                       tags$br(),
                       tags$br(),
                       tags$div(align="justify", tags$b("PharaohFUN"), "allows researchers to explore orthologous proteins across different evolutionarily 
                       distant species given a target protein. The tool focuses on viridiplantae microalgae
                       and land plants. A tutorial can be found", tags$a(href="https://www.youtube.com/watch?v=UniklbhHx3M", "here"), ". It offers flexibility in terms of the available search modalities, allowing the use of 
                             up to 5 different ones:", style = 'font-size: 18px; margin-left: 20px;'
                                
                       ),
                       fluidRow(tags$br()),
                       fluidRow(tags$br()),
                        fluidRow(
                          column(1),
                          column(12,
                                 valueBox(a("Gene ID", href="#shiny-tab-gene_search", "data-toggle" = "tab", "style" = "color:white"), 
                                          "Gene ID from one of the listed organisms",
                                          icon = icon("dna"), width = 4),
                                 
                                 valueBox(a("Sequence", href="#shiny-tab-seq_search", "data-toggle" = "tab", "style" = "color:white"), 
                                          "Protein sequence from one of the listed organisms",
                                          icon = icon("a"), width = 4, color = "lime"),
                                 
                                 valueBox(a("Orthogroup ID", href="#shiny-tab-og_id_search", "data-toggle" = "tab", "style" = "color:white"), 
                                          "Orthogroup ID (from STRING result)",
                                          icon = icon("dashboard"), width = 4, color = "red")
                          )
                        ),
                        fluidRow(
                          
                          column(2),
                          valueBox(a("Batch mode", href="#shiny-tab-batch_search", "data-toggle" = "tab", "style" = "color:white"), 
                                   "Sequences set from one of the listed organisms",
                                   icon = icon("layer-group"), width = 4, color = "orange"),
                          
                          valueBox(a("New organism", href="#shiny-tab-shoot_search", "data-toggle" = "tab", "style" = "color:white"), 
                                   "Protein sequence from any organism",
                                   icon = icon("leaf", lib = "glyphicon"), width = 4, color = "purple")
                        ),
                       
                       fluidRow(tags$br()),
                       tags$div(align="justify", style = 'font-size: 18px; margin-left: 30px;',
                                tags$ol(
                                  tags$li(tags$b("Gene ID-based search"), ". When the identifier of a gene in one of the species available 
                                       in the database is known, it allows the direct search for a gene. Only one gene
                                       per query corresponding to a species supported by the tool."),
                                  tags$li(tags$b("Sequence-based search"),". If an identifier is not known or differs from the nomenclature
                                       used by the tool, this modality maps an amino acid sequence onto the proteome of the 
                                       species, identifying the protein and eliminating possible ambiguities. Only one sequence
                                       per query corresponding to a species supported by the tool."),
                                  tags$li(tags$b("Orthogroup ID-based search"),". When the identifier of an orthogroup is known (typically 
                                       from STRING results of previous searches), it is possible to explore it directly
                                       without resorting to searching for individual genes. Only one orthogroup per query."),
                                  tags$li(tags$b("Batch mode search"),". It allows searching for multiple (up to 300) IDs or sequences 
                                       automatically, creating a folder with the sorted results. All genes must correspond
                                       to a single species supported by the tool."),
                                  tags$li(tags$b("New organism search"),". Unlike the previous cases, if you want to perform a study of 
                                       a protein that does not come from a species supported by the tool, this search method 
                                       allows the inclusion of the sequence in the gene tree and in subsequent analyses. Only
                                       one sequence per query from any organism.")
                                )
                       ),
                       
                       fluidRow(br()),
                       fluidRow(br()),
                       fluidRow(
                         column(4, 
                                tags$div(align="justify", style = 'font-size: 18px; margin-left: 20px;',  br(), 
                                         "The exploration of orthology is based on the construction of orthogroups, sets of genes that descend from a 
                         single gene in the common ancestor of the species under study. In this way, it is possible to trace the evolutionary 
                         history of these genes and to analyze the changes that the orthogroup has undergone from its appearance to 
                         its current situation in the extant species.", br(), br(), "Regarding the species supported by the tool, 
                         it currently integrates data from 167 species of the ", tags$b("Viridiplantae"), " clade, with representatives of the 
                         Prasinodermophyta, Chlorophyta and Streptophyta clades chosen to span the evolutionary set of these groups.",
                                         br(), br(),"After the construction of 
                         the tree, PharaohFUN allows the exploration of the proteins encoded by the orthogroup genes, implementing 
                         modules for interactive tree viewing, PFAM module determination, multiple sequence alignment, Gene Ontology 
                         terms annotation, KEGG pathways annotation, exploration of physical interactions between proteins 
                         and scientific literature annotation.", br(), br(), "The corresponding species tree is shown to the right,
                         including the rhodophyte clade used as outgroup.")
                         ),
                         column(1),
                         column(6,
                                tags$div(style = "margin-left: 20px;",
                                         img(
                                           src = "tree_pharaoh.png",
                                           alt = "species",
                                           width = 800,
                                           height = 1300, style="display: block; margin-left: 20px; margin-right: auto;"
                                         )
                                )
                         ),
                         column(1)
                         
                         
                       ),
                       fluidRow(br()),
                ),
                
                # Right Margin
                column(1)
              )
              
              
              
              
      ),
      
      # Second tab content
      tabItem(tabName = "gene_search",
              h2(""),
              fluidRow(valueBox("Gene ID-based search", 
                                subtitle = "Single gene, available organism",
                                icon = icon("dna"), width = 6)),
              br(),
              box(
                title = span(tags$b("Organism selection"), style = "color:#34c5d1; font-size: 20px; "), status = "info", width = "500",
                "Please select the desired organisms from the following list for performing the analysis.
                  Please take care of selecting the organism whose gene ID is being inputted.", br(), br(),
                column(3,
                       checkboxGroupInput(
                         "prasino_check_1",
                         p("Prasinodermophytes", class= "h4",
                           tags$img(
                             src = "palmophyllophyceae.png",
                             alt = "preasinodermophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column1,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "mami_check_1",
                         p("Mamiello/Pyramimonadophyceae", class= "h4",
                           tags$img(
                             src = "bathycoccus.png",
                             alt = "mamiellophyceae",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column2,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "chloro_check_1",
                         p("Other Chlorophytes", class= "h4",
                           tags$img(
                             src = "scenedesmus.png",
                             alt = "chlorophytes",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column3,
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "strepto_check_1",
                         p("Streptophyte algae", class= "h4",
                           tags$img(
                             src = "klebsormidium.png",
                             alt = "streptophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column4,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "bryo_check_1",
                         p("Bryophytes", class= "h4",
                           tags$img(
                             src = "marchantia.png",
                             alt = "bryophyta",
                             width = 32,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column5,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "lyco_check_1",
                         p("Lycophytes and Ferns", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "lycophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = c(column6, column7),
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "gymno_check_1",
                         p("Gymnosperms", class= "h4",
                           tags$img(
                             src = "gnetum.png",
                             alt = "gymnosperms",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column8,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "magno_check_1",
                         p("Other Angiosperms", class= "h4",
                           tags$img(
                             src = "amborella.png",
                             alt = "amborella",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column9,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "monocots_check_1",
                         p("Monocots", class= "h4",
                           tags$img(
                             src = "oryza.png",
                             alt = "monocots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column10,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "dicots_first_check_1",
                         p("Dicots", class= "h4",
                           tags$img(
                             src = "arabidopsis.png",
                             alt = "dicots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column11,
                         inline = F,
                         selected = NULL),
                       
                ),
                
                column(3,
                       checkboxGroupInput(
                         "dicots_second_check_1",
                         p("", class= "h4",
                           
                         ),
                         
                         choices = column12,
                         inline = F,
                         selected = NULL)
                ),
                
                fluidRow(br()),
                fluidRow(br()),
                
                span(tags$b("Tree building method"), style = "color:#34c5d1; font-size: 20px; "),
                div(br()),
                
                div("The first step in any study using PharaohFUN is the construction of the 
                    gene tree corresponding to the selected orthogroup. For this, four standard methods 
                    are offered, select one before continuing. Default is performed with FastTree using Orthofinder 2
                    pipeline, which renders trees by aproximate maximum-likelihood and then reconciliate
                    gene trees with the species tree shown in the Home tab. For the other three methods,
                    no reconciliation is performed but support for each branch is shown as bootstrap 
                    values."),
                
                shinyWidgets::awesomeRadio(
                  inputId = "build_trees_1",
                  label = "", 
                  choices = c("Maximum Likelihood", "Neighbour Joining", "UPGMA", "Bayesian Inference"),
                  selected = "Maximum Likelihood",
                  inline = TRUE, 
                  status = "info"
                ),
                
                fluidRow(br()),
                
                span(tags$b("Insert gene ID"), style = "color:#34c5d1; font-size: 20px; "),
                div(br()),
                
                div("Below you can write the ID associated to the protein whose evolutionary history you
                             wish to analyze."),
                
                fluidRow(
                  column(5, textInput("geneInt1", label = "", 
                                      width = "100%", placeholder = "AT2G46830")),
                  
                  column(1, div( style = "margin-top: 20px;", 
                                 shinyWidgets::actionBttn("run_button1", "Run", size = "sm", icon = icon("magnifying-glass"),
                                                          style = "float", color = "primary"))),
                  column(3)
                )
                
              ),
              
              br(),
              fluidRow(
                box(status = "info", width = 12, 
                    title = span(tags$b("Results"), style = "color:#34c5d1; font-size: 20px; "),
                    "Results for the query gene and species are displayed below, arranged in 
                      different tabs. The execution of each analysis is initiated from the start 
                      button within each of the tabs, with specific instructions for each analysis. First, go to Gene Tree tab,
                      which shows a gene tree containing the genes corresponding to the same orthogroup as the query in the
                      selected species. The rest of the tabs allow for a deeper exploration of these genes. Do not start any 
                      analysis before creating gene tree.",
                    div(br()),
                    tabsetPanel(type = "tabs",
                                tabPanel("Gene Tree",
                                         fluidRow(br()),
                                         div("This tab shows three different results. First of all, a complete list of the genes
                                               that are assigned to the same orthogroup as the query. Secondly, the proportion of
                                               genes of each species. Finally, a gene tree show the evolutionary relationships 
                                               between those genes. All results can be downloaded using the buttons at the bottom
                                               of the page."),
                                         fluidRow(br()),
                                         shinyjs::useShinyjs(),
                                         shinyjs::hidden(div(id='loading.tree1',h3('Please be patient, building tree ...'))),
                                         uiOutput(outputId = "error_tree1"),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_tree_text1"), tags$div(id = "box_tree_pie1")),
                                         fluidRow(tags$br()),
                                         fluidRow(tags$div(id = "box_tree_plot1")),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_tree1"),
                                                     tags$div(id = "download_newick1"),
                                                     tags$div(id = "download_tree_seqs1"))
                                ),
                                tabPanel("Collapsable tree",
                                         fluidRow(tags$br()),
                                         div("This tab allows an interactive visualization of the previous tree. 
                                               In particular, in situations where gene duplications have given rise 
                                               to several clades and you want to reduce the tree not in relation to 
                                               the species that appear, but to these clades, this visualization allows 
                                               the collapse of subtrees and the simple exploration of the areas of interest. 
                                               Press Show Collapsable Tree to show the tree."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("phylo_start1", "Show Collapsable Tree",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_phylo1"),
                                         fluidRow(tags$br())
                                ),
                                tabPanel("Expansion/Contraction",
                                         fluidRow(tags$br()),
                                         div("Click the Show Evolutionary History button to calculate the species tree showing 
                                               the reconstruction of ancestral states of the orthogroup, i.e. the number of genes 
                                               calculated for each common ancestor of each species. For orthogroups that
                                               have undergone expansions or contractions along any branch, these are marked 
                                               in red and blue on the tree, respectively. In turn, gains and losses of complete orthogroups
                                               are shown in green and purple, while the non-existence of orthogroup
                                               genes in the different clades is represented in gray."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("cafe_start1", "Show Evolutionary History",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         
                                         shinyjs::hidden(div(id='loading.cafe1', h3('Please be patient, reconstructing expansion/contraction events ...'))),
                                         tags$div(id = "error_cafe1"),
                                         tags$div(id = "box_mrca1"),
                                         tags$br(),
                                         tags$div(id = "box_cafe1"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "cafe_down_button1"),
                                                     tags$div(id = "download_ui_for_cafe_plot1"))
                                ),
                                tabPanel("PFAM Domains", 
                                         fluidRow(tags$br()),
                                         div("Next, click on the Show Gene Selection for Pfam button, select the desired proteins
                                               from the tree and click on the Show Pfam Domains button to determine their PFAM domains.
                                               A table with the domains of each protein and their positions will be displayed, as well 
                                               as a plot showing the same information. Links to each domain's data are provided.
                                               Warning: this process may take a long time in the case of selecting many proteins."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("pfam_start1", "Show Gene Selection for Pfam",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_pfams1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "pfam_selectionI1"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.pfam.pf1',h3('Please be patient, identifying domains ...'))),
                                         uiOutput(outputId = "error_pfam1"),
                                         tags$div(id = "box_pfam1"),
                                         tags$br(),
                                         tags$div(id = "box_pfplot1"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "pfam_down_button1"),
                                                     tags$div(id = "download_ui_for_pfam_table1")
                                         )
                                ),
                                tabPanel("Multiple Sequence Alignment",
                                         fluidRow(br()),
                                         div("Press the Show Gene Selection for MSA button to enable the selection of genes 
                                           from the tree to be aligned. Two alignment methods are supported: ClustalOmega uses this 
                                           algorithm to perform a de novo multiple sequence alignment of the selected proteins, while 
                                           the second option filters the precalculated MSA of the entire orthogroup (computed using the 
                                           software MAFFT) to retain 
                                           only the selected sequences, removing columns containing only gaps in the chosen subset. Thus,
                                           this second option retains the evolutionary framework to which the entire orthogroup is ascribed, 
                                           while the first aligns only the chosen sequences. The resulting alignment can be explored interactively,
                                           including searching for patterns USING Selection -> Find Motifs and different cholor schemes can be applied 
                                           based on different properties of amino acids. A consensus sequence is shown to summarize the alignment. For 
                                           a graphical representation with the different amino acids colored according to their chemical nature, click 
                                           on the Download Colored MSA button. The aligned sequences can also be download as a standard FASTA file.
                                           Warning: This process (specially the colored download) may take a long time in the case of selecting many proteins."),
                                         fluidRow(tags$br()),
                                         shinyWidgets::actionBttn("msa_start1", "Show Gene Selection for MSA",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_msa1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_method1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_selectionI1"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.msa1',h3('Please be patient, aligning sequences ...'))),
                                         uiOutput(outputId = "error_msa1"),
                                         tags$div(id = "box_msa1"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "msa_down_fasta1"),
                                                     tags$div(id = "msa_down_plot1"))
                                ),
                                tabPanel("GO Terms", 
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and select 
                                               the ontology type to display the GO terms associated with those genes. After
                                               pressing the Show GO terms button, the results are displayed in tabular form 
                                               and are accompanied by a GO association plot (each node is a GO and an edge 
                                               is drawn between two nodes if a gene from the set share both terms) and a treeplot
                                               showing the hierarchy of the identified terms and their summary. Links to each 
                                               individual GO term are avalaible from table and all results can be downloaded in
                                               standard PNG and TSV files."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("go_start1", "Show Gene Selection for GO Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos_mode1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "gos_selection1"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.go1',h3('Please be patient, preparing results ...'))),
                                         uiOutput(outputId = "error_gos1"),
                                         tags$div(id = "box_gos_table1"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_gos_plot1"), tags$div(id = "box_gos_treeplot1")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_gos_table1"),
                                                     tags$div(id = "gos_down_button1"),
                                                     tags$div(id = "tree_gos_down_button1"))
                                         
                                         
                                ),
                                tabPanel("KEGG Orthology",
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and press
                                               the Show KEGG information button to show the results. These include a table 
                                               showing the KEGG Orthology IDs, indicating how many and which of the selected 
                                               proteins correspond to each ID. In addition, the application performs an enrichment 
                                               in KEGG pathways from the identified KOs and allows for the plotting of these 
                                               pathways with the genes mapped onto them, using a selector to choose the pathway
                                               in case several enriched ones exist. Warning: this process may take a long time 
                                               in the case of selecting many proteins. Attempts to plot large pathways will
                                               produce an error, only more specific ones are supported i.e. Carbon metabolism 
                                               will return an error while Biotin metabolism will not."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("kegg_start1", "Show Gene Selection for KEGG Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_kos1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "kos_selection1"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.ko1',h3('Please be patient, exploring pathways ...'))),
                                         uiOutput(outputId = "error_kos1"),
                                         tags$div(id = "box_kos_table1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_kegg_table1"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "download_ui_for_kos_table1"),
                                                     tags$div(id = "download_ui_for_kegg_table1")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_paths1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "paths_button1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_path_image1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "path_download_ui1")
                                         
                                ),
                                tabPanel("STRING Interactions",
                                         fluidRow(tags$br()),
                                         div("This tab presents the results of interactions of the different proteins 
                                               for the species supported by the STRING database. After clicking the button,
                                               the proteins of the species included in the database are displayed. Once the
                                               desired selection has been made, press the Show String Interactions
                                               button and a table will be displayed
                                               indicating in each row an interaction and the type of evidence for its existence,
                                               either experimental (direct) or ortholog-based (interolog). In addition, a count
                                               of the orthogroups to which the target proteins belong is computed, presented in
                                               tabular form and by means of a pie chart, to assess preferential interaction of
                                               the orthogroup under study with other specific orthogroups. Finally, for
                                               specific proteins, an
                                               interaction network can be generated with the proteins that have the most reliable
                                               interactions with the selected protein. In case more than one protein is selected,
                                               all of them will appear in the network, which will contain their interactions to 
                                               determine pathways between them. The network is shown as an image and a button
                                               that creates the network is provided on the STRING page, where an interactive 
                                               network with additional protein structure information and SMART domains appears.
                                               Species currently supported by STRING: Aegilops, Arabidopsis, Bathycoccus, 
                                               Chara, Chlamydomonas, Coccomyxa,
                                               Klebsormidium, Micromonas, Oryza, Ostreococcus, 
                                               Physcomitrium, Raphidocelis, Scenedesmus, Selaginella, Solanum, Sorghum,
                                               Triticum and Volvox. Warning: the selection of many proteins in the first
                                               step can lead to particularly high waiting times."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("string_start1", "Show Gene Selection for STRING annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_string1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "string_selection1"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.string1',h3('Please be patient, retrieving interactions ...'))),
                                         uiOutput(outputId = "error_string1"),
                                         tags$div(id = "box_st_table1"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_count_table1"), tags$div(id = "box_count_plot1")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_st_table1"),
                                                     tags$div(id = "download_ui_for_count_table1"),
                                                     tags$div(id = "count_down_button1")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_network1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "network_button1"),
                                         fluidRow(tags$br()),
                                         uiOutput(outputId = "error_network1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_output_network1")
                                ),
                                tabPanel("Literature Annotation", 
                                         fluidRow(tags$br()),
                                         div("In the text box, type the search term to query your bibliographic information,
                                               i.e., CCA1. Then select one of the 4 search modes: Normal returns the entities 
                                               containing the term, Exact returns those that are identical, Alias returns all 
                                               aliases associated with the term and Substring returns all those containing a 
                                               given string. Associations found in the literature are returned in tabular form,
                                               with links to the papers from which the information was extracted. This
                                               functionality is based on PlantConnectome, to extend this analysis, it is recommended
                                               to go to https://connectome.plant.tools/."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("lit_start1", "Show Search Box for Literature Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "primary"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "query_lit1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_lit1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "lit_selection1"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.lit1',h3('Please be patient, browsing literature ...'))),
                                         uiOutput(outputId = "error_lit1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_lit_table1"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "download_ui_for_lit_table1")
                                         
                                )
                    ))
              )
      ),
      
      
      # Third tab content
      tabItem(tabName = "seq_search",
              h2(""),
              fluidRow(valueBox("Sequence-based search", 
                                subtitle = "Single gene, available organism",
                                icon = icon("a"), width = 6, color = "lime")),
              br(),
              box(
                title = span(tags$b("Organism selection"), style = "color:#25d04a; font-size: 20px; "), status = "success", width = "500",
                "Please select the desired organisms from the following list for performing the analysis.
                  Please take care of selecting the organism whose gene ID is being inputted.", br(), br(),
                column(3,
                       checkboxGroupInput(
                         "prasino_check_2",
                         p("Prasinodermophytes", class= "h4",
                           tags$img(
                             src = "palmophyllophyceae.png",
                             alt = "prasinodermophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column1,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "mami_check_2",
                         p("Mamiello/Pyramimonadophyceae", class= "h4",
                           tags$img(
                             src = "bathycoccus.png",
                             alt = "mamiellophyceae",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column2,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "chloro_check_2",
                         p("Other Chlorophytes", class= "h4",
                           tags$img(
                             src = "scenedesmus.png",
                             alt = "chlorophytes",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column3,
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "strepto_check_2",
                         p("Streptophyte algae", class= "h4",
                           tags$img(
                             src = "klebsormidium.png",
                             alt = "streptophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column4,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "bryo_check_2",
                         p("Bryophytes", class= "h4",
                           tags$img(
                             src = "marchantia.png",
                             alt = "bryophyta",
                             width = 32,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column5,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "lyco_check_2",
                         p("Lycophytes and Ferns", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "lycophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = c(column6, column7),
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "gymno_check_2",
                         p("Gymnosperms", class= "h4",
                           tags$img(
                             src = "gnetum.png",
                             alt = "gymnosperms",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column8,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "magno_check_2",
                         p("Other Angiosperms", class= "h4",
                           tags$img(
                             src = "amborella.png",
                             alt = "amborella",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column9,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "monocots_check_2",
                         p("Monocots", class= "h4",
                           tags$img(
                             src = "oryza.png",
                             alt = "monocots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column10,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "dicots_first_check_2",
                         p("Dicots", class= "h4",
                           tags$img(
                             src = "arabidopsis.png",
                             alt = "dicots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column11,
                         inline = F,
                         selected = NULL),
                       
                ),
                
                column(3,
                       checkboxGroupInput(
                         "dicots_second_check_2",
                         p("", class= "h4",
                           
                         ),
                         
                         choices = column12,
                         inline = F,
                         selected = NULL)
                ),
                
                fluidRow(br()),
                fluidRow(br()),
                
                span(tags$b("Tree building method"), style = "color:#25d04a; font-size: 20px; "),
                div(br()),
                
                div("The first step in any study using PharaohFUN is the construction of the 
                    gene tree corresponding to the selected orthogroup. For this, four standard methods 
                    are offered, select one before continuing. Default is performed with FastTree using Orthofinder
                    pipeline, which renders trees by aproximate maximum-likelihood and then reconciliate
                    gene trees with the species tree shown in the Home tab. For the other three methods,
                    no reconciliation is performed but support for each branch is shown as bootstrap 
                    values."),
                
                shinyWidgets::awesomeRadio(
                  inputId = "build_trees_2",
                  label = "", 
                  choices = c("Maximum Likelihood", "Neighbour Joining", "UPGMA", "Bayesian Inference"),
                  selected = "Maximum Likelihood",
                  inline = TRUE, 
                  status = "success"
                ),
                
                fluidRow(br()),
                
                span(tags$b("Insert a protein chain"), style = "color:#25d04a; font-size: 20px; "),
                br(),
                br(),
                div("Below you can paste the sequence of the protein whose evolutionary history you
                             wish to analyze. Then use the selection bar to choose the organism which the pasted sequence
                             belongs to."),
                fluidRow(
                  
                  column(8, 
                         textAreaInput(inputId = "geneInt2", 
                                       "", 
                                       width="200%", height = "200px", 
                                       value= "", resize = "vertical",
                                       placeholder = 
                                         "METNSSGEDLVIKTRKPYTITKQRERWTEEEHNRFIEALRLYGRAWQKIEEHVATKTAVQ...")),
                  column(4,
                         
                         fluidRow(
                           
                           column(9,
                                  div(style = "margin-top: 25px;",
                                      shinyWidgets::pickerInput("organism_input_2","Choose organism",
                                                                choices=names(organisms_values),
                                                                multiple = F, selected=names(organisms_values)[11])
                                  )
                                  
                                  
                           )),
                         
                         fluidRow(
                           column(3, div( style = "margin-top: 85px;", 
                                          shinyWidgets::actionBttn("run_button2", "Run", size = "sm", icon = icon("magnifying-glass"),
                                                                   style = "float", color = "success"))),
                           
                           column(9)
                           ),
                         
                  )
                  
                )),
              
              br(),
              fluidRow(
                box(status = "success", width = 12, 
                    title = span(tags$b("Results"), style = "color:#25d04a; font-size: 20px; "),
                    "Results for the query gene and species are displayed below, arranged in 
                      different tabs. The execution of each analysis is initiated from the start 
                      button within each of the tabs, with specific instructions for each analysis. First, go to Gene Tree tab,
                      which shows best matches for query sequence and a tree containing the genes corresponding to the same 
                      orthogroup as the query in the
                      selected species. The rest of the tabs allow for a deeper exploration of these genes. Do not start any 
                      analysis before creating gene tree.",
                    div(br()),
                    tabsetPanel(type = "tabs",
                                tabPanel("Gene Tree", 
                                         fluidRow(br()),
                                         div("This tab shows four different results. First of all, a table with up to 5 best 
                                         matches for the query sequence in the chosen proteome, with decreasing confidence. First row
                                         corresponds to best match, which is used to perform subsequent analysis. If you are interested
                                         in another one, please copy de ID and paste it in Gene ID-based search tab. Secondly, a complete list of the genes
                                         that are assigned to the same orthogroup as the query. Next, the proportion of  genes of each species. 
                                         Finally, a gene tree show the evolutionary relationships between those genes. All results can be downloaded 
                                         using the buttons at the bottom of the page."),
                                         fluidRow(br()),
                                         
                                         shinyjs::useShinyjs(),
                                         shinyjs::hidden(div(id='loading.tree2',h3('Please be patient, building tree ...'))),
                                         uiOutput(outputId = "error_tree2"),
                                         fluidRow(tags$div(id = "box_tree_seq_table2")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_tree_text2"), tags$div(id = "box_tree_pie2")),
                                         fluidRow(tags$br()),
                                         fluidRow(tags$div(id = "box_tree_plot2")),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_tree2"),
                                                     tags$div(id = "download_newick2"),
                                                     tags$div(id = "download_tree_seqs2"))
                                ),
                                tabPanel("Collapsable tree",
                                         fluidRow(tags$br()),
                                         div("This tab allows for an interactive visualization of the previous tree. 
                                               In particular, in situations where gene duplications have given rise 
                                               to several clades and you want to reduce the tree not in relation to 
                                               the species that appear, but to these clades, this visualization allows 
                                               the collapse of subtrees and the simple exploration of the areas of interest. 
                                               Press Show Collapsable Tree to show the tree."),
                                         fluidRow(tags$br()),
                                         shinyWidgets::actionBttn("phylo_start2", "Show Collapsable Tree",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_phylo2"),
                                         fluidRow(tags$br())
                                ),
                                tabPanel("Expansion/Contraction",
                                         fluidRow(tags$br()),
                                         div("Click the Show Evolutionary History button to calculate the species tree showing 
                                               the reconstruction of ancestral states of the orthogroup, i.e. the number of genes 
                                               calculated for each common ancestor of each species. For orthogroups that
                                               have undergone expansions or contractions along any branch, these are marked 
                                               in red and blue on the tree, respectively. In turn, gains and losses of complete orthogroups
                                               are shown in green and purple, while the non-existence of orthogroup
                                               genes in the different clades is represented in gray."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("cafe_start2", "Show Evolutionary History",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         
                                         shinyjs::hidden(div(id='loading.cafe2', h3('Please be patient, reconstructing expansion/contraction events ...'))),
                                         tags$div(id = "error_cafe2"),
                                         tags$div(id = "box_mrca2"),
                                         tags$br(),
                                         tags$div(id = "box_cafe2"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "cafe_down_button2"),
                                                     tags$div(id = "download_ui_for_cafe_plot2"))
                                ),
                                tabPanel("PFAM Domains", 
                                         fluidRow(tags$br()),
                                         div("Next, click on the Show Gene Selection for Pfam button, select the desired proteins
                                               from the tree and click on the Show Pfam Domains button to determine their PFAM domains.
                                               A table with the domains of each protein and their positions will be displayed, as well 
                                               as a plot showing the same information. Links to each domain's data are provided.
                                               Warning: this process may take a long time in the case of selecting many proteins."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("pfam_start2", "Show Gene Selection for Pfam",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_pfams2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "pfam_selectionI2"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.pfam.pf2',h3('Please be patient, identifying domains ...'))),
                                         uiOutput(outputId = "error_pfam2"),
                                         tags$div(id = "box_pfam2"),
                                         tags$br(),
                                         tags$div(id = "box_pfplot2"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "pfam_down_button2"),
                                                     tags$div(id = "download_ui_for_pfam_table2")
                                         )
                                ),
                                tabPanel("Multiple Sequence Alignment",
                                         fluidRow(tags$br()),
                                         div("Press the Show Gene Selection for MSA button to enable the selection of genes 
                                           from the tree to be aligned. Two alignment methods are supported: ClustalOmega uses this 
                                           algorithm to perform a de novo multiple sequence alignment of the selected proteins, while 
                                           the second option filters the precalculated MSA of the entire orthogroup (computed using the 
                                           software MAFFT) to retain 
                                           only the selected sequences, removing columns containing only gaps in the chosen subset. Thus,
                                           this second option retains the evolutionary framework to which the entire orthogroup is ascribed, 
                                           while the first aligns only the chosen sequences. The resulting alignment can be explored interactively,
                                           including searching for patterns USING Selection -> Find Motifs and different cholor schemes can be applied 
                                           based on different properties of amino acids. A consensus sequence is shown to summarize the alignment. For 
                                           a graphical representation with the different amino acids colored according to their chemical nature, click 
                                           on the Download Colored MSA button. The aligned sequences can also be download as a standard FASTA file.
                                           Warning: This process (specially the colored download) may take a long time in the case of selecting many proteins."),
                                         fluidRow(tags$br()),
                                         shinyWidgets::actionBttn("msa_start2", "Show Gene Selection for MSA",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_msa2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_method2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_selectionI2"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.msa2',h3('Please be patient, aligning sequences ...'))),
                                         uiOutput(outputId = "error_msa2"),
                                         tags$div(id = "box_msa2"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "msa_down_fasta2"),
                                                     tags$div(id = "msa_down_plot2"))
                                ),
                                tabPanel("GO Terms", 
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and select 
                                               the ontology type to display the GO terms associated with those genes. After
                                               pressing the Show GO terms button, the results are displayed in tabular form 
                                               and are accompanied by a GO association plot (each node is a GO and an edge 
                                               is drawn between two nodes if a gene from the set share both terms) and a treeplot
                                               showing the hierarchy of the identified terms and their summary. Links to each 
                                               individual GO term are avalaible from table and all results can be downloaded in
                                               standard PNG and TSV files."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("go_start2", "Show Gene Selection for GO Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos_mode2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "gos_selection2"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.go2',h3('Please be patient, preparing results ...'))),
                                         uiOutput(outputId = "error_gos2"),
                                         tags$div(id = "box_gos_table2"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_gos_plot2"), tags$div(id = "box_gos_treeplot2")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_gos_table2"),
                                                     tags$div(id = "gos_down_button2"),
                                                     tags$div(id = "tree_gos_down_button2"))
                                         
                                         
                                ),
                                tabPanel("KEGG Orthology",
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and press
                                               the Show KEGG information button to show the results. These include a table 
                                               showing the KEGG Orthology IDs, indicating how many and which of the selected 
                                               proteins correspond to each ID. In addition, the application performs an enrichment 
                                               in KEGG pathways from the identified KOs and allows for the plotting of these 
                                               pathways with the genes mapped onto them, using a selector to choose the pathway
                                               in case several enriched ones exist. Warning: this process may take a long time 
                                               in the case of selecting many proteins. Attempts to plot large pathways will
                                               produce an error, only more specific ones are supported i.e. Carbon metabolism 
                                               will return an error while Biotin metabolism will not."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("kegg_start2", "Show Gene Selection for KEGG Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_kos2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "kos_selection2"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.ko2',h3('Please be patient, exploring pathways ...'))),
                                         uiOutput(outputId = "error_kos2"),
                                         tags$div(id = "box_kos_table2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_kegg_table2"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "download_ui_for_kos_table2"),
                                                     tags$div(id = "download_ui_for_kegg_table2")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_paths2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "paths_button2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_path_image2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "path_download_ui2")
                                         
                                ),
                                tabPanel("STRING Interactions",
                                         fluidRow(tags$br()),
                                         div("This tab presents the results of interactions of the different proteins 
                                               for the species supported by the STRING database. After clicking the button,
                                               the proteins of the species included in the database are displayed. Once the
                                               desired selection has been made, press the Show String Interactions
                                               button and a table will be displayed
                                               indicating in each row an interaction and the type of evidence for its existence,
                                               either experimental (direct) or ortholog-based (interolog). In addition, a count
                                               of the orthogroups to which the target proteins belong is computed, presented in
                                               tabular form and by means of a pie chart, to assess preferential interaction of
                                               the orthogroup under study with other specific orthogroups. Finally, for
                                               specific proteins, an
                                               interaction network can be generated with the proteins that have the most reliable
                                               interactions with the selected protein. In case more than one protein is selected,
                                               all of them will appear in the network, which will contain their interactions to 
                                               determine pathways between them. The network is shown as an image and a button
                                               that creates the network is provided on the STRING page, where an interactive 
                                               network with additional protein structure information and SMART domains appears.
                                               Species currently supported by STRING: Aegilops, Arabidopsis, Bathycoccus, 
                                               Chara, Chlamydomonas, Coccomyxa,
                                               Klebsormidium, Micromonas, Oryza, Ostreococcus,
                                               Physcomitrium, Raphidocelis, Scenedesmus, Selaginella, Solanum, Sorghum,
                                               Triticum and Volvox. Warning: the selection of many proteins in the first
                                               step can lead to particularly high waiting times."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("string_start2", "Show Gene Selection for STRING annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_string2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "string_selection2"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.string2',h3('Please be patient, retrieving interactions ...'))),
                                         uiOutput(outputId = "error_string2"),
                                         tags$div(id = "box_st_table2"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_count_table2"), tags$div(id = "box_count_plot2")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_st_table2"),
                                                     tags$div(id = "download_ui_for_count_table2"),
                                                     tags$div(id = "count_down_button2")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_network2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "network_button2"),
                                         fluidRow(tags$br()),
                                         uiOutput(outputId = "error_network2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_output_network2")
                                         
                                ),
                                tabPanel("Literature Annotation", 
                                         fluidRow(tags$br()),
                                         div("In the text box, type the search term to query your bibliographic information,
                                               i.e., CCA1. Then select one of the 4 search modes: Normal returns the entities 
                                               containing the term, Exact returns those that are identical, Alias returns all 
                                               aliases associated with the term and Substring returns all those containing a 
                                               given string. Associations found in the literature are returned in tabular form,
                                               with links to the papers from which the information was extracted. This
                                               functionality is based on PlantConnectome, to extend this analysis, it is recommended
                                               to go to https://connectome.plant.tools/."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("lit_start2", "Show Search Box for Literature Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "success"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "query_lit2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_lit2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "lit_selection2"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.lit2',h3('Please be patient, browsing literature ...'))),
                                         uiOutput(outputId = "error_lit2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_lit_table2"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "download_ui_for_lit_table2")
                                         
                                )
                    ))
              )
      ),
      
      # Fourth tab content
      tabItem(tabName = "og_id_search",
              h2(""),
              fluidRow(valueBox("Orthogroup ID-based search", 
                                subtitle = "Single orthogroup, available organism",
                                icon = icon("dashboard"), width = 6, color = "red")),
              br(),
              box(
                title = span(tags$b("Organism selection"), style = "color:#d5251d; font-size: 20px; "), status = "danger", width = "500",
                "Please select the desired organisms from the following list for performing the analysis.
                  Please take care of selecting organisms that are present in the selected orthogroup.", br(), br(),
                column(3,
                       checkboxGroupInput(
                         "prasino_check_3",
                         p("Prasinodermophytes", class= "h4",
                           tags$img(
                             src = "palmophyllophyceae.png",
                             alt = "prasinodermophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column1,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "mami_check_3",
                         p("Mamiello/Pyramimonadophyceae", class= "h4",
                           tags$img(
                             src = "bathycoccus.png",
                             alt = "mamiellophyceae",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column2,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "chloro_check_3",
                         p("Other Chlorophytes", class= "h4",
                           tags$img(
                             src = "scenedesmus.png",
                             alt = "chlorophytes",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column3,
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "strepto_check_3",
                         p("Streptophyte algae", class= "h4",
                           tags$img(
                             src = "klebsormidium.png",
                             alt = "streptophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column4,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "bryo_check_3",
                         p("Bryophytes", class= "h4",
                           tags$img(
                             src = "marchantia.png",
                             alt = "bryophyta",
                             width = 32,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column5,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "lyco_check_3",
                         p("Lycophytes and Ferns", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "lycophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = c(column6, column7),
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "gymno_check_3",
                         p("Gymnosperms", class= "h4",
                           tags$img(
                             src = "gnetum.png",
                             alt = "gymnosperms",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column8,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "magno_check_3",
                         p("Other Angiosperms", class= "h4",
                           tags$img(
                             src = "amborella.png",
                             alt = "amborella",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column9,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "monocots_check_3",
                         p("Monocots", class= "h4",
                           tags$img(
                             src = "oryza.png",
                             alt = "monocots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column10,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "dicots_first_check_3",
                         p("Dicots", class= "h4",
                           tags$img(
                             src = "arabidopsis.png",
                             alt = "dicots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column11,
                         inline = F,
                         selected = NULL),
                       
                ),
                
                column(3,
                       checkboxGroupInput(
                         "dicots_second_check_3",
                         p("", class= "h4",
                           
                         ),
                         
                         choices = column12,
                         inline = F,
                         selected = NULL)
                ),
                
                fluidRow(br()),
                fluidRow(br()),
                
                span(tags$b("Tree building method"), style = "color:#d5251d; font-size: 20px; "),
                div(br()),
                
                div("The first step in any study using PharaohFUN is the construction of the 
                    gene tree corresponding to the selected orthogroup. For this, four standard methods 
                    are offered, select one before continuing. Default is performed with FastTree using Orthofinder 2
                    pipeline, which renders trees by aproximate maximum-likelihood and then reconciliate
                    gene trees with the species tree shown in the Home tab. For the other three methods,
                    no reconciliation is performed but support for each branch is shown as bootstrap 
                    values."),
                
                shinyWidgets::awesomeRadio(
                  inputId = "build_trees_3",
                  label = "", 
                  choices = c("Maximum Likelihood", "Neighbour Joining", "UPGMA", "Bayesian Inference"),
                  selected = "Maximum Likelihood",
                  inline = TRUE, 
                  status = "danger"
                ),
                
                fluidRow(br()),
                
                span(tags$b("Insert Orthogroup ID"), style = "color:#d5251d; font-size: 20px; "),
                div(br()),
                
                div("Below you can write the ID corresponding to the desired orthogroup. It is specially intended
                             to study STRING's results from previous queries."),
                
                
                fluidRow(
                  column(5, textInput("geneInt3", label = "", width = "100%", placeholder = "OG0001709")),
                  
                  column(1, div( style = "margin-top: 20px;", 
                                 shinyWidgets::actionBttn("run_button3", "Run", size = "sm", icon = icon("magnifying-glass"),
                                                          style = "float", color = "danger"))),
                  column(3)
                  
                  
                  
                )),
              
              br(),
              fluidRow(
                box(status = "danger", width = 12,
                    title = span(tags$b("Results"), style = "color:#d5251d; font-size: 20px; "),
                    "Results for the query orthogroup and species are displayed below, arranged in 
                      different tabs. The execution of each analysis is initiated from the start 
                      button within each of the tabs, with specific instructions for each analysis. First, go to Gene Tree tab,
                      which shows a tree containing the genes corresponding to the orthogroup in the selected species. 
                      The rest of the tabs allow for a deeper exploration of these genes. Do not start any analysis before 
                    creating gene tree.",
                    div(br()),
                    tabsetPanel(type = "tabs",
                                tabPanel("Gene Tree",
                                         fluidRow(br()),
                                         div("This tab shows three different results. First of all, a complete list of the genes
                                               that are assigned to the query orthogroup. Secondly, the proportion of
                                               genes of each species. Finally, a gene tree show the evolutionary relationships 
                                               between those genes. All results can be downloaded using the buttons at the bottom
                                               of the page."),
                                         fluidRow(br()),
                                         shinyjs::useShinyjs(),
                                         shinyjs::hidden(div(id='loading.tree3',h3('Please be patient, building tree ...'))),
                                         uiOutput(outputId = "error_tree3"),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_tree_text3"), tags$div(id = "box_tree_pie3")),
                                         fluidRow(tags$br()),
                                         fluidRow(tags$div(id = "box_tree_plot3")),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_tree3"),
                                                     tags$div(id = "download_newick3"),
                                                     tags$div(id = "download_tree_seqs3"))
                                ),
                                tabPanel("Collapsable tree",
                                         fluidRow(tags$br()),
                                         div("This tab allows an interactive visualization of the previous tree. 
                                               In particular, in situations where gene duplications have given rise 
                                               to several clades and you want to reduce the tree not in relation to 
                                               the species that appear, but to these clades, this visualization allows 
                                               the collapse of subtrees and the simple exploration of the areas of interest. 
                                               Press Show Collapsable Tree to show the tree."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("phylo_start3", "Show Collapsable Tree",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_phylo3"),
                                         fluidRow(tags$br())
                                ),
                                tabPanel("Expansion/Contraction",
                                         fluidRow(tags$br()),
                                         div("Click the Show Evolutionary History button to calculate the species tree showing 
                                               the reconstruction of ancestral states of the orthogroup, i.e. the number of genes 
                                               calculated for each common ancestor of each species. For orthogroups that
                                               have undergone expansions or contractions along any branch, these are marked 
                                               in red and blue on the tree, respectively. In turn, gains and losses of complete orthogroups
                                               are shown in green and purple, while the non-existence of orthogroup
                                               genes in the different clades is represented in gray."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("cafe_start3", "Show Evolutionary History",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         
                                         shinyjs::hidden(div(id='loading.cafe3', h3('Please be patient, reconstructing expansion/contraction events ...'))),
                                         tags$div(id = "error_cafe3"),
                                         tags$div(id = "box_mrca3"),
                                         tags$br(),
                                         tags$div(id = "box_cafe3"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "cafe_down_button3"),
                                                     tags$div(id = "download_ui_for_cafe_plot3"))
                                ),
                                tabPanel("PFAM Domains", 
                                         fluidRow(tags$br()),
                                         div("Next, click on the Show Gene Selection for Pfam button, select the desired proteins
                                               from the tree and click on the Show Pfam Domains button to determine their PFAM domains.
                                               A table with the domains of each protein and their positions will be displayed, as well 
                                               as a plot showing the same information. Links to each domain's data are provided.
                                               Warning: this process may take a long time in the case of selecting many proteins."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("pfam_start3", "Show Gene Selection for Pfam",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_pfams3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "pfam_selectionI3"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.pfam.pf3',h3('Please be patient, identifying domains ...'))),
                                         uiOutput(outputId = "error_pfam3"),
                                         tags$div(id = "box_pfam3"),
                                         tags$br(),
                                         tags$div(id = "box_pfplot3"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "pfam_down_button3"),
                                                     tags$div(id = "download_ui_for_pfam_table3")
                                         )
                                ),
                                tabPanel("Multiple Sequence Alignment",
                                         fluidRow(tags$br()),
                                         div("Press the Show Gene Selection for MSA button to enable the selection of genes 
                                           from the tree to be aligned. Two alignment methods are supported: ClustalOmega uses this 
                                           algorithm to perform a de novo multiple sequence alignment of the selected proteins, while 
                                           the second option filters the precalculated MSA of the entire orthogroup (computed using the 
                                           software MAFFT) to retain 
                                           only the selected sequences, removing columns containing only gaps in the chosen subset. Thus,
                                           this second option retains the evolutionary framework to which the entire orthogroup is ascribed, 
                                           while the first aligns only the chosen sequences. The resulting alignment can be explored interactively,
                                           including searching for patterns USING Selection -> Find Motifs and different cholor schemes can be applied 
                                           based on different properties of amino acids. A consensus sequence is shown to summarize the alignment. For 
                                           a graphical representation with the different amino acids colored according to their chemical nature, click 
                                           on the Download Colored MSA button. The aligned sequences can also be download as a standard FASTA file.
                                           Warning: This process (specially the colored download) may take a long time in the case of selecting many proteins."),
                                         fluidRow(tags$br()),
                                         shinyWidgets::actionBttn("msa_start3", "Show Gene Selection for MSA",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_msa3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_method3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_selectionI3"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.msa3',h3('Please be patient, aligning sequences ...'))),
                                         uiOutput(outputId = "error_msa3"),
                                         tags$div(id = "box_msa3"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "msa_down_fasta3"),
                                                     tags$div(id = "msa_down_plot3"))
                                ),
                                tabPanel("GO Terms", 
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and select 
                                               the ontology type to display the GO terms associated with those genes. After
                                               pressing the Show GO terms button, the results are displayed in tabular form 
                                               and are accompanied by a GO association plot (each node is a GO and an edge 
                                               is drawn between two nodes if a gene from the set share both terms) and a treeplot
                                               showing the hierarchy of the identified terms and their summary. Links to each 
                                               individual GO term are avalaible from table and all results can be downloaded in
                                               standard PNG and TSV files."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("go_start3", "Show Gene Selection for GO Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos_mode3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "gos_selection3"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.go3',h3('Please be patient, preparing results ...'))),
                                         uiOutput(outputId = "error_gos3"),
                                         tags$div(id = "box_gos_table3"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_gos_plot3"), tags$div(id = "box_gos_treeplot3")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_gos_table3"),
                                                     tags$div(id = "gos_down_button3"),
                                                     tags$div(id = "tree_gos_down_button3"))
                                         
                                         
                                ),
                                tabPanel("KEGG Orthology",
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and press
                                               the Show KEGG information button to show the results. These include a table 
                                               showing the KEGG Orthology IDs, indicating how many and which of the selected 
                                               proteins correspond to each ID. In addition, the application performs an enrichment 
                                               in KEGG pathways from the identified KOs and allows for the plotting of these 
                                               pathways with the genes mapped onto them, using a selector to choose the pathway
                                               in case several enriched ones exist. Warning: this process may take a long time 
                                               in the case of selecting many proteins. Attempts to plot large pathways will
                                               produce an error, only more specific ones are supported i.e. Carbon metabolism 
                                               will return an error while Biotin metabolism will not."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("kegg_start3", "Show Gene Selection for KEGG Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_kos3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "kos_selection3"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.ko3',h3('Please be patient, exploring pathways ...'))),
                                         uiOutput(outputId = "error_kos3"),
                                         tags$div(id = "box_kos_table3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_kegg_table3"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "download_ui_for_kos_table3"),
                                                     tags$div(id = "download_ui_for_kegg_table3")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_paths3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "paths_button3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_path_image3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "path_download_ui3")
                                         
                                ),
                                tabPanel("STRING Interactions",
                                         fluidRow(tags$br()),
                                         div("This tab presents the results of interactions of the different proteins 
                                               for the species supported by the STRING database. After clicking the button,
                                               the proteins of the species included in the database are displayed. Once the
                                               desired selection has been made, press the Show String Interactions
                                               button and a table will be displayed
                                               indicating in each row an interaction and the type of evidence for its existence,
                                               either experimental (direct) or ortholog-based (interolog). In addition, a count
                                               of the orthogroups to which the target proteins belong is computed, presented in
                                               tabular form and by means of a pie chart, to assess preferential interaction of
                                               the orthogroup under study with other specific orthogroups. Finally, for
                                               specific proteins, an
                                               interaction network can be generated with the proteins that have the most reliable
                                               interactions with the selected protein. In case more than one protein is selected,
                                               all of them will appear in the network, which will contain their interactions to 
                                               determine pathways between them. The network is shown as an image and a button
                                               that creates the network is provided on the STRING page, where an interactive 
                                               network with additional protein structure information and SMART domains appears.
                                               Species currently supported by STRING: Aegilops, Arabidopsis, Bathycoccus, 
                                               Chara, Chlamydomonas, Coccomyxa, 
                                               Klebsormidium, Micromonas, Oryza, Ostreococcus, 
                                               Physcomitrium, Raphidocelis, Scenedesmus, Selaginella, Solanum, Sorghum,
                                               Triticum and Volvox. Warning: the selection of many proteins in the first
                                               step can lead to particularly high waiting times."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("string_start3", "Show Gene Selection for STRING annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_string3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "string_selection3"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.string3',h3('Please be patient, retrieving interactions ...'))),
                                         uiOutput(outputId = "error_string3"),
                                         tags$div(id = "box_st_table3"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_count_table3"), tags$div(id = "box_count_plot3")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_st_table3"),
                                                     tags$div(id = "download_ui_for_count_table3"),
                                                     tags$div(id = "count_down_button3")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_network3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "network_button3"),
                                         fluidRow(tags$br()),
                                         uiOutput(outputId = "error_network3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_output_network3")
                                         
                                ),
                                tabPanel("Literature Annotation", 
                                         fluidRow(tags$br()),
                                         div("In the text box, type the search term to query your bibliographic information,
                                               i.e., CCA1. Then select one of the 4 search modes: Normal returns the entities 
                                               containing the term, Exact returns those that are identical, Alias returns all 
                                               aliases associated with the term and Substring returns all those containing a 
                                               given string. Associations found in the literature are returned in tabular form,
                                               with links to the papers from which the information was extracted. This
                                               functionality is based on PlantConnectome, to extend this analysis, it is recommended
                                               to go to https://connectome.plant.tools/."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("lit_start3", "Show Search Box for Literature Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "danger"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "query_lit3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_lit3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "lit_selection3"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.lit3',h3('Please be patient, browsing literature ...'))),
                                         uiOutput(outputId = "error_lit3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_lit_table3"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "download_ui_for_lit_table3")
                                         
                                )
                    ))
              )
      ),
      
      # Fifth tab content
      tabItem(tabName = "batch_search",
              h2(""),
              fluidRow(valueBox("Batch mode search", 
                                subtitle = "Set of genes, available organism",
                                icon = icon("layer-group"), width = 6, color = "orange")),
              br(),
              box(
                title = span(tags$b("Organism selection"), style = "color:#e37326; font-size: 20px; "), status = "warning", width = "500",
                "Please select the desired organisms from the following list for performing the analysis.
                  Please take care of selecting the organism whose sequences or IDs are being used as input. The 
                selected organisms will be the only ones that appear in heatmap and gene tables.", br(), br(),
                       column(3,
                              checkboxGroupInput(
                                "prasino_check_4",
                                p("Prasinodermophytes", class= "h4",
                                  tags$img(
                                    src = "palmophyllophyceae.png",
                                    alt = "prasinodermophytes",
                                    width = 25,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                choices = column1,
                                inline = F,
                                selected = NULL),
                              
                              checkboxGroupInput(
                                "mami_check_4",
                                p("Mamiello/Pyramimonadophyceae", class= "h4",
                                  tags$img(
                                    src = "bathycoccus.png",
                                    alt = "mamiellophyceae",
                                    width = 25,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                choices = column2,
                                inline = F,
                                selected = NULL),
                              
                              checkboxGroupInput(
                                "chloro_check_4",
                                p("Other Chlorophytes", class= "h4",
                                  tags$img(
                                    src = "scenedesmus.png",
                                    alt = "chlorophytes",
                                    width = 50,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                choices = column3,
                                inline = F,
                                selected = NULL)
                              
                       ),
                       column(3,
                              checkboxGroupInput(
                                "strepto_check_4",
                                p("Streptophyte algae", class= "h4",
                                  tags$img(
                                    src = "klebsormidium.png",
                                    alt = "streptophytes",
                                    width = 25,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                choices = column4,
                                inline = F,
                                selected = NULL),
                              
                              checkboxGroupInput(
                                "bryo_check_4",
                                p("Bryophytes", class= "h4",
                                  tags$img(
                                    src = "marchantia.png",
                                    alt = "bryophyta",
                                    width = 32,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                choices = column5,
                                inline = F,
                                selected = NULL),
                              
                              checkboxGroupInput(
                                "lyco_check_4",
                                p("Lycophytes and Ferns", class= "h4",
                                  tags$img(
                                    src = "selaginella.png",
                                    alt = "lycophytes",
                                    width = 25,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                choices = c(column6, column7),
                                inline = F,
                                selected = NULL)
                              
                       ),
                       column(3,
                              checkboxGroupInput(
                                "gymno_check_4",
                                p("Gymnosperms", class= "h4",
                                  tags$img(
                                    src = "gnetum.png",
                                    alt = "gymnosperms",
                                    width = 50,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                
                                choices = column8,
                                inline = F,
                                selected = NULL),
                              
                              checkboxGroupInput(
                                "magno_check_4",
                                p("Other Angiosperms", class= "h4",
                                  tags$img(
                                    src = "amborella.png",
                                    alt = "amborella",
                                    width = 50,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                
                                choices = column9,
                                inline = F,
                                selected = NULL),
                              
                              checkboxGroupInput(
                                "monocots_check_4",
                                p("Monocots", class= "h4",
                                  tags$img(
                                    src = "oryza.png",
                                    alt = "monocots",
                                    width = 50,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                
                                choices = column10,
                                inline = F,
                                selected = NULL),
                              
                              checkboxGroupInput(
                                "dicots_first_check_4",
                                p("Dicots", class= "h4",
                                  tags$img(
                                    src = "arabidopsis.png",
                                    alt = "dicots",
                                    width = 50,
                                    height = 25, style = "margin-left: 10px;"
                                  )
                                ),
                                
                                choices = column11,
                                inline = F,
                                selected = NULL),
                              
                       ),
                       
                       column(3,
                              checkboxGroupInput(
                                "dicots_second_check_4",
                                p("", class= "h4",
                                  
                                ),
                                
                                choices = column12,
                                inline = F,
                                selected = NULL)
                       ),
                
                fluidRow(br()),
                fluidRow(br()),
                
                span(tags$b("Choose Sequence File to Upload"), style = "color:#e37326; font-size: 20px; "),
                div(br()),
                
                div("Below you can upload a file with a list of gene IDs (with a single ID per line) or a FASTA file
                    containing custom identifiers for each sequence. These IDs or sequences must correspond to one
                    of the available organisms, and the correct species must be selected prior to analysis on the 
                    selection bar to the right of the Run button. Finally, choose if the upload file
                    contains IDs or Sequences and click Run to begin computations."),
                fluidRow(br()),
                
                
                fluidRow(
                  column(4, fileInput(inputId = "geneInt4",label = "")),
                  column(1, div( style = "margin-top: 20px;", 
                                 shinyWidgets::actionBttn("run_button4", "Run", size = "sm", icon = icon("magnifying-glass"),
                                                          style = "float", color = "warning"))),
                  column(3, div(style = "margin-top: 0px;",
                                
                                shinyWidgets::pickerInput("organism_input_4","",
                                                          choices=names(organisms_values),
                                                          multiple = F, selected=names(organisms_values)[11]))
                         
                         
                  ),
                  column(2,div(style = "margin-top: 24px;",
                               shinyWidgets::materialSwitch(inputId = "search_mode4", label = "IDs", 
                                                            value = T, status = "warning", inline = TRUE),
                               span("Sequences"))),
                  
                  column(2)
                  
                  
                  
                )),
              
              br(),
              
              fluidRow(
                box(status = "warning", width = 12, 
                    title = span(tags$b("Results"), style = "color:#e37326; font-size: 20px; "),
                    "Results for the query sequences or IDs and species are displayed below. Using the download button, user can
                    access a compressed file with results arranged in different subfolders. First, a table in TSV format indicates
                    the mapping of each sequence against the proteome and its associated metrics (in case sequence sarch mode is 
                    selected). Each of the subfolders corresponds to a query sequence/ID and contains the gene tree of its 
                    orthogroup (without filtering by species), the MSA of the complete orthogroup, the expansion/contraction 
                    results, the KO and GO annotations for the genes from the selected species and a table with the genes
                    forming the orthogroup in the selected species. Additionally, a visual summary of the number of genes 
                    within each orthogroup relative to each of the sequences/IDs is presented in the form of a heatmap, 
                    which allows to determine if in any of the selected species some orthogroups have undergone expansions 
                    or if there are orthogroups that are not present in all species.",
                    div(br()),
                    fluidRow(br()),
                    shinyjs::useShinyjs(),
                    shinyjs::hidden(div(id='loading.batch4',h3('Please be patient, preparing your folders ...'))),
                    uiOutput(outputId = "error_batch4"),
                    tags$div(id = "download_batch4"),
                    fluidRow(tags$br()),
                    tags$div(id = "box_heatmap4")
                    
                ))
              
      ),
      
      # Sixth tab content
      tabItem(tabName = "shoot_search",
              h2(""),
              fluidRow(valueBox("New organism sequence search", 
                                subtitle = "Single gene, custom organism",
                                icon = icon("leaf", lib = "glyphicon"), width = 6, color = "purple")),
              br(),
              box(
                title = span(tags$b("Organism selection"), style = "color:#5e3587; font-size: 20px; "), status = "primary", width = "500",
                "Please select the desired organisms from the following list for performing the analysis.
                  It is recommended to select organisms that span the evolutionary placement of the custom
                organism to allow for better resolution of the created tree.", br(), br(),
                column(3,
                       checkboxGroupInput(
                         "prasino_check_5",
                         p("Prasinodermophytes", class= "h4",
                           tags$img(
                             src = "palmophyllophyceae.png",
                             alt = "prasinodermophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column1,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "mami_check_5",
                         p("Mamiello/Pyramimonadophyceae", class= "h4",
                           tags$img(
                             src = "bathycoccus.png",
                             alt = "mamiellophyceae",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column2,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "chloro_check_5",
                         p("Other Chlorophytes", class= "h4",
                           tags$img(
                             src = "scenedesmus.png",
                             alt = "chlorophytes",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column3,
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "strepto_check_5",
                         p("Streptophyte algae", class= "h4",
                           tags$img(
                             src = "klebsormidium.png",
                             alt = "streptophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column4,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "bryo_check_5",
                         p("Bryophytes", class= "h4",
                           tags$img(
                             src = "marchantia.png",
                             alt = "bryophyta",
                             width = 32,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column5,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "lyco_check_5",
                         p("Lycophytes and Ferns", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "lycophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = c(column6, column7),
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "gymno_check_5",
                         p("Gymnosperms", class= "h4",
                           tags$img(
                             src = "gnetum.png",
                             alt = "gymnosperms",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column8,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "magno_check_5",
                         p("Other Angiosperms", class= "h4",
                           tags$img(
                             src = "amborella.png",
                             alt = "amborella",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column9,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "monocots_check_5",
                         p("Monocots", class= "h4",
                           tags$img(
                             src = "oryza.png",
                             alt = "monocots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column10,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "dicots_first_check_5",
                         p("Dicots", class= "h4",
                           tags$img(
                             src = "arabidopsis.png",
                             alt = "dicots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column11,
                         inline = F,
                         selected = NULL),
                       
                ),
                
                column(3,
                       checkboxGroupInput(
                         "dicots_second_check_5",
                         p("", class= "h4",
                           
                         ),
                         
                         choices = column12,
                         inline = F,
                         selected = NULL)
                ),
                
                fluidRow(br()),
                fluidRow(br()),
                
                span(tags$b("Insert a protein chain"), style = "color:#5e3587; font-size: 20px; "),
                br(),
                br(),
                div("Below you can paste the sequence of the protein whose evolutionary history you
                             wish to analyze."),
                
                
                fluidRow(
                  column(8, textAreaInput(inputId = "geneInt5", 
                                          "", 
                                          width="200%", height = "200px", 
                                          value= "", resize = "vertical",
                                          placeholder = 
                                            "METNSSGEDLVIKTRKPYTITKQRERWTEEEHNRFIEALRLYGRAWQKIEEHVATKTAVQ...")),
                  #column(1, div( style = "margin-top: 20px;", actionButton("run", "Run", icon = icon("magnifying-glass")))))
                  column(1, div( style = "margin-top: 185px;", 
                                 shinyWidgets::actionBttn("run_button5", "Run", size = "sm", icon = icon("magnifying-glass"),
                                                          style = "float", color = "royal"))),
                  column(3)
                  
                  
                  
                )),
              
              br(),
              fluidRow(
                box(status = "primary", width = 12,
                    title = span(tags$b("Results"), style = "color:#5e3587; font-size: 20px; "),
                    "Results for the query sequence and species are displayed below, arranged in 
                      different tabs. The execution of each analysis is initiated from the start 
                      button within each of the tabs, with specific instructions for each analysis. First, go to Gene Tree tab,
                      which shows a tree containing the genes corresponding to the same orthogroup as the query in the
                      selected species. The rest of the tabs allow for a deeper exploration of these genes. Do not start any 
                      analysis before creating gene tree.",
                    div(br()),
                    tabsetPanel(type = "tabs",
                                tabPanel("Gene Tree",
                                         fluidRow(br()),
                                         div("This tab shows three different results. First of all, a complete list of the genes
                                               that are assigned to the same orthogroup as the query. Secondly, the proportion of
                                               genes of each species. Finally, a gene tree show the evolutionary relationships 
                                               between those genes. Query sequence placement is highlighted in red. All results 
                                               can be downloaded using the buttons at the bottom
                                               of the page."),
                                         fluidRow(br()),
                                         shinyjs::useShinyjs(),
                                         shinyjs::hidden(div(id='loading.tree5',h3('Please be patient, building tree ...'))),
                                         uiOutput(outputId = "error_tree5"),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_tree_text5"), tags$div(id = "box_tree_pie5")),
                                         fluidRow(tags$br()),
                                         fluidRow(tags$div(id = "box_tree_plot5")),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_tree5"),
                                                     tags$div(id = "download_newick5"),
                                                     tags$div(id = "download_tree_seqs5"))
                                ),
                                tabPanel("Collapsable tree",
                                         fluidRow(tags$br()),
                                         div("This tab allows an interactive visualization of the previous tree. 
                                               In particular, in situations where gene duplications have given rise 
                                               to several clades and you want to reduce the tree not in relation to 
                                               the species that appear, but to these clades, this visualization allows 
                                               the collapse of subtrees and the simple exploration of the areas of interest. 
                                               Press Show Collapsable Tree to show the tree."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("phylo_start5", "Show Collapsable Tree",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_phylo5"),
                                         fluidRow(tags$br())
                                ),
                                tabPanel("Expansion/Contraction",
                                         fluidRow(tags$br()),
                                         div("Click the Show Evolutionary History button to calculate the species tree showing 
                                               the reconstruction of ancestral states of the orthogroup, i.e. the number of genes 
                                               calculated for each common ancestor of each species. For orthogroups that
                                               have undergone expansions or contractions along any branch, these are marked 
                                               in red and blue on the tree, respectively. In turn, gains and losses of complete orthogroups
                                               are shown in green and purple, while the non-existence of orthogroup
                                               genes in the different clades is represented in gray."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("cafe_start5", "Show Evolutionary History",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         
                                         shinyjs::hidden(div(id='loading.cafe5', h3('Please be patient, reconstructing expansion/contraction events ...'))),
                                         tags$div(id = "error_cafe5"),
                                         tags$div(id = "box_mrca5"),
                                         tags$br(),
                                         tags$div(id = "box_cafe5"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "cafe_down_button5"),
                                                     tags$div(id = "download_ui_for_cafe_plot5"))
                                ),
                                tabPanel("PFAM Domains", 
                                         fluidRow(tags$br()),
                                         div("Next, click on the Show Gene Selection for Pfam button, select the desired proteins
                                               from the tree and click on the Show Pfam Domains button to determine their PFAM domains.
                                               A table with the domains of each protein and their positions will be displayed, as well 
                                               as a plot showing the same information. Links to each domain's data are provided.
                                               Warning: this process may take a long time in the case of selecting many proteins."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("pfam_start5", "Show Gene Selection for Pfam",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_pfams5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "pfam_selectionI5"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.pfam.pf5',h3('Please be patient, identifying domains ...'))),
                                         uiOutput(outputId = "error_pfam5"),
                                         tags$div(id = "box_pfam5"),
                                         tags$br(),
                                         tags$div(id = "box_pfplot5"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "pfam_down_button5"),
                                                     tags$div(id = "download_ui_for_pfam_table5")
                                         )
                                ),
                                tabPanel("Multiple Sequence Alignment",
                                         fluidRow(tags$br()),
                                         div("Press the Show Gene Selection for MSA button to enable the selection of genes 
                                           from the tree to be aligned. Two alignment methods are supported: ClustalOmega uses this 
                                           algorithm to perform a de novo multiple sequence alignment of the selected proteins, while 
                                           the second option filters the precalculated MSA of the entire orthogroup (computed using the 
                                           software MAFFT) to retain 
                                           only the selected sequences, removing columns containing only gaps in the chosen subset. Thus,
                                           this second option retains the evolutionary framework to which the entire orthogroup is ascribed, 
                                           while the first aligns only the chosen sequences. The resulting alignment can be explored interactively,
                                           including searching for patterns USING Selection -> Find Motifs and different cholor schemes can be applied 
                                           based on different properties of amino acids. A consensus sequence is shown to summarize the alignment. For 
                                           a graphical representation with the different amino acids colored according to their chemical nature, click 
                                           on the Download Colored MSA button. The aligned sequences can also be download as a standard FASTA file.
                                           Warning: This process (specially the colored download) may take a long time in the case of selecting many proteins."),
                                         fluidRow(tags$br()),
                                         shinyWidgets::actionBttn("msa_start5", "Show Gene Selection for MSA",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_msa5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_method5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "msa_selectionI5"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.msa5',h3('Please be patient, aligning sequences ...'))),
                                         uiOutput(outputId = "error_msa5"),
                                         tags$div(id = "box_msa5"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "msa_down_fasta5"),
                                                     tags$div(id = "msa_down_plot5"))
                                ),
                                tabPanel("GO Terms", 
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and select 
                                               the ontology type to display the GO terms associated with those genes. After
                                               pressing the Show GO terms button, the results are displayed in tabular form 
                                               and are accompanied by a GO association plot (each node is a GO and an edge 
                                               is drawn between two nodes if a gene from the set share both terms) and a treeplot
                                               showing the hierarchy of the identified terms and their summary. Links to each 
                                               individual GO term are avalaible from table and all results can be downloaded in
                                               standard PNG and TSV files."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("go_start5", "Show Gene Selection for GO Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_gos_mode5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "gos_selection5"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.go5',h3('Please be patient, preparing results ...'))),
                                         uiOutput(outputId = "error_gos5"),
                                         tags$div(id = "box_gos_table5"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_gos_plot5"), tags$div(id = "box_gos_treeplot5")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_gos_table5"),
                                                     tags$div(id = "gos_down_button5"),
                                                     tags$div(id = "tree_gos_down_button5"))
                                         
                                         
                                ),
                                tabPanel("KEGG Orthology",
                                         fluidRow(tags$br()),
                                         div("Click on the button to select the genes of interest from the tree and press
                                               the Show KEGG information button to show the results. These include a table 
                                               showing the KEGG Orthology IDs, indicating how many and which of the selected 
                                               proteins correspond to each ID. In addition, the application performs an enrichment 
                                               in KEGG pathways from the identified KOs and allows for the plotting of these 
                                               pathways with the genes mapped onto them, using a selector to choose the pathway
                                               in case several enriched ones exist. Warning: this process may take a long time 
                                               in the case of selecting many proteins. Attempts to plot large pathways will
                                               produce an error, only more specific ones are supported i.e. Carbon metabolism 
                                               will return an error while Biotin metabolism will not."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("kegg_start5", "Show Gene Selection for KEGG Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_kos5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "kos_selection5"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.ko5',h3('Please be patient, exploring pathways ...'))),
                                         uiOutput(outputId = "error_kos5"),
                                         tags$div(id = "box_kos_table5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_kegg_table5"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"), 
                                                     tags$div(id = "download_ui_for_kos_table5"),
                                                     tags$div(id = "download_ui_for_kegg_table5")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_paths5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "paths_button5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_path_image5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "path_download_ui5")
                                         
                                ),
                                tabPanel("STRING Interactions",
                                         fluidRow(tags$br()),
                                         div("This tab presents the results of interactions of the different proteins 
                                               for the species supported by the STRING database. After clicking the button,
                                               the proteins of the species included in the database are displayed. Once the
                                               desired selection has been made, press the Show String Interactions
                                               button and a table will be displayed
                                               indicating in each row an interaction and the type of evidence for its existence,
                                               either experimental (direct) or ortholog-based (interolog). In addition, a count
                                               of the orthogroups to which the target proteins belong is computed, presented in
                                               tabular form and by means of a pie chart, to assess preferential interaction of
                                               the orthogroup under study with other specific orthogroups. Finally, for
                                               specific proteins, an
                                               interaction network can be generated with the proteins that have the most reliable
                                               interactions with the selected protein. In case more than one protein is selected,
                                               all of them will appear in the network, which will contain their interactions to 
                                               determine pathways between them. The network is shown as an image and a button
                                               that creates the network is provided on the STRING page, where an interactive 
                                               network with additional protein structure information and SMART domains appears.
                                               Species currently supported by STRING: Aegilops, Arabidopsis, Bathycoccus, 
                                               Chara, Chlamydomonas, Coccomyxa,
                                               Klebsormidium, Micromonas, Oryza, Ostreococcus,
                                               Physcomitrium, Raphidocelis, Scenedesmus, Selaginella, Solanum, Sorghum,
                                               Triticum and Volvox. Warning: the selection of many proteins in the first
                                               step can lead to particularly high waiting times."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("string_start5", "Show Gene Selection for STRING annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_string5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "string_selection5"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.string5',h3('Please be patient, retrieving interactions ...'))),
                                         uiOutput(outputId = "error_string5"),
                                         tags$div(id = "box_st_table5"),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("50%", "50%"),
                                                     tags$div(id = "box_count_table5"), tags$div(id = "box_count_plot5")),
                                         fluidRow(tags$br()),
                                         splitLayout(cellWidths = c("33%", "33%", "33%"), 
                                                     tags$div(id = "download_ui_for_st_table5"),
                                                     tags$div(id = "download_ui_for_count_table5"),
                                                     tags$div(id = "count_down_button5")),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_network5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "network_button5"),
                                         fluidRow(tags$br()),
                                         uiOutput(outputId = "error_network5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_output_network5")
                                         
                                ),
                                tabPanel("Literature Annotation", 
                                         fluidRow(tags$br()),
                                         div("In the text box, type the search term to query your bibliographic information,
                                               i.e., CCA1. Then select one of the 4 search modes: Normal returns the entities 
                                               containing the term, Exact returns those that are identical, Alias returns all 
                                               aliases associated with the term and Substring returns all those containing a 
                                               given string. Associations found in the literature are returned in tabular form,
                                               with links to the papers from which the information was extracted. This
                                               functionality is based on PlantConnectome, to extend this analysis, it is recommended
                                               to go to https://connectome.plant.tools/."),
                                         fluidRow(br()),
                                         shinyWidgets::actionBttn("lit_start5", "Show Search Box for Literature Annotation",
                                                                  size = "sm", icon = icon("magnifying-glass"),
                                                                  style = "float", color = "royal"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "query_lit5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "selected_lit5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "lit_selection5"),
                                         fluidRow(tags$br()),
                                         shinyjs::hidden(div(id='loading.lit5',h3('Please be patient, browsing literature ...'))),
                                         uiOutput(outputId = "error_lit5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "box_lit_table5"),
                                         fluidRow(tags$br()),
                                         tags$div(id = "download_ui_for_lit_table5")
                                         
                                )
                    ))
              )
      ),
      
      # Complete dataset tab
      tabItem(tabName = "whole_data", 
              
              h2(""),
              fluidRow(valueBox("Complete Datasets", 
                                subtitle = "Complete data, available organisms",
                                icon = icon("th-list", lib = "glyphicon"), width = 6, color = "maroon")),
              br(),
              box(
                title = span(tags$b("Organism selection"), style = "color:#c21860; font-size: 20px; "), status = "danger", width = "500",
                "Please select the desired organisms from the following list for filtering the complete dataset. The 
                selected organisms will be the only ones that appear in the final table.", br(), br(),
                column(3,
                       checkboxGroupInput(
                         "prasino_check_5",
                         p("Prasinodermophytes", class= "h4",
                           tags$img(
                             src = "palmophyllophyceae.png",
                             alt = "preasinodermophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column1,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "mami_check_5",
                         p("Mamiellophyceae/Pyramimonadophyceae", class= "h4",
                           tags$img(
                             src = "bathycoccus.png",
                             alt = "mamiellophyceae",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column2,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "chloro_check_5",
                         p("Other Chlorophytes", class= "h4",
                           tags$img(
                             src = "scenedesmus.png",
                             alt = "chlorophytes",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column3,
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "strepto_check_5",
                         p("Streptophyte algae", class= "h4",
                           tags$img(
                             src = "klebsormidium.png",
                             alt = "streptophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column4,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "bryo_check_5",
                         p("Bryophytes", class= "h4",
                           tags$img(
                             src = "marchantia.png",
                             alt = "bryophyta",
                             width = 32,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = column5,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "lyco_check_5",
                         p("Lycophytes and Ferns", class= "h4",
                           tags$img(
                             src = "klebsormidium.png",
                             alt = "lycophytes",
                             width = 25,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         choices = c(column6, column7),
                         inline = F,
                         selected = NULL)
                       
                ),
                column(3,
                       checkboxGroupInput(
                         "gymno_check_5",
                         p("Gymnosperms", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "gymnosperms",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column8,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "magno_check_5",
                         p("Other Angiosperms", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "gymnosperms",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column9,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "monocots_check_5",
                         p("Monocots", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "monocots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column10,
                         inline = F,
                         selected = NULL),
                       
                       checkboxGroupInput(
                         "dicots_first_check_5",
                         p("Dicots", class= "h4",
                           tags$img(
                             src = "selaginella.png",
                             alt = "monocots",
                             width = 50,
                             height = 25, style = "margin-left: 10px;"
                           )
                         ),
                         
                         choices = column11,
                         inline = F,
                         selected = NULL),
                       
                ),
                
                column(3,
                       checkboxGroupInput(
                         "dicots_second_check_5",
                         p("", class= "h4",
                           
                         ),
                         
                         choices = column12,
                         inline = F,
                         selected = NULL)
                ),
                
                fluidRow(br()),
                fluidRow(br()),
                
                span(tags$b("Choose Model Data to Download"), style = "color:#c21860; font-size: 20px; "),
                div(br()),
                
                div("Once all species are selected, click Run to create the table."),
                fluidRow(br()),
                
                fluidRow(column(5), 
                        
                         column(1, div( style = "margin-top: 9px;", 
                                        shinyWidgets::actionBttn("run_button6", "Run", size = "sm", icon = icon("magnifying-glass"),
                                                                 style = "float", color = "danger"))),
                         column(6))
                
               ),
              
              br(),
              
              fluidRow(
                box(status = "danger", width = 12,
                    title = span(tags$b("Results"), style = "color:#c21860; font-size: 20px; "),
                    "Use the button to download the filtered dataset.",
                    div(br()),
                   
                    div(id='down.text6', uiOutput(outputId = "text_dataset6")),
                    div(br()),
                    div(br()),
                    div(id='down.whole6',
                                        style = "margin-left: 600px;", 
                                        shinyWidgets::downloadBttn(outputId= "downloadDataset6", "Download Filtered Dataset",
                                                                                                  size = "sm", color = "danger"))
                                        
                                        ),
                div(br())
              )
              
      ),
      
      # Download genomes tab
      tabItem(tabName = "down_genomes", 
              h2(""),
              fluidRow(valueBox("Download genomes", 
                                subtitle = "Complete genome, available organism",
                                icon = icon("download"), width = 6, color = "olive")),
              br(),
              box(
                title = span(tags$b("Organism selection"), style = "color:#08a266; font-size: 20px; "), status = "success", width = "500",
                "Please select the desired organism to download the corresponding genome.", br(), br(),
                
                fluidRow(
                  column(3),
                  column(2, div(style = "margin-top: 4px;",
                              
                              shinyWidgets::pickerInput("organism_down_7","",
                                                        choices=names(organisms_values),
                                                        multiple = F, selected=names(organisms_values)[11]))),
                  column(1),
                  column(3, div(style = "margin-top: 22px;",
                         shinyWidgets::downloadBttn(outputId= "downloadGenomes7", "Download Genome",
                                                    size = "sm", color = "success"))),
                  column(3)
                         ),
                  
                )
              
              ),
      
      # Last tab
      tabItem(tabName = "contact_tutorial", 
              
              h2(""),
              fluidRow(valueBox("Contact and Info", 
                                subtitle = "Acknowledgments and other information",
                                icon = icon("envelope"), width = 6, color = "teal")),
              br(),
              
              tags$div(style = 'font-size: 18px; margin-left: 20px;',"Authors: Marcos Ramos González, Víctor
                       Ramos González, Emma Serrano Pérez, Christina Arvanitidou, Jorge Hernández García, Mercedes García González and Francisco José Romero Campero."),
              tags$br(),
              
              tags$div(style = 'font-size: 18px; margin-left: 20px;',"We are strongly committed to open access software and open science. PharaohFUN's source code is available
                       at GitHub following the lateral panel link and is released under a GNU General Public License v3.0. If you 
                       experience any problem using PharaohFUN, please create an issue in GitHub and we will address it. For other
                       inquiries, send an email to mramos5@us.es."),
              tags$br(),
              
              fluidRow(
                column(8, 
                       tags$div(style = 'font-size: 18px; margin-left: 20px;', 
                                "All organisms images where acquired from PhyloPic. Next, we present a list with the authors of each one:",
                                tags$ul(
                                  tags$li("Phaeodactylum by Jonathan Wells."),
                                  tags$li("Porphyra by Guillaume Dera."),
                                  tags$li("Cyanophora by Guillaume Dera."),
                                  tags$li("Ostreococcus by Guillaume Dera."),
                                  tags$li("Scenedesmus by Sergio A. Muñoz-Gómez."),
                                  tags$li("Klebsormidium by Matt Crook."),
                                  tags$li("Marchantia by Guillaume Dera."),
                                  tags$li("Selaginella by Mason McNair."),
                                  tags$li("Arabidopsis by Mason McNair."),
                                  tags$li("Chlamydomonas by Sergio A. Muñoz-Gómez."),
                                  tags$li("Haematococcus by Matthew Crook."),
                                  tags$li("Volvox by Matthew Crook."),
                                  tags$li("Zygnema by Matthew Crook"),
                                  tags$li("Marchantia by T. Michael Keesey."),
                                  tags$li("Araucaria by T. Michael Keesey."),
                                  tags$li("Gnetum by T. Michael Keesey."),
                                  
                                ),
                       )
                ),
                column(4, tags$div(align="center",width=60,
                                   HTML('<script type="text/javascript" src="//rf.revolvermaps.com/0/0/6.js?i=5yb18atzqxr&amp;m=7&amp;c=e63100&amp;cr1=ffffff&amp;f=arial&amp;l=0&amp;bv=90&amp;lx=-420&amp;ly=420&amp;hi=20&amp;he=7&amp;hc=a8ddff&amp;rs=80" async="async"></script>'))
                       
                )
              ),
              tags$div(style = 'font-size: 18px; margin-left: 20px;', 
                       "Some of these images are licensed under an ", tags$a(href="https://creativecommons.org/licenses/by-nc-sa/3.0/", 
                                                                             "Attribution-NonCommercial-ShareAlike 3.0 Unported"), "or ",
                       
                       tags$a(href="https://creativecommons.org/licenses/by-sa/3.0/", "Attribution-ShareAlike 3.0 Unported"), " license.")
      )
      # Close tabs, body and UI code   
      
    )
    
  )
)

server <- function(input, output) {
  
  ############## GENE ID-BASED SEARCH ##############
  
  # Set global variables for tracking changes in output
  UI_exist_pfam1 <<- F
  UI_exist_tree1 <<- F
  UI_exist_phylo1 <<- F
  UI_exist_cafe1 <<- F
  UI_exist_error_cafe1 <<- F
  UI_exist_msa1 <<- F
  UI_exist_go1 <<- F
  UI_exist_kegg1 <<- F
  UI_exist_kegg_path1 <<- F
  UI_exist_pathview1 <<- F
  UI_exist_lit1 <<-  F
  UI_exist_string1 <<-  F
  UI_exist_network1 <<-  F
    
  # Clear previous error outputs
    
  # observeEvent(input$run_button1, {
  #   
  #                output$error_pfam1 <- NULL
  #                output$error_tree1 <- NULL
  #                
  #              })
  
  # To avoid autoupdating some inputs, define variables with its values
  model.selected1 <- reactive({
    model.selected <- F
    return(model.selected)
  })%>% bindEvent(input$run_button1)
  
  build_trees1 <- reactive({
    build_trees <- as.character(input$build_trees_1)
    return(build_trees)
  }) %>% bindEvent(input$run_button1)
  
  # Load organisms selection based on the model selected
  selected_organisms1 <- reactive({
    selected_organisms <- c(input$prasino_check_1, input$mami_check_1,
                            input$chloro_check_1, input$strepto_check_1,
                            input$bryo_check_1, input$lyco_check_1,
                            input$gymno_check_1, input$magno_check_1,
                            input$monocots_check_1, input$dicots_first_check_1,
                            input$dicots_second_check_1)
    return(selected_organisms)
    
  }) %>% bindEvent(input$run_button1)
  
  selected_values_org1 <- reactive(organisms_values[selected_organisms1()]) %>% bindEvent(input$run_button1)
 
  
  og.name1 <- reactive({
    library(stringr)
    gene.name.tree <- str_replace_all(input$geneInt1, fixed(" "), "")
    shinyjs::showElement(id = 'loading.tree1')
    
    # Load table with orthogroups information depending on selected model
    ortho.table.search <- "Gene_Trees/Orthogroups.tsv"
    ortho.table <- read.csv(ortho.table.search,
                           header = T, sep = "\t", as.is = T,
                           fill = T, blank.lines.skip = F)

    
    # Find orthogroup of target gene
    found <- F
    file.name <- NULL
    i <- 1
    while (!(found) && (i <= nrow(ortho.table)))
    {
      object <- as.character(ortho.table[i,])
      gene.number <- grep(pattern = gene.name.tree, object)
      if (length(gene.number) != 0)
      {
        found <- T
        file.name <- ortho.table[i,1]
      }
      else
      {
        i <- i+1
      }
    }
    
    # Error message if gene ID does not belong to any OG
    if (is.null(file.name))
    {
      shinyjs::hideElement(id = 'loading.tree1')
      
      if (UI_exist_tree1)
      {
        removeUI(
          selector = "div:has(>> #treeTips1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>>> #presentorg1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #tree_image1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTree1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadNewick1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTreeSeqs1)",
          multiple = TRUE,
          immediate = TRUE
        )
      }
      
      UI_exist_tree1 <<- F
      output$error_tree1 <- renderUI({renderText({print("No results for this 
      query due to not supported gene name or lack of orthologs in the selected organisms.")})})
      validate(need(!is.null(file.name), " "))
    }
    
    return(file.name)
    
  }) %>% bindEvent(input$run_button1)
  
  tree1 <- reactive({
    file.name <- og.name1()
    # Load gene tree file depending on the input
    
    tree.name <- paste("Gene_Trees",paste(file.name, "tree.txt", sep = "_"), sep="/")
    
    # Error if tree file not found
    if (!(file.exists(tree.name)))
    {
      shinyjs::hideElement(id = 'loading.tree1')
      
      if (UI_exist_tree1)
      {
        removeUI(
          selector = "div:has(>> #treeTips1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>>> #presentorg1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #tree_image1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTree1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadNewick1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTreeSeqs1)",
          multiple = TRUE,
          immediate = TRUE
        )
      }
      
      UI_exist_tree1 <<- F
      output$error_tree1 <- renderUI({renderText({print("Unable to construct tree associated to
                                                        to an orthogroup with less than 4 genes.")})})
      validate(need(file.exists(tree.name), " "))
    }
    
    # Generate tree depending on the tree building method selected
    {
    if (build_trees1() == "Maximum Likelihood")
    {
      tree <- read.tree(tree.name)
      return(tree)
    }
    else if(build_trees1() == "Bayesian Inference")
    {
      tree.bayes <- paste("Tree_Bayes",paste0("bayes_tree_", file.name, ".nwk"), sep="/")
      
      # Error if tree file not found
      if (!(file.exists(tree.bayes)))
      {
        shinyjs::hideElement(id = 'loading.tree1')
        
        if (UI_exist_tree1)
        {
          removeUI(
            selector = "div:has(>> #treeTips1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>>> #presentorg1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #tree_image1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadTree1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadNewick1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadTreeSeqs1)",
            multiple = TRUE,
            immediate = TRUE
          )
        }
        
        UI_exist_tree1 <<- F
        output$error_tree1 <- renderUI({renderText({print("Bayesian tree inference is currently under development. 
                                                          We are working to offer this method for all 
                                                          orthogroups through regular updates. If this message appears,
                                                          the orthogroup of interest is not yet available. If you want it
                                                          to appear in the next update, please send an email to mramos5@us.es 
                                                          and we will try to prioritize it. Meanwhile, try the other methods
                                                          to build the gene tree.")})})
        validate(need(file.exists(tree.bayes), " "))
      }
      
      tree <- read.tree(tree.bayes)
      return(tree)
      
    }
    
    else
    {
      library(phangorn)
      
      # Read MSA
      multiple_file <- paste("MultipleSequenceAlignments", paste0(file.name, ".fa"), sep="/")
      
      
      mytree <- read.phyDat(file = multiple_file,
                            format="fasta", type = "AA")
      
      # To keep the same workflow, the reduced tree will be calculated, then the applied subsetting
      # will only reduce them if the selected method has been fasttree
      
      
      # Selection of genes from the selected organism
      
      # Create empty vectors
      
      tips_to_keep.praco1 = tips_to_keep.ot1 = tips_to_keep.ostlu1 = tips_to_keep.ostme1 = tips_to_keep.bp1 <- c()
      tips_to_keep.mi1 = tips_to_keep.micco1 = tips_to_keep.cymte1 = tips_to_keep.cr1 = tips_to_keep.chlin1 <- c()
      tips_to_keep.haema1 = tips_to_keep.vc1 = tips_to_keep.ds1 = tips_to_keep.gonpe1 = tips_to_keep.tetso1 <- c()
      tips_to_keep.cz1 = tips_to_keep.rs1 = tips_to_keep.sceobli1 = tips_to_keep.desar1 = tips_to_keep.monmi1 <- c()
      tips_to_keep.caule1 = tips_to_keep.ostqu1 = tips_to_keep.brysp1 = tips_to_keep.cocco1 <- c()
      tips_to_keep.ellbi1 = tips_to_keep.myrbi1 = tips_to_keep.apalo1 = tips_to_keep.astgl1 <- c()
      tips_to_keep.tresp1 = tips_to_keep.picso1 = tips_to_keep.chlso1 = tips_to_keep.auxpr1 <- c()
      tips_to_keep.helsp1 = tips_to_keep.um1 = tips_to_keep.tetst1 = tips_to_keep.pedsp1 <- c()
      tips_to_keep.chlpr1 = tips_to_keep.picsp1 = tips_to_keep.ca1 = tips_to_keep.mv1 = tips_to_keep.mesvb1 <- c()
      tips_to_keep.kn1 = tips_to_keep.chara1 = tips_to_keep.me1 = tips_to_keep.meskr1 = tips_to_keep.sp1 <- c()
      tips_to_keep.zygci1 = tips_to_keep.zygcb1 = tips_to_keep.zygcy1 <- c()
      tips_to_keep.cp1 = tips_to_keep.pp1 = tips_to_keep.plesc1 = tips_to_keep.funhy1 <- c()
      tips_to_keep.smag1 = tips_to_keep.sphfa1 = tips_to_keep.takle1 = tips_to_keep.mp1 <- c()
      tips_to_keep.marpa1 = tips_to_keep.ricfl1 = tips_to_keep.ricso1 = tips_to_keep.aa1 <- c()
      tips_to_keep.antpu1 = tips_to_keep.antfu1 = tips_to_keep.megfl1 = tips_to_keep.phach1 <- c()
      tips_to_keep.phyph1 = tips_to_keep.parpe1 = tips_to_keep.notor1 = tips_to_keep.phaca1 <- c()
      tips_to_keep.phasp1 = tips_to_keep.leidu1 <- c()
      tips_to_keep.sm1 = tips_to_keep.isosi1 = tips_to_keep.dipco1 = tips_to_keep.hupas1 = tips_to_keep.lyccl1 <- c()
      tips_to_keep.cri1 = tips_to_keep.adica1 = tips_to_keep.alssp1 = tips_to_keep.sc1 = tips_to_keep.af1 = tips_to_keep.marve1 <- c()
      tips_to_keep.tp1 = tips_to_keep.seqse1 = tips_to_keep.seqgi1 = tips_to_keep.taxch1 = tips_to_keep.abial1 <- c()
      tips_to_keep.picab1 = tips_to_keep.gnemo1 = tips_to_keep.welmi1 = tips_to_keep.cyc1 = tips_to_keep.ginbi1 <- c()
      tips_to_keep.ambtr1 = tips_to_keep.nymco1 = tips_to_keep.cinka1 = tips_to_keep.aegi1 <- c()
      tips_to_keep.ta1 = tips_to_keep.horvu1 = tips_to_keep.bradi1 = tips_to_keep.os1 <- c()
      tips_to_keep.setit1 = tips_to_keep.sb1 = tips_to_keep.zm1 = tips_to_keep.panha1 <- c()
      tips_to_keep.missi1 = tips_to_keep.oroth1 = tips_to_keep.eleco1 = tips_to_keep.anaco1 <- c()
      tips_to_keep.musac1 = tips_to_keep.dioal1 = tips_to_keep.spipo1 = tips_to_keep.aspof1 <- c()
      tips_to_keep.zosma1 = tips_to_keep.at1 = tips_to_keep.araly1 = tips_to_keep.capgr1 <- c()
      tips_to_keep.brara1 = tips_to_keep.braol1 = tips_to_keep.sisir1 = tips_to_keep.schpa1 <- c()
      tips_to_keep.eutsa1 = tips_to_keep.thlar1 = tips_to_keep.boest1 = tips_to_keep.carpa1 <- c()
      tips_to_keep.theca1 = tips_to_keep.goshi1 = tips_to_keep.gosra1 = tips_to_keep.citsi1 <- c()
      tips_to_keep.pontr1 = tips_to_keep.eucgr1 = tips_to_keep.maldo1 = tips_to_keep.prupe1 <- c()
      tips_to_keep.frave1 = tips_to_keep.corav1 = tips_to_keep.caril1 = tips_to_keep.queru1 <- c()
      tips_to_keep.cucsa1 = tips_to_keep.glyma1 = tips_to_keep.medtr1 = tips_to_keep.tripr1 <- c()
      tips_to_keep.cicar1 = tips_to_keep.lotja1 = tips_to_keep.phaac1 = tips_to_keep.vigun1 <- c()
      tips_to_keep.lupal1 = tips_to_keep.lencu1 = tips_to_keep.arahy1 = tips_to_keep.poptr1 <- c()
      tips_to_keep.manes1 = tips_to_keep.salpu1 = tips_to_keep.linus1 = tips_to_keep.corci1 <- c()
      tips_to_keep.vitvi1 = tips_to_keep.kalfe1 = tips_to_keep.vacda1 = tips_to_keep.oleeu1 <- c()
      tips_to_keep.mimgu1 = tips_to_keep.sl1 = tips_to_keep.soltu1 = tips_to_keep.cofar1 <- c()
      tips_to_keep.lacsa1 = tips_to_keep.dauca1 = tips_to_keep.betvu1 = tips_to_keep.amahy1 <- c()
      tips_to_keep.chequ1 = tips_to_keep.sapof1 = tips_to_keep.aquco1 <- c()
      
      
      organisms.list <- c(selected_values_org1())
      
      
      # Fill vectors if organisms are in list and change names
      {
        if ("mp" %in% organisms.list)
        {
          tips_to_keep.mp1 <- grep(pattern = "marchantia_polymorpha", names(mytree)) 
        }
        
        if ("ot" %in% organisms.list)
        {
          tips_to_keep.ot1 <- grep(pattern = "ostreococcus_tauri",names(mytree)) 
        }
        
        if ("at" %in% organisms.list)
        {
          tips_to_keep.at1 <- grep(pattern = "arabidopsis_thaliana",names(mytree)) 
        }
        
        if ("cp" %in% organisms.list)
        {
          tips_to_keep.cp1 <- grep(pattern = "ceratodon_purpureus",names(mytree)) 
        }
        
        if ("cr" %in% organisms.list)
        {
          tips_to_keep.cr1 <- grep(pattern = "chlamydomonas_reinhardtii",names(mytree))
        }
        
        if ("cz" %in% organisms.list)
        {
          tips_to_keep.cz1 <- grep(pattern = "chromochloris_zofingiensis",names(mytree)) 
        }
        
        if ("kn" %in% organisms.list)
        {
          tips_to_keep.kn1 <- grep(pattern = "klebsormidium_nitens",names(mytree)) 
        }
        
        if ("me" %in% organisms.list)
        {
          tips_to_keep.me1 <- grep(pattern = "mesotaenium_endlicherianum",names(mytree)) 
        }
        
        if ("mi" %in% organisms.list)
        {
          tips_to_keep.mi1 <- grep(pattern = "micromonas_pusilla",names(mytree)) 
        }
        
        if ("pp" %in% organisms.list)
        {
          tips_to_keep.pp1 <- grep(pattern = "physcomitrium_patens",names(mytree)) 
        }
        
        if ("sl" %in% organisms.list)
        {
          tips_to_keep.sl1 <- grep(pattern = "solanum_lycopersicum",names(mytree)) 
        }
        
        if ("sm" %in% organisms.list)
        {
          tips_to_keep.sm1 <- grep(pattern = "selaginella_moellendorffii",names(mytree))
        }
        
        if ("sp" %in% organisms.list)
        {
          tips_to_keep.sp1 <- grep(pattern = "spirogloea_muscicola",names(mytree)) 
        }
        
        if ("ta" %in% organisms.list)
        {
          tips_to_keep.ta1 <- grep(pattern = "triticum_aestivum",names(mytree)) 
        }
        
        if ("vc" %in% organisms.list)
        {
          tips_to_keep.vc1 <- grep(pattern = "volvox_carteri",names(mytree))
        }
        
        if ("bp" %in% organisms.list)
        {
          tips_to_keep.bp1 <- grep(pattern = "bathycoccus_prasinos",names(mytree))
        }
        
        if ("cri" %in% organisms.list)
        {
          tips_to_keep.cri1 <- grep(pattern = "ceratopteris_richardii",names(mytree))
        }
        
        if ("ds" %in% organisms.list)
        {
          tips_to_keep.ds1 <- grep(pattern = "dunaliella_salina",names(mytree))
        }
        
        if ("os" %in% organisms.list)
        {
          tips_to_keep.os1 <- grep(pattern = "oryza_sativa",names(mytree))
        }
        
        if ("smag" %in% organisms.list)
        {
          tips_to_keep.smag1 <- grep(pattern = "sphagnum_magellanicum",names(mytree))
        }
        
        if ("tp" %in% organisms.list)
        {
          tips_to_keep.tp1 <- grep(pattern = "thuja_plicata",names(mytree))
        }
        
        if ("aa" %in% organisms.list)
        {
          tips_to_keep.aa1 <- grep(pattern = "anthoceros_agrestis",names(mytree))
        }
        
        if ("um" %in% organisms.list)
        {
          tips_to_keep.um1 <- grep(pattern = "ulva_mutabilis",names(mytree))
        }
        
        if ("rs" %in% organisms.list)
        {
          tips_to_keep.rs1 <- grep(pattern = "raphidocelis_subcapitata",names(mytree))
        }
        
        if ("cyc" %in% organisms.list)
        {
          tips_to_keep.cyc1 <- grep(pattern = "cycas_panzhihuaensis",names(mytree))
        }
        
        
        if ("ca" %in% organisms.list)
        {
          tips_to_keep.ca1 <- grep(pattern = "chlorokybus_atmophyticus",names(mytree))
        }
        
        if ("mv" %in% organisms.list)
        {
          tips_to_keep.mv1 <- grep(pattern = "mesostigma_viride1140",names(mytree))
        }
        
        if ("af" %in% organisms.list)
        {
          tips_to_keep.af1 <- grep(pattern = "azolla_filiculoides",names(mytree))
        }
        
        if ("sc" %in% organisms.list)
        {
          tips_to_keep.sc1 <- grep(pattern = "salvinia_cucullata",names(mytree))
        }
        
        if ("aegi" %in% organisms.list)
        {
          tips_to_keep.aegi1 <- grep(pattern = "aegilops_tauschii",names(mytree))
        }
        
        if ("sb" %in% organisms.list)
        {
          tips_to_keep.sb1 <- grep(pattern = "sorghum_bicolor",names(mytree))
        }
        
        if ("chara" %in% organisms.list)
        {
          tips_to_keep.chara1 <- grep(pattern = "chara_braunii",names(mytree))
        }
        
        
        if ("sceobli" %in% organisms.list)
        {
          tips_to_keep.sceobli1 <- grep(pattern = "scenedesmus_obliquus",names(mytree))
        }
        
        if ("cocco" %in% organisms.list)
        {
          tips_to_keep.cocco1 <- grep(pattern = "coccomyxa_subellipsoidea",names(mytree))
        }
        
        if ("haema" %in% organisms.list)
        {
          tips_to_keep.haema1 <- grep(pattern = "haematococcus_lacustris",names(mytree))
        }
        
        if ("zm" %in% organisms.list)
        {
          tips_to_keep.zm1 <- grep(pattern = "zea_mays",names(mytree))
        }
        
        if ("praco" %in% organisms.list)
        {
          tips_to_keep.praco1 <- grep(pattern = "prasinoderma_coloniale",names(mytree))
        }
        
        if ("ostlu" %in% organisms.list)
        {
          tips_to_keep.ostlu1 <- grep(pattern = "ostreococcus_lucimarinus",names(mytree))
        }
        
        if ("ostme" %in% organisms.list)
        {
          tips_to_keep.ostme1 <- grep(pattern = "ostreococcus_mediterraneus",names(mytree))
        }
        
        if ("micco" %in% organisms.list)
        {
          tips_to_keep.micco1 <- grep(pattern = "micromonas_commoda",names(mytree))
        }
        
        if ("cymte" %in% organisms.list)
        {
          tips_to_keep.cymte1 <- grep(pattern = "cymbomonas_tetramitiformis",names(mytree))
        }
        
        
        if ("chlin" %in% organisms.list)
        {
          tips_to_keep.chlin1 <- grep(pattern = "chlamydomonas_incerta",names(mytree))
        }
        
        if ("gonpe" %in% organisms.list)
        {
          tips_to_keep.gonpe1 <- grep(pattern = "gonium_pectorale",names(mytree))
        }
        
        if ("tetso" %in% organisms.list)
        {
          tips_to_keep.tetso1 <- grep(pattern = "tetrabaena_socialis",names(mytree))
        }
        
        if ("desar" %in% organisms.list)
        {
          tips_to_keep.desar1 <- grep(pattern = "desmodesmus_armatus",names(mytree))
        }
        
        if ("monmi" %in% organisms.list)
        {
          tips_to_keep.monmi1 <- grep(pattern = "monoraphidium_minutum",names(mytree))
        }
        
        if ("caule" %in% organisms.list)
        {
          tips_to_keep.caule1 <- grep(pattern = "caulerpa_lentillifera",names(mytree))
        }
        
        if ("ostqu" %in% organisms.list)
        {
          tips_to_keep.ostqu1 <- grep(pattern = "ostreobium_quekettii",names(mytree))
        }
        
        if ("brysp" %in% organisms.list)
        {
          tips_to_keep.brysp1 <- grep(pattern = "bryopsis_sp",names(mytree))
        }
        
        if ("ellbi" %in% organisms.list)
        {
          tips_to_keep.ellbi1 <- grep(pattern = "elliptochloris_bilobata",names(mytree))
        }
        
        if ("myrbi" %in% organisms.list)
        {
          tips_to_keep.myrbi1 <- grep(pattern = "myrmecia_bisecta",names(mytree))
        }
        
        if ("apalo" %in% organisms.list)
        {
          tips_to_keep.apalo1 <- grep(pattern = "apatococcus_lobatus",names(mytree))
        }
        
        if ("astgl" %in% organisms.list)
        {
          tips_to_keep.astgl1 <- grep(pattern = "asterochloris_glomerata",names(mytree))
        }
        
        if ("tresp" %in% organisms.list)
        {
          tips_to_keep.tresp1 <- grep(pattern = "trebouxia_sp",names(mytree))
        }
        
        if ("picso" %in% organisms.list)
        {
          tips_to_keep.picso1 <- grep(pattern = "picochlorum_soloecismus",names(mytree))
        }
        
        if ("chlso" %in% organisms.list)
        {
          tips_to_keep.chlso1 <- grep(pattern = "chlorella_sorokiniana",names(mytree))
        }
        
        if ("auxpr" %in% organisms.list)
        {
          tips_to_keep.auxpr1 <- grep(pattern = "auxenochlorella_protothecoides",names(mytree))
        }
        
        if ("helsp" %in% organisms.list)
        {
          tips_to_keep.helsp1 <- grep(pattern = "helicosporidium_sp",names(mytree))
        }
        
        if ("tetst" %in% organisms.list)
        {
          tips_to_keep.tetst1 <- grep(pattern = "tetraselmis_striata",names(mytree))
        }
        
        if ("pedsp" %in% organisms.list)
        {
          tips_to_keep.pedsp1 <- grep(pattern = "pedinophyceae_sp",names(mytree))
        }
        
        if ("chlpr" %in% organisms.list)
        {
          tips_to_keep.chlpr1 <- grep(pattern = "chloropicon_primus",names(mytree))
        }
        
        if ("picsp" %in% organisms.list)
        {
          tips_to_keep.picsp1 <- grep(pattern = "picocystis_sp",names(mytree))
        }
        
        
        if ("mesvb" %in% organisms.list)
        {
          tips_to_keep.mesvb1 <- grep(pattern = "mesostigma_viride296",names(mytree))
        }
        
        if ("meskr" %in% organisms.list)
        {
          tips_to_keep.meskr1 <- grep(pattern = "mesotaenium_kramstae",names(mytree))
        }
        
        if ("zygci" %in% organisms.list)
        {
          tips_to_keep.zygci1 <- grep(pattern = "zygnema_circumcarinatum_",names(mytree))
        }
        
        if ("zygcb" %in% organisms.list)
        {
          tips_to_keep.zygcb1 <- grep(pattern = "zygnema_circumcarinatum1559",names(mytree))
        }
        
        if ("zygcy" %in% organisms.list)
        {
          tips_to_keep.zygcy1 <- grep(pattern = "zygnema_cylindricum",names(mytree))
        }
        
        if ("plesc" %in% organisms.list)
        {
          tips_to_keep.plesc1 <- grep(pattern = "pleurozium_schreberi",names(mytree))
        }
        
        if ("funhy" %in% organisms.list)
        {
          tips_to_keep.funhy1 <- grep(pattern = "funaria_hygrometrica",names(mytree))
        }
        
        if ("sphfa" %in% organisms.list)
        {
          tips_to_keep.sphfa1 <- grep(pattern = "sphagnum_fallax",names(mytree))
        }
        
        if ("takle" %in% organisms.list)
        {
          tips_to_keep.takle1 <- grep(pattern = "takakia_lepidozioides",names(mytree))
        }
        
        if ("marpa" %in% organisms.list)
        {
          tips_to_keep.marpa1 <- grep(pattern = "marchantia_paleacea",names(mytree))
        }
        
        if ("ricfl" %in% organisms.list)
        {
          tips_to_keep.ricfl1 <- grep(pattern = "riccia_fluitans",names(mytree))
        }
        
        if ("ricso" %in% organisms.list)
        {
          tips_to_keep.ricso1 <- grep(pattern = "riccia_sorocarpa",names(mytree))
        }
        
        if ("antpu" %in% organisms.list)
        {
          tips_to_keep.antpu1 <- grep(pattern = "anthoceros_punctatus",names(mytree))
        }
        
        if ("antfu" %in% organisms.list)
        {
          tips_to_keep.antfu1 <- grep(pattern = "anthoceros_fusiformis",names(mytree))
        }
        
        if ("megfl" %in% organisms.list)
        {
          tips_to_keep.megfl1 <- grep(pattern = "megaceros_flagellaris",names(mytree))
        }
        
        if ("phach" %in% organisms.list)
        {
          tips_to_keep.phach1 <- grep(pattern = "phaeomegaceros_chiloensis",names(mytree))
        }
        
        if ("phyph" %in% organisms.list)
        {
          tips_to_keep.phyph1 <- grep(pattern = "phymatoceros_phymatodes",names(mytree))
        }
        
        if ("parpe" %in% organisms.list)
        {
          tips_to_keep.parpe1 <- grep(pattern = "paraphymatoceros_pearsonii",names(mytree))
        }
        
        if ("notor" %in% organisms.list)
        {
          tips_to_keep.notor1 <- grep(pattern = "notothylas_orbicularis",names(mytree))
        }
        
        if ("phaca" %in% organisms.list)
        {
          tips_to_keep.phaca1 <- grep(pattern = "phaeoceros_carolinianus",names(mytree))
        }
        
        if ("phasp" %in% organisms.list)
        {
          tips_to_keep.phasp1 <- grep(pattern = "phaeoceros_sp",names(mytree))
        }
        
        if ("leidu" %in% organisms.list)
        {
          tips_to_keep.leidu1 <- grep(pattern = "leiosporoceros_dussii",names(mytree))
        }
        
        if ("isosi" %in% organisms.list)
        {
          tips_to_keep.isosi1 <- grep(pattern = "isoetes_sinensis",names(mytree))
        }
        
        if ("dipco" %in% organisms.list)
        {
          tips_to_keep.dipco1 <- grep(pattern = "diphasiastrum_complanatum",names(mytree))
        }
        
        if ("hupas" %in% organisms.list)
        {
          tips_to_keep.hupas1 <- grep(pattern = "huperzia_asiatica",names(mytree))
        }
        
        if ("lyccl" %in% organisms.list)
        {
          tips_to_keep.lyccl1 <- grep(pattern = "lycopodium_clavatum",names(mytree))
        }
        
        if ("adica" %in% organisms.list)
        {
          tips_to_keep.adica1 <- grep(pattern = "adiantum_capillus",names(mytree))
        }
        
        if ("alssp" %in% organisms.list)
        {
          tips_to_keep.alssp1 <- grep(pattern = "alsophila_spinulosa",names(mytree))
        }
        
        if ("marve" %in% organisms.list)
        {
          tips_to_keep.marve1 <- grep(pattern = "marsilea_vestita",names(mytree))
        }
        
        if ("seqse" %in% organisms.list)
        {
          tips_to_keep.seqse1 <- grep(pattern = "sequoia_sempervirens",names(mytree))
        }
        
        if ("seqgi" %in% organisms.list)
        {
          tips_to_keep.seqgi1 <- grep(pattern = "sequoiadendron_giganteum",names(mytree))
        }
        
        if ("taxch" %in% organisms.list)
        {
          tips_to_keep.taxch1 <- grep(pattern = "taxus_chinensis",names(mytree))
        }
        
        if ("abial" %in% organisms.list)
        {
          tips_to_keep.abial1 <- grep(pattern = "abies_alba",names(mytree))
        }
        
        if ("picab" %in% organisms.list)
        {
          tips_to_keep.picab1 <- grep(pattern = "picea_abies",names(mytree))
        }
        
        if ("gnemo" %in% organisms.list)
        {
          tips_to_keep.gnemo1 <- grep(pattern = "gnetum_montanum",names(mytree))
        }
        
        if ("welmi" %in% organisms.list)
        {
          tips_to_keep.welmi1 <- grep(pattern = "welwitschia_mirabilis",names(mytree))
        }
        
        if ("ginbi" %in% organisms.list)
        {
          tips_to_keep.ginbi1 <- grep(pattern = "ginkgo_biloba",names(mytree))
        }
        
        if ("ambtr" %in% organisms.list)
        {
          tips_to_keep.ambtr1 <- grep(pattern = "amborella_trichopoda",names(mytree))
        }
        
        if ("nymco" %in% organisms.list)
        {
          tips_to_keep.nymco1 <- grep(pattern = "nymphaea_colorata",names(mytree))
        }
        
        if ("cinka" %in% organisms.list)
        {
          tips_to_keep.cinka1 <- grep(pattern = "cinnamomum_kanehirae",names(mytree))
        }
        
        if ("horvu" %in% organisms.list)
        {
          tips_to_keep.horvu1 <- grep(pattern = "hordeum_vulgare",names(mytree))
        }
        
        if ("bradi" %in% organisms.list)
        {
          tips_to_keep.bradi1 <- grep(pattern = "brachypodium_distachyon",names(mytree))
        }
        
        if ("setit" %in% organisms.list)
        {
          tips_to_keep.setit1 <- grep(pattern = "setaria_italica",names(mytree))
        }
        
        if ("panha" %in% organisms.list)
        {
          tips_to_keep.panha1 <- grep(pattern = "panicum_hallii",names(mytree))
        }
        
        if ("missi" %in% organisms.list)
        {
          tips_to_keep.missi1 <- grep(pattern = "miscanthus_sinensis",names(mytree))
        }
        
        if ("oroth" %in% organisms.list)
        {
          tips_to_keep.oroth1 <- grep(pattern = "oropetium_thomaeum",names(mytree))
        }
        
        if ("eleco" %in% organisms.list)
        {
          tips_to_keep.eleco1 <- grep(pattern = "eleusine_coracana",names(mytree))
        }
        
        if ("anaco" %in% organisms.list)
        {
          tips_to_keep.anaco1 <- grep(pattern = "ananas_comosus",names(mytree))
        }
        
        if ("musac" %in% organisms.list)
        {
          tips_to_keep.musac1 <- grep(pattern = "musa_acuminata",names(mytree))
        }
        
        if ("dioal" %in% organisms.list)
        {
          tips_to_keep.dioal1 <- grep(pattern = "dioscorea_alata",names(mytree))
        }
        
        if ("spipo" %in% organisms.list)
        {
          tips_to_keep.spipo1 <- grep(pattern = "spirodela_polyrhiza",names(mytree))
        }
        
        if ("aspof" %in% organisms.list)
        {
          tips_to_keep.aspof1 <- grep(pattern = "asparagus_officinalis",names(mytree))
        }
        
        if ("zosma" %in% organisms.list)
        {
          tips_to_keep.zosma1 <- grep(pattern = "zostera_marina",names(mytree))
        }
        
        if ("araly" %in% organisms.list)
        {
          tips_to_keep.araly1 <- grep(pattern = "arabidopsis_lyrata",names(mytree))
        }
        
        if ("capgr" %in% organisms.list)
        {
          tips_to_keep.capgr1 <- grep(pattern = "capsella_grandiflora",names(mytree))
        }
        
        if ("brara" %in% organisms.list)
        {
          tips_to_keep.brara1 <- grep(pattern = "brassica_rapa",names(mytree))
        }
        
        if ("braol" %in% organisms.list)
        {
          tips_to_keep.braol1 <- grep(pattern = "brassica_oleracea",names(mytree))
        }
        
        if ("sisir" %in% organisms.list)
        {
          tips_to_keep.sisir1 <- grep(pattern = "sisymbrium_irio",names(mytree))
        }
        
        if ("schpa" %in% organisms.list)
        {
          tips_to_keep.schpa1 <- grep(pattern = "schrenkiella_parvula",names(mytree))
        }
        
        if ("eutsa" %in% organisms.list)
        {
          tips_to_keep.eutsa1 <- grep(pattern = "eutrema_salsugineum",names(mytree))
        }
        
        if ("thlar" %in% organisms.list)
        {
          tips_to_keep.thlar1 <- grep(pattern = "thlaspi_arvense",names(mytree))
        }
        
        if ("boest" %in% organisms.list)
        {
          tips_to_keep.boest1 <- grep(pattern = "boechera_stricta",names(mytree))
        }
        
        if ("carpa" %in% organisms.list)
        {
          tips_to_keep.carpa1 <- grep(pattern = "carica_papaya",names(mytree))
        }
        
        if ("theca" %in% organisms.list)
        {
          tips_to_keep.theca1 <- grep(pattern = "theobroma_cacao",names(mytree))
        }
        
        if ("goshi" %in% organisms.list)
        {
          tips_to_keep.goshi1 <- grep(pattern = "gossypium_hirsutum",names(mytree))
        }
        
        if ("gosra" %in% organisms.list)
        {
          tips_to_keep.gosra1 <- grep(pattern = "gossypium_raimondii",names(mytree))
        }
        
        if ("citsi" %in% organisms.list)
        {
          tips_to_keep.citsi1 <- grep(pattern = "citrus_sinensis",names(mytree))
        }
        
        if ("pontr" %in% organisms.list)
        {
          tips_to_keep.pontr1 <- grep(pattern = "poncirus_trifoliata",names(mytree))
        }
        
        if ("eucgr" %in% organisms.list)
        {
          tips_to_keep.eucgr1 <- grep(pattern = "eucalyptus_grandis",names(mytree))
        }
        
        if ("maldo" %in% organisms.list)
        {
          tips_to_keep.maldo1 <- grep(pattern = "malus_domestica",names(mytree))
        }
        
        if ("prupe" %in% organisms.list)
        {
          tips_to_keep.prupe1 <- grep(pattern = "prunus_persica",names(mytree))
        }
        
        if ("frave" %in% organisms.list)
        {
          tips_to_keep.frave1 <- grep(pattern = "fragaria_vesca",names(mytree))
        }
        
        if ("corav" %in% organisms.list)
        {
          tips_to_keep.corav1 <- grep(pattern = "corylus_avellana",names(mytree))
        }
        
        if ("caril" %in% organisms.list)
        {
          tips_to_keep.caril1 <- grep(pattern = "carya_illinoinensis",names(mytree))
        }
        
        if ("queru" %in% organisms.list)
        {
          tips_to_keep.queru1 <- grep(pattern = "quercus_rubra",names(mytree))
        }
        
        if ("cucsa" %in% organisms.list)
        {
          tips_to_keep.cucsa1 <- grep(pattern = "cucumis_sativus",names(mytree))
        }
        
        if ("glyma" %in% organisms.list)
        {
          tips_to_keep.glyma1 <- grep(pattern = "glycine_max",names(mytree))
        }
        
        if ("medtr" %in% organisms.list)
        {
          tips_to_keep.medtr1 <- grep(pattern = "medicago_truncatula",names(mytree))
        }
        
        if ("tripr" %in% organisms.list)
        {
          tips_to_keep.tripr1 <- grep(pattern = "trifolium_pratense",names(mytree))
        }
        
        if ("cicar" %in% organisms.list)
        {
          tips_to_keep.cicar1 <- grep(pattern = "cicer_arietinum",names(mytree))
        }
        
        if ("lotja" %in% organisms.list)
        {
          tips_to_keep.lotja1 <- grep(pattern = "lotus_japonicus",names(mytree))
        }
        
        if ("phaac" %in% organisms.list)
        {
          tips_to_keep.phaac1 <- grep(pattern = "phaseolus_acutifolius",names(mytree))
        }
        
        if ("vigun" %in% organisms.list)
        {
          tips_to_keep.vigun1 <- grep(pattern = "vigna_unguiculata",names(mytree))
        }
        
        if ("lupal" %in% organisms.list)
        {
          tips_to_keep.lupal1 <- grep(pattern = "lupinus_albus",names(mytree))
        }
        
        if ("lencu" %in% organisms.list)
        {
          tips_to_keep.lencu1 <- grep(pattern = "lens_culinaris",names(mytree))
        }
        
        if ("arahy" %in% organisms.list)
        {
          tips_to_keep.arahy1 <- grep(pattern = "arachis_hypogaea",names(mytree))
        }
        
        if ("poptr" %in% organisms.list)
        {
          tips_to_keep.poptr1 <- grep(pattern = "populus_trichocarpa",names(mytree))
        }
        
        if ("manes" %in% organisms.list)
        {
          tips_to_keep.manes1 <- grep(pattern = "manihot_esculenta",names(mytree))
        }
        
        if ("salpu" %in% organisms.list)
        {
          tips_to_keep.salpu1 <- grep(pattern = "salix_purpurea",names(mytree))
        }
        
        if ("linus" %in% organisms.list)
        {
          tips_to_keep.linus1 <- grep(pattern = "linum_usitatissimum",names(mytree))
        }
        
        if ("corci" %in% organisms.list)
        {
          tips_to_keep.corci1 <- grep(pattern = "corymbia_citriodora",names(mytree))
        }
        
        if ("vitvi" %in% organisms.list)
        {
          tips_to_keep.vitvi1 <- grep(pattern = "vitis_vinifera",names(mytree))
        }
        
        if ("kalfe" %in% organisms.list)
        {
          tips_to_keep.kalfe1 <- grep(pattern = "kalanchoe_fedtschenkoi",names(mytree))
        }
        
        if ("vacda" %in% organisms.list)
        {
          tips_to_keep.vacda1 <- grep(pattern = "vaccinium_darrowii",names(mytree))
        }
        
        if ("oleeu" %in% organisms.list)
        {
          tips_to_keep.oleeu1 <- grep(pattern = "olea_europaea",names(mytree))
        }
        
        if ("mimgu" %in% organisms.list)
        {
          tips_to_keep.mimgu1 <- grep(pattern = "mimulus_guttatus",names(mytree))
        }
        
        if ("soltu" %in% organisms.list)
        {
          tips_to_keep.soltu1 <- grep(pattern = "solanum_tuberosum",names(mytree))
        }
        
        if ("cofar" %in% organisms.list)
        {
          tips_to_keep.cofar1 <- grep(pattern = "coffea_arabica",names(mytree))
        }
        
        if ("lacsa" %in% organisms.list)
        {
          tips_to_keep.lacsa1 <- grep(pattern = "lactuca_sativa",names(mytree))
        }
        
        if ("dauca" %in% organisms.list)
        {
          tips_to_keep.dauca1 <- grep(pattern = "daucus_carota",names(mytree))
        }
        
        if ("betvu" %in% organisms.list)
        {
          tips_to_keep.betvu1 <- grep(pattern = "beta_vulgaris",names(mytree))
        }
        
        if ("amahy" %in% organisms.list)
        {
          tips_to_keep.amahy1 <- grep(pattern = "amaranthus_hypochondriacus",names(mytree))
        }
        
        if ("chequ" %in% organisms.list)
        {
          tips_to_keep.chequ1 <- grep(pattern = "chenopodium_quinoa",names(mytree))
        }
        
        if ("sapof" %in% organisms.list)
        {
          tips_to_keep.sapof1 <- grep(pattern = "saponaria_officinalis",names(mytree))
        }
        
        if ("aquco" %in% organisms.list)
        {
          tips_to_keep.aquco1 <- grep(pattern = "aquilegia_coerulea",names(mytree))
        }
      }
      
      
      # Concatenate indexes to keep and subset MSA
      tips_to_keep.global <- c(tips_to_keep.praco1, tips_to_keep.ot1, tips_to_keep.ostlu1, tips_to_keep.ostme1, tips_to_keep.bp1,
                               tips_to_keep.mi1, tips_to_keep.micco1, tips_to_keep.cymte1, tips_to_keep.cr1, tips_to_keep.chlin1,
                               tips_to_keep.haema1, tips_to_keep.vc1, tips_to_keep.ds1, tips_to_keep.gonpe1, tips_to_keep.tetso1,
                               tips_to_keep.cz1, tips_to_keep.rs1, tips_to_keep.sceobli1, tips_to_keep.desar1, tips_to_keep.monmi1,
                               tips_to_keep.caule1, tips_to_keep.ostqu1, tips_to_keep.brysp1, tips_to_keep.cocco1,
                               tips_to_keep.ellbi1, tips_to_keep.myrbi1, tips_to_keep.apalo1, tips_to_keep.astgl1,
                               tips_to_keep.tresp1, tips_to_keep.picso1, tips_to_keep.chlso1, tips_to_keep.auxpr1,
                               tips_to_keep.helsp1, tips_to_keep.um1, tips_to_keep.tetst1, tips_to_keep.pedsp1,
                               tips_to_keep.chlpr1, tips_to_keep.picsp1, tips_to_keep.ca1, tips_to_keep.mv1, tips_to_keep.mesvb1,
                               tips_to_keep.kn1, tips_to_keep.chara1, tips_to_keep.me1, tips_to_keep.meskr1, tips_to_keep.sp1,
                               tips_to_keep.zygci1, tips_to_keep.zygcb1, tips_to_keep.zygcy1,
                               tips_to_keep.cp1, tips_to_keep.pp1, tips_to_keep.plesc1, tips_to_keep.funhy1,
                               tips_to_keep.smag1, tips_to_keep.sphfa1, tips_to_keep.takle1, tips_to_keep.mp1,
                               tips_to_keep.marpa1, tips_to_keep.ricfl1, tips_to_keep.ricso1, tips_to_keep.aa1,
                               tips_to_keep.antpu1, tips_to_keep.antfu1, tips_to_keep.megfl1, tips_to_keep.phach1,
                               tips_to_keep.phyph1, tips_to_keep.parpe1, tips_to_keep.notor1, tips_to_keep.phaca1,
                               tips_to_keep.phasp1, tips_to_keep.leidu1, tips_to_keep.sm1, tips_to_keep.isosi1, 
                               tips_to_keep.dipco1, tips_to_keep.hupas1, tips_to_keep.lyccl1,
                               tips_to_keep.cri1, tips_to_keep.adica1, tips_to_keep.alssp1, tips_to_keep.sc1, 
                               tips_to_keep.af1, tips_to_keep.marve1, tips_to_keep.tp1, tips_to_keep.seqse1, 
                               tips_to_keep.seqgi1, tips_to_keep.taxch1, tips_to_keep.abial1, tips_to_keep.picab1,
                               tips_to_keep.gnemo1, tips_to_keep.welmi1, tips_to_keep.cyc1, tips_to_keep.ginbi1,
                               tips_to_keep.ambtr1, tips_to_keep.nymco1, tips_to_keep.cinka1, tips_to_keep.aegi1,
                               tips_to_keep.ta1, tips_to_keep.horvu1, tips_to_keep.bradi1, tips_to_keep.os1,
                               tips_to_keep.setit1, tips_to_keep.sb1, tips_to_keep.zm1, tips_to_keep.panha1,
                               tips_to_keep.missi1, tips_to_keep.oroth1, tips_to_keep.eleco1, tips_to_keep.anaco1,
                               tips_to_keep.musac1, tips_to_keep.dioal1, tips_to_keep.spipo1, tips_to_keep.aspof1,
                               tips_to_keep.zosma1, tips_to_keep.at1, tips_to_keep.araly1, tips_to_keep.capgr1,
                               tips_to_keep.brara1, tips_to_keep.braol1, tips_to_keep.sisir1, tips_to_keep.schpa1,
                               tips_to_keep.eutsa1, tips_to_keep.thlar1, tips_to_keep.boest1, tips_to_keep.carpa1,
                               tips_to_keep.theca1, tips_to_keep.goshi1, tips_to_keep.gosra1, tips_to_keep.citsi1,
                               tips_to_keep.pontr1, tips_to_keep.eucgr1, tips_to_keep.maldo1, tips_to_keep.prupe1,
                               tips_to_keep.frave1, tips_to_keep.corav1, tips_to_keep.caril1, tips_to_keep.queru1,
                               tips_to_keep.cucsa1, tips_to_keep.glyma1, tips_to_keep.medtr1, tips_to_keep.tripr1,
                               tips_to_keep.cicar1, tips_to_keep.lotja1, tips_to_keep.phaac1, tips_to_keep.vigun1,
                               tips_to_keep.lupal1, tips_to_keep.lencu1, tips_to_keep.arahy1, tips_to_keep.poptr1,
                               tips_to_keep.manes1, tips_to_keep.salpu1, tips_to_keep.linus1, tips_to_keep.corci1,
                               tips_to_keep.vitvi1, tips_to_keep.kalfe1, tips_to_keep.vacda1, tips_to_keep.oleeu1,
                               tips_to_keep.mimgu1, tips_to_keep.sl1, tips_to_keep.soltu1, tips_to_keep.cofar1,
                               tips_to_keep.lacsa1, tips_to_keep.dauca1, tips_to_keep.betvu1, tips_to_keep.amahy1,
                               tips_to_keep.chequ1, tips_to_keep.sapof1, tips_to_keep.aquco1)
      
      
      my_subset_tree <- subset(mytree, tips_to_keep.global)
      
      {
      if (build_trees1() == "Neighbour Joining")
      {
        # Create dist matrix for NJ and build tree
        dm <- dist.ml(my_subset_tree)
        treeNJ  <- NJ(dm)
        
        # Bootstrap for NJ
        fun_nj <- function(x) NJ(dist.ml(x))
        bs_nj <- bootstrap.phyDat(my_subset_tree, fun_nj, bs=100)
        
        # Save bootstrap values to the tree
        tree <- plotBS(treeNJ, bs_nj, type = "n")
        # Rooting tree
        tree <- midpoint(tree)
        return(tree)
      }
      
      else if (build_trees1() == "UPGMA")
      {
        dm <- dist.ml(my_subset_tree)
        treeUPGMA  <- upgma(dm)
        
        # Bootstrap for UPGMA
        fun_upgma <- function(x) upgma(dist.ml(x))
        bs_upgma <- bootstrap.phyDat(my_subset_tree, fun_upgma, bs=100)
        
        # Save bootstrap values to the tree
        tree <- addConfidences(treeUPGMA, bs_upgma)
        return(tree)
      }
      
      }
      
    }
    }
  }) %>% bindEvent(input$run_button1)
  
  ortho_seq1 <- reactive({
    file.name <- og.name1()
  
    # Load orthogroup sequences file
    ortho.seq.name <- paste("Orthogroup_Sequences",paste(file.name, "fa", sep = "."), sep="/")
    
    ortho_seq <- seqinr::read.fasta(ortho.seq.name, seqtype = "AA")
    return(ortho_seq)
  })
  
  # Tips to keep of each species with proper notation
  tips_to_keep.mp1 <- reactive({
    
    tree <- tree1()
    # Selection of organisms
    organisms.list <- c(selected_values_org1())
    
    # Selection of genes from the selected organism
    tips_to_keep.mp <- c()
    if ("mp" %in% organisms.list)
    {
      tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label)
    }
    return(tips_to_keep.mp)
  })
  
  tips_to_keep.ot1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ot <- c()
    if ("ot" %in% organisms.list)
    {
      tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label)
    }
    return(tips_to_keep.ot)
  })
  
  tips_to_keep.at1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.at <- c()
    if ("at" %in% organisms.list)
    {
      tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label)
    }
    return(tips_to_keep.at)
  })
  
  tips_to_keep.cp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cp <- c()
    if ("cp" %in% organisms.list)
    {
      tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label)
    }
    
    return(tips_to_keep.cp)
  })
  
  tips_to_keep.cr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cr <- c()
    if ("cr" %in% organisms.list)
    {
      tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
    }
    return(tips_to_keep.cr)
  })
  
  tips_to_keep.cz1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cz <- c()
    if ("cz" %in% organisms.list)
    {
      tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label)
    }
    return(tips_to_keep.cz)
  })
  
  tips_to_keep.kn1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.kn <- c()
    if ("kn" %in% organisms.list)
    {
      tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label)
    }
    return(tips_to_keep.kn)
  })
  
  tips_to_keep.me1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.me <- c()
    if ("me" %in% organisms.list)
    {
      tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label)
    }
    return(tips_to_keep.me)
  })
  
  tips_to_keep.mi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.mi <- c()
    if ("mi" %in% organisms.list)
    {
      tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label)
    }
    return(tips_to_keep.mi)
  })
  
  tips_to_keep.pp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.pp <- c()
    if ("pp" %in% organisms.list)
    {
      tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label)
    }
    return(tips_to_keep.pp)
  })
  
  tips_to_keep.sl1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sl <- c()
    if ("sl" %in% organisms.list)
    {
      tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label)
    }
    return(tips_to_keep.sl)
  })
  
  tips_to_keep.sm1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sm <- c()
    if ("sm" %in% organisms.list)
    {
      tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label)
    }
    return(tips_to_keep.sm)
  })
  
  tips_to_keep.sp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sp <- c()
    if ("sp" %in% organisms.list)
    {
      tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label)
    }
    return(tips_to_keep.sp)
  })
  
  tips_to_keep.ta1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ta <- c()
    if ("ta" %in% organisms.list)
    {
      tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label)
    }
    return(tips_to_keep.ta)
  })
  
  tips_to_keep.vc1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.vc <- c()
    if ("vc" %in% organisms.list)
    {
      tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
    }
    return(tips_to_keep.vc)
  })
  
  tips_to_keep.bp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.bp <- c()
    if ("bp" %in% organisms.list)
    {
      tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
    }
    return(tips_to_keep.bp)
  })
  
  tips_to_keep.cri1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cri <- c()
    if ("cri" %in% organisms.list)
    {
      tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
    }
    return(tips_to_keep.cri)
  })
  
  tips_to_keep.ds1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ds <- c()
    if ("ds" %in% organisms.list)
    {
      tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
    }
    return(tips_to_keep.ds)
  })
  
  tips_to_keep.os1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.os <- c()
    if ("os" %in% organisms.list)
    {
      tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
    }
    return(tips_to_keep.os)
  })
  
  tips_to_keep.smag1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.smag <- c()
    if ("smag" %in% organisms.list)
    {
      tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
    }
    return(tips_to_keep.smag)
  })
  
  tips_to_keep.tp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.tp <- c()
    if ("tp" %in% organisms.list)
    {
      tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
    }
    return(tips_to_keep.tp)
  })
  
  tips_to_keep.aa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.aa <- c()
    if ("aa" %in% organisms.list)
    {
      tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
    }
    return(tips_to_keep.aa)
  })
  
  tips_to_keep.um1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.um <- c()
    if ("um" %in% organisms.list)
    {
      tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
    }
    return(tips_to_keep.um)
  })
  
  tips_to_keep.rs1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.rs <- c()
    if ("rs" %in% organisms.list)
    {
      tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
    }
    return(tips_to_keep.rs)
  })
  
  tips_to_keep.cyc1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cyc <- c()
    if ("cyc" %in% organisms.list)
    {
      tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
    }
    return(tips_to_keep.cyc)
  })
 
  
  tips_to_keep.ca1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ca <- c()
    if ("ca" %in% organisms.list)
    {
      tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
    }
    return(tips_to_keep.ca)
  })
  
  tips_to_keep.mv1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.mv <- c()
    if ("mv" %in% organisms.list)
    {
      tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
    }
    return(tips_to_keep.mv)
  })
  
  tips_to_keep.af1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.af <- c()
    if ("af" %in% organisms.list)
    {
      tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
    }
    return(tips_to_keep.af)
  })
  
  tips_to_keep.sc1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sc <- c()
    if ("sc" %in% organisms.list)
    {
      tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
    }
    return(tips_to_keep.sc)
  })
  
  tips_to_keep.aegi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.aegi <- c()
    if ("aegi" %in% organisms.list)
    {
      tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
    }
    return(tips_to_keep.aegi)
  })
  
  tips_to_keep.sb1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sb <- c()
    if ("sb" %in% organisms.list)
    {
      tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
    }
    return(tips_to_keep.sb)
  })
  
  tips_to_keep.chara1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.chara <- c()
    if ("chara" %in% organisms.list)
    {
      tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
    }
    return(tips_to_keep.chara)
  })
  
  tips_to_keep.sceobli1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sceobli <- c()
    if ("sceobli" %in% organisms.list)
    {
      tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
    }
    return(tips_to_keep.sceobli)
  })
  
  tips_to_keep.cocco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cocco <- c()
    if ("cocco" %in% organisms.list)
    {
      tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
    }
    return(tips_to_keep.cocco)
  })

  
  tips_to_keep.haema1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.haema <- c()
    if ("haema" %in% organisms.list)
    {
      tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
    }
    return(tips_to_keep.haema)
  })
  
  tips_to_keep.zm1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.zm <- c()
    if ("zm" %in% organisms.list)
    {
      tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
    }
    return(tips_to_keep.zm)
  })
  
  tips_to_keep.praco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.praco <- c()
    if ("praco" %in% organisms.list)
    {
      tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
    }
    return(tips_to_keep.praco)
  })
  
  tips_to_keep.ostlu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ostlu <- c()
    if ("ostlu" %in% organisms.list)
    {
      tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
    }
    return(tips_to_keep.ostlu)
  })
  
  tips_to_keep.ostme1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ostme <- c()
    if ("ostme" %in% organisms.list)
    {
      tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
    }
    return(tips_to_keep.ostme)
  })
  
  tips_to_keep.micco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.micco <- c()
    if ("micco" %in% organisms.list)
    {
      tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
    }
    return(tips_to_keep.micco)
  })
  
  tips_to_keep.cymte1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cymte <- c()
    if ("cymte" %in% organisms.list)
    {
      tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
    }
    return(tips_to_keep.cymte)
  })
  
  tips_to_keep.chlin1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.chlin <- c()
    if ("chlin" %in% organisms.list)
    {
      tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
    }
    return(tips_to_keep.chlin)
  })
  
  tips_to_keep.gonpe1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.gonpe <- c()
    if ("gonpe" %in% organisms.list)
    {
      tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
    }
    return(tips_to_keep.gonpe)
  })
  
  tips_to_keep.tetso1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.tetso <- c()
    if ("tetso" %in% organisms.list)
    {
      tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
    }
    return(tips_to_keep.tetso)
  })
  
  tips_to_keep.desar1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.desar <- c()
    if ("desar" %in% organisms.list)
    {
      tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
    }
    return(tips_to_keep.desar)
  })
  
  tips_to_keep.monmi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.monmi <- c()
    if ("monmi" %in% organisms.list)
    {
      tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
    }
    return(tips_to_keep.monmi)
  })
  
  tips_to_keep.caule1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.caule <- c()
    if ("caule" %in% organisms.list)
    {
      tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
    }
    return(tips_to_keep.caule)
  })
  
  tips_to_keep.ostqu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ostqu <- c()
    if ("ostqu" %in% organisms.list)
    {
      tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
    }
    return(tips_to_keep.ostqu)
  })
  
  tips_to_keep.brysp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.brysp <- c()
    if ("brysp" %in% organisms.list)
    {
      tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
    }
    return(tips_to_keep.brysp)
  })
  
  tips_to_keep.ellbi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ellbi <- c()
    if ("ellbi" %in% organisms.list)
    {
      tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
    }
    return(tips_to_keep.ellbi)
  })
  
  tips_to_keep.myrbi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.myrbi <- c()
    if ("myrbi" %in% organisms.list)
    {
      tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
    }
    return(tips_to_keep.myrbi)
  })
  
  tips_to_keep.apalo1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.apalo <- c()
    if ("apalo" %in% organisms.list)
    {
      tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
    }
    return(tips_to_keep.apalo)
  })
  
  tips_to_keep.astgl1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.astgl <- c()
    if ("astgl" %in% organisms.list)
    {
      tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
    }
    return(tips_to_keep.astgl)
  })
  
  tips_to_keep.tresp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.tresp <- c()
    if ("tresp" %in% organisms.list)
    {
      tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
    }
    return(tips_to_keep.tresp)
  })
  
  tips_to_keep.picso1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.picso <- c()
    if ("picso" %in% organisms.list)
    {
      tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
    }
    return(tips_to_keep.picso)
  })
  
  tips_to_keep.chlso1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.chlso <- c()
    if ("chlso" %in% organisms.list)
    {
      tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
    }
    return(tips_to_keep.chlso)
  })
  
  tips_to_keep.auxpr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.auxpr <- c()
    if ("auxpr" %in% organisms.list)
    {
      tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
    }
    return(tips_to_keep.auxpr)
  })
  
  tips_to_keep.helsp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.helsp <- c()
    if ("helsp" %in% organisms.list)
    {
      tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
    }
    return(tips_to_keep.helsp)
  })
  
  tips_to_keep.tetst1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.tetst <- c()
    if ("tetst" %in% organisms.list)
    {
      tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
    }
    return(tips_to_keep.tetst)
  })
  
  tips_to_keep.pedsp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.pedsp <- c()
    if ("pedsp" %in% organisms.list)
    {
      tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
    }
    return(tips_to_keep.pedsp)
  })
  
  tips_to_keep.chlpr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.chlpr <- c()
    if ("chlpr" %in% organisms.list)
    {
      tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
    }
    return(tips_to_keep.chlpr)
  })
  
  tips_to_keep.picsp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.picsp <- c()
    if ("picsp" %in% organisms.list)
    {
      tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
    }
    return(tips_to_keep.picsp)
  })
  
  tips_to_keep.mesvb1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.mesvb <- c()
    if ("mesvb" %in% organisms.list)
    {
      tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
    }
    return(tips_to_keep.mesvb)
  })
  
  tips_to_keep.meskr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.meskr <- c()
    if ("meskr" %in% organisms.list)
    {
      tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
    }
    return(tips_to_keep.meskr)
  })
  
  tips_to_keep.zygci1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.zygci <- c()
    if ("zygci" %in% organisms.list)
    {
      tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
    }
    return(tips_to_keep.zygci)
  })
  
  tips_to_keep.zygcb1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.zygcb <- c()
    if ("zygcb" %in% organisms.list)
    {
      tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
    }
    return(tips_to_keep.zygcb)
  })
  
  tips_to_keep.zygcy1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.zygcy <- c()
    if ("zygcy" %in% organisms.list)
    {
      tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
    }
    return(tips_to_keep.zygcy)
  })
  
  tips_to_keep.plesc1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.plesc <- c()
    if ("plesc" %in% organisms.list)
    {
      tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
    }
    return(tips_to_keep.plesc)
  })
  
  tips_to_keep.funhy1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.funhy <- c()
    if ("funhy" %in% organisms.list)
    {
      tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
    }
    return(tips_to_keep.funhy)
  })
  
  tips_to_keep.sphfa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sphfa <- c()
    if ("sphfa" %in% organisms.list)
    {
      tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
    }
    return(tips_to_keep.sphfa)
  })
  
  tips_to_keep.takle1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.takle <- c()
    if ("takle" %in% organisms.list)
    {
      tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
    }
    return(tips_to_keep.takle)
  })
  
  tips_to_keep.marpa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.marpa <- c()
    if ("marpa" %in% organisms.list)
    {
      tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
    }
    return(tips_to_keep.marpa)
  })
  
  tips_to_keep.ricfl1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ricfl <- c()
    if ("ricfl" %in% organisms.list)
    {
      tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
    }
    return(tips_to_keep.ricfl)
  })
  
  tips_to_keep.ricso1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ricso <- c()
    if ("ricso" %in% organisms.list)
    {
      tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
    }
    return(tips_to_keep.ricso)
  })
  
  tips_to_keep.antpu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.antpu <- c()
    if ("antpu" %in% organisms.list)
    {
      tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
    }
    return(tips_to_keep.antpu)
  })
  
  tips_to_keep.antfu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.antfu <- c()
    if ("antfu" %in% organisms.list)
    {
      tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
    }
    return(tips_to_keep.antfu)
  })
  
  tips_to_keep.megfl1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.megfl <- c()
    if ("megfl" %in% organisms.list)
    {
      tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
    }
    return(tips_to_keep.megfl)
  })
  
  tips_to_keep.phach1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.phach <- c()
    if ("phach" %in% organisms.list)
    {
      tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
    }
    return(tips_to_keep.phach)
  })
  
  tips_to_keep.phyph1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.phyph <- c()
    if ("phyph" %in% organisms.list)
    {
      tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
    }
    return(tips_to_keep.phyph)
  })
  
  tips_to_keep.parpe1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.parpe <- c()
    if ("parpe" %in% organisms.list)
    {
      tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
    }
    return(tips_to_keep.parpe)
  })
  
  tips_to_keep.notor1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.notor <- c()
    if ("notor" %in% organisms.list)
    {
      tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
    }
    return(tips_to_keep.notor)
  })
  
  tips_to_keep.phaca1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.phaca <- c()
    if ("phaca" %in% organisms.list)
    {
      tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
    }
    return(tips_to_keep.phaca)
  })
  
  tips_to_keep.phasp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.phasp <- c()
    if ("phasp" %in% organisms.list)
    {
      tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
    }
    return(tips_to_keep.phasp)
  })
  
  tips_to_keep.leidu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.leidu <- c()
    if ("leidu" %in% organisms.list)
    {
      tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
    }
    return(tips_to_keep.leidu)
  })
  
  tips_to_keep.isosi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.isosi <- c()
    if ("isosi" %in% organisms.list)
    {
      tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
    }
    return(tips_to_keep.isosi)
  })
  
  tips_to_keep.dipco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.dipco <- c()
    if ("dipco" %in% organisms.list)
    {
      tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
    }
    return(tips_to_keep.dipco)
  })
  
  tips_to_keep.hupas1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.hupas <- c()
    if ("hupas" %in% organisms.list)
    {
      tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
    }
    return(tips_to_keep.hupas)
  })
  
  tips_to_keep.lyccl1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.lyccl <- c()
    if ("lyccl" %in% organisms.list)
    {
      tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
    }
    return(tips_to_keep.lyccl)
  })
  
  tips_to_keep.adica1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.adica <- c()
    if ("adica" %in% organisms.list)
    {
      tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
    }
    return(tips_to_keep.adica)
  })
  
  tips_to_keep.alssp1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.alssp <- c()
    if ("alssp" %in% organisms.list)
    {
      tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
    }
    return(tips_to_keep.alssp)
  })
  
  tips_to_keep.marve1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.marve <- c()
    if ("marve" %in% organisms.list)
    {
      tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
    }
    return(tips_to_keep.marve)
  })
  
  tips_to_keep.seqse1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.seqse <- c()
    if ("seqse" %in% organisms.list)
    {
      tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
    }
    return(tips_to_keep.seqse)
  })
  
  tips_to_keep.seqgi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.seqgi <- c()
    if ("seqgi" %in% organisms.list)
    {
      tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
    }
    return(tips_to_keep.seqgi)
  })
  
  tips_to_keep.taxch1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.taxch <- c()
    if ("taxch" %in% organisms.list)
    {
      tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
    }
    return(tips_to_keep.taxch)
  })
  
  tips_to_keep.abial1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.abial <- c()
    if ("abial" %in% organisms.list)
    {
      tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
    }
    return(tips_to_keep.abial)
  })
  
  tips_to_keep.picab1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.picab <- c()
    if ("picab" %in% organisms.list)
    {
      tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
    }
    return(tips_to_keep.picab)
  })
  
  tips_to_keep.gnemo1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.gnemo <- c()
    if ("gnemo" %in% organisms.list)
    {
      tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
    }
    return(tips_to_keep.gnemo)
  })
  
  tips_to_keep.welmi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.welmi <- c()
    if ("welmi" %in% organisms.list)
    {
      tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
    }
    return(tips_to_keep.welmi)
  })
  
  tips_to_keep.ginbi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ginbi <- c()
    if ("ginbi" %in% organisms.list)
    {
      tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
    }
    return(tips_to_keep.ginbi)
  })
  
  tips_to_keep.ambtr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.ambtr <- c()
    if ("ambtr" %in% organisms.list)
    {
      tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
    }
    return(tips_to_keep.ambtr)
  })
  
  tips_to_keep.nymco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.nymco <- c()
    if ("nymco" %in% organisms.list)
    {
      tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
    }
    return(tips_to_keep.nymco)
  })
  
  tips_to_keep.cinka1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cinka <- c()
    if ("cinka" %in% organisms.list)
    {
      tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
    }
    return(tips_to_keep.cinka)
  })
  
  tips_to_keep.horvu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.horvu <- c()
    if ("horvu" %in% organisms.list)
    {
      tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
    }
    return(tips_to_keep.horvu)
  })
  
  tips_to_keep.bradi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.bradi <- c()
    if ("bradi" %in% organisms.list)
    {
      tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
    }
    return(tips_to_keep.bradi)
  })
  
  tips_to_keep.setit1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.setit <- c()
    if ("setit" %in% organisms.list)
    {
      tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
    }
    return(tips_to_keep.setit)
  })
  
  tips_to_keep.panha1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.panha <- c()
    if ("panha" %in% organisms.list)
    {
      tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
    }
    return(tips_to_keep.panha)
  })
  
  tips_to_keep.missi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.missi <- c()
    if ("missi" %in% organisms.list)
    {
      tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
    }
    return(tips_to_keep.missi)
  })
  
  tips_to_keep.oroth1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.oroth <- c()
    if ("oroth" %in% organisms.list)
    {
      tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
    }
    return(tips_to_keep.oroth)
  })
  
  tips_to_keep.eleco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.eleco <- c()
    if ("eleco" %in% organisms.list)
    {
      tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
    }
    return(tips_to_keep.eleco)
  })
  
  tips_to_keep.anaco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.anaco <- c()
    if ("anaco" %in% organisms.list)
    {
      tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
    }
    return(tips_to_keep.anaco)
  })
  
  tips_to_keep.musac1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.musac <- c()
    if ("musac" %in% organisms.list)
    {
      tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
    }
    return(tips_to_keep.musac)
  })
  
  tips_to_keep.dioal1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.dioal <- c()
    if ("dioal" %in% organisms.list)
    {
      tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
    }
    return(tips_to_keep.dioal)
  })
  
  tips_to_keep.spipo1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.spipo <- c()
    if ("spipo" %in% organisms.list)
    {
      tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
    }
    return(tips_to_keep.spipo)
  })
  
  tips_to_keep.aspof1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.aspof <- c()
    if ("aspof" %in% organisms.list)
    {
      tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
    }
    return(tips_to_keep.aspof)
  })
  
  tips_to_keep.zosma1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.zosma <- c()
    if ("zosma" %in% organisms.list)
    {
      tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
    }
    return(tips_to_keep.zosma)
  })
  
  tips_to_keep.araly1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.araly <- c()
    if ("araly" %in% organisms.list)
    {
      tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
    }
    return(tips_to_keep.araly)
  })
  
  tips_to_keep.capgr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.capgr <- c()
    if ("capgr" %in% organisms.list)
    {
      tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
    }
    return(tips_to_keep.capgr)
  })
  
  tips_to_keep.brara1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.brara <- c()
    if ("brara" %in% organisms.list)
    {
      tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
    }
    return(tips_to_keep.brara)
  })
  
  tips_to_keep.braol1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.braol <- c()
    if ("braol" %in% organisms.list)
    {
      tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
    }
    return(tips_to_keep.braol)
  })
  
  tips_to_keep.sisir1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sisir <- c()
    if ("sisir" %in% organisms.list)
    {
      tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
    }
    return(tips_to_keep.sisir)
  })
  
  tips_to_keep.schpa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.schpa <- c()
    if ("schpa" %in% organisms.list)
    {
      tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
    }
    return(tips_to_keep.schpa)
  })
  
  tips_to_keep.eutsa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.eutsa <- c()
    if ("eutsa" %in% organisms.list)
    {
      tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
    }
    return(tips_to_keep.eutsa)
  })
  
  tips_to_keep.thlar1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.thlar <- c()
    if ("thlar" %in% organisms.list)
    {
      tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
    }
    return(tips_to_keep.thlar)
  })
  
  tips_to_keep.boest1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.boest <- c()
    if ("boest" %in% organisms.list)
    {
      tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
    }
    return(tips_to_keep.boest)
  })
  
  tips_to_keep.carpa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.carpa <- c()
    if ("carpa" %in% organisms.list)
    {
      tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
    }
    return(tips_to_keep.carpa)
  })
  
  tips_to_keep.theca1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.theca <- c()
    if ("theca" %in% organisms.list)
    {
      tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
    }
    return(tips_to_keep.theca)
  })
  
  tips_to_keep.goshi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.goshi <- c()
    if ("goshi" %in% organisms.list)
    {
      tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
    }
    return(tips_to_keep.goshi)
  })
  
  tips_to_keep.gosra1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.gosra <- c()
    if ("gosra" %in% organisms.list)
    {
      tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
    }
    return(tips_to_keep.gosra)
  })
  
  tips_to_keep.citsi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.citsi <- c()
    if ("citsi" %in% organisms.list)
    {
      tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
    }
    return(tips_to_keep.citsi)
  })
  
  tips_to_keep.pontr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.pontr <- c()
    if ("pontr" %in% organisms.list)
    {
      tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
    }
    return(tips_to_keep.pontr)
  })
  
  tips_to_keep.eucgr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.eucgr <- c()
    if ("eucgr" %in% organisms.list)
    {
      tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
    }
    return(tips_to_keep.eucgr)
  })
  
  tips_to_keep.maldo1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.maldo <- c()
    if ("maldo" %in% organisms.list)
    {
      tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
    }
    return(tips_to_keep.maldo)
  })
  
  tips_to_keep.prupe1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.prupe <- c()
    if ("prupe" %in% organisms.list)
    {
      tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
    }
    return(tips_to_keep.prupe)
  })
  
  tips_to_keep.frave1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.frave <- c()
    if ("frave" %in% organisms.list)
    {
      tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
    }
    return(tips_to_keep.frave)
  })
  
  tips_to_keep.corav1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.corav <- c()
    if ("corav" %in% organisms.list)
    {
      tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
    }
    return(tips_to_keep.corav)
  })
  
  tips_to_keep.caril1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.caril <- c()
    if ("caril" %in% organisms.list)
    {
      tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
    }
    return(tips_to_keep.caril)
  })
  
  tips_to_keep.queru1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.queru <- c()
    if ("queru" %in% organisms.list)
    {
      tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
    }
    return(tips_to_keep.queru)
  })
  
  tips_to_keep.cucsa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cucsa <- c()
    if ("cucsa" %in% organisms.list)
    {
      tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
    }
    return(tips_to_keep.cucsa)
  })
  
  tips_to_keep.glyma1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.glyma <- c()
    if ("glyma" %in% organisms.list)
    {
      tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
    }
    return(tips_to_keep.glyma)
  })
  
  tips_to_keep.medtr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.medtr <- c()
    if ("medtr" %in% organisms.list)
    {
      tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
    }
    return(tips_to_keep.medtr)
  })
  
  tips_to_keep.tripr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.tripr <- c()
    if ("tripr" %in% organisms.list)
    {
      tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
    }
    return(tips_to_keep.tripr)
  })
  
  tips_to_keep.cicar1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cicar <- c()
    if ("cicar" %in% organisms.list)
    {
      tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
    }
    return(tips_to_keep.cicar)
  })
  
  tips_to_keep.lotja1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.lotja <- c()
    if ("lotja" %in% organisms.list)
    {
      tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
    }
    return(tips_to_keep.lotja)
  })
  
  tips_to_keep.phaac1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.phaac <- c()
    if ("phaac" %in% organisms.list)
    {
      tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
    }
    return(tips_to_keep.phaac)
  })
  
  tips_to_keep.vigun1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.vigun <- c()
    if ("vigun" %in% organisms.list)
    {
      tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
    }
    return(tips_to_keep.vigun)
  })
  
  tips_to_keep.lupal1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.lupal <- c()
    if ("lupal" %in% organisms.list)
    {
      tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
    }
    return(tips_to_keep.lupal)
  })
  
  tips_to_keep.lencu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.lencu <- c()
    if ("lencu" %in% organisms.list)
    {
      tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
    }
    return(tips_to_keep.lencu)
  })
  
  tips_to_keep.arahy1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.arahy <- c()
    if ("arahy" %in% organisms.list)
    {
      tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
    }
    return(tips_to_keep.arahy)
  })
  
  tips_to_keep.poptr1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.poptr <- c()
    if ("poptr" %in% organisms.list)
    {
      tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
    }
    return(tips_to_keep.poptr)
  })
  
  tips_to_keep.manes1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.manes <- c()
    if ("manes" %in% organisms.list)
    {
      tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
    }
    return(tips_to_keep.manes)
  })
  
  tips_to_keep.salpu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.salpu <- c()
    if ("salpu" %in% organisms.list)
    {
      tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
    }
    return(tips_to_keep.salpu)
  })
  
  tips_to_keep.linus1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.linus <- c()
    if ("linus" %in% organisms.list)
    {
      tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
    }
    return(tips_to_keep.linus)
  })
  
  tips_to_keep.corci1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.corci <- c()
    if ("corci" %in% organisms.list)
    {
      tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
    }
    return(tips_to_keep.corci)
  })
  
  tips_to_keep.vitvi1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.vitvi <- c()
    if ("vitvi" %in% organisms.list)
    {
      tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
    }
    return(tips_to_keep.vitvi)
  })
  
  tips_to_keep.kalfe1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.kalfe <- c()
    if ("kalfe" %in% organisms.list)
    {
      tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
    }
    return(tips_to_keep.kalfe)
  })
  
  tips_to_keep.vacda1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.vacda <- c()
    if ("vacda" %in% organisms.list)
    {
      tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
    }
    return(tips_to_keep.vacda)
  })
  
  tips_to_keep.oleeu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.oleeu <- c()
    if ("oleeu" %in% organisms.list)
    {
      tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
    }
    return(tips_to_keep.oleeu)
  })
  
  tips_to_keep.mimgu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.mimgu <- c()
    if ("mimgu" %in% organisms.list)
    {
      tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
    }
    return(tips_to_keep.mimgu)
  })
  
  tips_to_keep.soltu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.soltu <- c()
    if ("soltu" %in% organisms.list)
    {
      tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
    }
    return(tips_to_keep.soltu)
  })
  
  tips_to_keep.cofar1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.cofar <- c()
    if ("cofar" %in% organisms.list)
    {
      tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
    }
    return(tips_to_keep.cofar)
  })
  
  tips_to_keep.lacsa1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.lacsa <- c()
    if ("lacsa" %in% organisms.list)
    {
      tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
    }
    return(tips_to_keep.lacsa)
  })
  
  tips_to_keep.dauca1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.dauca <- c()
    if ("dauca" %in% organisms.list)
    {
      tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
    }
    return(tips_to_keep.dauca)
  })
  
  tips_to_keep.betvu1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.betvu <- c()
    if ("betvu" %in% organisms.list)
    {
      tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
    }
    return(tips_to_keep.betvu)
  })
  
  tips_to_keep.amahy1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.amahy <- c()
    if ("amahy" %in% organisms.list)
    {
      tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
    }
    return(tips_to_keep.amahy)
  })
  
  tips_to_keep.chequ1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.chequ <- c()
    if ("chequ" %in% organisms.list)
    {
      tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
    }
    return(tips_to_keep.chequ)
  })
  
  tips_to_keep.sapof1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.sapof <- c()
    if ("sapof" %in% organisms.list)
    {
      tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
    }
    return(tips_to_keep.sapof)
  })
  
  tips_to_keep.aquco1 <- reactive({
    
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    tips_to_keep.aquco <- c()
    if ("aquco" %in% organisms.list)
    {
      tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
    }
    return(tips_to_keep.aquco)
  }) %>% bindEvent(input$run_button1)
  
  
  # Create complete gene tree with the proper name for each gene
  # For this, we split the species name apart from the gene name
  tree_adj1 <- reactive({
    tree <- tree1()
    organisms.list <- c(selected_values_org1())
    
    if ("mp" %in% organisms.list)
    {
      tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label) 
      if (length(tips_to_keep.mp) != 0)
      {
        mp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mp]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.mp] <- mp.v
      }
    }
    
    if ("ot" %in% organisms.list)
    {
      tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label) 
      if (length(tips_to_keep.ot) != 0)
      {
        ost.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ot]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ot] <- ost.v
      }
    }
    
    if ("at" %in% organisms.list)
    {
      tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label) 
      if (length(tips_to_keep.at) != 0)
      {
        arabi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.at]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.at] <- arabi.v
      }
    }
    
    if ("cp" %in% organisms.list)
    {
      tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label) 
      if (length(tips_to_keep.cp) != 0)
      {
        cer.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cp] <- cer.v
      }
    }
    
    if ("cr" %in% organisms.list)
    {
      tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
      if (length(tips_to_keep.cr) != 0)
      {
        chlamy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cr]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cr] <- chlamy.v
      }
    }
    
    if ("cz" %in% organisms.list)
    {
      tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label) 
      if (length(tips_to_keep.cz) != 0)
      {
        chromo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cz]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cz] <- chromo.v
      }
    }
    
    if ("kn" %in% organisms.list)
    {
      tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label) 
      if (length(tips_to_keep.kn) != 0)
      {
        klebs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kn]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.kn] <- klebs.v
      }
    }
    
    if ("me" %in% organisms.list)
    {
      tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label) 
      if (length(tips_to_keep.me) != 0)
      {
        meso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.me]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.me] <- meso.v
      }
    }
    
    if ("mi" %in% organisms.list)
    {
      tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label) 
      if (length(tips_to_keep.mi) != 0)
      {
        micro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mi]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.mi] <- micro.v
      }
    }
    
    if ("pp" %in% organisms.list)
    {
      tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label) 
      if (length(tips_to_keep.pp) != 0)
      {
        phys.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pp]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.pp] <- phys.v
      }
    }
    
    if ("sl" %in% organisms.list)
    {
      tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label) 
      if (length(tips_to_keep.sl) != 0)
      {
        sola.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sl]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sl] <- sola.v
      }
    }
    
    if ("sm" %in% organisms.list)
    {
      tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label) 
      if (length(tips_to_keep.sm) != 0)
      {
        sel.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sm]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sm] <- sel.v
      }
    }
    
    if ("sp" %in% organisms.list)
    {
      tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label) 
      if (length(tips_to_keep.sp) != 0)
      {
        spiro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sp]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sp] <- spiro.v
      }
    }
    
    if ("ta" %in% organisms.list)
    {
      tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label) 
      if (length(tips_to_keep.ta) != 0)
      {
        tri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ta]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ta] <- tri.v
      }
    }
    
    if ("vc" %in% organisms.list)
    {
      tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
      if (length(tips_to_keep.vc) != 0)
      {
        vc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vc]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.vc] <- vc.v
      }
    }
    
    if ("bp" %in% organisms.list)
    {
      tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
      if (length(tips_to_keep.bp) != 0)
      {
        bp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bp]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.bp] <- bp.v
      }
    }
    
    if ("cri" %in% organisms.list)
    {
      tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
      if (length(tips_to_keep.cri) != 0)
      {
        cri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cri]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cri] <- cri.v
      }
    }
    
    if ("ds" %in% organisms.list)
    {
      tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
      if (length(tips_to_keep.ds) != 0)
      {
        ds.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ds]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ds] <- ds.v
      }
    }
    
    if ("os" %in% organisms.list)
    {
      tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
      if (length(tips_to_keep.os) != 0)
      {
        os.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.os]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.os] <- os.v
      }
    }
    
    if ("smag" %in% organisms.list)
    {
      tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
      if (length(tips_to_keep.smag) != 0)
      {
        smag.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.smag]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.smag] <- smag.v
      }
    }
    
    if ("tp" %in% organisms.list)
    {
      tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
      if (length(tips_to_keep.tp) != 0)
      {
        tp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tp]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.tp] <- tp.v
      }
    }
    
    if ("aa" %in% organisms.list)
    {
      tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
      if (length(tips_to_keep.aa) != 0)
      {
        aa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aa]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.aa] <- aa.v
      }
    }
    
    if ("um" %in% organisms.list)
    {
      tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
      if (length(tips_to_keep.um) != 0)
      {
        um.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.um]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.um] <- um.v
      }
    }
    
    if ("rs" %in% organisms.list)
    {
      tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
      if (length(tips_to_keep.rs) != 0)
      {
        rs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.rs]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.rs] <- rs.v
      }
    }
    
    if ("cyc" %in% organisms.list)
    {
      tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
      if (length(tips_to_keep.cyc) != 0)
      {
        cyc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cyc]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cyc] <- cyc.v
      }
    }
    
    if ("ca" %in% organisms.list)
    {
      tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
      if (length(tips_to_keep.ca) != 0)
      {
        ca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ca]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ca] <- ca.v
      }
    }
    
    if ("mv" %in% organisms.list)
    {
      tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
      if (length(tips_to_keep.mv) != 0)
      {
        mv.vec1 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mv]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.mv] <- mv.vec1
      }
    }
    
    if ("af" %in% organisms.list)
    {
      tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
      if (length(tips_to_keep.af) != 0)
      {
        af.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.af]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.af] <- af.v
      }
    }
    
    if ("sc" %in% organisms.list)
    {
      tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
      if (length(tips_to_keep.sc) != 0)
      {
        sc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sc]), "_"), 
                       function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sc] <- sc.v
      }
    }
    
    if ("aegi" %in% organisms.list)
    {
      tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
      if (length(tips_to_keep.aegi) != 0)
      {
        aegi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aegi]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.aegi] <- aegi.v
      }
    }
    
    if ("sb" %in% organisms.list)
    {
      tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
      if (length(tips_to_keep.sb) != 0)
      {
        sb.vec1 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sb]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sb] <- sb.vec1
      }
    }
    
    if ("chara" %in% organisms.list)
    {
      tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
      if (length(tips_to_keep.chara) != 0)
      {
        chara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chara]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.chara] <- chara.v
      }
    }
    
    if ("sceobli" %in% organisms.list)
    {
      tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
      if (length(tips_to_keep.sceobli) != 0)
      {
        sceobli.vec1 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sceobli]), "_"), 
                               function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sceobli] <- sceobli.vec1
      }
    }
    
    if ("cocco" %in% organisms.list)
    {
      tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
      if (length(tips_to_keep.cocco) != 0)
      {
        cocco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cocco]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cocco] <- cocco.v
      }
    }
    
    if ("haema" %in% organisms.list)
    {
      tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
      if (length(tips_to_keep.haema) != 0)
      {
        haema.vec1 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.haema]), "_"), 
                             function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.haema] <- haema.vec1
      }
    }
    
    if ("zm" %in% organisms.list)
    {
      tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
      if (length(tips_to_keep.zm) != 0)
      {
        zm.vec1 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zm]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.zm] <- zm.vec1
      }
    }
    
    
    if ("praco" %in% organisms.list)
    {
      tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
      if (length(tips_to_keep.praco) != 0)
      {
        praco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.praco]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.praco] <- praco.v
      }
    }
    
    if ("ostlu" %in% organisms.list)
    {
      tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
      if (length(tips_to_keep.ostlu) != 0)
      {
        ostlu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostlu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ostlu] <- ostlu.v
      }
    }
    
    if ("ostme" %in% organisms.list)
    {
      tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
      if (length(tips_to_keep.ostme) != 0)
      {
        ostme.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostme]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ostme] <- ostme.v
      }
    }
    
    if ("micco" %in% organisms.list)
    {
      tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
      if (length(tips_to_keep.micco) != 0)
      {
        micco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.micco]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.micco] <- micco.v
      }
    }
    
    if ("cymte" %in% organisms.list)
    {
      tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
      if (length(tips_to_keep.cymte) != 0)
      {
        cymte.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cymte]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cymte] <- cymte.v
      }
    }
    
    if ("chlin" %in% organisms.list)
    {
      tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
      if (length(tips_to_keep.chlin) != 0)
      {
        chlin.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlin]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.chlin] <- chlin.v
      }
    }
    
    if ("gonpe" %in% organisms.list)
    {
      tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
      if (length(tips_to_keep.gonpe) != 0)
      {
        gonpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gonpe]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.gonpe] <- gonpe.v
      }
    }
    
    if ("tetso" %in% organisms.list)
    {
      tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
      if (length(tips_to_keep.tetso) != 0)
      {
        tetso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetso]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.tetso] <- tetso.v
      }
    }
    
    if ("desar" %in% organisms.list)
    {
      tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
      if (length(tips_to_keep.desar) != 0)
      {
        desar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.desar]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.desar] <- desar.v
      }
    }
    
    if ("monmi" %in% organisms.list)
    {
      tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
      if (length(tips_to_keep.monmi) != 0)
      {
        monmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.monmi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.monmi] <- monmi.v
      }
    }
    
    if ("caule" %in% organisms.list)
    {
      tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
      if (length(tips_to_keep.caule) != 0)
      {
        caule.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caule]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.caule] <- caule.v
      }
    }
    
    if ("ostqu" %in% organisms.list)
    {
      tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
      if (length(tips_to_keep.ostqu) != 0)
      {
        ostqu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostqu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ostqu] <- ostqu.v
      }
    }
    
    if ("brysp" %in% organisms.list)
    {
      tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
      if (length(tips_to_keep.brysp) != 0)
      {
        brysp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brysp]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.brysp] <- brysp.v
      }
    }
    
    if ("ellbi" %in% organisms.list)
    {
      tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
      if (length(tips_to_keep.ellbi) != 0)
      {
        ellbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ellbi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ellbi] <- ellbi.v
      }
    }
    
    if ("myrbi" %in% organisms.list)
    {
      tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
      if (length(tips_to_keep.myrbi) != 0)
      {
        myrbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.myrbi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.myrbi] <- myrbi.v
      }
    }
    
    if ("apalo" %in% organisms.list)
    {
      tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
      if (length(tips_to_keep.apalo) != 0)
      {
        apalo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.apalo]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.apalo] <- apalo.v
      }
    }
    
    if ("astgl" %in% organisms.list)
    {
      tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
      if (length(tips_to_keep.astgl) != 0)
      {
        astgl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.astgl]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.astgl] <- astgl.v
      }
    }
    
    if ("tresp" %in% organisms.list)
    {
      tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
      if (length(tips_to_keep.tresp) != 0)
      {
        tresp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tresp]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.tresp] <- tresp.v
      }
    }
    
    if ("picso" %in% organisms.list)
    {
      tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
      if (length(tips_to_keep.picso) != 0)
      {
        picso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picso]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.picso] <- picso.v
      }
    }
    
    if ("chlso" %in% organisms.list)
    {
      tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
      if (length(tips_to_keep.chlso) != 0)
      {
        chlso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlso]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.chlso] <- chlso.v
      }
    }
    
    if ("auxpr" %in% organisms.list)
    {
      tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
      if (length(tips_to_keep.auxpr) != 0)
      {
        auxpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.auxpr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.auxpr] <- auxpr.v
      }
    }
    
    if ("helsp" %in% organisms.list)
    {
      tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
      if (length(tips_to_keep.helsp) != 0)
      {
        helsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.helsp]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.helsp] <- helsp.v
      }
    }
    
    if ("tetst" %in% organisms.list)
    {
      tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
      if (length(tips_to_keep.tetst) != 0)
      {
        tetst.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetst]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.tetst] <- tetst.v
      }
    }
    
    if ("pedsp" %in% organisms.list)
    {
      tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
      if (length(tips_to_keep.pedsp) != 0)
      {
        pedsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pedsp]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.pedsp] <- pedsp.v
      }
    }
    
    if ("chlpr" %in% organisms.list)
    {
      tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
      if (length(tips_to_keep.chlpr) != 0)
      {
        chlpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlpr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.chlpr] <- chlpr.v
      }
    }
    
    if ("picsp" %in% organisms.list)
    {
      tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
      if (length(tips_to_keep.picsp) != 0)
      {
        picsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picsp]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.picsp] <- picsp.v
      }
    }
    
    if ("mesvb" %in% organisms.list)
    {
      tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
      if (length(tips_to_keep.mesvb) != 0)
      {
        mesvb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mesvb]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.mesvb] <- mesvb.v
      }
    }
    
    if ("meskr" %in% organisms.list)
    {
      tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
      if (length(tips_to_keep.meskr) != 0)
      {
        meskr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.meskr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.meskr] <- meskr.v
      }
    }
    
    if ("zygci" %in% organisms.list)
    {
      tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
      if (length(tips_to_keep.zygci) != 0)
      {
        zygci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygci]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.zygci] <- zygci.v
      }
    }
    
    if ("zygcb" %in% organisms.list)
    {
      tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
      if (length(tips_to_keep.zygcb) != 0)
      {
        zygcb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcb]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.zygcb] <- zygcb.v
      }
    }
    
    if ("zygcy" %in% organisms.list)
    {
      tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
      if (length(tips_to_keep.zygcy) != 0)
      {
        zygcy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcy]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.zygcy] <- zygcy.v
      }
    }
    
    if ("plesc" %in% organisms.list)
    {
      tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
      if (length(tips_to_keep.plesc) != 0)
      {
        plesc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.plesc]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.plesc] <- plesc.v
      }
    }
    
    if ("funhy" %in% organisms.list)
    {
      tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
      if (length(tips_to_keep.funhy) != 0)
      {
        funhy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.funhy]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.funhy] <- funhy.v
      }
    }
    
    if ("sphfa" %in% organisms.list)
    {
      tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
      if (length(tips_to_keep.sphfa) != 0)
      {
        sphfa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sphfa]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sphfa] <- sphfa.v
      }
    }
    
    if ("takle" %in% organisms.list)
    {
      tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
      if (length(tips_to_keep.takle) != 0)
      {
        takle.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.takle]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.takle] <- takle.v
      }
    }
    
    if ("marpa" %in% organisms.list)
    {
      tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
      if (length(tips_to_keep.marpa) != 0)
      {
        marpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marpa]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.marpa] <- marpa.v
      }
    }
    
    if ("ricfl" %in% organisms.list)
    {
      tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
      if (length(tips_to_keep.ricfl) != 0)
      {
        ricfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricfl]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ricfl] <- ricfl.v
      }
    }
    
    if ("ricso" %in% organisms.list)
    {
      tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
      if (length(tips_to_keep.ricso) != 0)
      {
        ricso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricso]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ricso] <- ricso.v
      }
    }
    
    if ("antpu" %in% organisms.list)
    {
      tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
      if (length(tips_to_keep.antpu) != 0)
      {
        antpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antpu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.antpu] <- antpu.v
      }
    }
    
    if ("antfu" %in% organisms.list)
    {
      tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
      if (length(tips_to_keep.antfu) != 0)
      {
        antfu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antfu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.antfu] <- antfu.v
      }
    }
    
    if ("megfl" %in% organisms.list)
    {
      tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
      if (length(tips_to_keep.megfl) != 0)
      {
        megfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.megfl]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.megfl] <- megfl.v
      }
    }
    
    if ("phach" %in% organisms.list)
    {
      tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
      if (length(tips_to_keep.phach) != 0)
      {
        phach.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phach]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.phach] <- phach.v
      }
    }
    
    if ("phyph" %in% organisms.list)
    {
      tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
      if (length(tips_to_keep.phyph) != 0)
      {
        phyph.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phyph]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.phyph] <- phyph.v
      }
    }
    
    if ("parpe" %in% organisms.list)
    {
      tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
      if (length(tips_to_keep.parpe) != 0)
      {
        parpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.parpe]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.parpe] <- parpe.v
      }
    }
    
    if ("notor" %in% organisms.list)
    {
      tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
      if (length(tips_to_keep.notor) != 0)
      {
        notor.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.notor]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.notor] <- notor.v
      }
    }
    
    if ("phaca" %in% organisms.list)
    {
      tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
      if (length(tips_to_keep.phaca) != 0)
      {
        phaca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaca]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.phaca] <- phaca.v
      }
    }
    
    if ("phasp" %in% organisms.list)
    {
      tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
      if (length(tips_to_keep.phasp) != 0)
      {
        phasp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phasp]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.phasp] <- phasp.v
      }
    }
    
    if ("leidu" %in% organisms.list)
    {
      tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
      if (length(tips_to_keep.leidu) != 0)
      {
        leidu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.leidu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.leidu] <- leidu.v
      }
    }
    
    if ("isosi" %in% organisms.list)
    {
      tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
      if (length(tips_to_keep.isosi) != 0)
      {
        isosi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.isosi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.isosi] <- isosi.v
      }
    }
    
    if ("dipco" %in% organisms.list)
    {
      tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
      if (length(tips_to_keep.dipco) != 0)
      {
        dipco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dipco]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.dipco] <- dipco.v
      }
    }
    
    if ("hupas" %in% organisms.list)
    {
      tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
      if (length(tips_to_keep.hupas) != 0)
      {
        hupas.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.hupas]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.hupas] <- hupas.v
      }
    }
    
    if ("lyccl" %in% organisms.list)
    {
      tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
      if (length(tips_to_keep.lyccl) != 0)
      {
        lyccl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lyccl]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.lyccl] <- lyccl.v
      }
    }
    
    if ("adica" %in% organisms.list)
    {
      tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
      if (length(tips_to_keep.adica) != 0)
      {
        adica.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.adica]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.adica] <- adica.v
      }
    }
    
    if ("alssp" %in% organisms.list)
    {
      tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
      if (length(tips_to_keep.alssp) != 0)
      {
        alssp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.alssp]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.alssp] <- alssp.v
      }
    }
    
    if ("marve" %in% organisms.list)
    {
      tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
      if (length(tips_to_keep.marve) != 0)
      {
        marve.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marve]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.marve] <- marve.v
      }
    }
    
    if ("seqse" %in% organisms.list)
    {
      tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
      if (length(tips_to_keep.seqse) != 0)
      {
        seqse.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqse]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.seqse] <- seqse.v
      }
    }
    
    if ("seqgi" %in% organisms.list)
    {
      tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
      if (length(tips_to_keep.seqgi) != 0)
      {
        seqgi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqgi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.seqgi] <- seqgi.v
      }
    }
    
    if ("taxch" %in% organisms.list)
    {
      tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
      if (length(tips_to_keep.taxch) != 0)
      {
        taxch.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.taxch]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.taxch] <- taxch.v
      }
    }
    
    if ("abial" %in% organisms.list)
    {
      tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
      if (length(tips_to_keep.abial) != 0)
      {
        abial.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.abial]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.abial] <- abial.v
      }
    }
    
    if ("picab" %in% organisms.list)
    {
      tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
      if (length(tips_to_keep.picab) != 0)
      {
        picab.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picab]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.picab] <- picab.v
      }
    }
    
    if ("gnemo" %in% organisms.list)
    {
      tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
      if (length(tips_to_keep.gnemo) != 0)
      {
        gnemo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gnemo]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.gnemo] <- gnemo.v
      }
    }
    
    if ("welmi" %in% organisms.list)
    {
      tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
      if (length(tips_to_keep.welmi) != 0)
      {
        welmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.welmi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.welmi] <- welmi.v
      }
    }
    
    if ("ginbi" %in% organisms.list)
    {
      tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
      if (length(tips_to_keep.ginbi) != 0)
      {
        ginbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ginbi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ginbi] <- ginbi.v
      }
    }
    
    if ("ambtr" %in% organisms.list)
    {
      tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
      if (length(tips_to_keep.ambtr) != 0)
      {
        ambtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ambtr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.ambtr] <- ambtr.v
      }
    }
    
    if ("nymco" %in% organisms.list)
    {
      tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
      if (length(tips_to_keep.nymco) != 0)
      {
        nymco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.nymco]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.nymco] <- nymco.v
      }
    }
    
    if ("cinka" %in% organisms.list)
    {
      tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
      if (length(tips_to_keep.cinka) != 0)
      {
        cinka.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cinka]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cinka] <- cinka.v
      }
    }
    
    
    if ("horvu" %in% organisms.list)
    {
      tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
      if (length(tips_to_keep.horvu) != 0)
      {
        horvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.horvu]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.horvu] <- horvu.v
      }
    }
    
    if ("bradi" %in% organisms.list)
    {
      tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
      if (length(tips_to_keep.bradi) != 0)
      {
        bradi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bradi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.bradi] <- bradi.v
      }
    }
    
    if ("setit" %in% organisms.list)
    {
      tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
      if (length(tips_to_keep.setit) != 0)
      {
        setit.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.setit]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.setit] <- setit.v
      }
    }
    
    if ("panha" %in% organisms.list)
    {
      tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
      if (length(tips_to_keep.panha) != 0)
      {
        panha.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.panha]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.panha] <- panha.v
      }
    }
    
    if ("missi" %in% organisms.list)
    {
      tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
      if (length(tips_to_keep.missi) != 0)
      {
        missi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.missi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.missi] <- missi.v
      }
    }
    
    if ("oroth" %in% organisms.list)
    {
      tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
      if (length(tips_to_keep.oroth) != 0)
      {
        oroth.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oroth]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.oroth] <- oroth.v
      }
    }
    
    if ("eleco" %in% organisms.list)
    {
      tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
      if (length(tips_to_keep.eleco) != 0)
      {
        eleco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eleco]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.eleco] <- eleco.v
      }
    }
    
    if ("anaco" %in% organisms.list)
    {
      tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
      if (length(tips_to_keep.anaco) != 0)
      {
        anaco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.anaco]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.anaco] <- anaco.v
      }
    }
    
    if ("musac" %in% organisms.list)
    {
      tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
      if (length(tips_to_keep.musac) != 0)
      {
        musac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.musac]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.musac] <- musac.v
      }
    }
    
    if ("dioal" %in% organisms.list)
    {
      tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
      if (length(tips_to_keep.dioal) != 0)
      {
        dioal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dioal]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.dioal] <- dioal.v
      }
    }
    
    if ("spipo" %in% organisms.list)
    {
      tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
      if (length(tips_to_keep.spipo) != 0)
      {
        spipo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.spipo]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.spipo] <- spipo.v
      }
    }
    
    if ("aspof" %in% organisms.list)
    {
      tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
      if (length(tips_to_keep.aspof) != 0)
      {
        aspof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aspof]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.aspof] <- aspof.v
      }
    }
    
    if ("zosma" %in% organisms.list)
    {
      tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
      if (length(tips_to_keep.zosma) != 0)
      {
        zosma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zosma]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.zosma] <- zosma.v
      }
    }
    
    if ("araly" %in% organisms.list)
    {
      tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
      if (length(tips_to_keep.araly) != 0)
      {
        araly.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.araly]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.araly] <- araly.v
      }
    }
    
    if ("capgr" %in% organisms.list)
    {
      tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
      if (length(tips_to_keep.capgr) != 0)
      {
        capgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.capgr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.capgr] <- capgr.v
      }
    }
    
    if ("brara" %in% organisms.list)
    {
      tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
      if (length(tips_to_keep.brara) != 0)
      {
        brara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brara]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.brara] <- brara.v
      }
    }
    
    if ("braol" %in% organisms.list)
    {
      tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
      if (length(tips_to_keep.braol) != 0)
      {
        braol.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.braol]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.braol] <- braol.v
      }
    }
    
    if ("sisir" %in% organisms.list)
    {
      tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
      if (length(tips_to_keep.sisir) != 0)
      {
        sisir.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sisir]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sisir] <- sisir.v
      }
    }
    
    if ("schpa" %in% organisms.list)
    {
      tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
      if (length(tips_to_keep.schpa) != 0)
      {
        schpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.schpa]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.schpa] <- schpa.v
      }
    }
    
    if ("eutsa" %in% organisms.list)
    {
      tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
      if (length(tips_to_keep.eutsa) != 0)
      {
        eutsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eutsa]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.eutsa] <- eutsa.v
      }
    }
    
    if ("thlar" %in% organisms.list)
    {
      tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
      if (length(tips_to_keep.thlar) != 0)
      {
        thlar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.thlar]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.thlar] <- thlar.v
      }
    }
    
    if ("boest" %in% organisms.list)
    {
      tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
      if (length(tips_to_keep.boest) != 0)
      {
        boest.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.boest]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.boest] <- boest.v
      }
    }
    
    if ("carpa" %in% organisms.list)
    {
      tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
      if (length(tips_to_keep.carpa) != 0)
      {
        carpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.carpa]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.carpa] <- carpa.v
      }
    }
    
    if ("theca" %in% organisms.list)
    {
      tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
      if (length(tips_to_keep.theca) != 0)
      {
        theca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.theca]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.theca] <- theca.v
      }
    }
    
    if ("goshi" %in% organisms.list)
    {
      tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
      if (length(tips_to_keep.goshi) != 0)
      {
        goshi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.goshi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.goshi] <- goshi.v
      }
    }
    
    if ("gosra" %in% organisms.list)
    {
      tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
      if (length(tips_to_keep.gosra) != 0)
      {
        gosra.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gosra]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.gosra] <- gosra.v
      }
    }
    
    if ("citsi" %in% organisms.list)
    {
      tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
      if (length(tips_to_keep.citsi) != 0)
      {
        citsi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.citsi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.citsi] <- citsi.v
      }
    }
    
    if ("pontr" %in% organisms.list)
    {
      tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
      if (length(tips_to_keep.pontr) != 0)
      {
        pontr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pontr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.pontr] <- pontr.v
      }
    }
    
    if ("eucgr" %in% organisms.list)
    {
      tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
      if (length(tips_to_keep.eucgr) != 0)
      {
        eucgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eucgr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.eucgr] <- eucgr.v
      }
    }
    
    if ("maldo" %in% organisms.list)
    {
      tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
      if (length(tips_to_keep.maldo) != 0)
      {
        maldo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.maldo]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.maldo] <- maldo.v
      }
    }
    
    if ("prupe" %in% organisms.list)
    {
      tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
      if (length(tips_to_keep.prupe) != 0)
      {
        prupe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.prupe]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.prupe] <- prupe.v
      }
    }
    
    if ("frave" %in% organisms.list)
    {
      tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
      if (length(tips_to_keep.frave) != 0)
      {
        frave.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.frave]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.frave] <- frave.v
      }
    }
    
    if ("corav" %in% organisms.list)
    {
      tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
      if (length(tips_to_keep.corav) != 0)
      {
        corav.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corav]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.corav] <- corav.v
      }
    }
    
    if ("caril" %in% organisms.list)
    {
      tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
      if (length(tips_to_keep.caril) != 0)
      {
        caril.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caril]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.caril] <- caril.v
      }
    }
    
    if ("queru" %in% organisms.list)
    {
      tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
      if (length(tips_to_keep.queru) != 0)
      {
        queru.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.queru]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.queru] <- queru.v
      }
    }
    
    if ("cucsa" %in% organisms.list)
    {
      tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
      if (length(tips_to_keep.cucsa) != 0)
      {
        cucsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cucsa]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cucsa] <- cucsa.v
      }
    }
    
    if ("glyma" %in% organisms.list)
    {
      tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
      if (length(tips_to_keep.glyma) != 0)
      {
        glyma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.glyma]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.glyma] <- glyma.v
      }
    }
    
    if ("medtr" %in% organisms.list)
    {
      tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
      if (length(tips_to_keep.medtr) != 0)
      {
        medtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.medtr]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.medtr] <- medtr.v
      }
    }
    
    if ("tripr" %in% organisms.list)
    {
      tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
      if (length(tips_to_keep.tripr) != 0)
      {
        tripr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tripr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.tripr] <- tripr.v
      }
    }
    
    if ("cicar" %in% organisms.list)
    {
      tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
      if (length(tips_to_keep.cicar) != 0)
      {
        cicar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cicar]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cicar] <- cicar.v
      }
    }
    
    if ("lotja" %in% organisms.list)
    {
      tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
      if (length(tips_to_keep.lotja) != 0)
      {
        lotja.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lotja]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.lotja] <- lotja.v
      }
    }
    
    if ("phaac" %in% organisms.list)
    {
      tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
      if (length(tips_to_keep.phaac) != 0)
      {
        phaac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaac]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.phaac] <- phaac.v
      }
    }
    
    if ("vigun" %in% organisms.list)
    {
      tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
      if (length(tips_to_keep.vigun) != 0)
      {
        vigun.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vigun]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.vigun] <- vigun.v
      }
    }
    
    if ("lupal" %in% organisms.list)
    {
      tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
      if (length(tips_to_keep.lupal) != 0)
      {
        lupal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lupal]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.lupal] <- lupal.v
      }
    }
    
    if ("lencu" %in% organisms.list)
    {
      tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
      if (length(tips_to_keep.lencu) != 0)
      {
        lencu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lencu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.lencu] <- lencu.v
      }
    }
    
    if ("arahy" %in% organisms.list)
    {
      tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
      if (length(tips_to_keep.arahy) != 0)
      {
        arahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.arahy]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.arahy] <- arahy.v
      }
    }
    
    if ("poptr" %in% organisms.list)
    {
      tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
      if (length(tips_to_keep.poptr) != 0)
      {
        poptr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.poptr]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.poptr] <- poptr.v
      }
    }
    
    if ("manes" %in% organisms.list)
    {
      tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
      if (length(tips_to_keep.manes) != 0)
      {
        manes.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.manes]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.manes] <- manes.v
      }
    }
    
    if ("salpu" %in% organisms.list)
    {
      tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
      if (length(tips_to_keep.salpu) != 0)
      {
        salpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.salpu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.salpu] <- salpu.v
      }
    }
    
    if ("linus" %in% organisms.list)
    {
      tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
      if (length(tips_to_keep.linus) != 0)
      {
        linus.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.linus]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.linus] <- linus.v
      }
    }
    
    if ("corci" %in% organisms.list)
    {
      tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
      if (length(tips_to_keep.corci) != 0)
      {
        corci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corci]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.corci] <- corci.v
      }
    }
    
    if ("vitvi" %in% organisms.list)
    {
      tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
      if (length(tips_to_keep.vitvi) != 0)
      {
        vitvi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vitvi]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.vitvi] <- vitvi.v
      }
    }
    
    if ("kalfe" %in% organisms.list)
    {
      tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
      if (length(tips_to_keep.kalfe) != 0)
      {
        kalfe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kalfe]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.kalfe] <- kalfe.v
      }
    }
    
    if ("vacda" %in% organisms.list)
    {
      tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
      if (length(tips_to_keep.vacda) != 0)
      {
        vacda.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vacda]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.vacda] <- vacda.v
      }
    }
    
    if ("oleeu" %in% organisms.list)
    {
      tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
      if (length(tips_to_keep.oleeu) != 0)
      {
        oleeu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oleeu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.oleeu] <- oleeu.v
      }
    }
    
    if ("mimgu" %in% organisms.list)
    {
      tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
      if (length(tips_to_keep.mimgu) != 0)
      {
        mimgu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mimgu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.mimgu] <- mimgu.v
      }
    }
    
    if ("soltu" %in% organisms.list)
    {
      tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
      if (length(tips_to_keep.soltu) != 0)
      {
        soltu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.soltu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.soltu] <- soltu.v
      }
    }
    
    if ("cofar" %in% organisms.list)
    {
      tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
      if (length(tips_to_keep.cofar) != 0)
      {
        cofar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cofar]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.cofar] <- cofar.v
      }
    }
    
    if ("lacsa" %in% organisms.list)
    {
      tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
      if (length(tips_to_keep.lacsa) != 0)
      {
        lacsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lacsa]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.lacsa] <- lacsa.v
      }
    }
    
    if ("dauca" %in% organisms.list)
    {
      tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
      if (length(tips_to_keep.dauca) != 0)
      {
        dauca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dauca]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.dauca] <- dauca.v
      }
    }
    
    if ("betvu" %in% organisms.list)
    {
      tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
      if (length(tips_to_keep.betvu) != 0)
      {
        betvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.betvu]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.betvu] <- betvu.v
      }
    }
    
    if ("amahy" %in% organisms.list)
    {
      tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
      if (length(tips_to_keep.amahy) != 0)
      {
        amahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.amahy]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.amahy] <- amahy.v
      }
    }
    
    if ("chequ" %in% organisms.list)
    {
      tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
      if (length(tips_to_keep.chequ) != 0)
      {
        chequ.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chequ]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.chequ] <- chequ.v
      }
    }
    
    if ("sapof" %in% organisms.list)
    {
      tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
      if (length(tips_to_keep.sapof) != 0)
      {
        sapof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sapof]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.sapof] <- sapof.v
      }
    }
    
    if ("aquco" %in% organisms.list)
    {
      tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
      if (length(tips_to_keep.aquco) != 0)
      {
        aquco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aquco]), "_"),
                          function(x) paste0(x[-c(1,2)], collapse="_"))
        tree$tip.label[tips_to_keep.aquco] <- aquco.v
      }
    }
    
    return(tree)
  }) %>% bindEvent(input$run_button1)
  
  # Generate reduced tree when the corresponding button is activated
  tree_reduced1 <- reactive({
    
    tree <- tree_adj1()
    # Define tips to keep (selected organisms) and generate the reduced tree
    tips_to_keep.global <- c(tips_to_keep.praco1(), tips_to_keep.ot1(), tips_to_keep.ostlu1(), tips_to_keep.ostme1(), tips_to_keep.bp1(),
                             tips_to_keep.mi1(), tips_to_keep.micco1(), tips_to_keep.cymte1(), tips_to_keep.cr1(), tips_to_keep.chlin1(),
                             tips_to_keep.haema1(), tips_to_keep.vc1(), tips_to_keep.ds1(), tips_to_keep.gonpe1(), tips_to_keep.tetso1(),
                             tips_to_keep.cz1(), tips_to_keep.rs1(), tips_to_keep.sceobli1(), tips_to_keep.desar1(), tips_to_keep.monmi1(),
                             tips_to_keep.caule1(), tips_to_keep.ostqu1(), tips_to_keep.brysp1(), tips_to_keep.cocco1(),
                             tips_to_keep.ellbi1(), tips_to_keep.myrbi1(), tips_to_keep.apalo1(), tips_to_keep.astgl1(),
                             tips_to_keep.tresp1(), tips_to_keep.picso1(), tips_to_keep.chlso1(), tips_to_keep.auxpr1(),
                             tips_to_keep.helsp1(), tips_to_keep.um1(), tips_to_keep.tetst1(), tips_to_keep.pedsp1(),
                             tips_to_keep.chlpr1(), tips_to_keep.picsp1(), tips_to_keep.ca1(), tips_to_keep.mv1(), tips_to_keep.mesvb1(),
                             tips_to_keep.kn1(), tips_to_keep.chara1(), tips_to_keep.me1(), tips_to_keep.meskr1(), tips_to_keep.sp1(),
                             tips_to_keep.zygci1(), tips_to_keep.zygcb1(), tips_to_keep.zygcy1(),
                             tips_to_keep.cp1(), tips_to_keep.pp1(), tips_to_keep.plesc1(), tips_to_keep.funhy1(),
                             tips_to_keep.smag1(), tips_to_keep.sphfa1(), tips_to_keep.takle1(), tips_to_keep.mp1(),
                             tips_to_keep.marpa1(), tips_to_keep.ricfl1(), tips_to_keep.ricso1(), tips_to_keep.aa1(),
                             tips_to_keep.antpu1(), tips_to_keep.antfu1(), tips_to_keep.megfl1(), tips_to_keep.phach1(),
                             tips_to_keep.phyph1(), tips_to_keep.parpe1(), tips_to_keep.notor1(), tips_to_keep.phaca1(),
                             tips_to_keep.phasp1(), tips_to_keep.leidu1(), tips_to_keep.sm1(), tips_to_keep.isosi1(), 
                             tips_to_keep.dipco1(), tips_to_keep.hupas1(), tips_to_keep.lyccl1(),
                             tips_to_keep.cri1(), tips_to_keep.adica1(), tips_to_keep.alssp1(), tips_to_keep.sc1(), 
                             tips_to_keep.af1(), tips_to_keep.marve1(), tips_to_keep.tp1(), tips_to_keep.seqse1(), 
                             tips_to_keep.seqgi1(), tips_to_keep.taxch1(), tips_to_keep.abial1(), tips_to_keep.picab1(),
                             tips_to_keep.gnemo1(), tips_to_keep.welmi1(), tips_to_keep.cyc1(), tips_to_keep.ginbi1(),
                             tips_to_keep.ambtr1(), tips_to_keep.nymco1(), tips_to_keep.cinka1(), tips_to_keep.aegi1(),
                             tips_to_keep.ta1(), tips_to_keep.horvu1(), tips_to_keep.bradi1(), tips_to_keep.os1(),
                             tips_to_keep.setit1(), tips_to_keep.sb1(), tips_to_keep.zm1(), tips_to_keep.panha1(),
                             tips_to_keep.missi1(), tips_to_keep.oroth1(), tips_to_keep.eleco1(), tips_to_keep.anaco1(),
                             tips_to_keep.musac1(), tips_to_keep.dioal1(), tips_to_keep.spipo1(), tips_to_keep.aspof1(),
                             tips_to_keep.zosma1(), tips_to_keep.at1(), tips_to_keep.araly1(), tips_to_keep.capgr1(),
                             tips_to_keep.brara1(), tips_to_keep.braol1(), tips_to_keep.sisir1(), tips_to_keep.schpa1(),
                             tips_to_keep.eutsa1(), tips_to_keep.thlar1(), tips_to_keep.boest1(), tips_to_keep.carpa1(),
                             tips_to_keep.theca1(), tips_to_keep.goshi1(), tips_to_keep.gosra1(), tips_to_keep.citsi1(),
                             tips_to_keep.pontr1(), tips_to_keep.eucgr1(), tips_to_keep.maldo1(), tips_to_keep.prupe1(),
                             tips_to_keep.frave1(), tips_to_keep.corav1(), tips_to_keep.caril1(), tips_to_keep.queru1(),
                             tips_to_keep.cucsa1(), tips_to_keep.glyma1(), tips_to_keep.medtr1(), tips_to_keep.tripr1(),
                             tips_to_keep.cicar1(), tips_to_keep.lotja1(), tips_to_keep.phaac1(), tips_to_keep.vigun1(),
                             tips_to_keep.lupal1(), tips_to_keep.lencu1(), tips_to_keep.arahy1(), tips_to_keep.poptr1(),
                             tips_to_keep.manes1(), tips_to_keep.salpu1(), tips_to_keep.linus1(), tips_to_keep.corci1(),
                             tips_to_keep.vitvi1(), tips_to_keep.kalfe1(), tips_to_keep.vacda1(), tips_to_keep.oleeu1(),
                             tips_to_keep.mimgu1(), tips_to_keep.sl1(), tips_to_keep.soltu1(), tips_to_keep.cofar1(),
                             tips_to_keep.lacsa1(), tips_to_keep.dauca1(), tips_to_keep.betvu1(), tips_to_keep.amahy1(),
                             tips_to_keep.chequ1(), tips_to_keep.sapof1(), tips_to_keep.aquco1())
    
    # Error message if trying to build tree with less than two tips
    if (length(tips_to_keep.global) < 2)
    {
      shinyjs::hideElement(id = 'loading.tree1')
      
      if (UI_exist_tree1)
      {
        removeUI(
          selector = "div:has(>> #treeTips1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>>> #presentorg1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #tree_image1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTree1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadNewick1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTreeSeqs1)",
          multiple = TRUE,
          immediate = TRUE
        )
      }
      
      UI_exist_tree1 <<- F
      output$error_tree1 <- renderUI({renderText({print("Unable to construct 
      tree with a single tip, please select more organisms.")})})
      validate(need(length(tips_to_keep.global) > 1, " "))
    }
    
      
  # Error message if query gene does not belong to selected organisms
    if (!(str_replace_all(input$geneInt1, fixed(" "), "") %in% tree$tip.label))
    {
      shinyjs::hideElement(id = 'loading.tree1')
      
      if (UI_exist_tree1)
      {
        removeUI(
          selector = "div:has(>> #treeTips1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>>> #presentorg1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #tree_image1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTree1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadNewick1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadTreeSeqs1)",
          multiple = TRUE,
          immediate = TRUE
        )
      }
      
      UI_exist_tree1 <<- F
      output$error_tree1 <- renderUI({renderText({print("Please select the species corresponding
      to the query gene. Read the instructions for clarification.")})})
      validate(need(str_replace_all(input$geneInt1, fixed(" "), "") %in% tree$tip.label, " "))
    }
    
    
    tips_to_drop <- setdiff(1:length(tree$tip.label), tips_to_keep.global)
    tree_reduced <- drop.tip(tree, tips_to_drop)
    
    return(tree_reduced)
  }) %>% bindEvent(input$run_button1)
  
  ### Select orthogroup sequences based on the reduced tree
  ortho_reduced1 <- reactive({
    
    tree_reduced <- tree_reduced1()
    ortho_seq <- ortho_seq1()
    ortho_reduced <- ortho_seq[tree_reduced$tip.label]
    return(ortho_reduced)
  }) %>% bindEvent(input$run_button1)
  
  organims_reduced1 <- reactive({
    
    tree_reduced <- tree_reduced1()
    
    len.mp <- length(tips_to_keep.mp1())
    len.ot <- length(tips_to_keep.ot1())
    len.at <- length(tips_to_keep.at1())
    len.cp <- length(tips_to_keep.cp1())
    len.cr <- length(tips_to_keep.cr1())
    len.cz <- length(tips_to_keep.cz1())
    len.kn <- length(tips_to_keep.kn1())
    len.me <- length(tips_to_keep.me1())
    len.mi <- length(tips_to_keep.mi1())
    len.pp <- length(tips_to_keep.pp1())
    len.sl <- length(tips_to_keep.sl1())
    len.sm <- length(tips_to_keep.sm1())
    len.sp <- length(tips_to_keep.sp1())
    len.ta <- length(tips_to_keep.ta1())
    len.vc <- length(tips_to_keep.vc1())
    len.bp <- length(tips_to_keep.bp1())
    len.cri <- length(tips_to_keep.cri1())
    len.ds <- length(tips_to_keep.ds1())
    len.os <- length(tips_to_keep.os1())
    len.smag <- length(tips_to_keep.smag1())
    len.tp <- length(tips_to_keep.tp1())
    len.aa <- length(tips_to_keep.aa1())
    len.um <- length(tips_to_keep.um1())
    len.rs <- length(tips_to_keep.rs1())
    len.cyc <- length(tips_to_keep.cyc1())
    len.ca <- length(tips_to_keep.ca1())
    len.mv <- length(tips_to_keep.mv1())
    len.af <- length(tips_to_keep.af1())
    len.sc <- length(tips_to_keep.sc1())
    len.aegi <- length(tips_to_keep.aegi1())
    len.sb <- length(tips_to_keep.sb1())
    len.chara <- length(tips_to_keep.chara1())
    len.sceobli <- length(tips_to_keep.sceobli1())
    len.cocco <- length(tips_to_keep.cocco1())
    len.haema <- length(tips_to_keep.haema1())
    len.zea <- length(tips_to_keep.zm1())
    len.praco <- length(tips_to_keep.praco1())
    len.ostlu <- length(tips_to_keep.ostlu1())
    len.ostme <- length(tips_to_keep.ostme1())
    len.micco <- length(tips_to_keep.micco1())
    len.cymte <- length(tips_to_keep.cymte1())
    len.chlin <- length(tips_to_keep.chlin1())
    len.gonpe <- length(tips_to_keep.gonpe1())
    len.tetso <- length(tips_to_keep.tetso1())
    len.desar <- length(tips_to_keep.desar1())
    len.monmi <- length(tips_to_keep.monmi1())
    len.caule <- length(tips_to_keep.caule1())
    len.ostqu <- length(tips_to_keep.ostqu1())
    len.brysp <- length(tips_to_keep.brysp1())
    len.ellbi <- length(tips_to_keep.ellbi1())
    len.myrbi <- length(tips_to_keep.myrbi1())
    len.apalo <- length(tips_to_keep.apalo1())
    len.astgl <- length(tips_to_keep.astgl1())
    len.tresp <- length(tips_to_keep.tresp1())
    len.picso <- length(tips_to_keep.picso1())
    len.chlso <- length(tips_to_keep.chlso1())
    len.auxpr <- length(tips_to_keep.auxpr1())
    len.helsp <- length(tips_to_keep.helsp1())
    len.tetst <- length(tips_to_keep.tetst1())
    len.pedsp <- length(tips_to_keep.pedsp1())
    len.chlpr <- length(tips_to_keep.chlpr1())
    len.picsp <- length(tips_to_keep.picsp1())
    len.mesvb <- length(tips_to_keep.mesvb1())
    len.meskr <- length(tips_to_keep.meskr1())
    len.zygci <- length(tips_to_keep.zygci1())
    len.zygcb <- length(tips_to_keep.zygcb1())
    len.zygcy <- length(tips_to_keep.zygcy1())
    len.plesc <- length(tips_to_keep.plesc1())
    len.funhy <- length(tips_to_keep.funhy1())
    len.sphfa <- length(tips_to_keep.sphfa1())
    len.takle <- length(tips_to_keep.takle1())
    len.marpa <- length(tips_to_keep.marpa1())
    len.ricfl <- length(tips_to_keep.ricfl1())
    len.ricso <- length(tips_to_keep.ricso1())
    len.antpu <- length(tips_to_keep.antpu1())
    len.antfu <- length(tips_to_keep.antfu1())
    len.megfl <- length(tips_to_keep.megfl1())
    len.phach <- length(tips_to_keep.phach1())
    len.phyph <- length(tips_to_keep.phyph1())
    len.parpe <- length(tips_to_keep.parpe1())
    len.notor <- length(tips_to_keep.notor1())
    len.phaca <- length(tips_to_keep.phaca1())
    len.phasp <- length(tips_to_keep.phasp1())
    len.leidu <- length(tips_to_keep.leidu1())
    len.isosi <- length(tips_to_keep.isosi1())
    len.dipco <- length(tips_to_keep.dipco1())
    len.hupas <- length(tips_to_keep.hupas1())
    len.lyccl <- length(tips_to_keep.lyccl1())
    len.adica <- length(tips_to_keep.adica1())
    len.alssp <- length(tips_to_keep.alssp1())
    len.marve <- length(tips_to_keep.marve1())
    len.seqse <- length(tips_to_keep.seqse1())
    len.seqgi <- length(tips_to_keep.seqgi1())
    len.taxch <- length(tips_to_keep.taxch1())
    len.abial <- length(tips_to_keep.abial1())
    len.picab <- length(tips_to_keep.picab1())
    len.gnemo <- length(tips_to_keep.gnemo1())
    len.welmi <- length(tips_to_keep.welmi1())
    len.ginbi <- length(tips_to_keep.ginbi1())
    len.ambtr <- length(tips_to_keep.ambtr1())
    len.nymco <- length(tips_to_keep.nymco1())
    len.cinka <- length(tips_to_keep.cinka1())
    len.horvu <- length(tips_to_keep.horvu1())
    len.bradi <- length(tips_to_keep.bradi1())
    len.setit <- length(tips_to_keep.setit1())
    len.panha <- length(tips_to_keep.panha1())
    len.missi <- length(tips_to_keep.missi1())
    len.oroth <- length(tips_to_keep.oroth1())
    len.eleco <- length(tips_to_keep.eleco1())
    len.anaco <- length(tips_to_keep.anaco1())
    len.musac <- length(tips_to_keep.musac1())
    len.dioal <- length(tips_to_keep.dioal1())
    len.spipo <- length(tips_to_keep.spipo1())
    len.aspof <- length(tips_to_keep.aspof1())
    len.zosma <- length(tips_to_keep.zosma1())
    len.araly <- length(tips_to_keep.araly1())
    len.capgr <- length(tips_to_keep.capgr1())
    len.brara <- length(tips_to_keep.brara1())
    len.braol <- length(tips_to_keep.braol1())
    len.sisir <- length(tips_to_keep.sisir1())
    len.schpa <- length(tips_to_keep.schpa1())
    len.eutsa <- length(tips_to_keep.eutsa1())
    len.thlar <- length(tips_to_keep.thlar1())
    len.boest <- length(tips_to_keep.boest1())
    len.carpa <- length(tips_to_keep.carpa1())
    len.theca <- length(tips_to_keep.theca1())
    len.goshi <- length(tips_to_keep.goshi1())
    len.gosra <- length(tips_to_keep.gosra1())
    len.citsi <- length(tips_to_keep.citsi1())
    len.pontr <- length(tips_to_keep.pontr1())
    len.eucgr <- length(tips_to_keep.eucgr1())
    len.maldo <- length(tips_to_keep.maldo1())
    len.prupe <- length(tips_to_keep.prupe1())
    len.frave <- length(tips_to_keep.frave1())
    len.corav <- length(tips_to_keep.corav1())
    len.caril <- length(tips_to_keep.caril1())
    len.queru <- length(tips_to_keep.queru1())
    len.cucsa <- length(tips_to_keep.cucsa1())
    len.glyma <- length(tips_to_keep.glyma1())
    len.medtr <- length(tips_to_keep.medtr1())
    len.tripr <- length(tips_to_keep.tripr1())
    len.cicar <- length(tips_to_keep.cicar1())
    len.lotja <- length(tips_to_keep.lotja1())
    len.phaac <- length(tips_to_keep.phaac1())
    len.vigun <- length(tips_to_keep.vigun1())
    len.lupal <- length(tips_to_keep.lupal1())
    len.lencu <- length(tips_to_keep.lencu1())
    len.arahy <- length(tips_to_keep.arahy1())
    len.poptr <- length(tips_to_keep.poptr1())
    len.manes <- length(tips_to_keep.manes1())
    len.salpu <- length(tips_to_keep.salpu1())
    len.linus <- length(tips_to_keep.linus1())
    len.corci <- length(tips_to_keep.corci1())
    len.vitvi <- length(tips_to_keep.vitvi1())
    len.kalfe <- length(tips_to_keep.kalfe1())
    len.vacda <- length(tips_to_keep.vacda1())
    len.oleeu <- length(tips_to_keep.oleeu1())
    len.mimgu <- length(tips_to_keep.mimgu1())
    len.soltu <- length(tips_to_keep.soltu1())
    len.cofar <- length(tips_to_keep.cofar1())
    len.lacsa <- length(tips_to_keep.lacsa1())
    len.dauca <- length(tips_to_keep.dauca1())
    len.betvu <- length(tips_to_keep.betvu1())
    len.amahy <- length(tips_to_keep.amahy1())
    len.chequ <- length(tips_to_keep.chequ1())
    len.sapof <- length(tips_to_keep.sapof1())
    len.aquco <- length(tips_to_keep.aquco1())
    
                             
    organims_reduced <- c(rep("Prasinoderma coloniale", len.praco),rep("Ostreococcus tauri", len.ot),
                          rep("Ostreococcus lucimarinus", len.ostlu),rep("Ostreococcus mediterraneus", len.ostme),
                          rep("Bathycoccus prasinos", len.bp),rep("Micromonas pusilla", len.mi),
                          rep("Micromonas commoda", len.micco),rep("Cymbomonas tetramitiformis", len.cymte),
                          rep("Chlamydomonas reinhardtii", len.cr),rep("Chlamydomonas incerta", len.chlin),
                          rep("Haematococcus lacustris", len.haema),rep("Volvox carteri", len.vc),
                          rep("Dunaliella salina", len.ds),rep("Gonium pectorale", len.gonpe),
                          rep("Tetrabaena socialis", len.tetso),rep("Chromochloris zofingiensis", len.cz),
                          rep("Raphidocelis subcapitata", len.rs),rep("Scenedesmus obliquus", len.sceobli),
                          rep("Desmodesmus armatus", len.desar),rep("Monoraphidium minutum", len.monmi),
                          rep("Caulerpa lentillifera", len.caule),rep("Ostreobium quekettii", len.ostqu),
                          rep("Bryopsis sp.", len.brysp),rep("Coccomyxa subellipsoidea", len.cocco),
                          rep("Elliptochloris bilobata", len.ellbi),rep("Myrmecia bisecta", len.myrbi),
                          rep("Apatococcus lobatus", len.apalo),rep("Asterochloris glomerata", len.astgl),
                          rep("Trebouxia sp.", len.tresp),rep("Picochlorum soloecismus", len.picso),
                          rep("Chlorella sorokiniana", len.chlso),rep("Auxenochlorella protothecoides", len.auxpr),
                          rep("Helicosporidium sp.", len.helsp),rep("Ulva mutabilis", len.um),
                          rep("Tetraselmis striata", len.tetst),rep("Pedinophyceae sp.", len.pedsp),
                          rep("Chloropicon primus", len.chlpr),rep("Picocystis sp.", len.picsp),
                          rep("Chlorokybus atmophyticus", len.ca),rep("Mesostigma viride CCAC 1140", len.mv),
                          rep("Mesostigma viride NIES 296", len.mesvb),rep("Klebsormidium nitens", len.kn),
                          rep("Chara braunii", len.chara),rep("Mesotaenium endlicherianum", len.me),
                          rep("Mesotaenium kramstae", len.meskr),rep("Spirogloea muscicola", len.sp),
                          rep("Zygnema circumcarinatum SAG", len.zygci),rep("Zygnema circumcarinatum UTEX 1559", len.zygcb),
                          rep("Zygnema cylindricum", len.zygcy),rep("Ceratodon purpureus", len.cp),
                          rep("Physcomitrium patens", len.pp),rep("Pleurozium schreberi", len.plesc),
                          rep("Funaria hygrometrica", len.funhy),rep("Sphagnum magellanicum", len.smag),
                          rep("Sphagnum fallax", len.sphfa),rep("Takakia lepidozioides", len.takle),
                          rep("Marchantia polymorpha", len.mp),rep("Marchantia paleacea", len.marpa),
                          rep("Riccia fluitans", len.ricfl),rep("Riccia sorocarpa", len.ricso),
                          rep("Anthoceros agrestis", len.aa),rep("Anthoceros punctatus", len.antpu),
                          rep("Anthoceros fusiformis", len.antfu),rep("Megaceros flagellaris", len.megfl),
                          rep("Phaeomegaceros chiloensis", len.phach),rep("Phymatoceros phymatodes", len.phyph),
                          rep("Paraphymatoceros pearsonii", len.parpe),rep("Notothylas orbicularis", len.notor),
                          rep("Phaeoceros carolinianus", len.phaca),rep("Phaeoceros sp.", len.phasp),
                          rep("Leiosporoceros dussii", len.leidu),rep("Selaginella moellendorffii", len.sm),
                          rep("Isoetes sinensis", len.isosi),rep("Diphasiastrum complanatum", len.dipco),
                          rep("Huperzia asiatica", len.hupas),rep("Lycopodium clavatum", len.lyccl),
                          rep("Ceratopteris richardii", len.cri),rep("Adiantum capillus", len.adica),
                          rep("Alsophila spinulosa", len.alssp),rep("Salvinia cucullata", len.sc),
                          rep("Azolla filiculoides", len.af),rep("Marsilea vestita", len.marve),
                          rep("Thuja plicata", len.tp),rep("Sequoia sempervirens", len.seqse),
                          rep("Sequoiadendron giganteum", len.seqgi),rep("Taxus chinensis", len.taxch),
                          rep("Abies alba", len.abial),rep("Picea abies", len.picab),rep("Gnetum montanum", len.gnemo),
                          rep("Welwitschia mirabilis", len.welmi),rep("Cycas panzhihuaensis", len.cyc),
                          rep("Ginkgo biloba", len.ginbi),rep("Amborella trichopoda", len.ambtr),
                          rep("Nymphaea colorata", len.nymco),rep("Cinnamomum kanehirae", len.cinka),
                          rep("Aegilops tauschii", len.aegi),rep("Triticum aestivum", len.ta),
                          rep("Hordeum vulgare", len.horvu),rep("Brachypodium distachyon", len.bradi),
                          rep("Oryza sativa", len.os),rep("Setaria italica", len.setit),rep("Sorghum bicolor", len.sb),
                          rep("Zea mays", len.zea),rep("Panicum hallii", len.panha),rep("Miscanthus sinensis", len.missi),
                          rep("Oropetium thomaeum", len.oroth),rep("Eleusine coracana", len.eleco),
                          rep("Ananas comosus", len.anaco),rep("Musa acuminata", len.musac),rep("Dioscorea alata", len.dioal),
                          rep("Spirodela polyrhiza", len.spipo),rep("Asparagus officinalis", len.aspof),
                          rep("Zostera marina", len.zosma),rep("Arabidopsis thaliana", len.at),
                          rep("Arabidopsis lyrata", len.araly),rep("Capsella grandiflora", len.capgr),
                          rep("Brassica rapa", len.brara),rep("Brassica oleracea", len.braol),
                          rep("Sisymbrium irio", len.sisir),rep("Schrenkiella parvula", len.schpa),
                          rep("Eutrema salsugineum", len.eutsa),rep("Thlaspi arvense", len.thlar),
                          rep("Boechera stricta", len.boest),rep("Carica papaya", len.carpa),rep("Theobroma cacao", len.theca),
                          rep("Gossypium hirsutum", len.goshi),rep("Gossypium raimondii", len.gosra),
                          rep("Citrus sinensis", len.citsi),rep("Poncirus trifoliata", len.pontr),
                          rep("Eucalyptus grandis", len.eucgr),rep("Malus domestica", len.maldo),
                          rep("Prunus persica", len.prupe),rep("Fragaria vesca", len.frave),rep("Corylus avellana", len.corav),
                          rep("Carya illinoinensis", len.caril),rep("Quercus rubra", len.queru),
                          rep("Cucumis sativus", len.cucsa),rep("Glycine max", len.glyma),rep("Medicago truncatula", len.medtr),
                          rep("Trifolium pratense", len.tripr),rep("Cicer arietinum", len.cicar),
                          rep("Lotus japonicus", len.lotja),rep("Phaseolus acutifolius", len.phaac),
                          rep("Vigna unguiculata", len.vigun),rep("Lupinus albus", len.lupal),rep("Lens culinaris", len.lencu),
                          rep("Arachis hypogaea", len.arahy),rep("Populus trichocarpa", len.poptr),
                          rep("Manihot esculenta", len.manes),rep("Salix purpurea", len.salpu),
                          rep("Linum usitatissimum", len.linus),rep("Corymbia citriodora", len.corci),
                          rep("Vitis vinifera", len.vitvi),rep("Kalanchoe fedtschenkoi", len.kalfe),
                          rep("Vaccinium darrowii", len.vacda),rep("Olea europaea", len.oleeu),
                          rep("Mimulus guttatus", len.mimgu),rep("Solanum lycopersicum", len.sl),
                          rep("Solanum tuberosum", len.soltu),rep("Coffea arabica", len.cofar),
                          rep("Lactuca sativa", len.lacsa),rep("Daucus carota", len.dauca),rep("Beta vulgaris", len.betvu),
                          rep("Amaranthus hypochondriacus", len.amahy),rep("Chenopodium quinoa", len.chequ),
                          rep("Saponaria officinalis", len.sapof),rep("Aquilegia coerulea", len.aquco))
    
    
    
    
    return(organims_reduced)
  }) %>% bindEvent(input$run_button1)
  
  # Create plot of subset tree
  tree_plot1 <- reactive({
    
    # Define previous variables
    tree_reduced <- tree_reduced1()
    gene.name.tree <- str_replace_all(input$geneInt1, fixed(" "), "")
    tree <- tree_adj1()
    
    tips_to_keep.mp <- tips_to_keep.mp1()
    tips_to_keep.ot <- tips_to_keep.ot1()
    tips_to_keep.at <- tips_to_keep.at1()
    tips_to_keep.cp <- tips_to_keep.cp1()
    tips_to_keep.cr <- tips_to_keep.cr1()
    tips_to_keep.cz <- tips_to_keep.cz1()
    tips_to_keep.kn <- tips_to_keep.kn1()
    tips_to_keep.me <- tips_to_keep.me1()
    tips_to_keep.mi <- tips_to_keep.mi1()
    tips_to_keep.pp <- tips_to_keep.pp1()
    tips_to_keep.sl <- tips_to_keep.sl1()
    tips_to_keep.sm <- tips_to_keep.sm1()
    tips_to_keep.sp <- tips_to_keep.sp1()
    tips_to_keep.ta <- tips_to_keep.ta1()
    tips_to_keep.vc <- tips_to_keep.vc1()
    tips_to_keep.bp <- tips_to_keep.bp1()
    tips_to_keep.cri <- tips_to_keep.cri1()
    tips_to_keep.ds <- tips_to_keep.ds1()
    tips_to_keep.os <- tips_to_keep.os1()
    tips_to_keep.smag <- tips_to_keep.smag1()
    tips_to_keep.tp <- tips_to_keep.tp1()
    tips_to_keep.aa <- tips_to_keep.aa1()
    tips_to_keep.um <- tips_to_keep.um1()
    tips_to_keep.rs <- tips_to_keep.rs1()
    tips_to_keep.cyc <- tips_to_keep.cyc1()
    tips_to_keep.ca <- tips_to_keep.ca1()
    tips_to_keep.mv <- tips_to_keep.mv1()
    tips_to_keep.af <- tips_to_keep.af1()
    tips_to_keep.sc <- tips_to_keep.sc1()
    tips_to_keep.aegi <- tips_to_keep.aegi1()
    tips_to_keep.sb <- tips_to_keep.sb1()
    tips_to_keep.chara <- tips_to_keep.chara1()
    tips_to_keep.sceobli <- tips_to_keep.sceobli1()
    tips_to_keep.cocco <- tips_to_keep.cocco1()
    tips_to_keep.haema <- tips_to_keep.haema1()
    tips_to_keep.zm <- tips_to_keep.zm1()
    tips_to_keep.praco <- tips_to_keep.praco1()
    tips_to_keep.ostlu <- tips_to_keep.ostlu1()
    tips_to_keep.ostme <- tips_to_keep.ostme1()
    tips_to_keep.micco <- tips_to_keep.micco1()
    tips_to_keep.cymte <- tips_to_keep.cymte1()
    tips_to_keep.chlin <- tips_to_keep.chlin1()
    tips_to_keep.gonpe <- tips_to_keep.gonpe1()
    tips_to_keep.tetso <- tips_to_keep.tetso1()
    tips_to_keep.desar <- tips_to_keep.desar1()
    tips_to_keep.monmi <- tips_to_keep.monmi1()
    tips_to_keep.caule <- tips_to_keep.caule1()
    tips_to_keep.ostqu <- tips_to_keep.ostqu1()
    tips_to_keep.brysp <- tips_to_keep.brysp1()
    tips_to_keep.ellbi <- tips_to_keep.ellbi1()
    tips_to_keep.myrbi <- tips_to_keep.myrbi1()
    tips_to_keep.apalo <- tips_to_keep.apalo1()
    tips_to_keep.astgl <- tips_to_keep.astgl1()
    tips_to_keep.tresp <- tips_to_keep.tresp1()
    tips_to_keep.picso <- tips_to_keep.picso1()
    tips_to_keep.chlso <- tips_to_keep.chlso1()
    tips_to_keep.auxpr <- tips_to_keep.auxpr1()
    tips_to_keep.helsp <- tips_to_keep.helsp1()
    tips_to_keep.tetst <- tips_to_keep.tetst1()
    tips_to_keep.pedsp <- tips_to_keep.pedsp1()
    tips_to_keep.chlpr <- tips_to_keep.chlpr1()
    tips_to_keep.picsp <- tips_to_keep.picsp1()
    tips_to_keep.mesvb <- tips_to_keep.mesvb1()
    tips_to_keep.meskr <- tips_to_keep.meskr1()
    tips_to_keep.zygci <- tips_to_keep.zygci1()
    tips_to_keep.zygcb <- tips_to_keep.zygcb1()
    tips_to_keep.zygcy <- tips_to_keep.zygcy1()
    tips_to_keep.plesc <- tips_to_keep.plesc1()
    tips_to_keep.funhy <- tips_to_keep.funhy1()
    tips_to_keep.sphfa <- tips_to_keep.sphfa1()
    tips_to_keep.takle <- tips_to_keep.takle1()
    tips_to_keep.marpa <- tips_to_keep.marpa1()
    tips_to_keep.ricfl <- tips_to_keep.ricfl1()
    tips_to_keep.ricso <- tips_to_keep.ricso1()
    tips_to_keep.antpu <- tips_to_keep.antpu1()
    tips_to_keep.antfu <- tips_to_keep.antfu1()
    tips_to_keep.megfl <- tips_to_keep.megfl1()
    tips_to_keep.phach <- tips_to_keep.phach1()
    tips_to_keep.phyph <- tips_to_keep.phyph1()
    tips_to_keep.parpe <- tips_to_keep.parpe1()
    tips_to_keep.notor <- tips_to_keep.notor1()
    tips_to_keep.phaca <- tips_to_keep.phaca1()
    tips_to_keep.phasp <- tips_to_keep.phasp1()
    tips_to_keep.leidu <- tips_to_keep.leidu1()
    tips_to_keep.isosi <- tips_to_keep.isosi1()
    tips_to_keep.dipco <- tips_to_keep.dipco1()
    tips_to_keep.hupas <- tips_to_keep.hupas1()
    tips_to_keep.lyccl <- tips_to_keep.lyccl1()
    tips_to_keep.adica <- tips_to_keep.adica1()
    tips_to_keep.alssp <- tips_to_keep.alssp1()
    tips_to_keep.marve <- tips_to_keep.marve1()
    tips_to_keep.seqse <- tips_to_keep.seqse1()
    tips_to_keep.seqgi <- tips_to_keep.seqgi1()
    tips_to_keep.taxch <- tips_to_keep.taxch1()
    tips_to_keep.abial <- tips_to_keep.abial1()
    tips_to_keep.picab <- tips_to_keep.picab1()
    tips_to_keep.gnemo <- tips_to_keep.gnemo1()
    tips_to_keep.welmi <- tips_to_keep.welmi1()
    tips_to_keep.ginbi <- tips_to_keep.ginbi1()
    tips_to_keep.ambtr <- tips_to_keep.ambtr1()
    tips_to_keep.nymco <- tips_to_keep.nymco1()
    tips_to_keep.cinka <- tips_to_keep.cinka1()
    tips_to_keep.horvu <- tips_to_keep.horvu1()
    tips_to_keep.bradi <- tips_to_keep.bradi1()
    tips_to_keep.setit <- tips_to_keep.setit1()
    tips_to_keep.panha <- tips_to_keep.panha1()
    tips_to_keep.missi <- tips_to_keep.missi1()
    tips_to_keep.oroth <- tips_to_keep.oroth1()
    tips_to_keep.eleco <- tips_to_keep.eleco1()
    tips_to_keep.anaco <- tips_to_keep.anaco1()
    tips_to_keep.musac <- tips_to_keep.musac1()
    tips_to_keep.dioal <- tips_to_keep.dioal1()
    tips_to_keep.spipo <- tips_to_keep.spipo1()
    tips_to_keep.aspof <- tips_to_keep.aspof1()
    tips_to_keep.zosma <- tips_to_keep.zosma1()
    tips_to_keep.araly <- tips_to_keep.araly1()
    tips_to_keep.capgr <- tips_to_keep.capgr1()
    tips_to_keep.brara <- tips_to_keep.brara1()
    tips_to_keep.braol <- tips_to_keep.braol1()
    tips_to_keep.sisir <- tips_to_keep.sisir1()
    tips_to_keep.schpa <- tips_to_keep.schpa1()
    tips_to_keep.eutsa <- tips_to_keep.eutsa1()
    tips_to_keep.thlar <- tips_to_keep.thlar1()
    tips_to_keep.boest <- tips_to_keep.boest1()
    tips_to_keep.carpa <- tips_to_keep.carpa1()
    tips_to_keep.theca <- tips_to_keep.theca1()
    tips_to_keep.goshi <- tips_to_keep.goshi1()
    tips_to_keep.gosra <- tips_to_keep.gosra1()
    tips_to_keep.citsi <- tips_to_keep.citsi1()
    tips_to_keep.pontr <- tips_to_keep.pontr1()
    tips_to_keep.eucgr <- tips_to_keep.eucgr1()
    tips_to_keep.maldo <- tips_to_keep.maldo1()
    tips_to_keep.prupe <- tips_to_keep.prupe1()
    tips_to_keep.frave <- tips_to_keep.frave1()
    tips_to_keep.corav <- tips_to_keep.corav1()
    tips_to_keep.caril <- tips_to_keep.caril1()
    tips_to_keep.queru <- tips_to_keep.queru1()
    tips_to_keep.cucsa <- tips_to_keep.cucsa1()
    tips_to_keep.glyma <- tips_to_keep.glyma1()
    tips_to_keep.medtr <- tips_to_keep.medtr1()
    tips_to_keep.tripr <- tips_to_keep.tripr1()
    tips_to_keep.cicar <- tips_to_keep.cicar1()
    tips_to_keep.lotja <- tips_to_keep.lotja1()
    tips_to_keep.phaac <- tips_to_keep.phaac1()
    tips_to_keep.vigun <- tips_to_keep.vigun1()
    tips_to_keep.lupal <- tips_to_keep.lupal1()
    tips_to_keep.lencu <- tips_to_keep.lencu1()
    tips_to_keep.arahy <- tips_to_keep.arahy1()
    tips_to_keep.poptr <- tips_to_keep.poptr1()
    tips_to_keep.manes <- tips_to_keep.manes1()
    tips_to_keep.salpu <- tips_to_keep.salpu1()
    tips_to_keep.linus <- tips_to_keep.linus1()
    tips_to_keep.corci <- tips_to_keep.corci1()
    tips_to_keep.vitvi <- tips_to_keep.vitvi1()
    tips_to_keep.kalfe <- tips_to_keep.kalfe1()
    tips_to_keep.vacda <- tips_to_keep.vacda1()
    tips_to_keep.oleeu <- tips_to_keep.oleeu1()
    tips_to_keep.mimgu <- tips_to_keep.mimgu1()
    tips_to_keep.soltu <- tips_to_keep.soltu1()
    tips_to_keep.cofar <- tips_to_keep.cofar1()
    tips_to_keep.lacsa <- tips_to_keep.lacsa1()
    tips_to_keep.dauca <- tips_to_keep.dauca1()
    tips_to_keep.betvu <- tips_to_keep.betvu1()
    tips_to_keep.amahy <- tips_to_keep.amahy1()
    tips_to_keep.chequ <- tips_to_keep.chequ1()
    tips_to_keep.sapof <- tips_to_keep.sapof1()
    tips_to_keep.aquco <- tips_to_keep.aquco1()
    
    
    if (length(tree_reduced$tip.label) < 2)
    {
      cat("")
    }
    else 
    {
      # Highlight the target gene
      high.gene <<- tree_reduced$tip.label[grep(pattern = gene.name.tree, tree_reduced$tip.label)]
      
      
      # Color asignment per species
      col.factor <- c()
      org.factor <- c()
      
      library(glue)
      library(ggtree)
      library(ggplot2)
      
      for (i in 1:length(tree_reduced$tip.label))
      {
        if (tree_reduced$tip.label[i] %in% high.gene)
        {
          col.factor <- c(col.factor,"#CD0000")
          org.factor <- c(org.factor,"Gen of interest")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
        {
          col.factor <- c(col.factor,"#006400")
          org.factor <- c(org.factor,"Marchantia polymorpha")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
        {
          col.factor <- c(col.factor,"#00008B")
          org.factor <- c(org.factor,"Ostreococcus tauri")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
        {
          col.factor <- c(col.factor,"#CD661D")
          org.factor <- c(org.factor,"Arabidopsis thaliana")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
        {
          col.factor <- c(col.factor,"#458B74")
          org.factor <- c(org.factor,"Ceratodon purpureus")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
        {
          col.factor <- c(col.factor,"#8B7355")
          org.factor <- c(org.factor,"Chlamydomonas reinhardtii")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
        {
          col.factor <- c(col.factor,"#458B00")
          org.factor <- c(org.factor,"Chromochloris zofingiensis")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
        {
          col.factor <- c(col.factor,"#CD1076")
          org.factor <- c(org.factor,"Klebsormidium nitens")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
        {
          col.factor <- c(col.factor,"#8B8878")
          org.factor <- c(org.factor,"Mesotaenium endlicherianum")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
        {
          col.factor <- c(col.factor,"#666666")
          org.factor <- c(org.factor,"Micromonas pusilla")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
        {
          col.factor <- c(col.factor,"#B8860B")
          org.factor <- c(org.factor,"Physcomitrium patens")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
        {
          col.factor <- c(col.factor,"#8B008B")
          org.factor <- c(org.factor,"Solanum lycopersicum")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
        {
          col.factor <- c(col.factor,"#6E8B3D")
          org.factor <- c(org.factor,"Selaginella moellendorffii")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
        {
          col.factor <- c(col.factor,"#79CDCD")
          org.factor <- c(org.factor,"Spirogloea muscicola")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
        {
          col.factor <- c(col.factor,"#CDCD00")
          org.factor <- c(org.factor,"Triticum aestivum")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
        {
          col.factor <- c(col.factor,"#16317d")
          org.factor <- c(org.factor,"Volvox carteri")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
        {
          col.factor <- c(col.factor,"#007e2f")
          org.factor <- c(org.factor,"Bathycoccus prasinos")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
        {
          col.factor <- c(col.factor,"#ffcd12")
          org.factor <- c(org.factor,"Ceratopteris richardii")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
        {
          col.factor <- c(col.factor,"#b86092")
          org.factor <- c(org.factor,"Dunaliella salina")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
        {
          col.factor <- c(col.factor,"#721b3e")
          org.factor <- c(org.factor,"Oryza sativa")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
        {
          col.factor <- c(col.factor,"#00b7a7")
          org.factor <- c(org.factor,"Sphagnum magellanicum")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
        {
          col.factor <- c(col.factor,"#67000d")
          org.factor <- c(org.factor,"Thuja plicata")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
        {
          col.factor <- c(col.factor,"#5b2c6f")
          org.factor <- c(org.factor,"Anthoceros agrestis")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
        {
          col.factor <- c(col.factor,"#15e71b")
          org.factor <- c(org.factor,"Ulva mutabilis")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
        {
          col.factor <- c(col.factor,"#2d04f9")
          org.factor <- c(org.factor,"Raphidocelis subcapitata")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
        {
          col.factor <- c(col.factor,"#873600")
          org.factor <- c(org.factor,"Cycas panzhihuaensis")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
        {
          col.factor <- c(col.factor,"#0b5345")
          org.factor <- c(org.factor,"Chlorokybus atmophyticus")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
        {
          col.factor <- c(col.factor,"#283747")
          org.factor <- c(org.factor,"Mesostigma viride CCAC 1140")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
        {
          col.factor <- c(col.factor,"#145a32")
          org.factor <- c(org.factor,"Azolla filiculoides")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
        {
          col.factor <- c(col.factor,"#3339e6")
          org.factor <- c(org.factor,"Salvinia cucullata")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
        {
          col.factor <- c(col.factor,"#e6338f")
          org.factor <- c(org.factor,"Aegilops tauschii")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
        {
          col.factor <- c(col.factor,"#cd016a")
          org.factor <- c(org.factor,"Sorghum bicolor")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
        {
          col.factor <- c(col.factor,"#117a65")
          org.factor <- c(org.factor,"Chara braunii")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
        {
          col.factor <- c(col.factor,"#148f77")
          org.factor <- c(org.factor,"Scenedesmus obliquus")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
        {
          col.factor <- c(col.factor,"#9c640c")
          org.factor <- c(org.factor,"Coccomyxa subellipsoidea")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
        {
          col.factor <- c(col.factor,"#196f3d")
          org.factor <- c(org.factor,"Haematococcus lacustris")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
        {
          col.factor <- c(col.factor,"#666909")
          org.factor <- c(org.factor,"Zea mays")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.praco])
        {
          col.factor <- c(col.factor,"#ADFF2F")
          org.factor <- c(org.factor,"Prasinoderma coloniale")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostlu])
        {
          col.factor <- c(col.factor,"#483D8B")
          org.factor <- c(org.factor,"Ostreococcus lucimarinus")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostme])
        {
          col.factor <- c(col.factor,"#7FFFD4")
          org.factor <- c(org.factor,"Ostreococcus mediterraneus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.micco])
        {
          col.factor <- c(col.factor,"#228B22")
          org.factor <- c(org.factor,"Micromonas commoda")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymte])
        {
          col.factor <- c(col.factor,"#00F5FF")
          org.factor <- c(org.factor,"Cymbomonas tetramitiformis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlin])
        {
          col.factor <- c(col.factor,"#FFE4B5")
          org.factor <- c(org.factor,"Chlamydomonas incerta")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gonpe])
        {
          col.factor <- c(col.factor,"#8B3A62")
          org.factor <- c(org.factor,"Gonium pectorale")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetso])
        {
          col.factor <- c(col.factor,"#FFDAB9")
          org.factor <- c(org.factor,"Tetrabaena socialis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.desar])
        {
          col.factor <- c(col.factor,"#54FF9F")
          org.factor <- c(org.factor,"Desmodesmus armatus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.monmi])
        {
          col.factor <- c(col.factor,"#20B2AA")
          org.factor <- c(org.factor,"Monoraphidium minutum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caule])
        {
          col.factor <- c(col.factor,"#9BCD9B")
          org.factor <- c(org.factor,"Caulerpa lentillifera")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostqu])
        {
          col.factor <- c(col.factor,"#5CACEE")
          org.factor <- c(org.factor,"Ostreobium quekettii")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brysp])
        {
          col.factor <- c(col.factor,"#B8860B")
          org.factor <- c(org.factor,"Bryopsis sp.")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ellbi])
        {
          col.factor <- c(col.factor,"#EEDD82")
          org.factor <- c(org.factor,"Elliptochloris bilobata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.myrbi])
        {
          col.factor <- c(col.factor,"#F0FFFF")
          org.factor <- c(org.factor,"Myrmecia bisecta")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.apalo])
        {
          col.factor <- c(col.factor,"#C71585")
          org.factor <- c(org.factor,"Apatococcus lobatus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.astgl])
        {
          col.factor <- c(col.factor,"#8B4789")
          org.factor <- c(org.factor,"Asterochloris glomerata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tresp])
        {
          col.factor <- c(col.factor,"#8B008B")
          org.factor <- c(org.factor,"Trebouxia sp.")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picso])
        {
          col.factor <- c(col.factor,"#D2691E")
          org.factor <- c(org.factor,"Picochlorum soloecismus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlso])
        {
          col.factor <- c(col.factor,"#8B7D6B")
          org.factor <- c(org.factor,"Chlorella sorokiniana")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.auxpr])
        {
          col.factor <- c(col.factor,"#000080")
          org.factor <- c(org.factor,"Auxenochlorella protothecoides")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.helsp])
        {
          col.factor <- c(col.factor,"#A020F0")
          org.factor <- c(org.factor,"Helicosporidium sp.")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetst])
        {
          col.factor <- c(col.factor,"#EE00EE")
          org.factor <- c(org.factor,"Tetraselmis striata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pedsp])
        {
          col.factor <- c(col.factor,"#8B1C62")
          org.factor <- c(org.factor,"Pedinophyceae sp.")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlpr])
        {
          col.factor <- c(col.factor,"#FFFACD")
          org.factor <- c(org.factor,"Chloropicon primus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picsp])
        {
          col.factor <- c(col.factor,"#FFFFE0")
          org.factor <- c(org.factor,"Picocystis sp.")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mesvb])
        {
          col.factor <- c(col.factor,"#A52A2A")
          org.factor <- c(org.factor,"Mesostigma viride NIES 296")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.meskr])
        {
          col.factor <- c(col.factor,"#E0FFFF")
          org.factor <- c(org.factor,"Mesotaenium kramstae")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygci])
        {
          col.factor <- c(col.factor,"#E6E6FA")
          org.factor <- c(org.factor,"Zygnema circumcarinatum SAG")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcb])
        {
          col.factor <- c(col.factor,"#DCDCDC")
          org.factor <- c(org.factor,"Zygnema circumcarinatum UTEX 1559")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcy])
        {
          col.factor <- c(col.factor,"#8B636C")
          org.factor <- c(org.factor,"Zygnema cylindricum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.plesc])
        {
          col.factor <- c(col.factor,"#CD2990")
          org.factor <- c(org.factor,"Pleurozium schreberi")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.funhy])
        {
          col.factor <- c(col.factor,"#CDC9C9")
          org.factor <- c(org.factor,"Funaria hygrometrica")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sphfa])
        {
          col.factor <- c(col.factor,"#1C86EE")
          org.factor <- c(org.factor,"Sphagnum fallax")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.takle])
        {
          col.factor <- c(col.factor,"#8B7500")
          org.factor <- c(org.factor,"Takakia lepidozioides")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marpa])
        {
          col.factor <- c(col.factor,"#EED5B7")
          org.factor <- c(org.factor,"Marchantia paleacea")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricfl])
        {
          col.factor <- c(col.factor,"#EE5C42")
          org.factor <- c(org.factor,"Riccia fluitans")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricso])
        {
          col.factor <- c(col.factor,"#FFFAFA")
          org.factor <- c(org.factor,"Riccia sorocarpa")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antpu])
        {
          col.factor <- c(col.factor,"#D02090")
          org.factor <- c(org.factor,"Anthoceros punctatus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antfu])
        {
          col.factor <- c(col.factor,"#FFF8DC")
          org.factor <- c(org.factor,"Anthoceros fusiformis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.megfl])
        {
          col.factor <- c(col.factor,"#EEA2AD")
          org.factor <- c(org.factor,"Megaceros flagellaris")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phach])
        {
          col.factor <- c(col.factor,"#BCD2EE")
          org.factor <- c(org.factor,"Phaeomegaceros chiloensis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phyph])
        {
          col.factor <- c(col.factor,"#FF00FF")
          org.factor <- c(org.factor,"Phymatoceros phymatodes")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.parpe])
        {
          col.factor <- c(col.factor,"#68228B")
          org.factor <- c(org.factor,"Paraphymatoceros pearsonii")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.notor])
        {
          col.factor <- c(col.factor,"#006400")
          org.factor <- c(org.factor,"Notothylas orbicularis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaca])
        {
          col.factor <- c(col.factor,"#00FFFF")
          org.factor <- c(org.factor,"Phaeoceros carolinianus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phasp])
        {
          col.factor <- c(col.factor,"#EED2EE")
          org.factor <- c(org.factor,"Phaeoceros sp.")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.leidu])
        {
          col.factor <- c(col.factor,"#FFEFD5")
          org.factor <- c(org.factor,"Leiosporoceros dussii")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.isosi])
        {
          col.factor <- c(col.factor,"#CDC8B1")
          org.factor <- c(org.factor,"Isoetes sinensis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dipco])
        {
          col.factor <- c(col.factor,"#8B6508")
          org.factor <- c(org.factor,"Diphasiastrum complanatum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.hupas])
        {
          col.factor <- c(col.factor,"#FFA54F")
          org.factor <- c(org.factor,"Huperzia asiatica")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lyccl])
        {
          col.factor <- c(col.factor,"#EEAEEE")
          org.factor <- c(org.factor,"Lycopodium clavatum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.adica])
        {
          col.factor <- c(col.factor,"#FF0000")
          org.factor <- c(org.factor,"Adiantum capillus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.alssp])
        {
          col.factor <- c(col.factor,"#9B30FF")
          org.factor <- c(org.factor,"Alsophila spinulosa")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marve])
        {
          col.factor <- c(col.factor,"#CD661D")
          org.factor <- c(org.factor,"Marsilea vestita")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqse])
        {
          col.factor <- c(col.factor,"#7AC5CD")
          org.factor <- c(org.factor,"Sequoia sempervirens")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqgi])
        {
          col.factor <- c(col.factor,"#CDC0B0")
          org.factor <- c(org.factor,"Sequoiadendron giganteum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.taxch])
        {
          col.factor <- c(col.factor,"#4682B4")
          org.factor <- c(org.factor,"Taxus chinensis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.abial])
        {
          col.factor <- c(col.factor,"#CDBA96")
          org.factor <- c(org.factor,"Abies alba")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picab])
        {
          col.factor <- c(col.factor,"#A2CD5A")
          org.factor <- c(org.factor,"Picea abies")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gnemo])
        {
          col.factor <- c(col.factor,"#FAFAD2")
          org.factor <- c(org.factor,"Gnetum montanum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.welmi])
        {
          col.factor <- c(col.factor,"#E0EEEE")
          org.factor <- c(org.factor,"Welwitschia mirabilis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ginbi])
        {
          col.factor <- c(col.factor,"#7A378B")
          org.factor <- c(org.factor,"Ginkgo biloba")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ambtr])
        {
          col.factor <- c(col.factor,"#FFDEAD")
          org.factor <- c(org.factor,"Amborella trichopoda")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.nymco])
        {
          col.factor <- c(col.factor,"#C0FF3E")
          org.factor <- c(org.factor,"Nymphaea colorata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cinka])
        {
          col.factor <- c(col.factor,"#7FFFD4")
          org.factor <- c(org.factor,"Cinnamomum kanehirae")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.horvu])
        {
          col.factor <- c(col.factor,"#1E90FF")
          org.factor <- c(org.factor,"Hordeum vulgare")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bradi])
        {
          col.factor <- c(col.factor,"#76EE00")
          org.factor <- c(org.factor,"Brachypodium distachyon")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.setit])
        {
          col.factor <- c(col.factor,"#EE9A49")
          org.factor <- c(org.factor,"Setaria italica")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.panha])
        {
          col.factor <- c(col.factor,"#FFF0F5")
          org.factor <- c(org.factor,"Panicum hallii")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.missi])
        {
          col.factor <- c(col.factor,"#CD9B1D")
          org.factor <- c(org.factor,"Miscanthus sinensis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oroth])
        {
          col.factor <- c(col.factor,"#36648B")
          org.factor <- c(org.factor,"Oropetium thomaeum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eleco])
        {
          col.factor <- c(col.factor,"#EEA9B8")
          org.factor <- c(org.factor,"Eleusine coracana")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.anaco])
        {
          col.factor <- c(col.factor,"#ADD8E6")
          org.factor <- c(org.factor,"Ananas comosus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.musac])
        {
          col.factor <- c(col.factor,"#66CDAA")
          org.factor <- c(org.factor,"Musa acuminata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dioal])
        {
          col.factor <- c(col.factor,"#FFF0F5")
          org.factor <- c(org.factor,"Dioscorea alata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.spipo])
        {
          col.factor <- c(col.factor,"#CD5555")
          org.factor <- c(org.factor,"Spirodela polyrhiza")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aspof])
        {
          col.factor <- c(col.factor,"#FFEC8B")
          org.factor <- c(org.factor,"Asparagus officinalis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zosma])
        {
          col.factor <- c(col.factor,"#668B8B")
          org.factor <- c(org.factor,"Zostera marina")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.araly])
        {
          col.factor <- c(col.factor,"#CDC673")
          org.factor <- c(org.factor,"Arabidopsis lyrata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.capgr])
        {
          col.factor <- c(col.factor,"#8FBC8F")
          org.factor <- c(org.factor,"Capsella grandiflora")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brara])
        {
          col.factor <- c(col.factor,"#4EEE94")
          org.factor <- c(org.factor,"Brassica rapa")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.braol])
        {
          col.factor <- c(col.factor,"#BA55D3")
          org.factor <- c(org.factor,"Brassica oleracea")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sisir])
        {
          col.factor <- c(col.factor,"#A4D3EE")
          org.factor <- c(org.factor,"Sisymbrium irio")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.schpa])
        {
          col.factor <- c(col.factor,"#8B2323")
          org.factor <- c(org.factor,"Schrenkiella parvula")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eutsa])
        {
          col.factor <- c(col.factor,"#8B2252")
          org.factor <- c(org.factor,"Eutrema salsugineum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.thlar])
        {
          col.factor <- c(col.factor,"#FF6A6A")
          org.factor <- c(org.factor,"Thlaspi arvense")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.boest])
        {
          col.factor <- c(col.factor,"#EE7942")
          org.factor <- c(org.factor,"Boechera stricta")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.carpa])
        {
          col.factor <- c(col.factor,"#BBFFFF")
          org.factor <- c(org.factor,"Carica papaya")
        }
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.theca])
        {
          col.factor <- c(col.factor,"#00CD66")
          org.factor <- c(org.factor,"Theobroma cacao")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.goshi])
        {
          col.factor <- c(col.factor,"#FFE4C4")
          org.factor <- c(org.factor,"Gossypium hirsutum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gosra])
        {
          col.factor <- c(col.factor,"#8B8386")
          org.factor <- c(org.factor,"Gossypium raimondii")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.citsi])
        {
          col.factor <- c(col.factor,"#CD4F39")
          org.factor <- c(org.factor,"Citrus sinensis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pontr])
        {
          col.factor <- c(col.factor,"#8B4726")
          org.factor <- c(org.factor,"Poncirus trifoliata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eucgr])
        {
          col.factor <- c(col.factor,"#FF6EB4")
          org.factor <- c(org.factor,"Eucalyptus grandis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.maldo])
        {
          col.factor <- c(col.factor,"#00008B")
          org.factor <- c(org.factor,"Malus domestica")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.prupe])
        {
          col.factor <- c(col.factor,"#6495ED")
          org.factor <- c(org.factor,"Prunus persica")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.frave])
        {
          col.factor <- c(col.factor,"#8DB6CD")
          org.factor <- c(org.factor,"Fragaria vesca")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corav])
        {
          col.factor <- c(col.factor,"#EE9572")
          org.factor <- c(org.factor,"Corylus avellana")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caril])
        {
          col.factor <- c(col.factor,"#FFEBCD")
          org.factor <- c(org.factor,"Carya illinoinensis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.queru])
        {
          col.factor <- c(col.factor,"#9ACD32")
          org.factor <- c(org.factor,"Quercus rubra")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cucsa])
        {
          col.factor <- c(col.factor,"#EE6AA7")
          org.factor <- c(org.factor,"Cucumis sativus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.glyma])
        {
          col.factor <- c(col.factor,"#D15FEE")
          org.factor <- c(org.factor,"Glycine max")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.medtr])
        {
          col.factor <- c(col.factor,"#6B8E23")
          org.factor <- c(org.factor,"Medicago truncatula")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tripr])
        {
          col.factor <- c(col.factor,"#8B1A1A")
          org.factor <- c(org.factor,"Trifolium pratense")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cicar])
        {
          col.factor <- c(col.factor,"#F0FFF0")
          org.factor <- c(org.factor,"Cicer arietinum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lotja])
        {
          col.factor <- c(col.factor,"#DEB887")
          org.factor <- c(org.factor,"Lotus japonicus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaac])
        {
          col.factor <- c(col.factor,"#FFA07A")
          org.factor <- c(org.factor,"Phaseolus acutifolius")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vigun])
        {
          col.factor <- c(col.factor,"#FFD700")
          org.factor <- c(org.factor,"Vigna unguiculata")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lupal])
        {
          col.factor <- c(col.factor,"#8B668B")
          org.factor <- c(org.factor,"Lupinus albus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lencu])
        {
          col.factor <- c(col.factor,"#FF7F50")
          org.factor <- c(org.factor,"Lens culinaris")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.arahy])
        {
          col.factor <- c(col.factor,"#8B8B7A")
          org.factor <- c(org.factor,"Arachis hypogaea")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.poptr])
        {
          col.factor <- c(col.factor,"#FAF0E6")
          org.factor <- c(org.factor,"Populus trichocarpa")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.manes])
        {
          col.factor <- c(col.factor,"#EE7621")
          org.factor <- c(org.factor,"Manihot esculenta")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.salpu])
        {
          col.factor <- c(col.factor,"#CDAA7D")
          org.factor <- c(org.factor,"Salix purpurea")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.linus])
        {
          col.factor <- c(col.factor,"#EE799F")
          org.factor <- c(org.factor,"Linum usitatissimum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corci])
        {
          col.factor <- c(col.factor,"#551A8B")
          org.factor <- c(org.factor,"Corymbia citriodora")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vitvi])
        {
          col.factor <- c(col.factor,"#90EE90")
          org.factor <- c(org.factor,"Vitis vinifera")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kalfe])
        {
          col.factor <- c(col.factor,"#8B4C39")
          org.factor <- c(org.factor,"Kalanchoe fedtschenkoi")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vacda])
        {
          col.factor <- c(col.factor,"#CD5C5C")
          org.factor <- c(org.factor,"Vaccinium darrowii")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oleeu])
        {
          col.factor <- c(col.factor,"#CD3278")
          org.factor <- c(org.factor,"Olea europaea")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mimgu])
        {
          col.factor <- c(col.factor,"#B3EE3A")
          org.factor <- c(org.factor,"Mimulus guttatus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.soltu])
        {
          col.factor <- c(col.factor,"#EE3B3B")
          org.factor <- c(org.factor,"Solanum tuberosum")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cofar])
        {
          col.factor <- c(col.factor,"#FFA500")
          org.factor <- c(org.factor,"Coffea arabica")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lacsa])
        {
          col.factor <- c(col.factor,"#66CD00")
          org.factor <- c(org.factor,"Lactuca sativa")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dauca])
        {
          col.factor <- c(col.factor,"#FF4500")
          org.factor <- c(org.factor,"Daucus carota")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.betvu])
        {
          col.factor <- c(col.factor,"#00FF00")
          org.factor <- c(org.factor,"Beta vulgaris")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.amahy])
        {
          col.factor <- c(col.factor,"#EEC900")
          org.factor <- c(org.factor,"Amaranthus hypochondriacus")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chequ])
        {
          col.factor <- c(col.factor,"#7B68EE")
          org.factor <- c(org.factor,"Chenopodium quinoa")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sapof])
        {
          col.factor <- c(col.factor,"#FFA07A")
          org.factor <- c(org.factor,"Saponaria officinalis")
        }
        
        else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aquco])
        {
          col.factor <- c(col.factor,"#DA70D6")
          org.factor <- c(org.factor,"Aquilegia coerulea")
        }
        
        
      }
      
      #Matrix with labels and colors and transform to dplyr format
      data.tree <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                              col = col.factor, org = org.factor)
      
      d2 <- dplyr::mutate(data.tree, lab = data.tree$label,
                          color = data.tree$col,
                          organism = data.tree$org,
                          name = glue("<i style='color:{color}'> {lab} </i>"))
      { 
      if (build_trees1() == "Maximum Likelihood")
      {
      tree_plot <- ggtree(tree_reduced) %<+% d2 + geom_tiplab() + theme(legend.position =) +
        xlim(0, max(tree_reduced$edge.length)*2.2) + geom_tiplab(aes(label = label, color = organism)) +
        scale_color_manual(values = unique(d2$col), breaks = unique(d2$org)) +
        geom_highlight(mapping=aes(subset = label %in% high.gene,
                                   node = node,
                                   fill = as.factor(node)), extend = 0.8) + 
        labs(fill = "Node of interest")
      
      }
      else
      {
        tree_plot <- ggtree(tree_reduced) %<+% d2 + geom_tiplab() + theme(legend.position =) +
          xlim(0, max(tree_reduced$edge.length)*2.2) + geom_tiplab(aes(label = label, color = organism)) +
          geom_nodelab() +
          scale_color_manual(values = unique(d2$col), breaks = unique(d2$org)) +
          geom_highlight(mapping=aes(subset = label %in% high.gene,
                                     node = node,
                                     fill = as.factor(node)), extend = 0.8) + 
          labs(fill = "Node of interest")
      }
      }
      shinyjs::hideElement(id = 'loading.tree1')
      return(tree_plot)
    }}) %>% bindEvent(input$run_button1)
  

  # Outputs
  observeEvent(isTruthy(tree_plot1()), {
    output$error_tree1 <- NULL
  })
  
  # Create boxes
  observeEvent(isTruthy(tree_plot1()), {
    
    if (UI_exist_tree1)
    {
      removeUI(
        selector = "div:has(>> #treeTips1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>>> #presentorg1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #tree_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadTree1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadNewick1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadTreeSeqs1)",
        multiple = TRUE,
        immediate = TRUE
      )
    }
    
    
    insertUI("#box_tree_text1", "afterEnd", ui = {
      box(
        title = "Genes in Orthogroup", status = "info", solidHeader = TRUE, width = 12,
        collapsible = TRUE,
        verbatimTextOutput("treeTips1")
      )
    }) 
    
    insertUI("#box_tree_pie1", "afterEnd", ui = {
      box(
        title = "Present Organisms", status = "info", solidHeader = TRUE,
        collapsible = TRUE, width = 12,
        plotlyOutput("presentorg1")
      )
    }) 
    
    insertUI("#box_tree_plot1", "afterEnd", ui = {
      image_height <- 300 + 15*length(tree_reduced1()$tip.label)
      box(width = 12, height = image_height + 100,
          title = "Gene Tree", status = "info", solidHeader = TRUE,
          collapsible = TRUE, 
          plotOutput("tree_image1", height = image_height, width = 1100)
      )
    })
      
      insertUI("#download_tree1", "afterEnd", ui = {
        tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTree1", 
                                                                           "Download Tree Plot",
                                                                           size = "sm", color = "primary"))
      })
      
      insertUI("#download_newick1", "afterEnd", ui = {
        tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadNewick1", 
                                                                           "Download NEWICK Tree",
                                                                           size = "sm", color = "primary"))
      })
      
      insertUI("#download_tree_seqs1", "afterEnd", ui = {
        tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTreeSeqs1", 
                                                                           "Download Protein Sequences",
                                                                           size = "sm", color = "primary"))
      })
 
    UI_exist_tree1 <<- TRUE
    shinyjs::hideElement(id = 'loading.tree1')
  })
  
  # Fill boxes with output
  output$treeTips1 <- renderPrint({
    print(tree_reduced1()$tip.label)
  }, width = 400) # %>% bindEvent(input$run_button1)

  # Render pie chart
  output$presentorg1 <- renderPlotly({

    {library(ggplot2)
    library(dplyr)

    data <- data.frame(table(organims_reduced1()))
    colnames(data) <- c("group", "value")
  
  # Compute the position of labels
    data <- data %>%
      arrange(desc(group)) %>%
      mutate(prop = value / sum(data$value) *100) %>%
      mutate(ypos = cumsum(prop)- 0.5*prop )

  # Create plot
    
    plotly::plot_ly(data=data,values=~prop,labels=~factor(group),
                    marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                           floor(nrow(data)/9)+1)),
                    type="pie",showlegend = F, text= ~group,
                    textinfo = "none", hoverinfo = "text")} 

  })

  # Render tree image
  output$tree_image1 <- renderImage({
    image_height <- 300 + 15*length(tree_reduced1()$tip.label)
    png("tree1.png", height = image_height, width = 1100)
    plot(tree_plot1())
    dev.off()

    list(src = "tree1.png",
         contentType="image/png", width=1100,height=image_height)
  }, deleteFile = T)

 # Download results
  output$downloadTree1 <- downloadHandler(
    filename= function() {
      paste("tree", ".png", sep="")
    },
    content= function(file) {
      image_height <- (300 + 11*length(tree_reduced1()$tip.label))*3
      image_width <- (200 + 400*max(tree_reduced1()$edge.length))*3
      png(file, height = image_height, width = image_width, res = (70 + 0.1*length(tree_reduced1()$tip.label))*3)
      plot(tree_plot1())
      dev.off()
    })

 # Create and download tree in newick format
 output$downloadNewick1 <- downloadHandler(
    filename= function() {
      paste("tree_newick", ".txt", sep="")
    },
    content= function(file) {
      write.tree(tree_reduced1(), file)
    })

 #  # Create and download sequences for genes in tree
  output$downloadTreeSeqs1 <- downloadHandler(
    filename= function() {
      paste("tree_seqs", ".fa", sep="")
    },
    content= function(file) {
      seqinr::write.fasta(sequences = seqinr::getSequence(ortho_reduced1()),
                  names = seqinr::getName(ortho_reduced1()), file.out = file)
    })
  
  
  ####################### PHYLOWIDGET ############################
  # Remove previous outputs when updated by a new search
  observeEvent(input$run_button1, {
    if (UI_exist_phylo1)
    {
      removeUI(
        selector = "div:has(>>> #phylo_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_phylo1 <<- F
    }
    
  })
  
  
  phylo_tree1 <- reactive({
    
    library(ape)
    tree_phylo <- tree_reduced1()
    
    # Normalize tree depth
    root_id <- length(tree_phylo$tip.label)+1
    norm_factor <- max(dist.nodes(tree_phylo)[root_id,])
    tree_phylo$edge.length <- tree_phylo$edge.length/norm_factor
    
    return(tree_phylo)
    
  }) %>% bindEvent(input$phylo_start1)
  
  observeEvent(isTruthy(phylo_tree1()),{
    
    if(UI_exist_phylo1)
    {
      removeUI(
        selector = "div:has(>>> #phylo_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_phylo1 <<- F
    }
    
    
    insertUI("#box_phylo1", "afterEnd", ui = {
      phylo_tree <- phylo_tree1()
      phylo_height <- length(phylo_tree$tip.label) *14 + 220
      box(width = 12,
          title = "Interactive Tree", status = "info", solidHeader = TRUE, height = phylo_height + 100,
          collapsible = TRUE,
          tags$div(id = "phylo_pocket1", style = paste0("width: 1300px; height: ",  phylo_height + 50, "px"),
                   phylowidgetOutput("phylo_plot1", height = paste0(phylo_height,"px"), width = "98%"))
      )
      })
    
    UI_exist_phylo1 <<- T
    
  })
  
  output$phylo_plot1 <- renderPhylowidget({
    
    phylo_tree <- phylo_tree1()
    phylowidget(phylo_tree)
  })
  
  #########################  PFAM  ###############################
  
  observeEvent(input$run_button1, {
    removeUI(
      selector = "div:has(>> #selected_pfamsI1)",
      multiple = TRUE,
      immediate = TRUE
    )
  })
  
  observeEvent(input$pfam_start1, {
    insertUI("#selected_pfams1", "afterEnd", ui = {
      
      shinyWidgets::pickerInput("selected_pfamsI1","Select the desired genes from the tree",
                                choices=isolate({tree_reduced1()$tip.label}), options = list(`actions-box` = TRUE),
                                multiple = T, selected = isolate({str_replace_all(input$geneInt1, fixed(" "), "")}))
    })
  })
  
  observeEvent(input$run_button1, {
    removeUI("#pfam_selection1")
  })
  
  observeEvent(input$pfam_start1, {
    insertUI("#pfam_selectionI1", "afterEnd", ui = {
      
      shinyWidgets::actionBttn("pfam_selection1", "Show Pfam Domains", size = "sm",
                               style = "float", color = "primary")
    })
  })
  
  
  observeEvent(input$run_button1, {
    if (UI_exist_pfam1)
    {
      removeUI(
        selector = "div:has(>> #output_pfam_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #pfam_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #pfam_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadPFAMTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_pfam1 <<- F
    }
  })
  
  
  total_table_pfam1 <- reactive({
    shinyjs::showElement(id = 'loading.pfam.pf1')
    ortho_reduced <- ortho_reduced1()
    sel_genes <- as.vector(input$selected_pfamsI1)
    
    if (length(sel_genes) < 1)
    {
      shinyjs::hideElement(id = 'loading.pfam.pf1')
      removeUI(
        selector = "div:has(>> #output_pfam_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #pfam_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #pfam_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadPFAMTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      UI_exist_pfam1 <<- F
      output$error_pfam1 <- renderUI({renderText({print("Please select at least one gene.")})})
      validate(need(length(sel_genes) > 0, "Please select at least one gene."))
    }
    
    output$error_pfam1 <- NULL
    #library(bio3d)
    library(RCurl)
    library(drawProteins)
    library(ggplot2)
    library(jsonlite)
    
    # Get the sequences as a vector of strings
    
    
    # Create data frame with proper columns
    total_table_pfam <- data.frame(type=NA,
                                   description=NA,
                                   begin=NA, end=NA,
                                   length=NA,
                                   accession=NA, entryName=NA,
                                   taxid=NA, order=NA)
    
    
    # Fill data frame with the information about domains obtained with hmmer
    for (i in 1:length(sel_genes))
    {
      ortho_comp <- ortho_reduced[[sel_genes[i]]]
      ortho_str <- seqinr::getSequence(ortho_comp, as.string = T)
      ortho_cha <- unlist(ortho_str)
      
      
      # Curl POST request
      url <- "https://www.ebi.ac.uk/Tools/hmmer/api/v1/search/hmmscan"
      
      data <- list(
        database = "pfam",
        input = ortho_cha
      )
      
      json_data <- toJSON(data, auto_unbox = T)
      
      response <- postForm(url, 
                           .opts = list(
                             postfields = json_data,
                             httpheader = c(
                               "Content-Type" = "application/json",
                               "Accept" = "application/json"
                             )
                           ))
      
      
      new_id <- strsplit(as.character(response), '"')[[1]][4]
      
      
      # Results URL
      url_res <- paste0("https://www.ebi.ac.uk/Tools/hmmer/api/v1/result/", new_id)
      
      # Iterate until job is finished, for that, check if SUCCESS is present in GET result
      nap_again <- T
      while(nap_again)
      {
        get_response <- getURL(url_res,
                               httpheader = c(
                                 "Accept" = "application/json"
                               ))
        
        if (grepl("SUCCESS", get_response))
        {
          nap_again <- F
          
          # Once computation is over, parse JSON response
          json_resp <- jsonlite::fromJSON(url_res, simplifyDataFrame = TRUE)
          json_hits <- data.frame(json_resp$result$hits)
          
          hits_acc <- json_hits$acc
          hits_desc <- json_hits$metadata$description
          hits_domains <- json_hits$domains
          
          df_acc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_acc, y = hits_domains, USE.NAMES = F))
          df_desc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_desc, y = hits_domains, USE.NAMES = F))
          df_begin <- unlist(sapply(hits_domains, function(x) x$iali))
          df_end <- unlist(sapply(hits_domains, function(x) x$jali))
          
          # Create PFAM table
          pfam_table <- data.frame(type=c("CHAIN", rep("DOMAIN", length(df_acc))),
                                   description=c("Protein chain",df_acc),
                                   begin=c(1, df_begin), end=c(nchar(ortho_cha),df_end),
                                   length=c(nchar(ortho_cha)-1, df_end - df_begin),
                                   accession=sel_genes[i], entryName=sel_genes[i],
                                   taxid=c("Chain", df_desc), order=i)
          
          total_table_pfam <- rbind(total_table_pfam, pfam_table)
          
        }
        else if (grepl("PENDING", get_response))
        {
          Sys.sleep(10)
        }
        # If there is a problem with server
        else
        {
          nap_again <- F
          pfam_table <- data.frame(type="CHAIN",
                                   description="Protein chain",
                                   begin=1, end=nchar(ortho_cha),
                                   length=nchar(ortho_cha)-1,
                                   accession=sel_genes[i], entryName=sel_genes[i],
                                   taxid="Chain", order=i)
          total_table_pfam <- rbind(total_table_pfam, pfam_table)
        }
      }
    }
    
    
    total_table_pfam <- total_table_pfam[-1,]
    total_table_pfam <- total_table_pfam[!duplicated(total_table_pfam),]
    # Remove protein chain results
    total_table_pfam <- subset(total_table_pfam, !(type=="DOMAIN" & description=="Protein chain"))
    
    return(total_table_pfam)
    
  }) %>% bindEvent(input$pfam_selection1)
  
  pfplot1 <- reactive({
    
    total_table_pfam <- total_table_pfam1()
    # Now we can plot domains information as chains
    pfplot <- draw_canvas(total_table_pfam)
    pfplot <- draw_chains(pfplot, total_table_pfam)
    pfplot <- draw_domains(pfplot, total_table_pfam, label_domains = F)
    pfplot <- pfplot + theme_bw(base_size = 20) + # white background
      theme(panel.grid.minor=element_blank(),
            panel.grid.major=element_blank()) +
      theme(axis.ticks = element_blank(),
            axis.text.y = element_blank()) +
      theme(panel.border = element_blank())
    #pfplot <- pfplot + labs(title = "Pfam domains")
    pfplot <- pfplot + theme(legend.position="top") + labs(fill="")
    
  }) %>% bindEvent(input$pfam_selection1)
  
  
  # Outputs
  
  observeEvent(isTruthy(pfplot1()), {
    
    if (UI_exist_pfam1)
    {
     removeUI(
           selector = "div:has(>> #output_pfam_table1)",
           multiple = TRUE,
           immediate = TRUE
         )
      
      removeUI(
        selector = "div:has(>> #pfam_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #pfam_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadPFAMTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
    
    }
    
      insertUI("#box_pfam1", "afterEnd", ui = {
        
        box(
          title = "PFAM Table", status = "info", solidHeader = TRUE, width = 12,
          collapsible = TRUE,
          dataTableOutput(outputId = "output_pfam_table1"))
    }) 
      
      insertUI("#box_pfplot1", "afterEnd", ui = {
        total_table_pfam <- total_table_pfam1()
        box_pfplot_height <- 150 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
        box(
          title = "Domains Localization", status = "info", solidHeader = TRUE, width = 12, height = box_pfplot_height+10,
          collapsible = TRUE,
          plotOutput("pfam_plot1", height = box_pfplot_height)
          )
        
      }) 
      
      insertUI("#pfam_down_button1", "afterEnd", ui = {
        tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "pfam_download1", "Download PFAM figure",
                                              size = "sm", color = "primary"))
        })
      
      insertUI("#download_ui_for_pfam_table1", "afterEnd", ui = {
        tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadPFAMTable1", "Download PFAM Table",
                                            size = "sm", color = "primary"))
      })
      
    UI_exist_pfam1 <<- TRUE
    shinyjs::hideElement(id = 'loading.pfam.pf1')
  })
  
  output$output_pfam_table1 <- renderDataTable({
    total_table_pfam <- total_table_pfam1()
    out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
    out_pf_table$description <- sapply(out_pf_table$description, function(x) pfam.link(x))
    colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
    return(out_pf_table)
  }, escape=FALSE, options = list(pageLength = 5))
  
  output$pfam_plot1 <- renderImage({
    total_table_pfam <- total_table_pfam1()
    pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
    pfam_width <- 1000
    pfplot <- pfplot1()
    png("pharaoh_folder/pfam.png",  width = pfam_width, height = pfam_height)
    plot(pfplot)
    dev.off()
    list(src = "pharaoh_folder/pfam.png",
         contentType="image/png")
  }, deleteFile = T
  )
  
  # Download results

  output$pfam_download1 <- downloadHandler(
    filename= function() {
      paste("pfam", ".png", sep="")
    },
    content= function(file) {
      total_table_pfam <- total_table_pfam1()
      pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
      pfam_width <- 1150
      pfplot <- pfplot1()

      png(file, height = pfam_height, width = pfam_width)
      plot(pfplot)
      dev.off()
    })

  output$downloadPFAMTable1<- downloadHandler(
    filename= function() {
      paste("pfam_table", ".tsv", sep="")
    },
    content= function(file) {
      total_table_pfam <- total_table_pfam1()
      out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
      colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
      write.table(x = out_pf_table, quote = F,sep = "\t",
                  file=file,row.names=FALSE,col.names=TRUE)
    })
  
  
  ####################### CAFE #################################
  
  # Remove previous outputs when updated by a new search
  observeEvent(input$run_button1, {
    if (UI_exist_cafe1)
    {
      removeUI(
        selector = "div:has(>> #cafe_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #cafe_mrca1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #cafe_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadCAFEPlot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_cafe1 <<- F
    }
    
    if (UI_exist_error_cafe1)
    {
      removeUI(
        selector = "div:has(>> #cafe_error_message1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_error_cafe1 <<- F
    }
  })
  
  ### CAFE parser and tree generator
  cafe_table1 <- reactive({
    
    shinyjs::showElement(id = 'loading.cafe1')
    
    # Import OG name
    og.cafe <- og.name1()
    
    # Define path to COUNT file
    cafe_comp_tree_file <- "pharaoh_folder/family_table_count.tsv"
    
    # Extract COUNT vector for the current OG
    library(data.table)
    library(ape)
    genecount <- fread(cafe_comp_tree_file, sep="\t", header = T)
    genecountdf <- as.data.frame(genecount)
    rownames(genecountdf) <- genecountdf$Orthogroup
    new_cols <- gsub(" ", "_", colnames(genecountdf))
    colnames(genecountdf) <- new_cols
    
    cafe.vector <- genecountdf[og.cafe,]
    
    if (sum(is.na(cafe.vector)) != 0)
    {
      shinyjs::hideElement(id = 'loading.cafe1')
      if (UI_exist_cafe1)
      {
        removeUI(
          selector = "div:has(>> #cafe_plot1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #cafe_mrca1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #cafe_download1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadCAFEPlot1)",
          multiple = TRUE,
          immediate = TRUE
        )
      }
      
      UI_exist_cafe1 <<- F
      
      if(UI_exist_error_cafe1)
      {
        removeUI(
          selector = "div:has(>> #cafe_error_message1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
      }
      insertUI("#error_cafe1", "afterEnd", ui = {
        box(width = 12,
            title = "Ancestral State Reconstruction", status = "info", solidHeader = TRUE,
            collapsible = TRUE,
            textOutput("cafe_error_message1"))
      })
      
      output$cafe_error_message1 <- renderText({print("No expansions/contraction detected for this orthogroup,
                                                        or infeasible computation due to large size and variance across
                                                        species.")})
      UI_exist_error_cafe1 <<- T
      
      validate(need(sum(is.na(cafe.vector)) == 0 , ""))
    }
    
    return(genecountdf)
  }) %>% bindEvent(input$cafe_start1)
  
  mrca.tree1 <- reactive({
    
    og.cafe <- og.name1()
    genecountdf <- cafe_table1()
    
    # Create phylogenomic tree with internal nodes names
    mrca.tree <- read.tree("pharaoh_folder/tree_with_internal_nodes.txt")
    
    return(mrca.tree)
    
  }) %>% bindEvent(input$cafe_start1)
  
  cafe_tree1 <- reactive({
    og.cafe <- og.name1()
    genecountdf <- cafe_table1()
    cafe.tree <- mrca.tree1()
    
    # Add counts to nodes
    tip.count <- as.numeric(genecountdf[og.cafe,2:172])
    names(tip.count) <- colnames(genecountdf)[2:172]
    
    node.count <- as.numeric(genecountdf[og.cafe,173:342])
    names(node.count) <- colnames(genecountdf)[173:342]
    
    cafe.tree$tip.label[match(names(tip.count), cafe.tree$tip.label)] <- tip.count
    cafe.tree$node.label[match(names(node.count), cafe.tree$node.label)] <- node.count
    
    return(cafe.tree)
    
  }) %>% bindEvent(input$cafe_start1)
  
  evo_plot_data1 <- reactive({
    
    og.cafe <- og.name1()
    genecountdf <- cafe_table1()
    cafe.tree <- mrca.tree1()
    
    # Add counts to nodes
    tip.count <- as.numeric(genecountdf[og.cafe,2:172])
    names(tip.count) <- colnames(genecountdf)[2:172]
    
    node.count <- as.numeric(genecountdf[og.cafe,173:342])
    names(node.count) <- colnames(genecountdf)[173:342]
    
    # Identify parental node to determine if a general expansion or contraction 
    # occured at the level of each node
    edge_table <- as.data.frame(cafe.tree$edge)
    rownames(edge_table) <- paste("edge", 1:nrow(edge_table), sep = "")
    colnames(edge_table) <- c("parent", "child")
    
    exp_cont_tip <- sapply(1:length(tip.count), function(x)
      if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) <
         as.numeric(tip.count[x])) "Expansion"
      else if (as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) >
               as.numeric(tip.count[x])) "Contraction"
      else "No change"
    )
    names(exp_cont_tip) <- names(tip.count)
    
    exp_cont_node <- sapply(names(node.count)[-170], function(x) # 170 is root, no parent
      if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) <
         as.numeric(node.count[x])) "Expansion"
      else if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) >
              as.numeric(node.count[x])) "Contraction"
      else "No change"
    )
    names(exp_cont_node) <- names(node.count)[-170]
    
    # A gain occurs when a node experiences an expansion and its parental exhibits a value of 0
    exp_cont_tip <- sapply(1:length(tip.count), function(x)
      if (exp_cont_tip[x] == "Expansion" &
          as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) == 0
      ) "Gain"
      else exp_cont_tip[x]
    )
    names(exp_cont_tip) <- names(tip.count)
    
    exp_cont_node <- sapply(names(node.count)[-170], function(x)
      if (exp_cont_node[match(x,names(exp_cont_node))] == "Expansion" &
          as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) == 0
      ) "Gain"
      else exp_cont_node[x]
    )
    names(exp_cont_node) <- names(node.count)[-170]
    
    # A loss occurs when a node experiences a contraction and takes the value 0
    exp_cont_tip <- sapply(1:length(tip.count), function(x)
      if (exp_cont_tip[x] == "Contraction" &
          tip.count[x] == 0
      ) "Loss"
      else exp_cont_tip[x]
    )
    names(exp_cont_tip) <- names(tip.count)
    
    exp_cont_node <- sapply(names(node.count)[-170], function(x)
      if (exp_cont_node[match(x,names(exp_cont_node))] == "Contraction" &
          node.count[x] == 0
      ) "Loss"
      else exp_cont_node[x]
    )
    names(exp_cont_node) <- names(node.count)[-170]
    
    # Create complete sorted vectors
    exp_cont_node["root"] <- "No change"
    node_ordered <- exp_cont_node[order(match(names(exp_cont_node),cafe.tree$node.label))]
    tip_ordered <- exp_cont_tip[order(match(names(exp_cont_tip),cafe.tree$tip.label))]
    
    change_vector <- c(tip_ordered, node_ordered)
    
    # Merge tips and nodes reconstruction
    node_ordered_values <- node.count[order(match(names(node.count),cafe.tree$node.label))]
    tip_ordered_values <- tip.count[order(match(names(tip.count),cafe.tree$tip.label))]
    cafe.count <- c(tip_ordered_values, node_ordered_values)
   
    # Create a timeline for a given OG
    tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
    tree.ancestor <- read.tree(tree.name)
    tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
    tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
    tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
    
    mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
    evo.paths <- c()
    for (i in 1:length(unique(tips.orgs)))
    {
      evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
    }
    
    evo.paths <- unique(evo.paths)
    evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
    
    # Associate gray and 0 to reconstruction for nodes not in allowed paths
    change_vector[setdiff(1:length(change_vector), evo.paths)] <- 
    sapply(change_vector[setdiff(1:length(change_vector), evo.paths)], function(x) if(x != "Loss") "OG not present" 
             else x)  
    
    change_vector <- unlist(change_vector)
    
    cafe.count[setdiff(1:length(change_vector), evo.paths)] <- 0
    
    unique(change_vector)
    color_cafe <- sapply(change_vector, function(x) if (x == "No change") "black" 
                         else if (x == "Expansion") "red" else if (x == "Contraction") "blue"
                         else if (x == "Gain") "green3" else if (x == "Loss") "purple"
                         else "gray", USE.NAMES = F)
    
    # Create tree representation
    cafe.table.tips <- data.frame(node = 1:length(cafe.tree$tip.label), label = cafe.tree$tip.label,
                                  col = color_cafe[1:length(cafe.tree$tip.label)], reconst = change_vector[1:length(cafe.tree$tip.label)],
                                  dup_number = cafe.count[1:length(cafe.tree$tip.label)])
    
    cafe.table.nodes <- data.frame(node = (Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label)), label = cafe.tree$node.label,
                                   col = color_cafe[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                   reconst = change_vector[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                   dup_number = cafe.count[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))])
    
    cafe.table.node.comp <- rbind(cafe.table.tips, cafe.table.nodes)
    
    d <- dplyr::mutate(cafe.table.node.comp)
    d$text <- d$label
    
    return(d)
    
  }) %>% bindEvent(input$cafe_start1)
  
  evo_plot1 <- reactive({
    
    d <- evo_plot_data1() 
    mrca.tree <- mrca.tree1()
    
    library(ggtree)
    library(ggplot2)
    
    evo_plot <- ggtree(mrca.tree) %<+% d + aes(colour = I(d$col)) +
      geom_tiplab(aes(label=gsub("_", " ", tools::toTitleCase(d$label))), offset = 30) +
      theme(legend.position = "none") +
      xlim(0, max(mrca.tree$edge.length)*1.85) +
      geom_nodepoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                     alpha = .75) +
      scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
      geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                    alpha = .75)
    
    return(evo_plot)
    
  }) %>% bindEvent(input$cafe_start1)
  
  evo_plotly1 <- reactive({
    
    d <- evo_plot_data1() 
    cafe.tree <- mrca.tree1()
    
    library(ggtree)
    library(ggplot2)
    
    # Remove internal nodes from direct visualization, retaining names in panels
    d$label[172:341] <- NA
    
    evo_plotly <- ggtree(cafe.tree) %<+% d + aes(colour = I(d$col),text=paste0("</br> Duplications: ",dup_number,
                                                                               "</br> Name: ",gsub("_", " ", tools::toTitleCase(d$text)))) + 
      geom_text(aes(x =  1970, label=gsub("_", " ", tools::toTitleCase(d$label)))) + 
      theme(legend.position = "none") +
      xlim(0, max(cafe.tree$edge.length)*1.25) +
      geom_point(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                 alpha = .75) +
      scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
      geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                    alpha = .75)
    
    p <- ggplotly(evo_plotly, tooltip = "text", width = 1400, height = 2200) 
    p <- p %>%
      plotly::style(textposition = "right",xanchor="right")
    
    return(p)
    
  }) %>% bindEvent(input$cafe_start1)
  
  evo.paths.id1 <- reactive({
    
    # Create phylogenomic tree with internal nodes names
    og.cafe <- og.name1()
    cafe.tree <- mrca.tree1()
    
    # Create timeline
    tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
    tree.ancestor <- read.tree(tree.name)
    tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
    tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
    tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
    
    mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
    evo.paths <- c()
    for (i in 1:length(unique(tips.orgs)))
    {
      evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
    }
    
    evo.paths <- unique(evo.paths)
    evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
    
    #print(paste0("Most recent common ancestor for this orthogroup is ", evo.paths.id[1]))
    
    return(evo.paths.id)
    
  }) %>% bindEvent(input$cafe_start1)
  
  # Outputs
  
  # Remove previous boxes if they exist and create new ones
  observeEvent(isTruthy(evo_plot1()), {
    
    if (UI_exist_cafe1)
    {
      removeUI(
        selector = "div:has(>> #cafe_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #cafe_mrca1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #cafe_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadCAFEPlot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
    }
    
    if (UI_exist_error_cafe1)
    {
      removeUI(
        selector = "div:has(>> #cafe_error_message1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
    }
    
    insertUI("#box_cafe1", "afterEnd", ui = {
      box(width = 12,
          title = "Ancestral State Reconstruction", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          plotlyOutput("cafe_plot1", height = ifelse(model.selected1(), "2300px", "2300px")))
    })
    
    insertUI("#box_mrca1", "afterEnd", ui = {
      box(width = 8,
          title = "Most Recent Common Ancestor", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          textOutput("cafe_mrca1")
      )
    })
    
    insertUI("#cafe_down_button1", "afterEnd", ui = {
      tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "cafe_download1", "Download NEWICK tree",
                                                                         size = "sm", color = "primary"))
    })
    
    insertUI("#download_ui_for_cafe_plot1", "afterEnd", ui = {
      tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadCAFEPlot1", "Download Ancestral State Plot",
                                                                         size = "sm", color = "primary"))
    })
    
    UI_exist_cafe1 <<- TRUE
    shinyjs::hideElement(id = 'loading.cafe1')
  })
  
  # Fill outputs
  
  output$cafe_plot1 <- renderPlotly({
    evo_plotly1()
  })
  
  output$cafe_mrca1 <- renderText({
    print(paste0("Most recent common ancestor for this orthogroup is the
                   ancestor of the clade: ", evo.paths.id1()[1]))
  })
  
  # Download tab's results
  
  output$cafe_download1 <- downloadHandler(
    filename= function() {
      paste("ancestral_newick", ".txt", sep="")
    },
    content= function(file) {
      cafe_tree <- cafe_tree1()
      
      write.tree(cafe_tree, file)
    })
  
  output$downloadCAFEPlot1<- downloadHandler(
    filename= function() {
      paste("ancestral_plot", ".png", sep="")
    },
    content= function(file) {
      evo_plot <- evo_plot1()
      
      png(file, width = 1400, height = 2800, res = 100)
      plot(evo_plot)
      dev.off()
    })
  
  ####################### MSA #################################
  
  observeEvent(input$run_button1, {
      removeUI(
        selector = "div:has(>> #selected_msaI1)",
        multiple = TRUE,
        immediate = TRUE
      )
    
    removeUI(
      selector = "div:has(>> #msa_methodI1)",
      multiple = TRUE,
      immediate = TRUE
    )
    
    removeUI("#msa_selection1")
    
    if (UI_exist_msa1)
    {
      removeUI(
        selector = "div:has(>>> #msa_print1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #msa_download_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #msa_download_fa1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_msa1 <<- F
    }
    
  })
  
  observeEvent(input$msa_start1, {
    insertUI("#selected_msa1", "afterEnd", ui = {
      
      shinyWidgets::pickerInput("selected_msaI1","Select the desired genes from the tree to align",
                                choices=isolate({tree_reduced1()$tip.label}), options = list(`actions-box` = TRUE),
                                multiple = T, selected = isolate({str_replace_all(input$geneInt1, fixed(" "), "")}))
      
    })
    
    insertUI("#msa_method1", "afterEnd", ui = {
      shinyWidgets::pickerInput(inputId = "msa_methodI1", label = "Choose alignment method", 
                                choices = c("ClustalOmega", "MAFFT"), selected = "ClustalOmega")
      
    })
    
    insertUI("#msa_selectionI1", "afterEnd", ui = {
      
      shinyWidgets::actionBttn("msa_selection1", "Align Sequences", size = "sm",
                               style = "float", color = "primary")
    })
    
  })
  
  alignseqs1 <- reactive({
    
    library(msa)
    shinyjs::showElement(id = 'loading.msa1')
    
    selected_genes <- as.vector(input$selected_msaI1)
    selected_method <- as.character(input$msa_methodI1)
    file.name <- og.name1()
    
    if (length(selected_genes) < 2)
    {
      shinyjs::hideElement(id = 'loading.msa1')
      if (UI_exist_msa1)
      {
      removeUI(
        selector = "div:has(>>> #msa_print1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #msa_download_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #msa_download_fa1)",
        multiple = TRUE,
        immediate = TRUE
      )
      }
      UI_exist_msa1 <<- F
      output$error_msa1 <- renderUI({renderText({print("Please select at least two genes.")})})
      validate(need(length(selected_genes) > 1, "Please select at least two genes."))
    }
    
    output$error_msa1 <- NULL
    
    # If de novo alignment is selected
    {
    if(selected_method == "ClustalOmega")
    {
    # Define path to orthogroup sequences file
    ortho.seq.name <- paste("Orthogroup_Sequences", paste(file.name, "fa", sep = "."), sep="/")
    
    
    # Read orthogroup sequences file and select the genes for alignment
    mySequences1 <- Biostrings::readAAStringSet(ortho.seq.name)
    mysubseqs <- mySequences1[selected_genes]
    
    alignseqs <- msa(mysubseqs, verbose = F, method = "ClustalOmega")
    }
    
    # If MAFFT alignment is selected
    else
    {
      ortho.seq.name <- paste("MultipleSequenceAlignments", paste(file.name, "fa", sep = "."), sep="/")
      mySequences1 <- seqinr::read.fasta(ortho.seq.name, seqtype = "AA")
      mysubseqs <- mySequences1[selected_genes]
      mysubnames <- seqinr::getName(mySequences1)
      
      # Identify indexes associated with reduced names
      indexes_msa <- sapply(selected_genes, function(x) grep(mysubnames, pattern = x))
      
      # Retrieve those sequences from alignment keeping gaps
      mysubseqs <- mySequences1[indexes_msa]
      names(mysubseqs) <- names(indexes_msa)
      
      # Remove columns with gaps andremove empty spaces in last positions
      seqs_mysubseqs <- seqinr::getSequence(mysubseqs)
      last <- seqs_mysubseqs[[length(seqs_mysubseqs)]]
      last <- last[which(last != " ")]
      seqs_mysubseqs[[length(seqs_mysubseqs)]] <- last
      seqs_mysubseqs <- remove_gaps(seqs_mysubseqs)
      names(seqs_mysubseqs) <- names(mysubseqs)
      
      mysubseqs2 <- unlist(lapply(seqs_mysubseqs, function(x) paste(x, collapse="")))
      
      alignseqs <- Biostrings::AAMultipleAlignment(mysubseqs2, use.names = T)
      
    }
    }
    
    detach("package:msa", unload=TRUE)
    
    return(alignseqs)
    
    }) %>% bindEvent(input$msa_selection1)
  
  # Create boxes for outputs
  observeEvent(isTruthy(alignseqs1()), {
    
    if (UI_exist_msa1)
    {
      removeUI(
        selector = "div:has(>>> #msa_print1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #msa_download_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #msa_download_fa1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
    }
    
    insertUI("#box_msa1", "afterEnd", ui = {
      selected_msa <- isolate({input$selected_msaI1})
      msa_height <- ifelse(length(selected_msa) > 14, 550, 400 + 5*length(selected_msa))
      box(width = 12,
          title = "MSA Explorer", status = "info", solidHeader = TRUE, height = msa_height,
          collapsible = TRUE,
          tags$div(id = "msa_pocket1", style = "width: 1300px; height: 400px",
                   msaROutput("msa_print1"))
          
      )
    })
    
    insertUI("#msa_down_plot1", "afterEnd", ui = {
      tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_plot1", "Download Colored MSA",
                                                                         size = "sm", color = "primary"))
    })
    
    insertUI("#msa_down_fasta1", "afterEnd", ui = {
      tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_fa1", "Download MSA FASTA",
                                                                         size = "sm", color = "primary"))
    })
    
    UI_exist_msa1 <<- TRUE
    #shinyjs::hideElement(id = 'loading.msa1')
  })
  

  # Fill Output
  output$msa_print1 <- renderMsaR({
    alignseqs <- alignseqs1()
    msaout <- msa::msaConvert(alignseqs, "ape::AAbin")
    msaR(msaout, menu=T, overviewbox = F,  colorscheme = "clustal")
  })
  
  # Prepare variables for pdf construction
  observeEvent(isTruthy(alignseqs1()), {
  alignseqs <- alignseqs1()
  
  library(ggmsa)
  class(alignseqs) <- "AAMultipleAlignment"
  
  for(i in 1:(ncol(alignseqs)%/%100 +1)){
    assign(paste("msap", i, sep = ""), ggmsa(alignseqs, 1+(100*(i-1)), i*100, seq_name = TRUE, char_width = 0.5) +
             geom_seqlogo(color = "Chemistry_AA"), envir = as.environment(1), pos=1)
  }
  shinyjs::hideElement(id = 'loading.msa1')
  })
  
  # Download tab's results
  # Download colored MSA in pdf
  output$msa_download_plot1 <- downloadHandler(
    filename= function() {
      paste("msa", ".pdf", sep="")
    },
    content= function(file) {
      selected_msa <- input$selected_msaI1
      alignseqs <- alignseqs1()
      pdf(file, height = 2+length(selected_msa)*0.25, width = 16)
      {
        for(i in 1:(ncol(alignseqs)%/%100 +1)){
          print(mget(paste0("msap", i), envir = as.environment(1)))
        }
        dev.off()
      }
    })
  
  # Download MSA in FASTA format
  output$msa_download_fa1<- downloadHandler(
    filename= function() {
      paste("msa", ".fa", sep="")
    },
    content= function(file) {
      alignseqs <- alignseqs1()
      writeXStringSet(as(unmasked(alignseqs), "XStringSet"), file)
    })
  
  
  ####################### GO #################################
  
  observeEvent(input$run_button1, {
    removeUI(
      selector = "div:has(>> #selected_gosI1)",
      multiple = TRUE,
      immediate = TRUE
    )
    
    removeUI(
      selector = "div:has(>> #selected_gos_modeI1)",
      multiple = TRUE,
      immediate = TRUE
    )
    
    removeUI("#gos_selectionI1")
    
    if (UI_exist_go1)
    {
      removeUI(
        selector = "div:has(>> #output_gos_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #gos_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #gos_treeplot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadGOSTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #gos_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #tree_gos_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_go1 <<- F
    }
    
  })
  
  observeEvent(input$go_start1, {
    insertUI("#selected_gos1", "afterEnd", ui = {
      
      shinyWidgets::pickerInput("selected_gosI1","Select the desired genes from the tree",
                                choices=isolate({tree_reduced1()$tip.label}), options = list(`actions-box` = TRUE),
                                multiple = T, selected = isolate({str_replace_all(input$geneInt1, fixed(" "), "")}))
      
    })
    
    insertUI("#selected_gos_mode1", "afterEnd", ui = {
      
      selectInput(inputId = "selected_gos_modeI1",
                  choices=c("Biological Processes" = "bp",
                            "Molecular Functions" = "mf",
                            "Cellular Components" = "cc"),
                  label = "Select the gene ontology to use",
                  multiple = F, selected = c("bp"))
      
    })
    
    insertUI("#gos_selection1", "afterEnd", ui = {
      
      shinyWidgets::actionBttn("gos_selectionI1", "Show GO terms", size = "sm",
                               style = "float", color = "primary")
    })
    
  })
  
  total_table_gos1 <- reactive({
    
    shinyjs::showElement(id = 'loading.go1')
    library(data.table)
    gos_anot <- as.data.frame(fread("pharaoh_folder/final_anot_table.tsv", sep="\t", header = T))
    sel.genes.go <- as.vector(input$selected_gosI1)
    selected_gos_mode <- as.character(isolate({input$selected_gos_modeI1}))
    
    total_table_gos <- subset(gos_anot, gos_anot$name %in% sel.genes.go)
    
    # Show an error if no terms are identified in the input
      if (nrow(total_table_gos) == 0) 
      {
        if (UI_exist_go1)
        {
          removeUI(
            selector = "div:has(>> #output_gos_table1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #gos_plot1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #gos_treeplot1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadGOSTable1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #gos_download1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #tree_gos_download1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          UI_exist_go1 <<- F
        }
        
        shinyjs::hideElement(id = 'loading.go1')
        output$error_gos1 <- renderUI({
          renderPrint({cat("0 GO terms identified.")})
        })
        
        validate(need(nrow(total_table_gos) != 0, " "))
      }
    
    
    gos_sel <- paste("gos", selected_gos_mode, sep="_")
    terms_sel <- paste("terms", selected_gos_mode, sep="_")
    
    total_table_gos <- total_table_gos[,c("organism", "id", "name", gos_sel, terms_sel)]
    
    
    # Once removed the two GO categories not selected, remove rows with blank cells
    #total_table_gos_clean <- subset(total_table_gos, gos_sel != "")
    total_table_gos_clean <- total_table_gos[ total_table_gos[[gos_sel]] != "" , ]
    
    # Show an error if no terms are identified after the previous operation
    if (nrow(total_table_gos_clean) == 0) 
    {
      shinyjs::hideElement(id = 'loading.go1')
      if (UI_exist_go1)
      {
        removeUI(
          selector = "div:has(>> #output_gos_table1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #gos_plot1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #gos_treeplot1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadGOSTable1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #gos_download1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #tree_gos_download1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        UI_exist_go1 <<- F
      }
      
      output$error_gos1 <- renderUI({
        renderPrint({cat("0 GO terms identified.")})
      })
      
      validate(need(nrow(total_table_gos_clean) != 0, " "))
    }
    
    output$error_gos1 <- NULL
    
    return(total_table_gos_clean)
    
  }) %>% bindEvent(input$gos_selectionI1)
  
  enr_table1 <- reactive({
    
    total_table_gos <- total_table_gos1()
    selected_gos_mode <- as.character(isolate({input$selected_gos_modeI1}))
    
    # Create the plot
    
    # Create a list of GO terms vector of each gene
    gos_sel <- paste("gos", selected_gos_mode, sep="_")
    gos_list <- apply(total_table_gos,MARGIN=1,FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
    names(gos_list) <- total_table_gos$name
    
    # Count GOs for each gene and create a matrix of gene-GO pairs
    count_gos_in_genes <- sapply(gos_list, FUN = function(x) length(x))
    comp_data <- data.frame(gene = rep(names(count_gos_in_genes), count_gos_in_genes), gos = as.character(unlist(gos_list)))
    
    # Collapse genes that share a same GO
    gene.v <- c()
    for (i in 1:length(unique(comp_data$gos)))
    {
      new.table <- subset(comp_data, gos == unique(comp_data$gos)[i])
      new.cha <- as.character(new.table$gene)
      gene.v <- c(gene.v, paste(new.cha, collapse = "/"))
    }
    
    names(gene.v) <- unique(comp_data$gos)
    
    # Load libraries and create gene chains, count and GO IDs fields (same order)
    library(GO.db)
    library("multienrichjam")
    library(clusterProfiler)
    library(enrichplot)
    library(ggplot2)
    
    count_go <- table(comp_data$gos)
    geneids <- gene.v[names(count_go)]
    count_terms <- mapply(function(x) {Term(x)}, names(count_go), USE.NAMES = F)
    
    # If there  exists any obsolote term
    count_terms <- count_terms[which(!is.na(count_terms))] 
    count_go <- count_go[which(!is.na(count_terms))]
    geneids <- geneids[which(!is.na(count_terms))]
    
    # Create pseudo-enrichment table
    enr_table <- data.frame(ID=names(count_go), Description=count_terms, GeneRatio="90/100", BgRatio="90/10000", 
                            pvalue=0.000005, p.adjust=0.000005, qvalue=0.000005, geneID=geneids, Count = as.vector(count_go))
    
    return(enr_table)
    
  }) %>% bindEvent(input$gos_selectionI1)
  
  ema_gos_plot1 <- reactive({
    
    enr_table <- enr_table1()
    
    # Transform to enrichResult object
    enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                 geneColname = "geneID", pvalueColname = "p.adjust",
                                 descriptionColname = "Description", pvalueCutoff = 0.05)
    
    # Create plot
    ema_gos_plot <- emapplot(pairwise_termsim(enr), showCategory = 15) + theme(legend.position='none')
    return(ema_gos_plot)
  })
  
  tree_gos_plot1 <- reactive({
    
    enr_table <- enr_table1()
    
    # Transform to enrichResult object
    enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                 geneColname = "geneID", pvalueColname = "p.adjust",
                                 descriptionColname = "Description", pvalueCutoff = 0.05)
    {
      if (nrow(enr_table) > 4)
      {
        tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(label_words_n = 3)) +
          theme(legend.position='none')
      }
      else if (nrow(enr_table) > 2)
      {
        tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(n = 2, label_words_n = 3)) + 
          theme(legend.position='none')
      }
      else
      {
        text <- paste("\n  Unable to create treeplot with less than 3 GO terms \n")
        tree_gos_plot <- ggplot() + 
          annotate("text", x = 4, y = 25, size=8, label = text) + 
          theme_void()
      }
    }
    
    shinyjs::hideElement(id = 'loading.go1')
    return(tree_gos_plot)
  })
  
  # Create boxes for outputs
  observeEvent(isTruthy(tree_gos_plot1()), {
    
    if (UI_exist_go1)
    {
      removeUI(
        selector = "div:has(>> #output_gos_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #gos_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #gos_treeplot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadGOSTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #gos_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #tree_gos_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
    }
    
    insertUI("#box_gos_table1", "afterEnd", ui = {
      box(width = 12,
          title = "GO Terms Table", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          dataTableOutput("output_gos_table1")
      )
    })
    
    insertUI("#box_gos_plot1", "afterEnd", ui = {
      box(width = 12,
          title = "GO Terms Plot", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          plotOutput("gos_plot1", height = 610)
      )
    })
    
    insertUI("#box_gos_treeplot1", "afterEnd", ui = {
      box(width = 12,
          title = "GO Terms Treeplot", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          plotOutput("gos_treeplot1", height = 610)
      )
    })
    
    
    insertUI("#download_ui_for_gos_table1", "afterEnd", ui = {
      tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadGOSTable1", "Download GO Table",
                                                                         size = "sm", color = "primary"))
    })
    
    insertUI("#gos_down_button1", "afterEnd", ui = {
      tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "gos_download1", "Download GO Plot",
                                                                         size = "sm", color = "primary"))
    })
    
    insertUI("#tree_gos_down_button1", "afterEnd", ui = {
      tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "tree_gos_download1", "Download GO Treeplot",
                                                                         size = "sm", color = "primary"))
    })
    
    UI_exist_go1 <<- TRUE
  })
  
  
  # Fill outputs
  # Table
  output$output_gos_table1 <- renderDataTable({
    total_table_gos <- total_table_gos1()
    selected_gos_mode <- as.character(isolate({input$selected_gos_modeI1}))
    gos_sel <- paste("gos", selected_gos_mode, sep="_")
    gos_list <- apply(total_table_gos,MARGIN=1,
                      FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
    {
    if(!is.list(gos_list))
    {
      gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
      gos_formatted <- paste0(gos_links, collapse = " | ")
      total_table_gos[,gos_sel] <- gos_formatted
      total_table_gos 
    }
    else
    {
      gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
      gos_formatted <- unlist(sapply(gos_links, function(x) paste0(x, collapse = " | ")))
      total_table_gos[,gos_sel] <- gos_formatted
      total_table_gos  
    }
    }
    
  },escape=FALSE, rownames=F, options =list(pageLength = 5))
  
  # First plot
  output$gos_plot1 <- renderImage({
    ema_gos_plot <- ema_gos_plot1()
    
    png("pharaoh_folder/gosplot.png", width = 590, height = 590, res = 90)
    plot(ema_gos_plot)
    dev.off()
    list(src = "pharaoh_folder/gosplot.png",
         contentType="image/png")
    
  }, deleteFile=T
  )
  
  # Second plot
  output$gos_treeplot1 <- renderImage({
    tree_gos_plot <- tree_gos_plot1()
    
    png("pharaoh_folder/treeplot.png", width = 620, height = 590, res = 80)
    plot(tree_gos_plot)
    dev.off()
    list(src = "pharaoh_folder/treeplot.png",
         contentType="image/png")
    
  }, deleteFile=T
  )
  
  # Download tab's results
  # Download GO table
  output$downloadGOSTable1 <- downloadHandler(
    filename= function() {
      paste("GOS_table", ".tsv", sep="")
    },
    content= function(file) {
      total_table_gos <- total_table_gos1()
      
      write.table(x = total_table_gos,quote = F,sep = "\t",
                  file=file,row.names=FALSE,col.names=TRUE)
    })
  
  # Download emaplot
  output$gos_download1 <- downloadHandler(
    filename= function() {
      paste("gos_plot", ".png", sep="")
    },
    content= function(file) {
      ema_gos_plot <- ema_gos_plot1()
      
      png(file, height = 1200, width = 1500, res=140)
      plot(ema_gos_plot)
      dev.off()
    })
  
  # Download treeplot
  output$tree_gos_download1 <- downloadHandler(
    filename= function() {
      paste("gos_treeplot", ".png", sep="")
    },
    content= function(file) {
      tree_gos_plot <- tree_gos_plot1()
      
      png(file, height = 1200, width = 1500, res=140)
      plot(tree_gos_plot)
      dev.off()
    })
  

###################### KEGG ###########################
  
  observeEvent(input$run_button1, {
    removeUI(
      selector = "div:has(>> #selected_kosI1)",
      multiple = TRUE,
      immediate = TRUE
    )
    
    removeUI("#kos_selectionI1")
    
    if (UI_exist_kegg1)
    {
      removeUI(
        selector = "div:has(>> #output_kos_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKOSTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      
      UI_exist_kegg1 <<- F
    }
    
    if (UI_exist_kegg_path1)
    {
      removeUI(
        selector = "div:has(>> #output_kegg_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #selected_pathsI1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKEGGTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI("#paths_buttonI1")
      
      UI_exist_kegg_path1 <<- F
    }
    
    if (UI_exist_pathview1)
    {
      removeUI(
        selector = "div:has(>>> #path_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKEGGpathway1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_pathview1 <<- F
    }
    
    output$error_kos1 <- NULL
    
  })
  
  observeEvent(input$kegg_start1, {
    
    
    if (UI_exist_kegg1)
    {
      removeUI(
        selector = "div:has(>> #output_kos_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKOSTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      
      UI_exist_kegg1 <<- F
    }
    
    if (UI_exist_kegg_path1)
    {
      removeUI(
        selector = "div:has(>> #output_kegg_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #selected_pathsI1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKEGGTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI("#paths_buttonI1")
      
      UI_exist_kegg_path1 <<- F
    }
    
    if (UI_exist_pathview1)
    {
      removeUI(
        selector = "div:has(>>> #path_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKEGGpathway1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_pathview1 <<- F
    }
    
    output$error_kos1 <- NULL
    
    insertUI("#selected_kos1", "afterEnd", ui = {
      
      shinyWidgets::pickerInput("selected_kosI1","Select the desired genes from the tree",
                                choices=isolate({tree_reduced1()$tip.label}), options = list(`actions-box` = TRUE),
                                multiple = T, selected = isolate({str_replace_all(input$geneInt1, fixed(" "), "")}))
      
      
    })
    
    
    insertUI("#kos_selection1", "afterEnd", ui = {
      
      shinyWidgets::actionBttn("kos_selectionI1", "Show KEGG pathways", size = "sm",
                               style = "float", color = "primary")
    })
    
  })
  
  tab_kegg1 <- reactive({
    
    shinyjs::showElement(id = 'loading.ko1')
    # Create KOs set
    kos_anot <- read.csv("pharaoh_folder/ko_table_funtree.tsv", sep="\t", header = T)
    sel.genes.ko <- as.vector(isolate({input$selected_kosI1}))
    
    tab_kegg <- subset(kos_anot, gene %in% sel.genes.ko)
    set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
    
    # Show an error if no terms are identified in the input
    {
      if (length(set_kegg) == 0) 
      {
        shinyjs::hideElement(id = 'loading.ko1')
        output$error_kos1 <- renderUI({
          renderPrint({cat("0 KO terms identified. Please select more genes. If this 
        message persists, it should be interpreted as a lack of KO annotation for this orthogroup")})
        })
        
        if (UI_exist_kegg1)
        {
          removeUI(
            selector = "div:has(>> #output_kos_table1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadKOSTable1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          
          UI_exist_kegg1 <<- F
        }
        
        if (UI_exist_kegg_path1)
        {
          removeUI(
            selector = "div:has(>> #output_kegg_table1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #selected_pathsI1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadKEGGTable1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI("#paths_buttonI1")
          
          UI_exist_kegg_path1 <<- F
        }
        
        if (UI_exist_pathview1)
        {
          removeUI(
            selector = "div:has(>>> #path_image1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadKEGGpathway1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          UI_exist_pathview1 <<- F
        }
        
        
        validate("No KO terms detected")
      }
    }
    
    #output$error_kos1 <- NULL
    
    return(tab_kegg)
    
  }) %>% bindEvent(input$kos_selectionI1)
  
  total_table_kegg1 <- reactive({
    
    tab_kegg <- tab_kegg1()
    set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
    
    # Load libraries
    library(clusterProfiler)
    library(enrichplot)
    
    # Enrich with pvalue cutoff = 1 to show all paths
    kos_enrich <- enrichKEGG(gene         = unique(unlist(strsplit(set_kegg, "/"))),
                             organism     = 'ko',
                             pvalueCutoff = 1)
    
    
    total_table_kegg <- as.data.frame(kos_enrich)
    
    # Show an error if the KOs are not mapped to any KEGG pathway
    {
      if (nrow(total_table_kegg) == 0) 
      {
        shinyjs::hideElement(id = 'loading.ko1')
        output$error_kos1 <- renderUI({
          renderPrint({cat("No KEGG pathway appears in this OG.")})
        })
        
        
        if (UI_exist_kegg_path1)
        {
          removeUI(
            selector = "div:has(>> #output_kegg_table1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #selected_pathsI1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadKEGGTable1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI("#paths_buttonI1")
          
          UI_exist_kegg_path1 <<- F
        }
        
        if (UI_exist_pathview1)
        {
          removeUI(
            selector = "div:has(>>> #path_image1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          removeUI(
            selector = "div:has(>> #downloadKEGGpathway1)",
            multiple = TRUE,
            immediate = TRUE
          )
          
          UI_exist_pathview1 <<- F
        }
        
        validate("No KEGG pathway appears in this OG.")
      }
    }
    
    output$error_kos1 <- NULL
    
    #pruebatest <- sapply(strsplit(total_table_kegg$ID, "map"), function(x) x[[2]])#only for testing
    #total_table_kegg$ID <- paste0("ko", pruebatest)
    # Filter out pathways that are not present in plants
    kegg_plants <- read.csv("pharaoh_folder/pathways_plant.ids", sep = "\t", header = T)$x
    total_table_kegg <- subset(total_table_kegg, ID %in% kegg_plants)
    
    
    return(total_table_kegg)
    
  }) %>% bindEvent(input$kos_selectionI1)
  
  total_table_kos1 <- reactive({
    
    tab_kegg <- tab_kegg1()
    
    library(KEGGREST)
    
    # Collapse genes that share KOs
    
    tab_kegg_for_ko <- subset(tab_kegg, ko != "")
    gene.v.ko <- c()
    individual_kos <- unique(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
    
    for (i in 1:length(individual_kos))
    {
      new.table <- subset(tab_kegg_for_ko, grepl(individual_kos[i],ko))
      new.cha <- as.character(new.table$gene)
      gene.v.ko <- c(gene.v.ko, paste(new.cha, collapse = "/"))
    }
    
    names(gene.v.ko) <- individual_kos
    
    # Create gene chains, count and KO IDs fields (same order)
    count_ko <- table(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
    geneids.ko <- gene.v.ko[names(count_ko)]
    count_terms.ko <- mapply(function(x) {keggFind("ko", x)}, names(count_ko), USE.NAMES = F)
    total_table_kos <- data.frame(ko=names(count_ko), name=count_terms.ko, count=as.numeric(count_ko),
                                  genes=geneids.ko)
    
    return(total_table_kos)
    
  }) %>% bindEvent(input$kos_selectionI1)
  
  # Create boxes for outputs
  observeEvent(isTruthy(total_table_kos1()), {
    
    if (UI_exist_kegg1)
    {
      removeUI(
        selector = "div:has(>> #output_kos_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKOSTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      
      UI_exist_kegg1 <<- F
    }
    
    insertUI("#box_kos_table1", "afterEnd", ui = {
      box(width = 12,
          title = "KO Terms Table", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          dataTableOutput("output_kos_table1")
      )
    })
    
    insertUI("#download_ui_for_kos_table1", "afterEnd", ui = {
      tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKOSTable1", "Download KO Table",
                                                                         size = "sm", color = "primary"))
    })
    
    UI_exist_kegg1 <<- TRUE
  })
  
  observeEvent(isTruthy(total_table_kegg1()), {
    
    if (UI_exist_kegg_path1)
    {
      removeUI(
        selector = "div:has(>> #output_kegg_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #selected_pathsI1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKEGGTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI("#paths_buttonI1")
      
      UI_exist_kegg_path1 <<- F
    }
    
    
    insertUI("#box_kegg_table1", "afterEnd", ui = {
      box(width = 12,
          title = "KEGG Pathways Table", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          dataTableOutput("output_kegg_table1")
      )
    })
    
    
    
    insertUI("#download_ui_for_kegg_table1", "afterEnd", ui = {
      tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKEGGTable1", "Download Pathways Table",
                                                                         size = "sm", color = "primary"))
    })
    
    UI_exist_kegg_path1 <<- TRUE
    
    # Remove previous results for pathview
    if (UI_exist_pathview1)
    {
      removeUI(
        selector = "div:has(>>> #path_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKEGGpathway1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_pathview1 <<- F
    }
  })
  
  # Fill outputs
  # Render KO table
  output$output_kos_table1 <- renderDataTable({
    total_table_kos <- total_table_kos1()
    total_table_kos$ko <- sapply(total_table_kos$ko, ko.link)
    total_table_kos
  },escape=FALSE, rownames= F, options =list(pageLength = 5))
  
  # Render KEGG table
  output$output_kegg_table1 <- renderDataTable({
    total_table_kegg <- total_table_kegg1()
    total_table_kegg$ID <- sapply(total_table_kegg$ID, kegg.link)
    total_table_kegg[,c("ID", "Description", "geneID")]
  },escape=FALSE, rownames= F, options =list(pageLength = 5))
  
  # Download tab's outputs
  # Download KO table
  output$downloadKOSTable1 <- downloadHandler(
    filename= function() {
      paste("KO_table", ".tsv", sep="")
    },
    content= function(file) {
      total_table_kos <- total_table_kos1()
      write.table(x = total_table_kos,quote = F,sep = "\t",
                  file=file,row.names=FALSE,col.names=TRUE)
    })
  
  # Download KEGG table
  output$downloadKEGGTable1 <- downloadHandler(
    filename= function() {
      paste("KEGG_table", ".tsv", sep="")
    },
    content= function(file) {
      total_table_kegg <- total_table_kegg1()
      write.table(x = total_table_kegg[,c("ID", "Description", "geneID")],quote = F,sep = "\t",
                  file=file,row.names=FALSE,col.names=TRUE)
    })
  
  # Create pathway selector and button
  observeEvent(input$kos_selectionI1,{
    
    total_table_kos <- total_table_kos1()
    total_table_kegg <- total_table_kegg1()
    
    if(nrow(total_table_kegg) != 0)
    {
      paths.options <- sapply(strsplit(total_table_kegg$ID, split = "ko"), function(x) x[[2]])
      
      
      insertUI("#selected_paths1", "afterEnd", ui = {
        shinyWidgets::pickerInput(inputId = "selected_pathsI1", label = "Select the pathway to plot", 
                                  choices = paths.options, selected = paths.options[1], multiple = F)
        
      })
        
      
      insertUI("#paths_button1", "afterEnd", ui = {
        
        shinyWidgets::actionBttn("paths_buttonI1", "Plot Pathway", size = "sm",
                                 style = "float", color = "primary")
      })
      
      shinyjs::hideElement(id = 'loading.ko1')
    }
  })
  
  observeEvent(input$paths_buttonI1,{
    
    if (UI_exist_pathview1)
    {
      removeUI(
        selector = "div:has(>>> #path_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadKEGGpathway1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_pathview1 <<- F
    }
    
    
  })
  
  # Create Image to Render and save path name
  pathway.current.id1 <- reactive({
    pathway.current.id <- input$selected_pathsI1
    total_table_kos <- total_table_kos1()
    
    kos_unique <- unique(total_table_kos$ko)
    gene.pathway <- rep(0, length(kos_unique))
    names(gene.pathway) <-  kos_unique
    gene.pathway[kos_unique] <-1
    
    library(pathview)
    pathview(gene.data = sort(gene.pathway,decreasing = TRUE),kegg.dir = "pharaoh_folder",
             pathway.id = pathway.current.id,
             species = "ko",
             limit = list(gene=max(abs(gene.pathway)), cpd=1),
             gene.idtype ="kegg")
    
    return(pathway.current.id)
    
  }) %>% bindEvent(input$paths_buttonI1)
  
  # Create output box and download button
  observeEvent(isTruthy(pathway.current.id1()),{
    
    insertUI("#box_path_image1", "afterEnd", ui = {
      box(width = 12,
          title = "KEGG Pathway Plot", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          fluidRow(column(1), imageOutput("path_image1", width = "100%", height = "700px"))
      )
    })
    
    insertUI("#path_download_ui1", "afterEnd", ui = {
      tags$div(shinyWidgets::downloadBttn(outputId= "downloadKEGGpathway1", "Download KEGG Pathway Plot",
                                          size = "sm", color = "primary"))
    })
    
    UI_exist_pathview1 <<- T
    
  })
  
  # Fill path image output
  output$path_image1 <- renderImage({
    
    pathway.current.id <- pathway.current.id1()
    list(src = paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."),
         contentType="image/png",width=900,height=700)
  },deleteFile = F)
  
  # Download and remove path image output
  output$downloadKEGGpathway1 <- downloadHandler(
    filename= function() {
      paste("path_plot", ".png", sep="")
    },
    content= function(file) {
      pathway.current.id <- pathway.current.id1()
      file.copy(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."), file)
      file.remove(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."))
    })
  
############################# LITERATURE ANNOTATION ##########################################
  
  observeEvent(input$run_button1, {
    removeUI(
      selector = "div:has(>> #selected_litI1)",
      multiple = TRUE,
      immediate = TRUE
    )
    
    removeUI(
      selector = "div:has(> #query_litI1)",
      multiple = TRUE,
      immediate = TRUE
    )
    
    removeUI("#lit_selectionI1")
    
    
    if (UI_exist_lit1)
    {
      removeUI(
        selector = "div:has(>> #output_lit_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadLITTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_lit1 <<- F
    }
    
  })
  
  observeEvent(input$lit_start1, {
    
    insertUI("#query_lit1", "afterEnd", ui = {
      
    textInput(inputId = "query_litI1",value = "", label = "Enter search term", placeholder = "CCA1")
      
    })
    
    
    insertUI("#selected_lit1", "afterEnd", ui = {
      
      shinyWidgets::pickerInput("selected_litI1","Select the search mode",
                                choices=c("Normal","Exact", "Substring", "Alias"), 
                                #options = list(`actions-box` = TRUE),
                                multiple = F, selected = "Normal")
      
      
    })
    
    
    insertUI("#lit_selection1", "afterEnd", ui = {
      
      shinyWidgets::actionBttn("lit_selectionI1", "Get biological information and papers", size = "sm",
                               style = "float", color = "primary")
    })
    
  })
  
  pc_result1 <- reactive({
    
    shinyjs::showElement(id = 'loading.lit1')
    pc_search <- as.character(input$query_litI1)
    pc_search <- gsub(" ", "%20", pc_search) # To make spaces interpretable
    pc_modality <- tolower(as.character(input$selected_litI1))
    
    # Get PlantConnectome URL for query
    pc_url <- paste(c("https://connectome.plant.tools", pc_modality, pc_search), collapse = "/")
    

    library(RCurl)
    
    pc_res <- getURL(pc_url)
    
    if (!length(grep("No hits", pc_res)) == 0)
    {
       shinyjs::hideElement(id = 'loading.lit1')
        
      if (UI_exist_lit1)
      {
        removeUI(
          selector = "div:has(>> #output_lit_table1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        removeUI(
          selector = "div:has(>> #downloadLITTable1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        UI_exist_lit1 <<- F
      }
      
      output$error_lit1 <- renderUI({
        renderPrint({cat("No results found for this query")})
        })
      
      validate(" ")
      
    }
    
    output$error_lit1 <- NULL
        
    # Isolate data frame from complete HTML file
    pc_split <- strsplit(as.character(pc_res), split = "<tbody")[[1]][2]
    pc_split <- strsplit(as.character(pc_split), split = "</tbody>")[[1]][1]
        
    pc_clean <- gsub("</tr>", "", pc_split)
    pc_clean <- gsub("[\r\n\t]", "", pc_clean)
        
    pc_vector <- strsplit(pc_clean, split = "<tr>")[[1]][-1]
    pc_vector2 <- sapply(pc_vector, FUN=function(x) strsplit(x, split = "<td> | </td>"))
    pc_table <- sapply(pc_vector2, FUN=function(x) as.character(unlist(x)))
        
    colnames(pc_table) <- NULL
    pc_result <- data.frame(t(pc_table[c(2,4,6,8),]))
    colnames(pc_result) <- c("Source", "Interaction Type", "Target", "Pubmed ID")

    return(pc_result)
    
  }) %>% bindEvent(input$lit_selectionI1)
  
  pc_result_show1 <- reactive({
    
    pc_result <- pc_result1()
    
    # Add links to papers
    urls_connect <- sapply(pc_result$`Pubmed ID`, FUN = function(x) paste0(c("<a href=\"",
                                                                             "https://pubmed.ncbi.nlm.nih.gov/",x,"/",
                                                                             "\" target=\"_blank\">", x,
                                                                             "</a>"),
                                                                           collapse=""))
    pc_result_show <- pc_result
    pc_result_show$`Pubmed ID` <- urls_connect
    
    #shinyjs::hideElement(id = 'loading.lit1')
    
    return(pc_result_show)
    
  })
  
  # Create boxes for outputs
  observeEvent(isTruthy(pc_result_show1()), {
    
    if (UI_exist_lit1)
    {
      removeUI(
        selector = "div:has(>> #output_lit_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadLITTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
    }
    
    insertUI("#box_lit_table1", "afterEnd", ui = {
      box(width = 12,
          title = "Literature Table", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          dataTableOutput("output_lit_table1")
      )
    })
    
    insertUI("#download_ui_for_lit_table1", "afterEnd", ui = {
      tags$div(style = "margin-left: 400px;", shinyWidgets::downloadBttn(outputId= "downloadLITTable1", "Download Literature Table",
                                                                         size = "sm", color = "primary"))
    })
    
    shinyjs::hideElement(id = 'loading.lit1')
    UI_exist_lit1 <<- TRUE
    
  })
  
  # Fill outputs
  # Render table
  output$output_lit_table1 <- renderDataTable({
    pc_result_show1()
  },escape=FALSE, rownames= F, options =list(pageLength = 10))
  
  # Download results
  output$downloadLITTable1 <- downloadHandler(
    filename= function() {
      paste("literature_table", ".tsv", sep="")
    },
    content= function(file) {
      pc_result <- pc_result1()
      write.table(x = pc_result,quote = F,sep = "\t",
                  file=file,row.names=FALSE,col.names=TRUE)
    })
  
  ######################## STRING ###########################
  
  observeEvent(input$run_button1, {
    removeUI(
      selector = "div:has(>> #selected_stringI1)",
      multiple = TRUE,
      immediate = TRUE
    )
    
    removeUI("#string_selectionI1")
    
    shinyjs::hideElement("error_string1")
    
    
    if (UI_exist_string1)
    {
      removeUI(
        selector = "div:has(>> #output_st_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #output_count_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>>> #count_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadSTRINGTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadCOUNTTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #count_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #selected_networkI1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI("#network_buttonI1")
      
      UI_exist_string1 <<- F
    }
    
    if (UI_exist_network1)
    {
      removeUI(
        selector = "div:has(>>>> #network_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_network1 <<- F
    }
    
  })
  
  # First, create a table that links reduced gene names to its species
  string_sel_table1 <- reactive({
    
    # Define previous variables
    tree_reduced <- tree_reduced1()
    tree <- tree_adj1()
    
    tips_to_keep.mp <- tips_to_keep.mp1()
    tips_to_keep.ot <- tips_to_keep.ot1()
    tips_to_keep.at <- tips_to_keep.at1()
    tips_to_keep.cp <- tips_to_keep.cp1()
    tips_to_keep.cr <- tips_to_keep.cr1()
    tips_to_keep.cz <- tips_to_keep.cz1()
    tips_to_keep.kn <- tips_to_keep.kn1()
    tips_to_keep.me <- tips_to_keep.me1()
    tips_to_keep.mi <- tips_to_keep.mi1()
    tips_to_keep.pp <- tips_to_keep.pp1()
    tips_to_keep.sl <- tips_to_keep.sl1()
    tips_to_keep.sm <- tips_to_keep.sm1()
    tips_to_keep.sp <- tips_to_keep.sp1()
    tips_to_keep.ta <- tips_to_keep.ta1()
    tips_to_keep.vc <- tips_to_keep.vc1()
    tips_to_keep.bp <- tips_to_keep.bp1()
    tips_to_keep.cri <- tips_to_keep.cri1()
    tips_to_keep.ds <- tips_to_keep.ds1()
    tips_to_keep.os <- tips_to_keep.os1()
    tips_to_keep.smag <- tips_to_keep.smag1()
    tips_to_keep.tp <- tips_to_keep.tp1()
    tips_to_keep.aa <- tips_to_keep.aa1()
    tips_to_keep.um <- tips_to_keep.um1()
    tips_to_keep.rs <- tips_to_keep.rs1()
    tips_to_keep.cyc <- tips_to_keep.cyc1()
    tips_to_keep.ca <- tips_to_keep.ca1()
    tips_to_keep.mv <- tips_to_keep.mv1()
    tips_to_keep.af <- tips_to_keep.af1()
    tips_to_keep.sc <- tips_to_keep.sc1()
    tips_to_keep.aegi <- tips_to_keep.aegi1()
    tips_to_keep.sb <- tips_to_keep.sb1()
    tips_to_keep.chara <- tips_to_keep.chara1()
    tips_to_keep.sceobli <- tips_to_keep.sceobli1()
    tips_to_keep.cocco <- tips_to_keep.cocco1()
    tips_to_keep.haema <- tips_to_keep.haema1()
    tips_to_keep.zm <- tips_to_keep.zm1()
    tips_to_keep.praco <- tips_to_keep.praco1()
    tips_to_keep.ostlu <- tips_to_keep.ostlu1()
    tips_to_keep.ostme <- tips_to_keep.ostme1()
    tips_to_keep.micco <- tips_to_keep.micco1()
    tips_to_keep.cymte <- tips_to_keep.cymte1()
    tips_to_keep.chlin <- tips_to_keep.chlin1()
    tips_to_keep.gonpe <- tips_to_keep.gonpe1()
    tips_to_keep.tetso <- tips_to_keep.tetso1()
    tips_to_keep.desar <- tips_to_keep.desar1()
    tips_to_keep.monmi <- tips_to_keep.monmi1()
    tips_to_keep.caule <- tips_to_keep.caule1()
    tips_to_keep.ostqu <- tips_to_keep.ostqu1()
    tips_to_keep.brysp <- tips_to_keep.brysp1()
    tips_to_keep.ellbi <- tips_to_keep.ellbi1()
    tips_to_keep.myrbi <- tips_to_keep.myrbi1()
    tips_to_keep.apalo <- tips_to_keep.apalo1()
    tips_to_keep.astgl <- tips_to_keep.astgl1()
    tips_to_keep.tresp <- tips_to_keep.tresp1()
    tips_to_keep.picso <- tips_to_keep.picso1()
    tips_to_keep.chlso <- tips_to_keep.chlso1()
    tips_to_keep.auxpr <- tips_to_keep.auxpr1()
    tips_to_keep.helsp <- tips_to_keep.helsp1()
    tips_to_keep.tetst <- tips_to_keep.tetst1()
    tips_to_keep.pedsp <- tips_to_keep.pedsp1()
    tips_to_keep.chlpr <- tips_to_keep.chlpr1()
    tips_to_keep.picsp <- tips_to_keep.picsp1()
    tips_to_keep.mesvb <- tips_to_keep.mesvb1()
    tips_to_keep.meskr <- tips_to_keep.meskr1()
    tips_to_keep.zygci <- tips_to_keep.zygci1()
    tips_to_keep.zygcb <- tips_to_keep.zygcb1()
    tips_to_keep.zygcy <- tips_to_keep.zygcy1()
    tips_to_keep.plesc <- tips_to_keep.plesc1()
    tips_to_keep.funhy <- tips_to_keep.funhy1()
    tips_to_keep.sphfa <- tips_to_keep.sphfa1()
    tips_to_keep.takle <- tips_to_keep.takle1()
    tips_to_keep.marpa <- tips_to_keep.marpa1()
    tips_to_keep.ricfl <- tips_to_keep.ricfl1()
    tips_to_keep.ricso <- tips_to_keep.ricso1()
    tips_to_keep.antpu <- tips_to_keep.antpu1()
    tips_to_keep.antfu <- tips_to_keep.antfu1()
    tips_to_keep.megfl <- tips_to_keep.megfl1()
    tips_to_keep.phach <- tips_to_keep.phach1()
    tips_to_keep.phyph <- tips_to_keep.phyph1()
    tips_to_keep.parpe <- tips_to_keep.parpe1()
    tips_to_keep.notor <- tips_to_keep.notor1()
    tips_to_keep.phaca <- tips_to_keep.phaca1()
    tips_to_keep.phasp <- tips_to_keep.phasp1()
    tips_to_keep.leidu <- tips_to_keep.leidu1()
    tips_to_keep.isosi <- tips_to_keep.isosi1()
    tips_to_keep.dipco <- tips_to_keep.dipco1()
    tips_to_keep.hupas <- tips_to_keep.hupas1()
    tips_to_keep.lyccl <- tips_to_keep.lyccl1()
    tips_to_keep.adica <- tips_to_keep.adica1()
    tips_to_keep.alssp <- tips_to_keep.alssp1()
    tips_to_keep.marve <- tips_to_keep.marve1()
    tips_to_keep.seqse <- tips_to_keep.seqse1()
    tips_to_keep.seqgi <- tips_to_keep.seqgi1()
    tips_to_keep.taxch <- tips_to_keep.taxch1()
    tips_to_keep.abial <- tips_to_keep.abial1()
    tips_to_keep.picab <- tips_to_keep.picab1()
    tips_to_keep.gnemo <- tips_to_keep.gnemo1()
    tips_to_keep.welmi <- tips_to_keep.welmi1()
    tips_to_keep.ginbi <- tips_to_keep.ginbi1()
    tips_to_keep.ambtr <- tips_to_keep.ambtr1()
    tips_to_keep.nymco <- tips_to_keep.nymco1()
    tips_to_keep.cinka <- tips_to_keep.cinka1()
    tips_to_keep.horvu <- tips_to_keep.horvu1()
    tips_to_keep.bradi <- tips_to_keep.bradi1()
    tips_to_keep.setit <- tips_to_keep.setit1()
    tips_to_keep.panha <- tips_to_keep.panha1()
    tips_to_keep.missi <- tips_to_keep.missi1()
    tips_to_keep.oroth <- tips_to_keep.oroth1()
    tips_to_keep.eleco <- tips_to_keep.eleco1()
    tips_to_keep.anaco <- tips_to_keep.anaco1()
    tips_to_keep.musac <- tips_to_keep.musac1()
    tips_to_keep.dioal <- tips_to_keep.dioal1()
    tips_to_keep.spipo <- tips_to_keep.spipo1()
    tips_to_keep.aspof <- tips_to_keep.aspof1()
    tips_to_keep.zosma <- tips_to_keep.zosma1()
    tips_to_keep.araly <- tips_to_keep.araly1()
    tips_to_keep.capgr <- tips_to_keep.capgr1()
    tips_to_keep.brara <- tips_to_keep.brara1()
    tips_to_keep.braol <- tips_to_keep.braol1()
    tips_to_keep.sisir <- tips_to_keep.sisir1()
    tips_to_keep.schpa <- tips_to_keep.schpa1()
    tips_to_keep.eutsa <- tips_to_keep.eutsa1()
    tips_to_keep.thlar <- tips_to_keep.thlar1()
    tips_to_keep.boest <- tips_to_keep.boest1()
    tips_to_keep.carpa <- tips_to_keep.carpa1()
    tips_to_keep.theca <- tips_to_keep.theca1()
    tips_to_keep.goshi <- tips_to_keep.goshi1()
    tips_to_keep.gosra <- tips_to_keep.gosra1()
    tips_to_keep.citsi <- tips_to_keep.citsi1()
    tips_to_keep.pontr <- tips_to_keep.pontr1()
    tips_to_keep.eucgr <- tips_to_keep.eucgr1()
    tips_to_keep.maldo <- tips_to_keep.maldo1()
    tips_to_keep.prupe <- tips_to_keep.prupe1()
    tips_to_keep.frave <- tips_to_keep.frave1()
    tips_to_keep.corav <- tips_to_keep.corav1()
    tips_to_keep.caril <- tips_to_keep.caril1()
    tips_to_keep.queru <- tips_to_keep.queru1()
    tips_to_keep.cucsa <- tips_to_keep.cucsa1()
    tips_to_keep.glyma <- tips_to_keep.glyma1()
    tips_to_keep.medtr <- tips_to_keep.medtr1()
    tips_to_keep.tripr <- tips_to_keep.tripr1()
    tips_to_keep.cicar <- tips_to_keep.cicar1()
    tips_to_keep.lotja <- tips_to_keep.lotja1()
    tips_to_keep.phaac <- tips_to_keep.phaac1()
    tips_to_keep.vigun <- tips_to_keep.vigun1()
    tips_to_keep.lupal <- tips_to_keep.lupal1()
    tips_to_keep.lencu <- tips_to_keep.lencu1()
    tips_to_keep.arahy <- tips_to_keep.arahy1()
    tips_to_keep.poptr <- tips_to_keep.poptr1()
    tips_to_keep.manes <- tips_to_keep.manes1()
    tips_to_keep.salpu <- tips_to_keep.salpu1()
    tips_to_keep.linus <- tips_to_keep.linus1()
    tips_to_keep.corci <- tips_to_keep.corci1()
    tips_to_keep.vitvi <- tips_to_keep.vitvi1()
    tips_to_keep.kalfe <- tips_to_keep.kalfe1()
    tips_to_keep.vacda <- tips_to_keep.vacda1()
    tips_to_keep.oleeu <- tips_to_keep.oleeu1()
    tips_to_keep.mimgu <- tips_to_keep.mimgu1()
    tips_to_keep.soltu <- tips_to_keep.soltu1()
    tips_to_keep.cofar <- tips_to_keep.cofar1()
    tips_to_keep.lacsa <- tips_to_keep.lacsa1()
    tips_to_keep.dauca <- tips_to_keep.dauca1()
    tips_to_keep.betvu <- tips_to_keep.betvu1()
    tips_to_keep.amahy <- tips_to_keep.amahy1()
    tips_to_keep.chequ <- tips_to_keep.chequ1()
    tips_to_keep.sapof <- tips_to_keep.sapof1()
    tips_to_keep.aquco <- tips_to_keep.aquco1()
    
    # Table construction
    org.factor <- c()
    
    for (i in 1:length(tree_reduced$tip.label))
    {
      if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
      {
        org.factor <- c(org.factor,"Marchantia polymorpha")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
      {
        org.factor <- c(org.factor,"Ostreococcus tauri")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
      {
        org.factor <- c(org.factor,"Arabidopsis thaliana")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
      {
        org.factor <- c(org.factor,"Ceratodon purpureus")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
      {
        org.factor <- c(org.factor,"Chlamydomonas reinhardtii")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
      {
        org.factor <- c(org.factor,"Chromochloris zofingiensis")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
      {
        org.factor <- c(org.factor,"Klebsormidium nitens")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
      {
        org.factor <- c(org.factor,"Mesotaenium endlicherianum")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
      {
        org.factor <- c(org.factor,"Micromonas pusilla")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
      {
        org.factor <- c(org.factor,"Physcomitrium patens")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
      {
        org.factor <- c(org.factor,"Solanum lycopersicum")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
      {
        org.factor <- c(org.factor,"Selaginella moellendorffii")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
      {
        org.factor <- c(org.factor,"Spirogloea muscicola")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
      {
        org.factor <- c(org.factor,"Triticum aestivum")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
      {
        org.factor <- c(org.factor,"Volvox carteri")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
      {
        org.factor <- c(org.factor,"Bathycoccus prasinos")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
      {
        org.factor <- c(org.factor,"Ceratopteris richardii")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
      {
        org.factor <- c(org.factor,"Dunaliella salina")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
      {
        org.factor <- c(org.factor,"Oryza sativa")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
      {
        org.factor <- c(org.factor,"Sphagnum magellanicum")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
      {
        org.factor <- c(org.factor,"Thuja plicata")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
      {
        org.factor <- c(org.factor,"Anthoceros agrestis")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
      {
        org.factor <- c(org.factor,"Ulva mutabilis")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
      {
        org.factor <- c(org.factor,"Raphidocelis subcapitata")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
      {
        org.factor <- c(org.factor,"Cycas panzhihuaensis")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
      {
        org.factor <- c(org.factor,"Chlorokybus atmophyticus")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
      {
        org.factor <- c(org.factor,"Mesostigma viride CCAC 1140")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
      {
        org.factor <- c(org.factor,"Azolla filiculoides")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
      {
        org.factor <- c(org.factor,"Salvinia cucullata")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
      {
        org.factor <- c(org.factor,"Aegilops tauschii")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
      {
        org.factor <- c(org.factor,"Sorghum bicolor")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
      {
        org.factor <- c(org.factor,"Chara braunii")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
      {
        org.factor <- c(org.factor,"Scenedesmus obliquus")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
      {
        org.factor <- c(org.factor,"Coccomyxa subellipsoidea")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
      {
        org.factor <- c(org.factor,"Haematococcus lacustris")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
      {
        org.factor <- c(org.factor,"Zea mays")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.praco])
      {
        org.factor <- c(org.factor,"Prasinoderma coloniale")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostlu])
      {
        org.factor <- c(org.factor,"Ostreococcus lucimarinus")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostme])
      {
        org.factor <- c(org.factor,"Ostreococcus mediterraneus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.micco])
      {
        org.factor <- c(org.factor,"Micromonas commoda")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymte])
      {
        org.factor <- c(org.factor,"Cymbomonas tetramitiformis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlin])
      {
        org.factor <- c(org.factor,"Chlamydomonas incerta")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gonpe])
      {
        org.factor <- c(org.factor,"Gonium pectorale")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetso])
      {
        org.factor <- c(org.factor,"Tetrabaena socialis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.desar])
      {
        org.factor <- c(org.factor,"Desmodesmus armatus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.monmi])
      {
        org.factor <- c(org.factor,"Monoraphidium minutum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caule])
      {
        org.factor <- c(org.factor,"Caulerpa lentillifera")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostqu])
      {
        org.factor <- c(org.factor,"Ostreobium quekettii")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brysp])
      {
        org.factor <- c(org.factor,"Bryopsis sp.")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ellbi])
      {
        org.factor <- c(org.factor,"Elliptochloris bilobata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.myrbi])
      {
        org.factor <- c(org.factor,"Myrmecia bisecta")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.apalo])
      {
        org.factor <- c(org.factor,"Apatococcus lobatus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.astgl])
      {
        org.factor <- c(org.factor,"Asterochloris glomerata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tresp])
      {
        org.factor <- c(org.factor,"Trebouxia sp.")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picso])
      {
        org.factor <- c(org.factor,"Picochlorum soloecismus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlso])
      {
        org.factor <- c(org.factor,"Chlorella sorokiniana")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.auxpr])
      {
        org.factor <- c(org.factor,"Auxenochlorella protothecoides")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.helsp])
      {
        org.factor <- c(org.factor,"Helicosporidium sp.")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetst])
      {
        org.factor <- c(org.factor,"Tetraselmis striata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pedsp])
      {
        org.factor <- c(org.factor,"Pedinophyceae sp.")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlpr])
      {
        org.factor <- c(org.factor,"Chloropicon primus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picsp])
      {
        org.factor <- c(org.factor,"Picocystis sp.")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mesvb])
      {
        org.factor <- c(org.factor,"Mesostigma viride NIES 296")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.meskr])
      {
        org.factor <- c(org.factor,"Mesotaenium kramstae")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygci])
      {
        org.factor <- c(org.factor,"Zygnema circumcarinatum SAG")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcb])
      {
        org.factor <- c(org.factor,"Zygnema circumcarinatum UTEX 1559")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcy])
      {
        org.factor <- c(org.factor,"Zygnema cylindricum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.plesc])
      {
        org.factor <- c(org.factor,"Pleurozium schreberi")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.funhy])
      {
        org.factor <- c(org.factor,"Funaria hygrometrica")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sphfa])
      {
        org.factor <- c(org.factor,"Sphagnum fallax")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.takle])
      {
        org.factor <- c(org.factor,"Takakia lepidozioides")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marpa])
      {
        org.factor <- c(org.factor,"Marchantia paleacea")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricfl])
      {
        org.factor <- c(org.factor,"Riccia fluitans")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricso])
      {
        org.factor <- c(org.factor,"Riccia sorocarpa")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antpu])
      {
        org.factor <- c(org.factor,"Anthoceros punctatus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antfu])
      {
        org.factor <- c(org.factor,"Anthoceros fusiformis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.megfl])
      {
        org.factor <- c(org.factor,"Megaceros flagellaris")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phach])
      {
        org.factor <- c(org.factor,"Phaeomegaceros chiloensis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phyph])
      {
        org.factor <- c(org.factor,"Phymatoceros phymatodes")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.parpe])
      {
        org.factor <- c(org.factor,"Paraphymatoceros pearsonii")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.notor])
      {
        org.factor <- c(org.factor,"Notothylas orbicularis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaca])
      {
        org.factor <- c(org.factor,"Phaeoceros carolinianus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phasp])
      {
        org.factor <- c(org.factor,"Phaeoceros sp.")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.leidu])
      {
        org.factor <- c(org.factor,"Leiosporoceros dussii")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.isosi])
      {
        org.factor <- c(org.factor,"Isoetes sinensis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dipco])
      {
        org.factor <- c(org.factor,"Diphasiastrum complanatum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.hupas])
      {
        org.factor <- c(org.factor,"Huperzia asiatica")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lyccl])
      {
        org.factor <- c(org.factor,"Lycopodium clavatum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.adica])
      {
        org.factor <- c(org.factor,"Adiantum capillus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.alssp])
      {
        org.factor <- c(org.factor,"Alsophila spinulosa")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marve])
      {
        org.factor <- c(org.factor,"Marsilea vestita")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqse])
      {
        org.factor <- c(org.factor,"Sequoia sempervirens")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqgi])
      {
        org.factor <- c(org.factor,"Sequoiadendron giganteum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.taxch])
      {
        org.factor <- c(org.factor,"Taxus chinensis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.abial])
      {
        org.factor <- c(org.factor,"Abies alba")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picab])
      {
        org.factor <- c(org.factor,"Picea abies")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gnemo])
      {
        org.factor <- c(org.factor,"Gnetum montanum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.welmi])
      {
        org.factor <- c(org.factor,"Welwitschia mirabilis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ginbi])
      {
        org.factor <- c(org.factor,"Ginkgo biloba")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ambtr])
      {
        org.factor <- c(org.factor,"Amborella trichopoda")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.nymco])
      {
        org.factor <- c(org.factor,"Nymphaea colorata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cinka])
      {
        org.factor <- c(org.factor,"Cinnamomum kanehirae")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.horvu])
      {
        org.factor <- c(org.factor,"Hordeum vulgare")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bradi])
      {
        org.factor <- c(org.factor,"Brachypodium distachyon")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.setit])
      {
        org.factor <- c(org.factor,"Setaria italica")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.panha])
      {
        org.factor <- c(org.factor,"Panicum hallii")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.missi])
      {
        org.factor <- c(org.factor,"Miscanthus sinensis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oroth])
      {
        org.factor <- c(org.factor,"Oropetium thomaeum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eleco])
      {
        org.factor <- c(org.factor,"Eleusine coracana")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.anaco])
      {
        org.factor <- c(org.factor,"Ananas comosus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.musac])
      {
        org.factor <- c(org.factor,"Musa acuminata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dioal])
      {
        org.factor <- c(org.factor,"Dioscorea alata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.spipo])
      {
        org.factor <- c(org.factor,"Spirodela polyrhiza")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aspof])
      {
        org.factor <- c(org.factor,"Asparagus officinalis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zosma])
      {
        org.factor <- c(org.factor,"Zostera marina")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.araly])
      {
        org.factor <- c(org.factor,"Arabidopsis lyrata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.capgr])
      {
        org.factor <- c(org.factor,"Capsella grandiflora")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brara])
      {
        org.factor <- c(org.factor,"Brassica rapa")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.braol])
      {
        org.factor <- c(org.factor,"Brassica oleracea")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sisir])
      {
        org.factor <- c(org.factor,"Sisymbrium irio")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.schpa])
      {
        org.factor <- c(org.factor,"Schrenkiella parvula")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eutsa])
      {
        org.factor <- c(org.factor,"Eutrema salsugineum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.thlar])
      {
        org.factor <- c(org.factor,"Thlaspi arvense")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.boest])
      {
        org.factor <- c(org.factor,"Boechera stricta")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.carpa])
      {
        org.factor <- c(org.factor,"Carica papaya")
      }
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.theca])
      {
        org.factor <- c(org.factor,"Theobroma cacao")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.goshi])
      {
        org.factor <- c(org.factor,"Gossypium hirsutum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gosra])
      {
        org.factor <- c(org.factor,"Gossypium raimondii")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.citsi])
      {
        org.factor <- c(org.factor,"Citrus sinensis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pontr])
      {
        org.factor <- c(org.factor,"Poncirus trifoliata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eucgr])
      {
        org.factor <- c(org.factor,"Eucalyptus grandis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.maldo])
      {
        org.factor <- c(org.factor,"Malus domestica")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.prupe])
      {
        org.factor <- c(org.factor,"Prunus persica")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.frave])
      {
        org.factor <- c(org.factor,"Fragaria vesca")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corav])
      {
        org.factor <- c(org.factor,"Corylus avellana")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caril])
      {
        org.factor <- c(org.factor,"Carya illinoinensis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.queru])
      {
        org.factor <- c(org.factor,"Quercus rubra")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cucsa])
      {
        org.factor <- c(org.factor,"Cucumis sativus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.glyma])
      {
        org.factor <- c(org.factor,"Glycine max")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.medtr])
      {
        org.factor <- c(org.factor,"Medicago truncatula")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tripr])
      {
        org.factor <- c(org.factor,"Trifolium pratense")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cicar])
      {
        org.factor <- c(org.factor,"Cicer arietinum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lotja])
      {
        org.factor <- c(org.factor,"Lotus japonicus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaac])
      {
        org.factor <- c(org.factor,"Phaseolus acutifolius")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vigun])
      {
        org.factor <- c(org.factor,"Vigna unguiculata")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lupal])
      {
        org.factor <- c(org.factor,"Lupinus albus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lencu])
      {
        org.factor <- c(org.factor,"Lens culinaris")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.arahy])
      {
        org.factor <- c(org.factor,"Arachis hypogaea")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.poptr])
      {
        org.factor <- c(org.factor,"Populus trichocarpa")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.manes])
      {
        org.factor <- c(org.factor,"Manihot esculenta")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.salpu])
      {
        org.factor <- c(org.factor,"Salix purpurea")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.linus])
      {
        org.factor <- c(org.factor,"Linum usitatissimum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corci])
      {
        org.factor <- c(org.factor,"Corymbia citriodora")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vitvi])
      {
        org.factor <- c(org.factor,"Vitis vinifera")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kalfe])
      {
        org.factor <- c(org.factor,"Kalanchoe fedtschenkoi")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vacda])
      {
        org.factor <- c(org.factor,"Vaccinium darrowii")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oleeu])
      {
        org.factor <- c(org.factor,"Olea europaea")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mimgu])
      {
        org.factor <- c(org.factor,"Mimulus guttatus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.soltu])
      {
        org.factor <- c(org.factor,"Solanum tuberosum")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cofar])
      {
        org.factor <- c(org.factor,"Coffea arabica")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lacsa])
      {
        org.factor <- c(org.factor,"Lactuca sativa")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dauca])
      {
        org.factor <- c(org.factor,"Daucus carota")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.betvu])
      {
        org.factor <- c(org.factor,"Beta vulgaris")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.amahy])
      {
        org.factor <- c(org.factor,"Amaranthus hypochondriacus")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chequ])
      {
        org.factor <- c(org.factor,"Chenopodium quinoa")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sapof])
      {
        org.factor <- c(org.factor,"Saponaria officinalis")
      }
      
      else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aquco])
      {
        org.factor <- c(org.factor,"Aquilegia coerulea")
      }
      
      
    }
    
    #Matrix with labels and colors and transform to dplyr format
    string.sel.table <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                                   org = org.factor)
    
    return(string.sel.table)
    
  }) %>% bindEvent(input$string_start1)
  
  # Now, selection is allowed for genes of species with STRING support
  observeEvent(input$string_start1, {
    
    string_sel_table <- string_sel_table1()
    allow_string_species <- c("Aegilops tauschii", "Arabidopsis thaliana", "Bathycoccus prasinos", "Chara braunii", 
      "Chlamydomonas reinhardtii","Coccomyxa subellipsoidea", "Klebsormidium nitens", 
      "Micromonas commoda", "Oryza sativa", "Micromonas pusilla", "Ostreococcus lucimarinus",
      "Ostreococcus tauri", "Physcomitrium patens", "Raphidocelis subcapitata",
      "Scenedesmus obliquus", "Selaginella moellendorffii", "Solanum lycopersicum",
      "Sorghum bicolor", "Triticum aestivum", "Volvox carteri")
    
    st_genes <- subset(string_sel_table, string_sel_table$org %in% allow_string_species)$label
    
    if (length(st_genes) == 0)
    {
      shinyjs::showElement("error_string1")
      output$error_string1 <- renderUI({renderText({print("No results for this
      analysis due to lack of genes of STRING-supported species in the selection.")})})
      validate(" ")
    }
    
    output$error_string1 <- NULL
    
    insertUI("#selected_string1", "afterEnd", ui = {
      
      shinyWidgets::pickerInput("selected_stringI1","Select the desired genes from the tree",
                                choices=st_genes, options = list(`actions-box` = TRUE),
                                multiple = T)
      
    })
    
    
    insertUI("#string_selection1", "afterEnd", ui = {
      
      shinyWidgets::actionBttn("string_selectionI1", "Show STRING Interactions", size = "sm",
                               style = "float", color = "primary")
    })
    
  })
  
  phys_table1 <- reactive({
    
    shinyjs::showElement(id = 'loading.string1')
    query_genes <- input$selected_stringI1
    string_sel_table <- string_sel_table1()
    
    # Load complete STRING annotation table
    library(data.table)
    
    # # Load iteratively from split files using data.table format
    # data_phys <- fread("pharaoh_folder/string_physical/string_physical_1.tsv")
    # 
    # for (x in list.files("pharaoh_folder/string_physical/")[-1])
    # {
    #   data_phys <- rbind(data_phys, 
    #                      fread(paste0("pharaoh_folder/string_physical/", x)))
    # }
    
    # Load data only for selected organisms
    species_selected <- gsub("[.]", "", gsub(" ", "_", tolower(unique(string_sel_table$org))))
    
    # Get a List of all files named with a key word, say all `.csv` files
    filenames_string <- list.files("pharaoh_folder/string_physical/")
    files_selected <- filenames_string[unlist(sapply(species_selected, function(x) grep(x, filenames_string)))]
    
    # If in the future any of these species is added to STRING:
    # { 
    #   if (seq_org == "Zygnema circumcarinatum SAG")
    #   {
    #     org_search <- "zygnema_circumcarinatum"
    #   }
    #   else if (seq_org == "Zygnema circumcarinatum UTEX 1559")
    #   {
    #     org_search <- "zygnema_circumcarinatum1559"
    #   }
    #   else if (seq_org == "Mesostigma viride CCAC 1140")
    #   {
    #     org_search <- "mesostigma_viride1140"
    #   }
    #   else if (seq_org == "Mesostigma viride NIES 296")
    #   {
    #     org_search <- "mesostigma_viride296"
    #   }
    #   else
    #   {
    #    org_search <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(seq_org))))
    #   }
    # }
    
    # Load and bind all data sets
    data_phys <- rbindlist(lapply(paste0("pharaoh_folder/string_physical/", files_selected),fread))
    
    
    # Subset by query genes using data.table for speed
    #string_res <- subset(data_phys, data_phys$prot_query %in% query_genes)
    string_res <- data_phys[prot_query %in% query_genes,]
    string_res <- as.data.frame(string_res)
    
    # Assign OG ID to each target
    ortho_data_file <- "Gene_Trees/Orthogroups.tsv"
    
    ortho_data <- as.data.frame(fread(ortho_data_file))
    
    ortho_char <- apply(ortho_data, MARGIN = 1, function(x) paste(x, collapse = ","))
    ortho.numbers <- sapply(string_res$prot_interaction, function(x) grep(x, ortho_char), USE.NAMES = F)
    
    # If a pattern is found in several names, i.e., is a subpattern of several genes,
    # search for the exact match
    if(class(ortho.numbers) == "list")
    {
      # If a gene isn't associated to an OG
      index.none <- which(sapply(ortho.numbers, function(x) length(x) == 0))
      
      if (length(index.none) != 0)
      {
        # We create another row for OG table to associate those genes
        ortho_data <- rbind(ortho_data, "No OG")
        ortho.numbers[index.none] <- nrow(ortho_data)
      }
      
      # If a gene has more than one match due to subpatterns
      index.wrong <- which(sapply(ortho.numbers, function(x) length(x) > 1))
      {
        if (length(index.wrong) == 0)
        {
          ortho.numbers <- unlist(ortho.numbers, use.names = F)
        }
        else
        {
          for (i in index.wrong)
          {
            for (j in ortho.numbers[[i]])
            {
              ortho_char_split <- strsplit(ortho_char[j],split = ",")
              ortho_char_split_clean <- sapply(ortho_char_split, function(x) gsub(" ", "", x))
              if (string_res$prot_interaction[i] %in% ortho_char_split_clean)
              {
                ortho.numbers[i] <- j
                ortho.numbers <- unlist(ortho.numbers, use.names = F)
              }
            }
          }
        }
      }
    }
    
    
    
    # Create the final table
    ortho.string.names <- ortho_data$Orthogroup[ortho.numbers]
    phys_table <- data.frame(string_res, orthogroup = ortho.string.names)
    
    return(phys_table)
  }) %>% bindEvent(input$string_selectionI1)
  
  # Create count table to identify enriched OGs in STRING result
  string_counts1 <- reactive({
    
    phys_table <- phys_table1()
    string_counts <- sort(table(phys_table$orthogroup), decreasing = T)
    
    return(string_counts)
    
  }) %>% bindEvent(input$string_selectionI1)
  
  string_count_plot1 <- reactive({
    
    library(ggplot2)
    library(dplyr)
    
    data_count <- as.data.frame(string_counts1())
    colnames(data_count) <- c("orthogroup", "value")
    
    # Compute the position of labels
    data_count <- data_count %>%
      arrange(desc(orthogroup)) %>%
      mutate(prop = value / sum(data_count$value) *100) %>%
      mutate(ypos = cumsum(prop)- 0.5*prop )
    
    # Create plot
    count_plot <- ggplot(data_count, aes(x="", y=prop, fill=orthogroup)) +
      geom_bar(stat="identity", width=1, color="white") +
      coord_polar("y", start=0) +
      theme_void() +
      theme(legend.position="none") +
      #geom_text(aes(y = ypos, label = orthogroup), color = "white", size=6) +
      scale_fill_manual(values = rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"), 
                                     floor(nrow(data_count)/9)+1))
    
    return(count_plot)
    
  }) %>% bindEvent(input$string_selectionI1)
  
  string_count_plotly1 <- reactive({
    
    library(ggplot2)
    library(dplyr)
    
    data_count <- as.data.frame(string_counts1())
    colnames(data_count) <- c("orthogroup", "value")
    
    # Compute the position of labels
    data_count <- data_count %>%
      arrange(desc(orthogroup)) %>%
      mutate(prop = value / sum(data_count$value) *100) %>%
      mutate(ypos = cumsum(prop)- 0.5*prop )
    
    # Create plot
    count_plotly <- plotly::plot_ly(data=data_count,values=~prop,labels=~factor(orthogroup),
                    marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                           floor(nrow(data_count)/9)+1)),
                    type="pie",showlegend = F, text= ~paste0("</br> ", orthogroup,
                                                            "</br> ",prop, "%"),
                    textinfo = "none", hoverinfo = "text") 
    
    
    
    return(count_plotly)
    
  }) %>% bindEvent(input$string_selectionI1)
  
  # Create boxes for outputs
  observeEvent(isTruthy(string_count_plot1()), {
    
    if (UI_exist_string1)
    {
      removeUI(
        selector = "div:has(>> #output_st_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #output_count_table1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>>> #count_plot1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadSTRINGTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #downloadCOUNTTable1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #count_download1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI(
        selector = "div:has(>> #selected_networkI1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      removeUI("#network_buttonI1")
      
    }
    
    insertUI("#box_st_table1", "afterEnd", ui = {
      box(width = 12,
          title = "STRING Interactions Table", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          dataTableOutput("output_st_table1")
      )
    })
    
    insertUI("#box_count_table1", "afterEnd", ui = {
      box(width = 12,
          title = "Interacting Orthogroups Table", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          dataTableOutput("output_count_table1")
      )
    })
    
    insertUI("#box_count_plot1", "afterEnd", ui = {
      box(width = 12,
          title = "Interacting Orthogroups Plot", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          plotlyOutput("count_plot1")
      )
    })
    
    
    insertUI("#download_ui_for_st_table1", "afterEnd", ui = {
      tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadSTRINGTable1", "Download STRING Table",
                                                                         size = "sm", color = "primary"))
    })
    
    insertUI("#download_ui_for_count_table1", "afterEnd", ui = {
      tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadCOUNTTable1", "Download OG Count Table",
                                                                         size = "sm", color = "primary"))
    })
    
    insertUI("#count_down_button1", "afterEnd", ui = {
      tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "count_download1", "Download OG Count Plot",
                                                                         size = "sm", color = "primary"))
    })
    
    UI_exist_string1 <<- TRUE
    
    # Remove previous results for STRING network
    if (UI_exist_network1)
    {
      removeUI(
        selector = "div:has(>>>> #network_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
      UI_exist_network1 <<- F
    }
    
  })
  
  # Fill outputs
  # Render STRING table

  output$output_st_table1 <- renderDataTable({
    phys_table <- phys_table1()
    #phys_table$type <- sapply(phys_table$type, color_string)
    datatable(phys_table, escape=FALSE, rownames= F, options =list(pageLength = 10)) %>%
      formatStyle(
        'type',
        color = styleEqual(
          c("Direct interaction", "Interolog"), c('green', '#CA931B')
        )
      )
  }) 
  
  
  # Render OG count table
  output$output_count_table1 <- renderDataTable({
    string_counts <- as.data.frame(string_counts1())
    colnames(string_counts) <- c("orthogroup", "count")
    string_counts
  },escape=FALSE, rownames= F, options =list(pageLength = 7))
  
  # Render OG count pie chart
  output$count_plot1 <- renderPlotly({
    string_count_plotly1()
    
  })
  
  
  # Download tab's outputs
  # Download STRING table
  output$downloadSTRINGTable1 <- downloadHandler(
    filename= function() {
      paste("string_table", ".tsv", sep="")
    },
    content= function(file) {
      phys_table <- phys_table1()
      write.table(x = phys_table,quote = F,sep = "\t",
                  file=file,row.names=FALSE,col.names=TRUE)
    })
  
  # Download count table
  output$downloadCOUNTTable1 <- downloadHandler(
    filename= function() {
      paste("string_count_table", ".tsv", sep="")
    },
    content= function(file) {
      string_counts <- string_counts1()
      write.table(x = string_counts,quote = F,sep = "\t",
                  file=file,row.names=FALSE,col.names=TRUE)
    })
  
  # Download count plot
  output$count_download1 <- downloadHandler(
    filename= function() {
      paste("string_count", ".png", sep="")
    },
    content= function(file) {
      string_count_plot <- string_count_plot1()
      
      png(file, height = 450, width = 450)
      plot(string_count_plot)
      dev.off()
    })
  
  # Create gene selector and button for STRING network representation
  observeEvent(input$string_selectionI1,{
    
    phys_table <- phys_table1()
    network_genes <- unique(phys_table$prot_query)
    
    # Error message if no genes are allowed for selection should have been reported
    # earlier
    
    insertUI("#selected_network1", "afterEnd", ui = {
      
      shinyWidgets::pickerInput(inputId = "selected_networkI1", label = "Select the gene whose network you want to plot", 
                                choices = network_genes, selected = network_genes[1],
                                options = list(`actions-box` = TRUE), multiple = T)
      
    })
    
    
    insertUI("#network_button1", "afterEnd", ui = {
      
      shinyWidgets::actionBttn("network_buttonI1", "Plot STRING Network", size = "sm",
                               style = "float", color = "primary")
    })
    
    shinyjs::hideElement(id = 'loading.string1')
    
  })

  
  mapped_string1 <- reactive({
    
    # Load genes (PharaohFUN IDs) and convert to STRING IDs
    library(data.table)
    
    network_genes <- input$selected_networkI1
    map_table <- fread("pharaoh_folder/string_map.tsv")
    
    # Fast subset using data.table
    map_network <- map_table[pharaohfun_id %in% network_genes,]
    
    # Error if no genes from selection have an associated high fidelity STRING ID
    if (nrow(map_network) == 0)
    {
      
      if (UI_exist_network1)
      {
        removeUI(
          selector = "div:has(>>>> #network_image1)",
          multiple = TRUE,
          immediate = TRUE
        )
        
        UI_exist_network1 <<- F
      }
      
      output$error_network1 <- renderUI({renderText({print("It's not possible to map any
                               genes from selection to STRING IDs, please select different ones.")})})
      validate( " ")
      
    }
    
    output$error_network1 <- NULL
    
    # Get only STRING IDS and paste in a format interpretable by JS function
    string_ids <- as.data.frame(map_network)$string_id
    
    
    mapped_string <- paste0(string_ids, collapse = "%0d")
    return(mapped_string)
    
  }) %>% bindEvent(input$network_buttonI1)
  
  # Create boxes
  observeEvent(isTruthy(mapped_string1()),{
    
    mapped_string <- mapped_string1()
    
    if (UI_exist_network1)
    {
      removeUI(
        selector = "div:has(>>>> #network_image1)",
        multiple = TRUE,
        immediate = TRUE
      )
      
    }
    
    url_interactive <- paste0("https://string-db.org/cgi/network?identifiers=", 
                              mapped_string, 
                              "&add_color_nodes=25&network_flavor=confidence&show_query_node_labels=1")
    
    insertUI("#box_output_network1", "afterEnd", ui = {
      box(width = 12,
          title = "Image", status = "info", solidHeader = TRUE,
          collapsible = TRUE,
          fluidRow(column(1), 
                   column(8, htmlOutput("network_image1")), 
                   column(3, div( style = "margin-top: 300px;", 
                            shinyWidgets::actionBttn("network_link1", "Interactive Network", size = "md", 
                                                     icon = icon("circle-nodes"), style = "float", color = "primary", 
                                                        onclick=paste0("window.open('", url_interactive,"','_blank')")))
                          
                          ))
          
      )
    })
    
    UI_exist_network1 <<- T
    
  })
  
  # Fill network box
  
  output$network_image1 <- renderText({
    
    mapped_string <- mapped_string1()
    src_map <- paste0("https://string-db.org/api/image/network?identifiers=", mapped_string, 
                      "&add_color_nodes=25&network_flavor=confidence")
    
    c('<img src="',src_map,'"width="675" height="625">')
    })
  

# End of Gene ID-based search results
  
  
  
  ############### SEQUENCE-BASED SEARCH ####################
  
  UI_exist_pfam2 <<- F
  UI_exist_tree2 <<- F
  UI_exist_phylo2 <<- F
  UI_exist_cafe2 <<- F
  UI_exist_error_cafe2 <<- F
  UI_exist_msa2 <<- F
  UI_exist_go2 <<- F
  UI_exist_kegg2 <<- F
  UI_exist_kegg_path2 <<- F
  UI_exist_pathview2 <<- F
  UI_exist_lit2 <<-  F
  UI_exist_string2 <<-  F
  UI_exist_network2 <<-  F
  
  # Define variables for input values, update only when Run button is clicked
  model.selected2 <- reactive({
    model.selected <- F
    return(model.selected)
  })%>% bindEvent(input$run_button2)
  
  build_trees2 <- reactive({
    build_trees <- as.character(input$build_trees_2)
    return(build_trees)
  }) %>% bindEvent(input$run_button2)
  
  selected_organisms2 <- reactive({
    selected_organisms <- c(input$prasino_check_2, input$mami_check_2,
                            input$chloro_check_2, input$strepto_check_2,
                            input$bryo_check_2, input$lyco_check_2,
                            input$gymno_check_2, input$magno_check_2,
                            input$monocots_check_2, input$dicots_first_check_2,
                            input$dicots_second_check_2)
    return(selected_organisms)
    
  }) %>% bindEvent(input$run_button2)
  
  selected_values_org2 <- reactive(organisms_values[selected_organisms2()]) %>% bindEvent(input$run_button2)

  
   seq_org2 <- reactive({
     
     seq_org <- input$organism_input_2
     
     return(seq_org)
     
   }) %>% bindEvent(input$run_button2)
    
   # Start results computation
   # First, search for the most similar protein in the proteome of the species
   # chosen by the user to identify the corresponding ID. After this, workflow is
   # the same as on the Gene ID-based search
   
   diamond_table2 <- reactive({
     
     shinyjs::showElement(id = 'loading.tree2')
     
     seq_org <- seq_org2()
     model.selected <- model.selected2()
     
     # Check if selected model contains selected proteome
     # Clean chain input
     library(seqinr)
     library(stringr)
     
     seq_comp <- as.character(input$geneInt2)
     seq_comp_clean <- toupper(gsub("[\r\n\t]", "", seq_comp))
     seq_comp_clean2 <- str_replace_all(seq_comp_clean, fixed(" "), "")
     vec_comp <- stringr::str_split_1(seq_comp_clean2, pattern = "")
     
     # Create fasta por comparison
     random_name <- "new_diamond"
     write.fasta(vec_comp, names = "query_prot", paste0("pharaoh_folder/", random_name, ".fa"))
     
     # Access species proteome
     
     
     # Some proteomes are not covered by the previous command
     { 
     if (seq_org == "Zygnema circumcarinatum SAG")
     {
       org_search <- "zygnema_circumcarinatum"
     }
     else if (seq_org == "Zygnema circumcarinatum UTEX 1559")
     {
       org_search <- "zygnema_circumcarinatum1559"
     }
     else if (seq_org == "Mesostigma viride CCAC 1140")
     {
       org_search <- "mesostigma_viride1140"
     }
     else if (seq_org == "Mesostigma viride NIES 296")
     {
       org_search <- "mesostigma_viride296"
     }
     else
     {
       org_search <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(seq_org))))
     }
     }
     
     org_fasta <- paste("pharaohfun_proteomes", paste0(org_search, ".fa", sep=""), sep = "/")
     
     library(rdiamond)
     
     
     # Run diamond
     diamond_res <- NULL
     
     try(
     diamond_res <- rdiamond::diamond_protein_to_protein(
       query   = paste0("pharaoh_folder/", random_name, ".fa"),
       subject = org_fasta,
       sensitivity_mode = "sensitive",
       output_path = tempdir(),
       #db_import  = FALSE,
       #diamond_exec_path = "/usr/bin", 
       max_target_seqs = 5)
     )
    
     # Convert to data.frame
     diamond_table <- data.frame(ID=diamond_res$subject_id, Identity_perc=diamond_res$perc_identity,
                                 Num_identical_matches=diamond_res$num_ident_matches, 
                                 Length = diamond_res$alig_length, Mismatches = diamond_res$mismatches,
                                 Num_gaps=diamond_res$n_gaps, E_value=diamond_res$evalue)
    
     # Error if no sequence is detected
     if (nrow(diamond_table) == 0)
     {
       shinyjs::hideElement(id = 'loading.tree2')
       
       if (UI_exist_tree2)
       {
         removeUI(
           selector = "div:has(>> #tree_seq_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #treeTips2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs2)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree2 <<- F
       output$error_tree2 <- renderUI({renderText({print("No matches for the query were detected in the selected proteome.")})})
       validate(" ")
     }
     
     return(diamond_table)
     
     
   }) %>% bindEvent(input$run_button2)
  
   
   og.name2 <- reactive({
     
     diamond_table <- diamond_table2()
     gene.name.tree <- as.character(diamond_table$ID[1])
       
       
     # Load table with orthogroups information depending on selected model
     ortho.table.search <- "Gene_Trees/Orthogroups.tsv"
     ortho.table <- read.csv(ortho.table.search,
                             header = T, sep = "\t", as.is = T,
                             fill = T, blank.lines.skip = F)
     
     
     # Find orthogroup of target gene
     found <- F
     file.name <- NULL
     i <- 1
     while (!(found) && (i <= nrow(ortho.table)))
     {
       object <- as.character(ortho.table[i,])
       gene.number <- grep(pattern = gene.name.tree, object)
       if (length(gene.number) != 0)
       {
         found <- T
         file.name <- ortho.table[i,1]
       }
       else
       {
         i <- i+1
       }
     }
     
     # Error message if gene ID does not belong to any OG
     if (is.null(file.name))
     {
       shinyjs::hideElement(id = 'loading.tree2')
       
       if (UI_exist_tree2)
       {
         removeUI(
           selector = "div:has(>> #tree_seq_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #treeTips2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs2)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree2 <<- F
       output$error_tree2 <- renderUI({renderText({print("No results for this 
      query due to not supported gene name or lack of orthologs in the selected organisms.")})})
       validate(need(!is.null(file.name), " "))
     }
     
     return(file.name)
     
   }) %>% bindEvent(input$run_button2)
   
   
   tree2 <- reactive({
     file.name <- og.name2()
     # Load gene tree file depending on the input
     
     tree.name <- paste("Gene_Trees",paste(file.name, "tree.txt", sep = "_"), sep="/")
     
     # Error if tree file not found
     if (!(file.exists(tree.name)))
     {
       shinyjs::hideElement(id = 'loading.tree2')
       
       if (UI_exist_tree2)
       {
         removeUI(
           selector = "div:has(>> #tree_seq_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #treeTips2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs2)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree2 <<- F
       output$error_tree2 <- renderUI({renderText({print("Unable to construct tree associated to
                                                        to an orthogroup with less than 4 genes.")})})
       validate(need(file.exists(tree.name), " "))
     }
     
     # Generate tree depending on the tree building method selected
     {
       if (build_trees2() == "Maximum Likelihood")
       {
         tree <- read.tree(tree.name)
         return(tree)
       }
       
       else if(build_trees2() == "Bayesian Inference")
       {
         tree.bayes <-  paste("Tree_Bayes",paste0("bayes_tree_", file.name, ".nwk"), sep="/")
         
         # Error if tree file not found
         if (!(file.exists(tree.bayes)))
         {
           shinyjs::hideElement(id = 'loading.tree2')
           
           if (UI_exist_tree2)
           {
             removeUI(
               selector = "div:has(>> #tree_seq_table2)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #treeTips2)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>>> #presentorg2)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #tree_image2)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #downloadTree2)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #downloadNewick2)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #downloadTreeSeqs2)",
               multiple = TRUE,
               immediate = TRUE
             )
           }
           
           UI_exist_tree2 <<- F
           output$error_tree2 <- renderUI({renderText({print("Bayesian tree inference is currently under development. 
                                                          We are working to offer this method for all 
                                                          orthogroups through regular updates. If this message appears,
                                                          the orthogroup of interest is not yet available. If you want it
                                                          to appear in the next update, please send an email to mramos5@us.es 
                                                          and we will try to prioritize it. Meanwhile, try the other methods
                                                          to build the gene tree.")})})
           validate(need(file.exists(tree.bayes), " "))
         }
         
         tree <- read.tree(tree.bayes)
         return(tree)
         
       }
       
       else
       {
         library(phangorn)
         
         # Read MSA
         multiple_file <- paste("MultipleSequenceAlignments", paste0(file.name, ".fa"), sep="/")
         
         
         mytree <- read.phyDat(file = multiple_file,
                               format="fasta", type = "AA")
         
         # To keep the same workflow, the reduced tree will be calculated, then the applied subsetting
         # will only reduce them if the selected method has been fasttree
         
         
         # Selection of genes from the selected organism
         
         # Create empty vectors
         
         tips_to_keep.praco2 = tips_to_keep.ot2 = tips_to_keep.ostlu2 = tips_to_keep.ostme2 = tips_to_keep.bp2 <- c()
         tips_to_keep.mi2 = tips_to_keep.micco2 = tips_to_keep.cymte2 = tips_to_keep.cr2 = tips_to_keep.chlin2 <- c()
         tips_to_keep.haema2 = tips_to_keep.vc2 = tips_to_keep.ds2 = tips_to_keep.gonpe2 = tips_to_keep.tetso2 <- c()
         tips_to_keep.cz2 = tips_to_keep.rs2 = tips_to_keep.sceobli2 = tips_to_keep.desar2 = tips_to_keep.monmi2 <- c()
         tips_to_keep.caule2 = tips_to_keep.ostqu2 = tips_to_keep.brysp2 = tips_to_keep.cocco2 <- c()
         tips_to_keep.ellbi2 = tips_to_keep.myrbi2 = tips_to_keep.apalo2 = tips_to_keep.astgl2 <- c()
         tips_to_keep.tresp2 = tips_to_keep.picso2 = tips_to_keep.chlso2 = tips_to_keep.auxpr2 <- c()
         tips_to_keep.helsp2 = tips_to_keep.um2 = tips_to_keep.tetst2 = tips_to_keep.pedsp2 <- c()
         tips_to_keep.chlpr2 = tips_to_keep.picsp2 = tips_to_keep.ca2 = tips_to_keep.mv2 = tips_to_keep.mesvb2 <- c()
         tips_to_keep.kn2 = tips_to_keep.chara2 = tips_to_keep.me2 = tips_to_keep.meskr2 = tips_to_keep.sp2 <- c()
         tips_to_keep.zygci2 = tips_to_keep.zygcb2 = tips_to_keep.zygcy2 <- c()
         tips_to_keep.cp2 = tips_to_keep.pp2 = tips_to_keep.plesc2 = tips_to_keep.funhy2 <- c()
         tips_to_keep.smag2 = tips_to_keep.sphfa2 = tips_to_keep.takle2 = tips_to_keep.mp2 <- c()
         tips_to_keep.marpa2 = tips_to_keep.ricfl2 = tips_to_keep.ricso2 = tips_to_keep.aa2 <- c()
         tips_to_keep.antpu2 = tips_to_keep.antfu2 = tips_to_keep.megfl2 = tips_to_keep.phach2 <- c()
         tips_to_keep.phyph2 = tips_to_keep.parpe2 = tips_to_keep.notor2 = tips_to_keep.phaca2 <- c()
         tips_to_keep.phasp2 = tips_to_keep.leidu2 <- c()
         tips_to_keep.sm2 = tips_to_keep.isosi2 = tips_to_keep.dipco2 = tips_to_keep.hupas2 = tips_to_keep.lyccl2 <- c()
         tips_to_keep.cri2 = tips_to_keep.adica2 = tips_to_keep.alssp2 = tips_to_keep.sc2 = tips_to_keep.af2 = tips_to_keep.marve2 <- c()
         tips_to_keep.tp2 = tips_to_keep.seqse2 = tips_to_keep.seqgi2 = tips_to_keep.taxch2 = tips_to_keep.abial2 <- c()
         tips_to_keep.picab2 = tips_to_keep.gnemo2 = tips_to_keep.welmi2 = tips_to_keep.cyc2 = tips_to_keep.ginbi2 <- c()
         tips_to_keep.ambtr2 = tips_to_keep.nymco2 = tips_to_keep.cinka2 = tips_to_keep.aegi2 <- c()
         tips_to_keep.ta2 = tips_to_keep.horvu2 = tips_to_keep.bradi2 = tips_to_keep.os2 <- c()
         tips_to_keep.setit2 = tips_to_keep.sb2 = tips_to_keep.zm2 = tips_to_keep.panha2 <- c()
         tips_to_keep.missi2 = tips_to_keep.oroth2 = tips_to_keep.eleco2 = tips_to_keep.anaco2 <- c()
         tips_to_keep.musac2 = tips_to_keep.dioal2 = tips_to_keep.spipo2 = tips_to_keep.aspof2 <- c()
         tips_to_keep.zosma2 = tips_to_keep.at2 = tips_to_keep.araly2 = tips_to_keep.capgr2 <- c()
         tips_to_keep.brara2 = tips_to_keep.braol2 = tips_to_keep.sisir2 = tips_to_keep.schpa2 <- c()
         tips_to_keep.eutsa2 = tips_to_keep.thlar2 = tips_to_keep.boest2 = tips_to_keep.carpa2 <- c()
         tips_to_keep.theca2 = tips_to_keep.goshi2 = tips_to_keep.gosra2 = tips_to_keep.citsi2 <- c()
         tips_to_keep.pontr2 = tips_to_keep.eucgr2 = tips_to_keep.maldo2 = tips_to_keep.prupe2 <- c()
         tips_to_keep.frave2 = tips_to_keep.corav2 = tips_to_keep.caril2 = tips_to_keep.queru2 <- c()
         tips_to_keep.cucsa2 = tips_to_keep.glyma2 = tips_to_keep.medtr2 = tips_to_keep.tripr2 <- c()
         tips_to_keep.cicar2 = tips_to_keep.lotja2 = tips_to_keep.phaac2 = tips_to_keep.vigun2 <- c()
         tips_to_keep.lupal2 = tips_to_keep.lencu2 = tips_to_keep.arahy2 = tips_to_keep.poptr2 <- c()
         tips_to_keep.manes2 = tips_to_keep.salpu2 = tips_to_keep.linus2 = tips_to_keep.corci2 <- c()
         tips_to_keep.vitvi2 = tips_to_keep.kalfe2 = tips_to_keep.vacda2 = tips_to_keep.oleeu2 <- c()
         tips_to_keep.mimgu2 = tips_to_keep.sl2 = tips_to_keep.soltu2 = tips_to_keep.cofar2 <- c()
         tips_to_keep.lacsa2 = tips_to_keep.dauca2 = tips_to_keep.betvu2 = tips_to_keep.amahy2 <- c()
         tips_to_keep.chequ2 = tips_to_keep.sapof2 = tips_to_keep.aquco2 <- c()
         
         
         organisms.list <- c(selected_values_org2())
         
         
         # Fill vectors if organisms are in list and change names
         {
           if ("mp" %in% organisms.list)
           {
             tips_to_keep.mp2 <- grep(pattern = "marchantia_polymorpha", names(mytree)) 
           }
           
           if ("ot" %in% organisms.list)
           {
             tips_to_keep.ot2 <- grep(pattern = "ostreococcus_tauri",names(mytree)) 
           }
           
           if ("at" %in% organisms.list)
           {
             tips_to_keep.at2 <- grep(pattern = "arabidopsis_thaliana",names(mytree)) 
           }
           
           if ("cp" %in% organisms.list)
           {
             tips_to_keep.cp2 <- grep(pattern = "ceratodon_purpureus",names(mytree)) 
           }
           
           if ("cr" %in% organisms.list)
           {
             tips_to_keep.cr2 <- grep(pattern = "chlamydomonas_reinhardtii",names(mytree))
           }
           
           if ("cz" %in% organisms.list)
           {
             tips_to_keep.cz2 <- grep(pattern = "chromochloris_zofingiensis",names(mytree)) 
           }
           
           if ("kn" %in% organisms.list)
           {
             tips_to_keep.kn2 <- grep(pattern = "klebsormidium_nitens",names(mytree)) 
           }
           
           if ("me" %in% organisms.list)
           {
             tips_to_keep.me2 <- grep(pattern = "mesotaenium_endlicherianum",names(mytree)) 
           }
           
           if ("mi" %in% organisms.list)
           {
             tips_to_keep.mi2 <- grep(pattern = "micromonas_pusilla",names(mytree)) 
           }
           
           if ("pp" %in% organisms.list)
           {
             tips_to_keep.pp2 <- grep(pattern = "physcomitrium_patens",names(mytree)) 
           }
           
           if ("sl" %in% organisms.list)
           {
             tips_to_keep.sl2 <- grep(pattern = "solanum_lycopersicum",names(mytree)) 
           }
           
           if ("sm" %in% organisms.list)
           {
             tips_to_keep.sm2 <- grep(pattern = "selaginella_moellendorffii",names(mytree))
           }
           
           if ("sp" %in% organisms.list)
           {
             tips_to_keep.sp2 <- grep(pattern = "spirogloea_muscicola",names(mytree)) 
           }
           
           if ("ta" %in% organisms.list)
           {
             tips_to_keep.ta2 <- grep(pattern = "triticum_aestivum",names(mytree)) 
           }
           
           if ("vc" %in% organisms.list)
           {
             tips_to_keep.vc2 <- grep(pattern = "volvox_carteri",names(mytree))
           }
           
           if ("bp" %in% organisms.list)
           {
             tips_to_keep.bp2 <- grep(pattern = "bathycoccus_prasinos",names(mytree))
           }
           
           if ("cri" %in% organisms.list)
           {
             tips_to_keep.cri2 <- grep(pattern = "ceratopteris_richardii",names(mytree))
           }
           
           if ("ds" %in% organisms.list)
           {
             tips_to_keep.ds2 <- grep(pattern = "dunaliella_salina",names(mytree))
           }
           
           if ("os" %in% organisms.list)
           {
             tips_to_keep.os2 <- grep(pattern = "oryza_sativa",names(mytree))
           }
           
           if ("smag" %in% organisms.list)
           {
             tips_to_keep.smag2 <- grep(pattern = "sphagnum_magellanicum",names(mytree))
           }
           
           if ("tp" %in% organisms.list)
           {
             tips_to_keep.tp2 <- grep(pattern = "thuja_plicata",names(mytree))
           }
           
           if ("aa" %in% organisms.list)
           {
             tips_to_keep.aa2 <- grep(pattern = "anthoceros_agrestis",names(mytree))
           }
           
           if ("um" %in% organisms.list)
           {
             tips_to_keep.um2 <- grep(pattern = "ulva_mutabilis",names(mytree))
           }
           
           if ("rs" %in% organisms.list)
           {
             tips_to_keep.rs2 <- grep(pattern = "raphidocelis_subcapitata",names(mytree))
           }
           
           if ("cyc" %in% organisms.list)
           {
             tips_to_keep.cyc2 <- grep(pattern = "cycas_panzhihuaensis",names(mytree))
           }
           
           
           if ("ca" %in% organisms.list)
           {
             tips_to_keep.ca2 <- grep(pattern = "chlorokybus_atmophyticus",names(mytree))
           }
           
           if ("mv" %in% organisms.list)
           {
             tips_to_keep.mv2 <- grep(pattern = "mesostigma_viride1140",names(mytree))
           }
           
           if ("af" %in% organisms.list)
           {
             tips_to_keep.af2 <- grep(pattern = "azolla_filiculoides",names(mytree))
           }
           
           if ("sc" %in% organisms.list)
           {
             tips_to_keep.sc2 <- grep(pattern = "salvinia_cucullata",names(mytree))
           }
           
           if ("aegi" %in% organisms.list)
           {
             tips_to_keep.aegi2 <- grep(pattern = "aegilops_tauschii",names(mytree))
           }
           
           if ("sb" %in% organisms.list)
           {
             tips_to_keep.sb2 <- grep(pattern = "sorghum_bicolor",names(mytree))
           }
           
           if ("chara" %in% organisms.list)
           {
             tips_to_keep.chara2 <- grep(pattern = "chara_braunii",names(mytree))
           }
           
           
           if ("sceobli" %in% organisms.list)
           {
             tips_to_keep.sceobli2 <- grep(pattern = "scenedesmus_obliquus",names(mytree))
           }
           
           if ("cocco" %in% organisms.list)
           {
             tips_to_keep.cocco2 <- grep(pattern = "coccomyxa_subellipsoidea",names(mytree))
           }
           
           if ("haema" %in% organisms.list)
           {
             tips_to_keep.haema2 <- grep(pattern = "haematococcus_lacustris",names(mytree))
           }
           
           if ("zm" %in% organisms.list)
           {
             tips_to_keep.zm2 <- grep(pattern = "zea_mays",names(mytree))
           }
           
           if ("praco" %in% organisms.list)
           {
             tips_to_keep.praco2 <- grep(pattern = "prasinoderma_coloniale",names(mytree))
           }
           
           if ("ostlu" %in% organisms.list)
           {
             tips_to_keep.ostlu2 <- grep(pattern = "ostreococcus_lucimarinus",names(mytree))
           }
           
           if ("ostme" %in% organisms.list)
           {
             tips_to_keep.ostme2 <- grep(pattern = "ostreococcus_mediterraneus",names(mytree))
           }
           
           if ("micco" %in% organisms.list)
           {
             tips_to_keep.micco2 <- grep(pattern = "micromonas_commoda",names(mytree))
           }
           
           if ("cymte" %in% organisms.list)
           {
             tips_to_keep.cymte2 <- grep(pattern = "cymbomonas_tetramitiformis",names(mytree))
           }
           
           
           if ("chlin" %in% organisms.list)
           {
             tips_to_keep.chlin2 <- grep(pattern = "chlamydomonas_incerta",names(mytree))
           }
           
           if ("gonpe" %in% organisms.list)
           {
             tips_to_keep.gonpe2 <- grep(pattern = "gonium_pectorale",names(mytree))
           }
           
           if ("tetso" %in% organisms.list)
           {
             tips_to_keep.tetso2 <- grep(pattern = "tetrabaena_socialis",names(mytree))
           }
           
           if ("desar" %in% organisms.list)
           {
             tips_to_keep.desar2 <- grep(pattern = "desmodesmus_armatus",names(mytree))
           }
           
           if ("monmi" %in% organisms.list)
           {
             tips_to_keep.monmi2 <- grep(pattern = "monoraphidium_minutum",names(mytree))
           }
           
           if ("caule" %in% organisms.list)
           {
             tips_to_keep.caule2 <- grep(pattern = "caulerpa_lentillifera",names(mytree))
           }
           
           if ("ostqu" %in% organisms.list)
           {
             tips_to_keep.ostqu2 <- grep(pattern = "ostreobium_quekettii",names(mytree))
           }
           
           if ("brysp" %in% organisms.list)
           {
             tips_to_keep.brysp2 <- grep(pattern = "bryopsis_sp",names(mytree))
           }
           
           if ("ellbi" %in% organisms.list)
           {
             tips_to_keep.ellbi2 <- grep(pattern = "elliptochloris_bilobata",names(mytree))
           }
           
           if ("myrbi" %in% organisms.list)
           {
             tips_to_keep.myrbi2 <- grep(pattern = "myrmecia_bisecta",names(mytree))
           }
           
           if ("apalo" %in% organisms.list)
           {
             tips_to_keep.apalo2 <- grep(pattern = "apatococcus_lobatus",names(mytree))
           }
           
           if ("astgl" %in% organisms.list)
           {
             tips_to_keep.astgl2 <- grep(pattern = "asterochloris_glomerata",names(mytree))
           }
           
           if ("tresp" %in% organisms.list)
           {
             tips_to_keep.tresp2 <- grep(pattern = "trebouxia_sp",names(mytree))
           }
           
           if ("picso" %in% organisms.list)
           {
             tips_to_keep.picso2 <- grep(pattern = "picochlorum_soloecismus",names(mytree))
           }
           
           if ("chlso" %in% organisms.list)
           {
             tips_to_keep.chlso2 <- grep(pattern = "chlorella_sorokiniana",names(mytree))
           }
           
           if ("auxpr" %in% organisms.list)
           {
             tips_to_keep.auxpr2 <- grep(pattern = "auxenochlorella_protothecoides",names(mytree))
           }
           
           if ("helsp" %in% organisms.list)
           {
             tips_to_keep.helsp2 <- grep(pattern = "helicosporidium_sp",names(mytree))
           }
           
           if ("tetst" %in% organisms.list)
           {
             tips_to_keep.tetst2 <- grep(pattern = "tetraselmis_striata",names(mytree))
           }
           
           if ("pedsp" %in% organisms.list)
           {
             tips_to_keep.pedsp2 <- grep(pattern = "pedinophyceae_sp",names(mytree))
           }
           
           if ("chlpr" %in% organisms.list)
           {
             tips_to_keep.chlpr2 <- grep(pattern = "chloropicon_primus",names(mytree))
           }
           
           if ("picsp" %in% organisms.list)
           {
             tips_to_keep.picsp2 <- grep(pattern = "picocystis_sp",names(mytree))
           }
           
           
           if ("mesvb" %in% organisms.list)
           {
             tips_to_keep.mesvb2 <- grep(pattern = "mesostigma_viride296",names(mytree))
           }
           
           if ("meskr" %in% organisms.list)
           {
             tips_to_keep.meskr2 <- grep(pattern = "mesotaenium_kramstae",names(mytree))
           }
           
           if ("zygci" %in% organisms.list)
           {
             tips_to_keep.zygci2 <- grep(pattern = "zygnema_circumcarinatum_",names(mytree))
           }
           
           if ("zygcb" %in% organisms.list)
           {
             tips_to_keep.zygcb2 <- grep(pattern = "zygnema_circumcarinatum1559",names(mytree))
           }
           
           if ("zygcy" %in% organisms.list)
           {
             tips_to_keep.zygcy2 <- grep(pattern = "zygnema_cylindricum",names(mytree))
           }
           
           if ("plesc" %in% organisms.list)
           {
             tips_to_keep.plesc2 <- grep(pattern = "pleurozium_schreberi",names(mytree))
           }
           
           if ("funhy" %in% organisms.list)
           {
             tips_to_keep.funhy2 <- grep(pattern = "funaria_hygrometrica",names(mytree))
           }
           
           if ("sphfa" %in% organisms.list)
           {
             tips_to_keep.sphfa2 <- grep(pattern = "sphagnum_fallax",names(mytree))
           }
           
           if ("takle" %in% organisms.list)
           {
             tips_to_keep.takle2 <- grep(pattern = "takakia_lepidozioides",names(mytree))
           }
           
           if ("marpa" %in% organisms.list)
           {
             tips_to_keep.marpa2 <- grep(pattern = "marchantia_paleacea",names(mytree))
           }
           
           if ("ricfl" %in% organisms.list)
           {
             tips_to_keep.ricfl2 <- grep(pattern = "riccia_fluitans",names(mytree))
           }
           
           if ("ricso" %in% organisms.list)
           {
             tips_to_keep.ricso2 <- grep(pattern = "riccia_sorocarpa",names(mytree))
           }
           
           if ("antpu" %in% organisms.list)
           {
             tips_to_keep.antpu2 <- grep(pattern = "anthoceros_punctatus",names(mytree))
           }
           
           if ("antfu" %in% organisms.list)
           {
             tips_to_keep.antfu2 <- grep(pattern = "anthoceros_fusiformis",names(mytree))
           }
           
           if ("megfl" %in% organisms.list)
           {
             tips_to_keep.megfl2 <- grep(pattern = "megaceros_flagellaris",names(mytree))
           }
           
           if ("phach" %in% organisms.list)
           {
             tips_to_keep.phach2 <- grep(pattern = "phaeomegaceros_chiloensis",names(mytree))
           }
           
           if ("phyph" %in% organisms.list)
           {
             tips_to_keep.phyph2 <- grep(pattern = "phymatoceros_phymatodes",names(mytree))
           }
           
           if ("parpe" %in% organisms.list)
           {
             tips_to_keep.parpe2 <- grep(pattern = "paraphymatoceros_pearsonii",names(mytree))
           }
           
           if ("notor" %in% organisms.list)
           {
             tips_to_keep.notor2 <- grep(pattern = "notothylas_orbicularis",names(mytree))
           }
           
           if ("phaca" %in% organisms.list)
           {
             tips_to_keep.phaca2 <- grep(pattern = "phaeoceros_carolinianus",names(mytree))
           }
           
           if ("phasp" %in% organisms.list)
           {
             tips_to_keep.phasp2 <- grep(pattern = "phaeoceros_sp",names(mytree))
           }
           
           if ("leidu" %in% organisms.list)
           {
             tips_to_keep.leidu2 <- grep(pattern = "leiosporoceros_dussii",names(mytree))
           }
           
           if ("isosi" %in% organisms.list)
           {
             tips_to_keep.isosi2 <- grep(pattern = "isoetes_sinensis",names(mytree))
           }
           
           if ("dipco" %in% organisms.list)
           {
             tips_to_keep.dipco2 <- grep(pattern = "diphasiastrum_complanatum",names(mytree))
           }
           
           if ("hupas" %in% organisms.list)
           {
             tips_to_keep.hupas2 <- grep(pattern = "huperzia_asiatica",names(mytree))
           }
           
           if ("lyccl" %in% organisms.list)
           {
             tips_to_keep.lyccl2 <- grep(pattern = "lycopodium_clavatum",names(mytree))
           }
           
           if ("adica" %in% organisms.list)
           {
             tips_to_keep.adica2 <- grep(pattern = "adiantum_capillus",names(mytree))
           }
           
           if ("alssp" %in% organisms.list)
           {
             tips_to_keep.alssp2 <- grep(pattern = "alsophila_spinulosa",names(mytree))
           }
           
           if ("marve" %in% organisms.list)
           {
             tips_to_keep.marve2 <- grep(pattern = "marsilea_vestita",names(mytree))
           }
           
           if ("seqse" %in% organisms.list)
           {
             tips_to_keep.seqse2 <- grep(pattern = "sequoia_sempervirens",names(mytree))
           }
           
           if ("seqgi" %in% organisms.list)
           {
             tips_to_keep.seqgi2 <- grep(pattern = "sequoiadendron_giganteum",names(mytree))
           }
           
           if ("taxch" %in% organisms.list)
           {
             tips_to_keep.taxch2 <- grep(pattern = "taxus_chinensis",names(mytree))
           }
           
           if ("abial" %in% organisms.list)
           {
             tips_to_keep.abial2 <- grep(pattern = "abies_alba",names(mytree))
           }
           
           if ("picab" %in% organisms.list)
           {
             tips_to_keep.picab2 <- grep(pattern = "picea_abies",names(mytree))
           }
           
           if ("gnemo" %in% organisms.list)
           {
             tips_to_keep.gnemo2 <- grep(pattern = "gnetum_montanum",names(mytree))
           }
           
           if ("welmi" %in% organisms.list)
           {
             tips_to_keep.welmi2 <- grep(pattern = "welwitschia_mirabilis",names(mytree))
           }
           
           if ("ginbi" %in% organisms.list)
           {
             tips_to_keep.ginbi2 <- grep(pattern = "ginkgo_biloba",names(mytree))
           }
           
           if ("ambtr" %in% organisms.list)
           {
             tips_to_keep.ambtr2 <- grep(pattern = "amborella_trichopoda",names(mytree))
           }
           
           if ("nymco" %in% organisms.list)
           {
             tips_to_keep.nymco2 <- grep(pattern = "nymphaea_colorata",names(mytree))
           }
           
           if ("cinka" %in% organisms.list)
           {
             tips_to_keep.cinka2 <- grep(pattern = "cinnamomum_kanehirae",names(mytree))
           }
           
           if ("horvu" %in% organisms.list)
           {
             tips_to_keep.horvu2 <- grep(pattern = "hordeum_vulgare",names(mytree))
           }
           
           if ("bradi" %in% organisms.list)
           {
             tips_to_keep.bradi2 <- grep(pattern = "brachypodium_distachyon",names(mytree))
           }
           
           if ("setit" %in% organisms.list)
           {
             tips_to_keep.setit2 <- grep(pattern = "setaria_italica",names(mytree))
           }
           
           if ("panha" %in% organisms.list)
           {
             tips_to_keep.panha2 <- grep(pattern = "panicum_hallii",names(mytree))
           }
           
           if ("missi" %in% organisms.list)
           {
             tips_to_keep.missi2 <- grep(pattern = "miscanthus_sinensis",names(mytree))
           }
           
           if ("oroth" %in% organisms.list)
           {
             tips_to_keep.oroth2 <- grep(pattern = "oropetium_thomaeum",names(mytree))
           }
           
           if ("eleco" %in% organisms.list)
           {
             tips_to_keep.eleco2 <- grep(pattern = "eleusine_coracana",names(mytree))
           }
           
           if ("anaco" %in% organisms.list)
           {
             tips_to_keep.anaco2 <- grep(pattern = "ananas_comosus",names(mytree))
           }
           
           if ("musac" %in% organisms.list)
           {
             tips_to_keep.musac2 <- grep(pattern = "musa_acuminata",names(mytree))
           }
           
           if ("dioal" %in% organisms.list)
           {
             tips_to_keep.dioal2 <- grep(pattern = "dioscorea_alata",names(mytree))
           }
           
           if ("spipo" %in% organisms.list)
           {
             tips_to_keep.spipo2 <- grep(pattern = "spirodela_polyrhiza",names(mytree))
           }
           
           if ("aspof" %in% organisms.list)
           {
             tips_to_keep.aspof2 <- grep(pattern = "asparagus_officinalis",names(mytree))
           }
           
           if ("zosma" %in% organisms.list)
           {
             tips_to_keep.zosma2 <- grep(pattern = "zostera_marina",names(mytree))
           }
           
           if ("araly" %in% organisms.list)
           {
             tips_to_keep.araly2 <- grep(pattern = "arabidopsis_lyrata",names(mytree))
           }
           
           if ("capgr" %in% organisms.list)
           {
             tips_to_keep.capgr2 <- grep(pattern = "capsella_grandiflora",names(mytree))
           }
           
           if ("brara" %in% organisms.list)
           {
             tips_to_keep.brara2 <- grep(pattern = "brassica_rapa",names(mytree))
           }
           
           if ("braol" %in% organisms.list)
           {
             tips_to_keep.braol2 <- grep(pattern = "brassica_oleracea",names(mytree))
           }
           
           if ("sisir" %in% organisms.list)
           {
             tips_to_keep.sisir2 <- grep(pattern = "sisymbrium_irio",names(mytree))
           }
           
           if ("schpa" %in% organisms.list)
           {
             tips_to_keep.schpa2 <- grep(pattern = "schrenkiella_parvula",names(mytree))
           }
           
           if ("eutsa" %in% organisms.list)
           {
             tips_to_keep.eutsa2 <- grep(pattern = "eutrema_salsugineum",names(mytree))
           }
           
           if ("thlar" %in% organisms.list)
           {
             tips_to_keep.thlar2 <- grep(pattern = "thlaspi_arvense",names(mytree))
           }
           
           if ("boest" %in% organisms.list)
           {
             tips_to_keep.boest2 <- grep(pattern = "boechera_stricta",names(mytree))
           }
           
           if ("carpa" %in% organisms.list)
           {
             tips_to_keep.carpa2 <- grep(pattern = "carica_papaya",names(mytree))
           }
           
           if ("theca" %in% organisms.list)
           {
             tips_to_keep.theca2 <- grep(pattern = "theobroma_cacao",names(mytree))
           }
           
           if ("goshi" %in% organisms.list)
           {
             tips_to_keep.goshi2 <- grep(pattern = "gossypium_hirsutum",names(mytree))
           }
           
           if ("gosra" %in% organisms.list)
           {
             tips_to_keep.gosra2 <- grep(pattern = "gossypium_raimondii",names(mytree))
           }
           
           if ("citsi" %in% organisms.list)
           {
             tips_to_keep.citsi2 <- grep(pattern = "citrus_sinensis",names(mytree))
           }
           
           if ("pontr" %in% organisms.list)
           {
             tips_to_keep.pontr2 <- grep(pattern = "poncirus_trifoliata",names(mytree))
           }
           
           if ("eucgr" %in% organisms.list)
           {
             tips_to_keep.eucgr2 <- grep(pattern = "eucalyptus_grandis",names(mytree))
           }
           
           if ("maldo" %in% organisms.list)
           {
             tips_to_keep.maldo2 <- grep(pattern = "malus_domestica",names(mytree))
           }
           
           if ("prupe" %in% organisms.list)
           {
             tips_to_keep.prupe2 <- grep(pattern = "prunus_persica",names(mytree))
           }
           
           if ("frave" %in% organisms.list)
           {
             tips_to_keep.frave2 <- grep(pattern = "fragaria_vesca",names(mytree))
           }
           
           if ("corav" %in% organisms.list)
           {
             tips_to_keep.corav2 <- grep(pattern = "corylus_avellana",names(mytree))
           }
           
           if ("caril" %in% organisms.list)
           {
             tips_to_keep.caril2 <- grep(pattern = "carya_illinoinensis",names(mytree))
           }
           
           if ("queru" %in% organisms.list)
           {
             tips_to_keep.queru2 <- grep(pattern = "quercus_rubra",names(mytree))
           }
           
           if ("cucsa" %in% organisms.list)
           {
             tips_to_keep.cucsa2 <- grep(pattern = "cucumis_sativus",names(mytree))
           }
           
           if ("glyma" %in% organisms.list)
           {
             tips_to_keep.glyma2 <- grep(pattern = "glycine_max",names(mytree))
           }
           
           if ("medtr" %in% organisms.list)
           {
             tips_to_keep.medtr2 <- grep(pattern = "medicago_truncatula",names(mytree))
           }
           
           if ("tripr" %in% organisms.list)
           {
             tips_to_keep.tripr2 <- grep(pattern = "trifolium_pratense",names(mytree))
           }
           
           if ("cicar" %in% organisms.list)
           {
             tips_to_keep.cicar2 <- grep(pattern = "cicer_arietinum",names(mytree))
           }
           
           if ("lotja" %in% organisms.list)
           {
             tips_to_keep.lotja2 <- grep(pattern = "lotus_japonicus",names(mytree))
           }
           
           if ("phaac" %in% organisms.list)
           {
             tips_to_keep.phaac2 <- grep(pattern = "phaseolus_acutifolius",names(mytree))
           }
           
           if ("vigun" %in% organisms.list)
           {
             tips_to_keep.vigun2 <- grep(pattern = "vigna_unguiculata",names(mytree))
           }
           
           if ("lupal" %in% organisms.list)
           {
             tips_to_keep.lupal2 <- grep(pattern = "lupinus_albus",names(mytree))
           }
           
           if ("lencu" %in% organisms.list)
           {
             tips_to_keep.lencu2 <- grep(pattern = "lens_culinaris",names(mytree))
           }
           
           if ("arahy" %in% organisms.list)
           {
             tips_to_keep.arahy2 <- grep(pattern = "arachis_hypogaea",names(mytree))
           }
           
           if ("poptr" %in% organisms.list)
           {
             tips_to_keep.poptr2 <- grep(pattern = "populus_trichocarpa",names(mytree))
           }
           
           if ("manes" %in% organisms.list)
           {
             tips_to_keep.manes2 <- grep(pattern = "manihot_esculenta",names(mytree))
           }
           
           if ("salpu" %in% organisms.list)
           {
             tips_to_keep.salpu2 <- grep(pattern = "salix_purpurea",names(mytree))
           }
           
           if ("linus" %in% organisms.list)
           {
             tips_to_keep.linus2 <- grep(pattern = "linum_usitatissimum",names(mytree))
           }
           
           if ("corci" %in% organisms.list)
           {
             tips_to_keep.corci2 <- grep(pattern = "corymbia_citriodora",names(mytree))
           }
           
           if ("vitvi" %in% organisms.list)
           {
             tips_to_keep.vitvi2 <- grep(pattern = "vitis_vinifera",names(mytree))
           }
           
           if ("kalfe" %in% organisms.list)
           {
             tips_to_keep.kalfe2 <- grep(pattern = "kalanchoe_fedtschenkoi",names(mytree))
           }
           
           if ("vacda" %in% organisms.list)
           {
             tips_to_keep.vacda2 <- grep(pattern = "vaccinium_darrowii",names(mytree))
           }
           
           if ("oleeu" %in% organisms.list)
           {
             tips_to_keep.oleeu2 <- grep(pattern = "olea_europaea",names(mytree))
           }
           
           if ("mimgu" %in% organisms.list)
           {
             tips_to_keep.mimgu2 <- grep(pattern = "mimulus_guttatus",names(mytree))
           }
           
           if ("soltu" %in% organisms.list)
           {
             tips_to_keep.soltu2 <- grep(pattern = "solanum_tuberosum",names(mytree))
           }
           
           if ("cofar" %in% organisms.list)
           {
             tips_to_keep.cofar2 <- grep(pattern = "coffea_arabica",names(mytree))
           }
           
           if ("lacsa" %in% organisms.list)
           {
             tips_to_keep.lacsa2 <- grep(pattern = "lactuca_sativa",names(mytree))
           }
           
           if ("dauca" %in% organisms.list)
           {
             tips_to_keep.dauca2 <- grep(pattern = "daucus_carota",names(mytree))
           }
           
           if ("betvu" %in% organisms.list)
           {
             tips_to_keep.betvu2 <- grep(pattern = "beta_vulgaris",names(mytree))
           }
           
           if ("amahy" %in% organisms.list)
           {
             tips_to_keep.amahy2 <- grep(pattern = "amaranthus_hypochondriacus",names(mytree))
           }
           
           if ("chequ" %in% organisms.list)
           {
             tips_to_keep.chequ2 <- grep(pattern = "chenopodium_quinoa",names(mytree))
           }
           
           if ("sapof" %in% organisms.list)
           {
             tips_to_keep.sapof2 <- grep(pattern = "saponaria_officinalis",names(mytree))
           }
           
           if ("aquco" %in% organisms.list)
           {
             tips_to_keep.aquco2 <- grep(pattern = "aquilegia_coerulea",names(mytree))
           }
         }
         
         
         # Concatenate indexes to keep and subset MSA
         tips_to_keep.global <- c(tips_to_keep.praco2, tips_to_keep.ot2, tips_to_keep.ostlu2, tips_to_keep.ostme2, tips_to_keep.bp2,
                                  tips_to_keep.mi2, tips_to_keep.micco2, tips_to_keep.cymte2, tips_to_keep.cr2, tips_to_keep.chlin2,
                                  tips_to_keep.haema2, tips_to_keep.vc2, tips_to_keep.ds2, tips_to_keep.gonpe2, tips_to_keep.tetso2,
                                  tips_to_keep.cz2, tips_to_keep.rs2, tips_to_keep.sceobli2, tips_to_keep.desar2, tips_to_keep.monmi2,
                                  tips_to_keep.caule2, tips_to_keep.ostqu2, tips_to_keep.brysp2, tips_to_keep.cocco2,
                                  tips_to_keep.ellbi2, tips_to_keep.myrbi2, tips_to_keep.apalo2, tips_to_keep.astgl2,
                                  tips_to_keep.tresp2, tips_to_keep.picso2, tips_to_keep.chlso2, tips_to_keep.auxpr2,
                                  tips_to_keep.helsp2, tips_to_keep.um2, tips_to_keep.tetst2, tips_to_keep.pedsp2,
                                  tips_to_keep.chlpr2, tips_to_keep.picsp2, tips_to_keep.ca2, tips_to_keep.mv2, tips_to_keep.mesvb2,
                                  tips_to_keep.kn2, tips_to_keep.chara2, tips_to_keep.me2, tips_to_keep.meskr2, tips_to_keep.sp2,
                                  tips_to_keep.zygci2, tips_to_keep.zygcb2, tips_to_keep.zygcy2,
                                  tips_to_keep.cp2, tips_to_keep.pp2, tips_to_keep.plesc2, tips_to_keep.funhy2,
                                  tips_to_keep.smag2, tips_to_keep.sphfa2, tips_to_keep.takle2, tips_to_keep.mp2,
                                  tips_to_keep.marpa2, tips_to_keep.ricfl2, tips_to_keep.ricso2, tips_to_keep.aa2,
                                  tips_to_keep.antpu2, tips_to_keep.antfu2, tips_to_keep.megfl2, tips_to_keep.phach2,
                                  tips_to_keep.phyph2, tips_to_keep.parpe2, tips_to_keep.notor2, tips_to_keep.phaca2,
                                  tips_to_keep.phasp2, tips_to_keep.leidu2, tips_to_keep.sm2, tips_to_keep.isosi2, 
                                  tips_to_keep.dipco2, tips_to_keep.hupas2, tips_to_keep.lyccl2,
                                  tips_to_keep.cri2, tips_to_keep.adica2, tips_to_keep.alssp2, tips_to_keep.sc2, 
                                  tips_to_keep.af2, tips_to_keep.marve2, tips_to_keep.tp2, tips_to_keep.seqse2, 
                                  tips_to_keep.seqgi2, tips_to_keep.taxch2, tips_to_keep.abial2, tips_to_keep.picab2,
                                  tips_to_keep.gnemo2, tips_to_keep.welmi2, tips_to_keep.cyc2, tips_to_keep.ginbi2,
                                  tips_to_keep.ambtr2, tips_to_keep.nymco2, tips_to_keep.cinka2, tips_to_keep.aegi2,
                                  tips_to_keep.ta2, tips_to_keep.horvu2, tips_to_keep.bradi2, tips_to_keep.os2,
                                  tips_to_keep.setit2, tips_to_keep.sb2, tips_to_keep.zm2, tips_to_keep.panha2,
                                  tips_to_keep.missi2, tips_to_keep.oroth2, tips_to_keep.eleco2, tips_to_keep.anaco2,
                                  tips_to_keep.musac2, tips_to_keep.dioal2, tips_to_keep.spipo2, tips_to_keep.aspof2,
                                  tips_to_keep.zosma2, tips_to_keep.at2, tips_to_keep.araly2, tips_to_keep.capgr2,
                                  tips_to_keep.brara2, tips_to_keep.braol2, tips_to_keep.sisir2, tips_to_keep.schpa2,
                                  tips_to_keep.eutsa2, tips_to_keep.thlar2, tips_to_keep.boest2, tips_to_keep.carpa2,
                                  tips_to_keep.theca2, tips_to_keep.goshi2, tips_to_keep.gosra2, tips_to_keep.citsi2,
                                  tips_to_keep.pontr2, tips_to_keep.eucgr2, tips_to_keep.maldo2, tips_to_keep.prupe2,
                                  tips_to_keep.frave2, tips_to_keep.corav2, tips_to_keep.caril2, tips_to_keep.queru2,
                                  tips_to_keep.cucsa2, tips_to_keep.glyma2, tips_to_keep.medtr2, tips_to_keep.tripr2,
                                  tips_to_keep.cicar2, tips_to_keep.lotja2, tips_to_keep.phaac2, tips_to_keep.vigun2,
                                  tips_to_keep.lupal2, tips_to_keep.lencu2, tips_to_keep.arahy2, tips_to_keep.poptr2,
                                  tips_to_keep.manes2, tips_to_keep.salpu2, tips_to_keep.linus2, tips_to_keep.corci2,
                                  tips_to_keep.vitvi2, tips_to_keep.kalfe2, tips_to_keep.vacda2, tips_to_keep.oleeu2,
                                  tips_to_keep.mimgu2, tips_to_keep.sl2, tips_to_keep.soltu2, tips_to_keep.cofar2,
                                  tips_to_keep.lacsa2, tips_to_keep.dauca2, tips_to_keep.betvu2, tips_to_keep.amahy2,
                                  tips_to_keep.chequ2, tips_to_keep.sapof2, tips_to_keep.aquco2)
         
         
         my_subset_tree <- subset(mytree, tips_to_keep.global)
         
         {
           if (build_trees2() == "Neighbour Joining")
           {
             # Create dist matrix for NJ and build tree
             dm <- dist.ml(my_subset_tree)
             treeNJ  <- NJ(dm)
             
             # Bootstrap for NJ
             fun_nj <- function(x) NJ(dist.ml(x))
             bs_nj <- bootstrap.phyDat(my_subset_tree, fun_nj, bs=100)
             
             # Save bootstrap values to the tree
             tree <- plotBS(treeNJ, bs_nj, type = "n")
             # Rooting tree
             tree <- midpoint(tree)
             return(tree)
           }
           
           else if (build_trees2() == "UPGMA")
           {
             dm <- dist.ml(my_subset_tree)
             treeUPGMA  <- upgma(dm)
             
             # Bootstrap for UPGMA
             fun_upgma <- function(x) upgma(dist.ml(x))
             bs_upgma <- bootstrap.phyDat(my_subset_tree, fun_upgma, bs=100)
             
             # Save bootstrap values to the tree
             tree <- addConfidences(treeUPGMA, bs_upgma)
             return(tree)
           }
         }
         
       }
     }
   }) %>% bindEvent(input$run_button2)
   
   ortho_seq2 <- reactive({
     file.name <- og.name2()
     
     # Load orthogroup sequences file
     ortho.seq.name <- paste("Orthogroup_Sequences",paste(file.name, "fa", sep = "."), sep="/")
     
     ortho_seq <- seqinr::read.fasta(ortho.seq.name, seqtype = "AA")
     return(ortho_seq)
   })
   
   
   # Tips to keep of each species with proper notation
   tips_to_keep.mp2 <- reactive({
     
     tree <- tree2()
     # Selection of organisms
     organisms.list <- c(selected_values_org2())
     
     # Selection of genes from the selected organism
     tips_to_keep.mp <- c()
     if ("mp" %in% organisms.list)
     {
       tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label)
     }
     return(tips_to_keep.mp)
   })
   
   tips_to_keep.ot2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ot <- c()
     if ("ot" %in% organisms.list)
     {
       tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label)
     }
     return(tips_to_keep.ot)
   })
   
   tips_to_keep.at2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.at <- c()
     if ("at" %in% organisms.list)
     {
       tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label)
     }
     return(tips_to_keep.at)
   })
   
   tips_to_keep.cp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cp <- c()
     if ("cp" %in% organisms.list)
     {
       tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label)
     }
     
     return(tips_to_keep.cp)
   })
   
   tips_to_keep.cr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cr <- c()
     if ("cr" %in% organisms.list)
     {
       tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
     }
     return(tips_to_keep.cr)
   })
   
   tips_to_keep.cz2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cz <- c()
     if ("cz" %in% organisms.list)
     {
       tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label)
     }
     return(tips_to_keep.cz)
   })
   
   tips_to_keep.kn2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.kn <- c()
     if ("kn" %in% organisms.list)
     {
       tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label)
     }
     return(tips_to_keep.kn)
   })
   
   tips_to_keep.me2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.me <- c()
     if ("me" %in% organisms.list)
     {
       tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label)
     }
     return(tips_to_keep.me)
   })
   
   tips_to_keep.mi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.mi <- c()
     if ("mi" %in% organisms.list)
     {
       tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label)
     }
     return(tips_to_keep.mi)
   })
   
   tips_to_keep.pp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.pp <- c()
     if ("pp" %in% organisms.list)
     {
       tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label)
     }
     return(tips_to_keep.pp)
   })
   
   tips_to_keep.sl2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sl <- c()
     if ("sl" %in% organisms.list)
     {
       tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label)
     }
     return(tips_to_keep.sl)
   })
   
   tips_to_keep.sm2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sm <- c()
     if ("sm" %in% organisms.list)
     {
       tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label)
     }
     return(tips_to_keep.sm)
   })
   
   tips_to_keep.sp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sp <- c()
     if ("sp" %in% organisms.list)
     {
       tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label)
     }
     return(tips_to_keep.sp)
   })
   
   tips_to_keep.ta2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ta <- c()
     if ("ta" %in% organisms.list)
     {
       tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label)
     }
     return(tips_to_keep.ta)
   })
   
   tips_to_keep.vc2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.vc <- c()
     if ("vc" %in% organisms.list)
     {
       tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
     }
     return(tips_to_keep.vc)
   })
   
   tips_to_keep.bp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.bp <- c()
     if ("bp" %in% organisms.list)
     {
       tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
     }
     return(tips_to_keep.bp)
   })
   
   tips_to_keep.cri2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cri <- c()
     if ("cri" %in% organisms.list)
     {
       tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
     }
     return(tips_to_keep.cri)
   })
   
   tips_to_keep.ds2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ds <- c()
     if ("ds" %in% organisms.list)
     {
       tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
     }
     return(tips_to_keep.ds)
   })
   
   tips_to_keep.os2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.os <- c()
     if ("os" %in% organisms.list)
     {
       tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
     }
     return(tips_to_keep.os)
   })
   
   tips_to_keep.smag2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.smag <- c()
     if ("smag" %in% organisms.list)
     {
       tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
     }
     return(tips_to_keep.smag)
   })
   
   tips_to_keep.tp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.tp <- c()
     if ("tp" %in% organisms.list)
     {
       tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
     }
     return(tips_to_keep.tp)
   })
   
   tips_to_keep.aa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.aa <- c()
     if ("aa" %in% organisms.list)
     {
       tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
     }
     return(tips_to_keep.aa)
   })
   
   tips_to_keep.um2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.um <- c()
     if ("um" %in% organisms.list)
     {
       tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
     }
     return(tips_to_keep.um)
   })
   
   tips_to_keep.rs2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.rs <- c()
     if ("rs" %in% organisms.list)
     {
       tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
     }
     return(tips_to_keep.rs)
   })
   
   tips_to_keep.cyc2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cyc <- c()
     if ("cyc" %in% organisms.list)
     {
       tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
     }
     return(tips_to_keep.cyc)
   })
   
   
   tips_to_keep.ca2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ca <- c()
     if ("ca" %in% organisms.list)
     {
       tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
     }
     return(tips_to_keep.ca)
   })
   
   tips_to_keep.mv2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.mv <- c()
     if ("mv" %in% organisms.list)
     {
       tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
     }
     return(tips_to_keep.mv)
   })
   
   tips_to_keep.af2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.af <- c()
     if ("af" %in% organisms.list)
     {
       tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
     }
     return(tips_to_keep.af)
   })
   
   tips_to_keep.sc2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sc <- c()
     if ("sc" %in% organisms.list)
     {
       tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
     }
     return(tips_to_keep.sc)
   })
   
   tips_to_keep.aegi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.aegi <- c()
     if ("aegi" %in% organisms.list)
     {
       tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
     }
     return(tips_to_keep.aegi)
   })
   
   tips_to_keep.sb2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sb <- c()
     if ("sb" %in% organisms.list)
     {
       tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
     }
     return(tips_to_keep.sb)
   })
   
   tips_to_keep.chara2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.chara <- c()
     if ("chara" %in% organisms.list)
     {
       tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
     }
     return(tips_to_keep.chara)
   })
   
   tips_to_keep.sceobli2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sceobli <- c()
     if ("sceobli" %in% organisms.list)
     {
       tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
     }
     return(tips_to_keep.sceobli)
   })
   
   tips_to_keep.cocco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cocco <- c()
     if ("cocco" %in% organisms.list)
     {
       tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
     }
     return(tips_to_keep.cocco)
   })
   
   
   tips_to_keep.haema2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.haema <- c()
     if ("haema" %in% organisms.list)
     {
       tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
     }
     return(tips_to_keep.haema)
   })
   
   tips_to_keep.zm2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.zm <- c()
     if ("zm" %in% organisms.list)
     {
       tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
     }
     return(tips_to_keep.zm)
   })
   
   tips_to_keep.praco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.praco <- c()
     if ("praco" %in% organisms.list)
     {
       tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
     }
     return(tips_to_keep.praco)
   })
   
   tips_to_keep.ostlu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ostlu <- c()
     if ("ostlu" %in% organisms.list)
     {
       tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
     }
     return(tips_to_keep.ostlu)
   })
   
   tips_to_keep.ostme2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ostme <- c()
     if ("ostme" %in% organisms.list)
     {
       tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
     }
     return(tips_to_keep.ostme)
   })
   
   tips_to_keep.micco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.micco <- c()
     if ("micco" %in% organisms.list)
     {
       tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
     }
     return(tips_to_keep.micco)
   })
   
   tips_to_keep.cymte2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cymte <- c()
     if ("cymte" %in% organisms.list)
     {
       tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
     }
     return(tips_to_keep.cymte)
   })
   
   tips_to_keep.chlin2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.chlin <- c()
     if ("chlin" %in% organisms.list)
     {
       tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
     }
     return(tips_to_keep.chlin)
   })
   
   tips_to_keep.gonpe2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.gonpe <- c()
     if ("gonpe" %in% organisms.list)
     {
       tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
     }
     return(tips_to_keep.gonpe)
   })
   
   tips_to_keep.tetso2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.tetso <- c()
     if ("tetso" %in% organisms.list)
     {
       tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
     }
     return(tips_to_keep.tetso)
   })
   
   tips_to_keep.desar2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.desar <- c()
     if ("desar" %in% organisms.list)
     {
       tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
     }
     return(tips_to_keep.desar)
   })
   
   tips_to_keep.monmi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.monmi <- c()
     if ("monmi" %in% organisms.list)
     {
       tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
     }
     return(tips_to_keep.monmi)
   })
   
   tips_to_keep.caule2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.caule <- c()
     if ("caule" %in% organisms.list)
     {
       tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
     }
     return(tips_to_keep.caule)
   })
   
   tips_to_keep.ostqu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ostqu <- c()
     if ("ostqu" %in% organisms.list)
     {
       tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
     }
     return(tips_to_keep.ostqu)
   })
   
   tips_to_keep.brysp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.brysp <- c()
     if ("brysp" %in% organisms.list)
     {
       tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
     }
     return(tips_to_keep.brysp)
   })
   
   tips_to_keep.ellbi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ellbi <- c()
     if ("ellbi" %in% organisms.list)
     {
       tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
     }
     return(tips_to_keep.ellbi)
   })
   
   tips_to_keep.myrbi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.myrbi <- c()
     if ("myrbi" %in% organisms.list)
     {
       tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
     }
     return(tips_to_keep.myrbi)
   })
   
   tips_to_keep.apalo2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.apalo <- c()
     if ("apalo" %in% organisms.list)
     {
       tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
     }
     return(tips_to_keep.apalo)
   })
   
   tips_to_keep.astgl2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.astgl <- c()
     if ("astgl" %in% organisms.list)
     {
       tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
     }
     return(tips_to_keep.astgl)
   })
   
   tips_to_keep.tresp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.tresp <- c()
     if ("tresp" %in% organisms.list)
     {
       tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
     }
     return(tips_to_keep.tresp)
   })
   
   tips_to_keep.picso2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.picso <- c()
     if ("picso" %in% organisms.list)
     {
       tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
     }
     return(tips_to_keep.picso)
   })
   
   tips_to_keep.chlso2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.chlso <- c()
     if ("chlso" %in% organisms.list)
     {
       tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
     }
     return(tips_to_keep.chlso)
   })
   
   tips_to_keep.auxpr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.auxpr <- c()
     if ("auxpr" %in% organisms.list)
     {
       tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
     }
     return(tips_to_keep.auxpr)
   })
   
   tips_to_keep.helsp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.helsp <- c()
     if ("helsp" %in% organisms.list)
     {
       tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
     }
     return(tips_to_keep.helsp)
   })
   
   tips_to_keep.tetst2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.tetst <- c()
     if ("tetst" %in% organisms.list)
     {
       tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
     }
     return(tips_to_keep.tetst)
   })
   
   tips_to_keep.pedsp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.pedsp <- c()
     if ("pedsp" %in% organisms.list)
     {
       tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
     }
     return(tips_to_keep.pedsp)
   })
   
   tips_to_keep.chlpr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.chlpr <- c()
     if ("chlpr" %in% organisms.list)
     {
       tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
     }
     return(tips_to_keep.chlpr)
   })
   
   tips_to_keep.picsp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.picsp <- c()
     if ("picsp" %in% organisms.list)
     {
       tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
     }
     return(tips_to_keep.picsp)
   })
   
   tips_to_keep.mesvb2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.mesvb <- c()
     if ("mesvb" %in% organisms.list)
     {
       tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
     }
     return(tips_to_keep.mesvb)
   })
   
   tips_to_keep.meskr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.meskr <- c()
     if ("meskr" %in% organisms.list)
     {
       tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
     }
     return(tips_to_keep.meskr)
   })
   
   tips_to_keep.zygci2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.zygci <- c()
     if ("zygci" %in% organisms.list)
     {
       tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
     }
     return(tips_to_keep.zygci)
   })
   
   tips_to_keep.zygcb2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.zygcb <- c()
     if ("zygcb" %in% organisms.list)
     {
       tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
     }
     return(tips_to_keep.zygcb)
   })
   
   tips_to_keep.zygcy2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.zygcy <- c()
     if ("zygcy" %in% organisms.list)
     {
       tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
     }
     return(tips_to_keep.zygcy)
   })
   
   tips_to_keep.plesc2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.plesc <- c()
     if ("plesc" %in% organisms.list)
     {
       tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
     }
     return(tips_to_keep.plesc)
   })
   
   tips_to_keep.funhy2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.funhy <- c()
     if ("funhy" %in% organisms.list)
     {
       tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
     }
     return(tips_to_keep.funhy)
   })
   
   tips_to_keep.sphfa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sphfa <- c()
     if ("sphfa" %in% organisms.list)
     {
       tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
     }
     return(tips_to_keep.sphfa)
   })
   
   tips_to_keep.takle2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.takle <- c()
     if ("takle" %in% organisms.list)
     {
       tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
     }
     return(tips_to_keep.takle)
   })
   
   tips_to_keep.marpa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.marpa <- c()
     if ("marpa" %in% organisms.list)
     {
       tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
     }
     return(tips_to_keep.marpa)
   })
   
   tips_to_keep.ricfl2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ricfl <- c()
     if ("ricfl" %in% organisms.list)
     {
       tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
     }
     return(tips_to_keep.ricfl)
   })
   
   tips_to_keep.ricso2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ricso <- c()
     if ("ricso" %in% organisms.list)
     {
       tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
     }
     return(tips_to_keep.ricso)
   })
   
   tips_to_keep.antpu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.antpu <- c()
     if ("antpu" %in% organisms.list)
     {
       tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
     }
     return(tips_to_keep.antpu)
   })
   
   tips_to_keep.antfu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.antfu <- c()
     if ("antfu" %in% organisms.list)
     {
       tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
     }
     return(tips_to_keep.antfu)
   })
   
   tips_to_keep.megfl2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.megfl <- c()
     if ("megfl" %in% organisms.list)
     {
       tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
     }
     return(tips_to_keep.megfl)
   })
   
   tips_to_keep.phach2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.phach <- c()
     if ("phach" %in% organisms.list)
     {
       tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
     }
     return(tips_to_keep.phach)
   })
   
   tips_to_keep.phyph2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.phyph <- c()
     if ("phyph" %in% organisms.list)
     {
       tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
     }
     return(tips_to_keep.phyph)
   })
   
   tips_to_keep.parpe2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.parpe <- c()
     if ("parpe" %in% organisms.list)
     {
       tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
     }
     return(tips_to_keep.parpe)
   })
   
   tips_to_keep.notor2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.notor <- c()
     if ("notor" %in% organisms.list)
     {
       tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
     }
     return(tips_to_keep.notor)
   })
   
   tips_to_keep.phaca2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.phaca <- c()
     if ("phaca" %in% organisms.list)
     {
       tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
     }
     return(tips_to_keep.phaca)
   })
   
   tips_to_keep.phasp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.phasp <- c()
     if ("phasp" %in% organisms.list)
     {
       tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
     }
     return(tips_to_keep.phasp)
   })
   
   tips_to_keep.leidu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.leidu <- c()
     if ("leidu" %in% organisms.list)
     {
       tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
     }
     return(tips_to_keep.leidu)
   })
   
   tips_to_keep.isosi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.isosi <- c()
     if ("isosi" %in% organisms.list)
     {
       tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
     }
     return(tips_to_keep.isosi)
   })
   
   tips_to_keep.dipco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.dipco <- c()
     if ("dipco" %in% organisms.list)
     {
       tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
     }
     return(tips_to_keep.dipco)
   })
   
   tips_to_keep.hupas2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.hupas <- c()
     if ("hupas" %in% organisms.list)
     {
       tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
     }
     return(tips_to_keep.hupas)
   })
   
   tips_to_keep.lyccl2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.lyccl <- c()
     if ("lyccl" %in% organisms.list)
     {
       tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
     }
     return(tips_to_keep.lyccl)
   })
   
   tips_to_keep.adica2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.adica <- c()
     if ("adica" %in% organisms.list)
     {
       tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
     }
     return(tips_to_keep.adica)
   })
   
   tips_to_keep.alssp2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.alssp <- c()
     if ("alssp" %in% organisms.list)
     {
       tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
     }
     return(tips_to_keep.alssp)
   })
   
   tips_to_keep.marve2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.marve <- c()
     if ("marve" %in% organisms.list)
     {
       tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
     }
     return(tips_to_keep.marve)
   })
   
   tips_to_keep.seqse2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.seqse <- c()
     if ("seqse" %in% organisms.list)
     {
       tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
     }
     return(tips_to_keep.seqse)
   })
   
   tips_to_keep.seqgi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.seqgi <- c()
     if ("seqgi" %in% organisms.list)
     {
       tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
     }
     return(tips_to_keep.seqgi)
   })
   
   tips_to_keep.taxch2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.taxch <- c()
     if ("taxch" %in% organisms.list)
     {
       tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
     }
     return(tips_to_keep.taxch)
   })
   
   tips_to_keep.abial2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.abial <- c()
     if ("abial" %in% organisms.list)
     {
       tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
     }
     return(tips_to_keep.abial)
   })
   
   tips_to_keep.picab2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.picab <- c()
     if ("picab" %in% organisms.list)
     {
       tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
     }
     return(tips_to_keep.picab)
   })
   
   tips_to_keep.gnemo2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.gnemo <- c()
     if ("gnemo" %in% organisms.list)
     {
       tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
     }
     return(tips_to_keep.gnemo)
   })
   
   tips_to_keep.welmi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.welmi <- c()
     if ("welmi" %in% organisms.list)
     {
       tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
     }
     return(tips_to_keep.welmi)
   })
   
   tips_to_keep.ginbi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ginbi <- c()
     if ("ginbi" %in% organisms.list)
     {
       tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
     }
     return(tips_to_keep.ginbi)
   })
   
   tips_to_keep.ambtr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.ambtr <- c()
     if ("ambtr" %in% organisms.list)
     {
       tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
     }
     return(tips_to_keep.ambtr)
   })
   
   tips_to_keep.nymco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.nymco <- c()
     if ("nymco" %in% organisms.list)
     {
       tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
     }
     return(tips_to_keep.nymco)
   })
   
   tips_to_keep.cinka2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cinka <- c()
     if ("cinka" %in% organisms.list)
     {
       tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
     }
     return(tips_to_keep.cinka)
   })
   
   tips_to_keep.horvu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.horvu <- c()
     if ("horvu" %in% organisms.list)
     {
       tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
     }
     return(tips_to_keep.horvu)
   })
   
   tips_to_keep.bradi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.bradi <- c()
     if ("bradi" %in% organisms.list)
     {
       tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
     }
     return(tips_to_keep.bradi)
   })
   
   tips_to_keep.setit2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.setit <- c()
     if ("setit" %in% organisms.list)
     {
       tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
     }
     return(tips_to_keep.setit)
   })
   
   tips_to_keep.panha2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.panha <- c()
     if ("panha" %in% organisms.list)
     {
       tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
     }
     return(tips_to_keep.panha)
   })
   
   tips_to_keep.missi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.missi <- c()
     if ("missi" %in% organisms.list)
     {
       tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
     }
     return(tips_to_keep.missi)
   })
   
   tips_to_keep.oroth2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.oroth <- c()
     if ("oroth" %in% organisms.list)
     {
       tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
     }
     return(tips_to_keep.oroth)
   })
   
   tips_to_keep.eleco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.eleco <- c()
     if ("eleco" %in% organisms.list)
     {
       tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
     }
     return(tips_to_keep.eleco)
   })
   
   tips_to_keep.anaco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.anaco <- c()
     if ("anaco" %in% organisms.list)
     {
       tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
     }
     return(tips_to_keep.anaco)
   })
   
   tips_to_keep.musac2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.musac <- c()
     if ("musac" %in% organisms.list)
     {
       tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
     }
     return(tips_to_keep.musac)
   })
   
   tips_to_keep.dioal2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.dioal <- c()
     if ("dioal" %in% organisms.list)
     {
       tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
     }
     return(tips_to_keep.dioal)
   })
   
   tips_to_keep.spipo2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.spipo <- c()
     if ("spipo" %in% organisms.list)
     {
       tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
     }
     return(tips_to_keep.spipo)
   })
   
   tips_to_keep.aspof2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.aspof <- c()
     if ("aspof" %in% organisms.list)
     {
       tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
     }
     return(tips_to_keep.aspof)
   })
   
   tips_to_keep.zosma2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.zosma <- c()
     if ("zosma" %in% organisms.list)
     {
       tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
     }
     return(tips_to_keep.zosma)
   })
   
   tips_to_keep.araly2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.araly <- c()
     if ("araly" %in% organisms.list)
     {
       tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
     }
     return(tips_to_keep.araly)
   })
   
   tips_to_keep.capgr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.capgr <- c()
     if ("capgr" %in% organisms.list)
     {
       tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
     }
     return(tips_to_keep.capgr)
   })
   
   tips_to_keep.brara2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.brara <- c()
     if ("brara" %in% organisms.list)
     {
       tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
     }
     return(tips_to_keep.brara)
   })
   
   tips_to_keep.braol2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.braol <- c()
     if ("braol" %in% organisms.list)
     {
       tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
     }
     return(tips_to_keep.braol)
   })
   
   tips_to_keep.sisir2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sisir <- c()
     if ("sisir" %in% organisms.list)
     {
       tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
     }
     return(tips_to_keep.sisir)
   })
   
   tips_to_keep.schpa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.schpa <- c()
     if ("schpa" %in% organisms.list)
     {
       tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
     }
     return(tips_to_keep.schpa)
   })
   
   tips_to_keep.eutsa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.eutsa <- c()
     if ("eutsa" %in% organisms.list)
     {
       tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
     }
     return(tips_to_keep.eutsa)
   })
   
   tips_to_keep.thlar2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.thlar <- c()
     if ("thlar" %in% organisms.list)
     {
       tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
     }
     return(tips_to_keep.thlar)
   })
   
   tips_to_keep.boest2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.boest <- c()
     if ("boest" %in% organisms.list)
     {
       tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
     }
     return(tips_to_keep.boest)
   })
   
   tips_to_keep.carpa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.carpa <- c()
     if ("carpa" %in% organisms.list)
     {
       tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
     }
     return(tips_to_keep.carpa)
   })
   
   tips_to_keep.theca2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.theca <- c()
     if ("theca" %in% organisms.list)
     {
       tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
     }
     return(tips_to_keep.theca)
   })
   
   tips_to_keep.goshi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.goshi <- c()
     if ("goshi" %in% organisms.list)
     {
       tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
     }
     return(tips_to_keep.goshi)
   })
   
   tips_to_keep.gosra2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.gosra <- c()
     if ("gosra" %in% organisms.list)
     {
       tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
     }
     return(tips_to_keep.gosra)
   })
   
   tips_to_keep.citsi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.citsi <- c()
     if ("citsi" %in% organisms.list)
     {
       tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
     }
     return(tips_to_keep.citsi)
   })
   
   tips_to_keep.pontr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.pontr <- c()
     if ("pontr" %in% organisms.list)
     {
       tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
     }
     return(tips_to_keep.pontr)
   })
   
   tips_to_keep.eucgr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.eucgr <- c()
     if ("eucgr" %in% organisms.list)
     {
       tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
     }
     return(tips_to_keep.eucgr)
   })
   
   tips_to_keep.maldo2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.maldo <- c()
     if ("maldo" %in% organisms.list)
     {
       tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
     }
     return(tips_to_keep.maldo)
   })
   
   tips_to_keep.prupe2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.prupe <- c()
     if ("prupe" %in% organisms.list)
     {
       tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
     }
     return(tips_to_keep.prupe)
   })
   
   tips_to_keep.frave2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.frave <- c()
     if ("frave" %in% organisms.list)
     {
       tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
     }
     return(tips_to_keep.frave)
   })
   
   tips_to_keep.corav2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.corav <- c()
     if ("corav" %in% organisms.list)
     {
       tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
     }
     return(tips_to_keep.corav)
   })
   
   tips_to_keep.caril2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.caril <- c()
     if ("caril" %in% organisms.list)
     {
       tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
     }
     return(tips_to_keep.caril)
   })
   
   tips_to_keep.queru2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.queru <- c()
     if ("queru" %in% organisms.list)
     {
       tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
     }
     return(tips_to_keep.queru)
   })
   
   tips_to_keep.cucsa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cucsa <- c()
     if ("cucsa" %in% organisms.list)
     {
       tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
     }
     return(tips_to_keep.cucsa)
   })
   
   tips_to_keep.glyma2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.glyma <- c()
     if ("glyma" %in% organisms.list)
     {
       tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
     }
     return(tips_to_keep.glyma)
   })
   
   tips_to_keep.medtr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.medtr <- c()
     if ("medtr" %in% organisms.list)
     {
       tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
     }
     return(tips_to_keep.medtr)
   })
   
   tips_to_keep.tripr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.tripr <- c()
     if ("tripr" %in% organisms.list)
     {
       tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
     }
     return(tips_to_keep.tripr)
   })
   
   tips_to_keep.cicar2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cicar <- c()
     if ("cicar" %in% organisms.list)
     {
       tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
     }
     return(tips_to_keep.cicar)
   })
   
   tips_to_keep.lotja2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.lotja <- c()
     if ("lotja" %in% organisms.list)
     {
       tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
     }
     return(tips_to_keep.lotja)
   })
   
   tips_to_keep.phaac2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.phaac <- c()
     if ("phaac" %in% organisms.list)
     {
       tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
     }
     return(tips_to_keep.phaac)
   })
   
   tips_to_keep.vigun2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.vigun <- c()
     if ("vigun" %in% organisms.list)
     {
       tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
     }
     return(tips_to_keep.vigun)
   })
   
   tips_to_keep.lupal2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.lupal <- c()
     if ("lupal" %in% organisms.list)
     {
       tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
     }
     return(tips_to_keep.lupal)
   })
   
   tips_to_keep.lencu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.lencu <- c()
     if ("lencu" %in% organisms.list)
     {
       tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
     }
     return(tips_to_keep.lencu)
   })
   
   tips_to_keep.arahy2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.arahy <- c()
     if ("arahy" %in% organisms.list)
     {
       tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
     }
     return(tips_to_keep.arahy)
   })
   
   tips_to_keep.poptr2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.poptr <- c()
     if ("poptr" %in% organisms.list)
     {
       tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
     }
     return(tips_to_keep.poptr)
   })
   
   tips_to_keep.manes2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.manes <- c()
     if ("manes" %in% organisms.list)
     {
       tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
     }
     return(tips_to_keep.manes)
   })
   
   tips_to_keep.salpu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.salpu <- c()
     if ("salpu" %in% organisms.list)
     {
       tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
     }
     return(tips_to_keep.salpu)
   })
   
   tips_to_keep.linus2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.linus <- c()
     if ("linus" %in% organisms.list)
     {
       tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
     }
     return(tips_to_keep.linus)
   })
   
   tips_to_keep.corci2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.corci <- c()
     if ("corci" %in% organisms.list)
     {
       tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
     }
     return(tips_to_keep.corci)
   })
   
   tips_to_keep.vitvi2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.vitvi <- c()
     if ("vitvi" %in% organisms.list)
     {
       tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
     }
     return(tips_to_keep.vitvi)
   })
   
   tips_to_keep.kalfe2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.kalfe <- c()
     if ("kalfe" %in% organisms.list)
     {
       tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
     }
     return(tips_to_keep.kalfe)
   })
   
   tips_to_keep.vacda2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.vacda <- c()
     if ("vacda" %in% organisms.list)
     {
       tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
     }
     return(tips_to_keep.vacda)
   })
   
   tips_to_keep.oleeu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.oleeu <- c()
     if ("oleeu" %in% organisms.list)
     {
       tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
     }
     return(tips_to_keep.oleeu)
   })
   
   tips_to_keep.mimgu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.mimgu <- c()
     if ("mimgu" %in% organisms.list)
     {
       tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
     }
     return(tips_to_keep.mimgu)
   })
   
   tips_to_keep.soltu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.soltu <- c()
     if ("soltu" %in% organisms.list)
     {
       tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
     }
     return(tips_to_keep.soltu)
   })
   
   tips_to_keep.cofar2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.cofar <- c()
     if ("cofar" %in% organisms.list)
     {
       tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
     }
     return(tips_to_keep.cofar)
   })
   
   tips_to_keep.lacsa2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.lacsa <- c()
     if ("lacsa" %in% organisms.list)
     {
       tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
     }
     return(tips_to_keep.lacsa)
   })
   
   tips_to_keep.dauca2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.dauca <- c()
     if ("dauca" %in% organisms.list)
     {
       tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
     }
     return(tips_to_keep.dauca)
   })
   
   tips_to_keep.betvu2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.betvu <- c()
     if ("betvu" %in% organisms.list)
     {
       tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
     }
     return(tips_to_keep.betvu)
   })
   
   tips_to_keep.amahy2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.amahy <- c()
     if ("amahy" %in% organisms.list)
     {
       tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
     }
     return(tips_to_keep.amahy)
   })
   
   tips_to_keep.chequ2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.chequ <- c()
     if ("chequ" %in% organisms.list)
     {
       tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
     }
     return(tips_to_keep.chequ)
   })
   
   tips_to_keep.sapof2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.sapof <- c()
     if ("sapof" %in% organisms.list)
     {
       tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
     }
     return(tips_to_keep.sapof)
   })
   
   tips_to_keep.aquco2 <- reactive({
     
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     tips_to_keep.aquco <- c()
     if ("aquco" %in% organisms.list)
     {
       tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
     }
     return(tips_to_keep.aquco)
   }) %>% bindEvent(input$run_button2)
   
   
   # Create complete gene tree with the proper name for each gene
   # For this, we split the species name apart from the gene name
   tree_adj2 <- reactive({
     tree <- tree2()
     organisms.list <- c(selected_values_org2())
     
     if ("mp" %in% organisms.list)
     {
       tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label) 
       if (length(tips_to_keep.mp) != 0)
       {
         mp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mp] <- mp.v
       }
     }
     
     if ("ot" %in% organisms.list)
     {
       tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label) 
       if (length(tips_to_keep.ot) != 0)
       {
         ost.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ot]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ot] <- ost.v
       }
     }
     
     if ("at" %in% organisms.list)
     {
       tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label) 
       if (length(tips_to_keep.at) != 0)
       {
         arabi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.at]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.at] <- arabi.v
       }
     }
     
     if ("cp" %in% organisms.list)
     {
       tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label) 
       if (length(tips_to_keep.cp) != 0)
       {
         cer.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cp]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cp] <- cer.v
       }
     }
     
     if ("cr" %in% organisms.list)
     {
       tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
       if (length(tips_to_keep.cr) != 0)
       {
         chlamy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cr]), "_"), 
                            function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cr] <- chlamy.v
       }
     }
     
     if ("cz" %in% organisms.list)
     {
       tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label) 
       if (length(tips_to_keep.cz) != 0)
       {
         chromo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cz]), "_"), 
                            function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cz] <- chromo.v
       }
     }
     
     if ("kn" %in% organisms.list)
     {
       tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label) 
       if (length(tips_to_keep.kn) != 0)
       {
         klebs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kn]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.kn] <- klebs.v
       }
     }
     
     if ("me" %in% organisms.list)
     {
       tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label) 
       if (length(tips_to_keep.me) != 0)
       {
         meso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.me]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.me] <- meso.v
       }
     }
     
     if ("mi" %in% organisms.list)
     {
       tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label) 
       if (length(tips_to_keep.mi) != 0)
       {
         micro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mi]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mi] <- micro.v
       }
     }
     
     if ("pp" %in% organisms.list)
     {
       tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label) 
       if (length(tips_to_keep.pp) != 0)
       {
         phys.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pp]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pp] <- phys.v
       }
     }
     
     if ("sl" %in% organisms.list)
     {
       tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label) 
       if (length(tips_to_keep.sl) != 0)
       {
         sola.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sl]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sl] <- sola.v
       }
     }
     
     if ("sm" %in% organisms.list)
     {
       tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label) 
       if (length(tips_to_keep.sm) != 0)
       {
         sel.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sm]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sm] <- sel.v
       }
     }
     
     if ("sp" %in% organisms.list)
     {
       tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label) 
       if (length(tips_to_keep.sp) != 0)
       {
         spiro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sp]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sp] <- spiro.v
       }
     }
     
     if ("ta" %in% organisms.list)
     {
       tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label) 
       if (length(tips_to_keep.ta) != 0)
       {
         tri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ta]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ta] <- tri.v
       }
     }
     
     if ("vc" %in% organisms.list)
     {
       tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
       if (length(tips_to_keep.vc) != 0)
       {
         vc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vc]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vc] <- vc.v
       }
     }
     
     if ("bp" %in% organisms.list)
     {
       tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
       if (length(tips_to_keep.bp) != 0)
       {
         bp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.bp] <- bp.v
       }
     }
     
     if ("cri" %in% organisms.list)
     {
       tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
       if (length(tips_to_keep.cri) != 0)
       {
         cri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cri]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cri] <- cri.v
       }
     }
     
     if ("ds" %in% organisms.list)
     {
       tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
       if (length(tips_to_keep.ds) != 0)
       {
         ds.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ds]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ds] <- ds.v
       }
     }
     
     if ("os" %in% organisms.list)
     {
       tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
       if (length(tips_to_keep.os) != 0)
       {
         os.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.os]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.os] <- os.v
       }
     }
     
     if ("smag" %in% organisms.list)
     {
       tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
       if (length(tips_to_keep.smag) != 0)
       {
         smag.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.smag]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.smag] <- smag.v
       }
     }
     
     if ("tp" %in% organisms.list)
     {
       tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
       if (length(tips_to_keep.tp) != 0)
       {
         tp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tp] <- tp.v
       }
     }
     
     if ("aa" %in% organisms.list)
     {
       tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
       if (length(tips_to_keep.aa) != 0)
       {
         aa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aa]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aa] <- aa.v
       }
     }
     
     if ("um" %in% organisms.list)
     {
       tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
       if (length(tips_to_keep.um) != 0)
       {
         um.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.um]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.um] <- um.v
       }
     }
     
     if ("rs" %in% organisms.list)
     {
       tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
       if (length(tips_to_keep.rs) != 0)
       {
         rs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.rs]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.rs] <- rs.v
       }
     }
     
     if ("cyc" %in% organisms.list)
     {
       tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
       if (length(tips_to_keep.cyc) != 0)
       {
         cyc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cyc]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cyc] <- cyc.v
       }
     }
     
     if ("ca" %in% organisms.list)
     {
       tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
       if (length(tips_to_keep.ca) != 0)
       {
         ca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ca]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ca] <- ca.v
       }
     }
     
     if ("mv" %in% organisms.list)
     {
       tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
       if (length(tips_to_keep.mv) != 0)
       {
         mv.vec2 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mv]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mv] <- mv.vec2
       }
     }
     
     if ("af" %in% organisms.list)
     {
       tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
       if (length(tips_to_keep.af) != 0)
       {
         af.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.af]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.af] <- af.v
       }
     }
     
     if ("sc" %in% organisms.list)
     {
       tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
       if (length(tips_to_keep.sc) != 0)
       {
         sc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sc]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sc] <- sc.v
       }
     }
     
     if ("aegi" %in% organisms.list)
     {
       tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
       if (length(tips_to_keep.aegi) != 0)
       {
         aegi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aegi]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aegi] <- aegi.v
       }
     }
     
     if ("sb" %in% organisms.list)
     {
       tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
       if (length(tips_to_keep.sb) != 0)
       {
         sb.vec2 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sb]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sb] <- sb.vec2
       }
     }
     
     if ("chara" %in% organisms.list)
     {
       tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
       if (length(tips_to_keep.chara) != 0)
       {
         chara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chara]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chara] <- chara.v
       }
     }
     
     if ("sceobli" %in% organisms.list)
     {
       tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
       if (length(tips_to_keep.sceobli) != 0)
       {
         sceobli.vec2 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sceobli]), "_"), 
                                function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sceobli] <- sceobli.vec2
       }
     }
     
     if ("cocco" %in% organisms.list)
     {
       tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
       if (length(tips_to_keep.cocco) != 0)
       {
         cocco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cocco]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cocco] <- cocco.v
       }
     }
     
     if ("haema" %in% organisms.list)
     {
       tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
       if (length(tips_to_keep.haema) != 0)
       {
         haema.vec2 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.haema]), "_"), 
                              function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.haema] <- haema.vec2
       }
     }
     
     if ("zm" %in% organisms.list)
     {
       tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
       if (length(tips_to_keep.zm) != 0)
       {
         zm.vec2 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zm]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zm] <- zm.vec2
       }
     }
     
     
     if ("praco" %in% organisms.list)
     {
       tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
       if (length(tips_to_keep.praco) != 0)
       {
         praco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.praco]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.praco] <- praco.v
       }
     }
     
     if ("ostlu" %in% organisms.list)
     {
       tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
       if (length(tips_to_keep.ostlu) != 0)
       {
         ostlu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostlu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostlu] <- ostlu.v
       }
     }
     
     if ("ostme" %in% organisms.list)
     {
       tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
       if (length(tips_to_keep.ostme) != 0)
       {
         ostme.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostme]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostme] <- ostme.v
       }
     }
     
     if ("micco" %in% organisms.list)
     {
       tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
       if (length(tips_to_keep.micco) != 0)
       {
         micco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.micco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.micco] <- micco.v
       }
     }
     
     if ("cymte" %in% organisms.list)
     {
       tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
       if (length(tips_to_keep.cymte) != 0)
       {
         cymte.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cymte]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cymte] <- cymte.v
       }
     }
     
     if ("chlin" %in% organisms.list)
     {
       tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
       if (length(tips_to_keep.chlin) != 0)
       {
         chlin.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlin]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlin] <- chlin.v
       }
     }
     
     if ("gonpe" %in% organisms.list)
     {
       tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
       if (length(tips_to_keep.gonpe) != 0)
       {
         gonpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gonpe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gonpe] <- gonpe.v
       }
     }
     
     if ("tetso" %in% organisms.list)
     {
       tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
       if (length(tips_to_keep.tetso) != 0)
       {
         tetso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tetso] <- tetso.v
       }
     }
     
     if ("desar" %in% organisms.list)
     {
       tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
       if (length(tips_to_keep.desar) != 0)
       {
         desar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.desar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.desar] <- desar.v
       }
     }
     
     if ("monmi" %in% organisms.list)
     {
       tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
       if (length(tips_to_keep.monmi) != 0)
       {
         monmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.monmi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.monmi] <- monmi.v
       }
     }
     
     if ("caule" %in% organisms.list)
     {
       tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
       if (length(tips_to_keep.caule) != 0)
       {
         caule.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caule]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.caule] <- caule.v
       }
     }
     
     if ("ostqu" %in% organisms.list)
     {
       tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
       if (length(tips_to_keep.ostqu) != 0)
       {
         ostqu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostqu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostqu] <- ostqu.v
       }
     }
     
     if ("brysp" %in% organisms.list)
     {
       tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
       if (length(tips_to_keep.brysp) != 0)
       {
         brysp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brysp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.brysp] <- brysp.v
       }
     }
     
     if ("ellbi" %in% organisms.list)
     {
       tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
       if (length(tips_to_keep.ellbi) != 0)
       {
         ellbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ellbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ellbi] <- ellbi.v
       }
     }
     
     if ("myrbi" %in% organisms.list)
     {
       tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
       if (length(tips_to_keep.myrbi) != 0)
       {
         myrbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.myrbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.myrbi] <- myrbi.v
       }
     }
     
     if ("apalo" %in% organisms.list)
     {
       tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
       if (length(tips_to_keep.apalo) != 0)
       {
         apalo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.apalo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.apalo] <- apalo.v
       }
     }
     
     if ("astgl" %in% organisms.list)
     {
       tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
       if (length(tips_to_keep.astgl) != 0)
       {
         astgl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.astgl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.astgl] <- astgl.v
       }
     }
     
     if ("tresp" %in% organisms.list)
     {
       tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
       if (length(tips_to_keep.tresp) != 0)
       {
         tresp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tresp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tresp] <- tresp.v
       }
     }
     
     if ("picso" %in% organisms.list)
     {
       tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
       if (length(tips_to_keep.picso) != 0)
       {
         picso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picso] <- picso.v
       }
     }
     
     if ("chlso" %in% organisms.list)
     {
       tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
       if (length(tips_to_keep.chlso) != 0)
       {
         chlso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlso] <- chlso.v
       }
     }
     
     if ("auxpr" %in% organisms.list)
     {
       tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
       if (length(tips_to_keep.auxpr) != 0)
       {
         auxpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.auxpr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.auxpr] <- auxpr.v
       }
     }
     
     if ("helsp" %in% organisms.list)
     {
       tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
       if (length(tips_to_keep.helsp) != 0)
       {
         helsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.helsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.helsp] <- helsp.v
       }
     }
     
     if ("tetst" %in% organisms.list)
     {
       tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
       if (length(tips_to_keep.tetst) != 0)
       {
         tetst.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetst]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tetst] <- tetst.v
       }
     }
     
     if ("pedsp" %in% organisms.list)
     {
       tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
       if (length(tips_to_keep.pedsp) != 0)
       {
         pedsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pedsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pedsp] <- pedsp.v
       }
     }
     
     if ("chlpr" %in% organisms.list)
     {
       tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
       if (length(tips_to_keep.chlpr) != 0)
       {
         chlpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlpr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlpr] <- chlpr.v
       }
     }
     
     if ("picsp" %in% organisms.list)
     {
       tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
       if (length(tips_to_keep.picsp) != 0)
       {
         picsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picsp] <- picsp.v
       }
     }
     
     if ("mesvb" %in% organisms.list)
     {
       tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
       if (length(tips_to_keep.mesvb) != 0)
       {
         mesvb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mesvb]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mesvb] <- mesvb.v
       }
     }
     
     if ("meskr" %in% organisms.list)
     {
       tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
       if (length(tips_to_keep.meskr) != 0)
       {
         meskr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.meskr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.meskr] <- meskr.v
       }
     }
     
     if ("zygci" %in% organisms.list)
     {
       tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
       if (length(tips_to_keep.zygci) != 0)
       {
         zygci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygci]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygci] <- zygci.v
       }
     }
     
     if ("zygcb" %in% organisms.list)
     {
       tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
       if (length(tips_to_keep.zygcb) != 0)
       {
         zygcb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcb]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygcb] <- zygcb.v
       }
     }
     
     if ("zygcy" %in% organisms.list)
     {
       tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
       if (length(tips_to_keep.zygcy) != 0)
       {
         zygcy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygcy] <- zygcy.v
       }
     }
     
     if ("plesc" %in% organisms.list)
     {
       tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
       if (length(tips_to_keep.plesc) != 0)
       {
         plesc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.plesc]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.plesc] <- plesc.v
       }
     }
     
     if ("funhy" %in% organisms.list)
     {
       tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
       if (length(tips_to_keep.funhy) != 0)
       {
         funhy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.funhy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.funhy] <- funhy.v
       }
     }
     
     if ("sphfa" %in% organisms.list)
     {
       tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
       if (length(tips_to_keep.sphfa) != 0)
       {
         sphfa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sphfa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sphfa] <- sphfa.v
       }
     }
     
     if ("takle" %in% organisms.list)
     {
       tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
       if (length(tips_to_keep.takle) != 0)
       {
         takle.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.takle]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.takle] <- takle.v
       }
     }
     
     if ("marpa" %in% organisms.list)
     {
       tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
       if (length(tips_to_keep.marpa) != 0)
       {
         marpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.marpa] <- marpa.v
       }
     }
     
     if ("ricfl" %in% organisms.list)
     {
       tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
       if (length(tips_to_keep.ricfl) != 0)
       {
         ricfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricfl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ricfl] <- ricfl.v
       }
     }
     
     if ("ricso" %in% organisms.list)
     {
       tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
       if (length(tips_to_keep.ricso) != 0)
       {
         ricso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ricso] <- ricso.v
       }
     }
     
     if ("antpu" %in% organisms.list)
     {
       tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
       if (length(tips_to_keep.antpu) != 0)
       {
         antpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antpu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.antpu] <- antpu.v
       }
     }
     
     if ("antfu" %in% organisms.list)
     {
       tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
       if (length(tips_to_keep.antfu) != 0)
       {
         antfu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antfu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.antfu] <- antfu.v
       }
     }
     
     if ("megfl" %in% organisms.list)
     {
       tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
       if (length(tips_to_keep.megfl) != 0)
       {
         megfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.megfl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.megfl] <- megfl.v
       }
     }
     
     if ("phach" %in% organisms.list)
     {
       tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
       if (length(tips_to_keep.phach) != 0)
       {
         phach.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phach]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phach] <- phach.v
       }
     }
     
     if ("phyph" %in% organisms.list)
     {
       tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
       if (length(tips_to_keep.phyph) != 0)
       {
         phyph.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phyph]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phyph] <- phyph.v
       }
     }
     
     if ("parpe" %in% organisms.list)
     {
       tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
       if (length(tips_to_keep.parpe) != 0)
       {
         parpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.parpe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.parpe] <- parpe.v
       }
     }
     
     if ("notor" %in% organisms.list)
     {
       tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
       if (length(tips_to_keep.notor) != 0)
       {
         notor.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.notor]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.notor] <- notor.v
       }
     }
     
     if ("phaca" %in% organisms.list)
     {
       tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
       if (length(tips_to_keep.phaca) != 0)
       {
         phaca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phaca] <- phaca.v
       }
     }
     
     if ("phasp" %in% organisms.list)
     {
       tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
       if (length(tips_to_keep.phasp) != 0)
       {
         phasp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phasp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phasp] <- phasp.v
       }
     }
     
     if ("leidu" %in% organisms.list)
     {
       tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
       if (length(tips_to_keep.leidu) != 0)
       {
         leidu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.leidu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.leidu] <- leidu.v
       }
     }
     
     if ("isosi" %in% organisms.list)
     {
       tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
       if (length(tips_to_keep.isosi) != 0)
       {
         isosi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.isosi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.isosi] <- isosi.v
       }
     }
     
     if ("dipco" %in% organisms.list)
     {
       tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
       if (length(tips_to_keep.dipco) != 0)
       {
         dipco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dipco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dipco] <- dipco.v
       }
     }
     
     if ("hupas" %in% organisms.list)
     {
       tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
       if (length(tips_to_keep.hupas) != 0)
       {
         hupas.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.hupas]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.hupas] <- hupas.v
       }
     }
     
     if ("lyccl" %in% organisms.list)
     {
       tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
       if (length(tips_to_keep.lyccl) != 0)
       {
         lyccl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lyccl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lyccl] <- lyccl.v
       }
     }
     
     if ("adica" %in% organisms.list)
     {
       tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
       if (length(tips_to_keep.adica) != 0)
       {
         adica.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.adica]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.adica] <- adica.v
       }
     }
     
     if ("alssp" %in% organisms.list)
     {
       tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
       if (length(tips_to_keep.alssp) != 0)
       {
         alssp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.alssp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.alssp] <- alssp.v
       }
     }
     
     if ("marve" %in% organisms.list)
     {
       tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
       if (length(tips_to_keep.marve) != 0)
       {
         marve.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marve]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.marve] <- marve.v
       }
     }
     
     if ("seqse" %in% organisms.list)
     {
       tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
       if (length(tips_to_keep.seqse) != 0)
       {
         seqse.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqse]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.seqse] <- seqse.v
       }
     }
     
     if ("seqgi" %in% organisms.list)
     {
       tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
       if (length(tips_to_keep.seqgi) != 0)
       {
         seqgi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqgi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.seqgi] <- seqgi.v
       }
     }
     
     if ("taxch" %in% organisms.list)
     {
       tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
       if (length(tips_to_keep.taxch) != 0)
       {
         taxch.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.taxch]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.taxch] <- taxch.v
       }
     }
     
     if ("abial" %in% organisms.list)
     {
       tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
       if (length(tips_to_keep.abial) != 0)
       {
         abial.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.abial]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.abial] <- abial.v
       }
     }
     
     if ("picab" %in% organisms.list)
     {
       tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
       if (length(tips_to_keep.picab) != 0)
       {
         picab.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picab]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picab] <- picab.v
       }
     }
     
     if ("gnemo" %in% organisms.list)
     {
       tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
       if (length(tips_to_keep.gnemo) != 0)
       {
         gnemo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gnemo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gnemo] <- gnemo.v
       }
     }
     
     if ("welmi" %in% organisms.list)
     {
       tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
       if (length(tips_to_keep.welmi) != 0)
       {
         welmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.welmi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.welmi] <- welmi.v
       }
     }
     
     if ("ginbi" %in% organisms.list)
     {
       tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
       if (length(tips_to_keep.ginbi) != 0)
       {
         ginbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ginbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ginbi] <- ginbi.v
       }
     }
     
     if ("ambtr" %in% organisms.list)
     {
       tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
       if (length(tips_to_keep.ambtr) != 0)
       {
         ambtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ambtr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ambtr] <- ambtr.v
       }
     }
     
     if ("nymco" %in% organisms.list)
     {
       tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
       if (length(tips_to_keep.nymco) != 0)
       {
         nymco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.nymco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.nymco] <- nymco.v
       }
     }
     
     if ("cinka" %in% organisms.list)
     {
       tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
       if (length(tips_to_keep.cinka) != 0)
       {
         cinka.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cinka]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cinka] <- cinka.v
       }
     }
     
     
     if ("horvu" %in% organisms.list)
     {
       tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
       if (length(tips_to_keep.horvu) != 0)
       {
         horvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.horvu]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.horvu] <- horvu.v
       }
     }
     
     if ("bradi" %in% organisms.list)
     {
       tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
       if (length(tips_to_keep.bradi) != 0)
       {
         bradi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bradi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.bradi] <- bradi.v
       }
     }
     
     if ("setit" %in% organisms.list)
     {
       tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
       if (length(tips_to_keep.setit) != 0)
       {
         setit.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.setit]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.setit] <- setit.v
       }
     }
     
     if ("panha" %in% organisms.list)
     {
       tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
       if (length(tips_to_keep.panha) != 0)
       {
         panha.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.panha]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.panha] <- panha.v
       }
     }
     
     if ("missi" %in% organisms.list)
     {
       tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
       if (length(tips_to_keep.missi) != 0)
       {
         missi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.missi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.missi] <- missi.v
       }
     }
     
     if ("oroth" %in% organisms.list)
     {
       tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
       if (length(tips_to_keep.oroth) != 0)
       {
         oroth.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oroth]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.oroth] <- oroth.v
       }
     }
     
     if ("eleco" %in% organisms.list)
     {
       tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
       if (length(tips_to_keep.eleco) != 0)
       {
         eleco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eleco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eleco] <- eleco.v
       }
     }
     
     if ("anaco" %in% organisms.list)
     {
       tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
       if (length(tips_to_keep.anaco) != 0)
       {
         anaco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.anaco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.anaco] <- anaco.v
       }
     }
     
     if ("musac" %in% organisms.list)
     {
       tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
       if (length(tips_to_keep.musac) != 0)
       {
         musac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.musac]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.musac] <- musac.v
       }
     }
     
     if ("dioal" %in% organisms.list)
     {
       tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
       if (length(tips_to_keep.dioal) != 0)
       {
         dioal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dioal]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dioal] <- dioal.v
       }
     }
     
     if ("spipo" %in% organisms.list)
     {
       tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
       if (length(tips_to_keep.spipo) != 0)
       {
         spipo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.spipo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.spipo] <- spipo.v
       }
     }
     
     if ("aspof" %in% organisms.list)
     {
       tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
       if (length(tips_to_keep.aspof) != 0)
       {
         aspof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aspof]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aspof] <- aspof.v
       }
     }
     
     if ("zosma" %in% organisms.list)
     {
       tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
       if (length(tips_to_keep.zosma) != 0)
       {
         zosma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zosma]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zosma] <- zosma.v
       }
     }
     
     if ("araly" %in% organisms.list)
     {
       tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
       if (length(tips_to_keep.araly) != 0)
       {
         araly.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.araly]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.araly] <- araly.v
       }
     }
     
     if ("capgr" %in% organisms.list)
     {
       tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
       if (length(tips_to_keep.capgr) != 0)
       {
         capgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.capgr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.capgr] <- capgr.v
       }
     }
     
     if ("brara" %in% organisms.list)
     {
       tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
       if (length(tips_to_keep.brara) != 0)
       {
         brara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brara]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.brara] <- brara.v
       }
     }
     
     if ("braol" %in% organisms.list)
     {
       tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
       if (length(tips_to_keep.braol) != 0)
       {
         braol.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.braol]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.braol] <- braol.v
       }
     }
     
     if ("sisir" %in% organisms.list)
     {
       tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
       if (length(tips_to_keep.sisir) != 0)
       {
         sisir.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sisir]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sisir] <- sisir.v
       }
     }
     
     if ("schpa" %in% organisms.list)
     {
       tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
       if (length(tips_to_keep.schpa) != 0)
       {
         schpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.schpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.schpa] <- schpa.v
       }
     }
     
     if ("eutsa" %in% organisms.list)
     {
       tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
       if (length(tips_to_keep.eutsa) != 0)
       {
         eutsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eutsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eutsa] <- eutsa.v
       }
     }
     
     if ("thlar" %in% organisms.list)
     {
       tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
       if (length(tips_to_keep.thlar) != 0)
       {
         thlar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.thlar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.thlar] <- thlar.v
       }
     }
     
     if ("boest" %in% organisms.list)
     {
       tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
       if (length(tips_to_keep.boest) != 0)
       {
         boest.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.boest]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.boest] <- boest.v
       }
     }
     
     if ("carpa" %in% organisms.list)
     {
       tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
       if (length(tips_to_keep.carpa) != 0)
       {
         carpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.carpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.carpa] <- carpa.v
       }
     }
     
     if ("theca" %in% organisms.list)
     {
       tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
       if (length(tips_to_keep.theca) != 0)
       {
         theca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.theca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.theca] <- theca.v
       }
     }
     
     if ("goshi" %in% organisms.list)
     {
       tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
       if (length(tips_to_keep.goshi) != 0)
       {
         goshi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.goshi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.goshi] <- goshi.v
       }
     }
     
     if ("gosra" %in% organisms.list)
     {
       tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
       if (length(tips_to_keep.gosra) != 0)
       {
         gosra.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gosra]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gosra] <- gosra.v
       }
     }
     
     if ("citsi" %in% organisms.list)
     {
       tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
       if (length(tips_to_keep.citsi) != 0)
       {
         citsi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.citsi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.citsi] <- citsi.v
       }
     }
     
     if ("pontr" %in% organisms.list)
     {
       tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
       if (length(tips_to_keep.pontr) != 0)
       {
         pontr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pontr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pontr] <- pontr.v
       }
     }
     
     if ("eucgr" %in% organisms.list)
     {
       tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
       if (length(tips_to_keep.eucgr) != 0)
       {
         eucgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eucgr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eucgr] <- eucgr.v
       }
     }
     
     if ("maldo" %in% organisms.list)
     {
       tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
       if (length(tips_to_keep.maldo) != 0)
       {
         maldo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.maldo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.maldo] <- maldo.v
       }
     }
     
     if ("prupe" %in% organisms.list)
     {
       tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
       if (length(tips_to_keep.prupe) != 0)
       {
         prupe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.prupe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.prupe] <- prupe.v
       }
     }
     
     if ("frave" %in% organisms.list)
     {
       tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
       if (length(tips_to_keep.frave) != 0)
       {
         frave.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.frave]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.frave] <- frave.v
       }
     }
     
     if ("corav" %in% organisms.list)
     {
       tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
       if (length(tips_to_keep.corav) != 0)
       {
         corav.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corav]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.corav] <- corav.v
       }
     }
     
     if ("caril" %in% organisms.list)
     {
       tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
       if (length(tips_to_keep.caril) != 0)
       {
         caril.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caril]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.caril] <- caril.v
       }
     }
     
     if ("queru" %in% organisms.list)
     {
       tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
       if (length(tips_to_keep.queru) != 0)
       {
         queru.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.queru]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.queru] <- queru.v
       }
     }
     
     if ("cucsa" %in% organisms.list)
     {
       tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
       if (length(tips_to_keep.cucsa) != 0)
       {
         cucsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cucsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cucsa] <- cucsa.v
       }
     }
     
     if ("glyma" %in% organisms.list)
     {
       tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
       if (length(tips_to_keep.glyma) != 0)
       {
         glyma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.glyma]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.glyma] <- glyma.v
       }
     }
     
     if ("medtr" %in% organisms.list)
     {
       tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
       if (length(tips_to_keep.medtr) != 0)
       {
         medtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.medtr]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.medtr] <- medtr.v
       }
     }
     
     if ("tripr" %in% organisms.list)
     {
       tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
       if (length(tips_to_keep.tripr) != 0)
       {
         tripr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tripr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tripr] <- tripr.v
       }
     }
     
     if ("cicar" %in% organisms.list)
     {
       tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
       if (length(tips_to_keep.cicar) != 0)
       {
         cicar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cicar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cicar] <- cicar.v
       }
     }
     
     if ("lotja" %in% organisms.list)
     {
       tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
       if (length(tips_to_keep.lotja) != 0)
       {
         lotja.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lotja]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lotja] <- lotja.v
       }
     }
     
     if ("phaac" %in% organisms.list)
     {
       tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
       if (length(tips_to_keep.phaac) != 0)
       {
         phaac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaac]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phaac] <- phaac.v
       }
     }
     
     if ("vigun" %in% organisms.list)
     {
       tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
       if (length(tips_to_keep.vigun) != 0)
       {
         vigun.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vigun]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vigun] <- vigun.v
       }
     }
     
     if ("lupal" %in% organisms.list)
     {
       tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
       if (length(tips_to_keep.lupal) != 0)
       {
         lupal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lupal]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lupal] <- lupal.v
       }
     }
     
     if ("lencu" %in% organisms.list)
     {
       tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
       if (length(tips_to_keep.lencu) != 0)
       {
         lencu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lencu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lencu] <- lencu.v
       }
     }
     
     if ("arahy" %in% organisms.list)
     {
       tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
       if (length(tips_to_keep.arahy) != 0)
       {
         arahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.arahy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.arahy] <- arahy.v
       }
     }
     
     if ("poptr" %in% organisms.list)
     {
       tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
       if (length(tips_to_keep.poptr) != 0)
       {
         poptr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.poptr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.poptr] <- poptr.v
       }
     }
     
     if ("manes" %in% organisms.list)
     {
       tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
       if (length(tips_to_keep.manes) != 0)
       {
         manes.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.manes]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.manes] <- manes.v
       }
     }
     
     if ("salpu" %in% organisms.list)
     {
       tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
       if (length(tips_to_keep.salpu) != 0)
       {
         salpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.salpu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.salpu] <- salpu.v
       }
     }
     
     if ("linus" %in% organisms.list)
     {
       tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
       if (length(tips_to_keep.linus) != 0)
       {
         linus.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.linus]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.linus] <- linus.v
       }
     }
     
     if ("corci" %in% organisms.list)
     {
       tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
       if (length(tips_to_keep.corci) != 0)
       {
         corci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corci]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.corci] <- corci.v
       }
     }
     
     if ("vitvi" %in% organisms.list)
     {
       tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
       if (length(tips_to_keep.vitvi) != 0)
       {
         vitvi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vitvi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vitvi] <- vitvi.v
       }
     }
     
     if ("kalfe" %in% organisms.list)
     {
       tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
       if (length(tips_to_keep.kalfe) != 0)
       {
         kalfe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kalfe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.kalfe] <- kalfe.v
       }
     }
     
     if ("vacda" %in% organisms.list)
     {
       tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
       if (length(tips_to_keep.vacda) != 0)
       {
         vacda.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vacda]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vacda] <- vacda.v
       }
     }
     
     if ("oleeu" %in% organisms.list)
     {
       tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
       if (length(tips_to_keep.oleeu) != 0)
       {
         oleeu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oleeu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.oleeu] <- oleeu.v
       }
     }
     
     if ("mimgu" %in% organisms.list)
     {
       tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
       if (length(tips_to_keep.mimgu) != 0)
       {
         mimgu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mimgu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mimgu] <- mimgu.v
       }
     }
     
     if ("soltu" %in% organisms.list)
     {
       tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
       if (length(tips_to_keep.soltu) != 0)
       {
         soltu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.soltu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.soltu] <- soltu.v
       }
     }
     
     if ("cofar" %in% organisms.list)
     {
       tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
       if (length(tips_to_keep.cofar) != 0)
       {
         cofar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cofar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cofar] <- cofar.v
       }
     }
     
     if ("lacsa" %in% organisms.list)
     {
       tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
       if (length(tips_to_keep.lacsa) != 0)
       {
         lacsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lacsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lacsa] <- lacsa.v
       }
     }
     
     if ("dauca" %in% organisms.list)
     {
       tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
       if (length(tips_to_keep.dauca) != 0)
       {
         dauca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dauca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dauca] <- dauca.v
       }
     }
     
     if ("betvu" %in% organisms.list)
     {
       tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
       if (length(tips_to_keep.betvu) != 0)
       {
         betvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.betvu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.betvu] <- betvu.v
       }
     }
     
     if ("amahy" %in% organisms.list)
     {
       tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
       if (length(tips_to_keep.amahy) != 0)
       {
         amahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.amahy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.amahy] <- amahy.v
       }
     }
     
     if ("chequ" %in% organisms.list)
     {
       tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
       if (length(tips_to_keep.chequ) != 0)
       {
         chequ.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chequ]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chequ] <- chequ.v
       }
     }
     
     if ("sapof" %in% organisms.list)
     {
       tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
       if (length(tips_to_keep.sapof) != 0)
       {
         sapof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sapof]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sapof] <- sapof.v
       }
     }
     
     if ("aquco" %in% organisms.list)
     {
       tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
       if (length(tips_to_keep.aquco) != 0)
       {
         aquco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aquco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aquco] <- aquco.v
       }
     }
     
     return(tree)
   }) %>% bindEvent(input$run_button2)
   
   # Generate reduced tree when the corresponding button is activated
   tree_reduced2 <- reactive({
     
     tree <- tree_adj2()
     # Define tips to keep (selected organisms) and generate the reduced tree
     tips_to_keep.global <- c(tips_to_keep.praco2(), tips_to_keep.ot2(), tips_to_keep.ostlu2(), tips_to_keep.ostme2(), tips_to_keep.bp2(),
                              tips_to_keep.mi2(), tips_to_keep.micco2(), tips_to_keep.cymte2(), tips_to_keep.cr2(), tips_to_keep.chlin2(),
                              tips_to_keep.haema2(), tips_to_keep.vc2(), tips_to_keep.ds2(), tips_to_keep.gonpe2(), tips_to_keep.tetso2(),
                              tips_to_keep.cz2(), tips_to_keep.rs2(), tips_to_keep.sceobli2(), tips_to_keep.desar2(), tips_to_keep.monmi2(),
                              tips_to_keep.caule2(), tips_to_keep.ostqu2(), tips_to_keep.brysp2(), tips_to_keep.cocco2(),
                              tips_to_keep.ellbi2(), tips_to_keep.myrbi2(), tips_to_keep.apalo2(), tips_to_keep.astgl2(),
                              tips_to_keep.tresp2(), tips_to_keep.picso2(), tips_to_keep.chlso2(), tips_to_keep.auxpr2(),
                              tips_to_keep.helsp2(), tips_to_keep.um2(), tips_to_keep.tetst2(), tips_to_keep.pedsp2(),
                              tips_to_keep.chlpr2(), tips_to_keep.picsp2(), tips_to_keep.ca2(), tips_to_keep.mv2(), tips_to_keep.mesvb2(),
                              tips_to_keep.kn2(), tips_to_keep.chara2(), tips_to_keep.me2(), tips_to_keep.meskr2(), tips_to_keep.sp2(),
                              tips_to_keep.zygci2(), tips_to_keep.zygcb2(), tips_to_keep.zygcy2(),
                              tips_to_keep.cp2(), tips_to_keep.pp2(), tips_to_keep.plesc2(), tips_to_keep.funhy2(),
                              tips_to_keep.smag2(), tips_to_keep.sphfa2(), tips_to_keep.takle2(), tips_to_keep.mp2(),
                              tips_to_keep.marpa2(), tips_to_keep.ricfl2(), tips_to_keep.ricso2(), tips_to_keep.aa2(),
                              tips_to_keep.antpu2(), tips_to_keep.antfu2(), tips_to_keep.megfl2(), tips_to_keep.phach2(),
                              tips_to_keep.phyph2(), tips_to_keep.parpe2(), tips_to_keep.notor2(), tips_to_keep.phaca2(),
                              tips_to_keep.phasp2(), tips_to_keep.leidu2(), tips_to_keep.sm2(), tips_to_keep.isosi2(), 
                              tips_to_keep.dipco2(), tips_to_keep.hupas2(), tips_to_keep.lyccl2(),
                              tips_to_keep.cri2(), tips_to_keep.adica2(), tips_to_keep.alssp2(), tips_to_keep.sc2(), 
                              tips_to_keep.af2(), tips_to_keep.marve2(), tips_to_keep.tp2(), tips_to_keep.seqse2(), 
                              tips_to_keep.seqgi2(), tips_to_keep.taxch2(), tips_to_keep.abial2(), tips_to_keep.picab2(),
                              tips_to_keep.gnemo2(), tips_to_keep.welmi2(), tips_to_keep.cyc2(), tips_to_keep.ginbi2(),
                              tips_to_keep.ambtr2(), tips_to_keep.nymco2(), tips_to_keep.cinka2(), tips_to_keep.aegi2(),
                              tips_to_keep.ta2(), tips_to_keep.horvu2(), tips_to_keep.bradi2(), tips_to_keep.os2(),
                              tips_to_keep.setit2(), tips_to_keep.sb2(), tips_to_keep.zm2(), tips_to_keep.panha2(),
                              tips_to_keep.missi2(), tips_to_keep.oroth2(), tips_to_keep.eleco2(), tips_to_keep.anaco2(),
                              tips_to_keep.musac2(), tips_to_keep.dioal2(), tips_to_keep.spipo2(), tips_to_keep.aspof2(),
                              tips_to_keep.zosma2(), tips_to_keep.at2(), tips_to_keep.araly2(), tips_to_keep.capgr2(),
                              tips_to_keep.brara2(), tips_to_keep.braol2(), tips_to_keep.sisir2(), tips_to_keep.schpa2(),
                              tips_to_keep.eutsa2(), tips_to_keep.thlar2(), tips_to_keep.boest2(), tips_to_keep.carpa2(),
                              tips_to_keep.theca2(), tips_to_keep.goshi2(), tips_to_keep.gosra2(), tips_to_keep.citsi2(),
                              tips_to_keep.pontr2(), tips_to_keep.eucgr2(), tips_to_keep.maldo2(), tips_to_keep.prupe2(),
                              tips_to_keep.frave2(), tips_to_keep.corav2(), tips_to_keep.caril2(), tips_to_keep.queru2(),
                              tips_to_keep.cucsa2(), tips_to_keep.glyma2(), tips_to_keep.medtr2(), tips_to_keep.tripr2(),
                              tips_to_keep.cicar2(), tips_to_keep.lotja2(), tips_to_keep.phaac2(), tips_to_keep.vigun2(),
                              tips_to_keep.lupal2(), tips_to_keep.lencu2(), tips_to_keep.arahy2(), tips_to_keep.poptr2(),
                              tips_to_keep.manes2(), tips_to_keep.salpu2(), tips_to_keep.linus2(), tips_to_keep.corci2(),
                              tips_to_keep.vitvi2(), tips_to_keep.kalfe2(), tips_to_keep.vacda2(), tips_to_keep.oleeu2(),
                              tips_to_keep.mimgu2(), tips_to_keep.sl2(), tips_to_keep.soltu2(), tips_to_keep.cofar2(),
                              tips_to_keep.lacsa2(), tips_to_keep.dauca2(), tips_to_keep.betvu2(), tips_to_keep.amahy2(),
                              tips_to_keep.chequ2(), tips_to_keep.sapof2(), tips_to_keep.aquco2())
     
     # Error message if trying to build tree with less than two tips
     if (length(tips_to_keep.global) < 2)
     {
       shinyjs::hideElement(id = 'loading.tree2')
       
       if (UI_exist_tree2)
       {
         removeUI(
           selector = "div:has(>> #tree_seq_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #treeTips2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs2)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree2 <<- F
       output$error_tree2 <- renderUI({renderText({print("Unable to construct 
      tree with a single tip, please select more organisms.")})})
       validate(need(length(tips_to_keep.global) > 1, " "))
     }
     
     
     # Error message if query gene does not belong to selected organisms due to
     # not having selected chosen proteome in organisms list
     diamond_table <- diamond_table2()
     
     if (!(as.character(diamond_table$ID[1]) %in% tree$tip.label))
     {
       shinyjs::hideElement(id = 'loading.tree2')
       
       if (UI_exist_tree2)
       {
         removeUI(
           selector = "div:has(>> #tree_seq_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #treeTips2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs2)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree2 <<- F
       output$error_tree2 <- renderUI({renderText({print("Please select the species corresponding
      to the query gene. Read the instructions for clarification.")})})
       validate(need(as.character(diamond_table$ID[1]) %in% tree$tip.label, " "))
     }
     
     
     tips_to_drop <- setdiff(1:length(tree$tip.label), tips_to_keep.global)
     tree_reduced <- drop.tip(tree, tips_to_drop)
     
     return(tree_reduced)
   }) %>% bindEvent(input$run_button2)
   
   ### Select orthogroup sequences based on the reduced tree
   ortho_reduced2 <- reactive({
     
     tree_reduced <- tree_reduced2()
     ortho_seq <- ortho_seq2()
     ortho_reduced <- ortho_seq[tree_reduced$tip.label]
     return(ortho_reduced)
   }) %>% bindEvent(input$run_button2)
  
   organims_reduced2 <- reactive({
     
     tree_reduced <- tree_reduced2()
     
     len.mp <- length(tips_to_keep.mp2())
     len.ot <- length(tips_to_keep.ot2())
     len.at <- length(tips_to_keep.at2())
     len.cp <- length(tips_to_keep.cp2())
     len.cr <- length(tips_to_keep.cr2())
     len.cz <- length(tips_to_keep.cz2())
     len.kn <- length(tips_to_keep.kn2())
     len.me <- length(tips_to_keep.me2())
     len.mi <- length(tips_to_keep.mi2())
     len.pp <- length(tips_to_keep.pp2())
     len.sl <- length(tips_to_keep.sl2())
     len.sm <- length(tips_to_keep.sm2())
     len.sp <- length(tips_to_keep.sp2())
     len.ta <- length(tips_to_keep.ta2())
     len.vc <- length(tips_to_keep.vc2())
     len.bp <- length(tips_to_keep.bp2())
     len.cri <- length(tips_to_keep.cri2())
     len.ds <- length(tips_to_keep.ds2())
     len.os <- length(tips_to_keep.os2())
     len.smag <- length(tips_to_keep.smag2())
     len.tp <- length(tips_to_keep.tp2())
     len.aa <- length(tips_to_keep.aa2())
     len.um <- length(tips_to_keep.um2())
     len.rs <- length(tips_to_keep.rs2())
     len.cyc <- length(tips_to_keep.cyc2())
     len.ca <- length(tips_to_keep.ca2())
     len.mv <- length(tips_to_keep.mv2())
     len.af <- length(tips_to_keep.af2())
     len.sc <- length(tips_to_keep.sc2())
     len.aegi <- length(tips_to_keep.aegi2())
     len.sb <- length(tips_to_keep.sb2())
     len.chara <- length(tips_to_keep.chara2())
     len.sceobli <- length(tips_to_keep.sceobli2())
     len.cocco <- length(tips_to_keep.cocco2())
     len.haema <- length(tips_to_keep.haema2())
     len.zea <- length(tips_to_keep.zm2())
     len.praco <- length(tips_to_keep.praco2())
     len.ostlu <- length(tips_to_keep.ostlu2())
     len.ostme <- length(tips_to_keep.ostme2())
     len.micco <- length(tips_to_keep.micco2())
     len.cymte <- length(tips_to_keep.cymte2())
     len.chlin <- length(tips_to_keep.chlin2())
     len.gonpe <- length(tips_to_keep.gonpe2())
     len.tetso <- length(tips_to_keep.tetso2())
     len.desar <- length(tips_to_keep.desar2())
     len.monmi <- length(tips_to_keep.monmi2())
     len.caule <- length(tips_to_keep.caule2())
     len.ostqu <- length(tips_to_keep.ostqu2())
     len.brysp <- length(tips_to_keep.brysp2())
     len.ellbi <- length(tips_to_keep.ellbi2())
     len.myrbi <- length(tips_to_keep.myrbi2())
     len.apalo <- length(tips_to_keep.apalo2())
     len.astgl <- length(tips_to_keep.astgl2())
     len.tresp <- length(tips_to_keep.tresp2())
     len.picso <- length(tips_to_keep.picso2())
     len.chlso <- length(tips_to_keep.chlso2())
     len.auxpr <- length(tips_to_keep.auxpr2())
     len.helsp <- length(tips_to_keep.helsp2())
     len.tetst <- length(tips_to_keep.tetst2())
     len.pedsp <- length(tips_to_keep.pedsp2())
     len.chlpr <- length(tips_to_keep.chlpr2())
     len.picsp <- length(tips_to_keep.picsp2())
     len.mesvb <- length(tips_to_keep.mesvb2())
     len.meskr <- length(tips_to_keep.meskr2())
     len.zygci <- length(tips_to_keep.zygci2())
     len.zygcb <- length(tips_to_keep.zygcb2())
     len.zygcy <- length(tips_to_keep.zygcy2())
     len.plesc <- length(tips_to_keep.plesc2())
     len.funhy <- length(tips_to_keep.funhy2())
     len.sphfa <- length(tips_to_keep.sphfa2())
     len.takle <- length(tips_to_keep.takle2())
     len.marpa <- length(tips_to_keep.marpa2())
     len.ricfl <- length(tips_to_keep.ricfl2())
     len.ricso <- length(tips_to_keep.ricso2())
     len.antpu <- length(tips_to_keep.antpu2())
     len.antfu <- length(tips_to_keep.antfu2())
     len.megfl <- length(tips_to_keep.megfl2())
     len.phach <- length(tips_to_keep.phach2())
     len.phyph <- length(tips_to_keep.phyph2())
     len.parpe <- length(tips_to_keep.parpe2())
     len.notor <- length(tips_to_keep.notor2())
     len.phaca <- length(tips_to_keep.phaca2())
     len.phasp <- length(tips_to_keep.phasp2())
     len.leidu <- length(tips_to_keep.leidu2())
     len.isosi <- length(tips_to_keep.isosi2())
     len.dipco <- length(tips_to_keep.dipco2())
     len.hupas <- length(tips_to_keep.hupas2())
     len.lyccl <- length(tips_to_keep.lyccl2())
     len.adica <- length(tips_to_keep.adica2())
     len.alssp <- length(tips_to_keep.alssp2())
     len.marve <- length(tips_to_keep.marve2())
     len.seqse <- length(tips_to_keep.seqse2())
     len.seqgi <- length(tips_to_keep.seqgi2())
     len.taxch <- length(tips_to_keep.taxch2())
     len.abial <- length(tips_to_keep.abial2())
     len.picab <- length(tips_to_keep.picab2())
     len.gnemo <- length(tips_to_keep.gnemo2())
     len.welmi <- length(tips_to_keep.welmi2())
     len.ginbi <- length(tips_to_keep.ginbi2())
     len.ambtr <- length(tips_to_keep.ambtr2())
     len.nymco <- length(tips_to_keep.nymco2())
     len.cinka <- length(tips_to_keep.cinka2())
     len.horvu <- length(tips_to_keep.horvu2())
     len.bradi <- length(tips_to_keep.bradi2())
     len.setit <- length(tips_to_keep.setit2())
     len.panha <- length(tips_to_keep.panha2())
     len.missi <- length(tips_to_keep.missi2())
     len.oroth <- length(tips_to_keep.oroth2())
     len.eleco <- length(tips_to_keep.eleco2())
     len.anaco <- length(tips_to_keep.anaco2())
     len.musac <- length(tips_to_keep.musac2())
     len.dioal <- length(tips_to_keep.dioal2())
     len.spipo <- length(tips_to_keep.spipo2())
     len.aspof <- length(tips_to_keep.aspof2())
     len.zosma <- length(tips_to_keep.zosma2())
     len.araly <- length(tips_to_keep.araly2())
     len.capgr <- length(tips_to_keep.capgr2())
     len.brara <- length(tips_to_keep.brara2())
     len.braol <- length(tips_to_keep.braol2())
     len.sisir <- length(tips_to_keep.sisir2())
     len.schpa <- length(tips_to_keep.schpa2())
     len.eutsa <- length(tips_to_keep.eutsa2())
     len.thlar <- length(tips_to_keep.thlar2())
     len.boest <- length(tips_to_keep.boest2())
     len.carpa <- length(tips_to_keep.carpa2())
     len.theca <- length(tips_to_keep.theca2())
     len.goshi <- length(tips_to_keep.goshi2())
     len.gosra <- length(tips_to_keep.gosra2())
     len.citsi <- length(tips_to_keep.citsi2())
     len.pontr <- length(tips_to_keep.pontr2())
     len.eucgr <- length(tips_to_keep.eucgr2())
     len.maldo <- length(tips_to_keep.maldo2())
     len.prupe <- length(tips_to_keep.prupe2())
     len.frave <- length(tips_to_keep.frave2())
     len.corav <- length(tips_to_keep.corav2())
     len.caril <- length(tips_to_keep.caril2())
     len.queru <- length(tips_to_keep.queru2())
     len.cucsa <- length(tips_to_keep.cucsa2())
     len.glyma <- length(tips_to_keep.glyma2())
     len.medtr <- length(tips_to_keep.medtr2())
     len.tripr <- length(tips_to_keep.tripr2())
     len.cicar <- length(tips_to_keep.cicar2())
     len.lotja <- length(tips_to_keep.lotja2())
     len.phaac <- length(tips_to_keep.phaac2())
     len.vigun <- length(tips_to_keep.vigun2())
     len.lupal <- length(tips_to_keep.lupal2())
     len.lencu <- length(tips_to_keep.lencu2())
     len.arahy <- length(tips_to_keep.arahy2())
     len.poptr <- length(tips_to_keep.poptr2())
     len.manes <- length(tips_to_keep.manes2())
     len.salpu <- length(tips_to_keep.salpu2())
     len.linus <- length(tips_to_keep.linus2())
     len.corci <- length(tips_to_keep.corci2())
     len.vitvi <- length(tips_to_keep.vitvi2())
     len.kalfe <- length(tips_to_keep.kalfe2())
     len.vacda <- length(tips_to_keep.vacda2())
     len.oleeu <- length(tips_to_keep.oleeu2())
     len.mimgu <- length(tips_to_keep.mimgu2())
     len.soltu <- length(tips_to_keep.soltu2())
     len.cofar <- length(tips_to_keep.cofar2())
     len.lacsa <- length(tips_to_keep.lacsa2())
     len.dauca <- length(tips_to_keep.dauca2())
     len.betvu <- length(tips_to_keep.betvu2())
     len.amahy <- length(tips_to_keep.amahy2())
     len.chequ <- length(tips_to_keep.chequ2())
     len.sapof <- length(tips_to_keep.sapof2())
     len.aquco <- length(tips_to_keep.aquco2())
     
     
     organims_reduced <- c(rep("Prasinoderma coloniale", len.praco),rep("Ostreococcus tauri", len.ot),
                           rep("Ostreococcus lucimarinus", len.ostlu),rep("Ostreococcus mediterraneus", len.ostme),
                           rep("Bathycoccus prasinos", len.bp),rep("Micromonas pusilla", len.mi),
                           rep("Micromonas commoda", len.micco),rep("Cymbomonas tetramitiformis", len.cymte),
                           rep("Chlamydomonas reinhardtii", len.cr),rep("Chlamydomonas incerta", len.chlin),
                           rep("Haematococcus lacustris", len.haema),rep("Volvox carteri", len.vc),
                           rep("Dunaliella salina", len.ds),rep("Gonium pectorale", len.gonpe),
                           rep("Tetrabaena socialis", len.tetso),rep("Chromochloris zofingiensis", len.cz),
                           rep("Raphidocelis subcapitata", len.rs),rep("Scenedesmus obliquus", len.sceobli),
                           rep("Desmodesmus armatus", len.desar),rep("Monoraphidium minutum", len.monmi),
                           rep("Caulerpa lentillifera", len.caule),rep("Ostreobium quekettii", len.ostqu),
                           rep("Bryopsis sp.", len.brysp),rep("Coccomyxa subellipsoidea", len.cocco),
                           rep("Elliptochloris bilobata", len.ellbi),rep("Myrmecia bisecta", len.myrbi),
                           rep("Apatococcus lobatus", len.apalo),rep("Asterochloris glomerata", len.astgl),
                           rep("Trebouxia sp.", len.tresp),rep("Picochlorum soloecismus", len.picso),
                           rep("Chlorella sorokiniana", len.chlso),rep("Auxenochlorella protothecoides", len.auxpr),
                           rep("Helicosporidium sp.", len.helsp),rep("Ulva mutabilis", len.um),
                           rep("Tetraselmis striata", len.tetst),rep("Pedinophyceae sp.", len.pedsp),
                           rep("Chloropicon primus", len.chlpr),rep("Picocystis sp.", len.picsp),
                           rep("Chlorokybus atmophyticus", len.ca),rep("Mesostigma viride CCAC 1140", len.mv),
                           rep("Mesostigma viride NIES 296", len.mesvb),rep("Klebsormidium nitens", len.kn),
                           rep("Chara braunii", len.chara),rep("Mesotaenium endlicherianum", len.me),
                           rep("Mesotaenium kramstae", len.meskr),rep("Spirogloea muscicola", len.sp),
                           rep("Zygnema circumcarinatum SAG", len.zygci),rep("Zygnema circumcarinatum UTEX 1559", len.zygcb),
                           rep("Zygnema cylindricum", len.zygcy),rep("Ceratodon purpureus", len.cp),
                           rep("Physcomitrium patens", len.pp),rep("Pleurozium schreberi", len.plesc),
                           rep("Funaria hygrometrica", len.funhy),rep("Sphagnum magellanicum", len.smag),
                           rep("Sphagnum fallax", len.sphfa),rep("Takakia lepidozioides", len.takle),
                           rep("Marchantia polymorpha", len.mp),rep("Marchantia paleacea", len.marpa),
                           rep("Riccia fluitans", len.ricfl),rep("Riccia sorocarpa", len.ricso),
                           rep("Anthoceros agrestis", len.aa),rep("Anthoceros punctatus", len.antpu),
                           rep("Anthoceros fusiformis", len.antfu),rep("Megaceros flagellaris", len.megfl),
                           rep("Phaeomegaceros chiloensis", len.phach),rep("Phymatoceros phymatodes", len.phyph),
                           rep("Paraphymatoceros pearsonii", len.parpe),rep("Notothylas orbicularis", len.notor),
                           rep("Phaeoceros carolinianus", len.phaca),rep("Phaeoceros sp.", len.phasp),
                           rep("Leiosporoceros dussii", len.leidu),rep("Selaginella moellendorffii", len.sm),
                           rep("Isoetes sinensis", len.isosi),rep("Diphasiastrum complanatum", len.dipco),
                           rep("Huperzia asiatica", len.hupas),rep("Lycopodium clavatum", len.lyccl),
                           rep("Ceratopteris richardii", len.cri),rep("Adiantum capillus", len.adica),
                           rep("Alsophila spinulosa", len.alssp),rep("Salvinia cucullata", len.sc),
                           rep("Azolla filiculoides", len.af),rep("Marsilea vestita", len.marve),
                           rep("Thuja plicata", len.tp),rep("Sequoia sempervirens", len.seqse),
                           rep("Sequoiadendron giganteum", len.seqgi),rep("Taxus chinensis", len.taxch),
                           rep("Abies alba", len.abial),rep("Picea abies", len.picab),rep("Gnetum montanum", len.gnemo),
                           rep("Welwitschia mirabilis", len.welmi),rep("Cycas panzhihuaensis", len.cyc),
                           rep("Ginkgo biloba", len.ginbi),rep("Amborella trichopoda", len.ambtr),
                           rep("Nymphaea colorata", len.nymco),rep("Cinnamomum kanehirae", len.cinka),
                           rep("Aegilops tauschii", len.aegi),rep("Triticum aestivum", len.ta),
                           rep("Hordeum vulgare", len.horvu),rep("Brachypodium distachyon", len.bradi),
                           rep("Oryza sativa", len.os),rep("Setaria italica", len.setit),rep("Sorghum bicolor", len.sb),
                           rep("Zea mays", len.zea),rep("Panicum hallii", len.panha),rep("Miscanthus sinensis", len.missi),
                           rep("Oropetium thomaeum", len.oroth),rep("Eleusine coracana", len.eleco),
                           rep("Ananas comosus", len.anaco),rep("Musa acuminata", len.musac),rep("Dioscorea alata", len.dioal),
                           rep("Spirodela polyrhiza", len.spipo),rep("Asparagus officinalis", len.aspof),
                           rep("Zostera marina", len.zosma),rep("Arabidopsis thaliana", len.at),
                           rep("Arabidopsis lyrata", len.araly),rep("Capsella grandiflora", len.capgr),
                           rep("Brassica rapa", len.brara),rep("Brassica oleracea", len.braol),
                           rep("Sisymbrium irio", len.sisir),rep("Schrenkiella parvula", len.schpa),
                           rep("Eutrema salsugineum", len.eutsa),rep("Thlaspi arvense", len.thlar),
                           rep("Boechera stricta", len.boest),rep("Carica papaya", len.carpa),rep("Theobroma cacao", len.theca),
                           rep("Gossypium hirsutum", len.goshi),rep("Gossypium raimondii", len.gosra),
                           rep("Citrus sinensis", len.citsi),rep("Poncirus trifoliata", len.pontr),
                           rep("Eucalyptus grandis", len.eucgr),rep("Malus domestica", len.maldo),
                           rep("Prunus persica", len.prupe),rep("Fragaria vesca", len.frave),rep("Corylus avellana", len.corav),
                           rep("Carya illinoinensis", len.caril),rep("Quercus rubra", len.queru),
                           rep("Cucumis sativus", len.cucsa),rep("Glycine max", len.glyma),rep("Medicago truncatula", len.medtr),
                           rep("Trifolium pratense", len.tripr),rep("Cicer arietinum", len.cicar),
                           rep("Lotus japonicus", len.lotja),rep("Phaseolus acutifolius", len.phaac),
                           rep("Vigna unguiculata", len.vigun),rep("Lupinus albus", len.lupal),rep("Lens culinaris", len.lencu),
                           rep("Arachis hypogaea", len.arahy),rep("Populus trichocarpa", len.poptr),
                           rep("Manihot esculenta", len.manes),rep("Salix purpurea", len.salpu),
                           rep("Linum usitatissimum", len.linus),rep("Corymbia citriodora", len.corci),
                           rep("Vitis vinifera", len.vitvi),rep("Kalanchoe fedtschenkoi", len.kalfe),
                           rep("Vaccinium darrowii", len.vacda),rep("Olea europaea", len.oleeu),
                           rep("Mimulus guttatus", len.mimgu),rep("Solanum lycopersicum", len.sl),
                           rep("Solanum tuberosum", len.soltu),rep("Coffea arabica", len.cofar),
                           rep("Lactuca sativa", len.lacsa),rep("Daucus carota", len.dauca),rep("Beta vulgaris", len.betvu),
                           rep("Amaranthus hypochondriacus", len.amahy),rep("Chenopodium quinoa", len.chequ),
                           rep("Saponaria officinalis", len.sapof),rep("Aquilegia coerulea", len.aquco))
     
     return(organims_reduced)
   }) %>% bindEvent(input$run_button2)

   # Create plot of subset tree
   tree_plot2 <- reactive({
     
     # Define previous variables
     tree_reduced <- tree_reduced2()
     diamond_table <- diamond_table2()
     gene.name.tree <- as.character(diamond_table$ID[1])
     tree <- tree_adj2()
     
     tips_to_keep.mp <- tips_to_keep.mp2()
     tips_to_keep.ot <- tips_to_keep.ot2()
     tips_to_keep.at <- tips_to_keep.at2()
     tips_to_keep.cp <- tips_to_keep.cp2()
     tips_to_keep.cr <- tips_to_keep.cr2()
     tips_to_keep.cz <- tips_to_keep.cz2()
     tips_to_keep.kn <- tips_to_keep.kn2()
     tips_to_keep.me <- tips_to_keep.me2()
     tips_to_keep.mi <- tips_to_keep.mi2()
     tips_to_keep.pp <- tips_to_keep.pp2()
     tips_to_keep.sl <- tips_to_keep.sl2()
     tips_to_keep.sm <- tips_to_keep.sm2()
     tips_to_keep.sp <- tips_to_keep.sp2()
     tips_to_keep.ta <- tips_to_keep.ta2()
     tips_to_keep.vc <- tips_to_keep.vc2()
     tips_to_keep.bp <- tips_to_keep.bp2()
     tips_to_keep.cri <- tips_to_keep.cri2()
     tips_to_keep.ds <- tips_to_keep.ds2()
     tips_to_keep.os <- tips_to_keep.os2()
     tips_to_keep.smag <- tips_to_keep.smag2()
     tips_to_keep.tp <- tips_to_keep.tp2()
     tips_to_keep.aa <- tips_to_keep.aa2()
     tips_to_keep.um <- tips_to_keep.um2()
     tips_to_keep.rs <- tips_to_keep.rs2()
     tips_to_keep.cyc <- tips_to_keep.cyc2()
     tips_to_keep.ca <- tips_to_keep.ca2()
     tips_to_keep.mv <- tips_to_keep.mv2()
     tips_to_keep.af <- tips_to_keep.af2()
     tips_to_keep.sc <- tips_to_keep.sc2()
     tips_to_keep.aegi <- tips_to_keep.aegi2()
     tips_to_keep.sb <- tips_to_keep.sb2()
     tips_to_keep.chara <- tips_to_keep.chara2()
     tips_to_keep.sceobli <- tips_to_keep.sceobli2()
     tips_to_keep.cocco <- tips_to_keep.cocco2()
     tips_to_keep.haema <- tips_to_keep.haema2()
     tips_to_keep.zm <- tips_to_keep.zm2()
     tips_to_keep.praco <- tips_to_keep.praco2()
     tips_to_keep.ostlu <- tips_to_keep.ostlu2()
     tips_to_keep.ostme <- tips_to_keep.ostme2()
     tips_to_keep.micco <- tips_to_keep.micco2()
     tips_to_keep.cymte <- tips_to_keep.cymte2()
     tips_to_keep.chlin <- tips_to_keep.chlin2()
     tips_to_keep.gonpe <- tips_to_keep.gonpe2()
     tips_to_keep.tetso <- tips_to_keep.tetso2()
     tips_to_keep.desar <- tips_to_keep.desar2()
     tips_to_keep.monmi <- tips_to_keep.monmi2()
     tips_to_keep.caule <- tips_to_keep.caule2()
     tips_to_keep.ostqu <- tips_to_keep.ostqu2()
     tips_to_keep.brysp <- tips_to_keep.brysp2()
     tips_to_keep.ellbi <- tips_to_keep.ellbi2()
     tips_to_keep.myrbi <- tips_to_keep.myrbi2()
     tips_to_keep.apalo <- tips_to_keep.apalo2()
     tips_to_keep.astgl <- tips_to_keep.astgl2()
     tips_to_keep.tresp <- tips_to_keep.tresp2()
     tips_to_keep.picso <- tips_to_keep.picso2()
     tips_to_keep.chlso <- tips_to_keep.chlso2()
     tips_to_keep.auxpr <- tips_to_keep.auxpr2()
     tips_to_keep.helsp <- tips_to_keep.helsp2()
     tips_to_keep.tetst <- tips_to_keep.tetst2()
     tips_to_keep.pedsp <- tips_to_keep.pedsp2()
     tips_to_keep.chlpr <- tips_to_keep.chlpr2()
     tips_to_keep.picsp <- tips_to_keep.picsp2()
     tips_to_keep.mesvb <- tips_to_keep.mesvb2()
     tips_to_keep.meskr <- tips_to_keep.meskr2()
     tips_to_keep.zygci <- tips_to_keep.zygci2()
     tips_to_keep.zygcb <- tips_to_keep.zygcb2()
     tips_to_keep.zygcy <- tips_to_keep.zygcy2()
     tips_to_keep.plesc <- tips_to_keep.plesc2()
     tips_to_keep.funhy <- tips_to_keep.funhy2()
     tips_to_keep.sphfa <- tips_to_keep.sphfa2()
     tips_to_keep.takle <- tips_to_keep.takle2()
     tips_to_keep.marpa <- tips_to_keep.marpa2()
     tips_to_keep.ricfl <- tips_to_keep.ricfl2()
     tips_to_keep.ricso <- tips_to_keep.ricso2()
     tips_to_keep.antpu <- tips_to_keep.antpu2()
     tips_to_keep.antfu <- tips_to_keep.antfu2()
     tips_to_keep.megfl <- tips_to_keep.megfl2()
     tips_to_keep.phach <- tips_to_keep.phach2()
     tips_to_keep.phyph <- tips_to_keep.phyph2()
     tips_to_keep.parpe <- tips_to_keep.parpe2()
     tips_to_keep.notor <- tips_to_keep.notor2()
     tips_to_keep.phaca <- tips_to_keep.phaca2()
     tips_to_keep.phasp <- tips_to_keep.phasp2()
     tips_to_keep.leidu <- tips_to_keep.leidu2()
     tips_to_keep.isosi <- tips_to_keep.isosi2()
     tips_to_keep.dipco <- tips_to_keep.dipco2()
     tips_to_keep.hupas <- tips_to_keep.hupas2()
     tips_to_keep.lyccl <- tips_to_keep.lyccl2()
     tips_to_keep.adica <- tips_to_keep.adica2()
     tips_to_keep.alssp <- tips_to_keep.alssp2()
     tips_to_keep.marve <- tips_to_keep.marve2()
     tips_to_keep.seqse <- tips_to_keep.seqse2()
     tips_to_keep.seqgi <- tips_to_keep.seqgi2()
     tips_to_keep.taxch <- tips_to_keep.taxch2()
     tips_to_keep.abial <- tips_to_keep.abial2()
     tips_to_keep.picab <- tips_to_keep.picab2()
     tips_to_keep.gnemo <- tips_to_keep.gnemo2()
     tips_to_keep.welmi <- tips_to_keep.welmi2()
     tips_to_keep.ginbi <- tips_to_keep.ginbi2()
     tips_to_keep.ambtr <- tips_to_keep.ambtr2()
     tips_to_keep.nymco <- tips_to_keep.nymco2()
     tips_to_keep.cinka <- tips_to_keep.cinka2()
     tips_to_keep.horvu <- tips_to_keep.horvu2()
     tips_to_keep.bradi <- tips_to_keep.bradi2()
     tips_to_keep.setit <- tips_to_keep.setit2()
     tips_to_keep.panha <- tips_to_keep.panha2()
     tips_to_keep.missi <- tips_to_keep.missi2()
     tips_to_keep.oroth <- tips_to_keep.oroth2()
     tips_to_keep.eleco <- tips_to_keep.eleco2()
     tips_to_keep.anaco <- tips_to_keep.anaco2()
     tips_to_keep.musac <- tips_to_keep.musac2()
     tips_to_keep.dioal <- tips_to_keep.dioal2()
     tips_to_keep.spipo <- tips_to_keep.spipo2()
     tips_to_keep.aspof <- tips_to_keep.aspof2()
     tips_to_keep.zosma <- tips_to_keep.zosma2()
     tips_to_keep.araly <- tips_to_keep.araly2()
     tips_to_keep.capgr <- tips_to_keep.capgr2()
     tips_to_keep.brara <- tips_to_keep.brara2()
     tips_to_keep.braol <- tips_to_keep.braol2()
     tips_to_keep.sisir <- tips_to_keep.sisir2()
     tips_to_keep.schpa <- tips_to_keep.schpa2()
     tips_to_keep.eutsa <- tips_to_keep.eutsa2()
     tips_to_keep.thlar <- tips_to_keep.thlar2()
     tips_to_keep.boest <- tips_to_keep.boest2()
     tips_to_keep.carpa <- tips_to_keep.carpa2()
     tips_to_keep.theca <- tips_to_keep.theca2()
     tips_to_keep.goshi <- tips_to_keep.goshi2()
     tips_to_keep.gosra <- tips_to_keep.gosra2()
     tips_to_keep.citsi <- tips_to_keep.citsi2()
     tips_to_keep.pontr <- tips_to_keep.pontr2()
     tips_to_keep.eucgr <- tips_to_keep.eucgr2()
     tips_to_keep.maldo <- tips_to_keep.maldo2()
     tips_to_keep.prupe <- tips_to_keep.prupe2()
     tips_to_keep.frave <- tips_to_keep.frave2()
     tips_to_keep.corav <- tips_to_keep.corav2()
     tips_to_keep.caril <- tips_to_keep.caril2()
     tips_to_keep.queru <- tips_to_keep.queru2()
     tips_to_keep.cucsa <- tips_to_keep.cucsa2()
     tips_to_keep.glyma <- tips_to_keep.glyma2()
     tips_to_keep.medtr <- tips_to_keep.medtr2()
     tips_to_keep.tripr <- tips_to_keep.tripr2()
     tips_to_keep.cicar <- tips_to_keep.cicar2()
     tips_to_keep.lotja <- tips_to_keep.lotja2()
     tips_to_keep.phaac <- tips_to_keep.phaac2()
     tips_to_keep.vigun <- tips_to_keep.vigun2()
     tips_to_keep.lupal <- tips_to_keep.lupal2()
     tips_to_keep.lencu <- tips_to_keep.lencu2()
     tips_to_keep.arahy <- tips_to_keep.arahy2()
     tips_to_keep.poptr <- tips_to_keep.poptr2()
     tips_to_keep.manes <- tips_to_keep.manes2()
     tips_to_keep.salpu <- tips_to_keep.salpu2()
     tips_to_keep.linus <- tips_to_keep.linus2()
     tips_to_keep.corci <- tips_to_keep.corci2()
     tips_to_keep.vitvi <- tips_to_keep.vitvi2()
     tips_to_keep.kalfe <- tips_to_keep.kalfe2()
     tips_to_keep.vacda <- tips_to_keep.vacda2()
     tips_to_keep.oleeu <- tips_to_keep.oleeu2()
     tips_to_keep.mimgu <- tips_to_keep.mimgu2()
     tips_to_keep.soltu <- tips_to_keep.soltu2()
     tips_to_keep.cofar <- tips_to_keep.cofar2()
     tips_to_keep.lacsa <- tips_to_keep.lacsa2()
     tips_to_keep.dauca <- tips_to_keep.dauca2()
     tips_to_keep.betvu <- tips_to_keep.betvu2()
     tips_to_keep.amahy <- tips_to_keep.amahy2()
     tips_to_keep.chequ <- tips_to_keep.chequ2()
     tips_to_keep.sapof <- tips_to_keep.sapof2()
     tips_to_keep.aquco <- tips_to_keep.aquco2()
     
     if (length(tree_reduced$tip.label) < 2)
     {
       cat("")
     }
     else 
     {
       # Highlight the target gene
       high.gene2 <<- tree_reduced$tip.label[grep(pattern = gene.name.tree, tree_reduced$tip.label)]
       
       
       # Color asignment per species
       col.factor <- c()
       org.factor <- c()
       
       library(glue)
       library(ggtree)
       library(ggplot2)
       
       for (i in 1:length(tree_reduced$tip.label))
       {
         if (tree_reduced$tip.label[i] %in% high.gene2)
         {
           col.factor <- c(col.factor,"#CD0000")
           org.factor <- c(org.factor,"Gen of interest")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
         {
           col.factor <- c(col.factor,"#006400")
           org.factor <- c(org.factor,"Marchantia polymorpha")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
         {
           col.factor <- c(col.factor,"#00008B")
           org.factor <- c(org.factor,"Ostreococcus tauri")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
         {
           col.factor <- c(col.factor,"#CD661D")
           org.factor <- c(org.factor,"Arabidopsis thaliana")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
         {
           col.factor <- c(col.factor,"#458B74")
           org.factor <- c(org.factor,"Ceratodon purpureus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
         {
           col.factor <- c(col.factor,"#8B7355")
           org.factor <- c(org.factor,"Chlamydomonas reinhardtii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
         {
           col.factor <- c(col.factor,"#458B00")
           org.factor <- c(org.factor,"Chromochloris zofingiensis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
         {
           col.factor <- c(col.factor,"#CD1076")
           org.factor <- c(org.factor,"Klebsormidium nitens")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
         {
           col.factor <- c(col.factor,"#8B8878")
           org.factor <- c(org.factor,"Mesotaenium endlicherianum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
         {
           col.factor <- c(col.factor,"#666666")
           org.factor <- c(org.factor,"Micromonas pusilla")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
         {
           col.factor <- c(col.factor,"#B8860B")
           org.factor <- c(org.factor,"Physcomitrium patens")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
         {
           col.factor <- c(col.factor,"#8B008B")
           org.factor <- c(org.factor,"Solanum lycopersicum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
         {
           col.factor <- c(col.factor,"#6E8B3D")
           org.factor <- c(org.factor,"Selaginella moellendorffii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
         {
           col.factor <- c(col.factor,"#79CDCD")
           org.factor <- c(org.factor,"Spirogloea muscicola")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
         {
           col.factor <- c(col.factor,"#CDCD00")
           org.factor <- c(org.factor,"Triticum aestivum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
         {
           col.factor <- c(col.factor,"#16317d")
           org.factor <- c(org.factor,"Volvox carteri")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
         {
           col.factor <- c(col.factor,"#007e2f")
           org.factor <- c(org.factor,"Bathycoccus prasinos")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
         {
           col.factor <- c(col.factor,"#ffcd12")
           org.factor <- c(org.factor,"Ceratopteris richardii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
         {
           col.factor <- c(col.factor,"#b86092")
           org.factor <- c(org.factor,"Dunaliella salina")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
         {
           col.factor <- c(col.factor,"#721b3e")
           org.factor <- c(org.factor,"Oryza sativa")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
         {
           col.factor <- c(col.factor,"#00b7a7")
           org.factor <- c(org.factor,"Sphagnum magellanicum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
         {
           col.factor <- c(col.factor,"#67000d")
           org.factor <- c(org.factor,"Thuja plicata")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
         {
           col.factor <- c(col.factor,"#5b2c6f")
           org.factor <- c(org.factor,"Anthoceros agrestis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
         {
           col.factor <- c(col.factor,"#15e71b")
           org.factor <- c(org.factor,"Ulva mutabilis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
         {
           col.factor <- c(col.factor,"#2d04f9")
           org.factor <- c(org.factor,"Raphidocelis subcapitata")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
         {
           col.factor <- c(col.factor,"#873600")
           org.factor <- c(org.factor,"Cycas panzhihuaensis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
         {
           col.factor <- c(col.factor,"#0b5345")
           org.factor <- c(org.factor,"Chlorokybus atmophyticus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
         {
           col.factor <- c(col.factor,"#283747")
           org.factor <- c(org.factor,"Mesostigma viride CCAC 1140")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
         {
           col.factor <- c(col.factor,"#145a32")
           org.factor <- c(org.factor,"Azolla filiculoides")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
         {
           col.factor <- c(col.factor,"#3339e6")
           org.factor <- c(org.factor,"Salvinia cucullata")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
         {
           col.factor <- c(col.factor,"#e6338f")
           org.factor <- c(org.factor,"Aegilops tauschii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
         {
           col.factor <- c(col.factor,"#cd016a")
           org.factor <- c(org.factor,"Sorghum bicolor")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
         {
           col.factor <- c(col.factor,"#117a65")
           org.factor <- c(org.factor,"Chara braunii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
         {
           col.factor <- c(col.factor,"#148f77")
           org.factor <- c(org.factor,"Scenedesmus obliquus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
         {
           col.factor <- c(col.factor,"#9c640c")
           org.factor <- c(org.factor,"Coccomyxa subellipsoidea")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
         {
           col.factor <- c(col.factor,"#196f3d")
           org.factor <- c(org.factor,"Haematococcus lacustris")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
         {
           col.factor <- c(col.factor,"#666909")
           org.factor <- c(org.factor,"Zea mays")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.praco])
         {
           col.factor <- c(col.factor,"#ADFF2F")
           org.factor <- c(org.factor,"Prasinoderma coloniale")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostlu])
         {
           col.factor <- c(col.factor,"#483D8B")
           org.factor <- c(org.factor,"Ostreococcus lucimarinus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostme])
         {
           col.factor <- c(col.factor,"#7FFFD4")
           org.factor <- c(org.factor,"Ostreococcus mediterraneus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.micco])
         {
           col.factor <- c(col.factor,"#228B22")
           org.factor <- c(org.factor,"Micromonas commoda")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymte])
         {
           col.factor <- c(col.factor,"#00F5FF")
           org.factor <- c(org.factor,"Cymbomonas tetramitiformis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlin])
         {
           col.factor <- c(col.factor,"#FFE4B5")
           org.factor <- c(org.factor,"Chlamydomonas incerta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gonpe])
         {
           col.factor <- c(col.factor,"#8B3A62")
           org.factor <- c(org.factor,"Gonium pectorale")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetso])
         {
           col.factor <- c(col.factor,"#FFDAB9")
           org.factor <- c(org.factor,"Tetrabaena socialis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.desar])
         {
           col.factor <- c(col.factor,"#54FF9F")
           org.factor <- c(org.factor,"Desmodesmus armatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.monmi])
         {
           col.factor <- c(col.factor,"#20B2AA")
           org.factor <- c(org.factor,"Monoraphidium minutum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caule])
         {
           col.factor <- c(col.factor,"#9BCD9B")
           org.factor <- c(org.factor,"Caulerpa lentillifera")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostqu])
         {
           col.factor <- c(col.factor,"#5CACEE")
           org.factor <- c(org.factor,"Ostreobium quekettii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brysp])
         {
           col.factor <- c(col.factor,"#B8860B")
           org.factor <- c(org.factor,"Bryopsis sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ellbi])
         {
           col.factor <- c(col.factor,"#EEDD82")
           org.factor <- c(org.factor,"Elliptochloris bilobata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.myrbi])
         {
           col.factor <- c(col.factor,"#F0FFFF")
           org.factor <- c(org.factor,"Myrmecia bisecta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.apalo])
         {
           col.factor <- c(col.factor,"#C71585")
           org.factor <- c(org.factor,"Apatococcus lobatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.astgl])
         {
           col.factor <- c(col.factor,"#8B4789")
           org.factor <- c(org.factor,"Asterochloris glomerata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tresp])
         {
           col.factor <- c(col.factor,"#8B008B")
           org.factor <- c(org.factor,"Trebouxia sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picso])
         {
           col.factor <- c(col.factor,"#D2691E")
           org.factor <- c(org.factor,"Picochlorum soloecismus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlso])
         {
           col.factor <- c(col.factor,"#8B7D6B")
           org.factor <- c(org.factor,"Chlorella sorokiniana")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.auxpr])
         {
           col.factor <- c(col.factor,"#000080")
           org.factor <- c(org.factor,"Auxenochlorella protothecoides")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.helsp])
         {
           col.factor <- c(col.factor,"#A020F0")
           org.factor <- c(org.factor,"Helicosporidium sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetst])
         {
           col.factor <- c(col.factor,"#EE00EE")
           org.factor <- c(org.factor,"Tetraselmis striata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pedsp])
         {
           col.factor <- c(col.factor,"#8B1C62")
           org.factor <- c(org.factor,"Pedinophyceae sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlpr])
         {
           col.factor <- c(col.factor,"#FFFACD")
           org.factor <- c(org.factor,"Chloropicon primus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picsp])
         {
           col.factor <- c(col.factor,"#FFFFE0")
           org.factor <- c(org.factor,"Picocystis sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mesvb])
         {
           col.factor <- c(col.factor,"#A52A2A")
           org.factor <- c(org.factor,"Mesostigma viride NIES 296")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.meskr])
         {
           col.factor <- c(col.factor,"#E0FFFF")
           org.factor <- c(org.factor,"Mesotaenium kramstae")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygci])
         {
           col.factor <- c(col.factor,"#E6E6FA")
           org.factor <- c(org.factor,"Zygnema circumcarinatum SAG")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcb])
         {
           col.factor <- c(col.factor,"#DCDCDC")
           org.factor <- c(org.factor,"Zygnema circumcarinatum UTEX 1559")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcy])
         {
           col.factor <- c(col.factor,"#8B636C")
           org.factor <- c(org.factor,"Zygnema cylindricum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.plesc])
         {
           col.factor <- c(col.factor,"#CD2990")
           org.factor <- c(org.factor,"Pleurozium schreberi")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.funhy])
         {
           col.factor <- c(col.factor,"#CDC9C9")
           org.factor <- c(org.factor,"Funaria hygrometrica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sphfa])
         {
           col.factor <- c(col.factor,"#1C86EE")
           org.factor <- c(org.factor,"Sphagnum fallax")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.takle])
         {
           col.factor <- c(col.factor,"#8B7500")
           org.factor <- c(org.factor,"Takakia lepidozioides")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marpa])
         {
           col.factor <- c(col.factor,"#EED5B7")
           org.factor <- c(org.factor,"Marchantia paleacea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricfl])
         {
           col.factor <- c(col.factor,"#EE5C42")
           org.factor <- c(org.factor,"Riccia fluitans")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricso])
         {
           col.factor <- c(col.factor,"#FFFAFA")
           org.factor <- c(org.factor,"Riccia sorocarpa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antpu])
         {
           col.factor <- c(col.factor,"#D02090")
           org.factor <- c(org.factor,"Anthoceros punctatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antfu])
         {
           col.factor <- c(col.factor,"#FFF8DC")
           org.factor <- c(org.factor,"Anthoceros fusiformis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.megfl])
         {
           col.factor <- c(col.factor,"#EEA2AD")
           org.factor <- c(org.factor,"Megaceros flagellaris")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phach])
         {
           col.factor <- c(col.factor,"#BCD2EE")
           org.factor <- c(org.factor,"Phaeomegaceros chiloensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phyph])
         {
           col.factor <- c(col.factor,"#FF00FF")
           org.factor <- c(org.factor,"Phymatoceros phymatodes")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.parpe])
         {
           col.factor <- c(col.factor,"#68228B")
           org.factor <- c(org.factor,"Paraphymatoceros pearsonii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.notor])
         {
           col.factor <- c(col.factor,"#006400")
           org.factor <- c(org.factor,"Notothylas orbicularis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaca])
         {
           col.factor <- c(col.factor,"#00FFFF")
           org.factor <- c(org.factor,"Phaeoceros carolinianus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phasp])
         {
           col.factor <- c(col.factor,"#EED2EE")
           org.factor <- c(org.factor,"Phaeoceros sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.leidu])
         {
           col.factor <- c(col.factor,"#FFEFD5")
           org.factor <- c(org.factor,"Leiosporoceros dussii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.isosi])
         {
           col.factor <- c(col.factor,"#CDC8B1")
           org.factor <- c(org.factor,"Isoetes sinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dipco])
         {
           col.factor <- c(col.factor,"#8B6508")
           org.factor <- c(org.factor,"Diphasiastrum complanatum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.hupas])
         {
           col.factor <- c(col.factor,"#FFA54F")
           org.factor <- c(org.factor,"Huperzia asiatica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lyccl])
         {
           col.factor <- c(col.factor,"#EEAEEE")
           org.factor <- c(org.factor,"Lycopodium clavatum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.adica])
         {
           col.factor <- c(col.factor,"#FF0000")
           org.factor <- c(org.factor,"Adiantum capillus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.alssp])
         {
           col.factor <- c(col.factor,"#9B30FF")
           org.factor <- c(org.factor,"Alsophila spinulosa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marve])
         {
           col.factor <- c(col.factor,"#CD661D")
           org.factor <- c(org.factor,"Marsilea vestita")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqse])
         {
           col.factor <- c(col.factor,"#7AC5CD")
           org.factor <- c(org.factor,"Sequoia sempervirens")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqgi])
         {
           col.factor <- c(col.factor,"#CDC0B0")
           org.factor <- c(org.factor,"Sequoiadendron giganteum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.taxch])
         {
           col.factor <- c(col.factor,"#4682B4")
           org.factor <- c(org.factor,"Taxus chinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.abial])
         {
           col.factor <- c(col.factor,"#CDBA96")
           org.factor <- c(org.factor,"Abies alba")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picab])
         {
           col.factor <- c(col.factor,"#A2CD5A")
           org.factor <- c(org.factor,"Picea abies")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gnemo])
         {
           col.factor <- c(col.factor,"#FAFAD2")
           org.factor <- c(org.factor,"Gnetum montanum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.welmi])
         {
           col.factor <- c(col.factor,"#E0EEEE")
           org.factor <- c(org.factor,"Welwitschia mirabilis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ginbi])
         {
           col.factor <- c(col.factor,"#7A378B")
           org.factor <- c(org.factor,"Ginkgo biloba")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ambtr])
         {
           col.factor <- c(col.factor,"#FFDEAD")
           org.factor <- c(org.factor,"Amborella trichopoda")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.nymco])
         {
           col.factor <- c(col.factor,"#C0FF3E")
           org.factor <- c(org.factor,"Nymphaea colorata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cinka])
         {
           col.factor <- c(col.factor,"#7FFFD4")
           org.factor <- c(org.factor,"Cinnamomum kanehirae")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.horvu])
         {
           col.factor <- c(col.factor,"#1E90FF")
           org.factor <- c(org.factor,"Hordeum vulgare")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bradi])
         {
           col.factor <- c(col.factor,"#76EE00")
           org.factor <- c(org.factor,"Brachypodium distachyon")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.setit])
         {
           col.factor <- c(col.factor,"#EE9A49")
           org.factor <- c(org.factor,"Setaria italica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.panha])
         {
           col.factor <- c(col.factor,"#FFF0F5")
           org.factor <- c(org.factor,"Panicum hallii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.missi])
         {
           col.factor <- c(col.factor,"#CD9B1D")
           org.factor <- c(org.factor,"Miscanthus sinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oroth])
         {
           col.factor <- c(col.factor,"#36648B")
           org.factor <- c(org.factor,"Oropetium thomaeum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eleco])
         {
           col.factor <- c(col.factor,"#EEA9B8")
           org.factor <- c(org.factor,"Eleusine coracana")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.anaco])
         {
           col.factor <- c(col.factor,"#ADD8E6")
           org.factor <- c(org.factor,"Ananas comosus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.musac])
         {
           col.factor <- c(col.factor,"#66CDAA")
           org.factor <- c(org.factor,"Musa acuminata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dioal])
         {
           col.factor <- c(col.factor,"#FFF0F5")
           org.factor <- c(org.factor,"Dioscorea alata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.spipo])
         {
           col.factor <- c(col.factor,"#CD5555")
           org.factor <- c(org.factor,"Spirodela polyrhiza")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aspof])
         {
           col.factor <- c(col.factor,"#FFEC8B")
           org.factor <- c(org.factor,"Asparagus officinalis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zosma])
         {
           col.factor <- c(col.factor,"#668B8B")
           org.factor <- c(org.factor,"Zostera marina")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.araly])
         {
           col.factor <- c(col.factor,"#CDC673")
           org.factor <- c(org.factor,"Arabidopsis lyrata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.capgr])
         {
           col.factor <- c(col.factor,"#8FBC8F")
           org.factor <- c(org.factor,"Capsella grandiflora")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brara])
         {
           col.factor <- c(col.factor,"#4EEE94")
           org.factor <- c(org.factor,"Brassica rapa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.braol])
         {
           col.factor <- c(col.factor,"#BA55D3")
           org.factor <- c(org.factor,"Brassica oleracea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sisir])
         {
           col.factor <- c(col.factor,"#A4D3EE")
           org.factor <- c(org.factor,"Sisymbrium irio")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.schpa])
         {
           col.factor <- c(col.factor,"#8B2323")
           org.factor <- c(org.factor,"Schrenkiella parvula")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eutsa])
         {
           col.factor <- c(col.factor,"#8B2252")
           org.factor <- c(org.factor,"Eutrema salsugineum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.thlar])
         {
           col.factor <- c(col.factor,"#FF6A6A")
           org.factor <- c(org.factor,"Thlaspi arvense")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.boest])
         {
           col.factor <- c(col.factor,"#EE7942")
           org.factor <- c(org.factor,"Boechera stricta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.carpa])
         {
           col.factor <- c(col.factor,"#BBFFFF")
           org.factor <- c(org.factor,"Carica papaya")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.theca])
         {
           col.factor <- c(col.factor,"#00CD66")
           org.factor <- c(org.factor,"Theobroma cacao")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.goshi])
         {
           col.factor <- c(col.factor,"#FFE4C4")
           org.factor <- c(org.factor,"Gossypium hirsutum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gosra])
         {
           col.factor <- c(col.factor,"#8B8386")
           org.factor <- c(org.factor,"Gossypium raimondii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.citsi])
         {
           col.factor <- c(col.factor,"#CD4F39")
           org.factor <- c(org.factor,"Citrus sinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pontr])
         {
           col.factor <- c(col.factor,"#8B4726")
           org.factor <- c(org.factor,"Poncirus trifoliata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eucgr])
         {
           col.factor <- c(col.factor,"#FF6EB4")
           org.factor <- c(org.factor,"Eucalyptus grandis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.maldo])
         {
           col.factor <- c(col.factor,"#00008B")
           org.factor <- c(org.factor,"Malus domestica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.prupe])
         {
           col.factor <- c(col.factor,"#6495ED")
           org.factor <- c(org.factor,"Prunus persica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.frave])
         {
           col.factor <- c(col.factor,"#8DB6CD")
           org.factor <- c(org.factor,"Fragaria vesca")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corav])
         {
           col.factor <- c(col.factor,"#EE9572")
           org.factor <- c(org.factor,"Corylus avellana")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caril])
         {
           col.factor <- c(col.factor,"#FFEBCD")
           org.factor <- c(org.factor,"Carya illinoinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.queru])
         {
           col.factor <- c(col.factor,"#9ACD32")
           org.factor <- c(org.factor,"Quercus rubra")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cucsa])
         {
           col.factor <- c(col.factor,"#EE6AA7")
           org.factor <- c(org.factor,"Cucumis sativus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.glyma])
         {
           col.factor <- c(col.factor,"#D15FEE")
           org.factor <- c(org.factor,"Glycine max")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.medtr])
         {
           col.factor <- c(col.factor,"#6B8E23")
           org.factor <- c(org.factor,"Medicago truncatula")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tripr])
         {
           col.factor <- c(col.factor,"#8B1A1A")
           org.factor <- c(org.factor,"Trifolium pratense")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cicar])
         {
           col.factor <- c(col.factor,"#F0FFF0")
           org.factor <- c(org.factor,"Cicer arietinum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lotja])
         {
           col.factor <- c(col.factor,"#DEB887")
           org.factor <- c(org.factor,"Lotus japonicus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaac])
         {
           col.factor <- c(col.factor,"#FFA07A")
           org.factor <- c(org.factor,"Phaseolus acutifolius")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vigun])
         {
           col.factor <- c(col.factor,"#FFD700")
           org.factor <- c(org.factor,"Vigna unguiculata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lupal])
         {
           col.factor <- c(col.factor,"#8B668B")
           org.factor <- c(org.factor,"Lupinus albus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lencu])
         {
           col.factor <- c(col.factor,"#FF7F50")
           org.factor <- c(org.factor,"Lens culinaris")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.arahy])
         {
           col.factor <- c(col.factor,"#8B8B7A")
           org.factor <- c(org.factor,"Arachis hypogaea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.poptr])
         {
           col.factor <- c(col.factor,"#FAF0E6")
           org.factor <- c(org.factor,"Populus trichocarpa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.manes])
         {
           col.factor <- c(col.factor,"#EE7621")
           org.factor <- c(org.factor,"Manihot esculenta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.salpu])
         {
           col.factor <- c(col.factor,"#CDAA7D")
           org.factor <- c(org.factor,"Salix purpurea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.linus])
         {
           col.factor <- c(col.factor,"#EE799F")
           org.factor <- c(org.factor,"Linum usitatissimum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corci])
         {
           col.factor <- c(col.factor,"#551A8B")
           org.factor <- c(org.factor,"Corymbia citriodora")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vitvi])
         {
           col.factor <- c(col.factor,"#90EE90")
           org.factor <- c(org.factor,"Vitis vinifera")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kalfe])
         {
           col.factor <- c(col.factor,"#8B4C39")
           org.factor <- c(org.factor,"Kalanchoe fedtschenkoi")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vacda])
         {
           col.factor <- c(col.factor,"#CD5C5C")
           org.factor <- c(org.factor,"Vaccinium darrowii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oleeu])
         {
           col.factor <- c(col.factor,"#CD3278")
           org.factor <- c(org.factor,"Olea europaea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mimgu])
         {
           col.factor <- c(col.factor,"#B3EE3A")
           org.factor <- c(org.factor,"Mimulus guttatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.soltu])
         {
           col.factor <- c(col.factor,"#EE3B3B")
           org.factor <- c(org.factor,"Solanum tuberosum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cofar])
         {
           col.factor <- c(col.factor,"#FFA500")
           org.factor <- c(org.factor,"Coffea arabica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lacsa])
         {
           col.factor <- c(col.factor,"#66CD00")
           org.factor <- c(org.factor,"Lactuca sativa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dauca])
         {
           col.factor <- c(col.factor,"#FF4500")
           org.factor <- c(org.factor,"Daucus carota")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.betvu])
         {
           col.factor <- c(col.factor,"#00FF00")
           org.factor <- c(org.factor,"Beta vulgaris")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.amahy])
         {
           col.factor <- c(col.factor,"#EEC900")
           org.factor <- c(org.factor,"Amaranthus hypochondriacus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chequ])
         {
           col.factor <- c(col.factor,"#7B68EE")
           org.factor <- c(org.factor,"Chenopodium quinoa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sapof])
         {
           col.factor <- c(col.factor,"#FFA07A")
           org.factor <- c(org.factor,"Saponaria officinalis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aquco])
         {
           col.factor <- c(col.factor,"#DA70D6")
           org.factor <- c(org.factor,"Aquilegia coerulea")
         }
         
         
       }
       
       #Matrix with labels and colors and transform to dplyr format
       data.tree <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                               col = col.factor, org = org.factor)
       
       d2 <- dplyr::mutate(data.tree, lab = data.tree$label,
                           color = data.tree$col,
                           organism = data.tree$org,
                           name = glue("<i style='color:{color}'> {lab} </i>"))
       { 
         if (build_trees2() == "Maximum Likelihood")
         {
           tree_plot <- ggtree(tree_reduced) %<+% d2 + geom_tiplab() + theme(legend.position =) +
             xlim(0, max(tree_reduced$edge.length)*3) + geom_tiplab(aes(label = label, color = organism)) +
             scale_color_manual(values = unique(d2$col), breaks = unique(d2$org)) +
             geom_highlight(mapping=aes(subset = label %in% high.gene2,
                                        node = node,
                                        fill = as.factor(node)), extend = 0.8) + 
             labs(fill = "Node of interest")
           
         }
         else
         {
           tree_plot <- ggtree(tree_reduced) %<+% d2 + geom_tiplab() + theme(legend.position =) +
             xlim(0, max(tree_reduced$edge.length)*3) + geom_tiplab(aes(label = label, color = organism)) +
             geom_nodelab() +
             scale_color_manual(values = unique(d2$col), breaks = unique(d2$org)) +
             geom_highlight(mapping=aes(subset = label %in% high.gene2,
                                        node = node,
                                        fill = as.factor(node)), extend = 0.8) + 
             labs(fill = "Node of interest")
         }
       }
       shinyjs::hideElement(id = 'loading.tree2')
       return(tree_plot)
     }}) %>% bindEvent(input$run_button2)
   
   
   # Outputs
   observeEvent(isTruthy(tree_plot2()), {
     output$error_tree2 <- NULL
   })
   
   # Create boxes
   observeEvent(isTruthy(tree_plot2()), {
     
     if (UI_exist_tree2)
     {
       removeUI(
         selector = "div:has(>> #tree_seq_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #treeTips2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #presentorg2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadTree2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadNewick2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadTreeSeqs2)",
         multiple = TRUE,
         immediate = TRUE
       )
     }
     
     insertUI("#box_tree_seq_table2", "afterEnd", ui = {
       box(
         title = "Sequence Hits", status = "success", solidHeader = TRUE, width = 12,
         collapsible = TRUE,
         dataTableOutput(outputId = "tree_seq_table2")
         )
     }) 
     
     insertUI("#box_tree_text2", "afterEnd", ui = {
       box(
         title = "Genes in Orthogroup", status = "success", solidHeader = TRUE, width = 12,
         collapsible = TRUE,
         verbatimTextOutput("treeTips2")
       )
     }) 
     
     insertUI("#box_tree_pie2", "afterEnd", ui = {
       box(
         title = "Present Organisms", status = "success", solidHeader = TRUE,
         collapsible = TRUE, width = 12,
         plotlyOutput("presentorg2")
       )
     }) 
     
     insertUI("#box_tree_plot2", "afterEnd", ui = {
       image_height <- 300 + 15*length(tree_reduced2()$tip.label)
       box(width = 12,
           title = "Gene Tree", status = "success", solidHeader = TRUE,
           collapsible = TRUE, 
           plotOutput("tree_image2", height = image_height, width = 1100)
       )
     })
     
     insertUI("#download_tree2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTree2", 
                                                                          "Download Tree Plot",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#download_newick2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadNewick2", 
                                                                          "Download NEWICK Tree",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#download_tree_seqs2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTreeSeqs2", 
                                                                          "Download Protein Sequences",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_tree2 <<- TRUE
     shinyjs::hideElement(id = 'loading.tree2')
   })
   
   # Fill boxes with output
   # DIAMOND Table
   output$tree_seq_table2 <- renderDataTable({
     diamond_table <- diamond_table2()
     return(diamond_table)
   }, escape=FALSE, rownames= F, options = list(pageLength = 5))
   
   # Gene IDS
   output$treeTips2 <- renderPrint({
     print(tree_reduced2()$tip.label)
   }, width = 400) # %>% bindEvent(input$run_button2)
   
   # Render pie chart
   output$presentorg2 <- renderPlotly({
     
     {library(ggplot2)
       library(dplyr)
       
       data <- data.frame(table(organims_reduced2()))
       colnames(data) <- c("group", "value")
       
       # Compute the position of labels
       data <- data %>%
         arrange(desc(group)) %>%
         mutate(prop = value / sum(data$value) *100) %>%
         mutate(ypos = cumsum(prop)- 0.5*prop )
       
       # Create plot
       
       plotly::plot_ly(data=data,values=~prop,labels=~factor(group),
                       marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                              floor(nrow(data)/9)+1)),
                       type="pie",showlegend = F, text= ~group,
                       textinfo = "none", hoverinfo = "text")} 
     
   })
   
   # Render tree image
   output$tree_image2 <- renderImage({
     image_height <- 300 + 15*length(tree_reduced2()$tip.label)
     png("tree2.png", height = image_height, width = 1100)
     plot(tree_plot2())
     dev.off()
     
     list(src = "tree2.png",
          contentType="image/png", width=1100,height=image_height)
   }, deleteFile = T)
   
   # Download results
   output$downloadTree2 <- downloadHandler(
     filename= function() {
       paste("tree", ".png", sep="")
     },
     content= function(file) {
       image_height <- (300 + 11*length(tree_reduced2()$tip.label))*3
       image_width <- (200 + 400*max(tree_reduced2()$edge.length))*3
       png(file, height = image_height, width = image_width, res = (70 + 0.1*length(tree_reduced2()$tip.label))*3)
       plot(tree_plot2())
       dev.off()
     })
  
   
   # Download results
   output$downloadTree2 <- downloadHandler(
     filename= function() {
       paste("tree", ".png", sep="")
     },
     content= function(file) {
       image_height <- 300 + 11*length(tree_reduced2()$tip.label)
       image_width <- 200 + 400*max(tree_reduced2()$edge.length)
       png(file, height = image_height, width = image_width)
       plot(tree_plot2())
       dev.off()
     })
   
   # Create and download tree in newick format
   output$downloadNewick2 <- downloadHandler(
     filename= function() {
       paste("tree_newick", ".txt", sep="")
     },
     content= function(file) {
       write.tree(tree_reduced2(), file)
     })
   
   #  # Create and download sequences for genes in tree
   output$downloadTreeSeqs2 <- downloadHandler(
     filename= function() {
       paste("tree_seqs", ".fa", sep="")
     },
     content= function(file) {
       seqinr::write.fasta(sequences = seqinr::getSequence(ortho_reduced2()),
                           names = seqinr::getName(ortho_reduced2()), file.out = file)
     })

   
   ####################### PHYLOWIDGET ############################
   # Remove previous outputs when updated by a new search
   observeEvent(input$run_button2, {
     if (UI_exist_phylo2)
     {
       removeUI(
         selector = "div:has(>>> #phylo_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_phylo2 <<- F
     }
     
   })
   
   phylo_tree2 <- reactive({
     
     library(ape)
     tree_phylo <- tree_reduced2()
     
     # Normalize tree depth
     root_id <- length(tree_phylo$tip.label)+1
     norm_factor <- max(dist.nodes(tree_phylo)[root_id,])
     tree_phylo$edge.length <- tree_phylo$edge.length/norm_factor
     
     return(tree_phylo)
     
   }) %>% bindEvent(input$phylo_start2)
   
   observeEvent(isTruthy(phylo_tree2()),{
     
     if(UI_exist_phylo2)
     {
       removeUI(
         selector = "div:has(>>> #phylo_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_phylo2 <<- F
     }
     
     
     insertUI("#box_phylo2", "afterEnd", ui = {
       phylo_tree <- phylo_tree2()
       phylo_height <- length(phylo_tree$tip.label) *14 + 220
       box(width = 12,
           title = "Interactive Tree", status = "success", solidHeader = TRUE, height = phylo_height + 100,
           collapsible = TRUE,
           tags$div(id = "phylo_pocket2", style = paste0("width: 1300px; height: ",  phylo_height + 50, "px"),
                    phylowidgetOutput("phylo_plot2", height = paste0(phylo_height,"px"), width = "98%"))
       )
     })
     
     UI_exist_phylo2 <<- T
     
   })
   
   output$phylo_plot2 <- renderPhylowidget({
     
     phylo_tree <- phylo_tree2()
     phylowidget(phylo_tree)
   })
   
   #########################  PFAM  ###############################
   
   observeEvent(input$run_button2, {
     removeUI(
       selector = "div:has(>> #selected_pfamsI2)",
       multiple = TRUE,
       immediate = TRUE
     )
   })
   
   observeEvent(input$pfam_start2, {
     insertUI("#selected_pfams2", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_pfamsI2","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced2()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({diamond_table2()$ID[1]}))
     })
   })
   
   observeEvent(input$run_button2, {
     removeUI("#pfam_selection2")
   })
   
   observeEvent(input$pfam_start2, {
     insertUI("#pfam_selectionI2", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("pfam_selection2", "Show Pfam Domains", size = "sm",
                                style = "float", color = "success")
     })
   })
   
   
   observeEvent(input$run_button2, {
     if (UI_exist_pfam2)
     {
       removeUI(
         selector = "div:has(>> #output_pfam_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pfam2 <<- F
     }
   })
   
   
   total_table_pfam2 <- reactive({
     shinyjs::showElement(id = 'loading.pfam.pf2')
     ortho_reduced <- ortho_reduced2()
     sel_genes <- as.vector(input$selected_pfamsI2)
     
     if (length(sel_genes) < 1)
     {
       shinyjs::hideElement(id = 'loading.pfam.pf2')
       removeUI(
         selector = "div:has(>> #output_pfam_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       UI_exist_pfam2 <<- F
       output$error_pfam2 <- renderUI({renderText({print("Please select at least one gene.")})})
       validate(need(length(sel_genes) > 0, "Please select at least one gene."))
     }
     
     output$error_pfam2 <- NULL
     #library(bio3d)
     library(RCurl)
     library(drawProteins)
     library(ggplot2)
     library(jsonlite)
     
     # Get the sequences as a vector of strings
     
     
     # Create data frame with proper columns
     total_table_pfam <- data.frame(type=NA,
                                    description=NA,
                                    begin=NA, end=NA,
                                    length=NA,
                                    accession=NA, entryName=NA,
                                    taxid=NA, order=NA)
     
     
     # Fill data frame with the information about domains obtained with hmmer
     for (i in 1:length(sel_genes))
     {
       ortho_comp <- ortho_reduced[[sel_genes[i]]]
       ortho_str <- seqinr::getSequence(ortho_comp, as.string = T)
       ortho_cha <- unlist(ortho_str)
       
       
       # Curl POST request
       url <- "https://www.ebi.ac.uk/Tools/hmmer/api/v1/search/hmmscan"
       
       data <- list(
         database = "pfam",
         input = ortho_cha
       )
       
       json_data <- toJSON(data, auto_unbox = T)
       
       response <- postForm(url, 
                            .opts = list(
                              postfields = json_data,
                              httpheader = c(
                                "Content-Type" = "application/json",
                                "Accept" = "application/json"
                              )
                            ))
       
       
       new_id <- strsplit(as.character(response), '"')[[1]][4]
       
       
       # Results URL
       url_res <- paste0("https://www.ebi.ac.uk/Tools/hmmer/api/v1/result/", new_id)
       
       # Iterate until job is finished, for that, check if SUCCESS is present in GET result
       nap_again <- T
       while(nap_again)
       {
         get_response <- getURL(url_res,
                                httpheader = c(
                                  "Accept" = "application/json"
                                ))
         
         if (grepl("SUCCESS", get_response))
         {
           nap_again <- F
           
           # Once computation is over, parse JSON response
           json_resp <- jsonlite::fromJSON(url_res, simplifyDataFrame = TRUE)
           json_hits <- data.frame(json_resp$result$hits)
           
           hits_acc <- json_hits$acc
           hits_desc <- json_hits$metadata$description
           hits_domains <- json_hits$domains
           
           df_acc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_acc, y = hits_domains, USE.NAMES = F))
           df_desc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_desc, y = hits_domains, USE.NAMES = F))
           df_begin <- unlist(sapply(hits_domains, function(x) x$iali))
           df_end <- unlist(sapply(hits_domains, function(x) x$jali))
           
           # Create PFAM table
           pfam_table <- data.frame(type=c("CHAIN", rep("DOMAIN", length(df_acc))),
                                    description=c("Protein chain",df_acc),
                                    begin=c(1, df_begin), end=c(nchar(ortho_cha),df_end),
                                    length=c(nchar(ortho_cha)-1, df_end - df_begin),
                                    accession=sel_genes[i], entryName=sel_genes[i],
                                    taxid=c("Chain", df_desc), order=i)
           
           total_table_pfam <- rbind(total_table_pfam, pfam_table)
           
         }
         else if (grepl("PENDING", get_response))
         {
           Sys.sleep(10)
         }
         # If there is a problem with server
         else
         {
           nap_again <- F
           pfam_table <- data.frame(type="CHAIN",
                                    description="Protein chain",
                                    begin=1, end=nchar(ortho_cha),
                                    length=nchar(ortho_cha)-1,
                                    accession=sel_genes[i], entryName=sel_genes[i],
                                    taxid="Chain", order=i)
           total_table_pfam <- rbind(total_table_pfam, pfam_table)
         }
       }
     }
     
     total_table_pfam <- total_table_pfam[-1,]
     total_table_pfam <- total_table_pfam[!duplicated(total_table_pfam),]
     # Remove protein chain results
     total_table_pfam <- subset(total_table_pfam, !(type=="DOMAIN" & description=="Protein chain"))
     
     return(total_table_pfam)
     
   }) %>% bindEvent(input$pfam_selection2)
   
   pfplot2 <- reactive({
     
     total_table_pfam <- total_table_pfam2()
     # Now we can plot domains information as chains
     pfplot <- draw_canvas(total_table_pfam)
     pfplot <- draw_chains(pfplot, total_table_pfam)
     pfplot <- draw_domains(pfplot, total_table_pfam, label_domains = F)
     pfplot <- pfplot + theme_bw(base_size = 20) + # white background
       theme(panel.grid.minor=element_blank(),
             panel.grid.major=element_blank()) +
       theme(axis.ticks = element_blank(),
             axis.text.y = element_blank()) +
       theme(panel.border = element_blank())
     #pfplot <- pfplot + labs(title = "Pfam domains")
     pfplot <- pfplot + theme(legend.position="top") + labs(fill="")
     
   }) %>% bindEvent(input$pfam_selection2)
   
   
   # Outputs
   
   observeEvent(isTruthy(pfplot2()), {
     
     if (UI_exist_pfam2)
     {
       removeUI(
         selector = "div:has(>> #output_pfam_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_pfam2", "afterEnd", ui = {
       
       box(
         title = "PFAM Table", status = "success", solidHeader = TRUE, width = 12,
         collapsible = TRUE,
         dataTableOutput(outputId = "output_pfam_table2"))
     }) 
     
     insertUI("#box_pfplot2", "afterEnd", ui = {
       total_table_pfam <- total_table_pfam2()
       box_pfplot_height <- 150 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
       box(
         title = "Domains Localization", status = "success", solidHeader = TRUE, width = 12, #height = box_pfplot_height,
         collapsible = TRUE,
         plotOutput("pfam_plot2", height = box_pfplot_height))
     }) 
     
     insertUI("#pfam_down_button2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "pfam_download2", "Download PFAM figure",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#download_ui_for_pfam_table2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadPFAMTable2", "Download PFAM Table",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_pfam2 <<- TRUE
     shinyjs::hideElement(id = 'loading.pfam.pf2')
   })
   
   output$output_pfam_table2 <- renderDataTable({
     total_table_pfam <- total_table_pfam2()
     out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
     out_pf_table$description <- sapply(out_pf_table$description, function(x) pfam.link(x))
     colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
     return(out_pf_table)
   }, escape=FALSE, options = list(pageLength = 5))
   
   output$pfam_plot2 <- renderImage({
     total_table_pfam <- total_table_pfam2()
     pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
     pfam_width <- 1000
     pfplot <- pfplot2()
     png("pharaoh_folder/pfam.png",  width = pfam_width, height = pfam_height)
     plot(pfplot)
     dev.off()
     list(src = "pharaoh_folder/pfam.png",
          contentType="image/png")
   }, deleteFile = T
   )
   
   # Download results
   
   output$pfam_download2 <- downloadHandler(
     filename= function() {
       paste("pfam", ".png", sep="")
     },
     content= function(file) {
       total_table_pfam <- total_table_pfam2()
       pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
       pfam_width <- 1150
       pfplot <- pfplot2()
       
       png(file, height = pfam_height, width = pfam_width)
       plot(pfplot)
       dev.off()
     })
   
   output$downloadPFAMTable2<- downloadHandler(
     filename= function() {
       paste("pfam_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_pfam <- total_table_pfam2()
       out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
       colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
       write.table(x = out_pf_table, quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
  
   ####################### CAFE #################################
   
   # Remove previous outputs when updated by a new search
   observeEvent(input$run_button2, {
     if (UI_exist_cafe2)
     {
       removeUI(
         selector = "div:has(>> #cafe_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_mrca2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCAFEPlot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_cafe2 <<- F
     }
     
     if (UI_exist_error_cafe2)
     {
       removeUI(
         selector = "div:has(>> #cafe_error_message2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_error_cafe2 <<- F
     }
   })
   
   ### CAFE parser and tree generator
   cafe_table2 <- reactive({
     
     shinyjs::showElement(id = 'loading.cafe2')
     
     # Import OG name
     og.cafe <- og.name2()
     
     # Define path to COUNT file
     cafe_comp_tree_file <- "pharaoh_folder/family_table_count.tsv"
     
     # Extract COUNT vector for the current OG
     library(data.table)
     library(ape)
     genecount <- fread(cafe_comp_tree_file, sep="\t", header = T)
     genecountdf <- as.data.frame(genecount)
     rownames(genecountdf) <- genecountdf$Orthogroup
     new_cols <- gsub(" ", "_", colnames(genecountdf))
     colnames(genecountdf) <- new_cols
     
     cafe.vector <- genecountdf[og.cafe,]
     
     if (sum(is.na(cafe.vector)) != 0)
     {
       shinyjs::hideElement(id = 'loading.cafe2')
       if (UI_exist_cafe2)
       {
         removeUI(
           selector = "div:has(>> #cafe_plot2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #cafe_mrca2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #cafe_download2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadCAFEPlot2)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_cafe2 <<- F
       
       if(UI_exist_error_cafe2)
       {
         removeUI(
           selector = "div:has(>> #cafe_error_message2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
       }
       insertUI("#error_cafe2", "afterEnd", ui = {
         box(width = 12,
             title = "Ancestral State Reconstruction", status = "info", solidHeader = TRUE,
             collapsible = TRUE,
             textOutput("cafe_error_message2"))
       })
       
       output$cafe_error_message2 <- renderText({print("No expansions/contraction detected for this orthogroup,
                                                        or infeasible computation due to large size and variance across
                                                        species.")})
       UI_exist_error_cafe2 <<- T
       
       validate(need(sum(is.na(cafe.vector)) == 0 , ""))
     }
     
     return(genecountdf)
   }) %>% bindEvent(input$cafe_start2)
   
   mrca.tree2 <- reactive({
     
     og.cafe <- og.name2()
     genecountdf <- cafe_table2()
     
     # Create phylogenomic tree with internal nodes names
     mrca.tree <- read.tree("pharaoh_folder/tree_with_internal_nodes.txt")
     
     return(mrca.tree)
     
   }) %>% bindEvent(input$cafe_start2)
   
   cafe_tree2 <- reactive({
     og.cafe <- og.name2()
     genecountdf <- cafe_table2()
     cafe.tree <- mrca.tree2()
     
     # Add counts to nodes
     tip.count <- as.numeric(genecountdf[og.cafe,2:172])
     names(tip.count) <- colnames(genecountdf)[2:172]
     
     node.count <- as.numeric(genecountdf[og.cafe,173:342])
     names(node.count) <- colnames(genecountdf)[173:342]
     
     cafe.tree$tip.label[match(names(tip.count), cafe.tree$tip.label)] <- tip.count
     cafe.tree$node.label[match(names(node.count), cafe.tree$node.label)] <- node.count
     
     return(cafe.tree)
     
   }) %>% bindEvent(input$cafe_start2)
   
   evo_plot_data2 <- reactive({
     
     og.cafe <- og.name2()
     genecountdf <- cafe_table2()
     cafe.tree <- mrca.tree2()
     
     # Add counts to nodes
     tip.count <- as.numeric(genecountdf[og.cafe,2:172])
     names(tip.count) <- colnames(genecountdf)[2:172]
     
     node.count <- as.numeric(genecountdf[og.cafe,173:342])
     names(node.count) <- colnames(genecountdf)[173:342]
     
     # Identify parental node to determine if a general expansion or contraction 
     # occured at the level of each node
     edge_table <- as.data.frame(cafe.tree$edge)
     rownames(edge_table) <- paste("edge", 1:nrow(edge_table), sep = "")
     colnames(edge_table) <- c("parent", "child")
     
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) <
          as.numeric(tip.count[x])) "Expansion"
       else if (as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) >
                as.numeric(tip.count[x])) "Contraction"
       else "No change"
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x) # 170 is root, no parent
       if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) <
          as.numeric(node.count[x])) "Expansion"
       else if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) >
               as.numeric(node.count[x])) "Contraction"
       else "No change"
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # A gain occurs when a node experiences an expansion and its parental exhibits a value of 0
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if (exp_cont_tip[x] == "Expansion" &
           as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) == 0
       ) "Gain"
       else exp_cont_tip[x]
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x)
       if (exp_cont_node[match(x,names(exp_cont_node))] == "Expansion" &
           as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) == 0
       ) "Gain"
       else exp_cont_node[x]
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # A loss occurs when a node experiences a contraction and takes the value 0
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if (exp_cont_tip[x] == "Contraction" &
           tip.count[x] == 0
       ) "Loss"
       else exp_cont_tip[x]
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x)
       if (exp_cont_node[match(x,names(exp_cont_node))] == "Contraction" &
           node.count[x] == 0
       ) "Loss"
       else exp_cont_node[x]
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # Create complete sorted vectors
     exp_cont_node["root"] <- "No change"
     node_ordered <- exp_cont_node[order(match(names(exp_cont_node),cafe.tree$node.label))]
     tip_ordered <- exp_cont_tip[order(match(names(exp_cont_tip),cafe.tree$tip.label))]
     
     change_vector <- c(tip_ordered, node_ordered)
     
     # Merge tips and nodes reconstruction
     node_ordered_values <- node.count[order(match(names(node.count),cafe.tree$node.label))]
     tip_ordered_values <- tip.count[order(match(names(tip.count),cafe.tree$tip.label))]
     cafe.count <- c(tip_ordered_values, node_ordered_values)
     
     # Create a timeline for a given OG
     tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
     tree.ancestor <- read.tree(tree.name)
     tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
     tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
     tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
     
     mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
     evo.paths <- c()
     for (i in 1:length(unique(tips.orgs)))
     {
       evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
     }
     
     evo.paths <- unique(evo.paths)
     evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
     
     # Associate gray and 0 to reconstruction for nodes not in allowed paths
     change_vector[setdiff(1:length(change_vector), evo.paths)] <- 
       sapply(change_vector[setdiff(1:length(change_vector), evo.paths)], function(x) if(x != "Loss") "OG not present" 
              else x)  
     
     change_vector <- unlist(change_vector)
     
     cafe.count[setdiff(1:length(change_vector), evo.paths)] <- 0
     
     unique(change_vector)
     color_cafe <- sapply(change_vector, function(x) if (x == "No change") "black" 
                          else if (x == "Expansion") "red" else if (x == "Contraction") "blue"
                          else if (x == "Gain") "green3" else if (x == "Loss") "purple"
                          else "gray", USE.NAMES = F)
     
     # Create tree representation
     cafe.table.tips <- data.frame(node = 1:length(cafe.tree$tip.label), label = cafe.tree$tip.label,
                                   col = color_cafe[1:length(cafe.tree$tip.label)], reconst = change_vector[1:length(cafe.tree$tip.label)],
                                   dup_number = cafe.count[1:length(cafe.tree$tip.label)])
     
     cafe.table.nodes <- data.frame(node = (Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label)), label = cafe.tree$node.label,
                                    col = color_cafe[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                    reconst = change_vector[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                    dup_number = cafe.count[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))])
     
     cafe.table.node.comp <- rbind(cafe.table.tips, cafe.table.nodes)
     
     d <- dplyr::mutate(cafe.table.node.comp)
     d$text <- d$label
     
     return(d)
     
   }) %>% bindEvent(input$cafe_start2)
   
   evo_plot2 <- reactive({
     
     d <- evo_plot_data2() 
     mrca.tree <- mrca.tree2()
     
     library(ggtree)
     library(ggplot2)
     
     evo_plot <- ggtree(mrca.tree) %<+% d + aes(colour = I(d$col)) +
       geom_tiplab(aes(label=gsub("_", " ", tools::toTitleCase(d$label))), offset = 30) +
       theme(legend.position = "none") +
       xlim(0, max(mrca.tree$edge.length)*1.85) +
       geom_nodepoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                      alpha = .75) +
       scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
       geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                     alpha = .75)
     
     return(evo_plot)
     
   }) %>% bindEvent(input$cafe_start2)
   
   evo_plotly2 <- reactive({
     
     d <- evo_plot_data2() 
     cafe.tree <- mrca.tree2()
     
     library(ggtree)
     library(ggplot2)
     
     # Remove internal nodes from direct visualization, retaining names in panels
     d$label[172:341] <- NA
     
     evo_plotly <- ggtree(cafe.tree) %<+% d + aes(colour = I(d$col),text=paste0("</br> Duplications: ",dup_number,
                                                                                "</br> Name: ",gsub("_", " ", tools::toTitleCase(d$text)))) + 
       geom_text(aes(x =  1970, label=gsub("_", " ", tools::toTitleCase(d$label)))) + 
       theme(legend.position = "none") +
       xlim(0, max(cafe.tree$edge.length)*1.25) +
       geom_point(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                  alpha = .75) +
       scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
       geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                     alpha = .75)
     
     p <- ggplotly(evo_plotly, tooltip = "text", width = 1400, height = 2200) 
     p <- p %>%
       plotly::style(textposition = "right",xanchor="right")
     
     return(p)
     
   }) %>% bindEvent(input$cafe_start2)
   
   evo.paths.id2 <- reactive({
     
     # Create phylogenomic tree with internal nodes names
     og.cafe <- og.name2()
     cafe.tree <- mrca.tree2()
     
     # Create timeline
     tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
     tree.ancestor <- read.tree(tree.name)
     tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
     tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
     tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
     
     mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
     evo.paths <- c()
     for (i in 1:length(unique(tips.orgs)))
     {
       evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
     }
     
     evo.paths <- unique(evo.paths)
     evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
     
     #print(paste0("Most recent common ancestor for this orthogroup is ", evo.paths.id[1]))
     
     return(evo.paths.id)
     
   }) %>% bindEvent(input$cafe_start2)
   
   # Outputs
   
   # Remove previous boxes if they exist and create new ones
   observeEvent(isTruthy(evo_plot2()), {
     
     if (UI_exist_cafe2)
     {
       removeUI(
         selector = "div:has(>> #cafe_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_mrca2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCAFEPlot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     if (UI_exist_error_cafe2)
     {
       removeUI(
         selector = "div:has(>> #cafe_error_message2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_cafe2", "afterEnd", ui = {
       box(width = 12,
           title = "Ancestral State Reconstruction", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           plotlyOutput("cafe_plot2", height = ifelse(model.selected2(), "2300px", "2300px")))
     })
     
     insertUI("#box_mrca2", "afterEnd", ui = {
       box(width = 8,
           title = "Most Recent Common Ancestor", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           textOutput("cafe_mrca2")
       )
     })
     
     insertUI("#cafe_down_button2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "cafe_download2", "Download NEWICK tree",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#download_ui_for_cafe_plot2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadCAFEPlot2", "Download Ancestral State Plot",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_cafe2 <<- TRUE
     shinyjs::hideElement(id = 'loading.cafe2')
   })
   
   # Fill outputs
   output$cafe_plot2 <- renderPlotly({
     evo_plotly2()
   })
   
   output$cafe_mrca2 <- renderText({
     print(paste0("Most recent common ancestor for this orthogroup is the
                   ancestor of the clade: ", evo.paths.id2()[1]))
   })
   
   # Download tab's results
   
   output$cafe_download2 <- downloadHandler(
     filename= function() {
       paste("ancestral_newick", ".txt", sep="")
     },
     content= function(file) {
       cafe_tree <- cafe_tree2()
       
       write.tree(cafe_tree, file)
     })
   
   output$downloadCAFEPlot2<- downloadHandler(
     filename= function() {
       paste("ancestral_plot", ".png", sep="")
     },
     content= function(file) {
       evo_plot <- evo_plot2()
       
       png(file, width = 1400, height = 2800, res = 100)
       plot(evo_plot)
       dev.off()
     })
   

   ####################### MSA #################################
   
   observeEvent(input$run_button2, {
     removeUI(
       selector = "div:has(>> #selected_msaI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(>> #msa_methodI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#msa_selection2")
     
     if (UI_exist_msa2)
     {
       removeUI(
         selector = "div:has(>>> #msa_print2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_fa2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_msa2 <<- F
     }
     
   })
   
   observeEvent(input$msa_start2, {
     insertUI("#selected_msa2", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_msaI2","Select the desired genes from the tree to align",
                                 choices=isolate({tree_reduced2()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({diamond_table2()$ID[1]}))
       
     })
     
     insertUI("#msa_method2", "afterEnd", ui = {
       shinyWidgets::pickerInput(inputId = "msa_methodI2", label = "Choose alignment method", 
                                 choices = c("ClustalOmega", "MAFFT"), selected = "ClustalOmega")
       
     })
     
     insertUI("#msa_selectionI2", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("msa_selection2", "Align Sequences", size = "sm",
                                style = "float", color = "success")
     })
     
   })
   
   alignseqs2 <- reactive({
     
     library(msa)
     shinyjs::showElement(id = 'loading.msa2')
     
     selected_genes <- as.vector(input$selected_msaI2)
     selected_method <- as.character(input$msa_methodI2)
     file.name <- og.name2()
     
     if (length(selected_genes) < 2)
     {
       shinyjs::hideElement(id = 'loading.msa2')
       
       if (UI_exist_msa2)
       {
       removeUI(
         selector = "div:has(>>> #msa_print2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_fa2)",
         multiple = TRUE,
         immediate = TRUE
       )
       }
       UI_exist_msa2 <<- F
       output$error_msa2 <- renderUI({renderText({print("Please select at least two genes.")})})
       validate(need(length(selected_genes) > 1, "Please select at least two genes."))
     }
     
     output$error_msa2 <- NULL
     
     # If de novo alignment is selected
     {
       if(selected_method == "ClustalOmega")
       {
         # Define path to orthogroup sequences file
         ortho.seq.name <- paste("Orthogroup_Sequences", paste(file.name, "fa", sep = "."), sep="/")
         
         
         # Read orthogroup sequences file and select the genes for alignment
         mySequences1 <- Biostrings::readAAStringSet(ortho.seq.name)
         mysubseqs <- mySequences1[selected_genes]
         
         alignseqs <- msa(mysubseqs, verbose = F, method = "ClustalOmega")
       }
       
       # If MAFFT alignment is selected
       else
       {
         ortho.seq.name <- paste("MultipleSequenceAlignments", paste(file.name, "fa", sep = "."), sep="/")
         mySequences1 <- seqinr::read.fasta(ortho.seq.name, seqtype = "AA")
         mysubseqs <- mySequences1[selected_genes]
         mysubnames <- seqinr::getName(mySequences1)
         
         # Identify indexes associated with reduced names
         indexes_msa <- sapply(selected_genes, function(x) grep(mysubnames, pattern = x))
         
         # Retrieve those sequences from alignment keeping gaps
         mysubseqs <- mySequences1[indexes_msa]
         names(mysubseqs) <- names(indexes_msa)
         
         # Remove columns with gaps and remove empty spaces in last positions
         seqs_mysubseqs <- seqinr::getSequence(mysubseqs)
         last <- seqs_mysubseqs[[length(seqs_mysubseqs)]]
         last <- last[which(last != " ")]
         seqs_mysubseqs[[length(seqs_mysubseqs)]] <- last
         seqs_mysubseqs <- remove_gaps(seqs_mysubseqs)
         names(seqs_mysubseqs) <- names(mysubseqs)
         
         mysubseqs2 <- unlist(lapply(seqs_mysubseqs, function(x) paste(x, collapse="")))
         
         alignseqs <- Biostrings::AAMultipleAlignment(mysubseqs2, use.names = T)
         
       }
     }
     
     detach("package:msa", unload=TRUE)
     
     return(alignseqs)
     
   }) %>% bindEvent(input$msa_selection2)
   
   # Create boxes for outputs
   observeEvent(isTruthy(alignseqs2()), {
     
     if (UI_exist_msa2)
     {
       removeUI(
         selector = "div:has(>>> #msa_print2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_fa2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_msa2", "afterEnd", ui = {
       selected_msa <- isolate({input$selected_msaI2})
       msa_height <- ifelse(length(selected_msa) > 14, 550, 400 + 5*length(selected_msa))
       box(width = 12,
           title = "MSA Explorer", status = "success", solidHeader = TRUE, height = msa_height,
           collapsible = TRUE,
           tags$div(id = "msa_pocket2", style = "width: 1300px; height: 400px",
                    msaROutput("msa_print2"))
           
       )
     })
     
     insertUI("#msa_down_plot2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_plot2", "Download Colored MSA",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#msa_down_fasta2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_fa2", "Download MSA FASTA",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_msa2 <<- TRUE
   })
   
   
   # Fill Output
   output$msa_print2 <- renderMsaR({
     alignseqs <- alignseqs2()
     msaout <- msa::msaConvert(alignseqs, "ape::AAbin")
     msaR(msaout, menu=T, overviewbox = F,  colorscheme = "clustal")
   })
   
   # Prepare variables for pdf construction
   observeEvent(isTruthy(alignseqs2()), {
     alignseqs <- alignseqs2()
     
     library(ggmsa)
     class(alignseqs) <- "AAMultipleAlignment"
     
     for(i in 1:(ncol(alignseqs)%/%100 +1)){
       assign(paste("msapseq", i, sep = ""), ggmsa(alignseqs, 1+(100*(i-1)), i*100, seq_name = TRUE, char_width = 0.5) +
                geom_seqlogo(color = "Chemistry_AA"), envir = as.environment(1), pos=1)
     }
     shinyjs::hideElement(id = 'loading.msa2')
   })
   
   # Download tab's results
   # Download colored MSA in pdf
   output$msa_download_plot2 <- downloadHandler(
     filename= function() {
       paste("msa", ".pdf", sep="")
     },
     content= function(file) {
       selected_msa <- input$selected_msaI2
       alignseqs <- alignseqs2()
       pdf(file, height = 2+length(selected_msa)*0.25, width = 16)
       {
         for(i in 1:(ncol(alignseqs)%/%100 +1)){
           print(mget(paste0("msapseq", i), envir = as.environment(1)))
         }
         dev.off()
       }
     })
   
   # Download MSA in FASTA format
   output$msa_download_fa2<- downloadHandler(
     filename= function() {
       paste("msa", ".fa", sep="")
     },
     content= function(file) {
       alignseqs <- alignseqs2()
       writeXStringSet(as(unmasked(alignseqs), "XStringSet"), file)
     })
   
   ####################### GO #################################
   
   observeEvent(input$run_button2, {
     removeUI(
       selector = "div:has(>> #selected_gosI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(>> #selected_gos_modeI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#gos_selectionI2")
     
     if (UI_exist_go2)
     {
       removeUI(
         selector = "div:has(>> #output_gos_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_treeplot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadGOSTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_gos_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_go2 <<- F
     }
     
   })
   
   observeEvent(input$go_start2, {
     insertUI("#selected_gos2", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_gosI2","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced2()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({diamond_table2()$ID[1]}))
       
     })
     
     insertUI("#selected_gos_mode2", "afterEnd", ui = {
       
       selectInput(inputId = "selected_gos_modeI2",
                   choices=c("Biological Processes" = "bp",
                             "Molecular Functions" = "mf",
                             "Cellular Components" = "cc"),
                   label = "Select the gene ontology to use",
                   multiple = F, selected = c("bp"))
       
     })
     
     insertUI("#gos_selection2", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("gos_selectionI2", "Show GO terms", size = "sm",
                                style = "float", color = "success")
     })
     
   })
   
   total_table_gos2 <- reactive({
     
     shinyjs::showElement(id = 'loading.go2')
     gos_anot <- read.csv("pharaoh_folder/final_anot_table.tsv", sep="\t", header = T)
     sel.genes.go <- as.vector(input$selected_gosI2)
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI2}))
     
     total_table_gos <- subset(gos_anot, gos_anot$name %in% sel.genes.go)
     
     # Show an error if no terms are identified in the input
     if (nrow(total_table_gos) == 0) 
     {
       shinyjs::hideElement(id = 'loading.go2')
       if (UI_exist_go2)
       {
         removeUI(
           selector = "div:has(>> #output_gos_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_plot2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_treeplot2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadGOSTable2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_download2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_gos_download2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_go2 <<- F
       }
       output$error_gos2 <- renderUI({
         renderPrint({cat("0 GO terms identified.")})
       })
       
       validate(need(nrow(total_table_gos) != 0, " "))
     }
     
     
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     terms_sel <- paste("terms", selected_gos_mode, sep="_")
     total_table_gos <- total_table_gos[,c("organism", "id", "name", gos_sel, terms_sel)]
     
     # Once removed the two GO categories not selected, remove rows with blank cells
     total_table_gos_clean <- total_table_gos[ total_table_gos[[gos_sel]] != "" , ]
     
     # Show an error if no terms are identified after the previous operation
     if (nrow(total_table_gos_clean) == 0) 
     {
       shinyjs::hideElement(id = 'loading.go2')
       if (UI_exist_go2)
       {
         removeUI(
           selector = "div:has(>> #output_gos_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_plot2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_treeplot2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadGOSTable2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_download2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_gos_download2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_go2 <<- F
       }
       
       output$error_gos2 <- renderUI({
         renderPrint({cat("0 GO terms identified.")})
       })
       
       validate(need(nrow(total_table_gos_clean) != 0, " "))
     }
     
     output$error_gos2 <- NULL
     
     return(total_table_gos_clean)
     
   }) %>% bindEvent(input$gos_selectionI2)
   
   enr_table2 <- reactive({
     
     total_table_gos <- total_table_gos2()
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI2}))
     
     # Create the plot
     
     # Create a list of GO terms vector of each gene
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     gos_list <- apply(total_table_gos,MARGIN=1,FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
     names(gos_list) <- total_table_gos$name
     
     # Count GOs for each gene and create a matrix of gene-GO pairs
     count_gos_in_genes <- sapply(gos_list, FUN = function(x) length(x))
     comp_data <- data.frame(gene = rep(names(count_gos_in_genes), count_gos_in_genes), gos = as.character(unlist(gos_list)))
     
     # Collapse genes that share a same GO
     gene.v <- c()
     for (i in 1:length(unique(comp_data$gos)))
     {
       new.table <- subset(comp_data, gos == unique(comp_data$gos)[i])
       new.cha <- as.character(new.table$gene)
       gene.v <- c(gene.v, paste(new.cha, collapse = "/"))
     }
     
     names(gene.v) <- unique(comp_data$gos)
     
     # Load libraries and create gene chains, count and GO IDs fields (same order)
     library(GO.db)
     library("multienrichjam")
     library(clusterProfiler)
     library(enrichplot)
     library(ggplot2)
     
     count_go <- table(comp_data$gos)
     geneids <- gene.v[names(count_go)]
     count_terms <- mapply(function(x) {Term(x)}, names(count_go), USE.NAMES = F)
     
     # If there  exists any obsolote term
     count_terms <- count_terms[which(!is.na(count_terms))] 
     count_go <- count_go[which(!is.na(count_terms))]
     geneids <- geneids[which(!is.na(count_terms))]
     
     # Create pseudo-enrichment table
     enr_table <- data.frame(ID=names(count_go), Description=count_terms, GeneRatio="90/100", BgRatio="90/10000", 
                             pvalue=0.000005, p.adjust=0.000005, qvalue=0.000005, geneID=geneids, Count = as.vector(count_go))
     
     return(enr_table)
     
   }) %>% bindEvent(input$gos_selectionI2)
   
   ema_gos_plot2 <- reactive({
     
     enr_table <- enr_table2()
     
     # Transform to enrichResult object
     enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                  geneColname = "geneID", pvalueColname = "p.adjust",
                                  descriptionColname = "Description", pvalueCutoff = 0.05)
     
     # Create plot
     ema_gos_plot <- emapplot(pairwise_termsim(enr), showCategory = 15) + theme(legend.position='none')
     return(ema_gos_plot)
   })
   
   tree_gos_plot2 <- reactive({
     
     enr_table <- enr_table2()
     
     # Transform to enrichResult object
     enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                  geneColname = "geneID", pvalueColname = "p.adjust",
                                  descriptionColname = "Description", pvalueCutoff = 0.05)
     {
       if (nrow(enr_table) > 4)
       {
         tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(label_words_n = 3)) +
           theme(legend.position='none')
       }
       else if (nrow(enr_table) > 2)
       {
         tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(n = 2, label_words_n = 3)) + 
           theme(legend.position='none')
       }
       else
       {
         text <- paste("\n  Unable to create treeplot with less than 3 GO terms \n")
         tree_gos_plot <- ggplot() + 
           annotate("text", x = 4, y = 25, size=8, label = text) + 
           theme_void()
       }
     }
     
     shinyjs::hideElement(id = 'loading.go2')
     return(tree_gos_plot)
   })
   
   # Create boxes for outputs
   observeEvent(isTruthy(tree_gos_plot2()), {
     
     if (UI_exist_go2)
     {
       removeUI(
         selector = "div:has(>> #output_gos_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_treeplot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadGOSTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_gos_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_gos_table2", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Table", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_gos_table2")
       )
     })
     
     insertUI("#box_gos_plot2", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Plot", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           plotOutput("gos_plot2", height = 610)
       )
     })
     
     insertUI("#box_gos_treeplot2", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Treeplot", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           plotOutput("gos_treeplot2", height = 610)
       )
     })
     
     insertUI("#download_ui_for_gos_table2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadGOSTable2", "Download GO Table",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#gos_down_button2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "gos_download2", "Download GO Plot",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#tree_gos_down_button2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "tree_gos_download2", "Download GO Treeplot",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_go2 <<- TRUE
   })
   
   # Fill outputs
   # Table
   output$output_gos_table2 <- renderDataTable({
     total_table_gos <- total_table_gos2()
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI2}))
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     gos_list <- apply(total_table_gos,MARGIN=1,
                       FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
     {
       if(!is.list(gos_list))
       {
         gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
         gos_formatted <- paste0(gos_links, collapse = " | ")
         total_table_gos[,gos_sel] <- gos_formatted
         total_table_gos 
       }
       else
       {
         gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
         gos_formatted <- unlist(sapply(gos_links, function(x) paste0(x, collapse = " | ")))
         total_table_gos[,gos_sel] <- gos_formatted
         total_table_gos  
       }
     }
     
   },escape=FALSE, rownames=F, options =list(pageLength = 5))
   
   # First plot
   output$gos_plot2 <- renderImage({
     ema_gos_plot <- ema_gos_plot2()
     
     png("pharaoh_folder/gosplot.png", width = 590, height = 590, res = 90)
     plot(ema_gos_plot)
     dev.off()
     list(src = "pharaoh_folder/gosplot.png",
          contentType="image/png")
     
   }, deleteFile=T
   )
   
   # Second plot
   output$gos_treeplot2 <- renderImage({
     tree_gos_plot <- tree_gos_plot2()
     
     png("pharaoh_folder/treeplot.png", width = 620, height = 590, res = 80)
     plot(tree_gos_plot)
     dev.off()
     list(src = "pharaoh_folder/treeplot.png",
          contentType="image/png")
     
   }, deleteFile=T
   )
   
   # Download tab's results
   # Download GO table
   output$downloadGOSTable2 <- downloadHandler(
     filename= function() {
       paste("GOS_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_gos <- total_table_gos2()
       
       write.table(x = total_table_gos,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download emaplot
   output$gos_download2 <- downloadHandler(
     filename= function() {
       paste("gos_plot", ".png", sep="")
     },
     content= function(file) {
       ema_gos_plot <- ema_gos_plot2()
       
       png(file, height = 1200, width = 1500, res=140)
       plot(ema_gos_plot)
       dev.off()
     })
   
   # Download treeplot
   output$tree_gos_download2 <- downloadHandler(
     filename= function() {
       paste("gos_treeplot", ".png", sep="")
     },
     content= function(file) {
       tree_gos_plot <- tree_gos_plot2()
       
       png(file, height = 1200, width = 1500, res=140)
       plot(tree_gos_plot)
       dev.off()
     })
   
   
   
   ###################### KEGG ###########################
   
   observeEvent(input$run_button2, {
     removeUI(
       selector = "div:has(>> #selected_kosI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#kos_selectionI2")
     
     if (UI_exist_kegg2)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg2 <<- F
     }
     
     if (UI_exist_kegg_path2)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI2")
       
       UI_exist_kegg_path2 <<- F
     }
     
     if (UI_exist_pathview2)
     {
       removeUI(
         selector = "div:has(>>> #path_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview2 <<- F
     }
     
     output$error_kos2 <- NULL
     
     
   })
   
   observeEvent(input$kegg_start2, {
     
     if (UI_exist_kegg2)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg2 <<- F
     }
     
     if (UI_exist_kegg_path2)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI2")
       
       UI_exist_kegg_path2 <<- F
     }
     
     if (UI_exist_pathview2)
     {
       removeUI(
         selector = "div:has(>>> #path_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview2 <<- F
     }
     
     output$error_kos2 <- NULL
     
     insertUI("#selected_kos2", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_kosI2","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced2()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({diamond_table2()$ID[1]}))
       
       
     })
     
     
     insertUI("#kos_selection2", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("kos_selectionI2", "Show KEGG pathways", size = "sm",
                                style = "float", color = "success")
     })
     
   })
   
   tab_kegg2 <- reactive({
     
     shinyjs::showElement(id = 'loading.ko2')
     # Create KOs set
     kos_anot <- read.csv("pharaoh_folder/ko_table_funtree.tsv", sep="\t", header = T)
     sel.genes.ko <- as.vector(isolate({input$selected_kosI2}))
     
     tab_kegg <- subset(kos_anot, gene %in% sel.genes.ko)
     set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
     
     # Show an error if no terms are identified in the input
     {
       if (length(set_kegg) == 0) 
       {
         shinyjs::hideElement(id = 'loading.ko2')
         output$error_kos2 <- renderUI({
           renderPrint({cat("0 KO terms identified. Please select more genes. If this 
        message persists, it should be interpreted as a lack of KO annotation for this orthogroup")})
         })
         
         if (UI_exist_kegg2)
         {
           removeUI(
             selector = "div:has(>> #output_kos_table2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKOSTable2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           
           UI_exist_kegg2 <<- F
         }
         
         if (UI_exist_kegg_path2)
         {
           removeUI(
             selector = "div:has(>> #output_kegg_table2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #selected_pathsI2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGTable2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI("#paths_buttonI2")
           
           UI_exist_kegg_path2 <<- F
         }
         
         if (UI_exist_pathview2)
         {
           removeUI(
             selector = "div:has(>>> #path_image2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGpathway2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           UI_exist_pathview2 <<- F
         }
         
         
         validate("No KO terms detected")
       }
     }
     
     
     return(tab_kegg)
     
   }) %>% bindEvent(input$kos_selectionI2)
   
   total_table_kegg2 <- reactive({
     
     tab_kegg <- tab_kegg2()
     set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
     
     # Load libraries
     library(clusterProfiler)
     library(enrichplot)
     
     # Enrich with pvalue cutoff = 1 to show all paths
     os_enrich <- enrichKEGG(gene         = unique(unlist(strsplit(set_kegg, "/"))),
                             organism     = 'ko',
                             pvalueCutoff = 1)
     
     total_table_kegg <- as.data.frame(kos_enrich)
     
     # Show an error if the KOs are not mapped to any KEGG pathway
     {
       if (nrow(total_table_kegg) == 0) 
       {
         shinyjs::hideElement(id = 'loading.ko2')
         output$error_kos2 <- renderUI({
           renderPrint({cat("No KEGG pathway appears in this OG.")})
         })
         
         
         if (UI_exist_kegg_path2)
         {
           removeUI(
             selector = "div:has(>> #output_kegg_table2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #selected_pathsI2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGTable2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI("#paths_buttonI2")
           
           UI_exist_kegg_path2 <<- F
         }
         
         if (UI_exist_pathview2)
         {
           removeUI(
             selector = "div:has(>>> #path_image2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGpathway2)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           UI_exist_pathview2 <<- F
         }
         
         validate("No KEGG pathway appears in this OG.")
       }
     }
     output$error_kos2 <- NULL
     
     # Filter out pathways that are not present in plants
     kegg_plants <- read.csv("pharaoh_folder/pathways_plant.ids", sep = "\t", header = T)$x
     total_table_kegg <- subset(total_table_kegg, ID %in% kegg_plants)
     
     return(total_table_kegg)
     
   }) %>% bindEvent(input$kos_selectionI2)
   
   total_table_kos2 <- reactive({
     
     tab_kegg <- tab_kegg2()
     
     library(KEGGREST)
     
     # Collapse genes that share KOs
     
     tab_kegg_for_ko <- subset(tab_kegg, ko != "")
     gene.v.ko <- c()
     individual_kos <- unique(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
     
     for (i in 1:length(individual_kos))
     {
       new.table <- subset(tab_kegg_for_ko, grepl(individual_kos[i],ko))
       new.cha <- as.character(new.table$gene)
       gene.v.ko <- c(gene.v.ko, paste(new.cha, collapse = "/"))
     }
     
     names(gene.v.ko) <- individual_kos
     
     # Create gene chains, count and KO IDs fields (same order)
     count_ko <- table(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
     geneids.ko <- gene.v.ko[names(count_ko)]
     count_terms.ko <- mapply(function(x) {keggFind("ko", x)}, names(count_ko), USE.NAMES = F)
     total_table_kos <- data.frame(ko=names(count_ko), name=count_terms.ko, count=as.numeric(count_ko),
                                   genes=geneids.ko)
     
     return(total_table_kos)
     
   }) %>% bindEvent(input$kos_selectionI2)
   
   # Create boxes for outputs
   
   observeEvent(isTruthy(total_table_kos2()), {
     
     if (UI_exist_kegg2)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg2 <<- F
     }
     
     insertUI("#box_kos_table2", "afterEnd", ui = {
       box(width = 12,
           title = "KO Terms Table", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_kos_table2")
       )
     })
     
     insertUI("#download_ui_for_kos_table2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKOSTable2", "Download KO Table",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_kegg2 <<- TRUE
   })
   
   observeEvent(isTruthy(total_table_kegg2()), {
     
     if (UI_exist_kegg_path2)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI2")
       
       UI_exist_kegg_path2 <<- F
     }
     
     
     insertUI("#box_kegg_table2", "afterEnd", ui = {
       box(width = 12,
           title = "KEGG Pathways Table", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_kegg_table2")
       )
     })
     
     
     
     insertUI("#download_ui_for_kegg_table2", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKEGGTable2", "Download Pathways Table",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_kegg_path2 <<- TRUE
     
     # Remove previous results for pathview
     if (UI_exist_pathview2)
     {
       removeUI(
         selector = "div:has(>>> #path_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview2 <<- F
     }
   })
   
   # Fill outputs
   # Render KO table
   output$output_kos_table2 <- renderDataTable({
     total_table_kos <- total_table_kos2()
     total_table_kos$ko <- sapply(total_table_kos$ko, ko.link)
     total_table_kos
   },escape=FALSE, rownames= F, options =list(pageLength = 5))
   
   # Render KEGG table
   output$output_kegg_table2 <- renderDataTable({
     total_table_kegg <- total_table_kegg2()
     total_table_kegg$ID <- sapply(total_table_kegg$ID, kegg.link)
     total_table_kegg[,c("ID", "Description", "geneID")]
   },escape=FALSE, rownames= F, options =list(pageLength = 5))
   
   # Download tab's outputs
   # Download KO table
   output$downloadKOSTable2 <- downloadHandler(
     filename= function() {
       paste("KO_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_kos <- total_table_kos2()
       write.table(x = total_table_kos,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download KEGG table
   output$downloadKEGGTable2 <- downloadHandler(
     filename= function() {
       paste("KEGG_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_kegg <- total_table_kegg2()
       write.table(x = total_table_kegg[,c("ID", "Description", "geneID")],quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Create pathway selector and button
   observeEvent(input$kos_selectionI2,{
     
     total_table_kos <- total_table_kos2()
     total_table_kegg <- total_table_kegg2()
     
     if(nrow(total_table_kegg) != 0)
     {
       paths.options <- sapply(strsplit(total_table_kegg$ID, split = "ko"), function(x) x[[2]])
       
       
       insertUI("#selected_paths2", "afterEnd", ui = {
         shinyWidgets::pickerInput(inputId = "selected_pathsI2", label = "Select the pathway to plot", 
                                   choices = paths.options, selected = paths.options[1], multiple = F)
         
       })
       
       
       insertUI("#paths_button2", "afterEnd", ui = {
         
         shinyWidgets::actionBttn("paths_buttonI2", "Plot Pathway", size = "sm",
                                  style = "float", color = "success")
       })
       
       shinyjs::hideElement(id = 'loading.ko2')
     }
   })
   
   observeEvent(input$paths_buttonI2,{
     
     if (UI_exist_pathview2)
     {
       removeUI(
         selector = "div:has(>>> #path_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway2)",
         multiple = TRUE,
         immediate = TRUE
       )
     }
     
     UI_exist_pathview2 <<- F
     
   })
   
   # Create Image to Render and save path name
   pathway.current.id2 <- reactive({
     pathway.current.id <- input$selected_pathsI2
     total_table_kos <- total_table_kos2()
     
     kos_unique <- unique(total_table_kos$ko)
     gene.pathway <- rep(0, length(kos_unique))
     names(gene.pathway) <-  kos_unique
     gene.pathway[kos_unique] <-1
     
     library(pathview)
     pathview(gene.data = sort(gene.pathway,decreasing = TRUE),kegg.dir = "pharaoh_folder",
              pathway.id = pathway.current.id,
              species = "ko",
              limit = list(gene=max(abs(gene.pathway)), cpd=1),
              gene.idtype ="kegg")
     
     return(pathway.current.id)
     
   }) %>% bindEvent(input$paths_buttonI2)
   
   # Create output box and download button
   observeEvent(isTruthy(pathway.current.id2()),{
     
     insertUI("#box_path_image2", "afterEnd", ui = {
       box(width = 12,
           title = "KEGG Pathway Plot", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           fluidRow(column(1), imageOutput("path_image2", width = "100%", height = "700px"))
       )
     })
     
     insertUI("#path_download_ui2", "afterEnd", ui = {
       tags$div(shinyWidgets::downloadBttn(outputId= "downloadKEGGpathway2", "Download KEGG Pathway Plot",
                                           size = "sm", color = "success"))
     })
     
     UI_exist_pathview2 <<- T
     
   })
   
   # Fill path image output
   output$path_image2 <- renderImage({
     
     pathway.current.id <- pathway.current.id2()
     list(src = paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."),
          contentType="image/png",width=900,height=700)
   },deleteFile = F)
   
   # Download and remove path image output
   output$downloadKEGGpathway2 <- downloadHandler(
     filename= function() {
       paste("path_plot", ".png", sep="")
     },
     content= function(file) {
       pathway.current.id <- pathway.current.id2()
       file.copy(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."), file)
       file.remove(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."))
     })   
   
   
   
   ############################# LITERATURE ANNOTATION ##########################################
   
   observeEvent(input$run_button2, {
     removeUI(
       selector = "div:has(>> #selected_litI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(> #query_litI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#lit_selectionI2")
     
     
     if (UI_exist_lit2)
     {
       removeUI(
         selector = "div:has(>> #output_lit_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadLITTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_lit2 <<- F
     }
     
   })
   
   observeEvent(input$lit_start2, {
     
     insertUI("#query_lit2", "afterEnd", ui = {
       
       textInput(inputId = "query_litI2",value = "", label = "Enter search term", placeholder = "CCA1")
       
     })
     
     
     insertUI("#selected_lit2", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_litI2","Select the search mode",
                                 choices=c("Normal","Exact", "Substring", "Alias"),
                                 multiple = F, selected = "Normal")
       
       
     })
     
     
     insertUI("#lit_selection2", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("lit_selectionI2", "Get biological information and papers", size = "sm",
                                style = "float", color = "success")
     })
     
   })
   
   pc_result2 <- reactive({
     
     shinyjs::showElement(id = 'loading.lit2')
     pc_search <- as.character(input$query_litI2)
     pc_search <- gsub(" ", "%20", pc_search) 
     pc_modality <- tolower(as.character(input$selected_litI2))
     
     # Get PlantConnectome URL for query
     pc_url <- paste(c("https://connectome.plant.tools", pc_modality, pc_search), collapse = "/")
     
     library(RCurl)
     
     pc_res <- getURL(pc_url)
     
     if (!length(grep("No hits", pc_res)) == 0)
     {
       shinyjs::hideElement(id = 'loading.lit2')
       
       if (UI_exist_lit2)
       {
         removeUI(
           selector = "div:has(>> #output_lit_table2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadLITTable2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_lit2 <<- F
       }
       
       output$error_lit2 <- renderUI({
         renderPrint({cat("No results found for this query")})
       })
       
       validate(" ")
       
     }
     
     output$error_lit2 <- NULL
     
     # Isolate data frame from complete HTML file
     pc_split <- strsplit(as.character(pc_res), split = "<tbody")[[1]][2]
     pc_split <- strsplit(as.character(pc_split), split = "</tbody>")[[1]][1]
     
     pc_clean <- gsub("</tr>", "", pc_split)
     pc_clean <- gsub("[\r\n\t]", "", pc_clean)
     
     pc_vector <- strsplit(pc_clean, split = "<tr>")[[1]][-1]
     pc_vector2 <- sapply(pc_vector, FUN=function(x) strsplit(x, split = "<td> | </td>"))
     pc_table <- sapply(pc_vector2, FUN=function(x) as.character(unlist(x)))
     
     colnames(pc_table) <- NULL
     pc_result <- data.frame(t(pc_table[c(2,4,6,8),]))
     colnames(pc_result) <- c("Source", "Interaction Type", "Target", "Pubmed ID")
     
     return(pc_result)
     
   }) %>% bindEvent(input$lit_selectionI2)
   
   pc_result_show2 <- reactive({
     
     pc_result <- pc_result2()
     
     # Add links to papers
     urls_connect <- sapply(pc_result$`Pubmed ID`, FUN = function(x) paste0(c("<a href=\"",
                                                                              "https://pubmed.ncbi.nlm.nih.gov/",x,"/",
                                                                              "\" target=\"_blank\">", x,
                                                                              "</a>"),
                                                                            collapse=""))
     pc_result_show <- pc_result
     pc_result_show$`Pubmed ID` <- urls_connect
     
     return(pc_result_show)
     
   })
   
   # Create boxes for outputs
   observeEvent(isTruthy(pc_result_show2()), {
     
     if (UI_exist_lit2)
     {
       removeUI(
         selector = "div:has(>> #output_lit_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadLITTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_lit_table2", "afterEnd", ui = {
       box(width = 12,
           title = "Literature Table", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_lit_table2")
       )
     })
     
     insertUI("#download_ui_for_lit_table2", "afterEnd", ui = {
       tags$div(style = "margin-left: 400px;", shinyWidgets::downloadBttn(outputId= "downloadLITTable2", "Download Literature Table",
                                                                          size = "sm", color = "success"))
     })
     
     shinyjs::hideElement(id = 'loading.lit2')
     
     UI_exist_lit2 <<- TRUE
     
   })
   
   # Fill outputs
   # Render table
   output$output_lit_table2 <- renderDataTable({
     pc_result_show2()
   },escape=FALSE, rownames= F, options =list(pageLength = 10))
   
   # Download results
   output$downloadLITTable2 <- downloadHandler(
     filename= function() {
       paste("literature_table", ".tsv", sep="")
     },
     content= function(file) {
       pc_result <- pc_result2()
       write.table(x = pc_result,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })   
   
   
   ######################## STRING ###########################
   
   observeEvent(input$run_button2, {
     removeUI(
       selector = "div:has(>> #selected_stringI2)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#string_selectionI2")
     
     shinyjs::hideElement("error_string2")
     
     
     if (UI_exist_string2)
     {
       removeUI(
         selector = "div:has(>> #output_st_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #output_count_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #count_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadSTRINGTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCOUNTTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #count_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_networkI2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#network_buttonI2")
       
       UI_exist_string2 <<- F
     }
     
     if (UI_exist_network2)
     {
       removeUI(
         selector = "div:has(>>>> #network_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_network2 <<- F
     }
     
   })
   
   # First, create a table that links reduced gene names to its species
   string_sel_table2 <- reactive({
     
     # Define previous variables
     tree_reduced <- tree_reduced2()
     tree <- tree_adj2()
     
     tips_to_keep.mp <- tips_to_keep.mp2()
     tips_to_keep.ot <- tips_to_keep.ot2()
     tips_to_keep.at <- tips_to_keep.at2()
     tips_to_keep.cp <- tips_to_keep.cp2()
     tips_to_keep.cr <- tips_to_keep.cr2()
     tips_to_keep.cz <- tips_to_keep.cz2()
     tips_to_keep.kn <- tips_to_keep.kn2()
     tips_to_keep.me <- tips_to_keep.me2()
     tips_to_keep.mi <- tips_to_keep.mi2()
     tips_to_keep.pp <- tips_to_keep.pp2()
     tips_to_keep.sl <- tips_to_keep.sl2()
     tips_to_keep.sm <- tips_to_keep.sm2()
     tips_to_keep.sp <- tips_to_keep.sp2()
     tips_to_keep.ta <- tips_to_keep.ta2()
     tips_to_keep.vc <- tips_to_keep.vc2()
     tips_to_keep.bp <- tips_to_keep.bp2()
     tips_to_keep.cri <- tips_to_keep.cri2()
     tips_to_keep.ds <- tips_to_keep.ds2()
     tips_to_keep.os <- tips_to_keep.os2()
     tips_to_keep.smag <- tips_to_keep.smag2()
     tips_to_keep.tp <- tips_to_keep.tp2()
     tips_to_keep.aa <- tips_to_keep.aa2()
     tips_to_keep.um <- tips_to_keep.um2()
     tips_to_keep.rs <- tips_to_keep.rs2()
     tips_to_keep.cyc <- tips_to_keep.cyc2()
     tips_to_keep.ca <- tips_to_keep.ca2()
     tips_to_keep.mv <- tips_to_keep.mv2()
     tips_to_keep.af <- tips_to_keep.af2()
     tips_to_keep.sc <- tips_to_keep.sc2()
     tips_to_keep.aegi <- tips_to_keep.aegi2()
     tips_to_keep.sb <- tips_to_keep.sb2()
     tips_to_keep.chara <- tips_to_keep.chara2()
     tips_to_keep.sceobli <- tips_to_keep.sceobli2()
     tips_to_keep.cocco <- tips_to_keep.cocco2()
     tips_to_keep.haema <- tips_to_keep.haema2()
     tips_to_keep.zm <- tips_to_keep.zm2()
     tips_to_keep.praco <- tips_to_keep.praco2()
     tips_to_keep.ostlu <- tips_to_keep.ostlu2()
     tips_to_keep.ostme <- tips_to_keep.ostme2()
     tips_to_keep.micco <- tips_to_keep.micco2()
     tips_to_keep.cymte <- tips_to_keep.cymte2()
     tips_to_keep.chlin <- tips_to_keep.chlin2()
     tips_to_keep.gonpe <- tips_to_keep.gonpe2()
     tips_to_keep.tetso <- tips_to_keep.tetso2()
     tips_to_keep.desar <- tips_to_keep.desar2()
     tips_to_keep.monmi <- tips_to_keep.monmi2()
     tips_to_keep.caule <- tips_to_keep.caule2()
     tips_to_keep.ostqu <- tips_to_keep.ostqu2()
     tips_to_keep.brysp <- tips_to_keep.brysp2()
     tips_to_keep.ellbi <- tips_to_keep.ellbi2()
     tips_to_keep.myrbi <- tips_to_keep.myrbi2()
     tips_to_keep.apalo <- tips_to_keep.apalo2()
     tips_to_keep.astgl <- tips_to_keep.astgl2()
     tips_to_keep.tresp <- tips_to_keep.tresp2()
     tips_to_keep.picso <- tips_to_keep.picso2()
     tips_to_keep.chlso <- tips_to_keep.chlso2()
     tips_to_keep.auxpr <- tips_to_keep.auxpr2()
     tips_to_keep.helsp <- tips_to_keep.helsp2()
     tips_to_keep.tetst <- tips_to_keep.tetst2()
     tips_to_keep.pedsp <- tips_to_keep.pedsp2()
     tips_to_keep.chlpr <- tips_to_keep.chlpr2()
     tips_to_keep.picsp <- tips_to_keep.picsp2()
     tips_to_keep.mesvb <- tips_to_keep.mesvb2()
     tips_to_keep.meskr <- tips_to_keep.meskr2()
     tips_to_keep.zygci <- tips_to_keep.zygci2()
     tips_to_keep.zygcb <- tips_to_keep.zygcb2()
     tips_to_keep.zygcy <- tips_to_keep.zygcy2()
     tips_to_keep.plesc <- tips_to_keep.plesc2()
     tips_to_keep.funhy <- tips_to_keep.funhy2()
     tips_to_keep.sphfa <- tips_to_keep.sphfa2()
     tips_to_keep.takle <- tips_to_keep.takle2()
     tips_to_keep.marpa <- tips_to_keep.marpa2()
     tips_to_keep.ricfl <- tips_to_keep.ricfl2()
     tips_to_keep.ricso <- tips_to_keep.ricso2()
     tips_to_keep.antpu <- tips_to_keep.antpu2()
     tips_to_keep.antfu <- tips_to_keep.antfu2()
     tips_to_keep.megfl <- tips_to_keep.megfl2()
     tips_to_keep.phach <- tips_to_keep.phach2()
     tips_to_keep.phyph <- tips_to_keep.phyph2()
     tips_to_keep.parpe <- tips_to_keep.parpe2()
     tips_to_keep.notor <- tips_to_keep.notor2()
     tips_to_keep.phaca <- tips_to_keep.phaca2()
     tips_to_keep.phasp <- tips_to_keep.phasp2()
     tips_to_keep.leidu <- tips_to_keep.leidu2()
     tips_to_keep.isosi <- tips_to_keep.isosi2()
     tips_to_keep.dipco <- tips_to_keep.dipco2()
     tips_to_keep.hupas <- tips_to_keep.hupas2()
     tips_to_keep.lyccl <- tips_to_keep.lyccl2()
     tips_to_keep.adica <- tips_to_keep.adica2()
     tips_to_keep.alssp <- tips_to_keep.alssp2()
     tips_to_keep.marve <- tips_to_keep.marve2()
     tips_to_keep.seqse <- tips_to_keep.seqse2()
     tips_to_keep.seqgi <- tips_to_keep.seqgi2()
     tips_to_keep.taxch <- tips_to_keep.taxch2()
     tips_to_keep.abial <- tips_to_keep.abial2()
     tips_to_keep.picab <- tips_to_keep.picab2()
     tips_to_keep.gnemo <- tips_to_keep.gnemo2()
     tips_to_keep.welmi <- tips_to_keep.welmi2()
     tips_to_keep.ginbi <- tips_to_keep.ginbi2()
     tips_to_keep.ambtr <- tips_to_keep.ambtr2()
     tips_to_keep.nymco <- tips_to_keep.nymco2()
     tips_to_keep.cinka <- tips_to_keep.cinka2()
     tips_to_keep.horvu <- tips_to_keep.horvu2()
     tips_to_keep.bradi <- tips_to_keep.bradi2()
     tips_to_keep.setit <- tips_to_keep.setit2()
     tips_to_keep.panha <- tips_to_keep.panha2()
     tips_to_keep.missi <- tips_to_keep.missi2()
     tips_to_keep.oroth <- tips_to_keep.oroth2()
     tips_to_keep.eleco <- tips_to_keep.eleco2()
     tips_to_keep.anaco <- tips_to_keep.anaco2()
     tips_to_keep.musac <- tips_to_keep.musac2()
     tips_to_keep.dioal <- tips_to_keep.dioal2()
     tips_to_keep.spipo <- tips_to_keep.spipo2()
     tips_to_keep.aspof <- tips_to_keep.aspof2()
     tips_to_keep.zosma <- tips_to_keep.zosma2()
     tips_to_keep.araly <- tips_to_keep.araly2()
     tips_to_keep.capgr <- tips_to_keep.capgr2()
     tips_to_keep.brara <- tips_to_keep.brara2()
     tips_to_keep.braol <- tips_to_keep.braol2()
     tips_to_keep.sisir <- tips_to_keep.sisir2()
     tips_to_keep.schpa <- tips_to_keep.schpa2()
     tips_to_keep.eutsa <- tips_to_keep.eutsa2()
     tips_to_keep.thlar <- tips_to_keep.thlar2()
     tips_to_keep.boest <- tips_to_keep.boest2()
     tips_to_keep.carpa <- tips_to_keep.carpa2()
     tips_to_keep.theca <- tips_to_keep.theca2()
     tips_to_keep.goshi <- tips_to_keep.goshi2()
     tips_to_keep.gosra <- tips_to_keep.gosra2()
     tips_to_keep.citsi <- tips_to_keep.citsi2()
     tips_to_keep.pontr <- tips_to_keep.pontr2()
     tips_to_keep.eucgr <- tips_to_keep.eucgr2()
     tips_to_keep.maldo <- tips_to_keep.maldo2()
     tips_to_keep.prupe <- tips_to_keep.prupe2()
     tips_to_keep.frave <- tips_to_keep.frave2()
     tips_to_keep.corav <- tips_to_keep.corav2()
     tips_to_keep.caril <- tips_to_keep.caril2()
     tips_to_keep.queru <- tips_to_keep.queru2()
     tips_to_keep.cucsa <- tips_to_keep.cucsa2()
     tips_to_keep.glyma <- tips_to_keep.glyma2()
     tips_to_keep.medtr <- tips_to_keep.medtr2()
     tips_to_keep.tripr <- tips_to_keep.tripr2()
     tips_to_keep.cicar <- tips_to_keep.cicar2()
     tips_to_keep.lotja <- tips_to_keep.lotja2()
     tips_to_keep.phaac <- tips_to_keep.phaac2()
     tips_to_keep.vigun <- tips_to_keep.vigun2()
     tips_to_keep.lupal <- tips_to_keep.lupal2()
     tips_to_keep.lencu <- tips_to_keep.lencu2()
     tips_to_keep.arahy <- tips_to_keep.arahy2()
     tips_to_keep.poptr <- tips_to_keep.poptr2()
     tips_to_keep.manes <- tips_to_keep.manes2()
     tips_to_keep.salpu <- tips_to_keep.salpu2()
     tips_to_keep.linus <- tips_to_keep.linus2()
     tips_to_keep.corci <- tips_to_keep.corci2()
     tips_to_keep.vitvi <- tips_to_keep.vitvi2()
     tips_to_keep.kalfe <- tips_to_keep.kalfe2()
     tips_to_keep.vacda <- tips_to_keep.vacda2()
     tips_to_keep.oleeu <- tips_to_keep.oleeu2()
     tips_to_keep.mimgu <- tips_to_keep.mimgu2()
     tips_to_keep.soltu <- tips_to_keep.soltu2()
     tips_to_keep.cofar <- tips_to_keep.cofar2()
     tips_to_keep.lacsa <- tips_to_keep.lacsa2()
     tips_to_keep.dauca <- tips_to_keep.dauca2()
     tips_to_keep.betvu <- tips_to_keep.betvu2()
     tips_to_keep.amahy <- tips_to_keep.amahy2()
     tips_to_keep.chequ <- tips_to_keep.chequ2()
     tips_to_keep.sapof <- tips_to_keep.sapof2()
     tips_to_keep.aquco <- tips_to_keep.aquco2()
     
     # Table construction
     org.factor <- c()
     
     for (i in 1:length(tree_reduced$tip.label))
     {
       if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
       {
         org.factor <- c(org.factor,"Marchantia polymorpha")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
       {
         org.factor <- c(org.factor,"Ostreococcus tauri")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
       {
         org.factor <- c(org.factor,"Arabidopsis thaliana")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
       {
         org.factor <- c(org.factor,"Ceratodon purpureus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
       {
         org.factor <- c(org.factor,"Chlamydomonas reinhardtii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
       {
         org.factor <- c(org.factor,"Chromochloris zofingiensis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
       {
         org.factor <- c(org.factor,"Klebsormidium nitens")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
       {
         org.factor <- c(org.factor,"Mesotaenium endlicherianum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
       {
         org.factor <- c(org.factor,"Micromonas pusilla")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
       {
         org.factor <- c(org.factor,"Physcomitrium patens")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
       {
         org.factor <- c(org.factor,"Solanum lycopersicum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
       {
         org.factor <- c(org.factor,"Selaginella moellendorffii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
       {
         org.factor <- c(org.factor,"Spirogloea muscicola")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
       {
         org.factor <- c(org.factor,"Triticum aestivum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
       {
         org.factor <- c(org.factor,"Volvox carteri")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
       {
         org.factor <- c(org.factor,"Bathycoccus prasinos")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
       {
         org.factor <- c(org.factor,"Ceratopteris richardii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
       {
         org.factor <- c(org.factor,"Dunaliella salina")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
       {
         org.factor <- c(org.factor,"Oryza sativa")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
       {
         org.factor <- c(org.factor,"Sphagnum magellanicum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
       {
         org.factor <- c(org.factor,"Thuja plicata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
       {
         org.factor <- c(org.factor,"Anthoceros agrestis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
       {
         org.factor <- c(org.factor,"Ulva mutabilis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
       {
         org.factor <- c(org.factor,"Raphidocelis subcapitata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
       {
         org.factor <- c(org.factor,"Cycas panzhihuaensis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
       {
         org.factor <- c(org.factor,"Chlorokybus atmophyticus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
       {
         org.factor <- c(org.factor,"Mesostigma viride CCAC 1140")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
       {
         org.factor <- c(org.factor,"Azolla filiculoides")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
       {
         org.factor <- c(org.factor,"Salvinia cucullata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
       {
         org.factor <- c(org.factor,"Aegilops tauschii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
       {
         org.factor <- c(org.factor,"Sorghum bicolor")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
       {
         org.factor <- c(org.factor,"Chara braunii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
       {
         org.factor <- c(org.factor,"Scenedesmus obliquus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
       {
         org.factor <- c(org.factor,"Coccomyxa subellipsoidea")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
       {
         org.factor <- c(org.factor,"Haematococcus lacustris")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
       {
         org.factor <- c(org.factor,"Zea mays")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.praco])
       {
         org.factor <- c(org.factor,"Prasinoderma coloniale")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostlu])
       {
         org.factor <- c(org.factor,"Ostreococcus lucimarinus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostme])
       {
         org.factor <- c(org.factor,"Ostreococcus mediterraneus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.micco])
       {
         org.factor <- c(org.factor,"Micromonas commoda")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymte])
       {
         org.factor <- c(org.factor,"Cymbomonas tetramitiformis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlin])
       {
         org.factor <- c(org.factor,"Chlamydomonas incerta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gonpe])
       {
         org.factor <- c(org.factor,"Gonium pectorale")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetso])
       {
         org.factor <- c(org.factor,"Tetrabaena socialis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.desar])
       {
         org.factor <- c(org.factor,"Desmodesmus armatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.monmi])
       {
         org.factor <- c(org.factor,"Monoraphidium minutum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caule])
       {
         org.factor <- c(org.factor,"Caulerpa lentillifera")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostqu])
       {
         org.factor <- c(org.factor,"Ostreobium quekettii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brysp])
       {
         org.factor <- c(org.factor,"Bryopsis sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ellbi])
       {
         org.factor <- c(org.factor,"Elliptochloris bilobata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.myrbi])
       {
         org.factor <- c(org.factor,"Myrmecia bisecta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.apalo])
       {
         org.factor <- c(org.factor,"Apatococcus lobatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.astgl])
       {
         org.factor <- c(org.factor,"Asterochloris glomerata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tresp])
       {
         org.factor <- c(org.factor,"Trebouxia sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picso])
       {
         org.factor <- c(org.factor,"Picochlorum soloecismus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlso])
       {
         org.factor <- c(org.factor,"Chlorella sorokiniana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.auxpr])
       {
         org.factor <- c(org.factor,"Auxenochlorella protothecoides")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.helsp])
       {
         org.factor <- c(org.factor,"Helicosporidium sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetst])
       {
         org.factor <- c(org.factor,"Tetraselmis striata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pedsp])
       {
         org.factor <- c(org.factor,"Pedinophyceae sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlpr])
       {
         org.factor <- c(org.factor,"Chloropicon primus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picsp])
       {
         org.factor <- c(org.factor,"Picocystis sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mesvb])
       {
         org.factor <- c(org.factor,"Mesostigma viride NIES 296")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.meskr])
       {
         org.factor <- c(org.factor,"Mesotaenium kramstae")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygci])
       {
         org.factor <- c(org.factor,"Zygnema circumcarinatum SAG")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcb])
       {
         org.factor <- c(org.factor,"Zygnema circumcarinatum UTEX 1559")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcy])
       {
         org.factor <- c(org.factor,"Zygnema cylindricum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.plesc])
       {
         org.factor <- c(org.factor,"Pleurozium schreberi")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.funhy])
       {
         org.factor <- c(org.factor,"Funaria hygrometrica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sphfa])
       {
         org.factor <- c(org.factor,"Sphagnum fallax")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.takle])
       {
         org.factor <- c(org.factor,"Takakia lepidozioides")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marpa])
       {
         org.factor <- c(org.factor,"Marchantia paleacea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricfl])
       {
         org.factor <- c(org.factor,"Riccia fluitans")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricso])
       {
         org.factor <- c(org.factor,"Riccia sorocarpa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antpu])
       {
         org.factor <- c(org.factor,"Anthoceros punctatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antfu])
       {
         org.factor <- c(org.factor,"Anthoceros fusiformis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.megfl])
       {
         org.factor <- c(org.factor,"Megaceros flagellaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phach])
       {
         org.factor <- c(org.factor,"Phaeomegaceros chiloensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phyph])
       {
         org.factor <- c(org.factor,"Phymatoceros phymatodes")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.parpe])
       {
         org.factor <- c(org.factor,"Paraphymatoceros pearsonii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.notor])
       {
         org.factor <- c(org.factor,"Notothylas orbicularis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaca])
       {
         org.factor <- c(org.factor,"Phaeoceros carolinianus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phasp])
       {
         org.factor <- c(org.factor,"Phaeoceros sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.leidu])
       {
         org.factor <- c(org.factor,"Leiosporoceros dussii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.isosi])
       {
         org.factor <- c(org.factor,"Isoetes sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dipco])
       {
         org.factor <- c(org.factor,"Diphasiastrum complanatum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.hupas])
       {
         org.factor <- c(org.factor,"Huperzia asiatica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lyccl])
       {
         org.factor <- c(org.factor,"Lycopodium clavatum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.adica])
       {
         org.factor <- c(org.factor,"Adiantum capillus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.alssp])
       {
         org.factor <- c(org.factor,"Alsophila spinulosa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marve])
       {
         org.factor <- c(org.factor,"Marsilea vestita")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqse])
       {
         org.factor <- c(org.factor,"Sequoia sempervirens")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqgi])
       {
         org.factor <- c(org.factor,"Sequoiadendron giganteum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.taxch])
       {
         org.factor <- c(org.factor,"Taxus chinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.abial])
       {
         org.factor <- c(org.factor,"Abies alba")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picab])
       {
         org.factor <- c(org.factor,"Picea abies")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gnemo])
       {
         org.factor <- c(org.factor,"Gnetum montanum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.welmi])
       {
         org.factor <- c(org.factor,"Welwitschia mirabilis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ginbi])
       {
         org.factor <- c(org.factor,"Ginkgo biloba")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ambtr])
       {
         org.factor <- c(org.factor,"Amborella trichopoda")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.nymco])
       {
         org.factor <- c(org.factor,"Nymphaea colorata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cinka])
       {
         org.factor <- c(org.factor,"Cinnamomum kanehirae")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.horvu])
       {
         org.factor <- c(org.factor,"Hordeum vulgare")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bradi])
       {
         org.factor <- c(org.factor,"Brachypodium distachyon")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.setit])
       {
         org.factor <- c(org.factor,"Setaria italica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.panha])
       {
         org.factor <- c(org.factor,"Panicum hallii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.missi])
       {
         org.factor <- c(org.factor,"Miscanthus sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oroth])
       {
         org.factor <- c(org.factor,"Oropetium thomaeum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eleco])
       {
         org.factor <- c(org.factor,"Eleusine coracana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.anaco])
       {
         org.factor <- c(org.factor,"Ananas comosus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.musac])
       {
         org.factor <- c(org.factor,"Musa acuminata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dioal])
       {
         org.factor <- c(org.factor,"Dioscorea alata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.spipo])
       {
         org.factor <- c(org.factor,"Spirodela polyrhiza")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aspof])
       {
         org.factor <- c(org.factor,"Asparagus officinalis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zosma])
       {
         org.factor <- c(org.factor,"Zostera marina")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.araly])
       {
         org.factor <- c(org.factor,"Arabidopsis lyrata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.capgr])
       {
         org.factor <- c(org.factor,"Capsella grandiflora")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brara])
       {
         org.factor <- c(org.factor,"Brassica rapa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.braol])
       {
         org.factor <- c(org.factor,"Brassica oleracea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sisir])
       {
         org.factor <- c(org.factor,"Sisymbrium irio")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.schpa])
       {
         org.factor <- c(org.factor,"Schrenkiella parvula")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eutsa])
       {
         org.factor <- c(org.factor,"Eutrema salsugineum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.thlar])
       {
         org.factor <- c(org.factor,"Thlaspi arvense")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.boest])
       {
         org.factor <- c(org.factor,"Boechera stricta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.carpa])
       {
         org.factor <- c(org.factor,"Carica papaya")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.theca])
       {
         org.factor <- c(org.factor,"Theobroma cacao")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.goshi])
       {
         org.factor <- c(org.factor,"Gossypium hirsutum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gosra])
       {
         org.factor <- c(org.factor,"Gossypium raimondii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.citsi])
       {
         org.factor <- c(org.factor,"Citrus sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pontr])
       {
         org.factor <- c(org.factor,"Poncirus trifoliata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eucgr])
       {
         org.factor <- c(org.factor,"Eucalyptus grandis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.maldo])
       {
         org.factor <- c(org.factor,"Malus domestica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.prupe])
       {
         org.factor <- c(org.factor,"Prunus persica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.frave])
       {
         org.factor <- c(org.factor,"Fragaria vesca")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corav])
       {
         org.factor <- c(org.factor,"Corylus avellana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caril])
       {
         org.factor <- c(org.factor,"Carya illinoinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.queru])
       {
         org.factor <- c(org.factor,"Quercus rubra")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cucsa])
       {
         org.factor <- c(org.factor,"Cucumis sativus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.glyma])
       {
         org.factor <- c(org.factor,"Glycine max")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.medtr])
       {
         org.factor <- c(org.factor,"Medicago truncatula")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tripr])
       {
         org.factor <- c(org.factor,"Trifolium pratense")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cicar])
       {
         org.factor <- c(org.factor,"Cicer arietinum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lotja])
       {
         org.factor <- c(org.factor,"Lotus japonicus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaac])
       {
         org.factor <- c(org.factor,"Phaseolus acutifolius")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vigun])
       {
         org.factor <- c(org.factor,"Vigna unguiculata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lupal])
       {
         org.factor <- c(org.factor,"Lupinus albus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lencu])
       {
         org.factor <- c(org.factor,"Lens culinaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.arahy])
       {
         org.factor <- c(org.factor,"Arachis hypogaea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.poptr])
       {
         org.factor <- c(org.factor,"Populus trichocarpa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.manes])
       {
         org.factor <- c(org.factor,"Manihot esculenta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.salpu])
       {
         org.factor <- c(org.factor,"Salix purpurea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.linus])
       {
         org.factor <- c(org.factor,"Linum usitatissimum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corci])
       {
         org.factor <- c(org.factor,"Corymbia citriodora")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vitvi])
       {
         org.factor <- c(org.factor,"Vitis vinifera")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kalfe])
       {
         org.factor <- c(org.factor,"Kalanchoe fedtschenkoi")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vacda])
       {
         org.factor <- c(org.factor,"Vaccinium darrowii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oleeu])
       {
         org.factor <- c(org.factor,"Olea europaea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mimgu])
       {
         org.factor <- c(org.factor,"Mimulus guttatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.soltu])
       {
         org.factor <- c(org.factor,"Solanum tuberosum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cofar])
       {
         org.factor <- c(org.factor,"Coffea arabica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lacsa])
       {
         org.factor <- c(org.factor,"Lactuca sativa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dauca])
       {
         org.factor <- c(org.factor,"Daucus carota")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.betvu])
       {
         org.factor <- c(org.factor,"Beta vulgaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.amahy])
       {
         org.factor <- c(org.factor,"Amaranthus hypochondriacus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chequ])
       {
         org.factor <- c(org.factor,"Chenopodium quinoa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sapof])
       {
         org.factor <- c(org.factor,"Saponaria officinalis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aquco])
       {
         org.factor <- c(org.factor,"Aquilegia coerulea")
       }
       
       
     }
     
    
     
     #Matrix with labels and colors and transform to dplyr format
     string.sel.table <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                                    org = org.factor)
     
     return(string.sel.table)
     
   }) %>% bindEvent(input$string_start2)
   
   # Now, selection is allowed for genes of species with STRING support
   observeEvent(input$string_start2, {
     
     string_sel_table <- string_sel_table2()
     allow_string_species <- c("Aegilops tauschii", "Arabidopsis thaliana", "Bathycoccus prasinos", "Chara braunii", 
                               "Chlamydomonas reinhardtii","Coccomyxa subellipsoidea", "Klebsormidium nitens", 
                               "Micromonas commoda", "Oryza sativa", "Micromonas pusilla", "Ostreococcus lucimarinus",
                               "Ostreococcus tauri", "Physcomitrium patens", "Raphidocelis subcapitata",
                               "Scenedesmus obliquus", "Selaginella moellendorffii", "Solanum lycopersicum",
                               "Sorghum bicolor", "Triticum aestivum", "Volvox carteri")
     
     st_genes <- subset(string_sel_table, string_sel_table$org %in% allow_string_species)$label
     
     if (length(st_genes) == 0)
     {
       shinyjs::showElement("error_string2")
       output$error_string2 <- renderUI({renderText({print("No results for this
      analysis due to lack of genes of STRING-supported species in the selection.")})})
       validate(" ")
     }
     
     output$error_string2 <- NULL
     
     insertUI("#selected_string2", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_stringI2","Select the desired genes from the tree",
                                 choices=st_genes, options = list(`actions-box` = TRUE),
                                 multiple = T)
       
     })
     
     
     insertUI("#string_selection2", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("string_selectionI2", "Show STRING Interactions", size = "sm",
                                style = "float", color = "success")
     })
     
   })
   
   phys_table2 <- reactive({
     
     shinyjs::showElement(id = 'loading.string2')
     query_genes <- input$selected_stringI2
     string_sel_table <- string_sel_table2()
     
     # Load complete STRING annotation table
     library(data.table)
     
     # Load iteratively from split files using data.table format
     # data_phys <- fread("pharaoh_folder/string_physical/string_physical_1.tsv")
     # 
     # for (x in list.files("pharaoh_folder/string_physical/")[-1])
     # {
     #   data_phys <- rbind(data_phys, 
     #                      fread(paste0("pharaoh_folder/string_physical/", x)))
     # }
     
     # Load data only for selected organisms
     species_selected <- gsub("[.]", "", gsub(" ", "_", tolower(unique(string_sel_table$org))))
     
     # Get a List of all files named with a key word, say all `.csv` files
     filenames_string <- list.files("pharaoh_folder/string_physical/")
     files_selected <- filenames_string[unlist(sapply(species_selected, function(x) grep(x, filenames_string)))]
     
     # If in the future any of these species is added to STRING:
     # { 
     #   if (seq_org == "Zygnema circumcarinatum SAG")
     #   {
     #     org_search <- "zygnema_circumcarinatum"
     #   }
     #   else if (seq_org == "Zygnema circumcarinatum UTEX 1559")
     #   {
     #     org_search <- "zygnema_circumcarinatum1559"
     #   }
     #   else if (seq_org == "Mesostigma viride CCAC 1140")
     #   {
     #     org_search <- "mesostigma_viride1140"
     #   }
     #   else if (seq_org == "Mesostigma viride NIES 296")
     #   {
     #     org_search <- "mesostigma_viride296"
     #   }
     #   else
     #   {
     #    org_search <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(seq_org))))
     #   }
     # }
     
     # Load and bind all data sets
     data_phys <- rbindlist(lapply(paste0("pharaoh_folder/string_physical/", files_selected),fread))
     
     # Subset by query genes using data.table for speed
     #string_res <- subset(data_phys, data_phys$prot_query %in% query_genes)
     string_res <- data_phys[prot_query %in% query_genes,]
     string_res <- as.data.frame(string_res)
     
     # Assign OG ID to each target
     ortho_data_file <- "Gene_Trees/Orthogroups.tsv"
     
     ortho_data <- as.data.frame(fread(ortho_data_file))
     
     ortho_char <- apply(ortho_data, MARGIN = 1, function(x) paste(x, collapse = ","))
     ortho.numbers <- sapply(string_res$prot_interaction, function(x) grep(x, ortho_char), USE.NAMES = F)
     
     # If a pattern is found in several names, i.e., is a subpattern of several genes,
     # search for the exact match
     if(class(ortho.numbers) == "list")
     {
       # If a gene isn't associated to an OG
       index.none <- which(sapply(ortho.numbers, function(x) length(x) == 0))
       
       if (length(index.none) != 0)
       {
         # We create another row for OG table to associate those genes
         ortho_data <- rbind(ortho_data, "No OG")
         ortho.numbers[index.none] <- nrow(ortho_data)
       }
       
       # If a gene has more than one match due to subpatterns
       index.wrong <- which(sapply(ortho.numbers, function(x) length(x) > 1))
       {
         if (length(index.wrong) == 0)
         {
           ortho.numbers <- unlist(ortho.numbers, use.names = F)
         }
         else
         {
           for (i in index.wrong)
           {
             for (j in ortho.numbers[[i]])
             {
               ortho_char_split <- strsplit(ortho_char[j],split = ",")
               ortho_char_split_clean <- sapply(ortho_char_split, function(x) gsub(" ", "", x))
               if (string_res$prot_interaction[i] %in% ortho_char_split_clean)
               {
                 ortho.numbers[i] <- j
                 ortho.numbers <- unlist(ortho.numbers, use.names = F)
               }
             }
           }
         }
       }
     }
     
     
     
     # Create the final table
     ortho.string.names <- ortho_data$Orthogroup[ortho.numbers]
     phys_table <- data.frame(string_res, orthogroup = ortho.string.names)
     
     return(phys_table)
   }) %>% bindEvent(input$string_selectionI2)
   
   # Create count table to identify enriched OGs in STRING result
   string_counts2 <- reactive({
     
     phys_table <- phys_table2()
     string_counts <- sort(table(phys_table$orthogroup), decreasing = T)
     
     return(string_counts)
     
   }) %>% bindEvent(input$string_selectionI2)
   
   string_count_plot2 <- reactive({
     
     library(ggplot2)
     library(dplyr)
     
     data_count <- as.data.frame(string_counts2())
     colnames(data_count) <- c("orthogroup", "value")
     
     # Compute the position of labels
     data_count <- data_count %>%
       arrange(desc(orthogroup)) %>%
       mutate(prop = value / sum(data_count$value) *100) %>%
       mutate(ypos = cumsum(prop)- 0.5*prop )
     
     # Create plot
     count_plot <- ggplot(data_count, aes(x="", y=prop, fill=orthogroup)) +
       geom_bar(stat="identity", width=1, color="white") +
       coord_polar("y", start=0) +
       theme_void() +
       theme(legend.position="none") +
       #geom_text(aes(y = ypos, label = orthogroup), color = "white", size=6) +
       scale_fill_manual(values = rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"), 
                                      floor(nrow(data_count)/9)+1))
     
     return(count_plot)
     
   }) %>% bindEvent(input$string_selectionI2)
   
   string_count_plotly2 <- reactive({
     
     library(ggplot2)
     library(dplyr)
     
     data_count <- as.data.frame(string_counts2())
     colnames(data_count) <- c("orthogroup", "value")
     
     # Compute the position of labels
     data_count <- data_count %>%
       arrange(desc(orthogroup)) %>%
       mutate(prop = value / sum(data_count$value) *100) %>%
       mutate(ypos = cumsum(prop)- 0.5*prop )
     
     # Create plot
     count_plotly <- plotly::plot_ly(data=data_count,values=~prop,labels=~factor(orthogroup),
                                     marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                                            floor(nrow(data_count)/9)+1)),
                                     type="pie",showlegend = F, text= ~paste0("</br> ", orthogroup,
                                                                              "</br> ",prop, "%"),
                                     textinfo = "none", hoverinfo = "text") 
     
     
     
     return(count_plotly)
     
   }) %>% bindEvent(input$string_selectionI2)
   
   # Create boxes for outputs
   observeEvent(isTruthy(string_count_plot2()), {
     
     if (UI_exist_string2)
     {
       removeUI(
         selector = "div:has(>> #output_st_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #output_count_table2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #count_plot2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadSTRINGTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCOUNTTable2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #count_download2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_networkI2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#network_buttonI2")
       
     }
     
     insertUI("#box_st_table2", "afterEnd", ui = {
       box(width = 12,
           title = "STRING Interactions Table", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_st_table2")
       )
     })
     
     insertUI("#box_count_table2", "afterEnd", ui = {
       box(width = 12,
           title = "Interacting Orthogroups Table", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_count_table2")
       )
     })
     
     insertUI("#box_count_plot2", "afterEnd", ui = {
       box(width = 12,
           title = "Interacting Orthogroups Plot", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           plotlyOutput("count_plot2")
       )
     })
     
     
     insertUI("#download_ui_for_st_table2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadSTRINGTable2", "Download STRING Table",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#download_ui_for_count_table2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadCOUNTTable2", "Download OG Count Table",
                                                                          size = "sm", color = "success"))
     })
     
     insertUI("#count_down_button2", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "count_download2", "Download OG Count Plot",
                                                                          size = "sm", color = "success"))
     })
     
     UI_exist_string2 <<- TRUE
     
     # Remove previous results for STRING network
     if (UI_exist_network2)
     {
       removeUI(
         selector = "div:has(>>>> #network_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_network2 <<- F
     }
     
   })
   
   # Fill outputs
   # Render STRING table
   # output$output_st_table1 <- renderDataTable({
   #   phys_table <- phys_table1()
   #   #phys_table$type <- sapply(phys_table$type, color_string)
   #   datatable(phys_table) %>%
   #     formatStyle(
   #       'type',
   #       color = styleEqual(
   #         c("Direct interaction", "Interolog"), c('green', '#CA931B')
   #       )
   #       )
   # },escape=FALSE, rownames= F, options =list(pageLength = 10)) 
   
   
   output$output_st_table2 <- renderDataTable({
     phys_table <- phys_table2()
     #phys_table$type <- sapply(phys_table$type, color_string)
     datatable(phys_table, escape=FALSE, rownames= F, options =list(pageLength = 10)) %>%
       formatStyle(
         'type',
         color = styleEqual(
           c("Direct interaction", "Interolog"), c('green', '#CA931B')
         )
       )
   }) 
   
   
   # Render OG count table
   output$output_count_table2 <- renderDataTable({
     string_counts <- as.data.frame(string_counts2())
     colnames(string_counts) <- c("orthogroup", "count")
     string_counts
   },escape=FALSE, rownames= F, options =list(pageLength = 7))
   
   # Render OG count pie chart
   output$count_plot2 <- renderPlotly({
     string_count_plotly2()
   })
   
   
   # Download tab's outputs
   # Download STRING table
   output$downloadSTRINGTable2 <- downloadHandler(
     filename= function() {
       paste("string_table", ".tsv", sep="")
     },
     content= function(file) {
       phys_table <- phys_table2()
       write.table(x = phys_table,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download count table
   output$downloadCOUNTTable2 <- downloadHandler(
     filename= function() {
       paste("string_count_table", ".tsv", sep="")
     },
     content= function(file) {
       string_counts <- string_counts2()
       write.table(x = string_counts,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download count plot
   output$count_download2 <- downloadHandler(
     filename= function() {
       paste("string_count", ".png", sep="")
     },
     content= function(file) {
       string_count_plot <- string_count_plot2()
       
       png(file, height = 450, width = 450)
       plot(string_count_plot)
       dev.off()
     })
   
   # Create gene selector and button for STRING network representation
   observeEvent(input$string_selectionI2,{
     
     phys_table <- phys_table2()
     network_genes <- unique(phys_table$prot_query)
     
     # Error message if no genes are allowed for selection should have  been reported
     # earlier
     
     insertUI("#selected_network2", "afterEnd", ui = {
       
       shinyWidgets::pickerInput(inputId = "selected_networkI2", label = "Select the gene whose network you want to plot", 
                                 choices = network_genes, selected = network_genes[1],
                                 options = list(`actions-box` = TRUE), multiple = T)
       
     })
     
     
     insertUI("#network_button2", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("network_buttonI2", "Plot STRING Network", size = "sm",
                                style = "float", color = "success")
     })
     
     shinyjs::hideElement(id = 'loading.string2')
     
   })
   
   mapped_string2 <- reactive({
     
     # Load genes (PharaohFUN IDs) and convert to STRING IDs
     library(data.table)
     
     network_genes <- input$selected_networkI2
     map_table <- fread("pharaoh_folder/string_map.tsv")
     
     # Fast subset using data.table
     map_network <- map_table[pharaohfun_id %in% network_genes,]
     
     # Error if no genes from selection have an associated high fidelity STRING ID
     if (nrow(map_network) == 0)
     {
       
       if (UI_exist_network2)
       {
         removeUI(
           selector = "div:has(>>>> #network_image2)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_network2 <<- F
       }
       
       output$error_network2 <- renderUI({renderText({print("It's not possible to map any
                               genes from selection to STRING IDs, please select different ones.")})})
       validate( " ")
       
     }
     
     output$error_network2 <- NULL
     
     # Get only STRING IDS and paste in a format interpretable by JS function
     string_ids <- as.data.frame(map_network)$string_id
     
     
     mapped_string <- paste0(string_ids, collapse = "%0d")
     return(mapped_string)
     
   }) %>% bindEvent(input$network_buttonI2)
   
   # Create boxes
   observeEvent(isTruthy(mapped_string2()),{
     
     mapped_string <- mapped_string2()
     
     if (UI_exist_network2)
     {
       removeUI(
         selector = "div:has(>>>> #network_image2)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     url_interactive <- paste0("https://string-db.org/cgi/network?identifiers=", 
                               mapped_string, 
                               "&add_color_nodes=25&network_flavor=confidence&show_query_node_labels=1")
     
     insertUI("#box_output_network2", "afterEnd", ui = {
       box(width = 12,
           title = "Image", status = "success", solidHeader = TRUE,
           collapsible = TRUE,
           fluidRow(column(1), 
                    column(8, htmlOutput("network_image2")), 
                    column(3, div( style = "margin-top: 300px;", 
                                   shinyWidgets::actionBttn("network_link2", "Interactive Network", size = "md", 
                                                            icon = icon("circle-nodes"), style = "float", color = "success", 
                                                            onclick=paste0("window.open('", url_interactive,"','_blank')")))
                           
                    ))
           
       )
     })
     
     UI_exist_network2 <<- T
     
   })
   
   # Fill network box
   
   output$network_image2 <- renderText({
     
     mapped_string <- mapped_string2()
     src_map <- paste0("https://string-db.org/api/image/network?identifiers=", mapped_string, 
                       "&add_color_nodes=25&network_flavor=confidence")
     
     c('<img src="',src_map,'"width="675" height="625">')
   })
   
   
# End of Sequence-based search results 
   
   
     ############## ORTHOGROUP-BASED SEARCH ##############
   
   # Set global variables for tracking changes in output
   UI_exist_pfam3 <<- F
   UI_exist_tree3 <<- F
   UI_exist_phylo3 <<- F
   UI_exist_cafe3 <<- F
   UI_exist_error_cafe3 <<- F
   UI_exist_msa3 <<- F
   UI_exist_go3 <<- F
   UI_exist_kegg3 <<- F
   UI_exist_kegg_path3 <<- F
   UI_exist_pathview3 <<- F
   UI_exist_lit3 <<-  F
   UI_exist_string3 <<-  F
   UI_exist_network3 <<-  F
   
   model.selected3 <- reactive({
     model.selected <- F
     return(model.selected)
   })%>% bindEvent(input$run_button3)
   
   build_trees3 <- reactive({
     build_trees <- as.character(input$build_trees_3)
     return(build_trees)
   }) %>% bindEvent(input$run_button3)
   
   # Load organisms selection based on the model selected
   selected_organisms3 <- reactive({
     selected_organisms <- c(input$prasino_check_3, input$mami_check_3,
                             input$chloro_check_3, input$strepto_check_3,
                             input$bryo_check_3, input$lyco_check_3,
                             input$gymno_check_3, input$magno_check_3,
                             input$monocots_check_3, input$dicots_first_check_3,
                             input$dicots_second_check_3)
     
     return(selected_organisms)
     
   }) %>% bindEvent(input$run_button3)
   
   selected_values_org3 <- reactive(organisms_values[selected_organisms3()]) %>% bindEvent(input$run_button3)
   
   
   og.name3 <- reactive({
     shinyjs::showElement(id = 'loading.tree3')
     
     gene.name.tree <- input$geneInt3
     
     # Load table with orthogroups information depending on selected model
     ortho.table.search <- "Gene_Trees/Orthogroups.tsv"
     ortho.table <- read.csv(ortho.table.search,
                             header = T, sep = "\t", as.is = T,
                             fill = T, blank.lines.skip = F)
     
     
     # Error if OG does not correspond to any of OG IDs available
     if (!(gene.name.tree %in% ortho.table$Orthogroup))
     {
       shinyjs::hideElement(id = 'loading.tree3')
       
       if (UI_exist_tree3)
       {
         removeUI(
           selector = "div:has(>> #treeTips3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs3)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree3 <<- F
       output$error_tree3 <- renderUI({renderText({print("No results for this 
      query due to not supported OG ID, check that it is correct.")})})
       validate(need(gene.name.tree %in% ortho.table$Orthogroup, " "))
       
     }
     
     return(gene.name.tree)
     
   }) %>% bindEvent(input$run_button3)
   
   tree3 <- reactive({
     file.name <- og.name3()
     # Load gene tree file depending on the input
     
     tree.name <- paste("Gene_Trees",paste(file.name, "tree.txt", sep = "_"), sep="/")
     
     # Error if tree file not found
     if (!(file.exists(tree.name)))
     {
       shinyjs::hideElement(id = 'loading.tree3')
       
       if (UI_exist_tree3)
       {
         removeUI(
           selector = "div:has(>> #treeTips3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs3)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree3 <<- F
       output$error_tree3 <- renderUI({renderText({print("Unable to construct tree associated to
                                                        to an orthogroup with less than 4 genes.")})})
       validate(need(file.exists(tree.name), " "))
     }
     
     # Generate tree depending on the tree building method selected
     {
       if (build_trees3() == "Maximum Likelihood")
       {
         tree <- read.tree(tree.name)
         return(tree)
       }
       
       else if(build_trees3() == "Bayesian Inference")
       {
         tree.bayes <- paste("Tree_Bayes",paste0("bayes_tree_", file.name, ".nwk"), sep="/")
         
         # Error if tree file not found
         if (!(file.exists(tree.bayes)))
         {
           shinyjs::hideElement(id = 'loading.tree3')
           
           if (UI_exist_tree3)
           {
             removeUI(
               selector = "div:has(>> #treeTips3)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>>> #presentorg3)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #tree_image3)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #downloadTree3)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #downloadNewick3)",
               multiple = TRUE,
               immediate = TRUE
             )
             
             removeUI(
               selector = "div:has(>> #downloadTreeSeqs3)",
               multiple = TRUE,
               immediate = TRUE
             )
           }
           
           UI_exist_tree3 <<- F
           output$error_tree3 <- renderUI({renderText({print("Bayesian tree inference is currently under development. 
                                                          We are working to offer this method for all 
                                                          orthogroups through regular updates. If this message appears,
                                                          the orthogroup of interest is not yet available. If you want it
                                                          to appear in the next update, please send an email to mramos5@us.es 
                                                          and we will try to prioritize it. Meanwhile, try the other methods
                                                          to build the gene tree.")})})
           validate(need(file.exists(tree.bayes), " "))
         }
         
         tree <- read.tree(tree.bayes)
         return(tree)
         
       }
       
       else
       {
         library(phangorn)
         
         # Read MSA
         multiple_file <- paste("MultipleSequenceAlignments", paste0(file.name, ".fa"), sep="/")
         
         
         mytree <- read.phyDat(file = multiple_file,
                               format="fasta", type = "AA")
         
         # To keep the same workflow, the reduced tree will be calculated, then the applied subsetting
         # will only reduce them if the selected method has been fasttree
         
         
         # Selection of genes from the selected organism
         
         # Create empty vectors
         
         tips_to_keep.praco3 = tips_to_keep.ot3 = tips_to_keep.ostlu3 = tips_to_keep.ostme3 = tips_to_keep.bp3 <- c()
         tips_to_keep.mi3 = tips_to_keep.micco3 = tips_to_keep.cymte3 = tips_to_keep.cr3 = tips_to_keep.chlin3 <- c()
         tips_to_keep.haema3 = tips_to_keep.vc3 = tips_to_keep.ds3 = tips_to_keep.gonpe3 = tips_to_keep.tetso3 <- c()
         tips_to_keep.cz3 = tips_to_keep.rs3 = tips_to_keep.sceobli3 = tips_to_keep.desar3 = tips_to_keep.monmi3 <- c()
         tips_to_keep.caule3 = tips_to_keep.ostqu3 = tips_to_keep.brysp3 = tips_to_keep.cocco3 <- c()
         tips_to_keep.ellbi3 = tips_to_keep.myrbi3 = tips_to_keep.apalo3 = tips_to_keep.astgl3 <- c()
         tips_to_keep.tresp3 = tips_to_keep.picso3 = tips_to_keep.chlso3 = tips_to_keep.auxpr3 <- c()
         tips_to_keep.helsp3 = tips_to_keep.um3 = tips_to_keep.tetst3 = tips_to_keep.pedsp3 <- c()
         tips_to_keep.chlpr3 = tips_to_keep.picsp3 = tips_to_keep.ca3 = tips_to_keep.mv3 = tips_to_keep.mesvb3 <- c()
         tips_to_keep.kn3 = tips_to_keep.chara3 = tips_to_keep.me3 = tips_to_keep.meskr3 = tips_to_keep.sp3 <- c()
         tips_to_keep.zygci3 = tips_to_keep.zygcb3 = tips_to_keep.zygcy3 <- c()
         tips_to_keep.cp3 = tips_to_keep.pp3 = tips_to_keep.plesc3 = tips_to_keep.funhy3 <- c()
         tips_to_keep.smag3 = tips_to_keep.sphfa3 = tips_to_keep.takle3 = tips_to_keep.mp3 <- c()
         tips_to_keep.marpa3 = tips_to_keep.ricfl3 = tips_to_keep.ricso3 = tips_to_keep.aa3 <- c()
         tips_to_keep.antpu3 = tips_to_keep.antfu3 = tips_to_keep.megfl3 = tips_to_keep.phach3 <- c()
         tips_to_keep.phyph3 = tips_to_keep.parpe3 = tips_to_keep.notor3 = tips_to_keep.phaca3 <- c()
         tips_to_keep.phasp3 = tips_to_keep.leidu3 <- c()
         tips_to_keep.sm3 = tips_to_keep.isosi3 = tips_to_keep.dipco3 = tips_to_keep.hupas3 = tips_to_keep.lyccl3 <- c()
         tips_to_keep.cri3 = tips_to_keep.adica3 = tips_to_keep.alssp3 = tips_to_keep.sc3 = tips_to_keep.af3 = tips_to_keep.marve3 <- c()
         tips_to_keep.tp3 = tips_to_keep.seqse3 = tips_to_keep.seqgi3 = tips_to_keep.taxch3 = tips_to_keep.abial3 <- c()
         tips_to_keep.picab3 = tips_to_keep.gnemo3 = tips_to_keep.welmi3 = tips_to_keep.cyc3 = tips_to_keep.ginbi3 <- c()
         tips_to_keep.ambtr3 = tips_to_keep.nymco3 = tips_to_keep.cinka3 = tips_to_keep.aegi3 <- c()
         tips_to_keep.ta3 = tips_to_keep.horvu3 = tips_to_keep.bradi3 = tips_to_keep.os3 <- c()
         tips_to_keep.setit3 = tips_to_keep.sb3 = tips_to_keep.zm3 = tips_to_keep.panha3 <- c()
         tips_to_keep.missi3 = tips_to_keep.oroth3 = tips_to_keep.eleco3 = tips_to_keep.anaco3 <- c()
         tips_to_keep.musac3 = tips_to_keep.dioal3 = tips_to_keep.spipo3 = tips_to_keep.aspof3 <- c()
         tips_to_keep.zosma3 = tips_to_keep.at3 = tips_to_keep.araly3 = tips_to_keep.capgr3 <- c()
         tips_to_keep.brara3 = tips_to_keep.braol3 = tips_to_keep.sisir3 = tips_to_keep.schpa3 <- c()
         tips_to_keep.eutsa3 = tips_to_keep.thlar3 = tips_to_keep.boest3 = tips_to_keep.carpa3 <- c()
         tips_to_keep.theca3 = tips_to_keep.goshi3 = tips_to_keep.gosra3 = tips_to_keep.citsi3 <- c()
         tips_to_keep.pontr3 = tips_to_keep.eucgr3 = tips_to_keep.maldo3 = tips_to_keep.prupe3 <- c()
         tips_to_keep.frave3 = tips_to_keep.corav3 = tips_to_keep.caril3 = tips_to_keep.queru3 <- c()
         tips_to_keep.cucsa3 = tips_to_keep.glyma3 = tips_to_keep.medtr3 = tips_to_keep.tripr3 <- c()
         tips_to_keep.cicar3 = tips_to_keep.lotja3 = tips_to_keep.phaac3 = tips_to_keep.vigun3 <- c()
         tips_to_keep.lupal3 = tips_to_keep.lencu3 = tips_to_keep.arahy3 = tips_to_keep.poptr3 <- c()
         tips_to_keep.manes3 = tips_to_keep.salpu3 = tips_to_keep.linus3 = tips_to_keep.corci3 <- c()
         tips_to_keep.vitvi3 = tips_to_keep.kalfe3 = tips_to_keep.vacda3 = tips_to_keep.oleeu3 <- c()
         tips_to_keep.mimgu3 = tips_to_keep.sl3 = tips_to_keep.soltu3 = tips_to_keep.cofar3 <- c()
         tips_to_keep.lacsa3 = tips_to_keep.dauca3 = tips_to_keep.betvu3 = tips_to_keep.amahy3 <- c()
         tips_to_keep.chequ3 = tips_to_keep.sapof3 = tips_to_keep.aquco3 <- c()
         
         organisms.list <- c(selected_values_org3())
         
         
         # Fill vectors if organisms are in list and change names
         {
           if ("mp" %in% organisms.list)
           {
             tips_to_keep.mp3 <- grep(pattern = "marchantia_polymorpha", names(mytree)) 
           }
           
           if ("ot" %in% organisms.list)
           {
             tips_to_keep.ot3 <- grep(pattern = "ostreococcus_tauri",names(mytree)) 
           }
           
           if ("at" %in% organisms.list)
           {
             tips_to_keep.at3 <- grep(pattern = "arabidopsis_thaliana",names(mytree)) 
           }
           
           if ("cp" %in% organisms.list)
           {
             tips_to_keep.cp3 <- grep(pattern = "ceratodon_purpureus",names(mytree)) 
           }
           
           if ("cr" %in% organisms.list)
           {
             tips_to_keep.cr3 <- grep(pattern = "chlamydomonas_reinhardtii",names(mytree))
           }
           
           if ("cz" %in% organisms.list)
           {
             tips_to_keep.cz3 <- grep(pattern = "chromochloris_zofingiensis",names(mytree)) 
           }
           
           if ("kn" %in% organisms.list)
           {
             tips_to_keep.kn3 <- grep(pattern = "klebsormidium_nitens",names(mytree)) 
           }
           
           if ("me" %in% organisms.list)
           {
             tips_to_keep.me3 <- grep(pattern = "mesotaenium_endlicherianum",names(mytree)) 
           }
           
           if ("mi" %in% organisms.list)
           {
             tips_to_keep.mi3 <- grep(pattern = "micromonas_pusilla",names(mytree)) 
           }
           
           if ("pp" %in% organisms.list)
           {
             tips_to_keep.pp3 <- grep(pattern = "physcomitrium_patens",names(mytree)) 
           }
           
           if ("sl" %in% organisms.list)
           {
             tips_to_keep.sl3 <- grep(pattern = "solanum_lycopersicum",names(mytree)) 
           }
           
           if ("sm" %in% organisms.list)
           {
             tips_to_keep.sm3 <- grep(pattern = "selaginella_moellendorffii",names(mytree))
           }
           
           if ("sp" %in% organisms.list)
           {
             tips_to_keep.sp3 <- grep(pattern = "spirogloea_muscicola",names(mytree)) 
           }
           
           if ("ta" %in% organisms.list)
           {
             tips_to_keep.ta3 <- grep(pattern = "triticum_aestivum",names(mytree)) 
           }
           
           if ("vc" %in% organisms.list)
           {
             tips_to_keep.vc3 <- grep(pattern = "volvox_carteri",names(mytree))
           }
           
           if ("bp" %in% organisms.list)
           {
             tips_to_keep.bp3 <- grep(pattern = "bathycoccus_prasinos",names(mytree))
           }
           
           if ("cri" %in% organisms.list)
           {
             tips_to_keep.cri3 <- grep(pattern = "ceratopteris_richardii",names(mytree))
           }
           
           if ("ds" %in% organisms.list)
           {
             tips_to_keep.ds3 <- grep(pattern = "dunaliella_salina",names(mytree))
           }
           
           if ("os" %in% organisms.list)
           {
             tips_to_keep.os3 <- grep(pattern = "oryza_sativa",names(mytree))
           }
           
           if ("smag" %in% organisms.list)
           {
             tips_to_keep.smag3 <- grep(pattern = "sphagnum_magellanicum",names(mytree))
           }
           
           if ("tp" %in% organisms.list)
           {
             tips_to_keep.tp3 <- grep(pattern = "thuja_plicata",names(mytree))
           }
           
           if ("aa" %in% organisms.list)
           {
             tips_to_keep.aa3 <- grep(pattern = "anthoceros_agrestis",names(mytree))
           }
           
           if ("um" %in% organisms.list)
           {
             tips_to_keep.um3 <- grep(pattern = "ulva_mutabilis",names(mytree))
           }
           
           if ("rs" %in% organisms.list)
           {
             tips_to_keep.rs3 <- grep(pattern = "raphidocelis_subcapitata",names(mytree))
           }
           
           if ("cyc" %in% organisms.list)
           {
             tips_to_keep.cyc3 <- grep(pattern = "cycas_panzhihuaensis",names(mytree))
           }
           
           if ("ca" %in% organisms.list)
           {
             tips_to_keep.ca3 <- grep(pattern = "chlorokybus_atmophyticus",names(mytree))
           }
           
           if ("mv" %in% organisms.list)
           {
             tips_to_keep.mv3 <- grep(pattern = "mesostigma_viride1140",names(mytree))
           }
           
           if ("af" %in% organisms.list)
           {
             tips_to_keep.af3 <- grep(pattern = "azolla_filiculoides",names(mytree))
           }
           
           if ("sc" %in% organisms.list)
           {
             tips_to_keep.sc3 <- grep(pattern = "salvinia_cucullata",names(mytree))
           }
           
           if ("aegi" %in% organisms.list)
           {
             tips_to_keep.aegi3 <- grep(pattern = "aegilops_tauschii",names(mytree))
           }
           
           if ("sb" %in% organisms.list)
           {
             tips_to_keep.sb3 <- grep(pattern = "sorghum_bicolor",names(mytree))
           }
           
           if ("chara" %in% organisms.list)
           {
             tips_to_keep.chara3 <- grep(pattern = "chara_braunii",names(mytree))
           }
           
           
           if ("sceobli" %in% organisms.list)
           {
             tips_to_keep.sceobli3 <- grep(pattern = "scenedesmus_obliquus",names(mytree))
           }
           
           if ("cocco" %in% organisms.list)
           {
             tips_to_keep.cocco3 <- grep(pattern = "coccomyxa_subellipsoidea",names(mytree))
           }
           
           if ("haema" %in% organisms.list)
           {
             tips_to_keep.haema3 <- grep(pattern = "haematococcus_lacustris",names(mytree))
           }
           
           if ("zm" %in% organisms.list)
           {
             tips_to_keep.zm3 <- grep(pattern = "zea_mays",names(mytree))
           }
           
           if ("praco" %in% organisms.list)
           {
             tips_to_keep.praco3 <- grep(pattern = "prasinoderma_coloniale",names(mytree))
           }
           
           if ("ostlu" %in% organisms.list)
           {
             tips_to_keep.ostlu3 <- grep(pattern = "ostreococcus_lucimarinus",names(mytree))
           }
           
           if ("ostme" %in% organisms.list)
           {
             tips_to_keep.ostme3 <- grep(pattern = "ostreococcus_mediterraneus",names(mytree))
           }
           
           if ("micco" %in% organisms.list)
           {
             tips_to_keep.micco3 <- grep(pattern = "micromonas_commoda",names(mytree))
           }
           
           if ("cymte" %in% organisms.list)
           {
             tips_to_keep.cymte3 <- grep(pattern = "cymbomonas_tetramitiformis",names(mytree))
           }
           
           
           if ("chlin" %in% organisms.list)
           {
             tips_to_keep.chlin3 <- grep(pattern = "chlamydomonas_incerta",names(mytree))
           }
           
           if ("gonpe" %in% organisms.list)
           {
             tips_to_keep.gonpe3 <- grep(pattern = "gonium_pectorale",names(mytree))
           }
           
           if ("tetso" %in% organisms.list)
           {
             tips_to_keep.tetso3 <- grep(pattern = "tetrabaena_socialis",names(mytree))
           }
           
           if ("desar" %in% organisms.list)
           {
             tips_to_keep.desar3 <- grep(pattern = "desmodesmus_armatus",names(mytree))
           }
           
           if ("monmi" %in% organisms.list)
           {
             tips_to_keep.monmi3 <- grep(pattern = "monoraphidium_minutum",names(mytree))
           }
           
           if ("caule" %in% organisms.list)
           {
             tips_to_keep.caule3 <- grep(pattern = "caulerpa_lentillifera",names(mytree))
           }
           
           if ("ostqu" %in% organisms.list)
           {
             tips_to_keep.ostqu3 <- grep(pattern = "ostreobium_quekettii",names(mytree))
           }
           
           if ("brysp" %in% organisms.list)
           {
             tips_to_keep.brysp3 <- grep(pattern = "bryopsis_sp",names(mytree))
           }
           
           if ("ellbi" %in% organisms.list)
           {
             tips_to_keep.ellbi3 <- grep(pattern = "elliptochloris_bilobata",names(mytree))
           }
           
           if ("myrbi" %in% organisms.list)
           {
             tips_to_keep.myrbi3 <- grep(pattern = "myrmecia_bisecta",names(mytree))
           }
           
           if ("apalo" %in% organisms.list)
           {
             tips_to_keep.apalo3 <- grep(pattern = "apatococcus_lobatus",names(mytree))
           }
           
           if ("astgl" %in% organisms.list)
           {
             tips_to_keep.astgl3 <- grep(pattern = "asterochloris_glomerata",names(mytree))
           }
           
           if ("tresp" %in% organisms.list)
           {
             tips_to_keep.tresp3 <- grep(pattern = "trebouxia_sp",names(mytree))
           }
           
           if ("picso" %in% organisms.list)
           {
             tips_to_keep.picso3 <- grep(pattern = "picochlorum_soloecismus",names(mytree))
           }
           
           if ("chlso" %in% organisms.list)
           {
             tips_to_keep.chlso3 <- grep(pattern = "chlorella_sorokiniana",names(mytree))
           }
           
           if ("auxpr" %in% organisms.list)
           {
             tips_to_keep.auxpr3 <- grep(pattern = "auxenochlorella_protothecoides",names(mytree))
           }
           
           if ("helsp" %in% organisms.list)
           {
             tips_to_keep.helsp3 <- grep(pattern = "helicosporidium_sp",names(mytree))
           }
           
           if ("tetst" %in% organisms.list)
           {
             tips_to_keep.tetst3 <- grep(pattern = "tetraselmis_striata",names(mytree))
           }
           
           if ("pedsp" %in% organisms.list)
           {
             tips_to_keep.pedsp3 <- grep(pattern = "pedinophyceae_sp",names(mytree))
           }
           
           if ("chlpr" %in% organisms.list)
           {
             tips_to_keep.chlpr3 <- grep(pattern = "chloropicon_primus",names(mytree))
           }
           
           if ("picsp" %in% organisms.list)
           {
             tips_to_keep.picsp3 <- grep(pattern = "picocystis_sp",names(mytree))
           }
           
           if ("mesvb" %in% organisms.list)
           {
             tips_to_keep.mesvb3 <- grep(pattern = "mesostigma_viride296",names(mytree))
           }
           
           if ("meskr" %in% organisms.list)
           {
             tips_to_keep.meskr3 <- grep(pattern = "mesotaenium_kramstae",names(mytree))
           }
           
           if ("zygci" %in% organisms.list)
           {
             tips_to_keep.zygci3 <- grep(pattern = "zygnema_circumcarinatum_",names(mytree))
           }
           
           if ("zygcb" %in% organisms.list)
           {
             tips_to_keep.zygcb3 <- grep(pattern = "zygnema_circumcarinatum1559",names(mytree))
           }
           
           if ("zygcy" %in% organisms.list)
           {
             tips_to_keep.zygcy3 <- grep(pattern = "zygnema_cylindricum",names(mytree))
           }
           
           if ("plesc" %in% organisms.list)
           {
             tips_to_keep.plesc3 <- grep(pattern = "pleurozium_schreberi",names(mytree))
           }
           
           if ("funhy" %in% organisms.list)
           {
             tips_to_keep.funhy3 <- grep(pattern = "funaria_hygrometrica",names(mytree))
           }
           
           if ("sphfa" %in% organisms.list)
           {
             tips_to_keep.sphfa3 <- grep(pattern = "sphagnum_fallax",names(mytree))
           }
           
           if ("takle" %in% organisms.list)
           {
             tips_to_keep.takle3 <- grep(pattern = "takakia_lepidozioides",names(mytree))
           }
           
           if ("marpa" %in% organisms.list)
           {
             tips_to_keep.marpa3 <- grep(pattern = "marchantia_paleacea",names(mytree))
           }
           
           if ("ricfl" %in% organisms.list)
           {
             tips_to_keep.ricfl3 <- grep(pattern = "riccia_fluitans",names(mytree))
           }
           
           if ("ricso" %in% organisms.list)
           {
             tips_to_keep.ricso3 <- grep(pattern = "riccia_sorocarpa",names(mytree))
           }
           
           if ("antpu" %in% organisms.list)
           {
             tips_to_keep.antpu3 <- grep(pattern = "anthoceros_punctatus",names(mytree))
           }
           
           if ("antfu" %in% organisms.list)
           {
             tips_to_keep.antfu3 <- grep(pattern = "anthoceros_fusiformis",names(mytree))
           }
           
           if ("megfl" %in% organisms.list)
           {
             tips_to_keep.megfl3 <- grep(pattern = "megaceros_flagellaris",names(mytree))
           }
           
           if ("phach" %in% organisms.list)
           {
             tips_to_keep.phach3 <- grep(pattern = "phaeomegaceros_chiloensis",names(mytree))
           }
           
           if ("phyph" %in% organisms.list)
           {
             tips_to_keep.phyph3 <- grep(pattern = "phymatoceros_phymatodes",names(mytree))
           }
           
           if ("parpe" %in% organisms.list)
           {
             tips_to_keep.parpe3 <- grep(pattern = "paraphymatoceros_pearsonii",names(mytree))
           }
           
           if ("notor" %in% organisms.list)
           {
             tips_to_keep.notor3 <- grep(pattern = "notothylas_orbicularis",names(mytree))
           }
           
           if ("phaca" %in% organisms.list)
           {
             tips_to_keep.phaca3 <- grep(pattern = "phaeoceros_carolinianus",names(mytree))
           }
           
           if ("phasp" %in% organisms.list)
           {
             tips_to_keep.phasp3 <- grep(pattern = "phaeoceros_sp",names(mytree))
           }
           
           if ("leidu" %in% organisms.list)
           {
             tips_to_keep.leidu3 <- grep(pattern = "leiosporoceros_dussii",names(mytree))
           }
           
           if ("isosi" %in% organisms.list)
           {
             tips_to_keep.isosi3 <- grep(pattern = "isoetes_sinensis",names(mytree))
           }
           
           if ("dipco" %in% organisms.list)
           {
             tips_to_keep.dipco3 <- grep(pattern = "diphasiastrum_complanatum",names(mytree))
           }
           
           if ("hupas" %in% organisms.list)
           {
             tips_to_keep.hupas3 <- grep(pattern = "huperzia_asiatica",names(mytree))
           }
           
           if ("lyccl" %in% organisms.list)
           {
             tips_to_keep.lyccl3 <- grep(pattern = "lycopodium_clavatum",names(mytree))
           }
           
           if ("adica" %in% organisms.list)
           {
             tips_to_keep.adica3 <- grep(pattern = "adiantum_capillus",names(mytree))
           }
           
           if ("alssp" %in% organisms.list)
           {
             tips_to_keep.alssp3 <- grep(pattern = "alsophila_spinulosa",names(mytree))
           }
           
           if ("marve" %in% organisms.list)
           {
             tips_to_keep.marve3 <- grep(pattern = "marsilea_vestita",names(mytree))
           }
           
           if ("seqse" %in% organisms.list)
           {
             tips_to_keep.seqse3 <- grep(pattern = "sequoia_sempervirens",names(mytree))
           }
           
           if ("seqgi" %in% organisms.list)
           {
             tips_to_keep.seqgi3 <- grep(pattern = "sequoiadendron_giganteum",names(mytree))
           }
           
           if ("taxch" %in% organisms.list)
           {
             tips_to_keep.taxch3 <- grep(pattern = "taxus_chinensis",names(mytree))
           }
           
           if ("abial" %in% organisms.list)
           {
             tips_to_keep.abial3 <- grep(pattern = "abies_alba",names(mytree))
           }
           
           if ("picab" %in% organisms.list)
           {
             tips_to_keep.picab3 <- grep(pattern = "picea_abies",names(mytree))
           }
           
           if ("gnemo" %in% organisms.list)
           {
             tips_to_keep.gnemo3 <- grep(pattern = "gnetum_montanum",names(mytree))
           }
           
           if ("welmi" %in% organisms.list)
           {
             tips_to_keep.welmi3 <- grep(pattern = "welwitschia_mirabilis",names(mytree))
           }
           
           if ("ginbi" %in% organisms.list)
           {
             tips_to_keep.ginbi3 <- grep(pattern = "ginkgo_biloba",names(mytree))
           }
           
           if ("ambtr" %in% organisms.list)
           {
             tips_to_keep.ambtr3 <- grep(pattern = "amborella_trichopoda",names(mytree))
           }
           
           if ("nymco" %in% organisms.list)
           {
             tips_to_keep.nymco3 <- grep(pattern = "nymphaea_colorata",names(mytree))
           }
           
           if ("cinka" %in% organisms.list)
           {
             tips_to_keep.cinka3 <- grep(pattern = "cinnamomum_kanehirae",names(mytree))
           }
           
           if ("horvu" %in% organisms.list)
           {
             tips_to_keep.horvu3 <- grep(pattern = "hordeum_vulgare",names(mytree))
           }
           
           if ("bradi" %in% organisms.list)
           {
             tips_to_keep.bradi3 <- grep(pattern = "brachypodium_distachyon",names(mytree))
           }
           
           if ("setit" %in% organisms.list)
           {
             tips_to_keep.setit3 <- grep(pattern = "setaria_italica",names(mytree))
           }
           
           if ("panha" %in% organisms.list)
           {
             tips_to_keep.panha3 <- grep(pattern = "panicum_hallii",names(mytree))
           }
           
           if ("missi" %in% organisms.list)
           {
             tips_to_keep.missi3 <- grep(pattern = "miscanthus_sinensis",names(mytree))
           }
           
           if ("oroth" %in% organisms.list)
           {
             tips_to_keep.oroth3 <- grep(pattern = "oropetium_thomaeum",names(mytree))
           }
           
           if ("eleco" %in% organisms.list)
           {
             tips_to_keep.eleco3 <- grep(pattern = "eleusine_coracana",names(mytree))
           }
           
           if ("anaco" %in% organisms.list)
           {
             tips_to_keep.anaco3 <- grep(pattern = "ananas_comosus",names(mytree))
           }
           
           if ("musac" %in% organisms.list)
           {
             tips_to_keep.musac3 <- grep(pattern = "musa_acuminata",names(mytree))
           }
           
           if ("dioal" %in% organisms.list)
           {
             tips_to_keep.dioal3 <- grep(pattern = "dioscorea_alata",names(mytree))
           }
           
           if ("spipo" %in% organisms.list)
           {
             tips_to_keep.spipo3 <- grep(pattern = "spirodela_polyrhiza",names(mytree))
           }
           
           if ("aspof" %in% organisms.list)
           {
             tips_to_keep.aspof3 <- grep(pattern = "asparagus_officinalis",names(mytree))
           }
           
           if ("zosma" %in% organisms.list)
           {
             tips_to_keep.zosma3 <- grep(pattern = "zostera_marina",names(mytree))
           }
           
           if ("araly" %in% organisms.list)
           {
             tips_to_keep.araly3 <- grep(pattern = "arabidopsis_lyrata",names(mytree))
           }
           
           if ("capgr" %in% organisms.list)
           {
             tips_to_keep.capgr3 <- grep(pattern = "capsella_grandiflora",names(mytree))
           }
           
           if ("brara" %in% organisms.list)
           {
             tips_to_keep.brara3 <- grep(pattern = "brassica_rapa",names(mytree))
           }
           
           if ("braol" %in% organisms.list)
           {
             tips_to_keep.braol3 <- grep(pattern = "brassica_oleracea",names(mytree))
           }
           
           if ("sisir" %in% organisms.list)
           {
             tips_to_keep.sisir3 <- grep(pattern = "sisymbrium_irio",names(mytree))
           }
           
           if ("schpa" %in% organisms.list)
           {
             tips_to_keep.schpa3 <- grep(pattern = "schrenkiella_parvula",names(mytree))
           }
           
           if ("eutsa" %in% organisms.list)
           {
             tips_to_keep.eutsa3 <- grep(pattern = "eutrema_salsugineum",names(mytree))
           }
           
           if ("thlar" %in% organisms.list)
           {
             tips_to_keep.thlar3 <- grep(pattern = "thlaspi_arvense",names(mytree))
           }
           
           if ("boest" %in% organisms.list)
           {
             tips_to_keep.boest3 <- grep(pattern = "boechera_stricta",names(mytree))
           }
           
           if ("carpa" %in% organisms.list)
           {
             tips_to_keep.carpa3 <- grep(pattern = "carica_papaya",names(mytree))
           }
           
           if ("theca" %in% organisms.list)
           {
             tips_to_keep.theca3 <- grep(pattern = "theobroma_cacao",names(mytree))
           }
           
           if ("goshi" %in% organisms.list)
           {
             tips_to_keep.goshi3 <- grep(pattern = "gossypium_hirsutum",names(mytree))
           }
           
           if ("gosra" %in% organisms.list)
           {
             tips_to_keep.gosra3 <- grep(pattern = "gossypium_raimondii",names(mytree))
           }
           
           if ("citsi" %in% organisms.list)
           {
             tips_to_keep.citsi3 <- grep(pattern = "citrus_sinensis",names(mytree))
           }
           
           if ("pontr" %in% organisms.list)
           {
             tips_to_keep.pontr3 <- grep(pattern = "poncirus_trifoliata",names(mytree))
           }
           
           if ("eucgr" %in% organisms.list)
           {
             tips_to_keep.eucgr3 <- grep(pattern = "eucalyptus_grandis",names(mytree))
           }
           
           if ("maldo" %in% organisms.list)
           {
             tips_to_keep.maldo3 <- grep(pattern = "malus_domestica",names(mytree))
           }
           
           if ("prupe" %in% organisms.list)
           {
             tips_to_keep.prupe3 <- grep(pattern = "prunus_persica",names(mytree))
           }
           
           if ("frave" %in% organisms.list)
           {
             tips_to_keep.frave3 <- grep(pattern = "fragaria_vesca",names(mytree))
           }
           
           if ("corav" %in% organisms.list)
           {
             tips_to_keep.corav3 <- grep(pattern = "corylus_avellana",names(mytree))
           }
           
           if ("caril" %in% organisms.list)
           {
             tips_to_keep.caril3 <- grep(pattern = "carya_illinoinensis",names(mytree))
           }
           
           if ("queru" %in% organisms.list)
           {
             tips_to_keep.queru3 <- grep(pattern = "quercus_rubra",names(mytree))
           }
           
           if ("cucsa" %in% organisms.list)
           {
             tips_to_keep.cucsa3 <- grep(pattern = "cucumis_sativus",names(mytree))
           }
           
           if ("glyma" %in% organisms.list)
           {
             tips_to_keep.glyma3 <- grep(pattern = "glycine_max",names(mytree))
           }
           
           if ("medtr" %in% organisms.list)
           {
             tips_to_keep.medtr3 <- grep(pattern = "medicago_truncatula",names(mytree))
           }
           
           if ("tripr" %in% organisms.list)
           {
             tips_to_keep.tripr3 <- grep(pattern = "trifolium_pratense",names(mytree))
           }
           
           if ("cicar" %in% organisms.list)
           {
             tips_to_keep.cicar3 <- grep(pattern = "cicer_arietinum",names(mytree))
           }
           
           if ("lotja" %in% organisms.list)
           {
             tips_to_keep.lotja3 <- grep(pattern = "lotus_japonicus",names(mytree))
           }
           
           if ("phaac" %in% organisms.list)
           {
             tips_to_keep.phaac3 <- grep(pattern = "phaseolus_acutifolius",names(mytree))
           }
           
           if ("vigun" %in% organisms.list)
           {
             tips_to_keep.vigun3 <- grep(pattern = "vigna_unguiculata",names(mytree))
           }
           
           if ("lupal" %in% organisms.list)
           {
             tips_to_keep.lupal3 <- grep(pattern = "lupinus_albus",names(mytree))
           }
           
           if ("lencu" %in% organisms.list)
           {
             tips_to_keep.lencu3 <- grep(pattern = "lens_culinaris",names(mytree))
           }
           
           if ("arahy" %in% organisms.list)
           {
             tips_to_keep.arahy3 <- grep(pattern = "arachis_hypogaea",names(mytree))
           }
           
           if ("poptr" %in% organisms.list)
           {
             tips_to_keep.poptr3 <- grep(pattern = "populus_trichocarpa",names(mytree))
           }
           
           if ("manes" %in% organisms.list)
           {
             tips_to_keep.manes3 <- grep(pattern = "manihot_esculenta",names(mytree))
           }
           
           if ("salpu" %in% organisms.list)
           {
             tips_to_keep.salpu3 <- grep(pattern = "salix_purpurea",names(mytree))
           }
           
           if ("linus" %in% organisms.list)
           {
             tips_to_keep.linus3 <- grep(pattern = "linum_usitatissimum",names(mytree))
           }
           
           if ("corci" %in% organisms.list)
           {
             tips_to_keep.corci3 <- grep(pattern = "corymbia_citriodora",names(mytree))
           }
           
           if ("vitvi" %in% organisms.list)
           {
             tips_to_keep.vitvi3 <- grep(pattern = "vitis_vinifera",names(mytree))
           }
           
           if ("kalfe" %in% organisms.list)
           {
             tips_to_keep.kalfe3 <- grep(pattern = "kalanchoe_fedtschenkoi",names(mytree))
           }
           
           if ("vacda" %in% organisms.list)
           {
             tips_to_keep.vacda3 <- grep(pattern = "vaccinium_darrowii",names(mytree))
           }
           
           if ("oleeu" %in% organisms.list)
           {
             tips_to_keep.oleeu3 <- grep(pattern = "olea_europaea",names(mytree))
           }
           
           if ("mimgu" %in% organisms.list)
           {
             tips_to_keep.mimgu3 <- grep(pattern = "mimulus_guttatus",names(mytree))
           }
           
           if ("soltu" %in% organisms.list)
           {
             tips_to_keep.soltu3 <- grep(pattern = "solanum_tuberosum",names(mytree))
           }
           
           if ("cofar" %in% organisms.list)
           {
             tips_to_keep.cofar3 <- grep(pattern = "coffea_arabica",names(mytree))
           }
           
           if ("lacsa" %in% organisms.list)
           {
             tips_to_keep.lacsa3 <- grep(pattern = "lactuca_sativa",names(mytree))
           }
           
           if ("dauca" %in% organisms.list)
           {
             tips_to_keep.dauca3 <- grep(pattern = "daucus_carota",names(mytree))
           }
           
           if ("betvu" %in% organisms.list)
           {
             tips_to_keep.betvu3 <- grep(pattern = "beta_vulgaris",names(mytree))
           }
           
           if ("amahy" %in% organisms.list)
           {
             tips_to_keep.amahy3 <- grep(pattern = "amaranthus_hypochondriacus",names(mytree))
           }
           
           if ("chequ" %in% organisms.list)
           {
             tips_to_keep.chequ3 <- grep(pattern = "chenopodium_quinoa",names(mytree))
           }
           
           if ("sapof" %in% organisms.list)
           {
             tips_to_keep.sapof3 <- grep(pattern = "saponaria_officinalis",names(mytree))
           }
           
           if ("aquco" %in% organisms.list)
           {
             tips_to_keep.aquco3 <- grep(pattern = "aquilegia_coerulea",names(mytree))
           }
         }
         
         
         # Concatenate indexes to keep and subset MSA
         tips_to_keep.global <- c(tips_to_keep.praco3, tips_to_keep.ot3, tips_to_keep.ostlu3, tips_to_keep.ostme3, tips_to_keep.bp3,
                                  tips_to_keep.mi3, tips_to_keep.micco3, tips_to_keep.cymte3, tips_to_keep.cr3, tips_to_keep.chlin3,
                                  tips_to_keep.haema3, tips_to_keep.vc3, tips_to_keep.ds3, tips_to_keep.gonpe3, tips_to_keep.tetso3,
                                  tips_to_keep.cz3, tips_to_keep.rs3, tips_to_keep.sceobli3, tips_to_keep.desar3, tips_to_keep.monmi3,
                                  tips_to_keep.caule3, tips_to_keep.ostqu3, tips_to_keep.brysp3, tips_to_keep.cocco3,
                                  tips_to_keep.ellbi3, tips_to_keep.myrbi3, tips_to_keep.apalo3, tips_to_keep.astgl3,
                                  tips_to_keep.tresp3, tips_to_keep.picso3, tips_to_keep.chlso3, tips_to_keep.auxpr3,
                                  tips_to_keep.helsp3, tips_to_keep.um3, tips_to_keep.tetst3, tips_to_keep.pedsp3,
                                  tips_to_keep.chlpr3, tips_to_keep.picsp3, tips_to_keep.ca3, tips_to_keep.mv3, tips_to_keep.mesvb3,
                                  tips_to_keep.kn3, tips_to_keep.chara3, tips_to_keep.me3, tips_to_keep.meskr3, tips_to_keep.sp3,
                                  tips_to_keep.zygci3, tips_to_keep.zygcb3, tips_to_keep.zygcy3,
                                  tips_to_keep.cp3, tips_to_keep.pp3, tips_to_keep.plesc3, tips_to_keep.funhy3,
                                  tips_to_keep.smag3, tips_to_keep.sphfa3, tips_to_keep.takle3, tips_to_keep.mp3,
                                  tips_to_keep.marpa3, tips_to_keep.ricfl3, tips_to_keep.ricso3, tips_to_keep.aa3,
                                  tips_to_keep.antpu3, tips_to_keep.antfu3, tips_to_keep.megfl3, tips_to_keep.phach3,
                                  tips_to_keep.phyph3, tips_to_keep.parpe3, tips_to_keep.notor3, tips_to_keep.phaca3,
                                  tips_to_keep.phasp3, tips_to_keep.leidu3, tips_to_keep.sm3, tips_to_keep.isosi3, 
                                  tips_to_keep.dipco3, tips_to_keep.hupas3, tips_to_keep.lyccl3,
                                  tips_to_keep.cri3, tips_to_keep.adica3, tips_to_keep.alssp3, tips_to_keep.sc3, 
                                  tips_to_keep.af3, tips_to_keep.marve3, tips_to_keep.tp3, tips_to_keep.seqse3, 
                                  tips_to_keep.seqgi3, tips_to_keep.taxch3, tips_to_keep.abial3, tips_to_keep.picab3,
                                  tips_to_keep.gnemo3, tips_to_keep.welmi3, tips_to_keep.cyc3, tips_to_keep.ginbi3,
                                  tips_to_keep.ambtr3, tips_to_keep.nymco3, tips_to_keep.cinka3, tips_to_keep.aegi3,
                                  tips_to_keep.ta3, tips_to_keep.horvu3, tips_to_keep.bradi3, tips_to_keep.os3,
                                  tips_to_keep.setit3, tips_to_keep.sb3, tips_to_keep.zm3, tips_to_keep.panha3,
                                  tips_to_keep.missi3, tips_to_keep.oroth3, tips_to_keep.eleco3, tips_to_keep.anaco3,
                                  tips_to_keep.musac3, tips_to_keep.dioal3, tips_to_keep.spipo3, tips_to_keep.aspof3,
                                  tips_to_keep.zosma3, tips_to_keep.at3, tips_to_keep.araly3, tips_to_keep.capgr3,
                                  tips_to_keep.brara3, tips_to_keep.braol3, tips_to_keep.sisir3, tips_to_keep.schpa3,
                                  tips_to_keep.eutsa3, tips_to_keep.thlar3, tips_to_keep.boest3, tips_to_keep.carpa3,
                                  tips_to_keep.theca3, tips_to_keep.goshi3, tips_to_keep.gosra3, tips_to_keep.citsi3,
                                  tips_to_keep.pontr3, tips_to_keep.eucgr3, tips_to_keep.maldo3, tips_to_keep.prupe3,
                                  tips_to_keep.frave3, tips_to_keep.corav3, tips_to_keep.caril3, tips_to_keep.queru3,
                                  tips_to_keep.cucsa3, tips_to_keep.glyma3, tips_to_keep.medtr3, tips_to_keep.tripr3,
                                  tips_to_keep.cicar3, tips_to_keep.lotja3, tips_to_keep.phaac3, tips_to_keep.vigun3,
                                  tips_to_keep.lupal3, tips_to_keep.lencu3, tips_to_keep.arahy3, tips_to_keep.poptr3,
                                  tips_to_keep.manes3, tips_to_keep.salpu3, tips_to_keep.linus3, tips_to_keep.corci3,
                                  tips_to_keep.vitvi3, tips_to_keep.kalfe3, tips_to_keep.vacda3, tips_to_keep.oleeu3,
                                  tips_to_keep.mimgu3, tips_to_keep.sl3, tips_to_keep.soltu3, tips_to_keep.cofar3,
                                  tips_to_keep.lacsa3, tips_to_keep.dauca3, tips_to_keep.betvu3, tips_to_keep.amahy3,
                                  tips_to_keep.chequ3, tips_to_keep.sapof3, tips_to_keep.aquco3)
         
         
         my_subset_tree <- subset(mytree, tips_to_keep.global)
         
         {
           if (build_trees3() == "Neighbour Joining")
           {
             # Create dist matrix for NJ and build tree
             dm <- dist.ml(my_subset_tree)
             treeNJ  <- NJ(dm)
             
             # Bootstrap for NJ
             fun_nj <- function(x) NJ(dist.ml(x))
             bs_nj <- bootstrap.phyDat(my_subset_tree, fun_nj, bs=100)
             
             # Save bootstrap values to the tree
             tree <- plotBS(treeNJ, bs_nj, type = "n")
             # Rooting tree
             tree <- midpoint(tree)
             return(tree)
           }
           
           else if (build_trees3() == "UPGMA")
           {
             dm <- dist.ml(my_subset_tree)
             treeUPGMA  <- upgma(dm)
             
             # Bootstrap for UPGMA
             fun_upgma <- function(x) upgma(dist.ml(x))
             bs_upgma <- bootstrap.phyDat(my_subset_tree, fun_upgma, bs=100)
             
             # Save bootstrap values to the tree
             tree <- addConfidences(treeUPGMA, bs_upgma)
             return(tree)
           }
          
         }
         
       }
     }
   }) %>% bindEvent(input$run_button3)
   
   ortho_seq3 <- reactive({
     file.name <- og.name3()
     
     # Load orthogroup sequences file
     ortho.seq.name <- paste("Orthogroup_Sequences",paste(file.name, "fa", sep = "."), sep="/")
     
     ortho_seq <- seqinr::read.fasta(ortho.seq.name, seqtype = "AA")
     return(ortho_seq)
   })
   
   
   # Tips to keep of each species with proper notation
   tips_to_keep.mp3 <- reactive({
     
     tree <- tree3()
     # Selection of organisms
     organisms.list <- c(selected_values_org3())
     
     # Selection of genes from the selected organism
     tips_to_keep.mp <- c()
     if ("mp" %in% organisms.list)
     {
       tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label)
     }
     return(tips_to_keep.mp)
   })
   
   tips_to_keep.ot3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ot <- c()
     if ("ot" %in% organisms.list)
     {
       tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label)
     }
     return(tips_to_keep.ot)
   })
   
   tips_to_keep.at3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.at <- c()
     if ("at" %in% organisms.list)
     {
       tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label)
     }
     return(tips_to_keep.at)
   })
   
   tips_to_keep.cp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cp <- c()
     if ("cp" %in% organisms.list)
     {
       tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label)
     }
     
     return(tips_to_keep.cp)
   })
   
   tips_to_keep.cr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cr <- c()
     if ("cr" %in% organisms.list)
     {
       tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
     }
     return(tips_to_keep.cr)
   })
   
   tips_to_keep.cz3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cz <- c()
     if ("cz" %in% organisms.list)
     {
       tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label)
     }
     return(tips_to_keep.cz)
   })
   
   tips_to_keep.kn3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.kn <- c()
     if ("kn" %in% organisms.list)
     {
       tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label)
     }
     return(tips_to_keep.kn)
   })
   
   tips_to_keep.me3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.me <- c()
     if ("me" %in% organisms.list)
     {
       tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label)
     }
     return(tips_to_keep.me)
   })
   
   tips_to_keep.mi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.mi <- c()
     if ("mi" %in% organisms.list)
     {
       tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label)
     }
     return(tips_to_keep.mi)
   })
   
   tips_to_keep.pp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.pp <- c()
     if ("pp" %in% organisms.list)
     {
       tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label)
     }
     return(tips_to_keep.pp)
   })
   
   tips_to_keep.sl3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sl <- c()
     if ("sl" %in% organisms.list)
     {
       tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label)
     }
     return(tips_to_keep.sl)
   })
   
   tips_to_keep.sm3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sm <- c()
     if ("sm" %in% organisms.list)
     {
       tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label)
     }
     return(tips_to_keep.sm)
   })
   
   tips_to_keep.sp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sp <- c()
     if ("sp" %in% organisms.list)
     {
       tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label)
     }
     return(tips_to_keep.sp)
   })
   
   tips_to_keep.ta3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ta <- c()
     if ("ta" %in% organisms.list)
     {
       tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label)
     }
     return(tips_to_keep.ta)
   })
   
   tips_to_keep.vc3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.vc <- c()
     if ("vc" %in% organisms.list)
     {
       tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
     }
     return(tips_to_keep.vc)
   })
   
   tips_to_keep.bp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.bp <- c()
     if ("bp" %in% organisms.list)
     {
       tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
     }
     return(tips_to_keep.bp)
   })
   
   tips_to_keep.cri3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cri <- c()
     if ("cri" %in% organisms.list)
     {
       tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
     }
     return(tips_to_keep.cri)
   })
   
   tips_to_keep.ds3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ds <- c()
     if ("ds" %in% organisms.list)
     {
       tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
     }
     return(tips_to_keep.ds)
   })
   
   tips_to_keep.os3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.os <- c()
     if ("os" %in% organisms.list)
     {
       tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
     }
     return(tips_to_keep.os)
   })
   
   tips_to_keep.smag3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.smag <- c()
     if ("smag" %in% organisms.list)
     {
       tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
     }
     return(tips_to_keep.smag)
   })
   
   tips_to_keep.tp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.tp <- c()
     if ("tp" %in% organisms.list)
     {
       tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
     }
     return(tips_to_keep.tp)
   })
   
   tips_to_keep.aa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.aa <- c()
     if ("aa" %in% organisms.list)
     {
       tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
     }
     return(tips_to_keep.aa)
   })
   
   tips_to_keep.um3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.um <- c()
     if ("um" %in% organisms.list)
     {
       tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
     }
     return(tips_to_keep.um)
   })
   
   tips_to_keep.rs3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.rs <- c()
     if ("rs" %in% organisms.list)
     {
       tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
     }
     return(tips_to_keep.rs)
   })
   
   tips_to_keep.cyc3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cyc <- c()
     if ("cyc" %in% organisms.list)
     {
       tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
     }
     return(tips_to_keep.cyc)
   })
   
   
   tips_to_keep.ca3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ca <- c()
     if ("ca" %in% organisms.list)
     {
       tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
     }
     return(tips_to_keep.ca)
   })
   
   tips_to_keep.mv3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.mv <- c()
     if ("mv" %in% organisms.list)
     {
       tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
     }
     return(tips_to_keep.mv)
   })
   
   tips_to_keep.af3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.af <- c()
     if ("af" %in% organisms.list)
     {
       tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
     }
     return(tips_to_keep.af)
   })
   
   tips_to_keep.sc3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sc <- c()
     if ("sc" %in% organisms.list)
     {
       tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
     }
     return(tips_to_keep.sc)
   })
   
   tips_to_keep.aegi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.aegi <- c()
     if ("aegi" %in% organisms.list)
     {
       tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
     }
     return(tips_to_keep.aegi)
   })
   
   tips_to_keep.sb3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sb <- c()
     if ("sb" %in% organisms.list)
     {
       tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
     }
     return(tips_to_keep.sb)
   })
   
   tips_to_keep.chara3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.chara <- c()
     if ("chara" %in% organisms.list)
     {
       tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
     }
     return(tips_to_keep.chara)
   })
   
   tips_to_keep.sceobli3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sceobli <- c()
     if ("sceobli" %in% organisms.list)
     {
       tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
     }
     return(tips_to_keep.sceobli)
   })
   
   tips_to_keep.cocco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cocco <- c()
     if ("cocco" %in% organisms.list)
     {
       tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
     }
     return(tips_to_keep.cocco)
   })
   
   
   tips_to_keep.haema3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.haema <- c()
     if ("haema" %in% organisms.list)
     {
       tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
     }
     return(tips_to_keep.haema)
   })
   
   tips_to_keep.zm3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.zm <- c()
     if ("zm" %in% organisms.list)
     {
       tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
     }
     return(tips_to_keep.zm)
   })
   
   tips_to_keep.praco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.praco <- c()
     if ("praco" %in% organisms.list)
     {
       tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
     }
     return(tips_to_keep.praco)
   })
   
   tips_to_keep.ostlu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ostlu <- c()
     if ("ostlu" %in% organisms.list)
     {
       tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
     }
     return(tips_to_keep.ostlu)
   })
   
   tips_to_keep.ostme3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ostme <- c()
     if ("ostme" %in% organisms.list)
     {
       tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
     }
     return(tips_to_keep.ostme)
   })
   
   tips_to_keep.micco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.micco <- c()
     if ("micco" %in% organisms.list)
     {
       tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
     }
     return(tips_to_keep.micco)
   })
   
   tips_to_keep.cymte3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cymte <- c()
     if ("cymte" %in% organisms.list)
     {
       tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
     }
     return(tips_to_keep.cymte)
   })
   
   tips_to_keep.chlin3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.chlin <- c()
     if ("chlin" %in% organisms.list)
     {
       tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
     }
     return(tips_to_keep.chlin)
   })
   
   tips_to_keep.gonpe3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.gonpe <- c()
     if ("gonpe" %in% organisms.list)
     {
       tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
     }
     return(tips_to_keep.gonpe)
   })
   
   tips_to_keep.tetso3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.tetso <- c()
     if ("tetso" %in% organisms.list)
     {
       tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
     }
     return(tips_to_keep.tetso)
   })
   
   tips_to_keep.desar3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.desar <- c()
     if ("desar" %in% organisms.list)
     {
       tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
     }
     return(tips_to_keep.desar)
   })
   
   tips_to_keep.monmi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.monmi <- c()
     if ("monmi" %in% organisms.list)
     {
       tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
     }
     return(tips_to_keep.monmi)
   })
   
   tips_to_keep.caule3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.caule <- c()
     if ("caule" %in% organisms.list)
     {
       tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
     }
     return(tips_to_keep.caule)
   })
   
   tips_to_keep.ostqu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ostqu <- c()
     if ("ostqu" %in% organisms.list)
     {
       tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
     }
     return(tips_to_keep.ostqu)
   })
   
   tips_to_keep.brysp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.brysp <- c()
     if ("brysp" %in% organisms.list)
     {
       tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
     }
     return(tips_to_keep.brysp)
   })
   
   tips_to_keep.ellbi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ellbi <- c()
     if ("ellbi" %in% organisms.list)
     {
       tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
     }
     return(tips_to_keep.ellbi)
   })
   
   tips_to_keep.myrbi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.myrbi <- c()
     if ("myrbi" %in% organisms.list)
     {
       tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
     }
     return(tips_to_keep.myrbi)
   })
   
   tips_to_keep.apalo3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.apalo <- c()
     if ("apalo" %in% organisms.list)
     {
       tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
     }
     return(tips_to_keep.apalo)
   })
   
   tips_to_keep.astgl3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.astgl <- c()
     if ("astgl" %in% organisms.list)
     {
       tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
     }
     return(tips_to_keep.astgl)
   })
   
   tips_to_keep.tresp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.tresp <- c()
     if ("tresp" %in% organisms.list)
     {
       tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
     }
     return(tips_to_keep.tresp)
   })
   
   tips_to_keep.picso3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.picso <- c()
     if ("picso" %in% organisms.list)
     {
       tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
     }
     return(tips_to_keep.picso)
   })
   
   tips_to_keep.chlso3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.chlso <- c()
     if ("chlso" %in% organisms.list)
     {
       tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
     }
     return(tips_to_keep.chlso)
   })
   
   tips_to_keep.auxpr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.auxpr <- c()
     if ("auxpr" %in% organisms.list)
     {
       tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
     }
     return(tips_to_keep.auxpr)
   })
   
   tips_to_keep.helsp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.helsp <- c()
     if ("helsp" %in% organisms.list)
     {
       tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
     }
     return(tips_to_keep.helsp)
   })
   
   tips_to_keep.tetst3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.tetst <- c()
     if ("tetst" %in% organisms.list)
     {
       tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
     }
     return(tips_to_keep.tetst)
   })
   
   tips_to_keep.pedsp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.pedsp <- c()
     if ("pedsp" %in% organisms.list)
     {
       tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
     }
     return(tips_to_keep.pedsp)
   })
   
   tips_to_keep.chlpr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.chlpr <- c()
     if ("chlpr" %in% organisms.list)
     {
       tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
     }
     return(tips_to_keep.chlpr)
   })
   
   tips_to_keep.picsp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.picsp <- c()
     if ("picsp" %in% organisms.list)
     {
       tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
     }
     return(tips_to_keep.picsp)
   })
   
   tips_to_keep.mesvb3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.mesvb <- c()
     if ("mesvb" %in% organisms.list)
     {
       tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
     }
     return(tips_to_keep.mesvb)
   })
   
   tips_to_keep.meskr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.meskr <- c()
     if ("meskr" %in% organisms.list)
     {
       tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
     }
     return(tips_to_keep.meskr)
   })
   
   tips_to_keep.zygci3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.zygci <- c()
     if ("zygci" %in% organisms.list)
     {
       tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
     }
     return(tips_to_keep.zygci)
   })
   
   tips_to_keep.zygcb3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.zygcb <- c()
     if ("zygcb" %in% organisms.list)
     {
       tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
     }
     return(tips_to_keep.zygcb)
   })
   
   tips_to_keep.zygcy3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.zygcy <- c()
     if ("zygcy" %in% organisms.list)
     {
       tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
     }
     return(tips_to_keep.zygcy)
   })
   
   tips_to_keep.plesc3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.plesc <- c()
     if ("plesc" %in% organisms.list)
     {
       tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
     }
     return(tips_to_keep.plesc)
   })
   
   tips_to_keep.funhy3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.funhy <- c()
     if ("funhy" %in% organisms.list)
     {
       tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
     }
     return(tips_to_keep.funhy)
   })
   
   tips_to_keep.sphfa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sphfa <- c()
     if ("sphfa" %in% organisms.list)
     {
       tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
     }
     return(tips_to_keep.sphfa)
   })
   
   tips_to_keep.takle3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.takle <- c()
     if ("takle" %in% organisms.list)
     {
       tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
     }
     return(tips_to_keep.takle)
   })
   
   tips_to_keep.marpa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.marpa <- c()
     if ("marpa" %in% organisms.list)
     {
       tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
     }
     return(tips_to_keep.marpa)
   })
   
   tips_to_keep.ricfl3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ricfl <- c()
     if ("ricfl" %in% organisms.list)
     {
       tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
     }
     return(tips_to_keep.ricfl)
   })
   
   tips_to_keep.ricso3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ricso <- c()
     if ("ricso" %in% organisms.list)
     {
       tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
     }
     return(tips_to_keep.ricso)
   })
   
   tips_to_keep.antpu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.antpu <- c()
     if ("antpu" %in% organisms.list)
     {
       tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
     }
     return(tips_to_keep.antpu)
   })
   
   tips_to_keep.antfu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.antfu <- c()
     if ("antfu" %in% organisms.list)
     {
       tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
     }
     return(tips_to_keep.antfu)
   })
   
   tips_to_keep.megfl3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.megfl <- c()
     if ("megfl" %in% organisms.list)
     {
       tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
     }
     return(tips_to_keep.megfl)
   })
   
   tips_to_keep.phach3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.phach <- c()
     if ("phach" %in% organisms.list)
     {
       tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
     }
     return(tips_to_keep.phach)
   })
   
   tips_to_keep.phyph3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.phyph <- c()
     if ("phyph" %in% organisms.list)
     {
       tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
     }
     return(tips_to_keep.phyph)
   })
   
   tips_to_keep.parpe3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.parpe <- c()
     if ("parpe" %in% organisms.list)
     {
       tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
     }
     return(tips_to_keep.parpe)
   })
   
   tips_to_keep.notor3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.notor <- c()
     if ("notor" %in% organisms.list)
     {
       tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
     }
     return(tips_to_keep.notor)
   })
   
   tips_to_keep.phaca3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.phaca <- c()
     if ("phaca" %in% organisms.list)
     {
       tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
     }
     return(tips_to_keep.phaca)
   })
   
   tips_to_keep.phasp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.phasp <- c()
     if ("phasp" %in% organisms.list)
     {
       tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
     }
     return(tips_to_keep.phasp)
   })
   
   tips_to_keep.leidu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.leidu <- c()
     if ("leidu" %in% organisms.list)
     {
       tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
     }
     return(tips_to_keep.leidu)
   })
   
   tips_to_keep.isosi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.isosi <- c()
     if ("isosi" %in% organisms.list)
     {
       tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
     }
     return(tips_to_keep.isosi)
   })
   
   tips_to_keep.dipco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.dipco <- c()
     if ("dipco" %in% organisms.list)
     {
       tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
     }
     return(tips_to_keep.dipco)
   })
   
   tips_to_keep.hupas3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.hupas <- c()
     if ("hupas" %in% organisms.list)
     {
       tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
     }
     return(tips_to_keep.hupas)
   })
   
   tips_to_keep.lyccl3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.lyccl <- c()
     if ("lyccl" %in% organisms.list)
     {
       tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
     }
     return(tips_to_keep.lyccl)
   })
   
   tips_to_keep.adica3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.adica <- c()
     if ("adica" %in% organisms.list)
     {
       tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
     }
     return(tips_to_keep.adica)
   })
   
   tips_to_keep.alssp3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.alssp <- c()
     if ("alssp" %in% organisms.list)
     {
       tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
     }
     return(tips_to_keep.alssp)
   })
   
   tips_to_keep.marve3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.marve <- c()
     if ("marve" %in% organisms.list)
     {
       tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
     }
     return(tips_to_keep.marve)
   })
   
   tips_to_keep.seqse3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.seqse <- c()
     if ("seqse" %in% organisms.list)
     {
       tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
     }
     return(tips_to_keep.seqse)
   })
   
   tips_to_keep.seqgi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.seqgi <- c()
     if ("seqgi" %in% organisms.list)
     {
       tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
     }
     return(tips_to_keep.seqgi)
   })
   
   tips_to_keep.taxch3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.taxch <- c()
     if ("taxch" %in% organisms.list)
     {
       tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
     }
     return(tips_to_keep.taxch)
   })
   
   tips_to_keep.abial3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.abial <- c()
     if ("abial" %in% organisms.list)
     {
       tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
     }
     return(tips_to_keep.abial)
   })
   
   tips_to_keep.picab3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.picab <- c()
     if ("picab" %in% organisms.list)
     {
       tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
     }
     return(tips_to_keep.picab)
   })
   
   tips_to_keep.gnemo3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.gnemo <- c()
     if ("gnemo" %in% organisms.list)
     {
       tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
     }
     return(tips_to_keep.gnemo)
   })
   
   tips_to_keep.welmi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.welmi <- c()
     if ("welmi" %in% organisms.list)
     {
       tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
     }
     return(tips_to_keep.welmi)
   })
   
   tips_to_keep.ginbi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ginbi <- c()
     if ("ginbi" %in% organisms.list)
     {
       tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
     }
     return(tips_to_keep.ginbi)
   })
   
   tips_to_keep.ambtr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.ambtr <- c()
     if ("ambtr" %in% organisms.list)
     {
       tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
     }
     return(tips_to_keep.ambtr)
   })
   
   tips_to_keep.nymco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.nymco <- c()
     if ("nymco" %in% organisms.list)
     {
       tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
     }
     return(tips_to_keep.nymco)
   })
   
   tips_to_keep.cinka3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cinka <- c()
     if ("cinka" %in% organisms.list)
     {
       tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
     }
     return(tips_to_keep.cinka)
   })
   
   tips_to_keep.horvu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.horvu <- c()
     if ("horvu" %in% organisms.list)
     {
       tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
     }
     return(tips_to_keep.horvu)
   })
   
   tips_to_keep.bradi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.bradi <- c()
     if ("bradi" %in% organisms.list)
     {
       tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
     }
     return(tips_to_keep.bradi)
   })
   
   tips_to_keep.setit3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.setit <- c()
     if ("setit" %in% organisms.list)
     {
       tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
     }
     return(tips_to_keep.setit)
   })
   
   tips_to_keep.panha3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.panha <- c()
     if ("panha" %in% organisms.list)
     {
       tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
     }
     return(tips_to_keep.panha)
   })
   
   tips_to_keep.missi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.missi <- c()
     if ("missi" %in% organisms.list)
     {
       tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
     }
     return(tips_to_keep.missi)
   })
   
   tips_to_keep.oroth3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.oroth <- c()
     if ("oroth" %in% organisms.list)
     {
       tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
     }
     return(tips_to_keep.oroth)
   })
   
   tips_to_keep.eleco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.eleco <- c()
     if ("eleco" %in% organisms.list)
     {
       tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
     }
     return(tips_to_keep.eleco)
   })
   
   tips_to_keep.anaco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.anaco <- c()
     if ("anaco" %in% organisms.list)
     {
       tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
     }
     return(tips_to_keep.anaco)
   })
   
   tips_to_keep.musac3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.musac <- c()
     if ("musac" %in% organisms.list)
     {
       tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
     }
     return(tips_to_keep.musac)
   })
   
   tips_to_keep.dioal3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.dioal <- c()
     if ("dioal" %in% organisms.list)
     {
       tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
     }
     return(tips_to_keep.dioal)
   })
   
   tips_to_keep.spipo3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.spipo <- c()
     if ("spipo" %in% organisms.list)
     {
       tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
     }
     return(tips_to_keep.spipo)
   })
   
   tips_to_keep.aspof3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.aspof <- c()
     if ("aspof" %in% organisms.list)
     {
       tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
     }
     return(tips_to_keep.aspof)
   })
   
   tips_to_keep.zosma3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.zosma <- c()
     if ("zosma" %in% organisms.list)
     {
       tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
     }
     return(tips_to_keep.zosma)
   })
   
   tips_to_keep.araly3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.araly <- c()
     if ("araly" %in% organisms.list)
     {
       tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
     }
     return(tips_to_keep.araly)
   })
   
   tips_to_keep.capgr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.capgr <- c()
     if ("capgr" %in% organisms.list)
     {
       tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
     }
     return(tips_to_keep.capgr)
   })
   
   tips_to_keep.brara3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.brara <- c()
     if ("brara" %in% organisms.list)
     {
       tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
     }
     return(tips_to_keep.brara)
   })
   
   tips_to_keep.braol3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.braol <- c()
     if ("braol" %in% organisms.list)
     {
       tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
     }
     return(tips_to_keep.braol)
   })
   
   tips_to_keep.sisir3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sisir <- c()
     if ("sisir" %in% organisms.list)
     {
       tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
     }
     return(tips_to_keep.sisir)
   })
   
   tips_to_keep.schpa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.schpa <- c()
     if ("schpa" %in% organisms.list)
     {
       tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
     }
     return(tips_to_keep.schpa)
   })
   
   tips_to_keep.eutsa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.eutsa <- c()
     if ("eutsa" %in% organisms.list)
     {
       tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
     }
     return(tips_to_keep.eutsa)
   })
   
   tips_to_keep.thlar3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.thlar <- c()
     if ("thlar" %in% organisms.list)
     {
       tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
     }
     return(tips_to_keep.thlar)
   })
   
   tips_to_keep.boest3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.boest <- c()
     if ("boest" %in% organisms.list)
     {
       tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
     }
     return(tips_to_keep.boest)
   })
   
   tips_to_keep.carpa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.carpa <- c()
     if ("carpa" %in% organisms.list)
     {
       tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
     }
     return(tips_to_keep.carpa)
   })
   
   tips_to_keep.theca3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.theca <- c()
     if ("theca" %in% organisms.list)
     {
       tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
     }
     return(tips_to_keep.theca)
   })
   
   tips_to_keep.goshi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.goshi <- c()
     if ("goshi" %in% organisms.list)
     {
       tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
     }
     return(tips_to_keep.goshi)
   })
   
   tips_to_keep.gosra3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.gosra <- c()
     if ("gosra" %in% organisms.list)
     {
       tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
     }
     return(tips_to_keep.gosra)
   })
   
   tips_to_keep.citsi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.citsi <- c()
     if ("citsi" %in% organisms.list)
     {
       tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
     }
     return(tips_to_keep.citsi)
   })
   
   tips_to_keep.pontr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.pontr <- c()
     if ("pontr" %in% organisms.list)
     {
       tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
     }
     return(tips_to_keep.pontr)
   })
   
   tips_to_keep.eucgr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.eucgr <- c()
     if ("eucgr" %in% organisms.list)
     {
       tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
     }
     return(tips_to_keep.eucgr)
   })
   
   tips_to_keep.maldo3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.maldo <- c()
     if ("maldo" %in% organisms.list)
     {
       tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
     }
     return(tips_to_keep.maldo)
   })
   
   tips_to_keep.prupe3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.prupe <- c()
     if ("prupe" %in% organisms.list)
     {
       tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
     }
     return(tips_to_keep.prupe)
   })
   
   tips_to_keep.frave3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.frave <- c()
     if ("frave" %in% organisms.list)
     {
       tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
     }
     return(tips_to_keep.frave)
   })
   
   tips_to_keep.corav3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.corav <- c()
     if ("corav" %in% organisms.list)
     {
       tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
     }
     return(tips_to_keep.corav)
   })
   
   tips_to_keep.caril3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.caril <- c()
     if ("caril" %in% organisms.list)
     {
       tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
     }
     return(tips_to_keep.caril)
   })
   
   tips_to_keep.queru3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.queru <- c()
     if ("queru" %in% organisms.list)
     {
       tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
     }
     return(tips_to_keep.queru)
   })
   
   tips_to_keep.cucsa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cucsa <- c()
     if ("cucsa" %in% organisms.list)
     {
       tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
     }
     return(tips_to_keep.cucsa)
   })
   
   tips_to_keep.glyma3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.glyma <- c()
     if ("glyma" %in% organisms.list)
     {
       tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
     }
     return(tips_to_keep.glyma)
   })
   
   tips_to_keep.medtr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.medtr <- c()
     if ("medtr" %in% organisms.list)
     {
       tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
     }
     return(tips_to_keep.medtr)
   })
   
   tips_to_keep.tripr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.tripr <- c()
     if ("tripr" %in% organisms.list)
     {
       tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
     }
     return(tips_to_keep.tripr)
   })
   
   tips_to_keep.cicar3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cicar <- c()
     if ("cicar" %in% organisms.list)
     {
       tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
     }
     return(tips_to_keep.cicar)
   })
   
   tips_to_keep.lotja3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.lotja <- c()
     if ("lotja" %in% organisms.list)
     {
       tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
     }
     return(tips_to_keep.lotja)
   })
   
   tips_to_keep.phaac3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.phaac <- c()
     if ("phaac" %in% organisms.list)
     {
       tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
     }
     return(tips_to_keep.phaac)
   })
   
   tips_to_keep.vigun3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.vigun <- c()
     if ("vigun" %in% organisms.list)
     {
       tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
     }
     return(tips_to_keep.vigun)
   })
   
   tips_to_keep.lupal3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.lupal <- c()
     if ("lupal" %in% organisms.list)
     {
       tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
     }
     return(tips_to_keep.lupal)
   })
   
   tips_to_keep.lencu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.lencu <- c()
     if ("lencu" %in% organisms.list)
     {
       tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
     }
     return(tips_to_keep.lencu)
   })
   
   tips_to_keep.arahy3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.arahy <- c()
     if ("arahy" %in% organisms.list)
     {
       tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
     }
     return(tips_to_keep.arahy)
   })
   
   tips_to_keep.poptr3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.poptr <- c()
     if ("poptr" %in% organisms.list)
     {
       tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
     }
     return(tips_to_keep.poptr)
   })
   
   tips_to_keep.manes3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.manes <- c()
     if ("manes" %in% organisms.list)
     {
       tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
     }
     return(tips_to_keep.manes)
   })
   
   tips_to_keep.salpu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.salpu <- c()
     if ("salpu" %in% organisms.list)
     {
       tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
     }
     return(tips_to_keep.salpu)
   })
   
   tips_to_keep.linus3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.linus <- c()
     if ("linus" %in% organisms.list)
     {
       tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
     }
     return(tips_to_keep.linus)
   })
   
   tips_to_keep.corci3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.corci <- c()
     if ("corci" %in% organisms.list)
     {
       tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
     }
     return(tips_to_keep.corci)
   })
   
   tips_to_keep.vitvi3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.vitvi <- c()
     if ("vitvi" %in% organisms.list)
     {
       tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
     }
     return(tips_to_keep.vitvi)
   })
   
   tips_to_keep.kalfe3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.kalfe <- c()
     if ("kalfe" %in% organisms.list)
     {
       tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
     }
     return(tips_to_keep.kalfe)
   })
   
   tips_to_keep.vacda3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.vacda <- c()
     if ("vacda" %in% organisms.list)
     {
       tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
     }
     return(tips_to_keep.vacda)
   })
   
   tips_to_keep.oleeu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.oleeu <- c()
     if ("oleeu" %in% organisms.list)
     {
       tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
     }
     return(tips_to_keep.oleeu)
   })
   
   tips_to_keep.mimgu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.mimgu <- c()
     if ("mimgu" %in% organisms.list)
     {
       tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
     }
     return(tips_to_keep.mimgu)
   })
   
   tips_to_keep.soltu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.soltu <- c()
     if ("soltu" %in% organisms.list)
     {
       tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
     }
     return(tips_to_keep.soltu)
   })
   
   tips_to_keep.cofar3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.cofar <- c()
     if ("cofar" %in% organisms.list)
     {
       tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
     }
     return(tips_to_keep.cofar)
   })
   
   tips_to_keep.lacsa3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.lacsa <- c()
     if ("lacsa" %in% organisms.list)
     {
       tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
     }
     return(tips_to_keep.lacsa)
   })
   
   tips_to_keep.dauca3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.dauca <- c()
     if ("dauca" %in% organisms.list)
     {
       tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
     }
     return(tips_to_keep.dauca)
   })
   
   tips_to_keep.betvu3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.betvu <- c()
     if ("betvu" %in% organisms.list)
     {
       tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
     }
     return(tips_to_keep.betvu)
   })
   
   tips_to_keep.amahy3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.amahy <- c()
     if ("amahy" %in% organisms.list)
     {
       tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
     }
     return(tips_to_keep.amahy)
   })
   
   tips_to_keep.chequ3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.chequ <- c()
     if ("chequ" %in% organisms.list)
     {
       tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
     }
     return(tips_to_keep.chequ)
   })
   
   tips_to_keep.sapof3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.sapof <- c()
     if ("sapof" %in% organisms.list)
     {
       tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
     }
     return(tips_to_keep.sapof)
   })
   
   tips_to_keep.aquco3 <- reactive({
     
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     tips_to_keep.aquco <- c()
     if ("aquco" %in% organisms.list)
     {
       tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
     }
     return(tips_to_keep.aquco)
   }) %>% bindEvent(input$run_button3)
   
   
   # Create complete gene tree with the proper name for each gene
   # For this, we split the species name apart from the gene name
   tree_adj3 <- reactive({
     tree <- tree3()
     organisms.list <- c(selected_values_org3())
     
     if ("mp" %in% organisms.list)
     {
       tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label) 
       if (length(tips_to_keep.mp) != 0)
       {
         mp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mp] <- mp.v
       }
     }
     
     if ("ot" %in% organisms.list)
     {
       tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label) 
       if (length(tips_to_keep.ot) != 0)
       {
         ost.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ot]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ot] <- ost.v
       }
     }
     
     if ("at" %in% organisms.list)
     {
       tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label) 
       if (length(tips_to_keep.at) != 0)
       {
         arabi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.at]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.at] <- arabi.v
       }
     }
     
     if ("cp" %in% organisms.list)
     {
       tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label) 
       if (length(tips_to_keep.cp) != 0)
       {
         cer.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cp]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cp] <- cer.v
       }
     }
     
     if ("cr" %in% organisms.list)
     {
       tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
       if (length(tips_to_keep.cr) != 0)
       {
         chlamy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cr]), "_"), 
                            function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cr] <- chlamy.v
       }
     }
     
     if ("cz" %in% organisms.list)
     {
       tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label) 
       if (length(tips_to_keep.cz) != 0)
       {
         chromo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cz]), "_"), 
                            function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cz] <- chromo.v
       }
     }
     
     if ("kn" %in% organisms.list)
     {
       tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label) 
       if (length(tips_to_keep.kn) != 0)
       {
         klebs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kn]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.kn] <- klebs.v
       }
     }
     
     if ("me" %in% organisms.list)
     {
       tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label) 
       if (length(tips_to_keep.me) != 0)
       {
         meso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.me]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.me] <- meso.v
       }
     }
     
     if ("mi" %in% organisms.list)
     {
       tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label) 
       if (length(tips_to_keep.mi) != 0)
       {
         micro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mi]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mi] <- micro.v
       }
     }
     
     if ("pp" %in% organisms.list)
     {
       tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label) 
       if (length(tips_to_keep.pp) != 0)
       {
         phys.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pp]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pp] <- phys.v
       }
     }
     
     if ("sl" %in% organisms.list)
     {
       tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label) 
       if (length(tips_to_keep.sl) != 0)
       {
         sola.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sl]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sl] <- sola.v
       }
     }
     
     if ("sm" %in% organisms.list)
     {
       tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label) 
       if (length(tips_to_keep.sm) != 0)
       {
         sel.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sm]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sm] <- sel.v
       }
     }
     
     if ("sp" %in% organisms.list)
     {
       tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label) 
       if (length(tips_to_keep.sp) != 0)
       {
         spiro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sp]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sp] <- spiro.v
       }
     }
     
     if ("ta" %in% organisms.list)
     {
       tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label) 
       if (length(tips_to_keep.ta) != 0)
       {
         tri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ta]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ta] <- tri.v
       }
     }
     
     if ("vc" %in% organisms.list)
     {
       tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
       if (length(tips_to_keep.vc) != 0)
       {
         vc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vc]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vc] <- vc.v
       }
     }
     
     if ("bp" %in% organisms.list)
     {
       tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
       if (length(tips_to_keep.bp) != 0)
       {
         bp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.bp] <- bp.v
       }
     }
     
     if ("cri" %in% organisms.list)
     {
       tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
       if (length(tips_to_keep.cri) != 0)
       {
         cri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cri]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cri] <- cri.v
       }
     }
     
     if ("ds" %in% organisms.list)
     {
       tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
       if (length(tips_to_keep.ds) != 0)
       {
         ds.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ds]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ds] <- ds.v
       }
     }
     
     if ("os" %in% organisms.list)
     {
       tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
       if (length(tips_to_keep.os) != 0)
       {
         os.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.os]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.os] <- os.v
       }
     }
     
     if ("smag" %in% organisms.list)
     {
       tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
       if (length(tips_to_keep.smag) != 0)
       {
         smag.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.smag]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.smag] <- smag.v
       }
     }
     
     if ("tp" %in% organisms.list)
     {
       tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
       if (length(tips_to_keep.tp) != 0)
       {
         tp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tp] <- tp.v
       }
     }
     
     if ("aa" %in% organisms.list)
     {
       tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
       if (length(tips_to_keep.aa) != 0)
       {
         aa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aa]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aa] <- aa.v
       }
     }
     
     if ("um" %in% organisms.list)
     {
       tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
       if (length(tips_to_keep.um) != 0)
       {
         um.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.um]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.um] <- um.v
       }
     }
     
     if ("rs" %in% organisms.list)
     {
       tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
       if (length(tips_to_keep.rs) != 0)
       {
         rs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.rs]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.rs] <- rs.v
       }
     }
     
     if ("cyc" %in% organisms.list)
     {
       tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
       if (length(tips_to_keep.cyc) != 0)
       {
         cyc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cyc]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cyc] <- cyc.v
       }
     }
     
     if ("ca" %in% organisms.list)
     {
       tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
       if (length(tips_to_keep.ca) != 0)
       {
         ca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ca]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ca] <- ca.v
       }
     }
     
     if ("mv" %in% organisms.list)
     {
       tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
       if (length(tips_to_keep.mv) != 0)
       {
         mv.vec3 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mv]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mv] <- mv.vec3
       }
     }
     
     if ("af" %in% organisms.list)
     {
       tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
       if (length(tips_to_keep.af) != 0)
       {
         af.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.af]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.af] <- af.v
       }
     }
     
     if ("sc" %in% organisms.list)
     {
       tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
       if (length(tips_to_keep.sc) != 0)
       {
         sc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sc]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sc] <- sc.v
       }
     }
     
     if ("aegi" %in% organisms.list)
     {
       tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
       if (length(tips_to_keep.aegi) != 0)
       {
         aegi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aegi]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aegi] <- aegi.v
       }
     }
     
     if ("sb" %in% organisms.list)
     {
       tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
       if (length(tips_to_keep.sb) != 0)
       {
         sb.vec3 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sb]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sb] <- sb.vec3
       }
     }
     
     if ("chara" %in% organisms.list)
     {
       tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
       if (length(tips_to_keep.chara) != 0)
       {
         chara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chara]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chara] <- chara.v
       }
     }
     
     if ("sceobli" %in% organisms.list)
     {
       tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
       if (length(tips_to_keep.sceobli) != 0)
       {
         sceobli.vec3 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sceobli]), "_"), 
                                function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sceobli] <- sceobli.vec3
       }
     }
     
     if ("cocco" %in% organisms.list)
     {
       tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
       if (length(tips_to_keep.cocco) != 0)
       {
         cocco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cocco]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cocco] <- cocco.v
       }
     }
     
     if ("haema" %in% organisms.list)
     {
       tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
       if (length(tips_to_keep.haema) != 0)
       {
         haema.vec3 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.haema]), "_"), 
                              function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.haema] <- haema.vec3
       }
     }
     
     if ("zm" %in% organisms.list)
     {
       tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
       if (length(tips_to_keep.zm) != 0)
       {
         zm.vec3 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zm]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zm] <- zm.vec3
       }
     }
     
     
     if ("praco" %in% organisms.list)
     {
       tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
       if (length(tips_to_keep.praco) != 0)
       {
         praco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.praco]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.praco] <- praco.v
       }
     }
     
     if ("ostlu" %in% organisms.list)
     {
       tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
       if (length(tips_to_keep.ostlu) != 0)
       {
         ostlu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostlu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostlu] <- ostlu.v
       }
     }
     
     if ("ostme" %in% organisms.list)
     {
       tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
       if (length(tips_to_keep.ostme) != 0)
       {
         ostme.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostme]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostme] <- ostme.v
       }
     }
     
     if ("micco" %in% organisms.list)
     {
       tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
       if (length(tips_to_keep.micco) != 0)
       {
         micco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.micco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.micco] <- micco.v
       }
     }
     
     if ("cymte" %in% organisms.list)
     {
       tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
       if (length(tips_to_keep.cymte) != 0)
       {
         cymte.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cymte]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cymte] <- cymte.v
       }
     }
     
     if ("chlin" %in% organisms.list)
     {
       tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
       if (length(tips_to_keep.chlin) != 0)
       {
         chlin.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlin]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlin] <- chlin.v
       }
     }
     
     if ("gonpe" %in% organisms.list)
     {
       tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
       if (length(tips_to_keep.gonpe) != 0)
       {
         gonpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gonpe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gonpe] <- gonpe.v
       }
     }
     
     if ("tetso" %in% organisms.list)
     {
       tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
       if (length(tips_to_keep.tetso) != 0)
       {
         tetso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tetso] <- tetso.v
       }
     }
     
     if ("desar" %in% organisms.list)
     {
       tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
       if (length(tips_to_keep.desar) != 0)
       {
         desar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.desar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.desar] <- desar.v
       }
     }
     
     if ("monmi" %in% organisms.list)
     {
       tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
       if (length(tips_to_keep.monmi) != 0)
       {
         monmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.monmi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.monmi] <- monmi.v
       }
     }
     
     if ("caule" %in% organisms.list)
     {
       tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
       if (length(tips_to_keep.caule) != 0)
       {
         caule.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caule]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.caule] <- caule.v
       }
     }
     
     if ("ostqu" %in% organisms.list)
     {
       tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
       if (length(tips_to_keep.ostqu) != 0)
       {
         ostqu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostqu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostqu] <- ostqu.v
       }
     }
     
     if ("brysp" %in% organisms.list)
     {
       tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
       if (length(tips_to_keep.brysp) != 0)
       {
         brysp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brysp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.brysp] <- brysp.v
       }
     }
     
     if ("ellbi" %in% organisms.list)
     {
       tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
       if (length(tips_to_keep.ellbi) != 0)
       {
         ellbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ellbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ellbi] <- ellbi.v
       }
     }
     
     if ("myrbi" %in% organisms.list)
     {
       tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
       if (length(tips_to_keep.myrbi) != 0)
       {
         myrbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.myrbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.myrbi] <- myrbi.v
       }
     }
     
     if ("apalo" %in% organisms.list)
     {
       tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
       if (length(tips_to_keep.apalo) != 0)
       {
         apalo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.apalo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.apalo] <- apalo.v
       }
     }
     
     if ("astgl" %in% organisms.list)
     {
       tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
       if (length(tips_to_keep.astgl) != 0)
       {
         astgl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.astgl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.astgl] <- astgl.v
       }
     }
     
     if ("tresp" %in% organisms.list)
     {
       tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
       if (length(tips_to_keep.tresp) != 0)
       {
         tresp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tresp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tresp] <- tresp.v
       }
     }
     
     if ("picso" %in% organisms.list)
     {
       tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
       if (length(tips_to_keep.picso) != 0)
       {
         picso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picso] <- picso.v
       }
     }
     
     if ("chlso" %in% organisms.list)
     {
       tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
       if (length(tips_to_keep.chlso) != 0)
       {
         chlso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlso] <- chlso.v
       }
     }
     
     if ("auxpr" %in% organisms.list)
     {
       tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
       if (length(tips_to_keep.auxpr) != 0)
       {
         auxpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.auxpr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.auxpr] <- auxpr.v
       }
     }
     
     if ("helsp" %in% organisms.list)
     {
       tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
       if (length(tips_to_keep.helsp) != 0)
       {
         helsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.helsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.helsp] <- helsp.v
       }
     }
     
     if ("tetst" %in% organisms.list)
     {
       tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
       if (length(tips_to_keep.tetst) != 0)
       {
         tetst.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetst]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tetst] <- tetst.v
       }
     }
     
     if ("pedsp" %in% organisms.list)
     {
       tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
       if (length(tips_to_keep.pedsp) != 0)
       {
         pedsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pedsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pedsp] <- pedsp.v
       }
     }
     
     if ("chlpr" %in% organisms.list)
     {
       tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
       if (length(tips_to_keep.chlpr) != 0)
       {
         chlpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlpr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlpr] <- chlpr.v
       }
     }
     
     if ("picsp" %in% organisms.list)
     {
       tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
       if (length(tips_to_keep.picsp) != 0)
       {
         picsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picsp] <- picsp.v
       }
     }
     
     if ("mesvb" %in% organisms.list)
     {
       tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
       if (length(tips_to_keep.mesvb) != 0)
       {
         mesvb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mesvb]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mesvb] <- mesvb.v
       }
     }
     
     if ("meskr" %in% organisms.list)
     {
       tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
       if (length(tips_to_keep.meskr) != 0)
       {
         meskr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.meskr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.meskr] <- meskr.v
       }
     }
     
     if ("zygci" %in% organisms.list)
     {
       tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
       if (length(tips_to_keep.zygci) != 0)
       {
         zygci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygci]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygci] <- zygci.v
       }
     }
     
     if ("zygcb" %in% organisms.list)
     {
       tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
       if (length(tips_to_keep.zygcb) != 0)
       {
         zygcb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcb]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygcb] <- zygcb.v
       }
     }
     
     if ("zygcy" %in% organisms.list)
     {
       tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
       if (length(tips_to_keep.zygcy) != 0)
       {
         zygcy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygcy] <- zygcy.v
       }
     }
     
     if ("plesc" %in% organisms.list)
     {
       tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
       if (length(tips_to_keep.plesc) != 0)
       {
         plesc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.plesc]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.plesc] <- plesc.v
       }
     }
     
     if ("funhy" %in% organisms.list)
     {
       tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
       if (length(tips_to_keep.funhy) != 0)
       {
         funhy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.funhy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.funhy] <- funhy.v
       }
     }
     
     if ("sphfa" %in% organisms.list)
     {
       tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
       if (length(tips_to_keep.sphfa) != 0)
       {
         sphfa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sphfa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sphfa] <- sphfa.v
       }
     }
     
     if ("takle" %in% organisms.list)
     {
       tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
       if (length(tips_to_keep.takle) != 0)
       {
         takle.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.takle]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.takle] <- takle.v
       }
     }
     
     if ("marpa" %in% organisms.list)
     {
       tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
       if (length(tips_to_keep.marpa) != 0)
       {
         marpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.marpa] <- marpa.v
       }
     }
     
     if ("ricfl" %in% organisms.list)
     {
       tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
       if (length(tips_to_keep.ricfl) != 0)
       {
         ricfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricfl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ricfl] <- ricfl.v
       }
     }
     
     if ("ricso" %in% organisms.list)
     {
       tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
       if (length(tips_to_keep.ricso) != 0)
       {
         ricso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ricso] <- ricso.v
       }
     }
     
     if ("antpu" %in% organisms.list)
     {
       tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
       if (length(tips_to_keep.antpu) != 0)
       {
         antpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antpu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.antpu] <- antpu.v
       }
     }
     
     if ("antfu" %in% organisms.list)
     {
       tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
       if (length(tips_to_keep.antfu) != 0)
       {
         antfu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antfu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.antfu] <- antfu.v
       }
     }
     
     if ("megfl" %in% organisms.list)
     {
       tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
       if (length(tips_to_keep.megfl) != 0)
       {
         megfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.megfl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.megfl] <- megfl.v
       }
     }
     
     if ("phach" %in% organisms.list)
     {
       tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
       if (length(tips_to_keep.phach) != 0)
       {
         phach.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phach]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phach] <- phach.v
       }
     }
     
     if ("phyph" %in% organisms.list)
     {
       tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
       if (length(tips_to_keep.phyph) != 0)
       {
         phyph.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phyph]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phyph] <- phyph.v
       }
     }
     
     if ("parpe" %in% organisms.list)
     {
       tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
       if (length(tips_to_keep.parpe) != 0)
       {
         parpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.parpe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.parpe] <- parpe.v
       }
     }
     
     if ("notor" %in% organisms.list)
     {
       tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
       if (length(tips_to_keep.notor) != 0)
       {
         notor.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.notor]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.notor] <- notor.v
       }
     }
     
     if ("phaca" %in% organisms.list)
     {
       tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
       if (length(tips_to_keep.phaca) != 0)
       {
         phaca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phaca] <- phaca.v
       }
     }
     
     if ("phasp" %in% organisms.list)
     {
       tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
       if (length(tips_to_keep.phasp) != 0)
       {
         phasp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phasp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phasp] <- phasp.v
       }
     }
     
     if ("leidu" %in% organisms.list)
     {
       tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
       if (length(tips_to_keep.leidu) != 0)
       {
         leidu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.leidu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.leidu] <- leidu.v
       }
     }
     
     if ("isosi" %in% organisms.list)
     {
       tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
       if (length(tips_to_keep.isosi) != 0)
       {
         isosi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.isosi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.isosi] <- isosi.v
       }
     }
     
     if ("dipco" %in% organisms.list)
     {
       tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
       if (length(tips_to_keep.dipco) != 0)
       {
         dipco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dipco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dipco] <- dipco.v
       }
     }
     
     if ("hupas" %in% organisms.list)
     {
       tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
       if (length(tips_to_keep.hupas) != 0)
       {
         hupas.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.hupas]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.hupas] <- hupas.v
       }
     }
     
     if ("lyccl" %in% organisms.list)
     {
       tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
       if (length(tips_to_keep.lyccl) != 0)
       {
         lyccl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lyccl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lyccl] <- lyccl.v
       }
     }
     
     if ("adica" %in% organisms.list)
     {
       tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
       if (length(tips_to_keep.adica) != 0)
       {
         adica.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.adica]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.adica] <- adica.v
       }
     }
     
     if ("alssp" %in% organisms.list)
     {
       tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
       if (length(tips_to_keep.alssp) != 0)
       {
         alssp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.alssp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.alssp] <- alssp.v
       }
     }
     
     if ("marve" %in% organisms.list)
     {
       tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
       if (length(tips_to_keep.marve) != 0)
       {
         marve.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marve]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.marve] <- marve.v
       }
     }
     
     if ("seqse" %in% organisms.list)
     {
       tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
       if (length(tips_to_keep.seqse) != 0)
       {
         seqse.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqse]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.seqse] <- seqse.v
       }
     }
     
     if ("seqgi" %in% organisms.list)
     {
       tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
       if (length(tips_to_keep.seqgi) != 0)
       {
         seqgi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqgi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.seqgi] <- seqgi.v
       }
     }
     
     if ("taxch" %in% organisms.list)
     {
       tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
       if (length(tips_to_keep.taxch) != 0)
       {
         taxch.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.taxch]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.taxch] <- taxch.v
       }
     }
     
     if ("abial" %in% organisms.list)
     {
       tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
       if (length(tips_to_keep.abial) != 0)
       {
         abial.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.abial]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.abial] <- abial.v
       }
     }
     
     if ("picab" %in% organisms.list)
     {
       tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
       if (length(tips_to_keep.picab) != 0)
       {
         picab.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picab]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picab] <- picab.v
       }
     }
     
     if ("gnemo" %in% organisms.list)
     {
       tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
       if (length(tips_to_keep.gnemo) != 0)
       {
         gnemo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gnemo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gnemo] <- gnemo.v
       }
     }
     
     if ("welmi" %in% organisms.list)
     {
       tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
       if (length(tips_to_keep.welmi) != 0)
       {
         welmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.welmi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.welmi] <- welmi.v
       }
     }
     
     if ("ginbi" %in% organisms.list)
     {
       tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
       if (length(tips_to_keep.ginbi) != 0)
       {
         ginbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ginbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ginbi] <- ginbi.v
       }
     }
     
     if ("ambtr" %in% organisms.list)
     {
       tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
       if (length(tips_to_keep.ambtr) != 0)
       {
         ambtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ambtr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ambtr] <- ambtr.v
       }
     }
     
     if ("nymco" %in% organisms.list)
     {
       tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
       if (length(tips_to_keep.nymco) != 0)
       {
         nymco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.nymco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.nymco] <- nymco.v
       }
     }
     
     if ("cinka" %in% organisms.list)
     {
       tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
       if (length(tips_to_keep.cinka) != 0)
       {
         cinka.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cinka]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cinka] <- cinka.v
       }
     }
     
     
     if ("horvu" %in% organisms.list)
     {
       tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
       if (length(tips_to_keep.horvu) != 0)
       {
         horvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.horvu]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.horvu] <- horvu.v
       }
     }
     
     if ("bradi" %in% organisms.list)
     {
       tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
       if (length(tips_to_keep.bradi) != 0)
       {
         bradi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bradi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.bradi] <- bradi.v
       }
     }
     
     if ("setit" %in% organisms.list)
     {
       tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
       if (length(tips_to_keep.setit) != 0)
       {
         setit.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.setit]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.setit] <- setit.v
       }
     }
     
     if ("panha" %in% organisms.list)
     {
       tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
       if (length(tips_to_keep.panha) != 0)
       {
         panha.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.panha]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.panha] <- panha.v
       }
     }
     
     if ("missi" %in% organisms.list)
     {
       tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
       if (length(tips_to_keep.missi) != 0)
       {
         missi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.missi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.missi] <- missi.v
       }
     }
     
     if ("oroth" %in% organisms.list)
     {
       tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
       if (length(tips_to_keep.oroth) != 0)
       {
         oroth.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oroth]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.oroth] <- oroth.v
       }
     }
     
     if ("eleco" %in% organisms.list)
     {
       tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
       if (length(tips_to_keep.eleco) != 0)
       {
         eleco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eleco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eleco] <- eleco.v
       }
     }
     
     if ("anaco" %in% organisms.list)
     {
       tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
       if (length(tips_to_keep.anaco) != 0)
       {
         anaco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.anaco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.anaco] <- anaco.v
       }
     }
     
     if ("musac" %in% organisms.list)
     {
       tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
       if (length(tips_to_keep.musac) != 0)
       {
         musac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.musac]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.musac] <- musac.v
       }
     }
     
     if ("dioal" %in% organisms.list)
     {
       tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
       if (length(tips_to_keep.dioal) != 0)
       {
         dioal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dioal]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dioal] <- dioal.v
       }
     }
     
     if ("spipo" %in% organisms.list)
     {
       tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
       if (length(tips_to_keep.spipo) != 0)
       {
         spipo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.spipo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.spipo] <- spipo.v
       }
     }
     
     if ("aspof" %in% organisms.list)
     {
       tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
       if (length(tips_to_keep.aspof) != 0)
       {
         aspof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aspof]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aspof] <- aspof.v
       }
     }
     
     if ("zosma" %in% organisms.list)
     {
       tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
       if (length(tips_to_keep.zosma) != 0)
       {
         zosma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zosma]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zosma] <- zosma.v
       }
     }
     
     if ("araly" %in% organisms.list)
     {
       tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
       if (length(tips_to_keep.araly) != 0)
       {
         araly.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.araly]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.araly] <- araly.v
       }
     }
     
     if ("capgr" %in% organisms.list)
     {
       tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
       if (length(tips_to_keep.capgr) != 0)
       {
         capgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.capgr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.capgr] <- capgr.v
       }
     }
     
     if ("brara" %in% organisms.list)
     {
       tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
       if (length(tips_to_keep.brara) != 0)
       {
         brara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brara]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.brara] <- brara.v
       }
     }
     
     if ("braol" %in% organisms.list)
     {
       tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
       if (length(tips_to_keep.braol) != 0)
       {
         braol.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.braol]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.braol] <- braol.v
       }
     }
     
     if ("sisir" %in% organisms.list)
     {
       tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
       if (length(tips_to_keep.sisir) != 0)
       {
         sisir.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sisir]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sisir] <- sisir.v
       }
     }
     
     if ("schpa" %in% organisms.list)
     {
       tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
       if (length(tips_to_keep.schpa) != 0)
       {
         schpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.schpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.schpa] <- schpa.v
       }
     }
     
     if ("eutsa" %in% organisms.list)
     {
       tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
       if (length(tips_to_keep.eutsa) != 0)
       {
         eutsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eutsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eutsa] <- eutsa.v
       }
     }
     
     if ("thlar" %in% organisms.list)
     {
       tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
       if (length(tips_to_keep.thlar) != 0)
       {
         thlar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.thlar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.thlar] <- thlar.v
       }
     }
     
     if ("boest" %in% organisms.list)
     {
       tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
       if (length(tips_to_keep.boest) != 0)
       {
         boest.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.boest]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.boest] <- boest.v
       }
     }
     
     if ("carpa" %in% organisms.list)
     {
       tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
       if (length(tips_to_keep.carpa) != 0)
       {
         carpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.carpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.carpa] <- carpa.v
       }
     }
     
     if ("theca" %in% organisms.list)
     {
       tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
       if (length(tips_to_keep.theca) != 0)
       {
         theca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.theca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.theca] <- theca.v
       }
     }
     
     if ("goshi" %in% organisms.list)
     {
       tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
       if (length(tips_to_keep.goshi) != 0)
       {
         goshi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.goshi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.goshi] <- goshi.v
       }
     }
     
     if ("gosra" %in% organisms.list)
     {
       tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
       if (length(tips_to_keep.gosra) != 0)
       {
         gosra.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gosra]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gosra] <- gosra.v
       }
     }
     
     if ("citsi" %in% organisms.list)
     {
       tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
       if (length(tips_to_keep.citsi) != 0)
       {
         citsi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.citsi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.citsi] <- citsi.v
       }
     }
     
     if ("pontr" %in% organisms.list)
     {
       tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
       if (length(tips_to_keep.pontr) != 0)
       {
         pontr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pontr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pontr] <- pontr.v
       }
     }
     
     if ("eucgr" %in% organisms.list)
     {
       tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
       if (length(tips_to_keep.eucgr) != 0)
       {
         eucgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eucgr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eucgr] <- eucgr.v
       }
     }
     
     if ("maldo" %in% organisms.list)
     {
       tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
       if (length(tips_to_keep.maldo) != 0)
       {
         maldo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.maldo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.maldo] <- maldo.v
       }
     }
     
     if ("prupe" %in% organisms.list)
     {
       tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
       if (length(tips_to_keep.prupe) != 0)
       {
         prupe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.prupe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.prupe] <- prupe.v
       }
     }
     
     if ("frave" %in% organisms.list)
     {
       tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
       if (length(tips_to_keep.frave) != 0)
       {
         frave.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.frave]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.frave] <- frave.v
       }
     }
     
     if ("corav" %in% organisms.list)
     {
       tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
       if (length(tips_to_keep.corav) != 0)
       {
         corav.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corav]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.corav] <- corav.v
       }
     }
     
     if ("caril" %in% organisms.list)
     {
       tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
       if (length(tips_to_keep.caril) != 0)
       {
         caril.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caril]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.caril] <- caril.v
       }
     }
     
     if ("queru" %in% organisms.list)
     {
       tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
       if (length(tips_to_keep.queru) != 0)
       {
         queru.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.queru]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.queru] <- queru.v
       }
     }
     
     if ("cucsa" %in% organisms.list)
     {
       tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
       if (length(tips_to_keep.cucsa) != 0)
       {
         cucsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cucsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cucsa] <- cucsa.v
       }
     }
     
     if ("glyma" %in% organisms.list)
     {
       tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
       if (length(tips_to_keep.glyma) != 0)
       {
         glyma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.glyma]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.glyma] <- glyma.v
       }
     }
     
     if ("medtr" %in% organisms.list)
     {
       tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
       if (length(tips_to_keep.medtr) != 0)
       {
         medtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.medtr]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.medtr] <- medtr.v
       }
     }
     
     if ("tripr" %in% organisms.list)
     {
       tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
       if (length(tips_to_keep.tripr) != 0)
       {
         tripr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tripr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tripr] <- tripr.v
       }
     }
     
     if ("cicar" %in% organisms.list)
     {
       tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
       if (length(tips_to_keep.cicar) != 0)
       {
         cicar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cicar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cicar] <- cicar.v
       }
     }
     
     if ("lotja" %in% organisms.list)
     {
       tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
       if (length(tips_to_keep.lotja) != 0)
       {
         lotja.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lotja]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lotja] <- lotja.v
       }
     }
     
     if ("phaac" %in% organisms.list)
     {
       tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
       if (length(tips_to_keep.phaac) != 0)
       {
         phaac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaac]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phaac] <- phaac.v
       }
     }
     
     if ("vigun" %in% organisms.list)
     {
       tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
       if (length(tips_to_keep.vigun) != 0)
       {
         vigun.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vigun]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vigun] <- vigun.v
       }
     }
     
     if ("lupal" %in% organisms.list)
     {
       tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
       if (length(tips_to_keep.lupal) != 0)
       {
         lupal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lupal]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lupal] <- lupal.v
       }
     }
     
     if ("lencu" %in% organisms.list)
     {
       tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
       if (length(tips_to_keep.lencu) != 0)
       {
         lencu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lencu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lencu] <- lencu.v
       }
     }
     
     if ("arahy" %in% organisms.list)
     {
       tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
       if (length(tips_to_keep.arahy) != 0)
       {
         arahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.arahy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.arahy] <- arahy.v
       }
     }
     
     if ("poptr" %in% organisms.list)
     {
       tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
       if (length(tips_to_keep.poptr) != 0)
       {
         poptr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.poptr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.poptr] <- poptr.v
       }
     }
     
     if ("manes" %in% organisms.list)
     {
       tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
       if (length(tips_to_keep.manes) != 0)
       {
         manes.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.manes]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.manes] <- manes.v
       }
     }
     
     if ("salpu" %in% organisms.list)
     {
       tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
       if (length(tips_to_keep.salpu) != 0)
       {
         salpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.salpu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.salpu] <- salpu.v
       }
     }
     
     if ("linus" %in% organisms.list)
     {
       tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
       if (length(tips_to_keep.linus) != 0)
       {
         linus.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.linus]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.linus] <- linus.v
       }
     }
     
     if ("corci" %in% organisms.list)
     {
       tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
       if (length(tips_to_keep.corci) != 0)
       {
         corci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corci]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.corci] <- corci.v
       }
     }
     
     if ("vitvi" %in% organisms.list)
     {
       tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
       if (length(tips_to_keep.vitvi) != 0)
       {
         vitvi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vitvi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vitvi] <- vitvi.v
       }
     }
     
     if ("kalfe" %in% organisms.list)
     {
       tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
       if (length(tips_to_keep.kalfe) != 0)
       {
         kalfe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kalfe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.kalfe] <- kalfe.v
       }
     }
     
     if ("vacda" %in% organisms.list)
     {
       tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
       if (length(tips_to_keep.vacda) != 0)
       {
         vacda.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vacda]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vacda] <- vacda.v
       }
     }
     
     if ("oleeu" %in% organisms.list)
     {
       tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
       if (length(tips_to_keep.oleeu) != 0)
       {
         oleeu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oleeu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.oleeu] <- oleeu.v
       }
     }
     
     if ("mimgu" %in% organisms.list)
     {
       tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
       if (length(tips_to_keep.mimgu) != 0)
       {
         mimgu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mimgu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mimgu] <- mimgu.v
       }
     }
     
     if ("soltu" %in% organisms.list)
     {
       tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
       if (length(tips_to_keep.soltu) != 0)
       {
         soltu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.soltu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.soltu] <- soltu.v
       }
     }
     
     if ("cofar" %in% organisms.list)
     {
       tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
       if (length(tips_to_keep.cofar) != 0)
       {
         cofar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cofar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cofar] <- cofar.v
       }
     }
     
     if ("lacsa" %in% organisms.list)
     {
       tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
       if (length(tips_to_keep.lacsa) != 0)
       {
         lacsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lacsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lacsa] <- lacsa.v
       }
     }
     
     if ("dauca" %in% organisms.list)
     {
       tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
       if (length(tips_to_keep.dauca) != 0)
       {
         dauca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dauca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dauca] <- dauca.v
       }
     }
     
     if ("betvu" %in% organisms.list)
     {
       tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
       if (length(tips_to_keep.betvu) != 0)
       {
         betvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.betvu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.betvu] <- betvu.v
       }
     }
     
     if ("amahy" %in% organisms.list)
     {
       tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
       if (length(tips_to_keep.amahy) != 0)
       {
         amahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.amahy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.amahy] <- amahy.v
       }
     }
     
     if ("chequ" %in% organisms.list)
     {
       tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
       if (length(tips_to_keep.chequ) != 0)
       {
         chequ.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chequ]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chequ] <- chequ.v
       }
     }
     
     if ("sapof" %in% organisms.list)
     {
       tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
       if (length(tips_to_keep.sapof) != 0)
       {
         sapof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sapof]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sapof] <- sapof.v
       }
     }
     
     if ("aquco" %in% organisms.list)
     {
       tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
       if (length(tips_to_keep.aquco) != 0)
       {
         aquco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aquco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aquco] <- aquco.v
       }
     }
     
     return(tree)
   }) %>% bindEvent(input$run_button3)
   
   # Generate reduced tree when the corresponding button is activated
   tree_reduced3 <- reactive({
     
     tree <- tree_adj3()
     # Define tips to keep (selected organisms) and generate the reduced tree
     tips_to_keep.global <- c(tips_to_keep.praco3(), tips_to_keep.ot3(), tips_to_keep.ostlu3(), tips_to_keep.ostme3(), tips_to_keep.bp3(),
                              tips_to_keep.mi3(), tips_to_keep.micco3(), tips_to_keep.cymte3(), tips_to_keep.cr3(), tips_to_keep.chlin3(),
                              tips_to_keep.haema3(), tips_to_keep.vc3(), tips_to_keep.ds3(), tips_to_keep.gonpe3(), tips_to_keep.tetso3(),
                              tips_to_keep.cz3(), tips_to_keep.rs3(), tips_to_keep.sceobli3(), tips_to_keep.desar3(), tips_to_keep.monmi3(),
                              tips_to_keep.caule3(), tips_to_keep.ostqu3(), tips_to_keep.brysp3(), tips_to_keep.cocco3(),
                              tips_to_keep.ellbi3(), tips_to_keep.myrbi3(), tips_to_keep.apalo3(), tips_to_keep.astgl3(),
                              tips_to_keep.tresp3(), tips_to_keep.picso3(), tips_to_keep.chlso3(), tips_to_keep.auxpr3(),
                              tips_to_keep.helsp3(), tips_to_keep.um3(), tips_to_keep.tetst3(), tips_to_keep.pedsp3(),
                              tips_to_keep.chlpr3(), tips_to_keep.picsp3(), tips_to_keep.ca3(), tips_to_keep.mv3(), tips_to_keep.mesvb3(),
                              tips_to_keep.kn3(), tips_to_keep.chara3(), tips_to_keep.me3(), tips_to_keep.meskr3(), tips_to_keep.sp3(),
                              tips_to_keep.zygci3(), tips_to_keep.zygcb3(), tips_to_keep.zygcy3(),
                              tips_to_keep.cp3(), tips_to_keep.pp3(), tips_to_keep.plesc3(), tips_to_keep.funhy3(),
                              tips_to_keep.smag3(), tips_to_keep.sphfa3(), tips_to_keep.takle3(), tips_to_keep.mp3(),
                              tips_to_keep.marpa3(), tips_to_keep.ricfl3(), tips_to_keep.ricso3(), tips_to_keep.aa3(),
                              tips_to_keep.antpu3(), tips_to_keep.antfu3(), tips_to_keep.megfl3(), tips_to_keep.phach3(),
                              tips_to_keep.phyph3(), tips_to_keep.parpe3(), tips_to_keep.notor3(), tips_to_keep.phaca3(),
                              tips_to_keep.phasp3(), tips_to_keep.leidu3(), tips_to_keep.sm3(), tips_to_keep.isosi3(), 
                              tips_to_keep.dipco3(), tips_to_keep.hupas3(), tips_to_keep.lyccl3(),
                              tips_to_keep.cri3(), tips_to_keep.adica3(), tips_to_keep.alssp3(), tips_to_keep.sc3(), 
                              tips_to_keep.af3(), tips_to_keep.marve3(), tips_to_keep.tp3(), tips_to_keep.seqse3(), 
                              tips_to_keep.seqgi3(), tips_to_keep.taxch3(), tips_to_keep.abial3(), tips_to_keep.picab3(),
                              tips_to_keep.gnemo3(), tips_to_keep.welmi3(), tips_to_keep.cyc3(), tips_to_keep.ginbi3(),
                              tips_to_keep.ambtr3(), tips_to_keep.nymco3(), tips_to_keep.cinka3(), tips_to_keep.aegi3(),
                              tips_to_keep.ta3(), tips_to_keep.horvu3(), tips_to_keep.bradi3(), tips_to_keep.os3(),
                              tips_to_keep.setit3(), tips_to_keep.sb3(), tips_to_keep.zm3(), tips_to_keep.panha3(),
                              tips_to_keep.missi3(), tips_to_keep.oroth3(), tips_to_keep.eleco3(), tips_to_keep.anaco3(),
                              tips_to_keep.musac3(), tips_to_keep.dioal3(), tips_to_keep.spipo3(), tips_to_keep.aspof3(),
                              tips_to_keep.zosma3(), tips_to_keep.at3(), tips_to_keep.araly3(), tips_to_keep.capgr3(),
                              tips_to_keep.brara3(), tips_to_keep.braol3(), tips_to_keep.sisir3(), tips_to_keep.schpa3(),
                              tips_to_keep.eutsa3(), tips_to_keep.thlar3(), tips_to_keep.boest3(), tips_to_keep.carpa3(),
                              tips_to_keep.theca3(), tips_to_keep.goshi3(), tips_to_keep.gosra3(), tips_to_keep.citsi3(),
                              tips_to_keep.pontr3(), tips_to_keep.eucgr3(), tips_to_keep.maldo3(), tips_to_keep.prupe3(),
                              tips_to_keep.frave3(), tips_to_keep.corav3(), tips_to_keep.caril3(), tips_to_keep.queru3(),
                              tips_to_keep.cucsa3(), tips_to_keep.glyma3(), tips_to_keep.medtr3(), tips_to_keep.tripr3(),
                              tips_to_keep.cicar3(), tips_to_keep.lotja3(), tips_to_keep.phaac3(), tips_to_keep.vigun3(),
                              tips_to_keep.lupal3(), tips_to_keep.lencu3(), tips_to_keep.arahy3(), tips_to_keep.poptr3(),
                              tips_to_keep.manes3(), tips_to_keep.salpu3(), tips_to_keep.linus3(), tips_to_keep.corci3(),
                              tips_to_keep.vitvi3(), tips_to_keep.kalfe3(), tips_to_keep.vacda3(), tips_to_keep.oleeu3(),
                              tips_to_keep.mimgu3(), tips_to_keep.sl3(), tips_to_keep.soltu3(), tips_to_keep.cofar3(),
                              tips_to_keep.lacsa3(), tips_to_keep.dauca3(), tips_to_keep.betvu3(), tips_to_keep.amahy3(),
                              tips_to_keep.chequ3(), tips_to_keep.sapof3(), tips_to_keep.aquco3())
     
     # Error message if trying to build tree with less than two tips
     if (length(tips_to_keep.global) < 2)
     {
       shinyjs::hideElement(id = 'loading.tree3')
       
       if (UI_exist_tree3)
       {
         removeUI(
           selector = "div:has(>> #treeTips3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs3)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree3 <<- F
       output$error_tree3 <- renderUI({renderText({print("Unable to construct 
      tree with a single tip, please select more organisms.")})})
       validate(need(length(tips_to_keep.global) > 1, " "))
     }
     
     
     # Error message if query gene does not belong to selected organisms
     
     tips_to_drop <- setdiff(1:length(tree$tip.label), tips_to_keep.global)
     tree_reduced <- drop.tip(tree, tips_to_drop)
     
     return(tree_reduced)
   }) %>% bindEvent(input$run_button3)
   
   ### Select orthogroup sequences based on the reduced tree
   ortho_reduced3 <- reactive({
     
     tree_reduced <- tree_reduced3()
     ortho_seq <- ortho_seq3()
     ortho_reduced <- ortho_seq[tree_reduced$tip.label]
     return(ortho_reduced)
   }) %>% bindEvent(input$run_button3)
   
   organims_reduced3 <- reactive({
     
     tree_reduced <- tree_reduced3()
     
     len.mp <- length(tips_to_keep.mp3())
     len.ot <- length(tips_to_keep.ot3())
     len.at <- length(tips_to_keep.at3())
     len.cp <- length(tips_to_keep.cp3())
     len.cr <- length(tips_to_keep.cr3())
     len.cz <- length(tips_to_keep.cz3())
     len.kn <- length(tips_to_keep.kn3())
     len.me <- length(tips_to_keep.me3())
     len.mi <- length(tips_to_keep.mi3())
     len.pp <- length(tips_to_keep.pp3())
     len.sl <- length(tips_to_keep.sl3())
     len.sm <- length(tips_to_keep.sm3())
     len.sp <- length(tips_to_keep.sp3())
     len.ta <- length(tips_to_keep.ta3())
     len.vc <- length(tips_to_keep.vc3())
     len.bp <- length(tips_to_keep.bp3())
     len.cri <- length(tips_to_keep.cri3())
     len.ds <- length(tips_to_keep.ds3())
     len.os <- length(tips_to_keep.os3())
     len.smag <- length(tips_to_keep.smag3())
     len.tp <- length(tips_to_keep.tp3())
     len.aa <- length(tips_to_keep.aa3())
     len.um <- length(tips_to_keep.um3())
     len.rs <- length(tips_to_keep.rs3())
     len.cyc <- length(tips_to_keep.cyc3())
     len.ca <- length(tips_to_keep.ca3())
     len.mv <- length(tips_to_keep.mv3())
     len.af <- length(tips_to_keep.af3())
     len.sc <- length(tips_to_keep.sc3())
     len.aegi <- length(tips_to_keep.aegi3())
     len.sb <- length(tips_to_keep.sb3())
     len.chara <- length(tips_to_keep.chara3())
     len.sceobli <- length(tips_to_keep.sceobli3())
     len.cocco <- length(tips_to_keep.cocco3())
     len.haema <- length(tips_to_keep.haema3())
     len.zea <- length(tips_to_keep.zm3())
     len.praco <- length(tips_to_keep.praco3())
     len.ostlu <- length(tips_to_keep.ostlu3())
     len.ostme <- length(tips_to_keep.ostme3())
     len.micco <- length(tips_to_keep.micco3())
     len.cymte <- length(tips_to_keep.cymte3())
     len.chlin <- length(tips_to_keep.chlin3())
     len.gonpe <- length(tips_to_keep.gonpe3())
     len.tetso <- length(tips_to_keep.tetso3())
     len.desar <- length(tips_to_keep.desar3())
     len.monmi <- length(tips_to_keep.monmi3())
     len.caule <- length(tips_to_keep.caule3())
     len.ostqu <- length(tips_to_keep.ostqu3())
     len.brysp <- length(tips_to_keep.brysp3())
     len.ellbi <- length(tips_to_keep.ellbi3())
     len.myrbi <- length(tips_to_keep.myrbi3())
     len.apalo <- length(tips_to_keep.apalo3())
     len.astgl <- length(tips_to_keep.astgl3())
     len.tresp <- length(tips_to_keep.tresp3())
     len.picso <- length(tips_to_keep.picso3())
     len.chlso <- length(tips_to_keep.chlso3())
     len.auxpr <- length(tips_to_keep.auxpr3())
     len.helsp <- length(tips_to_keep.helsp3())
     len.tetst <- length(tips_to_keep.tetst3())
     len.pedsp <- length(tips_to_keep.pedsp3())
     len.chlpr <- length(tips_to_keep.chlpr3())
     len.picsp <- length(tips_to_keep.picsp3())
     len.mesvb <- length(tips_to_keep.mesvb3())
     len.meskr <- length(tips_to_keep.meskr3())
     len.zygci <- length(tips_to_keep.zygci3())
     len.zygcb <- length(tips_to_keep.zygcb3())
     len.zygcy <- length(tips_to_keep.zygcy3())
     len.plesc <- length(tips_to_keep.plesc3())
     len.funhy <- length(tips_to_keep.funhy3())
     len.sphfa <- length(tips_to_keep.sphfa3())
     len.takle <- length(tips_to_keep.takle3())
     len.marpa <- length(tips_to_keep.marpa3())
     len.ricfl <- length(tips_to_keep.ricfl3())
     len.ricso <- length(tips_to_keep.ricso3())
     len.antpu <- length(tips_to_keep.antpu3())
     len.antfu <- length(tips_to_keep.antfu3())
     len.megfl <- length(tips_to_keep.megfl3())
     len.phach <- length(tips_to_keep.phach3())
     len.phyph <- length(tips_to_keep.phyph3())
     len.parpe <- length(tips_to_keep.parpe3())
     len.notor <- length(tips_to_keep.notor3())
     len.phaca <- length(tips_to_keep.phaca3())
     len.phasp <- length(tips_to_keep.phasp3())
     len.leidu <- length(tips_to_keep.leidu3())
     len.isosi <- length(tips_to_keep.isosi3())
     len.dipco <- length(tips_to_keep.dipco3())
     len.hupas <- length(tips_to_keep.hupas3())
     len.lyccl <- length(tips_to_keep.lyccl3())
     len.adica <- length(tips_to_keep.adica3())
     len.alssp <- length(tips_to_keep.alssp3())
     len.marve <- length(tips_to_keep.marve3())
     len.seqse <- length(tips_to_keep.seqse3())
     len.seqgi <- length(tips_to_keep.seqgi3())
     len.taxch <- length(tips_to_keep.taxch3())
     len.abial <- length(tips_to_keep.abial3())
     len.picab <- length(tips_to_keep.picab3())
     len.gnemo <- length(tips_to_keep.gnemo3())
     len.welmi <- length(tips_to_keep.welmi3())
     len.ginbi <- length(tips_to_keep.ginbi3())
     len.ambtr <- length(tips_to_keep.ambtr3())
     len.nymco <- length(tips_to_keep.nymco3())
     len.cinka <- length(tips_to_keep.cinka3())
     len.horvu <- length(tips_to_keep.horvu3())
     len.bradi <- length(tips_to_keep.bradi3())
     len.setit <- length(tips_to_keep.setit3())
     len.panha <- length(tips_to_keep.panha3())
     len.missi <- length(tips_to_keep.missi3())
     len.oroth <- length(tips_to_keep.oroth3())
     len.eleco <- length(tips_to_keep.eleco3())
     len.anaco <- length(tips_to_keep.anaco3())
     len.musac <- length(tips_to_keep.musac3())
     len.dioal <- length(tips_to_keep.dioal3())
     len.spipo <- length(tips_to_keep.spipo3())
     len.aspof <- length(tips_to_keep.aspof3())
     len.zosma <- length(tips_to_keep.zosma3())
     len.araly <- length(tips_to_keep.araly3())
     len.capgr <- length(tips_to_keep.capgr3())
     len.brara <- length(tips_to_keep.brara3())
     len.braol <- length(tips_to_keep.braol3())
     len.sisir <- length(tips_to_keep.sisir3())
     len.schpa <- length(tips_to_keep.schpa3())
     len.eutsa <- length(tips_to_keep.eutsa3())
     len.thlar <- length(tips_to_keep.thlar3())
     len.boest <- length(tips_to_keep.boest3())
     len.carpa <- length(tips_to_keep.carpa3())
     len.theca <- length(tips_to_keep.theca3())
     len.goshi <- length(tips_to_keep.goshi3())
     len.gosra <- length(tips_to_keep.gosra3())
     len.citsi <- length(tips_to_keep.citsi3())
     len.pontr <- length(tips_to_keep.pontr3())
     len.eucgr <- length(tips_to_keep.eucgr3())
     len.maldo <- length(tips_to_keep.maldo3())
     len.prupe <- length(tips_to_keep.prupe3())
     len.frave <- length(tips_to_keep.frave3())
     len.corav <- length(tips_to_keep.corav3())
     len.caril <- length(tips_to_keep.caril3())
     len.queru <- length(tips_to_keep.queru3())
     len.cucsa <- length(tips_to_keep.cucsa3())
     len.glyma <- length(tips_to_keep.glyma3())
     len.medtr <- length(tips_to_keep.medtr3())
     len.tripr <- length(tips_to_keep.tripr3())
     len.cicar <- length(tips_to_keep.cicar3())
     len.lotja <- length(tips_to_keep.lotja3())
     len.phaac <- length(tips_to_keep.phaac3())
     len.vigun <- length(tips_to_keep.vigun3())
     len.lupal <- length(tips_to_keep.lupal3())
     len.lencu <- length(tips_to_keep.lencu3())
     len.arahy <- length(tips_to_keep.arahy3())
     len.poptr <- length(tips_to_keep.poptr3())
     len.manes <- length(tips_to_keep.manes3())
     len.salpu <- length(tips_to_keep.salpu3())
     len.linus <- length(tips_to_keep.linus3())
     len.corci <- length(tips_to_keep.corci3())
     len.vitvi <- length(tips_to_keep.vitvi3())
     len.kalfe <- length(tips_to_keep.kalfe3())
     len.vacda <- length(tips_to_keep.vacda3())
     len.oleeu <- length(tips_to_keep.oleeu3())
     len.mimgu <- length(tips_to_keep.mimgu3())
     len.soltu <- length(tips_to_keep.soltu3())
     len.cofar <- length(tips_to_keep.cofar3())
     len.lacsa <- length(tips_to_keep.lacsa3())
     len.dauca <- length(tips_to_keep.dauca3())
     len.betvu <- length(tips_to_keep.betvu3())
     len.amahy <- length(tips_to_keep.amahy3())
     len.chequ <- length(tips_to_keep.chequ3())
     len.sapof <- length(tips_to_keep.sapof3())
     len.aquco <- length(tips_to_keep.aquco3())
     
     
     organims_reduced <- c(rep("Prasinoderma coloniale", len.praco),rep("Ostreococcus tauri", len.ot),
                           rep("Ostreococcus lucimarinus", len.ostlu),rep("Ostreococcus mediterraneus", len.ostme),
                           rep("Bathycoccus prasinos", len.bp),rep("Micromonas pusilla", len.mi),
                           rep("Micromonas commoda", len.micco),rep("Cymbomonas tetramitiformis", len.cymte),
                           rep("Chlamydomonas reinhardtii", len.cr),rep("Chlamydomonas incerta", len.chlin),
                           rep("Haematococcus lacustris", len.haema),rep("Volvox carteri", len.vc),
                           rep("Dunaliella salina", len.ds),rep("Gonium pectorale", len.gonpe),
                           rep("Tetrabaena socialis", len.tetso),rep("Chromochloris zofingiensis", len.cz),
                           rep("Raphidocelis subcapitata", len.rs),rep("Scenedesmus obliquus", len.sceobli),
                           rep("Desmodesmus armatus", len.desar),rep("Monoraphidium minutum", len.monmi),
                           rep("Caulerpa lentillifera", len.caule),rep("Ostreobium quekettii", len.ostqu),
                           rep("Bryopsis sp.", len.brysp),rep("Coccomyxa subellipsoidea", len.cocco),
                           rep("Elliptochloris bilobata", len.ellbi),rep("Myrmecia bisecta", len.myrbi),
                           rep("Apatococcus lobatus", len.apalo),rep("Asterochloris glomerata", len.astgl),
                           rep("Trebouxia sp.", len.tresp),rep("Picochlorum soloecismus", len.picso),
                           rep("Chlorella sorokiniana", len.chlso),rep("Auxenochlorella protothecoides", len.auxpr),
                           rep("Helicosporidium sp.", len.helsp),rep("Ulva mutabilis", len.um),
                           rep("Tetraselmis striata", len.tetst),rep("Pedinophyceae sp.", len.pedsp),
                           rep("Chloropicon primus", len.chlpr),rep("Picocystis sp.", len.picsp),
                           rep("Chlorokybus atmophyticus", len.ca),rep("Mesostigma viride CCAC 1140", len.mv),
                           rep("Mesostigma viride NIES 296", len.mesvb),rep("Klebsormidium nitens", len.kn),
                           rep("Chara braunii", len.chara),rep("Mesotaenium endlicherianum", len.me),
                           rep("Mesotaenium kramstae", len.meskr),rep("Spirogloea muscicola", len.sp),
                           rep("Zygnema circumcarinatum SAG", len.zygci),rep("Zygnema circumcarinatum UTEX 1559", len.zygcb),
                           rep("Zygnema cylindricum", len.zygcy),rep("Ceratodon purpureus", len.cp),
                           rep("Physcomitrium patens", len.pp),rep("Pleurozium schreberi", len.plesc),
                           rep("Funaria hygrometrica", len.funhy),rep("Sphagnum magellanicum", len.smag),
                           rep("Sphagnum fallax", len.sphfa),rep("Takakia lepidozioides", len.takle),
                           rep("Marchantia polymorpha", len.mp),rep("Marchantia paleacea", len.marpa),
                           rep("Riccia fluitans", len.ricfl),rep("Riccia sorocarpa", len.ricso),
                           rep("Anthoceros agrestis", len.aa),rep("Anthoceros punctatus", len.antpu),
                           rep("Anthoceros fusiformis", len.antfu),rep("Megaceros flagellaris", len.megfl),
                           rep("Phaeomegaceros chiloensis", len.phach),rep("Phymatoceros phymatodes", len.phyph),
                           rep("Paraphymatoceros pearsonii", len.parpe),rep("Notothylas orbicularis", len.notor),
                           rep("Phaeoceros carolinianus", len.phaca),rep("Phaeoceros sp.", len.phasp),
                           rep("Leiosporoceros dussii", len.leidu),rep("Selaginella moellendorffii", len.sm),
                           rep("Isoetes sinensis", len.isosi),rep("Diphasiastrum complanatum", len.dipco),
                           rep("Huperzia asiatica", len.hupas),rep("Lycopodium clavatum", len.lyccl),
                           rep("Ceratopteris richardii", len.cri),rep("Adiantum capillus", len.adica),
                           rep("Alsophila spinulosa", len.alssp),rep("Salvinia cucullata", len.sc),
                           rep("Azolla filiculoides", len.af),rep("Marsilea vestita", len.marve),
                           rep("Thuja plicata", len.tp),rep("Sequoia sempervirens", len.seqse),
                           rep("Sequoiadendron giganteum", len.seqgi),rep("Taxus chinensis", len.taxch),
                           rep("Abies alba", len.abial),rep("Picea abies", len.picab),rep("Gnetum montanum", len.gnemo),
                           rep("Welwitschia mirabilis", len.welmi),rep("Cycas panzhihuaensis", len.cyc),
                           rep("Ginkgo biloba", len.ginbi),rep("Amborella trichopoda", len.ambtr),
                           rep("Nymphaea colorata", len.nymco),rep("Cinnamomum kanehirae", len.cinka),
                           rep("Aegilops tauschii", len.aegi),rep("Triticum aestivum", len.ta),
                           rep("Hordeum vulgare", len.horvu),rep("Brachypodium distachyon", len.bradi),
                           rep("Oryza sativa", len.os),rep("Setaria italica", len.setit),rep("Sorghum bicolor", len.sb),
                           rep("Zea mays", len.zea),rep("Panicum hallii", len.panha),rep("Miscanthus sinensis", len.missi),
                           rep("Oropetium thomaeum", len.oroth),rep("Eleusine coracana", len.eleco),
                           rep("Ananas comosus", len.anaco),rep("Musa acuminata", len.musac),rep("Dioscorea alata", len.dioal),
                           rep("Spirodela polyrhiza", len.spipo),rep("Asparagus officinalis", len.aspof),
                           rep("Zostera marina", len.zosma),rep("Arabidopsis thaliana", len.at),
                           rep("Arabidopsis lyrata", len.araly),rep("Capsella grandiflora", len.capgr),
                           rep("Brassica rapa", len.brara),rep("Brassica oleracea", len.braol),
                           rep("Sisymbrium irio", len.sisir),rep("Schrenkiella parvula", len.schpa),
                           rep("Eutrema salsugineum", len.eutsa),rep("Thlaspi arvense", len.thlar),
                           rep("Boechera stricta", len.boest),rep("Carica papaya", len.carpa),rep("Theobroma cacao", len.theca),
                           rep("Gossypium hirsutum", len.goshi),rep("Gossypium raimondii", len.gosra),
                           rep("Citrus sinensis", len.citsi),rep("Poncirus trifoliata", len.pontr),
                           rep("Eucalyptus grandis", len.eucgr),rep("Malus domestica", len.maldo),
                           rep("Prunus persica", len.prupe),rep("Fragaria vesca", len.frave),rep("Corylus avellana", len.corav),
                           rep("Carya illinoinensis", len.caril),rep("Quercus rubra", len.queru),
                           rep("Cucumis sativus", len.cucsa),rep("Glycine max", len.glyma),rep("Medicago truncatula", len.medtr),
                           rep("Trifolium pratense", len.tripr),rep("Cicer arietinum", len.cicar),
                           rep("Lotus japonicus", len.lotja),rep("Phaseolus acutifolius", len.phaac),
                           rep("Vigna unguiculata", len.vigun),rep("Lupinus albus", len.lupal),rep("Lens culinaris", len.lencu),
                           rep("Arachis hypogaea", len.arahy),rep("Populus trichocarpa", len.poptr),
                           rep("Manihot esculenta", len.manes),rep("Salix purpurea", len.salpu),
                           rep("Linum usitatissimum", len.linus),rep("Corymbia citriodora", len.corci),
                           rep("Vitis vinifera", len.vitvi),rep("Kalanchoe fedtschenkoi", len.kalfe),
                           rep("Vaccinium darrowii", len.vacda),rep("Olea europaea", len.oleeu),
                           rep("Mimulus guttatus", len.mimgu),rep("Solanum lycopersicum", len.sl),
                           rep("Solanum tuberosum", len.soltu),rep("Coffea arabica", len.cofar),
                           rep("Lactuca sativa", len.lacsa),rep("Daucus carota", len.dauca),rep("Beta vulgaris", len.betvu),
                           rep("Amaranthus hypochondriacus", len.amahy),rep("Chenopodium quinoa", len.chequ),
                           rep("Saponaria officinalis", len.sapof),rep("Aquilegia coerulea", len.aquco))
     
     return(organims_reduced)
   }) %>% bindEvent(input$run_button3)
   
   # Create tree plot
   tree_plot3 <- reactive({
     
     # Define previous variables
     tree_reduced <- tree_reduced3()
     gene.name.tree <- input$geneInt3
     tree <- tree_adj3()
     
     tips_to_keep.mp <- tips_to_keep.mp3()
     tips_to_keep.ot <- tips_to_keep.ot3()
     tips_to_keep.at <- tips_to_keep.at3()
     tips_to_keep.cp <- tips_to_keep.cp3()
     tips_to_keep.cr <- tips_to_keep.cr3()
     tips_to_keep.cz <- tips_to_keep.cz3()
     tips_to_keep.kn <- tips_to_keep.kn3()
     tips_to_keep.me <- tips_to_keep.me3()
     tips_to_keep.mi <- tips_to_keep.mi3()
     tips_to_keep.pp <- tips_to_keep.pp3()
     tips_to_keep.sl <- tips_to_keep.sl3()
     tips_to_keep.sm <- tips_to_keep.sm3()
     tips_to_keep.sp <- tips_to_keep.sp3()
     tips_to_keep.ta <- tips_to_keep.ta3()
     tips_to_keep.vc <- tips_to_keep.vc3()
     tips_to_keep.bp <- tips_to_keep.bp3()
     tips_to_keep.cri <- tips_to_keep.cri3()
     tips_to_keep.ds <- tips_to_keep.ds3()
     tips_to_keep.os <- tips_to_keep.os3()
     tips_to_keep.smag <- tips_to_keep.smag3()
     tips_to_keep.tp <- tips_to_keep.tp3()
     tips_to_keep.aa <- tips_to_keep.aa3()
     tips_to_keep.um <- tips_to_keep.um3()
     tips_to_keep.rs <- tips_to_keep.rs3()
     tips_to_keep.cyc <- tips_to_keep.cyc3()
     tips_to_keep.ca <- tips_to_keep.ca3()
     tips_to_keep.mv <- tips_to_keep.mv3()
     tips_to_keep.af <- tips_to_keep.af3()
     tips_to_keep.sc <- tips_to_keep.sc3()
     tips_to_keep.aegi <- tips_to_keep.aegi3()
     tips_to_keep.sb <- tips_to_keep.sb3()
     tips_to_keep.chara <- tips_to_keep.chara3()
     tips_to_keep.sceobli <- tips_to_keep.sceobli3()
     tips_to_keep.cocco <- tips_to_keep.cocco3()
     tips_to_keep.haema <- tips_to_keep.haema3()
     tips_to_keep.zm <- tips_to_keep.zm3()
     tips_to_keep.praco <- tips_to_keep.praco3()
     tips_to_keep.ostlu <- tips_to_keep.ostlu3()
     tips_to_keep.ostme <- tips_to_keep.ostme3()
     tips_to_keep.micco <- tips_to_keep.micco3()
     tips_to_keep.cymte <- tips_to_keep.cymte3()
     tips_to_keep.chlin <- tips_to_keep.chlin3()
     tips_to_keep.gonpe <- tips_to_keep.gonpe3()
     tips_to_keep.tetso <- tips_to_keep.tetso3()
     tips_to_keep.desar <- tips_to_keep.desar3()
     tips_to_keep.monmi <- tips_to_keep.monmi3()
     tips_to_keep.caule <- tips_to_keep.caule3()
     tips_to_keep.ostqu <- tips_to_keep.ostqu3()
     tips_to_keep.brysp <- tips_to_keep.brysp3()
     tips_to_keep.ellbi <- tips_to_keep.ellbi3()
     tips_to_keep.myrbi <- tips_to_keep.myrbi3()
     tips_to_keep.apalo <- tips_to_keep.apalo3()
     tips_to_keep.astgl <- tips_to_keep.astgl3()
     tips_to_keep.tresp <- tips_to_keep.tresp3()
     tips_to_keep.picso <- tips_to_keep.picso3()
     tips_to_keep.chlso <- tips_to_keep.chlso3()
     tips_to_keep.auxpr <- tips_to_keep.auxpr3()
     tips_to_keep.helsp <- tips_to_keep.helsp3()
     tips_to_keep.tetst <- tips_to_keep.tetst3()
     tips_to_keep.pedsp <- tips_to_keep.pedsp3()
     tips_to_keep.chlpr <- tips_to_keep.chlpr3()
     tips_to_keep.picsp <- tips_to_keep.picsp3()
     tips_to_keep.mesvb <- tips_to_keep.mesvb3()
     tips_to_keep.meskr <- tips_to_keep.meskr3()
     tips_to_keep.zygci <- tips_to_keep.zygci3()
     tips_to_keep.zygcb <- tips_to_keep.zygcb3()
     tips_to_keep.zygcy <- tips_to_keep.zygcy3()
     tips_to_keep.plesc <- tips_to_keep.plesc3()
     tips_to_keep.funhy <- tips_to_keep.funhy3()
     tips_to_keep.sphfa <- tips_to_keep.sphfa3()
     tips_to_keep.takle <- tips_to_keep.takle3()
     tips_to_keep.marpa <- tips_to_keep.marpa3()
     tips_to_keep.ricfl <- tips_to_keep.ricfl3()
     tips_to_keep.ricso <- tips_to_keep.ricso3()
     tips_to_keep.antpu <- tips_to_keep.antpu3()
     tips_to_keep.antfu <- tips_to_keep.antfu3()
     tips_to_keep.megfl <- tips_to_keep.megfl3()
     tips_to_keep.phach <- tips_to_keep.phach3()
     tips_to_keep.phyph <- tips_to_keep.phyph3()
     tips_to_keep.parpe <- tips_to_keep.parpe3()
     tips_to_keep.notor <- tips_to_keep.notor3()
     tips_to_keep.phaca <- tips_to_keep.phaca3()
     tips_to_keep.phasp <- tips_to_keep.phasp3()
     tips_to_keep.leidu <- tips_to_keep.leidu3()
     tips_to_keep.isosi <- tips_to_keep.isosi3()
     tips_to_keep.dipco <- tips_to_keep.dipco3()
     tips_to_keep.hupas <- tips_to_keep.hupas3()
     tips_to_keep.lyccl <- tips_to_keep.lyccl3()
     tips_to_keep.adica <- tips_to_keep.adica3()
     tips_to_keep.alssp <- tips_to_keep.alssp3()
     tips_to_keep.marve <- tips_to_keep.marve3()
     tips_to_keep.seqse <- tips_to_keep.seqse3()
     tips_to_keep.seqgi <- tips_to_keep.seqgi3()
     tips_to_keep.taxch <- tips_to_keep.taxch3()
     tips_to_keep.abial <- tips_to_keep.abial3()
     tips_to_keep.picab <- tips_to_keep.picab3()
     tips_to_keep.gnemo <- tips_to_keep.gnemo3()
     tips_to_keep.welmi <- tips_to_keep.welmi3()
     tips_to_keep.ginbi <- tips_to_keep.ginbi3()
     tips_to_keep.ambtr <- tips_to_keep.ambtr3()
     tips_to_keep.nymco <- tips_to_keep.nymco3()
     tips_to_keep.cinka <- tips_to_keep.cinka3()
     tips_to_keep.horvu <- tips_to_keep.horvu3()
     tips_to_keep.bradi <- tips_to_keep.bradi3()
     tips_to_keep.setit <- tips_to_keep.setit3()
     tips_to_keep.panha <- tips_to_keep.panha3()
     tips_to_keep.missi <- tips_to_keep.missi3()
     tips_to_keep.oroth <- tips_to_keep.oroth3()
     tips_to_keep.eleco <- tips_to_keep.eleco3()
     tips_to_keep.anaco <- tips_to_keep.anaco3()
     tips_to_keep.musac <- tips_to_keep.musac3()
     tips_to_keep.dioal <- tips_to_keep.dioal3()
     tips_to_keep.spipo <- tips_to_keep.spipo3()
     tips_to_keep.aspof <- tips_to_keep.aspof3()
     tips_to_keep.zosma <- tips_to_keep.zosma3()
     tips_to_keep.araly <- tips_to_keep.araly3()
     tips_to_keep.capgr <- tips_to_keep.capgr3()
     tips_to_keep.brara <- tips_to_keep.brara3()
     tips_to_keep.braol <- tips_to_keep.braol3()
     tips_to_keep.sisir <- tips_to_keep.sisir3()
     tips_to_keep.schpa <- tips_to_keep.schpa3()
     tips_to_keep.eutsa <- tips_to_keep.eutsa3()
     tips_to_keep.thlar <- tips_to_keep.thlar3()
     tips_to_keep.boest <- tips_to_keep.boest3()
     tips_to_keep.carpa <- tips_to_keep.carpa3()
     tips_to_keep.theca <- tips_to_keep.theca3()
     tips_to_keep.goshi <- tips_to_keep.goshi3()
     tips_to_keep.gosra <- tips_to_keep.gosra3()
     tips_to_keep.citsi <- tips_to_keep.citsi3()
     tips_to_keep.pontr <- tips_to_keep.pontr3()
     tips_to_keep.eucgr <- tips_to_keep.eucgr3()
     tips_to_keep.maldo <- tips_to_keep.maldo3()
     tips_to_keep.prupe <- tips_to_keep.prupe3()
     tips_to_keep.frave <- tips_to_keep.frave3()
     tips_to_keep.corav <- tips_to_keep.corav3()
     tips_to_keep.caril <- tips_to_keep.caril3()
     tips_to_keep.queru <- tips_to_keep.queru3()
     tips_to_keep.cucsa <- tips_to_keep.cucsa3()
     tips_to_keep.glyma <- tips_to_keep.glyma3()
     tips_to_keep.medtr <- tips_to_keep.medtr3()
     tips_to_keep.tripr <- tips_to_keep.tripr3()
     tips_to_keep.cicar <- tips_to_keep.cicar3()
     tips_to_keep.lotja <- tips_to_keep.lotja3()
     tips_to_keep.phaac <- tips_to_keep.phaac3()
     tips_to_keep.vigun <- tips_to_keep.vigun3()
     tips_to_keep.lupal <- tips_to_keep.lupal3()
     tips_to_keep.lencu <- tips_to_keep.lencu3()
     tips_to_keep.arahy <- tips_to_keep.arahy3()
     tips_to_keep.poptr <- tips_to_keep.poptr3()
     tips_to_keep.manes <- tips_to_keep.manes3()
     tips_to_keep.salpu <- tips_to_keep.salpu3()
     tips_to_keep.linus <- tips_to_keep.linus3()
     tips_to_keep.corci <- tips_to_keep.corci3()
     tips_to_keep.vitvi <- tips_to_keep.vitvi3()
     tips_to_keep.kalfe <- tips_to_keep.kalfe3()
     tips_to_keep.vacda <- tips_to_keep.vacda3()
     tips_to_keep.oleeu <- tips_to_keep.oleeu3()
     tips_to_keep.mimgu <- tips_to_keep.mimgu3()
     tips_to_keep.soltu <- tips_to_keep.soltu3()
     tips_to_keep.cofar <- tips_to_keep.cofar3()
     tips_to_keep.lacsa <- tips_to_keep.lacsa3()
     tips_to_keep.dauca <- tips_to_keep.dauca3()
     tips_to_keep.betvu <- tips_to_keep.betvu3()
     tips_to_keep.amahy <- tips_to_keep.amahy3()
     tips_to_keep.chequ <- tips_to_keep.chequ3()
     tips_to_keep.sapof <- tips_to_keep.sapof3()
     tips_to_keep.aquco <- tips_to_keep.aquco3()
     
     if (length(tree_reduced$tip.label) < 2)
     {
       cat("")
     }
     else 
     {
       
       # Color asignment per species
       col.factor <- c()
       org.factor <- c()
       
       library(glue)
       library(ggtree)
       library(ggplot2)
       
       for (i in 1:length(tree_reduced$tip.label))
       {
         if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
         {
           col.factor <- c(col.factor,"#006400")
           org.factor <- c(org.factor,"Marchantia polymorpha")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
         {
           col.factor <- c(col.factor,"#00008B")
           org.factor <- c(org.factor,"Ostreococcus tauri")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
         {
           col.factor <- c(col.factor,"#CD661D")
           org.factor <- c(org.factor,"Arabidopsis thaliana")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
         {
           col.factor <- c(col.factor,"#458B74")
           org.factor <- c(org.factor,"Ceratodon purpureus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
         {
           col.factor <- c(col.factor,"#8B7355")
           org.factor <- c(org.factor,"Chlamydomonas reinhardtii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
         {
           col.factor <- c(col.factor,"#458B00")
           org.factor <- c(org.factor,"Chromochloris zofingiensis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
         {
           col.factor <- c(col.factor,"#CD1076")
           org.factor <- c(org.factor,"Klebsormidium nitens")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
         {
           col.factor <- c(col.factor,"#8B8878")
           org.factor <- c(org.factor,"Mesotaenium endlicherianum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
         {
           col.factor <- c(col.factor,"#666666")
           org.factor <- c(org.factor,"Micromonas pusilla")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
         {
           col.factor <- c(col.factor,"#B8860B")
           org.factor <- c(org.factor,"Physcomitrium patens")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
         {
           col.factor <- c(col.factor,"#8B008B")
           org.factor <- c(org.factor,"Solanum lycopersicum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
         {
           col.factor <- c(col.factor,"#6E8B3D")
           org.factor <- c(org.factor,"Selaginella moellendorffii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
         {
           col.factor <- c(col.factor,"#79CDCD")
           org.factor <- c(org.factor,"Spirogloea muscicola")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
         {
           col.factor <- c(col.factor,"#CDCD00")
           org.factor <- c(org.factor,"Triticum aestivum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
         {
           col.factor <- c(col.factor,"#16317d")
           org.factor <- c(org.factor,"Volvox carteri")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
         {
           col.factor <- c(col.factor,"#007e2f")
           org.factor <- c(org.factor,"Bathycoccus prasinos")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
         {
           col.factor <- c(col.factor,"#ffcd12")
           org.factor <- c(org.factor,"Ceratopteris richardii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
         {
           col.factor <- c(col.factor,"#b86092")
           org.factor <- c(org.factor,"Dunaliella salina")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
         {
           col.factor <- c(col.factor,"#721b3e")
           org.factor <- c(org.factor,"Oryza sativa")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
         {
           col.factor <- c(col.factor,"#00b7a7")
           org.factor <- c(org.factor,"Sphagnum magellanicum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
         {
           col.factor <- c(col.factor,"#67000d")
           org.factor <- c(org.factor,"Thuja plicata")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
         {
           col.factor <- c(col.factor,"#5b2c6f")
           org.factor <- c(org.factor,"Anthoceros agrestis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
         {
           col.factor <- c(col.factor,"#15e71b")
           org.factor <- c(org.factor,"Ulva mutabilis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
         {
           col.factor <- c(col.factor,"#2d04f9")
           org.factor <- c(org.factor,"Raphidocelis subcapitata")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
         {
           col.factor <- c(col.factor,"#873600")
           org.factor <- c(org.factor,"Cycas panzhihuaensis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
         {
           col.factor <- c(col.factor,"#0b5345")
           org.factor <- c(org.factor,"Chlorokybus atmophyticus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
         {
           col.factor <- c(col.factor,"#283747")
           org.factor <- c(org.factor,"Mesostigma viride CCAC 1140")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
         {
           col.factor <- c(col.factor,"#145a32")
           org.factor <- c(org.factor,"Azolla filiculoides")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
         {
           col.factor <- c(col.factor,"#3339e6")
           org.factor <- c(org.factor,"Salvinia cucullata")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
         {
           col.factor <- c(col.factor,"#e6338f")
           org.factor <- c(org.factor,"Aegilops tauschii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
         {
           col.factor <- c(col.factor,"#cd016a")
           org.factor <- c(org.factor,"Sorghum bicolor")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
         {
           col.factor <- c(col.factor,"#117a65")
           org.factor <- c(org.factor,"Chara braunii")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
         {
           col.factor <- c(col.factor,"#148f77")
           org.factor <- c(org.factor,"Scenedesmus obliquus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
         {
           col.factor <- c(col.factor,"#9c640c")
           org.factor <- c(org.factor,"Coccomyxa subellipsoidea")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
         {
           col.factor <- c(col.factor,"#196f3d")
           org.factor <- c(org.factor,"Haematococcus lacustris")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
         {
           col.factor <- c(col.factor,"#666909")
           org.factor <- c(org.factor,"Zea mays")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.praco])
         {
           col.factor <- c(col.factor,"#ADFF2F")
           org.factor <- c(org.factor,"Prasinoderma coloniale")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostlu])
         {
           col.factor <- c(col.factor,"#483D8B")
           org.factor <- c(org.factor,"Ostreococcus lucimarinus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostme])
         {
           col.factor <- c(col.factor,"#7FFFD4")
           org.factor <- c(org.factor,"Ostreococcus mediterraneus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.micco])
         {
           col.factor <- c(col.factor,"#228B22")
           org.factor <- c(org.factor,"Micromonas commoda")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymte])
         {
           col.factor <- c(col.factor,"#00F5FF")
           org.factor <- c(org.factor,"Cymbomonas tetramitiformis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlin])
         {
           col.factor <- c(col.factor,"#FFE4B5")
           org.factor <- c(org.factor,"Chlamydomonas incerta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gonpe])
         {
           col.factor <- c(col.factor,"#8B3A62")
           org.factor <- c(org.factor,"Gonium pectorale")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetso])
         {
           col.factor <- c(col.factor,"#FFDAB9")
           org.factor <- c(org.factor,"Tetrabaena socialis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.desar])
         {
           col.factor <- c(col.factor,"#54FF9F")
           org.factor <- c(org.factor,"Desmodesmus armatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.monmi])
         {
           col.factor <- c(col.factor,"#20B2AA")
           org.factor <- c(org.factor,"Monoraphidium minutum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caule])
         {
           col.factor <- c(col.factor,"#9BCD9B")
           org.factor <- c(org.factor,"Caulerpa lentillifera")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostqu])
         {
           col.factor <- c(col.factor,"#5CACEE")
           org.factor <- c(org.factor,"Ostreobium quekettii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brysp])
         {
           col.factor <- c(col.factor,"#B8860B")
           org.factor <- c(org.factor,"Bryopsis sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ellbi])
         {
           col.factor <- c(col.factor,"#EEDD82")
           org.factor <- c(org.factor,"Elliptochloris bilobata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.myrbi])
         {
           col.factor <- c(col.factor,"#F0FFFF")
           org.factor <- c(org.factor,"Myrmecia bisecta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.apalo])
         {
           col.factor <- c(col.factor,"#C71585")
           org.factor <- c(org.factor,"Apatococcus lobatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.astgl])
         {
           col.factor <- c(col.factor,"#8B4789")
           org.factor <- c(org.factor,"Asterochloris glomerata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tresp])
         {
           col.factor <- c(col.factor,"#8B008B")
           org.factor <- c(org.factor,"Trebouxia sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picso])
         {
           col.factor <- c(col.factor,"#D2691E")
           org.factor <- c(org.factor,"Picochlorum soloecismus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlso])
         {
           col.factor <- c(col.factor,"#8B7D6B")
           org.factor <- c(org.factor,"Chlorella sorokiniana")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.auxpr])
         {
           col.factor <- c(col.factor,"#000080")
           org.factor <- c(org.factor,"Auxenochlorella protothecoides")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.helsp])
         {
           col.factor <- c(col.factor,"#A020F0")
           org.factor <- c(org.factor,"Helicosporidium sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetst])
         {
           col.factor <- c(col.factor,"#EE00EE")
           org.factor <- c(org.factor,"Tetraselmis striata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pedsp])
         {
           col.factor <- c(col.factor,"#8B1C62")
           org.factor <- c(org.factor,"Pedinophyceae sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlpr])
         {
           col.factor <- c(col.factor,"#FFFACD")
           org.factor <- c(org.factor,"Chloropicon primus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picsp])
         {
           col.factor <- c(col.factor,"#FFFFE0")
           org.factor <- c(org.factor,"Picocystis sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mesvb])
         {
           col.factor <- c(col.factor,"#A52A2A")
           org.factor <- c(org.factor,"Mesostigma viride NIES 296")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.meskr])
         {
           col.factor <- c(col.factor,"#E0FFFF")
           org.factor <- c(org.factor,"Mesotaenium kramstae")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygci])
         {
           col.factor <- c(col.factor,"#E6E6FA")
           org.factor <- c(org.factor,"Zygnema circumcarinatum SAG")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcb])
         {
           col.factor <- c(col.factor,"#DCDCDC")
           org.factor <- c(org.factor,"Zygnema circumcarinatum UTEX 1559")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcy])
         {
           col.factor <- c(col.factor,"#8B636C")
           org.factor <- c(org.factor,"Zygnema cylindricum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.plesc])
         {
           col.factor <- c(col.factor,"#CD2990")
           org.factor <- c(org.factor,"Pleurozium schreberi")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.funhy])
         {
           col.factor <- c(col.factor,"#CDC9C9")
           org.factor <- c(org.factor,"Funaria hygrometrica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sphfa])
         {
           col.factor <- c(col.factor,"#1C86EE")
           org.factor <- c(org.factor,"Sphagnum fallax")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.takle])
         {
           col.factor <- c(col.factor,"#8B7500")
           org.factor <- c(org.factor,"Takakia lepidozioides")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marpa])
         {
           col.factor <- c(col.factor,"#EED5B7")
           org.factor <- c(org.factor,"Marchantia paleacea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricfl])
         {
           col.factor <- c(col.factor,"#EE5C42")
           org.factor <- c(org.factor,"Riccia fluitans")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricso])
         {
           col.factor <- c(col.factor,"#FFFAFA")
           org.factor <- c(org.factor,"Riccia sorocarpa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antpu])
         {
           col.factor <- c(col.factor,"#D02090")
           org.factor <- c(org.factor,"Anthoceros punctatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antfu])
         {
           col.factor <- c(col.factor,"#FFF8DC")
           org.factor <- c(org.factor,"Anthoceros fusiformis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.megfl])
         {
           col.factor <- c(col.factor,"#EEA2AD")
           org.factor <- c(org.factor,"Megaceros flagellaris")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phach])
         {
           col.factor <- c(col.factor,"#BCD2EE")
           org.factor <- c(org.factor,"Phaeomegaceros chiloensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phyph])
         {
           col.factor <- c(col.factor,"#FF00FF")
           org.factor <- c(org.factor,"Phymatoceros phymatodes")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.parpe])
         {
           col.factor <- c(col.factor,"#68228B")
           org.factor <- c(org.factor,"Paraphymatoceros pearsonii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.notor])
         {
           col.factor <- c(col.factor,"#006400")
           org.factor <- c(org.factor,"Notothylas orbicularis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaca])
         {
           col.factor <- c(col.factor,"#00FFFF")
           org.factor <- c(org.factor,"Phaeoceros carolinianus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phasp])
         {
           col.factor <- c(col.factor,"#EED2EE")
           org.factor <- c(org.factor,"Phaeoceros sp.")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.leidu])
         {
           col.factor <- c(col.factor,"#FFEFD5")
           org.factor <- c(org.factor,"Leiosporoceros dussii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.isosi])
         {
           col.factor <- c(col.factor,"#CDC8B1")
           org.factor <- c(org.factor,"Isoetes sinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dipco])
         {
           col.factor <- c(col.factor,"#8B6508")
           org.factor <- c(org.factor,"Diphasiastrum complanatum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.hupas])
         {
           col.factor <- c(col.factor,"#FFA54F")
           org.factor <- c(org.factor,"Huperzia asiatica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lyccl])
         {
           col.factor <- c(col.factor,"#EEAEEE")
           org.factor <- c(org.factor,"Lycopodium clavatum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.adica])
         {
           col.factor <- c(col.factor,"#FF0000")
           org.factor <- c(org.factor,"Adiantum capillus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.alssp])
         {
           col.factor <- c(col.factor,"#9B30FF")
           org.factor <- c(org.factor,"Alsophila spinulosa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marve])
         {
           col.factor <- c(col.factor,"#CD661D")
           org.factor <- c(org.factor,"Marsilea vestita")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqse])
         {
           col.factor <- c(col.factor,"#7AC5CD")
           org.factor <- c(org.factor,"Sequoia sempervirens")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqgi])
         {
           col.factor <- c(col.factor,"#CDC0B0")
           org.factor <- c(org.factor,"Sequoiadendron giganteum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.taxch])
         {
           col.factor <- c(col.factor,"#4682B4")
           org.factor <- c(org.factor,"Taxus chinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.abial])
         {
           col.factor <- c(col.factor,"#CDBA96")
           org.factor <- c(org.factor,"Abies alba")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picab])
         {
           col.factor <- c(col.factor,"#A2CD5A")
           org.factor <- c(org.factor,"Picea abies")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gnemo])
         {
           col.factor <- c(col.factor,"#FAFAD2")
           org.factor <- c(org.factor,"Gnetum montanum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.welmi])
         {
           col.factor <- c(col.factor,"#E0EEEE")
           org.factor <- c(org.factor,"Welwitschia mirabilis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ginbi])
         {
           col.factor <- c(col.factor,"#7A378B")
           org.factor <- c(org.factor,"Ginkgo biloba")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ambtr])
         {
           col.factor <- c(col.factor,"#FFDEAD")
           org.factor <- c(org.factor,"Amborella trichopoda")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.nymco])
         {
           col.factor <- c(col.factor,"#C0FF3E")
           org.factor <- c(org.factor,"Nymphaea colorata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cinka])
         {
           col.factor <- c(col.factor,"#7FFFD4")
           org.factor <- c(org.factor,"Cinnamomum kanehirae")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.horvu])
         {
           col.factor <- c(col.factor,"#1E90FF")
           org.factor <- c(org.factor,"Hordeum vulgare")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bradi])
         {
           col.factor <- c(col.factor,"#76EE00")
           org.factor <- c(org.factor,"Brachypodium distachyon")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.setit])
         {
           col.factor <- c(col.factor,"#EE9A49")
           org.factor <- c(org.factor,"Setaria italica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.panha])
         {
           col.factor <- c(col.factor,"#FFF0F5")
           org.factor <- c(org.factor,"Panicum hallii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.missi])
         {
           col.factor <- c(col.factor,"#CD9B1D")
           org.factor <- c(org.factor,"Miscanthus sinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oroth])
         {
           col.factor <- c(col.factor,"#36648B")
           org.factor <- c(org.factor,"Oropetium thomaeum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eleco])
         {
           col.factor <- c(col.factor,"#EEA9B8")
           org.factor <- c(org.factor,"Eleusine coracana")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.anaco])
         {
           col.factor <- c(col.factor,"#ADD8E6")
           org.factor <- c(org.factor,"Ananas comosus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.musac])
         {
           col.factor <- c(col.factor,"#66CDAA")
           org.factor <- c(org.factor,"Musa acuminata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dioal])
         {
           col.factor <- c(col.factor,"#FFF0F5")
           org.factor <- c(org.factor,"Dioscorea alata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.spipo])
         {
           col.factor <- c(col.factor,"#CD5555")
           org.factor <- c(org.factor,"Spirodela polyrhiza")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aspof])
         {
           col.factor <- c(col.factor,"#FFEC8B")
           org.factor <- c(org.factor,"Asparagus officinalis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zosma])
         {
           col.factor <- c(col.factor,"#668B8B")
           org.factor <- c(org.factor,"Zostera marina")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.araly])
         {
           col.factor <- c(col.factor,"#CDC673")
           org.factor <- c(org.factor,"Arabidopsis lyrata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.capgr])
         {
           col.factor <- c(col.factor,"#8FBC8F")
           org.factor <- c(org.factor,"Capsella grandiflora")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brara])
         {
           col.factor <- c(col.factor,"#4EEE94")
           org.factor <- c(org.factor,"Brassica rapa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.braol])
         {
           col.factor <- c(col.factor,"#BA55D3")
           org.factor <- c(org.factor,"Brassica oleracea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sisir])
         {
           col.factor <- c(col.factor,"#A4D3EE")
           org.factor <- c(org.factor,"Sisymbrium irio")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.schpa])
         {
           col.factor <- c(col.factor,"#8B2323")
           org.factor <- c(org.factor,"Schrenkiella parvula")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eutsa])
         {
           col.factor <- c(col.factor,"#8B2252")
           org.factor <- c(org.factor,"Eutrema salsugineum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.thlar])
         {
           col.factor <- c(col.factor,"#FF6A6A")
           org.factor <- c(org.factor,"Thlaspi arvense")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.boest])
         {
           col.factor <- c(col.factor,"#EE7942")
           org.factor <- c(org.factor,"Boechera stricta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.carpa])
         {
           col.factor <- c(col.factor,"#BBFFFF")
           org.factor <- c(org.factor,"Carica papaya")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.theca])
         {
           col.factor <- c(col.factor,"#00CD66")
           org.factor <- c(org.factor,"Theobroma cacao")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.goshi])
         {
           col.factor <- c(col.factor,"#FFE4C4")
           org.factor <- c(org.factor,"Gossypium hirsutum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gosra])
         {
           col.factor <- c(col.factor,"#8B8386")
           org.factor <- c(org.factor,"Gossypium raimondii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.citsi])
         {
           col.factor <- c(col.factor,"#CD4F39")
           org.factor <- c(org.factor,"Citrus sinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pontr])
         {
           col.factor <- c(col.factor,"#8B4726")
           org.factor <- c(org.factor,"Poncirus trifoliata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eucgr])
         {
           col.factor <- c(col.factor,"#FF6EB4")
           org.factor <- c(org.factor,"Eucalyptus grandis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.maldo])
         {
           col.factor <- c(col.factor,"#00008B")
           org.factor <- c(org.factor,"Malus domestica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.prupe])
         {
           col.factor <- c(col.factor,"#6495ED")
           org.factor <- c(org.factor,"Prunus persica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.frave])
         {
           col.factor <- c(col.factor,"#8DB6CD")
           org.factor <- c(org.factor,"Fragaria vesca")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corav])
         {
           col.factor <- c(col.factor,"#EE9572")
           org.factor <- c(org.factor,"Corylus avellana")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caril])
         {
           col.factor <- c(col.factor,"#FFEBCD")
           org.factor <- c(org.factor,"Carya illinoinensis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.queru])
         {
           col.factor <- c(col.factor,"#9ACD32")
           org.factor <- c(org.factor,"Quercus rubra")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cucsa])
         {
           col.factor <- c(col.factor,"#EE6AA7")
           org.factor <- c(org.factor,"Cucumis sativus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.glyma])
         {
           col.factor <- c(col.factor,"#D15FEE")
           org.factor <- c(org.factor,"Glycine max")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.medtr])
         {
           col.factor <- c(col.factor,"#6B8E23")
           org.factor <- c(org.factor,"Medicago truncatula")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tripr])
         {
           col.factor <- c(col.factor,"#8B1A1A")
           org.factor <- c(org.factor,"Trifolium pratense")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cicar])
         {
           col.factor <- c(col.factor,"#F0FFF0")
           org.factor <- c(org.factor,"Cicer arietinum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lotja])
         {
           col.factor <- c(col.factor,"#DEB887")
           org.factor <- c(org.factor,"Lotus japonicus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaac])
         {
           col.factor <- c(col.factor,"#FFA07A")
           org.factor <- c(org.factor,"Phaseolus acutifolius")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vigun])
         {
           col.factor <- c(col.factor,"#FFD700")
           org.factor <- c(org.factor,"Vigna unguiculata")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lupal])
         {
           col.factor <- c(col.factor,"#8B668B")
           org.factor <- c(org.factor,"Lupinus albus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lencu])
         {
           col.factor <- c(col.factor,"#FF7F50")
           org.factor <- c(org.factor,"Lens culinaris")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.arahy])
         {
           col.factor <- c(col.factor,"#8B8B7A")
           org.factor <- c(org.factor,"Arachis hypogaea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.poptr])
         {
           col.factor <- c(col.factor,"#FAF0E6")
           org.factor <- c(org.factor,"Populus trichocarpa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.manes])
         {
           col.factor <- c(col.factor,"#EE7621")
           org.factor <- c(org.factor,"Manihot esculenta")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.salpu])
         {
           col.factor <- c(col.factor,"#CDAA7D")
           org.factor <- c(org.factor,"Salix purpurea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.linus])
         {
           col.factor <- c(col.factor,"#EE799F")
           org.factor <- c(org.factor,"Linum usitatissimum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corci])
         {
           col.factor <- c(col.factor,"#551A8B")
           org.factor <- c(org.factor,"Corymbia citriodora")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vitvi])
         {
           col.factor <- c(col.factor,"#90EE90")
           org.factor <- c(org.factor,"Vitis vinifera")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kalfe])
         {
           col.factor <- c(col.factor,"#8B4C39")
           org.factor <- c(org.factor,"Kalanchoe fedtschenkoi")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vacda])
         {
           col.factor <- c(col.factor,"#CD5C5C")
           org.factor <- c(org.factor,"Vaccinium darrowii")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oleeu])
         {
           col.factor <- c(col.factor,"#CD3278")
           org.factor <- c(org.factor,"Olea europaea")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mimgu])
         {
           col.factor <- c(col.factor,"#B3EE3A")
           org.factor <- c(org.factor,"Mimulus guttatus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.soltu])
         {
           col.factor <- c(col.factor,"#EE3B3B")
           org.factor <- c(org.factor,"Solanum tuberosum")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cofar])
         {
           col.factor <- c(col.factor,"#FFA500")
           org.factor <- c(org.factor,"Coffea arabica")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lacsa])
         {
           col.factor <- c(col.factor,"#66CD00")
           org.factor <- c(org.factor,"Lactuca sativa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dauca])
         {
           col.factor <- c(col.factor,"#FF4500")
           org.factor <- c(org.factor,"Daucus carota")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.betvu])
         {
           col.factor <- c(col.factor,"#00FF00")
           org.factor <- c(org.factor,"Beta vulgaris")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.amahy])
         {
           col.factor <- c(col.factor,"#EEC900")
           org.factor <- c(org.factor,"Amaranthus hypochondriacus")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chequ])
         {
           col.factor <- c(col.factor,"#7B68EE")
           org.factor <- c(org.factor,"Chenopodium quinoa")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sapof])
         {
           col.factor <- c(col.factor,"#FFA07A")
           org.factor <- c(org.factor,"Saponaria officinalis")
         }
         
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aquco])
         {
           col.factor <- c(col.factor,"#DA70D6")
           org.factor <- c(org.factor,"Aquilegia coerulea")
         }
         
         
       }
       
       
       
       #Matrix with labels and colors and transform to dplyr format
       data.tree <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                               col = col.factor, org = org.factor)
       
       d2 <- dplyr::mutate(data.tree, lab = data.tree$label,
                           color = data.tree$col,
                           organism = data.tree$org,
                           name = glue("<i style='color:{color}'> {lab} </i>"))
       { 
         if (build_trees3() == "Maximum Likelihood")
         {
           tree_plot <- ggtree(tree_reduced) %<+% d2 + geom_tiplab() + theme(legend.position =) +
             xlim(0, max(tree_reduced$edge.length)*3) + geom_tiplab(aes(label = label, color = organism)) +
             scale_color_manual(values = unique(d2$col), breaks = unique(d2$org)) +
             labs(fill = "Node of interest")
           
         }
         else
         {
           tree_plot <- ggtree(tree_reduced) %<+% d2 + geom_tiplab() + theme(legend.position =) +
             xlim(0, max(tree_reduced$edge.length)*3) + geom_tiplab(aes(label = label, color = organism)) +
             geom_nodelab() +
             scale_color_manual(values = unique(d2$col), breaks = unique(d2$org)) +
             labs(fill = "Node of interest")
         }
       }
       shinyjs::hideElement(id = 'loading.tree3')
       return(tree_plot)
     }}) %>% bindEvent(input$run_button3)
   
   
   # Outputs
   observeEvent(isTruthy(tree_plot3()), {
     output$error_tree3 <- NULL
   })
   
   # Create boxes
   observeEvent(isTruthy(tree_plot3()), {
     
     if (UI_exist_tree3)
     {
       removeUI(
         selector = "div:has(>> #treeTips3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #presentorg3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadTree3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadNewick3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadTreeSeqs3)",
         multiple = TRUE,
         immediate = TRUE
       )
     }
     
     
     insertUI("#box_tree_text3", "afterEnd", ui = {
       box(
         title = "Genes in Orthogroup", status = "danger", solidHeader = TRUE, width = 12,
         collapsible = TRUE,
         verbatimTextOutput("treeTips3")
       )
     }) 
     
     insertUI("#box_tree_pie3", "afterEnd", ui = {
       box(
         title = "Present Organisms", status = "danger", solidHeader = TRUE,
         collapsible = TRUE, width = 12,
         plotlyOutput("presentorg3")
       )
     }) 
     
     insertUI("#box_tree_plot3", "afterEnd", ui = {
       image_height <- 300 + 15*length(tree_reduced3()$tip.label)
       box(width = 12,
           title = "Gene Tree", status = "danger", solidHeader = TRUE,
           collapsible = TRUE, 
           plotOutput("tree_image3", height = image_height, width = 1100)
       )
     })
     
     insertUI("#download_tree3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTree3", 
                                                                          "Download Tree Plot",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#download_newick3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadNewick3", 
                                                                          "Download NEWICK Tree",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#download_tree_seqs3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTreeSeqs3", 
                                                                          "Download Protein Sequences",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_tree3 <<- TRUE
     shinyjs::hideElement(id = 'loading.tree3')
   })
   
   # Fill boxes with output
   output$treeTips3 <- renderPrint({
     print(tree_reduced3()$tip.label)
   }, width = 400) # %>% bindEvent(input$run_button3)
   
   # Render pie chart
   output$presentorg3 <- renderPlotly({
     
     {library(ggplot2)
       library(dplyr)
       
       data <- data.frame(table(organims_reduced3()))
       colnames(data) <- c("group", "value")
       
       # Compute the position of labels
       data <- data %>%
         arrange(desc(group)) %>%
         mutate(prop = value / sum(data$value) *100) %>%
         mutate(ypos = cumsum(prop)- 0.5*prop )
       
       # Create plot
       
       plotly::plot_ly(data=data,values=~prop,labels=~factor(group),
                       marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                              floor(nrow(data)/9)+1)),
                       type="pie",showlegend = F, text= ~group,
                       textinfo = "none", hoverinfo = "text")} 
     
   })
   
   # Render tree image
   output$tree_image3 <- renderImage({
     image_height <- 300 + 15*length(tree_reduced3()$tip.label)
     png("tree3.png", height = image_height, width = 1100)
     plot(tree_plot3())
     dev.off()
     
     list(src = "tree3.png",
          contentType="image/png", width=1100,height=image_height)
   }, deleteFile = T)
   
   # Download results
   output$downloadTree3 <- downloadHandler(
     filename= function() {
       paste("tree", ".png", sep="")
     },
     content= function(file) {
       image_height <- (300 + 11*length(tree_reduced3()$tip.label))*3
       image_width <- (200 + 400*max(tree_reduced3()$edge.length))*3
       png(file, height = image_height, width = image_width, res = (70 + 0.1*length(tree_reduced3()$tip.label))*3)
       plot(tree_plot3())
       dev.off()
     })
   
   # Create and download tree in newick format
   output$downloadNewick3 <- downloadHandler(
     filename= function() {
       paste("tree_newick", ".txt", sep="")
     },
     content= function(file) {
       write.tree(tree_reduced3(), file)
     })
   
   #  # Create and download sequences for genes in tree
   output$downloadTreeSeqs3 <- downloadHandler(
     filename= function() {
       paste("tree_seqs", ".fa", sep="")
     },
     content= function(file) {
       seqinr::write.fasta(sequences = seqinr::getSequence(ortho_reduced3()),
                           names = seqinr::getName(ortho_reduced3()), file.out = file)
     })
   
   
   ####################### PHYLOWIDGET ############################
   # Remove previous outputs when updated by a new search
   observeEvent(input$run_button3, {
     if (UI_exist_phylo3)
     {
       removeUI(
         selector = "div:has(>>> #phylo_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_phylo3 <<- F
     }
     
   })
   
   phylo_tree3 <- reactive({
     
     library(ape)
     tree_phylo <- tree_reduced3()
     
     # Normalize tree depth
     root_id <- length(tree_phylo$tip.label)+1
     norm_factor <- max(dist.nodes(tree_phylo)[root_id,])
     tree_phylo$edge.length <- tree_phylo$edge.length/norm_factor
     
     return(tree_phylo)
     
   }) %>% bindEvent(input$phylo_start3)
   
   observeEvent(isTruthy(phylo_tree3()),{
     
     if(UI_exist_phylo3)
     {
       removeUI(
         selector = "div:has(>>> #phylo_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_phylo3 <<- F
     }
     
     
     insertUI("#box_phylo3", "afterEnd", ui = {
       phylo_tree <- phylo_tree3()
       phylo_height <- length(phylo_tree$tip.label) *14 + 220
       box(width = 12,
           title = "Interactive Tree", status = "danger", solidHeader = TRUE, height = phylo_height + 100,
           collapsible = TRUE,
           tags$div(id = "phylo_pocket3", style = paste0("width: 1300px; height: ",  phylo_height + 50, "px"),
                    phylowidgetOutput("phylo_plot3", height = paste0(phylo_height,"px"), width = "98%"))
       )
     })
     
     UI_exist_phylo3 <<- T
     
   })
   
   output$phylo_plot3 <- renderPhylowidget({
     
     phylo_tree <- phylo_tree3()
     phylowidget(phylo_tree)
   })
   
   #########################  PFAM  ###############################
   
   observeEvent(input$run_button3, {
     removeUI(
       selector = "div:has(>> #selected_pfamsI3)",
       multiple = TRUE,
       immediate = TRUE
     )
   })
   
   observeEvent(input$pfam_start3, {
     insertUI("#selected_pfams3", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_pfamsI3","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced3()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({tree_reduced3()$tip.label[1]}))
     })
   })
   
   observeEvent(input$run_button3, {
     removeUI("#pfam_selection3")
   })
   
   observeEvent(input$pfam_start3, {
     insertUI("#pfam_selectionI3", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("pfam_selection3", "Show Pfam Domains", size = "sm",
                                style = "float", color = "danger")
     })
   })
   
   
   observeEvent(input$run_button3, {
     if (UI_exist_pfam3)
     {
       removeUI(
         selector = "div:has(>> #output_pfam_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pfam3 <<- F
     }
   })
   
   
   total_table_pfam3 <- reactive({
     shinyjs::showElement(id = 'loading.pfam.pf3')
     ortho_reduced <- ortho_reduced3()
     sel_genes <- as.vector(input$selected_pfamsI3)
     
     if (length(sel_genes) < 1)
     {
       shinyjs::hideElement(id = 'loading.pfam.pf3')
       removeUI(
         selector = "div:has(>> #output_pfam_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       UI_exist_pfam3 <<- F
       output$error_pfam3 <- renderUI({renderText({print("Please select at least one gene.")})})
       validate(need(length(sel_genes) > 0, "Please select at least one gene."))
     }
     
     output$error_pfam3 <- NULL

     library(RCurl)
     library(drawProteins)
     library(ggplot2)
     library(jsonlite)
     
     # Get the sequences as a vector of strings
     
     
     # Create data frame with proper columns
     total_table_pfam <- data.frame(type=NA,
                                    description=NA,
                                    begin=NA, end=NA,
                                    length=NA,
                                    accession=NA, entryName=NA,
                                    taxid=NA, order=NA)
     
     
     # Fill data frame with the information about domains obtained with hmmer
     for (i in 1:length(sel_genes))
     {
       ortho_comp <- ortho_reduced[[sel_genes[i]]]
       ortho_str <- seqinr::getSequence(ortho_comp, as.string = T)
       ortho_cha <- unlist(ortho_str)
       
       
       # Curl POST request
       url <- "https://www.ebi.ac.uk/Tools/hmmer/api/v1/search/hmmscan"
       
       data <- list(
         database = "pfam",
         input = ortho_cha
       )
       
       json_data <- toJSON(data, auto_unbox = T)
       
       response <- postForm(url, 
                            .opts = list(
                              postfields = json_data,
                              httpheader = c(
                                "Content-Type" = "application/json",
                                "Accept" = "application/json"
                              )
                            ))
       
       
       new_id <- strsplit(as.character(response), '"')[[1]][4]
       
       
       # Results URL
       url_res <- paste0("https://www.ebi.ac.uk/Tools/hmmer/api/v1/result/", new_id)
       
       # Iterate until job is finished, for that, check if SUCCESS is present in GET result
       nap_again <- T
       while(nap_again)
       {
         get_response <- getURL(url_res,
                                httpheader = c(
                                  "Accept" = "application/json"
                                ))
         
         if (grepl("SUCCESS", get_response))
         {
           nap_again <- F
           
           # Once computation is over, parse JSON response
           json_resp <- jsonlite::fromJSON(url_res, simplifyDataFrame = TRUE)
           json_hits <- data.frame(json_resp$result$hits)
           
           hits_acc <- json_hits$acc
           hits_desc <- json_hits$metadata$description
           hits_domains <- json_hits$domains
           
           df_acc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_acc, y = hits_domains, USE.NAMES = F))
           df_desc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_desc, y = hits_domains, USE.NAMES = F))
           df_begin <- unlist(sapply(hits_domains, function(x) x$iali))
           df_end <- unlist(sapply(hits_domains, function(x) x$jali))
           
           # Create PFAM table
           pfam_table <- data.frame(type=c("CHAIN", rep("DOMAIN", length(df_acc))),
                                    description=c("Protein chain",df_acc),
                                    begin=c(1, df_begin), end=c(nchar(ortho_cha),df_end),
                                    length=c(nchar(ortho_cha)-1, df_end - df_begin),
                                    accession=sel_genes[i], entryName=sel_genes[i],
                                    taxid=c("Chain", df_desc), order=i)
           
           total_table_pfam <- rbind(total_table_pfam, pfam_table)
           
         }
         else if (grepl("PENDING", get_response))
         {
           Sys.sleep(10)
         }
         # If there is a problem with server
         else
         {
           nap_again <- F
           pfam_table <- data.frame(type="CHAIN",
                                    description="Protein chain",
                                    begin=1, end=nchar(ortho_cha),
                                    length=nchar(ortho_cha)-1,
                                    accession=sel_genes[i], entryName=sel_genes[i],
                                    taxid="Chain", order=i)
           total_table_pfam <- rbind(total_table_pfam, pfam_table)
         }
       }
     }
     
     total_table_pfam <- total_table_pfam[-1,]
     total_table_pfam <- total_table_pfam[!duplicated(total_table_pfam),]
     # Remove protein chain results
     total_table_pfam <- subset(total_table_pfam, !(type=="DOMAIN" & description=="Protein chain"))
     
     return(total_table_pfam)
     
   }) %>% bindEvent(input$pfam_selection3)
   
   pfplot3 <- reactive({
     
     total_table_pfam <- total_table_pfam3()
     
     pfplot <- draw_canvas(total_table_pfam)
     pfplot <- draw_chains(pfplot, total_table_pfam)
     pfplot <- draw_domains(pfplot, total_table_pfam, label_domains = F)
     pfplot <- pfplot + theme_bw(base_size = 20) + # white background
       theme(panel.grid.minor=element_blank(),
             panel.grid.major=element_blank()) +
       theme(axis.ticks = element_blank(),
             axis.text.y = element_blank()) +
       theme(panel.border = element_blank())
     pfplot <- pfplot + theme(legend.position="top") + labs(fill="")
     
   }) %>% bindEvent(input$pfam_selection3)
   
   
   # Outputs
   
   observeEvent(isTruthy(pfplot3()), {
     
     if (UI_exist_pfam3)
     {
       removeUI(
         selector = "div:has(>> #output_pfam_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_pfam3", "afterEnd", ui = {
       
       box(
         title = "PFAM Table", status = "danger", solidHeader = TRUE, width = 12,
         collapsible = TRUE,
         dataTableOutput(outputId = "output_pfam_table3"))
     }) 
     
     insertUI("#box_pfplot3", "afterEnd", ui = {
       total_table_pfam <- total_table_pfam3()
       box_pfplot_height <- 150 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
       box(
         title = "Domains Localization", status = "danger", solidHeader = TRUE, width = 12, #height = box_pfplot_height,
         collapsible = TRUE,
         plotOutput("pfam_plot3", height = box_pfplot_height))
     }) 
     
     insertUI("#pfam_down_button3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "pfam_download3", "Download PFAM figure",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#download_ui_for_pfam_table3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadPFAMTable3", "Download PFAM Table",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_pfam3 <<- TRUE
     shinyjs::hideElement(id = 'loading.pfam.pf3')
   })
   
   output$output_pfam_table3 <- renderDataTable({
     total_table_pfam <- total_table_pfam3()
     out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
     out_pf_table$description <- sapply(out_pf_table$description, function(x) pfam.link(x))
     colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
     return(out_pf_table)
   }, escape=FALSE, options = list(pageLength = 5))
   
   output$pfam_plot3 <- renderImage({
     total_table_pfam <- total_table_pfam3()
     pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
     pfam_width <- 1000
     pfplot <- pfplot3()
     png("pharaoh_folder/pfam.png",  width = pfam_width, height = pfam_height)
     plot(pfplot)
     dev.off()
     list(src = "pharaoh_folder/pfam.png",
          contentType="image/png")
   }, deleteFile = T
   )
   
   # Download results
   
   output$pfam_download3 <- downloadHandler(
     filename= function() {
       paste("pfam", ".png", sep="")
     },
     content= function(file) {
       total_table_pfam <- total_table_pfam3()
       pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
       pfam_width <- 1150
       pfplot <- pfplot3()
       
       png(file, height = pfam_height, width = pfam_width)
       plot(pfplot)
       dev.off()
     })
   
   output$downloadPFAMTable3 <- downloadHandler(
     filename= function() {
       paste("pfam_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_pfam <- total_table_pfam3()
       out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
       colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
       write.table(x = out_pf_table, quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   ####################### CAFE #################################
   
   observeEvent(input$run_button3, {
     if (UI_exist_cafe3)
     {
       removeUI(
         selector = "div:has(>> #cafe_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_mrca3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCAFEPlot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_cafe3 <<- F
     }
     
     if (UI_exist_error_cafe3)
     {
       removeUI(
         selector = "div:has(>> #cafe_error_message3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_error_cafe3 <<- F
     }
   })
   
   ### CAFE parser and tree generator
   cafe_table3 <- reactive({
     
     shinyjs::showElement(id = 'loading.cafe3')
     
     # Import OG name
     og.cafe <- og.name3()
     
     # Define path to COUNT file
     cafe_comp_tree_file <- "pharaoh_folder/family_table_count.tsv"
     
     # Extract COUNT vector for the current OG
     library(data.table)
     library(ape)
     genecount <- fread(cafe_comp_tree_file, sep="\t", header = T)
     genecountdf <- as.data.frame(genecount)
     rownames(genecountdf) <- genecountdf$Orthogroup
     new_cols <- gsub(" ", "_", colnames(genecountdf))
     colnames(genecountdf) <- new_cols
     
     cafe.vector <- genecountdf[og.cafe,]
     
     if (sum(is.na(cafe.vector)) != 0)
     {
       shinyjs::hideElement(id = 'loading.cafe3')
       if (UI_exist_cafe3)
       {
         removeUI(
           selector = "div:has(>> #cafe_plot3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #cafe_mrca3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #cafe_download3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadCAFEPlot3)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_cafe3 <<- F
       
       if(UI_exist_error_cafe3)
       {
         removeUI(
           selector = "div:has(>> #cafe_error_message3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
       }
       insertUI("#error_cafe3", "afterEnd", ui = {
         box(width = 12,
             title = "Ancestral State Reconstruction", status = "info", solidHeader = TRUE,
             collapsible = TRUE,
             textOutput("cafe_error_message3"))
       })
       
       output$cafe_error_message3 <- renderText({print("No expansions/contraction detected for this orthogroup,
                                                        or infeasible computation due to large size and variance across
                                                        species.")})
       UI_exist_error_cafe3 <<- T
       
       validate(need(sum(is.na(cafe.vector)) == 0 , ""))
     }
     
     return(genecountdf)
   }) %>% bindEvent(input$cafe_start3)
   
   mrca.tree3 <- reactive({
     
     og.cafe <- og.name3()
     genecountdf <- cafe_table3()
     
     # Create phylogenomic tree with internal nodes names
     mrca.tree <- read.tree("pharaoh_folder/tree_with_internal_nodes.txt")
     
     return(mrca.tree)
     
   }) %>% bindEvent(input$cafe_start3)
   
   cafe_tree3 <- reactive({
     og.cafe <- og.name3()
     genecountdf <- cafe_table3()
     cafe.tree <- mrca.tree3()
     
     # Add counts to nodes
     tip.count <- as.numeric(genecountdf[og.cafe,2:172])
     names(tip.count) <- colnames(genecountdf)[2:172]
     
     node.count <- as.numeric(genecountdf[og.cafe,173:342])
     names(node.count) <- colnames(genecountdf)[173:342]
     
     cafe.tree$tip.label[match(names(tip.count), cafe.tree$tip.label)] <- tip.count
     cafe.tree$node.label[match(names(node.count), cafe.tree$node.label)] <- node.count
     
     return(cafe.tree)
     
   }) %>% bindEvent(input$cafe_start3)
   
   evo_plot_data3 <- reactive({
     
     og.cafe <- og.name3()
     genecountdf <- cafe_table3()
     cafe.tree <- mrca.tree3()
     
     # Add counts to nodes
     tip.count <- as.numeric(genecountdf[og.cafe,2:172])
     names(tip.count) <- colnames(genecountdf)[2:172]
     
     node.count <- as.numeric(genecountdf[og.cafe,173:342])
     names(node.count) <- colnames(genecountdf)[173:342]
     
     # Identify parental node to determine if a general expansion or contraction 
     # occured at the level of each node
     edge_table <- as.data.frame(cafe.tree$edge)
     rownames(edge_table) <- paste("edge", 1:nrow(edge_table), sep = "")
     colnames(edge_table) <- c("parent", "child")
     
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) <
          as.numeric(tip.count[x])) "Expansion"
       else if (as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) >
                as.numeric(tip.count[x])) "Contraction"
       else "No change"
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x) # 170 is root, no parent
       if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) <
          as.numeric(node.count[x])) "Expansion"
       else if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) >
               as.numeric(node.count[x])) "Contraction"
       else "No change"
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # A gain occurs when a node experiences an expansion and its parental exhibits a value of 0
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if (exp_cont_tip[x] == "Expansion" &
           as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) == 0
       ) "Gain"
       else exp_cont_tip[x]
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x)
       if (exp_cont_node[match(x,names(exp_cont_node))] == "Expansion" &
           as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) == 0
       ) "Gain"
       else exp_cont_node[x]
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # A loss occurs when a node experiences a contraction and takes the value 0
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if (exp_cont_tip[x] == "Contraction" &
           tip.count[x] == 0
       ) "Loss"
       else exp_cont_tip[x]
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x)
       if (exp_cont_node[match(x,names(exp_cont_node))] == "Contraction" &
           node.count[x] == 0
       ) "Loss"
       else exp_cont_node[x]
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # Create complete sorted vectors
     exp_cont_node["root"] <- "No change"
     node_ordered <- exp_cont_node[order(match(names(exp_cont_node),cafe.tree$node.label))]
     tip_ordered <- exp_cont_tip[order(match(names(exp_cont_tip),cafe.tree$tip.label))]
     
     change_vector <- c(tip_ordered, node_ordered)
     
     # Merge tips and nodes reconstruction
     node_ordered_values <- node.count[order(match(names(node.count),cafe.tree$node.label))]
     tip_ordered_values <- tip.count[order(match(names(tip.count),cafe.tree$tip.label))]
     cafe.count <- c(tip_ordered_values, node_ordered_values)
     
     # Create a timeline for a given OG
     tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
     tree.ancestor <- read.tree(tree.name)
     tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
     tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
     tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
     
     mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
     evo.paths <- c()
     for (i in 1:length(unique(tips.orgs)))
     {
       evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
     }
     
     evo.paths <- unique(evo.paths)
     evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
     
     # Associate gray and 0 to reconstruction for nodes not in allowed paths
     change_vector[setdiff(1:length(change_vector), evo.paths)] <- 
       sapply(change_vector[setdiff(1:length(change_vector), evo.paths)], function(x) if(x != "Loss") "OG not present" 
              else x)  
     
     change_vector <- unlist(change_vector)
     
     cafe.count[setdiff(1:length(change_vector), evo.paths)] <- 0
     
     unique(change_vector)
     color_cafe <- sapply(change_vector, function(x) if (x == "No change") "black" 
                          else if (x == "Expansion") "red" else if (x == "Contraction") "blue"
                          else if (x == "Gain") "green3" else if (x == "Loss") "purple"
                          else "gray", USE.NAMES = F)
     
     # Create tree representation
     cafe.table.tips <- data.frame(node = 1:length(cafe.tree$tip.label), label = cafe.tree$tip.label,
                                   col = color_cafe[1:length(cafe.tree$tip.label)], reconst = change_vector[1:length(cafe.tree$tip.label)],
                                   dup_number = cafe.count[1:length(cafe.tree$tip.label)])
     
     cafe.table.nodes <- data.frame(node = (Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label)), label = cafe.tree$node.label,
                                    col = color_cafe[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                    reconst = change_vector[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                    dup_number = cafe.count[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))])
     
     cafe.table.node.comp <- rbind(cafe.table.tips, cafe.table.nodes)
     
     d <- dplyr::mutate(cafe.table.node.comp)
     d$text <- d$label
     
     return(d)
     
   }) %>% bindEvent(input$cafe_start3)
   
   evo_plot3 <- reactive({
     
     d <- evo_plot_data3() 
     mrca.tree <- mrca.tree3()
     
     library(ggtree)
     library(ggplot2)
     
     evo_plot <- ggtree(mrca.tree) %<+% d + aes(colour = I(d$col)) +
       geom_tiplab(aes(label=gsub("_", " ", tools::toTitleCase(d$label))), offset = 30) +
       theme(legend.position = "none") +
       xlim(0, max(mrca.tree$edge.length)*1.85) +
       geom_nodepoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                      alpha = .75) +
       scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
       geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                     alpha = .75)
     
     return(evo_plot)
     
   }) %>% bindEvent(input$cafe_start3)
   
   evo_plotly3 <- reactive({
     
     d <- evo_plot_data3() 
     cafe.tree <- mrca.tree3()
     
     library(ggtree)
     library(ggplot2)
     
     # Remove internal nodes from direct visualization, retaining names in panels
     d$label[172:341] <- NA
     
     evo_plotly <- ggtree(cafe.tree) %<+% d + aes(colour = I(d$col),text=paste0("</br> Duplications: ",dup_number,
                                                                                "</br> Name: ",gsub("_", " ", tools::toTitleCase(d$text)))) + 
       geom_text(aes(x =  1970, label=gsub("_", " ", tools::toTitleCase(d$label)))) + 
       theme(legend.position = "none") +
       xlim(0, max(cafe.tree$edge.length)*1.25) +
       geom_point(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                  alpha = .75) +
       scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
       geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                     alpha = .75)
     
     p <- ggplotly(evo_plotly, tooltip = "text", width = 1400, height = 2200) 
     p <- p %>%
       plotly::style(textposition = "right",xanchor="right")
     
     return(p)
     
   }) %>% bindEvent(input$cafe_start3)
   
   evo.paths.id3 <- reactive({
     
     # Create phylogenomic tree with internal nodes names
     og.cafe <- og.name3()
     cafe.tree <- mrca.tree3()
     
     # Create timeline
     tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
     tree.ancestor <- read.tree(tree.name)
     tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
     tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
     tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
     
     mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
     evo.paths <- c()
     for (i in 1:length(unique(tips.orgs)))
     {
       evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
     }
     
     evo.paths <- unique(evo.paths)
     evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
     
     #print(paste0("Most recent common ancestor for this orthogroup is ", evo.paths.id[1]))
     
     return(evo.paths.id)
     
   }) %>% bindEvent(input$cafe_start3)
   
   # Outputs
   
   # Remove previous boxes if they exist and create new ones
   observeEvent(isTruthy(evo_plot3()), {
     
     if (UI_exist_cafe3)
     {
       removeUI(
         selector = "div:has(>> #cafe_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_mrca3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCAFEPlot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     if (UI_exist_error_cafe3)
     {
       removeUI(
         selector = "div:has(>> #cafe_error_message3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_cafe3", "afterEnd", ui = {
       box(width = 12,
           title = "Ancestral State Reconstruction", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           plotlyOutput("cafe_plot3", height = ifelse(model.selected3(), "2300px", "2300px")))
     })
     
     insertUI("#box_mrca3", "afterEnd", ui = {
       box(width = 8,
           title = "Most Recent Common Ancestor", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           textOutput("cafe_mrca3")
       )
     })
     
     insertUI("#cafe_down_button3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "cafe_download3", "Download NEWICK tree",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#download_ui_for_cafe_plot3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadCAFEPlot3", "Download Ancestral State Plot",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_cafe3 <<- TRUE
     shinyjs::hideElement(id = 'loading.cafe3')
   })
   
   # Fill outputs
   
   output$cafe_plot3 <- renderPlotly({
     evo_plotly3()
   })
   
   output$cafe_mrca3 <- renderText({
     print(paste0("Most recent common ancestor for this orthogroup is the
                   ancestor of the clade: ", evo.paths.id3()[1]))
   })
   
   # Download tab's results
   
   output$cafe_download3 <- downloadHandler(
     filename= function() {
       paste("ancestral_newick", ".txt", sep="")
     },
     content= function(file) {
       cafe_tree <- cafe_tree3()
       
       write.tree(cafe_tree, file)
     })
   
   output$downloadCAFEPlot3<- downloadHandler(
     filename= function() {
       paste("ancestral_plot", ".png", sep="")
     },
     content= function(file) {
       evo_plot <- evo_plot3()
       
       png(file, width = 1400, height = 2800, res = 100)
       plot(evo_plot)
       dev.off()
     })
   
   ####################### MSA #################################
   
   observeEvent(input$run_button3, {
     removeUI(
       selector = "div:has(>> #selected_msaI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(>> #msa_methodI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#msa_selection3")
     
     if (UI_exist_msa3)
     {
       removeUI(
         selector = "div:has(>>> #msa_print3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_fa3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_msa3 <<- F
     }
     
   })
   
   observeEvent(input$msa_start3, {
     insertUI("#selected_msa3", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_msaI3","Select the desired genes from the tree to align",
                                 choices=isolate({tree_reduced3()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({tree_reduced3()$tip.label[1]}))
       
     })
     
     insertUI("#msa_method3", "afterEnd", ui = {
       shinyWidgets::pickerInput(inputId = "msa_methodI3", label = "Choose alignment method", 
                                 choices = c("ClustalOmega", "MAFFT"), selected = "ClustalOmega")
       
     })
     
     insertUI("#msa_selectionI3", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("msa_selection3", "Align Sequences", size = "sm",
                                style = "float", color = "danger")
     })
     
   })
   
   alignseqs3 <- reactive({
     
     library(msa)
     shinyjs::showElement(id = 'loading.msa3')
     
     selected_genes <- as.vector(input$selected_msaI3)
     selected_method <- as.character(input$msa_methodI3)
     file.name <- og.name3()
     
     if (length(selected_genes) < 2)
     {
       shinyjs::hideElement(id = 'loading.msa3')
       
       if (UI_exist_msa3)
       {
         removeUI(
           selector = "div:has(>>> #msa_print3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #msa_download_plot3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #msa_download_fa3)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       UI_exist_msa3 <<- F
       output$error_msa3 <- renderUI({renderText({print("Please select at least two genes.")})})
       validate(need(length(selected_genes) > 1, "Please select at least two genes."))
     }
     
     output$error_msa3 <- NULL
     
     # If de novo alignment is selected
     {
       if(selected_method == "ClustalOmega")
       {
         # Define path to orthogroup sequences file
         ortho.seq.name <- paste("Orthogroup_Sequences", paste(file.name, "fa", sep = "."), sep="/")
         
         
         # Read orthogroup sequences file and select the genes for alignment
         mySequences1 <- Biostrings::readAAStringSet(ortho.seq.name)
         mysubseqs <- mySequences1[selected_genes]
         
         alignseqs <- msa(mysubseqs, verbose = F, method = "ClustalOmega")
       }
       
       # If MAFFT alignment is selected
       else
       {
         ortho.seq.name <- paste("MultipleSequenceAlignments", paste(file.name, "fa", sep = "."), sep="/")
         mySequences1 <- seqinr::read.fasta(ortho.seq.name, seqtype = "AA")
         mysubseqs <- mySequences1[selected_genes]
         mysubnames <- seqinr::getName(mySequences1)
         
         # Identify indexes associated with reduced names
         indexes_msa <- sapply(selected_genes, function(x) grep(mysubnames, pattern = x))
         
         # Retrieve those sequences from alignment keeping gaps
         mysubseqs <- mySequences1[indexes_msa]
         names(mysubseqs) <- names(indexes_msa)
         
         # Remove columns with gaps and remove empty spaces in last positions
         seqs_mysubseqs <- seqinr::getSequence(mysubseqs)
         last <- seqs_mysubseqs[[length(seqs_mysubseqs)]]
         last <- last[which(last != " ")]
         seqs_mysubseqs[[length(seqs_mysubseqs)]] <- last
         seqs_mysubseqs <- remove_gaps(seqs_mysubseqs)
         names(seqs_mysubseqs) <- names(mysubseqs)
         
         mysubseqs2 <- unlist(lapply(seqs_mysubseqs, function(x) paste(x, collapse="")))
         
         alignseqs <- Biostrings::AAMultipleAlignment(mysubseqs2, use.names = T)
         
       }
     }
     
     detach("package:msa", unload=TRUE)
     
     return(alignseqs)
     
   }) %>% bindEvent(input$msa_selection3)
   
   # Create boxes for outputs
   observeEvent(isTruthy(alignseqs3()), {
     
     if (UI_exist_msa3)
     {
       removeUI(
         selector = "div:has(>>> #msa_print3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_fa3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_msa3", "afterEnd", ui = {
       selected_msa <- isolate({input$selected_msaI3})
       msa_height <- ifelse(length(selected_msa) > 14, 550, 400 + 5*length(selected_msa))
       box(width = 12,
           title = "MSA Explorer", status = "danger", solidHeader = TRUE, height = msa_height,
           collapsible = TRUE,
           tags$div(id = "msa_pocket3", style = "width: 1300px; height: 400px",
                    msaROutput("msa_print3"))
           
       )
     })
     
     insertUI("#msa_down_plot3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_plot3", "Download Colored MSA",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#msa_down_fasta3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_fa3", "Download MSA FASTA",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_msa3 <<- TRUE
   })
   
   
   # Fill Output
   output$msa_print3 <- renderMsaR({
     alignseqs <- alignseqs3()
     msaout <- msa::msaConvert(alignseqs, "ape::AAbin")
     msaR(msaout, menu=T, overviewbox = F,  colorscheme = "clustal")
   })
   
   # Prepare variables for pdf construction
   observeEvent(isTruthy(alignseqs3()), {
     alignseqs <- alignseqs3()
     
     library(ggmsa)
     class(alignseqs) <- "AAMultipleAlignment"
     
     for(i in 1:(ncol(alignseqs)%/%100 +1)){
       assign(paste("msaportho", i, sep = ""), ggmsa(alignseqs, 1+(100*(i-1)), i*100, seq_name = TRUE, char_width = 0.5) +
                geom_seqlogo(color = "Chemistry_AA"), envir = as.environment(1), pos=1)
     }
     shinyjs::hideElement(id = 'loading.msa3')
   })
   
   # Download tab's results
   # Download colored MSA in pdf
   output$msa_download_plot3 <- downloadHandler(
     filename= function() {
       paste("msa", ".pdf", sep="")
     },
     content= function(file) {
       selected_msa <- input$selected_msaI3
       alignseqs <- alignseqs3()
       pdf(file, height = 2+length(selected_msa)*0.25, width = 16)
       {
         for(i in 1:(ncol(alignseqs)%/%100 +1)){
           print(mget(paste0("msaportho", i), envir = as.environment(1)))
         }
         dev.off()
       }
     })
   
   # Download MSA in FASTA format
   output$msa_download_fa3<- downloadHandler(
     filename= function() {
       paste("msa", ".fa", sep="")
     },
     content= function(file) {
       alignseqs <- alignseqs3()
       writeXStringSet(as(unmasked(alignseqs), "XStringSet"), file)
     })
   
   
   ####################### GO #################################
   
   observeEvent(input$run_button3, {
     removeUI(
       selector = "div:has(>> #selected_gosI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(>> #selected_gos_modeI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#gos_selectionI3")
     
     if (UI_exist_go3)
     {
       removeUI(
         selector = "div:has(>> #output_gos_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_treeplot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadGOSTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_gos_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_go3 <<- F
     }
     
   })
   
   observeEvent(input$go_start3, {
     insertUI("#selected_gos3", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_gosI3","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced3()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({tree_reduced3()$tip.label[1]}))
       
     })
     
     insertUI("#selected_gos_mode3", "afterEnd", ui = {
       
       selectInput(inputId = "selected_gos_modeI3",
                   choices=c("Biological Processes" = "bp",
                             "Molecular Functions" = "mf",
                             "Cellular Components" = "cc"),
                   label = "Select the gene ontology to use",
                   multiple = F, selected = c("bp"))
       
     })
     
     insertUI("#gos_selection3", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("gos_selectionI3", "Show GO terms", size = "sm",
                                style = "float", color = "danger")
     })
     
   })
   
   total_table_gos3 <- reactive({
     
     shinyjs::showElement(id = 'loading.go3')
     gos_anot <- read.csv("pharaoh_folder/final_anot_table.tsv", sep="\t", header = T)
     sel.genes.go <- as.vector(input$selected_gosI3)
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI3}))
     
     total_table_gos <- subset(gos_anot, gos_anot$name %in% sel.genes.go)
     
     # Show an error if no terms are identified in the input
     if (nrow(total_table_gos) == 0) 
     {
       shinyjs::hideElement(id = 'loading.go3')
       if (UI_exist_go3)
       {
         removeUI(
           selector = "div:has(>> #output_gos_table3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_plot3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_treeplot3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadGOSTable3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_download3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_gos_download3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_go3 <<- F
       }
       output$error_gos3 <- renderUI({
         renderPrint({cat("0 GO terms identified.")})
       })
       
       validate(need(nrow(total_table_gos) != 0, " "))
     }
     
     
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     terms_sel <- paste("terms", selected_gos_mode, sep="_")
     total_table_gos <- total_table_gos[,c("organism", "id", "name", gos_sel, terms_sel)]
     
     # Once removed the two GO categories not selected, remove rows with blank cells
     total_table_gos_clean <- total_table_gos[ total_table_gos[[gos_sel]] != "" , ]
     
     # Show an error if no terms are identified after the previous operation
     if (nrow(total_table_gos_clean) == 0) 
     {
       shinyjs::hideElement(id = 'loading.go3')
       if (UI_exist_go3)
       {
         removeUI(
           selector = "div:has(>> #output_gos_table3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_plot3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_treeplot3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadGOSTable3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_download3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_gos_download3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_go3 <<- F
       }
       
       output$error_gos3 <- renderUI({
         renderPrint({cat("0 GO terms identified.")})
       })
       
       validate(need(nrow(total_table_gos_clean) != 0, " "))
     }
     
     output$error_gos3 <- NULL
     
     return(total_table_gos_clean)
     
   }) %>% bindEvent(input$gos_selectionI3)
   
   enr_table3 <- reactive({
     
     total_table_gos <- total_table_gos3()
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI3}))
     
     # Create the plot
     
     # Create a list of GO terms vector of each gene
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     gos_list <- apply(total_table_gos,MARGIN=1,FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
     names(gos_list) <- total_table_gos$name
     
     # Count GOs for each gene and create a matrix of gene-GO pairs
     count_gos_in_genes <- sapply(gos_list, FUN = function(x) length(x))
     comp_data <- data.frame(gene = rep(names(count_gos_in_genes), count_gos_in_genes), gos = as.character(unlist(gos_list)))
     
     # Collapse genes that share a same GO
     gene.v <- c()
     for (i in 1:length(unique(comp_data$gos)))
     {
       new.table <- subset(comp_data, gos == unique(comp_data$gos)[i])
       new.cha <- as.character(new.table$gene)
       gene.v <- c(gene.v, paste(new.cha, collapse = "/"))
     }
     
     names(gene.v) <- unique(comp_data$gos)
     
     # Load libraries and create gene chains, count and GO IDs fields (same order)
     library(GO.db)
     library("multienrichjam")
     library(clusterProfiler)
     library(enrichplot)
     library(ggplot2)
     
     count_go <- table(comp_data$gos)
     geneids <- gene.v[names(count_go)]
     count_terms <- mapply(function(x) {Term(x)}, names(count_go), USE.NAMES = F)
     
     # If there  exists any obsolote term
     count_terms <- count_terms[which(!is.na(count_terms))] 
     count_go <- count_go[which(!is.na(count_terms))]
     geneids <- geneids[which(!is.na(count_terms))]
     
     # Create pseudo-enrichment table
     enr_table <- data.frame(ID=names(count_go), Description=count_terms, GeneRatio="90/100", BgRatio="90/10000", 
                             pvalue=0.000005, p.adjust=0.000005, qvalue=0.000005, geneID=geneids, Count = as.vector(count_go))
     
     return(enr_table)
     
   }) %>% bindEvent(input$gos_selectionI3)
   
   ema_gos_plot3 <- reactive({
     
     enr_table <- enr_table3()
     
     # Transform to enrichResult object
     enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                  geneColname = "geneID", pvalueColname = "p.adjust",
                                  descriptionColname = "Description", pvalueCutoff = 0.05)
     
     # Create plot
     ema_gos_plot <- emapplot(pairwise_termsim(enr), showCategory = 15) + theme(legend.position='none')
     return(ema_gos_plot)
   })
   
   tree_gos_plot3 <- reactive({
     
     enr_table <- enr_table3()
     
     # Transform to enrichResult object
     enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                  geneColname = "geneID", pvalueColname = "p.adjust",
                                  descriptionColname = "Description", pvalueCutoff = 0.05)
     {
       if (nrow(enr_table) > 4)
       {
         tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(label_words_n = 3)) +
           theme(legend.position='none')
       }
       else if (nrow(enr_table) > 2)
       {
         tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(n = 2, label_words_n = 3)) + 
           theme(legend.position='none')
       }
       else
       {
         text <- paste("\n  Unable to create treeplot with less than 3 GO terms \n")
         tree_gos_plot <- ggplot() + 
           annotate("text", x = 4, y = 25, size=8, label = text) + 
           theme_void()
       }
     }
     
     shinyjs::hideElement(id = 'loading.go3')
     return(tree_gos_plot)
   })
   
   # Create boxes for outputs
   observeEvent(isTruthy(tree_gos_plot3()), {
     
     if (UI_exist_go3)
     {
       removeUI(
         selector = "div:has(>> #output_gos_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_treeplot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadGOSTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_gos_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_gos_table3", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Table", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_gos_table3")
       )
     })
     
     insertUI("#box_gos_plot3", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Plot", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           plotOutput("gos_plot3", height = 610)
       )
     })
     
     insertUI("#box_gos_treeplot3", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Treeplot", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           plotOutput("gos_treeplot3", height = 610)
       )
     })
     
     insertUI("#download_ui_for_gos_table3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadGOSTable3", "Download GO Table",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#gos_down_button3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "gos_download3", "Download GO Plot",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#tree_gos_down_button3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "tree_gos_download3", "Download GO Treeplot",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_go3 <<- TRUE
   })
   
   # Fill outputs
   # Table
   output$output_gos_table3 <- renderDataTable({
     total_table_gos <- total_table_gos3()
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI3}))
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     gos_list <- apply(total_table_gos,MARGIN=1,
                       FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
     {
       if(!is.list(gos_list))
       {
         gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
         gos_formatted <- paste0(gos_links, collapse = " | ")
         total_table_gos[,gos_sel] <- gos_formatted
         total_table_gos 
       }
       else
       {
         gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
         gos_formatted <- unlist(sapply(gos_links, function(x) paste0(x, collapse = " | ")))
         total_table_gos[,gos_sel] <- gos_formatted
         total_table_gos  
       }
     }
   },escape=FALSE, rownames=F, options =list(pageLength = 5))
   
   # First plot
   output$gos_plot3 <- renderImage({
     ema_gos_plot <- ema_gos_plot3()
     
     png("pharaoh_folder/gosplot.png", width = 590, height = 590, res = 90)
     plot(ema_gos_plot)
     dev.off()
     list(src = "pharaoh_folder/gosplot.png",
          contentType="image/png")
     
   }, deleteFile=T
   )
   
   # Second plot
   output$gos_treeplot3 <- renderImage({
     tree_gos_plot <- tree_gos_plot3()
     
     png("pharaoh_folder/treeplot.png", width = 620, height = 590, res = 80)
     plot(tree_gos_plot)
     dev.off()
     list(src = "pharaoh_folder/treeplot.png",
          contentType="image/png")
     
   }, deleteFile=T
   )
   
   # Download tab's results
   # Download GO table
   output$downloadGOSTable3 <- downloadHandler(
     filename= function() {
       paste("GOS_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_gos <- total_table_gos3()
       
       write.table(x = total_table_gos,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download emaplot
   output$gos_download3 <- downloadHandler(
     filename= function() {
       paste("gos_plot", ".png", sep="")
     },
     content= function(file) {
       ema_gos_plot <- ema_gos_plot3()
       
       png(file,height = 1200, width = 1500, res=140)
       plot(ema_gos_plot)
       dev.off()
     })
   
   # Download treeplot
   output$tree_gos_download3 <- downloadHandler(
     filename= function() {
       paste("gos_treeplot", ".png", sep="")
     },
     content= function(file) {
       tree_gos_plot <- tree_gos_plot3()
       
       png(file, height = 1200, width = 1500, res=140)
       plot(tree_gos_plot)
       dev.off()
     })
  
   
   ###################### KEGG ###########################
   
   observeEvent(input$run_button3, {
     removeUI(
       selector = "div:has(>> #selected_kosI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#kos_selectionI3")
     
     
     if (UI_exist_kegg3)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg3 <<- F
     }
     
     if (UI_exist_kegg_path3)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI3")
       
       UI_exist_kegg_path3 <<- F
     }
     
     if (UI_exist_pathview3)
     {
       removeUI(
         selector = "div:has(>>> #path_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview3 <<- F
     }
     
     output$error_kos3 <- NULL
     
     
   })
   
   observeEvent(input$kegg_start3, {
     
     if (UI_exist_kegg3)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg3 <<- F
     }
     
     if (UI_exist_kegg_path3)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI3")
       
       UI_exist_kegg_path3 <<- F
     }
     
     if (UI_exist_pathview3)
     {
       removeUI(
         selector = "div:has(>>> #path_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview3 <<- F
     }
     
     output$error_kos3 <- NULL
     
     insertUI("#selected_kos3", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_kosI3","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced3()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({tree_reduced3()$tip.label[1]}))
       
       
     })
     
     
     insertUI("#kos_selection3", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("kos_selectionI3", "Show KEGG pathways", size = "sm",
                                style = "float", color = "danger")
     })
     
   })
   
   tab_kegg3 <- reactive({
     
     shinyjs::showElement(id = 'loading.ko3')
     # Create KOs set
     kos_anot <- read.csv("pharaoh_folder/ko_table_funtree.tsv", sep="\t", header = T)
     sel.genes.ko <- as.vector(isolate({input$selected_kosI3}))
     
     tab_kegg <- subset(kos_anot, gene %in% sel.genes.ko)
     set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
     
     # Show an error if no terms are identified in the input
     {
       if (length(set_kegg) == 0) 
       {
         shinyjs::hideElement(id = 'loading.ko3')
         output$error_kos3 <- renderUI({
           renderPrint({cat("0 KO terms identified. Please select more genes. If this 
        message persists, it should be interpreted as a lack of KO annotation for this orthogroup")})
         })
         
         if (UI_exist_kegg3)
         {
           removeUI(
             selector = "div:has(>> #output_kos_table3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKOSTable3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           
           UI_exist_kegg3 <<- F
         }
         
         if (UI_exist_kegg_path3)
         {
           removeUI(
             selector = "div:has(>> #output_kegg_table3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #selected_pathsI3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGTable3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI("#paths_buttonI3")
           
           UI_exist_kegg_path3 <<- F
         }
         
         if (UI_exist_pathview3)
         {
           removeUI(
             selector = "div:has(>>> #path_image3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGpathway3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           UI_exist_pathview3 <<- F
         }
         
         
         validate("No KO terms detected")
       }
     }
     
     return(tab_kegg)
     
   }) %>% bindEvent(input$kos_selectionI3)
   
   total_table_kegg3 <- reactive({
     
     tab_kegg <- tab_kegg3()
     set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
     
     # Load libraries
     library(clusterProfiler)
     library(enrichplot)
     
     # Enrich with pvalue cutoff = 1 to show all paths
     os_enrich <- enrichKEGG(gene         = unique(unlist(strsplit(set_kegg, "/"))),
                             organism     = 'ko',
                             pvalueCutoff = 1)
     
     total_table_kegg <- as.data.frame(kos_enrich)
     
     # Show an error if the KOs are not mapped to any KEGG pathway
     {
       if (nrow(total_table_kegg) == 0) 
       {
         shinyjs::hideElement(id = 'loading.ko3')
         output$error_kos3 <- renderUI({
           renderPrint({cat("No KEGG pathway appears in this OG.")})
         })
         
         
         if (UI_exist_kegg_path3)
         {
           removeUI(
             selector = "div:has(>> #output_kegg_table3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #selected_pathsI3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGTable3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI("#paths_buttonI3")
           
           UI_exist_kegg_path3 <<- F
         }
         
         if (UI_exist_pathview3)
         {
           removeUI(
             selector = "div:has(>>> #path_image3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGpathway3)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           UI_exist_pathview3 <<- F
         }
         
         validate("No KEGG pathway appears in this OG.")
       }
     }
     
     output$error_kos3 <- NULL
     
     # Filter out pathways that are not present in plants
     kegg_plants <- read.csv("pharaoh_folder/pathways_plant.ids", sep = "\t", header = T)$x
     total_table_kegg <- subset(total_table_kegg, ID %in% kegg_plants)
     
     return(total_table_kegg)
     
   }) %>% bindEvent(input$kos_selectionI3)
   
   total_table_kos3 <- reactive({
     
     tab_kegg <- tab_kegg3()
     
     library(KEGGREST)
     
     # Collapse genes that share KOs
     
     tab_kegg_for_ko <- subset(tab_kegg, ko != "")
     gene.v.ko <- c()
     individual_kos <- unique(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
     
     for (i in 1:length(individual_kos))
     {
       new.table <- subset(tab_kegg_for_ko, grepl(individual_kos[i],ko))
       new.cha <- as.character(new.table$gene)
       gene.v.ko <- c(gene.v.ko, paste(new.cha, collapse = "/"))
     }
     
     names(gene.v.ko) <- individual_kos
     
     # Create gene chains, count and KO IDs fields (same order)
     count_ko <- table(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
     geneids.ko <- gene.v.ko[names(count_ko)]
     count_terms.ko <- mapply(function(x) {keggFind("ko", x)}, names(count_ko), USE.NAMES = F)
     total_table_kos <- data.frame(ko=names(count_ko), name=count_terms.ko, count=as.numeric(count_ko),
                                   genes=geneids.ko)
     
     return(total_table_kos)
     
   }) %>% bindEvent(input$kos_selectionI3)
   
   # Create boxes for outputs
   observeEvent(isTruthy(total_table_kos3()), {
     
     if (UI_exist_kegg3)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg3 <<- F
     }
     
     insertUI("#box_kos_table3", "afterEnd", ui = {
       box(width = 12,
           title = "KO Terms Table", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_kos_table3")
       )
     })
     
     insertUI("#download_ui_for_kos_table3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKOSTable3", "Download KO Table",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_kegg3 <<- TRUE
   })
   
   observeEvent(isTruthy(total_table_kegg3()), {
     
     if (UI_exist_kegg_path3)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI3")
       
       UI_exist_kegg_path3 <<- F
     }
     
     
     insertUI("#box_kegg_table3", "afterEnd", ui = {
       box(width = 12,
           title = "KEGG Pathways Table", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_kegg_table3")
       )
     })
     
     
     
     insertUI("#download_ui_for_kegg_table3", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKEGGTable3", "Download Pathways Table",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_kegg_path3 <<- TRUE
     
     # Remove previous results for pathview
     if (UI_exist_pathview3)
     {
       removeUI(
         selector = "div:has(>>> #path_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview3 <<- F
     }
   })
   
   # Fill outputs
   # Render KO table
   output$output_kos_table3 <- renderDataTable({
     total_table_kos <- total_table_kos3()
     total_table_kos$ko <- sapply(total_table_kos$ko, ko.link)
     total_table_kos
   },escape=FALSE, rownames= F, options =list(pageLength = 5))
   
   # Render KEGG table
   output$output_kegg_table3 <- renderDataTable({
     total_table_kegg <- total_table_kegg3()
     total_table_kegg$ID <- sapply(total_table_kegg$ID, kegg.link)
     total_table_kegg[,c("ID", "Description", "geneID")]
   },escape=FALSE, rownames= F, options =list(pageLength = 5))
   
   # Download tab's outputs
   # Download KO table
   output$downloadKOSTable3 <- downloadHandler(
     filename= function() {
       paste("KO_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_kos <- total_table_kos3()
       write.table(x = total_table_kos,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download KEGG table
   output$downloadKEGGTable3 <- downloadHandler(
     filename= function() {
       paste("KEGG_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_kegg <- total_table_kegg3()
       write.table(x = total_table_kegg[,c("ID", "Description", "geneID")],quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Create pathway selector and button
   observeEvent(input$kos_selectionI3,{
     
     total_table_kos <- total_table_kos3()
     total_table_kegg <- total_table_kegg3()
     
     if(nrow(total_table_kegg) != 0)
     {
       paths.options <- sapply(strsplit(total_table_kegg$ID, split = "ko"), function(x) x[[2]])
       
       
       insertUI("#selected_paths3", "afterEnd", ui = {
         shinyWidgets::pickerInput(inputId = "selected_pathsI3", label = "Select the pathway to plot", 
                                   choices = paths.options, selected = paths.options[1], multiple = F)
         
       })
       
       
       insertUI("#paths_button3", "afterEnd", ui = {
         
         shinyWidgets::actionBttn("paths_buttonI3", "Plot Pathway", size = "sm",
                                  style = "float", color = "danger")
       })
       
       shinyjs::hideElement(id = 'loading.ko3')
     }
   })
   
   observeEvent(input$paths_buttonI3,{
     
     if (UI_exist_pathview3)
     {
       removeUI(
         selector = "div:has(>>> #path_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway3)",
         multiple = TRUE,
         immediate = TRUE
       )
     }
     
     UI_exist_pathview3 <<- F
     
   })
   
   # Create Image to Render and save path name
   pathway.current.id3 <- reactive({
     pathway.current.id <- input$selected_pathsI3
     total_table_kos <- total_table_kos3()
     
     kos_unique <- unique(total_table_kos$ko)
     gene.pathway <- rep(0, length(kos_unique))
     names(gene.pathway) <-  kos_unique
     gene.pathway[kos_unique] <-1
     
     library(pathview)
     pathview(gene.data = sort(gene.pathway,decreasing = TRUE),kegg.dir = "pharaoh_folder",
              pathway.id = pathway.current.id,
              species = "ko",
              limit = list(gene=max(abs(gene.pathway)), cpd=1),
              gene.idtype ="kegg")
     
     return(pathway.current.id)
     
   }) %>% bindEvent(input$paths_buttonI3)
   
   # Create output box and download button
   observeEvent(isTruthy(pathway.current.id3()),{
     
     insertUI("#box_path_image3", "afterEnd", ui = {
       box(width = 12,
           title = "KEGG Pathway Plot", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           fluidRow(column(1), imageOutput("path_image3", width = "100%", height = "700px"))
       )
     })
     
     insertUI("#path_download_ui3", "afterEnd", ui = {
       tags$div(shinyWidgets::downloadBttn(outputId= "downloadKEGGpathway3", "Download KEGG Pathway Plot",
                                           size = "sm", color = "danger"))
     })
     
     UI_exist_pathview3 <<- T
     
   })
   
   # Fill path image output
   output$path_image3 <- renderImage({
     
     pathway.current.id <- pathway.current.id3()
     list(src = paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."),
          contentType="image/png",width=900,height=700)
   },deleteFile = F)
   
   # Download and remove path image output
   output$downloadKEGGpathway3 <- downloadHandler(
     filename= function() {
       paste("path_plot", ".png", sep="")
     },
     content= function(file) {
       pathway.current.id <- pathway.current.id3()
       file.copy(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."), file)
       file.remove(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."))
     })   
   
   
   
   ############################# LITERATURE ANNOTATION ##########################################
   
   observeEvent(input$run_button3, {
     removeUI(
       selector = "div:has(>> #selected_litI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(> #query_litI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#lit_selectionI3")
     
     
     if (UI_exist_lit3)
     {
       removeUI(
         selector = "div:has(>> #output_lit_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadLITTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_lit3 <<- F
     }
     
   })
   
   observeEvent(input$lit_start3, {
     
     insertUI("#query_lit3", "afterEnd", ui = {
       
       textInput(inputId = "query_litI3",value = "", label = "Enter search term", placeholder = "CCA1")
       
     })
     
     
     insertUI("#selected_lit3", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_litI3","Select the search mode",
                                 choices=c("Normal","Exact", "Substring", "Alias"),
                                 multiple = F, selected = "Normal")
       
       
     })
     
     
     insertUI("#lit_selection3", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("lit_selectionI3", "Get biological information and papers", size = "sm",
                                style = "float", color = "danger")
     })
     
   })
   
   pc_result3 <- reactive({
     
     shinyjs::showElement(id = 'loading.lit3')
     pc_search <- as.character(input$query_litI3)
     pc_search <- gsub(" ", "%20", pc_search) # To make spaces interpretable
     pc_modality <- tolower(as.character(input$selected_litI3))
     
     # Get PlantConnectome URL for query
     pc_url <- paste(c("https://connectome.plant.tools", pc_modality, pc_search), collapse = "/")
     

     library(RCurl)
     
     pc_res <- getURL(pc_url)
     
     if (!length(grep("No hits", pc_res)) == 0)
     {
       shinyjs::hideElement(id = 'loading.lit3')
       
       if (UI_exist_lit3)
       {
         removeUI(
           selector = "div:has(>> #output_lit_table3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadLITTable3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_lit3 <<- F
       }
       
       output$error_lit3 <- renderUI({
         renderPrint({cat("No results found for this query")})
       })
       
       validate(" ")
       
     }
     
     output$error_lit3 <- NULL
     
     # Isolate data frame from complete HTML file
     pc_split <- strsplit(as.character(pc_res), split = "<tbody")[[1]][2]
     pc_split <- strsplit(as.character(pc_split), split = "</tbody>")[[1]][1]
     
     pc_clean <- gsub("</tr>", "", pc_split)
     pc_clean <- gsub("[\r\n\t]", "", pc_clean)
     
     pc_vector <- strsplit(pc_clean, split = "<tr>")[[1]][-1]
     pc_vector2 <- sapply(pc_vector, FUN=function(x) strsplit(x, split = "<td> | </td>"))
     pc_table <- sapply(pc_vector2, FUN=function(x) as.character(unlist(x)))
     
     colnames(pc_table) <- NULL
     pc_result <- data.frame(t(pc_table[c(2,4,6,8),]))
     colnames(pc_result) <- c("Source", "Interaction Type", "Target", "Pubmed ID")
     
     return(pc_result)
     
   }) %>% bindEvent(input$lit_selectionI3)
   
   pc_result_show3 <- reactive({
     
     pc_result <- pc_result3()
     
     # Add links to papers
     urls_connect <- sapply(pc_result$`Pubmed ID`, FUN = function(x) paste0(c("<a href=\"",
                                                                              "https://pubmed.ncbi.nlm.nih.gov/",x,"/",
                                                                              "\" target=\"_blank\">", x,
                                                                              "</a>"),
                                                                            collapse=""))
     pc_result_show <- pc_result
     pc_result_show$`Pubmed ID` <- urls_connect
     
     return(pc_result_show)
     
   })
   
   # Create boxes for outputs
   observeEvent(isTruthy(pc_result_show3()), {
     
     if (UI_exist_lit3)
     {
       removeUI(
         selector = "div:has(>> #output_lit_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadLITTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_lit_table3", "afterEnd", ui = {
       box(width = 12,
           title = "Literature Table", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_lit_table3")
       )
     })
     
     insertUI("#download_ui_for_lit_table3", "afterEnd", ui = {
       tags$div(style = "margin-left: 400px;", shinyWidgets::downloadBttn(outputId= "downloadLITTable3", "Download Literature Table",
                                                                          size = "sm", color = "danger"))
     })
     
     shinyjs::hideElement(id = 'loading.lit3')
     UI_exist_lit3 <<- TRUE
     
   })
   
   # Fill outputs
   # Render table
   output$output_lit_table3 <- renderDataTable({
     pc_result_show3()
   },escape=FALSE, rownames= F, options =list(pageLength = 10))
   
   # Download results
   output$downloadLITTable3 <- downloadHandler(
     filename= function() {
       paste("literature_table", ".tsv", sep="")
     },
     content= function(file) {
       pc_result <- pc_result3()
       write.table(x = pc_result,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })   
   
   
   ######################## STRING ###########################
   
   observeEvent(input$run_button3, {
     removeUI(
       selector = "div:has(>> #selected_stringI3)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#string_selectionI3")
     
     shinyjs::hideElement("error_string3")
     
     
     if (UI_exist_string3)
     {
       removeUI(
         selector = "div:has(>> #output_st_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #output_count_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #count_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadSTRINGTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCOUNTTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #count_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_networkI3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#network_buttonI3")
       
       UI_exist_string3 <<- F
     }
     
     if (UI_exist_network3)
     {
       removeUI(
         selector = "div:has(>>>> #network_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_network3 <<- F
     }
     
   })
   
   # First, create a table that links reduced gene names to its species
   string_sel_table3 <- reactive({
     
     # Define previous variables
     tree_reduced <- tree_reduced3()
     tree <- tree_adj3()
     
     tips_to_keep.mp <- tips_to_keep.mp3()
     tips_to_keep.ot <- tips_to_keep.ot3()
     tips_to_keep.at <- tips_to_keep.at3()
     tips_to_keep.cp <- tips_to_keep.cp3()
     tips_to_keep.cr <- tips_to_keep.cr3()
     tips_to_keep.cz <- tips_to_keep.cz3()
     tips_to_keep.kn <- tips_to_keep.kn3()
     tips_to_keep.me <- tips_to_keep.me3()
     tips_to_keep.mi <- tips_to_keep.mi3()
     tips_to_keep.pp <- tips_to_keep.pp3()
     tips_to_keep.sl <- tips_to_keep.sl3()
     tips_to_keep.sm <- tips_to_keep.sm3()
     tips_to_keep.sp <- tips_to_keep.sp3()
     tips_to_keep.ta <- tips_to_keep.ta3()
     tips_to_keep.vc <- tips_to_keep.vc3()
     tips_to_keep.bp <- tips_to_keep.bp3()
     tips_to_keep.cri <- tips_to_keep.cri3()
     tips_to_keep.ds <- tips_to_keep.ds3()
     tips_to_keep.os <- tips_to_keep.os3()
     tips_to_keep.smag <- tips_to_keep.smag3()
     tips_to_keep.tp <- tips_to_keep.tp3()
     tips_to_keep.aa <- tips_to_keep.aa3()
     tips_to_keep.um <- tips_to_keep.um3()
     tips_to_keep.rs <- tips_to_keep.rs3()
     tips_to_keep.cyc <- tips_to_keep.cyc3()
     tips_to_keep.ca <- tips_to_keep.ca3()
     tips_to_keep.mv <- tips_to_keep.mv3()
     tips_to_keep.af <- tips_to_keep.af3()
     tips_to_keep.sc <- tips_to_keep.sc3()
     tips_to_keep.aegi <- tips_to_keep.aegi3()
     tips_to_keep.sb <- tips_to_keep.sb3()
     tips_to_keep.chara <- tips_to_keep.chara3()
     tips_to_keep.sceobli <- tips_to_keep.sceobli3()
     tips_to_keep.cocco <- tips_to_keep.cocco3()
     tips_to_keep.haema <- tips_to_keep.haema3()
     tips_to_keep.zm <- tips_to_keep.zm3()
     tips_to_keep.praco <- tips_to_keep.praco3()
     tips_to_keep.ostlu <- tips_to_keep.ostlu3()
     tips_to_keep.ostme <- tips_to_keep.ostme3()
     tips_to_keep.micco <- tips_to_keep.micco3()
     tips_to_keep.cymte <- tips_to_keep.cymte3()
     tips_to_keep.chlin <- tips_to_keep.chlin3()
     tips_to_keep.gonpe <- tips_to_keep.gonpe3()
     tips_to_keep.tetso <- tips_to_keep.tetso3()
     tips_to_keep.desar <- tips_to_keep.desar3()
     tips_to_keep.monmi <- tips_to_keep.monmi3()
     tips_to_keep.caule <- tips_to_keep.caule3()
     tips_to_keep.ostqu <- tips_to_keep.ostqu3()
     tips_to_keep.brysp <- tips_to_keep.brysp3()
     tips_to_keep.ellbi <- tips_to_keep.ellbi3()
     tips_to_keep.myrbi <- tips_to_keep.myrbi3()
     tips_to_keep.apalo <- tips_to_keep.apalo3()
     tips_to_keep.astgl <- tips_to_keep.astgl3()
     tips_to_keep.tresp <- tips_to_keep.tresp3()
     tips_to_keep.picso <- tips_to_keep.picso3()
     tips_to_keep.chlso <- tips_to_keep.chlso3()
     tips_to_keep.auxpr <- tips_to_keep.auxpr3()
     tips_to_keep.helsp <- tips_to_keep.helsp3()
     tips_to_keep.tetst <- tips_to_keep.tetst3()
     tips_to_keep.pedsp <- tips_to_keep.pedsp3()
     tips_to_keep.chlpr <- tips_to_keep.chlpr3()
     tips_to_keep.picsp <- tips_to_keep.picsp3()
     tips_to_keep.mesvb <- tips_to_keep.mesvb3()
     tips_to_keep.meskr <- tips_to_keep.meskr3()
     tips_to_keep.zygci <- tips_to_keep.zygci3()
     tips_to_keep.zygcb <- tips_to_keep.zygcb3()
     tips_to_keep.zygcy <- tips_to_keep.zygcy3()
     tips_to_keep.plesc <- tips_to_keep.plesc3()
     tips_to_keep.funhy <- tips_to_keep.funhy3()
     tips_to_keep.sphfa <- tips_to_keep.sphfa3()
     tips_to_keep.takle <- tips_to_keep.takle3()
     tips_to_keep.marpa <- tips_to_keep.marpa3()
     tips_to_keep.ricfl <- tips_to_keep.ricfl3()
     tips_to_keep.ricso <- tips_to_keep.ricso3()
     tips_to_keep.antpu <- tips_to_keep.antpu3()
     tips_to_keep.antfu <- tips_to_keep.antfu3()
     tips_to_keep.megfl <- tips_to_keep.megfl3()
     tips_to_keep.phach <- tips_to_keep.phach3()
     tips_to_keep.phyph <- tips_to_keep.phyph3()
     tips_to_keep.parpe <- tips_to_keep.parpe3()
     tips_to_keep.notor <- tips_to_keep.notor3()
     tips_to_keep.phaca <- tips_to_keep.phaca3()
     tips_to_keep.phasp <- tips_to_keep.phasp3()
     tips_to_keep.leidu <- tips_to_keep.leidu3()
     tips_to_keep.isosi <- tips_to_keep.isosi3()
     tips_to_keep.dipco <- tips_to_keep.dipco3()
     tips_to_keep.hupas <- tips_to_keep.hupas3()
     tips_to_keep.lyccl <- tips_to_keep.lyccl3()
     tips_to_keep.adica <- tips_to_keep.adica3()
     tips_to_keep.alssp <- tips_to_keep.alssp3()
     tips_to_keep.marve <- tips_to_keep.marve3()
     tips_to_keep.seqse <- tips_to_keep.seqse3()
     tips_to_keep.seqgi <- tips_to_keep.seqgi3()
     tips_to_keep.taxch <- tips_to_keep.taxch3()
     tips_to_keep.abial <- tips_to_keep.abial3()
     tips_to_keep.picab <- tips_to_keep.picab3()
     tips_to_keep.gnemo <- tips_to_keep.gnemo3()
     tips_to_keep.welmi <- tips_to_keep.welmi3()
     tips_to_keep.ginbi <- tips_to_keep.ginbi3()
     tips_to_keep.ambtr <- tips_to_keep.ambtr3()
     tips_to_keep.nymco <- tips_to_keep.nymco3()
     tips_to_keep.cinka <- tips_to_keep.cinka3()
     tips_to_keep.horvu <- tips_to_keep.horvu3()
     tips_to_keep.bradi <- tips_to_keep.bradi3()
     tips_to_keep.setit <- tips_to_keep.setit3()
     tips_to_keep.panha <- tips_to_keep.panha3()
     tips_to_keep.missi <- tips_to_keep.missi3()
     tips_to_keep.oroth <- tips_to_keep.oroth3()
     tips_to_keep.eleco <- tips_to_keep.eleco3()
     tips_to_keep.anaco <- tips_to_keep.anaco3()
     tips_to_keep.musac <- tips_to_keep.musac3()
     tips_to_keep.dioal <- tips_to_keep.dioal3()
     tips_to_keep.spipo <- tips_to_keep.spipo3()
     tips_to_keep.aspof <- tips_to_keep.aspof3()
     tips_to_keep.zosma <- tips_to_keep.zosma3()
     tips_to_keep.araly <- tips_to_keep.araly3()
     tips_to_keep.capgr <- tips_to_keep.capgr3()
     tips_to_keep.brara <- tips_to_keep.brara3()
     tips_to_keep.braol <- tips_to_keep.braol3()
     tips_to_keep.sisir <- tips_to_keep.sisir3()
     tips_to_keep.schpa <- tips_to_keep.schpa3()
     tips_to_keep.eutsa <- tips_to_keep.eutsa3()
     tips_to_keep.thlar <- tips_to_keep.thlar3()
     tips_to_keep.boest <- tips_to_keep.boest3()
     tips_to_keep.carpa <- tips_to_keep.carpa3()
     tips_to_keep.theca <- tips_to_keep.theca3()
     tips_to_keep.goshi <- tips_to_keep.goshi3()
     tips_to_keep.gosra <- tips_to_keep.gosra3()
     tips_to_keep.citsi <- tips_to_keep.citsi3()
     tips_to_keep.pontr <- tips_to_keep.pontr3()
     tips_to_keep.eucgr <- tips_to_keep.eucgr3()
     tips_to_keep.maldo <- tips_to_keep.maldo3()
     tips_to_keep.prupe <- tips_to_keep.prupe3()
     tips_to_keep.frave <- tips_to_keep.frave3()
     tips_to_keep.corav <- tips_to_keep.corav3()
     tips_to_keep.caril <- tips_to_keep.caril3()
     tips_to_keep.queru <- tips_to_keep.queru3()
     tips_to_keep.cucsa <- tips_to_keep.cucsa3()
     tips_to_keep.glyma <- tips_to_keep.glyma3()
     tips_to_keep.medtr <- tips_to_keep.medtr3()
     tips_to_keep.tripr <- tips_to_keep.tripr3()
     tips_to_keep.cicar <- tips_to_keep.cicar3()
     tips_to_keep.lotja <- tips_to_keep.lotja3()
     tips_to_keep.phaac <- tips_to_keep.phaac3()
     tips_to_keep.vigun <- tips_to_keep.vigun3()
     tips_to_keep.lupal <- tips_to_keep.lupal3()
     tips_to_keep.lencu <- tips_to_keep.lencu3()
     tips_to_keep.arahy <- tips_to_keep.arahy3()
     tips_to_keep.poptr <- tips_to_keep.poptr3()
     tips_to_keep.manes <- tips_to_keep.manes3()
     tips_to_keep.salpu <- tips_to_keep.salpu3()
     tips_to_keep.linus <- tips_to_keep.linus3()
     tips_to_keep.corci <- tips_to_keep.corci3()
     tips_to_keep.vitvi <- tips_to_keep.vitvi3()
     tips_to_keep.kalfe <- tips_to_keep.kalfe3()
     tips_to_keep.vacda <- tips_to_keep.vacda3()
     tips_to_keep.oleeu <- tips_to_keep.oleeu3()
     tips_to_keep.mimgu <- tips_to_keep.mimgu3()
     tips_to_keep.soltu <- tips_to_keep.soltu3()
     tips_to_keep.cofar <- tips_to_keep.cofar3()
     tips_to_keep.lacsa <- tips_to_keep.lacsa3()
     tips_to_keep.dauca <- tips_to_keep.dauca3()
     tips_to_keep.betvu <- tips_to_keep.betvu3()
     tips_to_keep.amahy <- tips_to_keep.amahy3()
     tips_to_keep.chequ <- tips_to_keep.chequ3()
     tips_to_keep.sapof <- tips_to_keep.sapof3()
     tips_to_keep.aquco <- tips_to_keep.aquco3()
     
     # Table construction
     org.factor <- c()
     
     for (i in 1:length(tree_reduced$tip.label))
     {
       if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
       {
         org.factor <- c(org.factor,"Marchantia polymorpha")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
       {
         org.factor <- c(org.factor,"Ostreococcus tauri")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
       {
         org.factor <- c(org.factor,"Arabidopsis thaliana")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
       {
         org.factor <- c(org.factor,"Ceratodon purpureus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
       {
         org.factor <- c(org.factor,"Chlamydomonas reinhardtii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
       {
         org.factor <- c(org.factor,"Chromochloris zofingiensis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
       {
         org.factor <- c(org.factor,"Klebsormidium nitens")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
       {
         org.factor <- c(org.factor,"Mesotaenium endlicherianum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
       {
         org.factor <- c(org.factor,"Micromonas pusilla")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
       {
         org.factor <- c(org.factor,"Physcomitrium patens")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
       {
         org.factor <- c(org.factor,"Solanum lycopersicum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
       {
         org.factor <- c(org.factor,"Selaginella moellendorffii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
       {
         org.factor <- c(org.factor,"Spirogloea muscicola")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
       {
         org.factor <- c(org.factor,"Triticum aestivum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
       {
         org.factor <- c(org.factor,"Volvox carteri")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
       {
         org.factor <- c(org.factor,"Bathycoccus prasinos")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
       {
         org.factor <- c(org.factor,"Ceratopteris richardii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
       {
         org.factor <- c(org.factor,"Dunaliella salina")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
       {
         org.factor <- c(org.factor,"Oryza sativa")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
       {
         org.factor <- c(org.factor,"Sphagnum magellanicum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
       {
         org.factor <- c(org.factor,"Thuja plicata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
       {
         org.factor <- c(org.factor,"Anthoceros agrestis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
       {
         org.factor <- c(org.factor,"Ulva mutabilis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
       {
         org.factor <- c(org.factor,"Raphidocelis subcapitata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
       {
         org.factor <- c(org.factor,"Cycas panzhihuaensis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
       {
         org.factor <- c(org.factor,"Chlorokybus atmophyticus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
       {
         org.factor <- c(org.factor,"Mesostigma viride CCAC 1140")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
       {
         org.factor <- c(org.factor,"Azolla filiculoides")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
       {
         org.factor <- c(org.factor,"Salvinia cucullata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
       {
         org.factor <- c(org.factor,"Aegilops tauschii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
       {
         org.factor <- c(org.factor,"Sorghum bicolor")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
       {
         org.factor <- c(org.factor,"Chara braunii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
       {
         org.factor <- c(org.factor,"Scenedesmus obliquus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
       {
         org.factor <- c(org.factor,"Coccomyxa subellipsoidea")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
       {
         org.factor <- c(org.factor,"Haematococcus lacustris")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
       {
         org.factor <- c(org.factor,"Zea mays")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.praco])
       {
         org.factor <- c(org.factor,"Prasinoderma coloniale")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostlu])
       {
         org.factor <- c(org.factor,"Ostreococcus lucimarinus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostme])
       {
         org.factor <- c(org.factor,"Ostreococcus mediterraneus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.micco])
       {
         org.factor <- c(org.factor,"Micromonas commoda")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymte])
       {
         org.factor <- c(org.factor,"Cymbomonas tetramitiformis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlin])
       {
         org.factor <- c(org.factor,"Chlamydomonas incerta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gonpe])
       {
         org.factor <- c(org.factor,"Gonium pectorale")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetso])
       {
         org.factor <- c(org.factor,"Tetrabaena socialis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.desar])
       {
         org.factor <- c(org.factor,"Desmodesmus armatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.monmi])
       {
         org.factor <- c(org.factor,"Monoraphidium minutum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caule])
       {
         org.factor <- c(org.factor,"Caulerpa lentillifera")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostqu])
       {
         org.factor <- c(org.factor,"Ostreobium quekettii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brysp])
       {
         org.factor <- c(org.factor,"Bryopsis sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ellbi])
       {
         org.factor <- c(org.factor,"Elliptochloris bilobata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.myrbi])
       {
         org.factor <- c(org.factor,"Myrmecia bisecta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.apalo])
       {
         org.factor <- c(org.factor,"Apatococcus lobatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.astgl])
       {
         org.factor <- c(org.factor,"Asterochloris glomerata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tresp])
       {
         org.factor <- c(org.factor,"Trebouxia sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picso])
       {
         org.factor <- c(org.factor,"Picochlorum soloecismus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlso])
       {
         org.factor <- c(org.factor,"Chlorella sorokiniana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.auxpr])
       {
         org.factor <- c(org.factor,"Auxenochlorella protothecoides")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.helsp])
       {
         org.factor <- c(org.factor,"Helicosporidium sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetst])
       {
         org.factor <- c(org.factor,"Tetraselmis striata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pedsp])
       {
         org.factor <- c(org.factor,"Pedinophyceae sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlpr])
       {
         org.factor <- c(org.factor,"Chloropicon primus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picsp])
       {
         org.factor <- c(org.factor,"Picocystis sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mesvb])
       {
         org.factor <- c(org.factor,"Mesostigma viride NIES 296")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.meskr])
       {
         org.factor <- c(org.factor,"Mesotaenium kramstae")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygci])
       {
         org.factor <- c(org.factor,"Zygnema circumcarinatum SAG")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcb])
       {
         org.factor <- c(org.factor,"Zygnema circumcarinatum UTEX 1559")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcy])
       {
         org.factor <- c(org.factor,"Zygnema cylindricum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.plesc])
       {
         org.factor <- c(org.factor,"Pleurozium schreberi")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.funhy])
       {
         org.factor <- c(org.factor,"Funaria hygrometrica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sphfa])
       {
         org.factor <- c(org.factor,"Sphagnum fallax")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.takle])
       {
         org.factor <- c(org.factor,"Takakia lepidozioides")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marpa])
       {
         org.factor <- c(org.factor,"Marchantia paleacea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricfl])
       {
         org.factor <- c(org.factor,"Riccia fluitans")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricso])
       {
         org.factor <- c(org.factor,"Riccia sorocarpa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antpu])
       {
         org.factor <- c(org.factor,"Anthoceros punctatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antfu])
       {
         org.factor <- c(org.factor,"Anthoceros fusiformis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.megfl])
       {
         org.factor <- c(org.factor,"Megaceros flagellaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phach])
       {
         org.factor <- c(org.factor,"Phaeomegaceros chiloensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phyph])
       {
         org.factor <- c(org.factor,"Phymatoceros phymatodes")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.parpe])
       {
         org.factor <- c(org.factor,"Paraphymatoceros pearsonii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.notor])
       {
         org.factor <- c(org.factor,"Notothylas orbicularis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaca])
       {
         org.factor <- c(org.factor,"Phaeoceros carolinianus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phasp])
       {
         org.factor <- c(org.factor,"Phaeoceros sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.leidu])
       {
         org.factor <- c(org.factor,"Leiosporoceros dussii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.isosi])
       {
         org.factor <- c(org.factor,"Isoetes sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dipco])
       {
         org.factor <- c(org.factor,"Diphasiastrum complanatum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.hupas])
       {
         org.factor <- c(org.factor,"Huperzia asiatica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lyccl])
       {
         org.factor <- c(org.factor,"Lycopodium clavatum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.adica])
       {
         org.factor <- c(org.factor,"Adiantum capillus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.alssp])
       {
         org.factor <- c(org.factor,"Alsophila spinulosa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marve])
       {
         org.factor <- c(org.factor,"Marsilea vestita")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqse])
       {
         org.factor <- c(org.factor,"Sequoia sempervirens")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqgi])
       {
         org.factor <- c(org.factor,"Sequoiadendron giganteum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.taxch])
       {
         org.factor <- c(org.factor,"Taxus chinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.abial])
       {
         org.factor <- c(org.factor,"Abies alba")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picab])
       {
         org.factor <- c(org.factor,"Picea abies")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gnemo])
       {
         org.factor <- c(org.factor,"Gnetum montanum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.welmi])
       {
         org.factor <- c(org.factor,"Welwitschia mirabilis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ginbi])
       {
         org.factor <- c(org.factor,"Ginkgo biloba")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ambtr])
       {
         org.factor <- c(org.factor,"Amborella trichopoda")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.nymco])
       {
         org.factor <- c(org.factor,"Nymphaea colorata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cinka])
       {
         org.factor <- c(org.factor,"Cinnamomum kanehirae")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.horvu])
       {
         org.factor <- c(org.factor,"Hordeum vulgare")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bradi])
       {
         org.factor <- c(org.factor,"Brachypodium distachyon")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.setit])
       {
         org.factor <- c(org.factor,"Setaria italica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.panha])
       {
         org.factor <- c(org.factor,"Panicum hallii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.missi])
       {
         org.factor <- c(org.factor,"Miscanthus sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oroth])
       {
         org.factor <- c(org.factor,"Oropetium thomaeum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eleco])
       {
         org.factor <- c(org.factor,"Eleusine coracana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.anaco])
       {
         org.factor <- c(org.factor,"Ananas comosus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.musac])
       {
         org.factor <- c(org.factor,"Musa acuminata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dioal])
       {
         org.factor <- c(org.factor,"Dioscorea alata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.spipo])
       {
         org.factor <- c(org.factor,"Spirodela polyrhiza")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aspof])
       {
         org.factor <- c(org.factor,"Asparagus officinalis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zosma])
       {
         org.factor <- c(org.factor,"Zostera marina")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.araly])
       {
         org.factor <- c(org.factor,"Arabidopsis lyrata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.capgr])
       {
         org.factor <- c(org.factor,"Capsella grandiflora")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brara])
       {
         org.factor <- c(org.factor,"Brassica rapa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.braol])
       {
         org.factor <- c(org.factor,"Brassica oleracea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sisir])
       {
         org.factor <- c(org.factor,"Sisymbrium irio")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.schpa])
       {
         org.factor <- c(org.factor,"Schrenkiella parvula")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eutsa])
       {
         org.factor <- c(org.factor,"Eutrema salsugineum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.thlar])
       {
         org.factor <- c(org.factor,"Thlaspi arvense")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.boest])
       {
         org.factor <- c(org.factor,"Boechera stricta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.carpa])
       {
         org.factor <- c(org.factor,"Carica papaya")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.theca])
       {
         org.factor <- c(org.factor,"Theobroma cacao")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.goshi])
       {
         org.factor <- c(org.factor,"Gossypium hirsutum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gosra])
       {
         org.factor <- c(org.factor,"Gossypium raimondii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.citsi])
       {
         org.factor <- c(org.factor,"Citrus sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pontr])
       {
         org.factor <- c(org.factor,"Poncirus trifoliata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eucgr])
       {
         org.factor <- c(org.factor,"Eucalyptus grandis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.maldo])
       {
         org.factor <- c(org.factor,"Malus domestica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.prupe])
       {
         org.factor <- c(org.factor,"Prunus persica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.frave])
       {
         org.factor <- c(org.factor,"Fragaria vesca")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corav])
       {
         org.factor <- c(org.factor,"Corylus avellana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caril])
       {
         org.factor <- c(org.factor,"Carya illinoinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.queru])
       {
         org.factor <- c(org.factor,"Quercus rubra")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cucsa])
       {
         org.factor <- c(org.factor,"Cucumis sativus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.glyma])
       {
         org.factor <- c(org.factor,"Glycine max")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.medtr])
       {
         org.factor <- c(org.factor,"Medicago truncatula")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tripr])
       {
         org.factor <- c(org.factor,"Trifolium pratense")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cicar])
       {
         org.factor <- c(org.factor,"Cicer arietinum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lotja])
       {
         org.factor <- c(org.factor,"Lotus japonicus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaac])
       {
         org.factor <- c(org.factor,"Phaseolus acutifolius")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vigun])
       {
         org.factor <- c(org.factor,"Vigna unguiculata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lupal])
       {
         org.factor <- c(org.factor,"Lupinus albus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lencu])
       {
         org.factor <- c(org.factor,"Lens culinaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.arahy])
       {
         org.factor <- c(org.factor,"Arachis hypogaea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.poptr])
       {
         org.factor <- c(org.factor,"Populus trichocarpa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.manes])
       {
         org.factor <- c(org.factor,"Manihot esculenta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.salpu])
       {
         org.factor <- c(org.factor,"Salix purpurea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.linus])
       {
         org.factor <- c(org.factor,"Linum usitatissimum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corci])
       {
         org.factor <- c(org.factor,"Corymbia citriodora")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vitvi])
       {
         org.factor <- c(org.factor,"Vitis vinifera")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kalfe])
       {
         org.factor <- c(org.factor,"Kalanchoe fedtschenkoi")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vacda])
       {
         org.factor <- c(org.factor,"Vaccinium darrowii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oleeu])
       {
         org.factor <- c(org.factor,"Olea europaea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mimgu])
       {
         org.factor <- c(org.factor,"Mimulus guttatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.soltu])
       {
         org.factor <- c(org.factor,"Solanum tuberosum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cofar])
       {
         org.factor <- c(org.factor,"Coffea arabica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lacsa])
       {
         org.factor <- c(org.factor,"Lactuca sativa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dauca])
       {
         org.factor <- c(org.factor,"Daucus carota")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.betvu])
       {
         org.factor <- c(org.factor,"Beta vulgaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.amahy])
       {
         org.factor <- c(org.factor,"Amaranthus hypochondriacus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chequ])
       {
         org.factor <- c(org.factor,"Chenopodium quinoa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sapof])
       {
         org.factor <- c(org.factor,"Saponaria officinalis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aquco])
       {
         org.factor <- c(org.factor,"Aquilegia coerulea")
       }
       
       
     }
     
     #Matrix with labels and colors and transform to dplyr format
     string.sel.table <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                                    org = org.factor)
     
     return(string.sel.table)
     
   }) %>% bindEvent(input$string_start3)
   
   # Now, selection is allowed for genes of species with STRING support
   observeEvent(input$string_start3, {
     
     string_sel_table <- string_sel_table3()
     allow_string_species <- c("Aegilops tauschii", "Arabidopsis thaliana", "Bathycoccus prasinos", "Chara braunii", 
                               "Chlamydomonas reinhardtii","Coccomyxa subellipsoidea", "Klebsormidium nitens", 
                               "Micromonas commoda", "Oryza sativa", "Micromonas pusilla", "Ostreococcus lucimarinus",
                               "Ostreococcus tauri", "Physcomitrium patens", "Raphidocelis subcapitata",
                               "Scenedesmus obliquus", "Selaginella moellendorffii", "Solanum lycopersicum",
                               "Sorghum bicolor", "Triticum aestivum", "Volvox carteri")
     
     st_genes <- subset(string_sel_table, string_sel_table$org %in% allow_string_species)$label
     
     if (length(st_genes) == 0)
     {
       shinyjs::showElement("error_string3")
       output$error_string3 <- renderUI({renderText({print("No results for this
      analysis due to lack of genes of STRING-supported species in the selection.")})})
       validate(" ")
     }
     
     output$error_string3 <- NULL
     
     insertUI("#selected_string3", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_stringI3","Select the desired genes from the tree",
                                 choices=st_genes, options = list(`actions-box` = TRUE),
                                 multiple = T)
       
     })
     
     
     insertUI("#string_selection3", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("string_selectionI3", "Show STRING Interactions", size = "sm",
                                style = "float", color = "danger")
     })
     
   })
   
   phys_table3 <- reactive({
     
     shinyjs::showElement(id = 'loading.string3')
     query_genes <- input$selected_stringI3
     string_sel_table <- string_sel_table3()
     
     # Load complete STRING annotation table
     library(data.table)
     
     # Load iteratively from split files using data.table format
     # data_phys <- fread("pharaoh_folder/string_physical/string_physical_1.tsv")
     # 
     # for (x in list.files("pharaoh_folder/string_physical/")[-1])
     # {
     #   data_phys <- rbind(data_phys, 
     #                      fread(paste0("pharaoh_folder/string_physical/", x)))
     # }
     
     # Load data only for selected organisms
     species_selected <- gsub("[.]", "", gsub(" ", "_", tolower(unique(string_sel_table$org))))
     
     # Get a List of all files named with a key word, say all `.csv` files
     filenames_string <- list.files("pharaoh_folder/string_physical/")
     files_selected <- filenames_string[unlist(sapply(species_selected, function(x) grep(x, filenames_string)))]
     
     # If in the future any of these species is added to STRING:
     # { 
     #   if (seq_org == "Zygnema circumcarinatum SAG")
     #   {
     #     org_search <- "zygnema_circumcarinatum"
     #   }
     #   else if (seq_org == "Zygnema circumcarinatum UTEX 1559")
     #   {
     #     org_search <- "zygnema_circumcarinatum1559"
     #   }
     #   else if (seq_org == "Mesostigma viride CCAC 1140")
     #   {
     #     org_search <- "mesostigma_viride1140"
     #   }
     #   else if (seq_org == "Mesostigma viride NIES 296")
     #   {
     #     org_search <- "mesostigma_viride296"
     #   }
     #   else
     #   {
     #    org_search <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(seq_org))))
     #   }
     # }
     
     # Load and bind all data sets
     data_phys <- rbindlist(lapply(paste0("pharaoh_folder/string_physical/", files_selected),fread))
     
     # Subset by query genes using data.table for speed
     #string_res <- subset(data_phys, data_phys$prot_query %in% query_genes)
     string_res <- data_phys[prot_query %in% query_genes,]
     string_res <- as.data.frame(string_res)
     
     # Assign OG ID to each target
     ortho_data_file <- "Gene_Trees/Orthogroups.tsv"
     
     ortho_data <- as.data.frame(fread(ortho_data_file))
     
     ortho_char <- apply(ortho_data, MARGIN = 1, function(x) paste(x, collapse = ","))
     ortho.numbers <- sapply(string_res$prot_interaction, function(x) grep(x, ortho_char), USE.NAMES = F)
     
     # If a pattern is found in several names, i.e., is a subpattern of several genes,
     # search for the exact match
     if(class(ortho.numbers) == "list")
     {
       # If a gene isn't associated to an OG
       index.none <- which(sapply(ortho.numbers, function(x) length(x) == 0))
       
       if (length(index.none) != 0)
       {
         # We create another row for OG table to associate those genes
         ortho_data <- rbind(ortho_data, "No OG")
         ortho.numbers[index.none] <- nrow(ortho_data)
       }
       
       # If a gene has more than one match due to subpatterns
       index.wrong <- which(sapply(ortho.numbers, function(x) length(x) > 1))
       {
         if (length(index.wrong) == 0)
         {
           ortho.numbers <- unlist(ortho.numbers, use.names = F)
         }
         else
         {
           for (i in index.wrong)
           {
             for (j in ortho.numbers[[i]])
             {
               ortho_char_split <- strsplit(ortho_char[j],split = ",")
               ortho_char_split_clean <- sapply(ortho_char_split, function(x) gsub(" ", "", x))
               if (string_res$prot_interaction[i] %in% ortho_char_split_clean)
               {
                 ortho.numbers[i] <- j
                 ortho.numbers <- unlist(ortho.numbers, use.names = F)
               }
             }
           }
         }
       }
     }
     
     
     
     # Create the final table
     ortho.string.names <- ortho_data$Orthogroup[ortho.numbers]
     phys_table <- data.frame(string_res, orthogroup = ortho.string.names)
     
     return(phys_table)
   }) %>% bindEvent(input$string_selectionI3)
   
   # Create count table to identify enriched OGs in STRING result
   string_counts3 <- reactive({
     
     phys_table <- phys_table3()
     string_counts <- sort(table(phys_table$orthogroup), decreasing = T)
     
     return(string_counts)
     
   }) %>% bindEvent(input$string_selectionI3)
   
   string_count_plot3 <- reactive({
     
     library(ggplot2)
     library(dplyr)
     
     data_count <- as.data.frame(string_counts3())
     colnames(data_count) <- c("orthogroup", "value")
     
     # Compute the position of labels
     data_count <- data_count %>%
       arrange(desc(orthogroup)) %>%
       mutate(prop = value / sum(data_count$value) *100) %>%
       mutate(ypos = cumsum(prop)- 0.5*prop )
     
     # Create plot
     count_plot <- ggplot(data_count, aes(x="", y=prop, fill=orthogroup)) +
       geom_bar(stat="identity", width=1, color="white") +
       coord_polar("y", start=0) +
       theme_void() +
       theme(legend.position="none") +
       #geom_text(aes(y = ypos, label = orthogroup), color = "white", size=6) +
       scale_fill_manual(values = rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"), 
                                      floor(nrow(data_count)/9)+1))
     
     return(count_plot)
     
   }) %>% bindEvent(input$string_selectionI3)
   
   string_count_plotly3 <- reactive({
     
     library(ggplot2)
     library(dplyr)
     
     data_count <- as.data.frame(string_counts3())
     colnames(data_count) <- c("orthogroup", "value")
     
     # Compute the position of labels
     data_count <- data_count %>%
       arrange(desc(orthogroup)) %>%
       mutate(prop = value / sum(data_count$value) *100) %>%
       mutate(ypos = cumsum(prop)- 0.5*prop )
     
     # Create plot
     count_plotly <- plotly::plot_ly(data=data_count,values=~prop,labels=~factor(orthogroup),
                                     marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                                            floor(nrow(data_count)/9)+1)),
                                     type="pie",showlegend = F, text= ~paste0("</br> ", orthogroup,
                                                                              "</br> ",prop, "%"),
                                     textinfo = "none", hoverinfo = "text") 
     
     
     
     return(count_plotly)
     
   }) %>% bindEvent(input$string_selectionI3)
   
   # Create boxes for outputs
   observeEvent(isTruthy(string_count_plot3()), {
     
     if (UI_exist_string3)
     {
       removeUI(
         selector = "div:has(>> #output_st_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #output_count_table3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #count_plot3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadSTRINGTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCOUNTTable3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #count_download3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_networkI3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#network_buttonI3")
       
     }
     
     insertUI("#box_st_table3", "afterEnd", ui = {
       box(width = 12,
           title = "STRING Interactions Table", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_st_table3")
       )
     })
     
     insertUI("#box_count_table3", "afterEnd", ui = {
       box(width = 12,
           title = "Interacting Orthogroups Table", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_count_table3")
       )
     })
     
     insertUI("#box_count_plot3", "afterEnd", ui = {
       box(width = 12,
           title = "Interacting Orthogroups Plot", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           plotlyOutput("count_plot3")
       )
     })
     
     
     insertUI("#download_ui_for_st_table3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadSTRINGTable3", "Download STRING Table",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#download_ui_for_count_table3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadCOUNTTable3", "Download OG Count Table",
                                                                          size = "sm", color = "danger"))
     })
     
     insertUI("#count_down_button3", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "count_download3", "Download OG Count Plot",
                                                                          size = "sm", color = "danger"))
     })
     
     UI_exist_string3 <<- TRUE
     
     # Remove previous results for STRING network
     if (UI_exist_network3)
     {
       removeUI(
         selector = "div:has(>>>> #network_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_network3 <<- F
     }
     
   })
   
   # Fill outputs
   # Render STRING table
   output$output_st_table3 <- renderDataTable({
     phys_table <- phys_table3()
     
     datatable(phys_table, escape=FALSE, rownames= F, options =list(pageLength = 10)) %>%
       formatStyle(
         'type',
         color = styleEqual(
           c("Direct interaction", "Interolog"), c('green', '#CA931B')
         )
       )
   }) 
   
   
   # Render OG count table
   output$output_count_table3 <- renderDataTable({
     string_counts <- as.data.frame(string_counts3())
     colnames(string_counts) <- c("orthogroup", "count")
     string_counts
   },escape=FALSE, rownames= F, options =list(pageLength = 7))
   
   # Render OG count pie chart
   output$count_plot3 <- renderPlotly({
     string_count_plotly3()
   })
   
   
   # Download tab's outputs
   # Download STRING table
   output$downloadSTRINGTable3 <- downloadHandler(
     filename= function() {
       paste("string_table", ".tsv", sep="")
     },
     content= function(file) {
       phys_table <- phys_table3()
       write.table(x = phys_table,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download count table
   output$downloadCOUNTTable3 <- downloadHandler(
     filename= function() {
       paste("string_count_table", ".tsv", sep="")
     },
     content= function(file) {
       string_counts <- string_counts3()
       write.table(x = string_counts,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download count plot
   output$count_download3 <- downloadHandler(
     filename= function() {
       paste("string_count", ".png", sep="")
     },
     content= function(file) {
       string_count_plot <- string_count_plot3()
       
       png(file, height = 450, width = 450)
       plot(string_count_plot)
       dev.off()
     })
   
   # Create gene selector and button for STRING network representation
   observeEvent(input$string_selectionI3,{
     
     phys_table <- phys_table3()
     network_genes <- unique(phys_table$prot_query)
     
     # Error message if no genes are allowed for selection should have  been reported
     # earlier
     
     insertUI("#selected_network3", "afterEnd", ui = {
       
       shinyWidgets::pickerInput(inputId = "selected_networkI3", label = "Select the gene whose network you want to plot", 
                                 choices = network_genes, selected = network_genes[1],
                                 options = list(`actions-box` = TRUE), multiple = T)
       
     })
     
     
     insertUI("#network_button3", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("network_buttonI3", "Plot STRING Network", size = "sm",
                                style = "float", color = "danger")
     })
     
     shinyjs::hideElement(id = 'loading.string3')
     
   })
   
   mapped_string3 <- reactive({
     
     # Load genes (PharaohFUN IDs) and convert to STRING IDs
     library(data.table)
     
     network_genes <- input$selected_networkI3
     map_table <- fread("pharaoh_folder/string_map.tsv")
     
     # Fast subset using data.table
     map_network <- map_table[pharaohfun_id %in% network_genes,]
     
     # Error if no genes from selection have an associated high fidelity STRING ID
     if (nrow(map_network) == 0)
     {
       
       if (UI_exist_network3)
       {
         removeUI(
           selector = "div:has(>>>> #network_image3)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_network3 <<- F
       }
       
       output$error_network3 <- renderUI({renderText({print("It's not possible to map any
                               genes from selection to STRING IDs, please select different ones.")})})
       validate( " ")
       
     }
     
     output$error_network3 <- NULL
     
     # Get only STRING IDS and paste in a format interpretable by JS function
     string_ids <- as.data.frame(map_network)$string_id
     
     
     mapped_string <- paste0(string_ids, collapse = "%0d")
     return(mapped_string)
     
   }) %>% bindEvent(input$network_buttonI3)
   
   # Create boxes
   observeEvent(isTruthy(mapped_string3()),{
     
     mapped_string <- mapped_string3()
     
     if (UI_exist_network3)
     {
       removeUI(
         selector = "div:has(>>>> #network_image3)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     url_interactive <- paste0("https://string-db.org/cgi/network?identifiers=", 
                               mapped_string, 
                               "&add_color_nodes=25&network_flavor=confidence&show_query_node_labels=1")
     
     insertUI("#box_output_network3", "afterEnd", ui = {
       box(width = 12,
           title = "Image", status = "danger", solidHeader = TRUE,
           collapsible = TRUE,
           fluidRow(column(1), 
                    column(8, htmlOutput("network_image3")), 
                    column(3, div( style = "margin-top: 300px;", 
                                   shinyWidgets::actionBttn("network_link3", "Interactive Network", size = "md", 
                                                            icon = icon("circle-nodes"), style = "float", color = "danger", 
                                                            onclick=paste0("window.open('", url_interactive,"','_blank')")))
                           
                    ))
           
       )
     })
     
     UI_exist_network3 <<- T
     
   })
   
   # Fill network box
   
   output$network_image3 <- renderText({
     
     mapped_string <- mapped_string3()
     src_map <- paste0("https://string-db.org/api/image/network?identifiers=", mapped_string, 
                       "&add_color_nodes=25&network_flavor=confidence")
     
     c('<img src="',src_map,'"width="675" height="625">')
   })
   
# End of OG-based search results
   
##################### BATCH SEARCH ##########################
   
   # Set global variables for tracking changes in output
   UI_exist_batch4 <<- F
   
   # To avoid autoupdating some inputs, define variables with its values
   model.selected4 <- reactive({
     model.selected <- F
     return(model.selected)
   })%>% bindEvent(input$run_button4)
   
   search_mode4 <- reactive({
     search_mode <- input$search_mode4
     return(search_mode)
   }) %>% bindEvent(input$run_button4)
   
   seq_org4 <- reactive({
     seq_org <- input$organism_input_4
     return(seq_org)
     
   }) %>% bindEvent(input$run_button4)
   
   # Load organisms selection based on the model selected
   selected_organisms4 <- reactive({
     selected_organisms <- c(input$prasino_check_4, input$mami_check_4,
                             input$chloro_check_4, input$strepto_check_4,
                             input$bryo_check_4, input$lyco_check_4,
                             input$gymno_check_4, input$magno_check_4,
                             input$monocots_check_4, input$dicots_first_check_4,
                             input$dicots_second_check_4)
     return(selected_organisms)
     
   }) %>% bindEvent(input$run_button4)
   
   selected_values_org4 <- reactive(organisms_values[selected_organisms4()]) %>% bindEvent(input$run_button4)
   
   # Generate random name
   random.file4 <- reactive({
     
     random.file <- paste0(c("batch_", sample(LETTERS, 7, replace = T), sample.int(9, size=5, replace = T)), collapse = "")
     return(random.file)
   }) %>% bindEvent(input$run_button4)
   
   
   # Reactive for creating results folder when Run button is clicked and storing a value
   table_heat4 <- reactive({
     
     shinyjs::showElement(id = 'loading.batch4')
     library(data.table)
     library(stringr)
     
     random.file <- random.file4()
     search_mode4 <- search_mode4()
     # Access species proteome
     
     org_batch <- as.character(seq_org4())
     
     { 
       if (org_batch == "Zygnema circumcarinatum SAG")
       {
         org_search <- "zygnema_circumcarinatum"
       }
       else if (org_batch == "Zygnema circumcarinatum UTEX 1559")
       {
         org_search <- "zygnema_circumcarinatum1559"
       }
       else if (org_batch == "Mesostigma viride CCAC 1140")
       {
         org_search <- "mesostigma_viride1140"
       }
       else if (org_batch == "Mesostigma viride NIES 296")
       {
         org_search <- "mesostigma_viride296"
       }
       else
       {
         org_search <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(org_batch))))
       }
     }
     
     org_fasta <- paste("pharaohfun_proteomes", paste0(org_search, ".fa", sep=""), sep = "/")
     org_selection <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(selected_organisms4()))))
     
       if ("zygnema_circumcarinatum_sag" %in% org_selection)
       {
         org_selection[which(org_selection == "zygnema_circumcarinatum_sag")] <- "zygnema_circumcarinatum"
       }
       if ("zygnema_circumcarinatum_utex_1559" %in% org_selection)
       {
         org_selection[which(org_selection == "zygnema_circumcarinatum_utex_1559")] <- "zygnema_circumcarinatum1559"
       }
       if ("mesostigma_viride_ccac_1140" %in% org_selection)
       {
         org_selection[which(org_selection == "mesostigma_viride_ccac_1140")] <- "mesostigma_viride1140"
       }
       if ("mesostigma_viride_nies_296" %in% org_selection)
       {
         org_selection[which(org_selection == "mesostigma_viride_nies_296")] <- "mesostigma_viride296"
       }
     
     
     # Load data based on input mode
     # If input is entered as sequences
     {
       if (search_mode4)
       {
         # Load and clean fasta
         seqs_val <- seqinr::read.fasta(file = input$geneInt4$datapath, seqtype = "AA", forceDNAtolower = F, as.string = T) 
         names_seqs_val <- seqinr::getName(seqs_val)
         seqs_seqs_val <- seqinr::getSequence(seqs_val, as.string = T)
         seqs_seqs_val_clean <- lapply(seqs_seqs_val, function(x) as.character(toupper(gsub("[\r\n\t]", "", x))))
         seqs_seqs_val_clean2 <- lapply(seqs_seqs_val_clean, function(x) str_replace_all(x, fixed(" "), ""))
         vec_seq_val <- lapply(seqs_seqs_val_clean2, function(x) str_split_1(x, pattern = ""))
         
         # Create clean fasta file for comparison with random name to allow several users at once
         seqinr::write.fasta(vec_seq_val, names = names_seqs_val, paste0(random.file, "_new_diamond.fa"))
         
         # Run diamond
         library(rdiamond)
         
         diamond_res <- rdiamond::diamond_protein_to_protein(
           query   = paste0(random.file, "_new_diamond.fa"),
           subject = org_fasta,
           sensitivity_mode = "sensitive",
           output_path = tempdir(),
           #db_import  = FALSE,
           #diamond_exec_path = "/usr/bin", 
           max_target_seqs = 1)
         
         diamond_table <- data.frame(diamond_res)
         file.remove(paste0(random.file, "_new_diamond.fa"))
         
       }
       
       # If ID search mode is selected
       else
       {
         ids_vals <- fread(file = input$geneInt4$datapath, header = F)
         diamond_table <- data.frame(query_id = ids_vals$V1, subject_id = ids_vals$V1)
       }
     }
     
     # Error if an ID does not correspond to the selected proteome, it only
     # happens in ID search mode, but this chunk is placed here to preserve
     # a common framework for both search modes
     
     # Access species proteome
     prot_comp_fasta <- seqinr::read.fasta(file = org_fasta, seqtype = "AA", forceDNAtolower = F, as.string = T) 
     prot_ids <- seqinr::getName(prot_comp_fasta)
     
     {
     if (!all(diamond_table$subject_id %in% prot_ids))
     {
       # Error message and remove output
       if (UI_exist_batch4)
       {
         
         removeUI(
           selector = "div:has(>> #heat_image4)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadBatch4)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_batch4 <<- F
         
       }
       
       incorrect_ids <- diamond_table$subject_id[which(!(diamond_table$subject_id %in% prot_ids))]
       shinyjs::hideElement(id = 'loading.batch4')
       output$error_batch4 <- renderUI({
         renderPrint({cat(paste0("No results available due to wrong ID format. Please use
                          supported IDs or introduce the corresponding sequences
                          using a FASTA format with your own protein names. Remove the following
                          IDs: ", paste0(incorrect_ids, collapse = " / ")))})
       })
       
       validate(" ")
     }
     
     else
     {
       incorrect_ids <- NULL
       output$error_batch4 <- NULL
     }
     }
   
     # Once table is created and IDs checked, create a new results folder 
     # (remove previous one if it exists with the same random name)
     
     if (file.exists(paste0(random.file, ".zip")))
     {
       file.remove(paste0(random.file, ".zip"))
     }
     
     dir.create(random.file)
     
     # For each subject id in table, find OG name
     ortho.file <- "Gene_Trees/Orthogroups.tsv"
     
     ortho.line <- fread(ortho.file, select = c("Orthogroup",org_selection))
     
     
     gene.numbers <- sapply(diamond_table$subject_id, function(x) grep(x, ortho.line[[org_search]]), USE.NAMES = F)
     
     
     # If a pattern is found in several names, i.e., is a subpattern of several genes,
     # search for the exact match
     if(class(gene.numbers) == "list")
     {
       # Check if there are genes with more than one match due to subpatterns
       index.wrong <- which(sapply(gene.numbers, function(x) length(x) > 1))
       {
         if (length(index.wrong != 0))
         {
           for (i in index.wrong)
           {
             for (j in gene.numbers[[i]])
             {
               ortho_char_split <- strsplit(ortho.line[[org_search]][j],split = ",")
               ortho_char_split_clean <- sapply(ortho_char_split, function(x) gsub(" ", "", x))
               if (diamond_table$subject_id[i] %in% ortho_char_split_clean)
               {
                 gene.numbers[i] <- j
               }
             }
           }
         }
       }
     }
     
     
     # Write corresponding OG to results table, and indicate it if it is not clustered in any OG
     ogs.batch <- sapply(gene.numbers, function(x) ortho.line$Orthogroup[x], USE.NAMES = F)
     ogs.names <- sapply(ogs.batch, function(x) if (length(x) == 0) "Gene is not clustered in an OG" else x)
     
     diamond_table <- cbind(diamond_table, "OG" = ogs.names)
     fwrite(diamond_table, file = paste0(random.file, "/results.tsv"),
            sep = "\t")
     
     # Create subfolders for each of the input proteins
     apply(diamond_table, MARGIN = 1,
           FUN = function(x) if (x["OG"] != "Gene is not clustered in an OG") 
             dir.create(paste0(random.file, "/",
                               x["query_id"])))
     
     # Copy results to its folders
     folder_prefix <- ""
     
     apply(diamond_table, MARGIN = 1,
           FUN = function(x) if (x["OG"] != "Gene is not clustered in an OG") 
           {
             # Copy trees to its folders
             
             file.copy(from=paste0(folder_prefix, "Gene_Trees/", 
                                   x["OG"], "_tree.txt"), 
                       to=paste0(random.file, "/", x["query_id"], "/", x["query_id"], "_tree.txt"), 
                       overwrite = TRUE, recursive = FALSE, 
                       copy.mode = TRUE)
             
             # Copy MSAs to its folders
             file.copy(from=paste0(folder_prefix, "MultipleSequenceAlignments/",
                                   x["OG"], ".fa"), 
                       to=paste0(random.file, "/", x["query_id"], "/", x["query_id"], "_MSA.fa"), 
                       overwrite = TRUE, recursive = FALSE, 
                       copy.mode = TRUE)
             
           }
     )
     
     # Copy species tree to general folder
     file.copy(from="pharaoh_folder/tree_with_internal_nodes.txt", 
               to=paste0(random.file, "/", "species_tree.nwk"), 
               overwrite = TRUE, recursive = FALSE, 
               copy.mode = TRUE)
     
     # Functional annotation
     
     # Retrieve genes for each OG
     ortho_for_go <- apply(diamond_table, MARGIN = 1,
                           FUN = function(x) if (x["OG"] != "Gene is not clustered in an OG") 
                             as.character(subset(ortho.line, (ortho.line$Orthogroup == x["OG"]))), simplify = F)
     
     
     ortho_for_go2 <- lapply(ortho_for_go, function(x) paste0(x[-1][x[-1] != ""], collapse = ","))
     ortho_for_go_clean <- sapply(ortho_for_go2, function(x) strsplit(str_replace_all(x, fixed(" "), ""), split = ","))
     ortho_for_go_def <- ortho_for_go_clean[lapply(ortho_for_go_clean,length)>0]
     # Associate each vector of genes with its corresponding query id
     names(ortho_for_go_def) <- diamond_table$query_id[apply(diamond_table, MARGIN = 1,
                                                             FUN = function(x) (x["OG"] != "Gene is not clustered in an OG"))]
     
     
     # Subset the GO annotation and write TSVs
     annotation.go <- fread("pharaoh_folder/final_anot_table.tsv")
     
     mapply(function(x,y)
     {fwrite(annotation.go[annotation.go$name %in% x, ], 
             file = paste0(random.file, "/", y, "/", y, "_GO.tsv"), sep = "\t")}, 
     x=ortho_for_go_def, y=names(ortho_for_go_def))
     
     # Subset the KO annotation and write TSVs (using same set as in GO)
     annotation.ko <- fread("pharaoh_folder/ko_table_funtree.tsv")
     
     mapply(function(x,y)
     {fwrite(annotation.ko[annotation.ko$gene %in% x, ], 
             file = paste0(random.file, "/", y, "/", y, "_KO.tsv"), sep = "\t")}, 
     x=ortho_for_go_def, y=names(ortho_for_go_def))
     
     # CAFE files
     # Simply copy COUNT table entries in the general folder
     cafe_table <- fread("pharaoh_folder/family_table_count.tsv", sep = "\t", header = T)
     
     apply(diamond_table, MARGIN = 1,
           FUN = function(x) if (x["OG"] != "Gene is not clustered in an OG") 
           {
             # Copy trees to its folders
             new_line <- as.data.frame(cafe_table[Orthogroup == x["OG"]])
             fwrite(new_line, file = paste0(random.file, "/", x["query_id"], "/", x["query_id"], "_count_evo.tsv"),
                    sep = "\t", col.names = T, row.names = F)
           })
     
     
     
     # Create gene tables with reduced species for each OG
     table_ogs <- fread(ortho.file)
     org_index <- sapply(org_selection, function(x) which(x == colnames(table_ogs)), USE.NAMES = F)
     org_index <- as.numeric(c(1, org_index))
     table_ogs_red <- table_ogs[,c(org_index), with=F]
    
     apply(diamond_table, MARGIN = 1,
           FUN = function(x) if (x["OG"] != "Gene is not clustered in an OG") 
           {
             # Copy trees to its folders
             new_line <- as.data.frame(table_ogs_red[Orthogroup == x["OG"]])
             fwrite(new_line, file = paste0(random.file, "/", x["query_id"], "/", x["query_id"], "_gene_table.tsv"),
                    sep = "\t", col.names = T, row.names = F)
           })
     
     # Compress folder and remove uncompressed one
     zip(zipfile = paste0(random.file, ".zip"), files = random.file)
     unlink(random.file, recursive = TRUE)
     
     # Return table with query_id and Orthogroups for heatmap
     table_heat <- subset(diamond_table, OG != "Gene is not clustered in an OG")
     table_heat <- table_heat[, c("query_id", "OG")]
     return(table_heat)
     
   }) %>% bindEvent(input$run_button4)
   
   # Create heatmap for OGs in each species
   heat_plot4 <- reactive({
     
     library(ggplot2)
     library(dplyr)
     library(data.table)
     
     random.file <- random.file4()
     table_heat <- table_heat4()
     selected_organisms <- selected_organisms4()
     #org_sel <- as.character(sapply(selected_organisms, function(x) gsub(" ", "_", tolower(x))))
     #org_first <- sapply(strsplit(org_sel, split = "_"), function(x) x[[1]])
     org_selection <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(selected_organisms4()))))
     
     if ("zygnema_circumcarinatum_sag" %in% org_selection)
     {
       org_selection[which(org_selection == "zygnema_circumcarinatum_sag")] <- "zygnema_circumcarinatum"
     }
     if ("zygnema_circumcarinatum_utex_1559" %in% org_selection)
     {
       org_selection[which(org_selection == "zygnema_circumcarinatum_utex_1559")] <- "zygnema_circumcarinatum1559"
     }
     if ("mesostigma_viride_ccac_1140" %in% org_selection)
     {
       org_selection[which(org_selection == "mesostigma_viride_ccac_1140")] <- "mesostigma_viride1140"
     }
     if ("mesostigma_viride_nies_296" %in% org_selection)
     {
       org_selection[which(org_selection == "mesostigma_viride_nies_296")] <- "mesostigma_viride296"
     }
     
     
     # Load gene count for each OG and filter out species that are not selected by the user
     genecount_file <- "pharaoh_folder/genecount.tsv"
     genecount_data_raw <- fread(genecount_file)
     org_index <- sapply(org_selection, function(x) which(x == colnames(genecount_data_raw)), USE.NAMES = F)
     org_index <- as.numeric(c(1, org_index))
     genecount_data <- genecount_data_raw[,c(org_index), with=F]
     
     # Create index to retain order and dups in the subset
     genecount_index <- sapply(table_heat$OG, function(x) which(x == genecount_data$Orthogroup))
     genecount_filt <- genecount_data[genecount_index,]
     
     genecount_filt$Orthogroup <- table_heat$query_id
     colnames(genecount_filt) <- c("query", colnames(genecount_filt)[-1])
     
     # Convert from wide to long format for plotting
     data_plot <- melt(genecount_filt,
                       id.vars = c("query"),
                       measure.vars = patterns("*_"),
                       variable.name = "species",
                       value.name = "value")
     
     return(data_plot)
     
     # # Create heatmap
     # heat_plot <- ggplot(data_plot, aes(species, query, fill= value)) + 
     #   geom_tile(color = "white",
     #             linewidth = 1.5,
     #             linetype = 1) +
     #   scale_fill_gradientn(
     #     values = c(0, quantile(data_plot$value, 0.1), quantile(data_plot$value, 0.2),
     #                quantile(data_plot$value, 0.3), quantile(data_plot$value, 0.4),
     #                quantile(data_plot$value, 0.7)),
     #     guide = "colourbar",
     #     aesthetics = "fill",
     #     colors = c("white", "#fcd57a", "#fcc139", "#f48131", "#ed1411", "#ab0808")
     #   ) +
     #   theme(axis.text.x = element_text(angle = 65, hjust = 1, size=11),
     #         axis.text.y = element_text(size=12)) 
     # 
     # return(heat_plot)
     
   }) 
   
   # Create boxes
   
   observeEvent(isTruthy(heat_plot4()), {
     
     if (UI_exist_batch4)
     {
      
       removeUI(
         selector = "div:has(>> #heat_image4)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadBatch4)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     table_heat <- table_heat4()
     
     insertUI("#box_heatmap4", "afterEnd", ui = {
       box(width = 12, height = nrow(table_heat)*30+50,
           title = "Orthogroups Heatmap", status = "warning", solidHeader = TRUE,
           collapsible = TRUE, 
           plotlyOutput("heat_image4", height = nrow(table_heat)*30-20)
       )
     })
     
     insertUI("#download_batch4", "afterEnd", ui = {
       tags$div(style = "margin-left: 300px;", shinyWidgets::downloadBttn(outputId= "downloadBatch4", 
                                                                          "Download Results Folder",
                                                                          size = "sm", color = "warning"))
     })
     
     UI_exist_batch4 <<- TRUE
     shinyjs::hideElement(id = 'loading.batch4')
   })
   
   # Fill boxes with output
   
   # Render heatmap
   # output$heat_image4 <- renderImage({
   #   
   #   table_heat <- table_heat4()
   #   heat_plot <- heat_plot4()
   #   
   #   png("heatmap.png", height = nrow(table_heat)*30-20, width = 1000)
   #   plot(heat_plot)
   #   dev.off()
   #   
   #   list(src = "heatmap.png",
   #        contentType="image/png", width=1000,height=nrow(table_heat)*30-20)
   # }, deleteFile = T)
   
   output$heat_image4 <- renderPlotly({
     
     table_heat <- table_heat4()
     heat_plot <- heat_plot4()
     
     heat_plot %>%
       plot_ly(x = ~species, y = ~query, z = ~value, type = "heatmap", colors = c("#fcd57a", "#fcc139", "#f48131", "#ed1411", "#ab0808"),
               width = 1300, height = nrow(table_heat)*30-20) %>%
       layout(xaxis = list(side = "top"), 
              yaxis = list(title = ""), 
              showlegend = FALSE) %>%
       
       add_trace(data = heat_plot, x = ~species, y = ~query, z = ~value, type = "heatmap",
                 hoverinfo = 'text',
                 text = ~paste(" Species:", heat_plot$species,
                               "<br> Query gene:", heat_plot$query,
                               "<br> Gene number:", heat_plot$value)) %>%
      hide_colorbar()
     
   })
   
     
   
   # Download results
   output$downloadBatch4 <- downloadHandler(
     filename= function() {
       paste0("batch_results", ".zip")
     },
     content= function(file) {
       random.file <- random.file4()
       file.copy(paste0(random.file, ".zip"), file)
       file.remove(paste0(random.file, ".zip"))
     })
   

# End of Batch Mode Search
   
   
######################## SHOOT-BASED SEARCH #################
   
   # Set global variables for tracking changes in output
   UI_exist_pfam5 <<- F
   UI_exist_tree5 <<- F
   UI_exist_phylo5 <<- F
   UI_exist_cafe5 <<- F
   UI_exist_error_cafe5 <<- F
   UI_exist_msa5 <<- F
   UI_exist_go5 <<- F
   UI_exist_kegg5 <<- F
   UI_exist_kegg_path5 <<- F
   UI_exist_pathview5 <<- F
   UI_exist_lit5 <<-  F
   UI_exist_string5 <<-  F
   UI_exist_network5 <<-  F

   # To avoid autoupdating some inputs, define variables with its values
   model.selected5 <- reactive({
     model.selected <- F
     return(model.selected)
   })%>% bindEvent(input$run_button5)

   # Load organisms selection based on the model selected
   selected_organisms5 <- reactive({
     selected_organisms <- c(input$prasino_check_5, input$mami_check_5,
                             input$chloro_check_5, input$strepto_check_5,
                             input$bryo_check_5, input$lyco_check_5,
                             input$gymno_check_5, input$magno_check_5,
                             input$monocots_check_5, input$dicots_first_check_5,
                             input$dicots_second_check_5)
     return(selected_organisms)

   }) %>% bindEvent(input$run_button5)

   selected_values_org5 <- reactive(organisms_values[selected_organisms5()]) %>% bindEvent(input$run_button5)
   
   random.file5 <- reactive({

     random.file <- paste0(c("neworg_", sample(LETTERS, 7, replace = T), sample.int(9, size=5, replace = T)), collapse = "")
     return(random.file)
   }) %>% bindEvent(input$run_button5)
   
   # First, clean sequence 
   shoot_sequence5 <- reactive({
     
     library(stringr)
     output$error_tree5 <- NULL
     shinyjs::showElement(id = 'loading.tree5')
     seq_comp <- as.character(input$geneInt5)
     seq_comp_clean <- toupper(gsub("[\r\n\t]", "", seq_comp))
     seq_comp_clean2 <- str_replace_all(seq_comp_clean, fixed(" "), "")
     vec_comp <- str_split_1(seq_comp_clean2, pattern = "")
     return(vec_comp)
     
   }) %>% bindEvent(input$run_button5)

   
   tree5 <- reactive({

     library(seqinr)
     library(ape)
     vec_comp <- shoot_sequence5()
     random_name <- random.file5()
     
     # Create a FASTA file with the query seq (using a random file name to
     # avoid conflicts between simultaneous users) for comparison
     random_system <- paste0(random_name, "_system")
     write.fasta(vec_comp, names = "query_prot", paste0("pharaoh_folder/", random_system, ".fa"))

     # Then, execute DIAMOND in server to create .sh.ogs.txt.gz file and execute SHOOT

     database <- "Results_Apr03"
     shoot_works <- F
     
     try(
     system(paste0("cd /srv/shiny-server/PharaohFUN/pharaoh_folder;
            diamond blastp --db ", database, "/profile_sequences.all.fa -q ", random_system, ".fa -o ", random_system, ".fa.sh.ogs.txt.gz --quiet -e 0.001 --compress 1 -p 1 || true"))
     )
     
     gzip_file <- gzfile(paste0("pharaoh_folder/", random_system,".fa.sh.ogs.txt.gz"),'rt')  
     gzip_data <- readLines(gzip_file)
     shoot_test <- length(gzip_data)
     
     {
        if(shoot_test > 0)
        {
          shoot_works <- T
        }
     }
     
     # # If a normal diamond search does not provide results, repeat using ultra sensitive mode
     if(!shoot_works)
     {
       try(
       system(paste0("cd /srv/shiny-server/PharaohFUN/pharaoh_folder;
            diamond blastp --db ", database, "/profile_sequences.all.fa -q ", random_system, ".fa -o ", random_system, ".fa.sh.ogs.txt.gz --quiet -e 0.001 --compress 1 -p 1 --ultra-sensitive || true"))
       )
       
       gzip_file <- gzfile(paste0("pharaoh_folder/", random_system,".fa.sh.ogs.txt.gz"),'rt')  
       gzip_data <- readLines(gzip_file)
       shoot_test <- length(gzip_data)
       
       {
         if(shoot_test > 0)
         {
           shoot_works <- T
         }
       }
       
     }

     
     # Error if no sequence is detected after ultra sensitive mode
     {
       if(!shoot_works)
       {
         shinyjs::hideElement(id = 'loading.tree5')
         
         if (UI_exist_tree5)
         {
           removeUI(
             selector = "div:has(>> #treeTips5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>>> #presentorg5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #tree_image5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadTree5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadNewick5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadTreeSeqs5)",
             multiple = TRUE,
             immediate = TRUE
           )
         }
         
         UI_exist_tree5 <<- F
         output$error_tree5 <- renderUI({renderText({print("No orthogroup profile match the query sequence with E-value cutoff of 0.001.")})})
         validate(" ")
       }
     }
     
     system(paste0("cd /srv/shiny-server/PharaohFUN/pharaoh_folder;
            python3 /srv/shiny-server/PharaohFUN/SHOOT/shoot ", random_system, ".fa ", database, "/"))
     
    
     # Remove input fasta file once SHOOT results have been generated
     file.remove(paste0("pharaoh_folder/", random_system, ".fa"))
     
     # Load gene tree file depending on the input
     shoot_tree <- ape::read.tree(paste0("pharaoh_folder/", random_system, ".fa.shoot.tre"))
     
     return(shoot_tree)

   }) %>% bindEvent(input$run_button5)
   
   # Create variable for MSA and remove useless files
   shoot_msa5 <- reactive({
     tree <- tree5()
     random_name <- random.file5()
     random_system <- paste0(random_name, "_system")
     
     mySequences1 <- seqinr::read.fasta(paste0("pharaoh_folder/", random_system, ".fa.sh.msa.fa"), seqtype = "AA")
     
     # Remove useless files and folders
     system(paste0("cd /srv/shiny-server/PharaohFUN/pharaoh_folder/;
                   rm ", random_system, ".fa.*"))
     system(paste0("cd /srv/shiny-server/PharaohFUN/pharaoh_folder/;
                   rm -r ", random_system, ".fa.*"))
     
     return(mySequences1)
     
     
   }) %>% bindEvent(input$run_button5)


   # Tips to keep of each species with proper notation

   tips_to_keep.query5 <- reactive({

     tree <- tree5()

     # Selection of genes from the selected organism
     tips_to_keep.query <- grep(pattern = "query_prot",tree$tip.label)

     return(tips_to_keep.query)
   })

   tips_to_keep.mp5 <- reactive({
     
     tree <- tree5()
     # Selection of organisms
     organisms.list <- c(selected_values_org5())
     
     # Selection of genes from the selected organism
     tips_to_keep.mp <- c()
     if ("mp" %in% organisms.list)
     {
       tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label)
     }
     return(tips_to_keep.mp)
   })
   
   tips_to_keep.ot5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ot <- c()
     if ("ot" %in% organisms.list)
     {
       tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label)
     }
     return(tips_to_keep.ot)
   })
   
   tips_to_keep.at5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.at <- c()
     if ("at" %in% organisms.list)
     {
       tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label)
     }
     return(tips_to_keep.at)
   })
   
   tips_to_keep.cp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cp <- c()
     if ("cp" %in% organisms.list)
     {
       tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label)
     }
     
     return(tips_to_keep.cp)
   })
   
   tips_to_keep.cr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cr <- c()
     if ("cr" %in% organisms.list)
     {
       tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
     }
     return(tips_to_keep.cr)
   })
   
   tips_to_keep.cz5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cz <- c()
     if ("cz" %in% organisms.list)
     {
       tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label)
     }
     return(tips_to_keep.cz)
   })
   
   tips_to_keep.kn5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.kn <- c()
     if ("kn" %in% organisms.list)
     {
       tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label)
     }
     return(tips_to_keep.kn)
   })
   
   tips_to_keep.me5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.me <- c()
     if ("me" %in% organisms.list)
     {
       tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label)
     }
     return(tips_to_keep.me)
   })
   
   tips_to_keep.mi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.mi <- c()
     if ("mi" %in% organisms.list)
     {
       tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label)
     }
     return(tips_to_keep.mi)
   })
   
   tips_to_keep.pp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.pp <- c()
     if ("pp" %in% organisms.list)
     {
       tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label)
     }
     return(tips_to_keep.pp)
   })
   
   tips_to_keep.sl5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sl <- c()
     if ("sl" %in% organisms.list)
     {
       tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label)
     }
     return(tips_to_keep.sl)
   })
   
   tips_to_keep.sm5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sm <- c()
     if ("sm" %in% organisms.list)
     {
       tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label)
     }
     return(tips_to_keep.sm)
   })
   
   tips_to_keep.sp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sp <- c()
     if ("sp" %in% organisms.list)
     {
       tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label)
     }
     return(tips_to_keep.sp)
   })
   
   tips_to_keep.ta5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ta <- c()
     if ("ta" %in% organisms.list)
     {
       tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label)
     }
     return(tips_to_keep.ta)
   })
   
   tips_to_keep.vc5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.vc <- c()
     if ("vc" %in% organisms.list)
     {
       tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
     }
     return(tips_to_keep.vc)
   })
   
   tips_to_keep.bp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.bp <- c()
     if ("bp" %in% organisms.list)
     {
       tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
     }
     return(tips_to_keep.bp)
   })
   
   tips_to_keep.cri5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cri <- c()
     if ("cri" %in% organisms.list)
     {
       tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
     }
     return(tips_to_keep.cri)
   })
   
   tips_to_keep.ds5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ds <- c()
     if ("ds" %in% organisms.list)
     {
       tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
     }
     return(tips_to_keep.ds)
   })
   
   tips_to_keep.os5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.os <- c()
     if ("os" %in% organisms.list)
     {
       tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
     }
     return(tips_to_keep.os)
   })
   
   tips_to_keep.smag5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.smag <- c()
     if ("smag" %in% organisms.list)
     {
       tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
     }
     return(tips_to_keep.smag)
   })
   
   tips_to_keep.tp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.tp <- c()
     if ("tp" %in% organisms.list)
     {
       tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
     }
     return(tips_to_keep.tp)
   })
   
   tips_to_keep.aa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.aa <- c()
     if ("aa" %in% organisms.list)
     {
       tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
     }
     return(tips_to_keep.aa)
   })
   
   tips_to_keep.um5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.um <- c()
     if ("um" %in% organisms.list)
     {
       tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
     }
     return(tips_to_keep.um)
   })
   
   tips_to_keep.rs5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.rs <- c()
     if ("rs" %in% organisms.list)
     {
       tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
     }
     return(tips_to_keep.rs)
   })
   
   tips_to_keep.cyc5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cyc <- c()
     if ("cyc" %in% organisms.list)
     {
       tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
     }
     return(tips_to_keep.cyc)
   })
   
   
   tips_to_keep.ca5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ca <- c()
     if ("ca" %in% organisms.list)
     {
       tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
     }
     return(tips_to_keep.ca)
   })
   
   tips_to_keep.mv5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.mv <- c()
     if ("mv" %in% organisms.list)
     {
       tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
     }
     return(tips_to_keep.mv)
   })
   
   tips_to_keep.af5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.af <- c()
     if ("af" %in% organisms.list)
     {
       tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
     }
     return(tips_to_keep.af)
   })
   
   tips_to_keep.sc5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sc <- c()
     if ("sc" %in% organisms.list)
     {
       tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
     }
     return(tips_to_keep.sc)
   })
   
   tips_to_keep.aegi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.aegi <- c()
     if ("aegi" %in% organisms.list)
     {
       tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
     }
     return(tips_to_keep.aegi)
   })
   
   tips_to_keep.sb5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sb <- c()
     if ("sb" %in% organisms.list)
     {
       tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
     }
     return(tips_to_keep.sb)
   })
   
   tips_to_keep.chara5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.chara <- c()
     if ("chara" %in% organisms.list)
     {
       tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
     }
     return(tips_to_keep.chara)
   })
   
   tips_to_keep.sceobli5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sceobli <- c()
     if ("sceobli" %in% organisms.list)
     {
       tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
     }
     return(tips_to_keep.sceobli)
   })
   
   tips_to_keep.cocco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cocco <- c()
     if ("cocco" %in% organisms.list)
     {
       tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
     }
     return(tips_to_keep.cocco)
   })
   
   
   tips_to_keep.haema5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.haema <- c()
     if ("haema" %in% organisms.list)
     {
       tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
     }
     return(tips_to_keep.haema)
   })
   
   tips_to_keep.zm5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.zm <- c()
     if ("zm" %in% organisms.list)
     {
       tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
     }
     return(tips_to_keep.zm)
   })
   
   tips_to_keep.praco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.praco <- c()
     if ("praco" %in% organisms.list)
     {
       tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
     }
     return(tips_to_keep.praco)
   })
   
   tips_to_keep.ostlu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ostlu <- c()
     if ("ostlu" %in% organisms.list)
     {
       tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
     }
     return(tips_to_keep.ostlu)
   })
   
   tips_to_keep.ostme5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ostme <- c()
     if ("ostme" %in% organisms.list)
     {
       tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
     }
     return(tips_to_keep.ostme)
   })
   
   tips_to_keep.micco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.micco <- c()
     if ("micco" %in% organisms.list)
     {
       tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
     }
     return(tips_to_keep.micco)
   })
   
   tips_to_keep.cymte5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cymte <- c()
     if ("cymte" %in% organisms.list)
     {
       tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
     }
     return(tips_to_keep.cymte)
   })
   
   tips_to_keep.chlin5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.chlin <- c()
     if ("chlin" %in% organisms.list)
     {
       tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
     }
     return(tips_to_keep.chlin)
   })
   
   tips_to_keep.gonpe5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.gonpe <- c()
     if ("gonpe" %in% organisms.list)
     {
       tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
     }
     return(tips_to_keep.gonpe)
   })
   
   tips_to_keep.tetso5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.tetso <- c()
     if ("tetso" %in% organisms.list)
     {
       tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
     }
     return(tips_to_keep.tetso)
   })
   
   tips_to_keep.desar5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.desar <- c()
     if ("desar" %in% organisms.list)
     {
       tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
     }
     return(tips_to_keep.desar)
   })
   
   tips_to_keep.monmi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.monmi <- c()
     if ("monmi" %in% organisms.list)
     {
       tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
     }
     return(tips_to_keep.monmi)
   })
   
   tips_to_keep.caule5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.caule <- c()
     if ("caule" %in% organisms.list)
     {
       tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
     }
     return(tips_to_keep.caule)
   })
   
   tips_to_keep.ostqu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ostqu <- c()
     if ("ostqu" %in% organisms.list)
     {
       tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
     }
     return(tips_to_keep.ostqu)
   })
   
   tips_to_keep.brysp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.brysp <- c()
     if ("brysp" %in% organisms.list)
     {
       tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
     }
     return(tips_to_keep.brysp)
   })
   
   tips_to_keep.ellbi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ellbi <- c()
     if ("ellbi" %in% organisms.list)
     {
       tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
     }
     return(tips_to_keep.ellbi)
   })
   
   tips_to_keep.myrbi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.myrbi <- c()
     if ("myrbi" %in% organisms.list)
     {
       tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
     }
     return(tips_to_keep.myrbi)
   })
   
   tips_to_keep.apalo5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.apalo <- c()
     if ("apalo" %in% organisms.list)
     {
       tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
     }
     return(tips_to_keep.apalo)
   })
   
   tips_to_keep.astgl5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.astgl <- c()
     if ("astgl" %in% organisms.list)
     {
       tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
     }
     return(tips_to_keep.astgl)
   })
   
   tips_to_keep.tresp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.tresp <- c()
     if ("tresp" %in% organisms.list)
     {
       tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
     }
     return(tips_to_keep.tresp)
   })
   
   tips_to_keep.picso5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.picso <- c()
     if ("picso" %in% organisms.list)
     {
       tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
     }
     return(tips_to_keep.picso)
   })
   
   tips_to_keep.chlso5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.chlso <- c()
     if ("chlso" %in% organisms.list)
     {
       tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
     }
     return(tips_to_keep.chlso)
   })
   
   tips_to_keep.auxpr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.auxpr <- c()
     if ("auxpr" %in% organisms.list)
     {
       tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
     }
     return(tips_to_keep.auxpr)
   })
   
   tips_to_keep.helsp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.helsp <- c()
     if ("helsp" %in% organisms.list)
     {
       tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
     }
     return(tips_to_keep.helsp)
   })
   
   tips_to_keep.tetst5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.tetst <- c()
     if ("tetst" %in% organisms.list)
     {
       tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
     }
     return(tips_to_keep.tetst)
   })
   
   tips_to_keep.pedsp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.pedsp <- c()
     if ("pedsp" %in% organisms.list)
     {
       tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
     }
     return(tips_to_keep.pedsp)
   })
   
   tips_to_keep.chlpr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.chlpr <- c()
     if ("chlpr" %in% organisms.list)
     {
       tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
     }
     return(tips_to_keep.chlpr)
   })
   
   tips_to_keep.picsp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.picsp <- c()
     if ("picsp" %in% organisms.list)
     {
       tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
     }
     return(tips_to_keep.picsp)
   })
   
   tips_to_keep.mesvb5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.mesvb <- c()
     if ("mesvb" %in% organisms.list)
     {
       tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
     }
     return(tips_to_keep.mesvb)
   })
   
   tips_to_keep.meskr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.meskr <- c()
     if ("meskr" %in% organisms.list)
     {
       tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
     }
     return(tips_to_keep.meskr)
   })
   
   tips_to_keep.zygci5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.zygci <- c()
     if ("zygci" %in% organisms.list)
     {
       tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
     }
     return(tips_to_keep.zygci)
   })
   
   tips_to_keep.zygcb5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.zygcb <- c()
     if ("zygcb" %in% organisms.list)
     {
       tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
     }
     return(tips_to_keep.zygcb)
   })
   
   tips_to_keep.zygcy5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.zygcy <- c()
     if ("zygcy" %in% organisms.list)
     {
       tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
     }
     return(tips_to_keep.zygcy)
   })
   
   tips_to_keep.plesc5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.plesc <- c()
     if ("plesc" %in% organisms.list)
     {
       tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
     }
     return(tips_to_keep.plesc)
   })
   
   tips_to_keep.funhy5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.funhy <- c()
     if ("funhy" %in% organisms.list)
     {
       tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
     }
     return(tips_to_keep.funhy)
   })
   
   tips_to_keep.sphfa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sphfa <- c()
     if ("sphfa" %in% organisms.list)
     {
       tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
     }
     return(tips_to_keep.sphfa)
   })
   
   tips_to_keep.takle5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.takle <- c()
     if ("takle" %in% organisms.list)
     {
       tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
     }
     return(tips_to_keep.takle)
   })
   
   tips_to_keep.marpa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.marpa <- c()
     if ("marpa" %in% organisms.list)
     {
       tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
     }
     return(tips_to_keep.marpa)
   })
   
   tips_to_keep.ricfl5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ricfl <- c()
     if ("ricfl" %in% organisms.list)
     {
       tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
     }
     return(tips_to_keep.ricfl)
   })
   
   tips_to_keep.ricso5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ricso <- c()
     if ("ricso" %in% organisms.list)
     {
       tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
     }
     return(tips_to_keep.ricso)
   })
   
   tips_to_keep.antpu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.antpu <- c()
     if ("antpu" %in% organisms.list)
     {
       tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
     }
     return(tips_to_keep.antpu)
   })
   
   tips_to_keep.antfu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.antfu <- c()
     if ("antfu" %in% organisms.list)
     {
       tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
     }
     return(tips_to_keep.antfu)
   })
   
   tips_to_keep.megfl5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.megfl <- c()
     if ("megfl" %in% organisms.list)
     {
       tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
     }
     return(tips_to_keep.megfl)
   })
   
   tips_to_keep.phach5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.phach <- c()
     if ("phach" %in% organisms.list)
     {
       tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
     }
     return(tips_to_keep.phach)
   })
   
   tips_to_keep.phyph5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.phyph <- c()
     if ("phyph" %in% organisms.list)
     {
       tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
     }
     return(tips_to_keep.phyph)
   })
   
   tips_to_keep.parpe5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.parpe <- c()
     if ("parpe" %in% organisms.list)
     {
       tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
     }
     return(tips_to_keep.parpe)
   })
   
   tips_to_keep.notor5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.notor <- c()
     if ("notor" %in% organisms.list)
     {
       tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
     }
     return(tips_to_keep.notor)
   })
   
   tips_to_keep.phaca5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.phaca <- c()
     if ("phaca" %in% organisms.list)
     {
       tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
     }
     return(tips_to_keep.phaca)
   })
   
   tips_to_keep.phasp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.phasp <- c()
     if ("phasp" %in% organisms.list)
     {
       tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
     }
     return(tips_to_keep.phasp)
   })
   
   tips_to_keep.leidu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.leidu <- c()
     if ("leidu" %in% organisms.list)
     {
       tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
     }
     return(tips_to_keep.leidu)
   })
   
   tips_to_keep.isosi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.isosi <- c()
     if ("isosi" %in% organisms.list)
     {
       tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
     }
     return(tips_to_keep.isosi)
   })
   
   tips_to_keep.dipco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.dipco <- c()
     if ("dipco" %in% organisms.list)
     {
       tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
     }
     return(tips_to_keep.dipco)
   })
   
   tips_to_keep.hupas5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.hupas <- c()
     if ("hupas" %in% organisms.list)
     {
       tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
     }
     return(tips_to_keep.hupas)
   })
   
   tips_to_keep.lyccl5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.lyccl <- c()
     if ("lyccl" %in% organisms.list)
     {
       tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
     }
     return(tips_to_keep.lyccl)
   })
   
   tips_to_keep.adica5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.adica <- c()
     if ("adica" %in% organisms.list)
     {
       tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
     }
     return(tips_to_keep.adica)
   })
   
   tips_to_keep.alssp5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.alssp <- c()
     if ("alssp" %in% organisms.list)
     {
       tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
     }
     return(tips_to_keep.alssp)
   })
   
   tips_to_keep.marve5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.marve <- c()
     if ("marve" %in% organisms.list)
     {
       tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
     }
     return(tips_to_keep.marve)
   })
   
   tips_to_keep.seqse5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.seqse <- c()
     if ("seqse" %in% organisms.list)
     {
       tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
     }
     return(tips_to_keep.seqse)
   })
   
   tips_to_keep.seqgi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.seqgi <- c()
     if ("seqgi" %in% organisms.list)
     {
       tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
     }
     return(tips_to_keep.seqgi)
   })
   
   tips_to_keep.taxch5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.taxch <- c()
     if ("taxch" %in% organisms.list)
     {
       tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
     }
     return(tips_to_keep.taxch)
   })
   
   tips_to_keep.abial5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.abial <- c()
     if ("abial" %in% organisms.list)
     {
       tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
     }
     return(tips_to_keep.abial)
   })
   
   tips_to_keep.picab5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.picab <- c()
     if ("picab" %in% organisms.list)
     {
       tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
     }
     return(tips_to_keep.picab)
   })
   
   tips_to_keep.gnemo5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.gnemo <- c()
     if ("gnemo" %in% organisms.list)
     {
       tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
     }
     return(tips_to_keep.gnemo)
   })
   
   tips_to_keep.welmi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.welmi <- c()
     if ("welmi" %in% organisms.list)
     {
       tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
     }
     return(tips_to_keep.welmi)
   })
   
   tips_to_keep.ginbi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ginbi <- c()
     if ("ginbi" %in% organisms.list)
     {
       tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
     }
     return(tips_to_keep.ginbi)
   })
   
   tips_to_keep.ambtr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.ambtr <- c()
     if ("ambtr" %in% organisms.list)
     {
       tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
     }
     return(tips_to_keep.ambtr)
   })
   
   tips_to_keep.nymco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.nymco <- c()
     if ("nymco" %in% organisms.list)
     {
       tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
     }
     return(tips_to_keep.nymco)
   })
   
   tips_to_keep.cinka5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cinka <- c()
     if ("cinka" %in% organisms.list)
     {
       tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
     }
     return(tips_to_keep.cinka)
   })
   
   tips_to_keep.horvu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.horvu <- c()
     if ("horvu" %in% organisms.list)
     {
       tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
     }
     return(tips_to_keep.horvu)
   })
   
   tips_to_keep.bradi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.bradi <- c()
     if ("bradi" %in% organisms.list)
     {
       tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
     }
     return(tips_to_keep.bradi)
   })
   
   tips_to_keep.setit5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.setit <- c()
     if ("setit" %in% organisms.list)
     {
       tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
     }
     return(tips_to_keep.setit)
   })
   
   tips_to_keep.panha5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.panha <- c()
     if ("panha" %in% organisms.list)
     {
       tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
     }
     return(tips_to_keep.panha)
   })
   
   tips_to_keep.missi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.missi <- c()
     if ("missi" %in% organisms.list)
     {
       tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
     }
     return(tips_to_keep.missi)
   })
   
   tips_to_keep.oroth5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.oroth <- c()
     if ("oroth" %in% organisms.list)
     {
       tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
     }
     return(tips_to_keep.oroth)
   })
   
   tips_to_keep.eleco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.eleco <- c()
     if ("eleco" %in% organisms.list)
     {
       tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
     }
     return(tips_to_keep.eleco)
   })
   
   tips_to_keep.anaco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.anaco <- c()
     if ("anaco" %in% organisms.list)
     {
       tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
     }
     return(tips_to_keep.anaco)
   })
   
   tips_to_keep.musac5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.musac <- c()
     if ("musac" %in% organisms.list)
     {
       tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
     }
     return(tips_to_keep.musac)
   })
   
   tips_to_keep.dioal5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.dioal <- c()
     if ("dioal" %in% organisms.list)
     {
       tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
     }
     return(tips_to_keep.dioal)
   })
   
   tips_to_keep.spipo5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.spipo <- c()
     if ("spipo" %in% organisms.list)
     {
       tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
     }
     return(tips_to_keep.spipo)
   })
   
   tips_to_keep.aspof5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.aspof <- c()
     if ("aspof" %in% organisms.list)
     {
       tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
     }
     return(tips_to_keep.aspof)
   })
   
   tips_to_keep.zosma5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.zosma <- c()
     if ("zosma" %in% organisms.list)
     {
       tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
     }
     return(tips_to_keep.zosma)
   })
   
   tips_to_keep.araly5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.araly <- c()
     if ("araly" %in% organisms.list)
     {
       tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
     }
     return(tips_to_keep.araly)
   })
   
   tips_to_keep.capgr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.capgr <- c()
     if ("capgr" %in% organisms.list)
     {
       tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
     }
     return(tips_to_keep.capgr)
   })
   
   tips_to_keep.brara5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.brara <- c()
     if ("brara" %in% organisms.list)
     {
       tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
     }
     return(tips_to_keep.brara)
   })
   
   tips_to_keep.braol5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.braol <- c()
     if ("braol" %in% organisms.list)
     {
       tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
     }
     return(tips_to_keep.braol)
   })
   
   tips_to_keep.sisir5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sisir <- c()
     if ("sisir" %in% organisms.list)
     {
       tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
     }
     return(tips_to_keep.sisir)
   })
   
   tips_to_keep.schpa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.schpa <- c()
     if ("schpa" %in% organisms.list)
     {
       tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
     }
     return(tips_to_keep.schpa)
   })
   
   tips_to_keep.eutsa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.eutsa <- c()
     if ("eutsa" %in% organisms.list)
     {
       tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
     }
     return(tips_to_keep.eutsa)
   })
   
   tips_to_keep.thlar5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.thlar <- c()
     if ("thlar" %in% organisms.list)
     {
       tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
     }
     return(tips_to_keep.thlar)
   })
   
   tips_to_keep.boest5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.boest <- c()
     if ("boest" %in% organisms.list)
     {
       tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
     }
     return(tips_to_keep.boest)
   })
   
   tips_to_keep.carpa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.carpa <- c()
     if ("carpa" %in% organisms.list)
     {
       tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
     }
     return(tips_to_keep.carpa)
   })
   
   tips_to_keep.theca5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.theca <- c()
     if ("theca" %in% organisms.list)
     {
       tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
     }
     return(tips_to_keep.theca)
   })
   
   tips_to_keep.goshi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.goshi <- c()
     if ("goshi" %in% organisms.list)
     {
       tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
     }
     return(tips_to_keep.goshi)
   })
   
   tips_to_keep.gosra5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.gosra <- c()
     if ("gosra" %in% organisms.list)
     {
       tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
     }
     return(tips_to_keep.gosra)
   })
   
   tips_to_keep.citsi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.citsi <- c()
     if ("citsi" %in% organisms.list)
     {
       tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
     }
     return(tips_to_keep.citsi)
   })
   
   tips_to_keep.pontr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.pontr <- c()
     if ("pontr" %in% organisms.list)
     {
       tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
     }
     return(tips_to_keep.pontr)
   })
   
   tips_to_keep.eucgr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.eucgr <- c()
     if ("eucgr" %in% organisms.list)
     {
       tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
     }
     return(tips_to_keep.eucgr)
   })
   
   tips_to_keep.maldo5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.maldo <- c()
     if ("maldo" %in% organisms.list)
     {
       tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
     }
     return(tips_to_keep.maldo)
   })
   
   tips_to_keep.prupe5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.prupe <- c()
     if ("prupe" %in% organisms.list)
     {
       tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
     }
     return(tips_to_keep.prupe)
   })
   
   tips_to_keep.frave5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.frave <- c()
     if ("frave" %in% organisms.list)
     {
       tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
     }
     return(tips_to_keep.frave)
   })
   
   tips_to_keep.corav5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.corav <- c()
     if ("corav" %in% organisms.list)
     {
       tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
     }
     return(tips_to_keep.corav)
   })
   
   tips_to_keep.caril5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.caril <- c()
     if ("caril" %in% organisms.list)
     {
       tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
     }
     return(tips_to_keep.caril)
   })
   
   tips_to_keep.queru5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.queru <- c()
     if ("queru" %in% organisms.list)
     {
       tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
     }
     return(tips_to_keep.queru)
   })
   
   tips_to_keep.cucsa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cucsa <- c()
     if ("cucsa" %in% organisms.list)
     {
       tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
     }
     return(tips_to_keep.cucsa)
   })
   
   tips_to_keep.glyma5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.glyma <- c()
     if ("glyma" %in% organisms.list)
     {
       tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
     }
     return(tips_to_keep.glyma)
   })
   
   tips_to_keep.medtr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.medtr <- c()
     if ("medtr" %in% organisms.list)
     {
       tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
     }
     return(tips_to_keep.medtr)
   })
   
   tips_to_keep.tripr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.tripr <- c()
     if ("tripr" %in% organisms.list)
     {
       tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
     }
     return(tips_to_keep.tripr)
   })
   
   tips_to_keep.cicar5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cicar <- c()
     if ("cicar" %in% organisms.list)
     {
       tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
     }
     return(tips_to_keep.cicar)
   })
   
   tips_to_keep.lotja5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.lotja <- c()
     if ("lotja" %in% organisms.list)
     {
       tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
     }
     return(tips_to_keep.lotja)
   })
   
   tips_to_keep.phaac5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.phaac <- c()
     if ("phaac" %in% organisms.list)
     {
       tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
     }
     return(tips_to_keep.phaac)
   })
   
   tips_to_keep.vigun5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.vigun <- c()
     if ("vigun" %in% organisms.list)
     {
       tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
     }
     return(tips_to_keep.vigun)
   })
   
   tips_to_keep.lupal5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.lupal <- c()
     if ("lupal" %in% organisms.list)
     {
       tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
     }
     return(tips_to_keep.lupal)
   })
   
   tips_to_keep.lencu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.lencu <- c()
     if ("lencu" %in% organisms.list)
     {
       tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
     }
     return(tips_to_keep.lencu)
   })
   
   tips_to_keep.arahy5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.arahy <- c()
     if ("arahy" %in% organisms.list)
     {
       tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
     }
     return(tips_to_keep.arahy)
   })
   
   tips_to_keep.poptr5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.poptr <- c()
     if ("poptr" %in% organisms.list)
     {
       tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
     }
     return(tips_to_keep.poptr)
   })
   
   tips_to_keep.manes5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.manes <- c()
     if ("manes" %in% organisms.list)
     {
       tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
     }
     return(tips_to_keep.manes)
   })
   
   tips_to_keep.salpu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.salpu <- c()
     if ("salpu" %in% organisms.list)
     {
       tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
     }
     return(tips_to_keep.salpu)
   })
   
   tips_to_keep.linus5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.linus <- c()
     if ("linus" %in% organisms.list)
     {
       tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
     }
     return(tips_to_keep.linus)
   })
   
   tips_to_keep.corci5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.corci <- c()
     if ("corci" %in% organisms.list)
     {
       tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
     }
     return(tips_to_keep.corci)
   })
   
   tips_to_keep.vitvi5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.vitvi <- c()
     if ("vitvi" %in% organisms.list)
     {
       tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
     }
     return(tips_to_keep.vitvi)
   })
   
   tips_to_keep.kalfe5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.kalfe <- c()
     if ("kalfe" %in% organisms.list)
     {
       tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
     }
     return(tips_to_keep.kalfe)
   })
   
   tips_to_keep.vacda5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.vacda <- c()
     if ("vacda" %in% organisms.list)
     {
       tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
     }
     return(tips_to_keep.vacda)
   })
   
   tips_to_keep.oleeu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.oleeu <- c()
     if ("oleeu" %in% organisms.list)
     {
       tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
     }
     return(tips_to_keep.oleeu)
   })
   
   tips_to_keep.mimgu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.mimgu <- c()
     if ("mimgu" %in% organisms.list)
     {
       tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
     }
     return(tips_to_keep.mimgu)
   })
   
   tips_to_keep.soltu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.soltu <- c()
     if ("soltu" %in% organisms.list)
     {
       tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
     }
     return(tips_to_keep.soltu)
   })
   
   tips_to_keep.cofar5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.cofar <- c()
     if ("cofar" %in% organisms.list)
     {
       tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
     }
     return(tips_to_keep.cofar)
   })
   
   tips_to_keep.lacsa5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.lacsa <- c()
     if ("lacsa" %in% organisms.list)
     {
       tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
     }
     return(tips_to_keep.lacsa)
   })
   
   tips_to_keep.dauca5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.dauca <- c()
     if ("dauca" %in% organisms.list)
     {
       tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
     }
     return(tips_to_keep.dauca)
   })
   
   tips_to_keep.betvu5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.betvu <- c()
     if ("betvu" %in% organisms.list)
     {
       tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
     }
     return(tips_to_keep.betvu)
   })
   
   tips_to_keep.amahy5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.amahy <- c()
     if ("amahy" %in% organisms.list)
     {
       tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
     }
     return(tips_to_keep.amahy)
   })
   
   tips_to_keep.chequ5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.chequ <- c()
     if ("chequ" %in% organisms.list)
     {
       tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
     }
     return(tips_to_keep.chequ)
   })
   
   tips_to_keep.sapof5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.sapof <- c()
     if ("sapof" %in% organisms.list)
     {
       tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
     }
     return(tips_to_keep.sapof)
   })
   
   tips_to_keep.aquco5 <- reactive({
     
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     tips_to_keep.aquco <- c()
     if ("aquco" %in% organisms.list)
     {
       tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
     }
     return(tips_to_keep.aquco)
   }) %>% bindEvent(input$run_button5)
   
   # Create complete gene tree with the proper name for each gene
   # For this, we split the species name apart from the gene name
   tree_adj5 <- reactive({
     tree <- tree5()
     organisms.list <- c(selected_values_org5())
     
     if ("mp" %in% organisms.list)
     {
       tips_to_keep.mp <- grep(pattern = "marchantia_polymorpha",tree$tip.label) 
       if (length(tips_to_keep.mp) != 0)
       {
         mp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mp] <- mp.v
       }
     }
     
     if ("ot" %in% organisms.list)
     {
       tips_to_keep.ot <- grep(pattern = "ostreococcus_tauri",tree$tip.label) 
       if (length(tips_to_keep.ot) != 0)
       {
         ost.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ot]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ot] <- ost.v
       }
     }
     
     if ("at" %in% organisms.list)
     {
       tips_to_keep.at <- grep(pattern = "arabidopsis_thaliana",tree$tip.label) 
       if (length(tips_to_keep.at) != 0)
       {
         arabi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.at]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.at] <- arabi.v
       }
     }
     
     if ("cp" %in% organisms.list)
     {
       tips_to_keep.cp <- grep(pattern = "ceratodon_purpureus",tree$tip.label) 
       if (length(tips_to_keep.cp) != 0)
       {
         cer.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cp]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cp] <- cer.v
       }
     }
     
     if ("cr" %in% organisms.list)
     {
       tips_to_keep.cr <- grep(pattern = "chlamydomonas_reinhardtii",tree$tip.label)
       if (length(tips_to_keep.cr) != 0)
       {
         chlamy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cr]), "_"), 
                            function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cr] <- chlamy.v
       }
     }
     
     if ("cz" %in% organisms.list)
     {
       tips_to_keep.cz <- grep(pattern = "chromochloris_zofingiensis",tree$tip.label) 
       if (length(tips_to_keep.cz) != 0)
       {
         chromo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cz]), "_"), 
                            function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cz] <- chromo.v
       }
     }
     
     if ("kn" %in% organisms.list)
     {
       tips_to_keep.kn <- grep(pattern = "klebsormidium_nitens",tree$tip.label) 
       if (length(tips_to_keep.kn) != 0)
       {
         klebs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kn]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.kn] <- klebs.v
       }
     }
     
     if ("me" %in% organisms.list)
     {
       tips_to_keep.me <- grep(pattern = "mesotaenium_endlicherianum",tree$tip.label) 
       if (length(tips_to_keep.me) != 0)
       {
         meso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.me]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.me] <- meso.v
       }
     }
     
     if ("mi" %in% organisms.list)
     {
       tips_to_keep.mi <- grep(pattern = "micromonas_pusilla",tree$tip.label) 
       if (length(tips_to_keep.mi) != 0)
       {
         micro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mi]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mi] <- micro.v
       }
     }
     
     if ("pp" %in% organisms.list)
     {
       tips_to_keep.pp <- grep(pattern = "physcomitrium_patens",tree$tip.label) 
       if (length(tips_to_keep.pp) != 0)
       {
         phys.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pp]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pp] <- phys.v
       }
     }
     
     if ("sl" %in% organisms.list)
     {
       tips_to_keep.sl <- grep(pattern = "solanum_lycopersicum",tree$tip.label) 
       if (length(tips_to_keep.sl) != 0)
       {
         sola.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sl]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sl] <- sola.v
       }
     }
     
     if ("sm" %in% organisms.list)
     {
       tips_to_keep.sm <- grep(pattern = "selaginella_moellendorffii",tree$tip.label) 
       if (length(tips_to_keep.sm) != 0)
       {
         sel.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sm]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sm] <- sel.v
       }
     }
     
     if ("sp" %in% organisms.list)
     {
       tips_to_keep.sp <- grep(pattern = "spirogloea_muscicola",tree$tip.label) 
       if (length(tips_to_keep.sp) != 0)
       {
         spiro.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sp]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sp] <- spiro.v
       }
     }
     
     if ("ta" %in% organisms.list)
     {
       tips_to_keep.ta <- grep(pattern = "triticum_aestivum",tree$tip.label) 
       if (length(tips_to_keep.ta) != 0)
       {
         tri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ta]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ta] <- tri.v
       }
     }
     
     if ("vc" %in% organisms.list)
     {
       tips_to_keep.vc <- grep(pattern = "volvox_carteri",tree$tip.label)
       if (length(tips_to_keep.vc) != 0)
       {
         vc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vc]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vc] <- vc.v
       }
     }
     
     if ("bp" %in% organisms.list)
     {
       tips_to_keep.bp <- grep(pattern = "bathycoccus_prasinos",tree$tip.label)
       if (length(tips_to_keep.bp) != 0)
       {
         bp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.bp] <- bp.v
       }
     }
     
     if ("cri" %in% organisms.list)
     {
       tips_to_keep.cri <- grep(pattern = "ceratopteris_richardii",tree$tip.label)
       if (length(tips_to_keep.cri) != 0)
       {
         cri.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cri]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cri] <- cri.v
       }
     }
     
     if ("ds" %in% organisms.list)
     {
       tips_to_keep.ds <- grep(pattern = "dunaliella_salina",tree$tip.label)
       if (length(tips_to_keep.ds) != 0)
       {
         ds.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ds]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ds] <- ds.v
       }
     }
     
     if ("os" %in% organisms.list)
     {
       tips_to_keep.os <- grep(pattern = "oryza_sativa",tree$tip.label)
       if (length(tips_to_keep.os) != 0)
       {
         os.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.os]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.os] <- os.v
       }
     }
     
     if ("smag" %in% organisms.list)
     {
       tips_to_keep.smag <- grep(pattern = "sphagnum_magellanicum",tree$tip.label)
       if (length(tips_to_keep.smag) != 0)
       {
         smag.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.smag]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.smag] <- smag.v
       }
     }
     
     if ("tp" %in% organisms.list)
     {
       tips_to_keep.tp <- grep(pattern = "thuja_plicata",tree$tip.label)
       if (length(tips_to_keep.tp) != 0)
       {
         tp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tp]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tp] <- tp.v
       }
     }
     
     if ("aa" %in% organisms.list)
     {
       tips_to_keep.aa <- grep(pattern = "anthoceros_agrestis",tree$tip.label)
       if (length(tips_to_keep.aa) != 0)
       {
         aa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aa]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aa] <- aa.v
       }
     }
     
     if ("um" %in% organisms.list)
     {
       tips_to_keep.um <- grep(pattern = "ulva_mutabilis",tree$tip.label)
       if (length(tips_to_keep.um) != 0)
       {
         um.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.um]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.um] <- um.v
       }
     }
     
     if ("rs" %in% organisms.list)
     {
       tips_to_keep.rs <- grep(pattern = "raphidocelis_subcapitata",tree$tip.label)
       if (length(tips_to_keep.rs) != 0)
       {
         rs.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.rs]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.rs] <- rs.v
       }
     }
     
     if ("cyc" %in% organisms.list)
     {
       tips_to_keep.cyc <- grep(pattern = "cycas_panzhihuaensis",tree$tip.label)
       if (length(tips_to_keep.cyc) != 0)
       {
         cyc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cyc]), "_"), 
                         function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cyc] <- cyc.v
       }
     }
     
     if ("ca" %in% organisms.list)
     {
       tips_to_keep.ca <- grep(pattern = "chlorokybus_atmophyticus",tree$tip.label)
       if (length(tips_to_keep.ca) != 0)
       {
         ca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ca]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ca] <- ca.v
       }
     }
     
     if ("mv" %in% organisms.list)
     {
       tips_to_keep.mv <- grep(pattern = "mesostigma_viride1140",tree$tip.label)
       if (length(tips_to_keep.mv) != 0)
       {
         mv.vec5 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mv]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mv] <- mv.vec5
       }
     }
     
     if ("af" %in% organisms.list)
     {
       tips_to_keep.af <- grep(pattern = "azolla_filiculoides",tree$tip.label)
       if (length(tips_to_keep.af) != 0)
       {
         af.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.af]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.af] <- af.v
       }
     }
     
     if ("sc" %in% organisms.list)
     {
       tips_to_keep.sc <- grep(pattern = "salvinia_cucullata",tree$tip.label)
       if (length(tips_to_keep.sc) != 0)
       {
         sc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sc]), "_"), 
                        function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sc] <- sc.v
       }
     }
     
     if ("aegi" %in% organisms.list)
     {
       tips_to_keep.aegi <- grep(pattern = "aegilops_tauschii",tree$tip.label)
       if (length(tips_to_keep.aegi) != 0)
       {
         aegi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aegi]), "_"), 
                          function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aegi] <- aegi.v
       }
     }
     
     if ("sb" %in% organisms.list)
     {
       tips_to_keep.sb <- grep(pattern = "sorghum_bicolor",tree$tip.label)
       if (length(tips_to_keep.sb) != 0)
       {
         sb.vec5 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sb]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sb] <- sb.vec5
       }
     }
     
     if ("chara" %in% organisms.list)
     {
       tips_to_keep.chara <- grep(pattern = "chara_braunii",tree$tip.label)
       if (length(tips_to_keep.chara) != 0)
       {
         chara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chara]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chara] <- chara.v
       }
     }
     
     if ("sceobli" %in% organisms.list)
     {
       tips_to_keep.sceobli <- grep(pattern = "scenedesmus_obliquus",tree$tip.label)
       if (length(tips_to_keep.sceobli) != 0)
       {
         sceobli.vec5 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sceobli]), "_"), 
                                function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sceobli] <- sceobli.vec5
       }
     }
     
     if ("cocco" %in% organisms.list)
     {
       tips_to_keep.cocco <- grep(pattern = "coccomyxa_subellipsoidea",tree$tip.label)
       if (length(tips_to_keep.cocco) != 0)
       {
         cocco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cocco]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cocco] <- cocco.v
       }
     }
     
     if ("haema" %in% organisms.list)
     {
       tips_to_keep.haema <- grep(pattern = "haematococcus_lacustris",tree$tip.label)
       if (length(tips_to_keep.haema) != 0)
       {
         haema.vec5 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.haema]), "_"), 
                              function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.haema] <- haema.vec5
       }
     }
     
     if ("zm" %in% organisms.list)
     {
       tips_to_keep.zm <- grep(pattern = "zea_mays",tree$tip.label)
       if (length(tips_to_keep.zm) != 0)
       {
         zm.vec5 <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zm]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zm] <- zm.vec5
       }
     }
     
     
     if ("praco" %in% organisms.list)
     {
       tips_to_keep.praco <- grep(pattern = "prasinoderma_coloniale",tree$tip.label)
       if (length(tips_to_keep.praco) != 0)
       {
         praco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.praco]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.praco] <- praco.v
       }
     }
     
     if ("ostlu" %in% organisms.list)
     {
       tips_to_keep.ostlu <- grep(pattern = "ostreococcus_lucimarinus",tree$tip.label)
       if (length(tips_to_keep.ostlu) != 0)
       {
         ostlu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostlu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostlu] <- ostlu.v
       }
     }
     
     if ("ostme" %in% organisms.list)
     {
       tips_to_keep.ostme <- grep(pattern = "ostreococcus_mediterraneus",tree$tip.label)
       if (length(tips_to_keep.ostme) != 0)
       {
         ostme.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostme]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostme] <- ostme.v
       }
     }
     
     if ("micco" %in% organisms.list)
     {
       tips_to_keep.micco <- grep(pattern = "micromonas_commoda",tree$tip.label)
       if (length(tips_to_keep.micco) != 0)
       {
         micco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.micco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.micco] <- micco.v
       }
     }
     
     if ("cymte" %in% organisms.list)
     {
       tips_to_keep.cymte <- grep(pattern = "cymbomonas_tetramitiformis",tree$tip.label)
       if (length(tips_to_keep.cymte) != 0)
       {
         cymte.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cymte]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cymte] <- cymte.v
       }
     }
     
     if ("chlin" %in% organisms.list)
     {
       tips_to_keep.chlin <- grep(pattern = "chlamydomonas_incerta",tree$tip.label)
       if (length(tips_to_keep.chlin) != 0)
       {
         chlin.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlin]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlin] <- chlin.v
       }
     }
     
     if ("gonpe" %in% organisms.list)
     {
       tips_to_keep.gonpe <- grep(pattern = "gonium_pectorale",tree$tip.label)
       if (length(tips_to_keep.gonpe) != 0)
       {
         gonpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gonpe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gonpe] <- gonpe.v
       }
     }
     
     if ("tetso" %in% organisms.list)
     {
       tips_to_keep.tetso <- grep(pattern = "tetrabaena_socialis",tree$tip.label)
       if (length(tips_to_keep.tetso) != 0)
       {
         tetso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tetso] <- tetso.v
       }
     }
     
     if ("desar" %in% organisms.list)
     {
       tips_to_keep.desar <- grep(pattern = "desmodesmus_armatus",tree$tip.label)
       if (length(tips_to_keep.desar) != 0)
       {
         desar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.desar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.desar] <- desar.v
       }
     }
     
     if ("monmi" %in% organisms.list)
     {
       tips_to_keep.monmi <- grep(pattern = "monoraphidium_minutum",tree$tip.label)
       if (length(tips_to_keep.monmi) != 0)
       {
         monmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.monmi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.monmi] <- monmi.v
       }
     }
     
     if ("caule" %in% organisms.list)
     {
       tips_to_keep.caule <- grep(pattern = "caulerpa_lentillifera",tree$tip.label)
       if (length(tips_to_keep.caule) != 0)
       {
         caule.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caule]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.caule] <- caule.v
       }
     }
     
     if ("ostqu" %in% organisms.list)
     {
       tips_to_keep.ostqu <- grep(pattern = "ostreobium_quekettii",tree$tip.label)
       if (length(tips_to_keep.ostqu) != 0)
       {
         ostqu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ostqu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ostqu] <- ostqu.v
       }
     }
     
     if ("brysp" %in% organisms.list)
     {
       tips_to_keep.brysp <- grep(pattern = "bryopsis_sp",tree$tip.label)
       if (length(tips_to_keep.brysp) != 0)
       {
         brysp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brysp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.brysp] <- brysp.v
       }
     }
     
     if ("ellbi" %in% organisms.list)
     {
       tips_to_keep.ellbi <- grep(pattern = "elliptochloris_bilobata",tree$tip.label)
       if (length(tips_to_keep.ellbi) != 0)
       {
         ellbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ellbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ellbi] <- ellbi.v
       }
     }
     
     if ("myrbi" %in% organisms.list)
     {
       tips_to_keep.myrbi <- grep(pattern = "myrmecia_bisecta",tree$tip.label)
       if (length(tips_to_keep.myrbi) != 0)
       {
         myrbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.myrbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.myrbi] <- myrbi.v
       }
     }
     
     if ("apalo" %in% organisms.list)
     {
       tips_to_keep.apalo <- grep(pattern = "apatococcus_lobatus",tree$tip.label)
       if (length(tips_to_keep.apalo) != 0)
       {
         apalo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.apalo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.apalo] <- apalo.v
       }
     }
     
     if ("astgl" %in% organisms.list)
     {
       tips_to_keep.astgl <- grep(pattern = "asterochloris_glomerata",tree$tip.label)
       if (length(tips_to_keep.astgl) != 0)
       {
         astgl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.astgl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.astgl] <- astgl.v
       }
     }
     
     if ("tresp" %in% organisms.list)
     {
       tips_to_keep.tresp <- grep(pattern = "trebouxia_sp",tree$tip.label)
       if (length(tips_to_keep.tresp) != 0)
       {
         tresp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tresp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tresp] <- tresp.v
       }
     }
     
     if ("picso" %in% organisms.list)
     {
       tips_to_keep.picso <- grep(pattern = "picochlorum_soloecismus",tree$tip.label)
       if (length(tips_to_keep.picso) != 0)
       {
         picso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picso] <- picso.v
       }
     }
     
     if ("chlso" %in% organisms.list)
     {
       tips_to_keep.chlso <- grep(pattern = "chlorella_sorokiniana",tree$tip.label)
       if (length(tips_to_keep.chlso) != 0)
       {
         chlso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlso] <- chlso.v
       }
     }
     
     if ("auxpr" %in% organisms.list)
     {
       tips_to_keep.auxpr <- grep(pattern = "auxenochlorella_protothecoides",tree$tip.label)
       if (length(tips_to_keep.auxpr) != 0)
       {
         auxpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.auxpr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.auxpr] <- auxpr.v
       }
     }
     
     if ("helsp" %in% organisms.list)
     {
       tips_to_keep.helsp <- grep(pattern = "helicosporidium_sp",tree$tip.label)
       if (length(tips_to_keep.helsp) != 0)
       {
         helsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.helsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.helsp] <- helsp.v
       }
     }
     
     if ("tetst" %in% organisms.list)
     {
       tips_to_keep.tetst <- grep(pattern = "tetraselmis_striata",tree$tip.label)
       if (length(tips_to_keep.tetst) != 0)
       {
         tetst.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tetst]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tetst] <- tetst.v
       }
     }
     
     if ("pedsp" %in% organisms.list)
     {
       tips_to_keep.pedsp <- grep(pattern = "pedinophyceae_sp",tree$tip.label)
       if (length(tips_to_keep.pedsp) != 0)
       {
         pedsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pedsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pedsp] <- pedsp.v
       }
     }
     
     if ("chlpr" %in% organisms.list)
     {
       tips_to_keep.chlpr <- grep(pattern = "chloropicon_primus",tree$tip.label)
       if (length(tips_to_keep.chlpr) != 0)
       {
         chlpr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chlpr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chlpr] <- chlpr.v
       }
     }
     
     if ("picsp" %in% organisms.list)
     {
       tips_to_keep.picsp <- grep(pattern = "picocystis_sp",tree$tip.label)
       if (length(tips_to_keep.picsp) != 0)
       {
         picsp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picsp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picsp] <- picsp.v
       }
     }
     
     if ("mesvb" %in% organisms.list)
     {
       tips_to_keep.mesvb <- grep(pattern = "mesostigma_viride296",tree$tip.label)
       if (length(tips_to_keep.mesvb) != 0)
       {
         mesvb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mesvb]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mesvb] <- mesvb.v
       }
     }
     
     if ("meskr" %in% organisms.list)
     {
       tips_to_keep.meskr <- grep(pattern = "mesotaenium_kramstae",tree$tip.label)
       if (length(tips_to_keep.meskr) != 0)
       {
         meskr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.meskr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.meskr] <- meskr.v
       }
     }
     
     if ("zygci" %in% organisms.list)
     {
       tips_to_keep.zygci <- grep(pattern = "zygnema_circumcarinatum_",tree$tip.label)
       if (length(tips_to_keep.zygci) != 0)
       {
         zygci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygci]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygci] <- zygci.v
       }
     }
     
     if ("zygcb" %in% organisms.list)
     {
       tips_to_keep.zygcb <- grep(pattern = "zygnema_circumcarinatum1559",tree$tip.label)
       if (length(tips_to_keep.zygcb) != 0)
       {
         zygcb.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcb]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygcb] <- zygcb.v
       }
     }
     
     if ("zygcy" %in% organisms.list)
     {
       tips_to_keep.zygcy <- grep(pattern = "zygnema_cylindricum",tree$tip.label)
       if (length(tips_to_keep.zygcy) != 0)
       {
         zygcy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zygcy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zygcy] <- zygcy.v
       }
     }
     
     if ("plesc" %in% organisms.list)
     {
       tips_to_keep.plesc <- grep(pattern = "pleurozium_schreberi",tree$tip.label)
       if (length(tips_to_keep.plesc) != 0)
       {
         plesc.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.plesc]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.plesc] <- plesc.v
       }
     }
     
     if ("funhy" %in% organisms.list)
     {
       tips_to_keep.funhy <- grep(pattern = "funaria_hygrometrica",tree$tip.label)
       if (length(tips_to_keep.funhy) != 0)
       {
         funhy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.funhy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.funhy] <- funhy.v
       }
     }
     
     if ("sphfa" %in% organisms.list)
     {
       tips_to_keep.sphfa <- grep(pattern = "sphagnum_fallax",tree$tip.label)
       if (length(tips_to_keep.sphfa) != 0)
       {
         sphfa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sphfa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sphfa] <- sphfa.v
       }
     }
     
     if ("takle" %in% organisms.list)
     {
       tips_to_keep.takle <- grep(pattern = "takakia_lepidozioides",tree$tip.label)
       if (length(tips_to_keep.takle) != 0)
       {
         takle.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.takle]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.takle] <- takle.v
       }
     }
     
     if ("marpa" %in% organisms.list)
     {
       tips_to_keep.marpa <- grep(pattern = "marchantia_paleacea",tree$tip.label)
       if (length(tips_to_keep.marpa) != 0)
       {
         marpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.marpa] <- marpa.v
       }
     }
     
     if ("ricfl" %in% organisms.list)
     {
       tips_to_keep.ricfl <- grep(pattern = "riccia_fluitans",tree$tip.label)
       if (length(tips_to_keep.ricfl) != 0)
       {
         ricfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricfl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ricfl] <- ricfl.v
       }
     }
     
     if ("ricso" %in% organisms.list)
     {
       tips_to_keep.ricso <- grep(pattern = "riccia_sorocarpa",tree$tip.label)
       if (length(tips_to_keep.ricso) != 0)
       {
         ricso.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ricso]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ricso] <- ricso.v
       }
     }
     
     if ("antpu" %in% organisms.list)
     {
       tips_to_keep.antpu <- grep(pattern = "anthoceros_punctatus",tree$tip.label)
       if (length(tips_to_keep.antpu) != 0)
       {
         antpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antpu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.antpu] <- antpu.v
       }
     }
     
     if ("antfu" %in% organisms.list)
     {
       tips_to_keep.antfu <- grep(pattern = "anthoceros_fusiformis",tree$tip.label)
       if (length(tips_to_keep.antfu) != 0)
       {
         antfu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.antfu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.antfu] <- antfu.v
       }
     }
     
     if ("megfl" %in% organisms.list)
     {
       tips_to_keep.megfl <- grep(pattern = "megaceros_flagellaris",tree$tip.label)
       if (length(tips_to_keep.megfl) != 0)
       {
         megfl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.megfl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.megfl] <- megfl.v
       }
     }
     
     if ("phach" %in% organisms.list)
     {
       tips_to_keep.phach <- grep(pattern = "phaeomegaceros_chiloensis",tree$tip.label)
       if (length(tips_to_keep.phach) != 0)
       {
         phach.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phach]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phach] <- phach.v
       }
     }
     
     if ("phyph" %in% organisms.list)
     {
       tips_to_keep.phyph <- grep(pattern = "phymatoceros_phymatodes",tree$tip.label)
       if (length(tips_to_keep.phyph) != 0)
       {
         phyph.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phyph]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phyph] <- phyph.v
       }
     }
     
     if ("parpe" %in% organisms.list)
     {
       tips_to_keep.parpe <- grep(pattern = "paraphymatoceros_pearsonii",tree$tip.label)
       if (length(tips_to_keep.parpe) != 0)
       {
         parpe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.parpe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.parpe] <- parpe.v
       }
     }
     
     if ("notor" %in% organisms.list)
     {
       tips_to_keep.notor <- grep(pattern = "notothylas_orbicularis",tree$tip.label)
       if (length(tips_to_keep.notor) != 0)
       {
         notor.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.notor]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.notor] <- notor.v
       }
     }
     
     if ("phaca" %in% organisms.list)
     {
       tips_to_keep.phaca <- grep(pattern = "phaeoceros_carolinianus",tree$tip.label)
       if (length(tips_to_keep.phaca) != 0)
       {
         phaca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phaca] <- phaca.v
       }
     }
     
     if ("phasp" %in% organisms.list)
     {
       tips_to_keep.phasp <- grep(pattern = "phaeoceros_sp",tree$tip.label)
       if (length(tips_to_keep.phasp) != 0)
       {
         phasp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phasp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phasp] <- phasp.v
       }
     }
     
     if ("leidu" %in% organisms.list)
     {
       tips_to_keep.leidu <- grep(pattern = "leiosporoceros_dussii",tree$tip.label)
       if (length(tips_to_keep.leidu) != 0)
       {
         leidu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.leidu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.leidu] <- leidu.v
       }
     }
     
     if ("isosi" %in% organisms.list)
     {
       tips_to_keep.isosi <- grep(pattern = "isoetes_sinensis",tree$tip.label)
       if (length(tips_to_keep.isosi) != 0)
       {
         isosi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.isosi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.isosi] <- isosi.v
       }
     }
     
     if ("dipco" %in% organisms.list)
     {
       tips_to_keep.dipco <- grep(pattern = "diphasiastrum_complanatum",tree$tip.label)
       if (length(tips_to_keep.dipco) != 0)
       {
         dipco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dipco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dipco] <- dipco.v
       }
     }
     
     if ("hupas" %in% organisms.list)
     {
       tips_to_keep.hupas <- grep(pattern = "huperzia_asiatica",tree$tip.label)
       if (length(tips_to_keep.hupas) != 0)
       {
         hupas.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.hupas]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.hupas] <- hupas.v
       }
     }
     
     if ("lyccl" %in% organisms.list)
     {
       tips_to_keep.lyccl <- grep(pattern = "lycopodium_clavatum",tree$tip.label)
       if (length(tips_to_keep.lyccl) != 0)
       {
         lyccl.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lyccl]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lyccl] <- lyccl.v
       }
     }
     
     if ("adica" %in% organisms.list)
     {
       tips_to_keep.adica <- grep(pattern = "adiantum_capillus",tree$tip.label)
       if (length(tips_to_keep.adica) != 0)
       {
         adica.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.adica]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.adica] <- adica.v
       }
     }
     
     if ("alssp" %in% organisms.list)
     {
       tips_to_keep.alssp <- grep(pattern = "alsophila_spinulosa",tree$tip.label)
       if (length(tips_to_keep.alssp) != 0)
       {
         alssp.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.alssp]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.alssp] <- alssp.v
       }
     }
     
     if ("marve" %in% organisms.list)
     {
       tips_to_keep.marve <- grep(pattern = "marsilea_vestita",tree$tip.label)
       if (length(tips_to_keep.marve) != 0)
       {
         marve.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.marve]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.marve] <- marve.v
       }
     }
     
     if ("seqse" %in% organisms.list)
     {
       tips_to_keep.seqse <- grep(pattern = "sequoia_sempervirens",tree$tip.label)
       if (length(tips_to_keep.seqse) != 0)
       {
         seqse.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqse]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.seqse] <- seqse.v
       }
     }
     
     if ("seqgi" %in% organisms.list)
     {
       tips_to_keep.seqgi <- grep(pattern = "sequoiadendron_giganteum",tree$tip.label)
       if (length(tips_to_keep.seqgi) != 0)
       {
         seqgi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.seqgi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.seqgi] <- seqgi.v
       }
     }
     
     if ("taxch" %in% organisms.list)
     {
       tips_to_keep.taxch <- grep(pattern = "taxus_chinensis",tree$tip.label)
       if (length(tips_to_keep.taxch) != 0)
       {
         taxch.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.taxch]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.taxch] <- taxch.v
       }
     }
     
     if ("abial" %in% organisms.list)
     {
       tips_to_keep.abial <- grep(pattern = "abies_alba",tree$tip.label)
       if (length(tips_to_keep.abial) != 0)
       {
         abial.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.abial]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.abial] <- abial.v
       }
     }
     
     if ("picab" %in% organisms.list)
     {
       tips_to_keep.picab <- grep(pattern = "picea_abies",tree$tip.label)
       if (length(tips_to_keep.picab) != 0)
       {
         picab.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.picab]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.picab] <- picab.v
       }
     }
     
     if ("gnemo" %in% organisms.list)
     {
       tips_to_keep.gnemo <- grep(pattern = "gnetum_montanum",tree$tip.label)
       if (length(tips_to_keep.gnemo) != 0)
       {
         gnemo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gnemo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gnemo] <- gnemo.v
       }
     }
     
     if ("welmi" %in% organisms.list)
     {
       tips_to_keep.welmi <- grep(pattern = "welwitschia_mirabilis",tree$tip.label)
       if (length(tips_to_keep.welmi) != 0)
       {
         welmi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.welmi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.welmi] <- welmi.v
       }
     }
     
     if ("ginbi" %in% organisms.list)
     {
       tips_to_keep.ginbi <- grep(pattern = "ginkgo_biloba",tree$tip.label)
       if (length(tips_to_keep.ginbi) != 0)
       {
         ginbi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ginbi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ginbi] <- ginbi.v
       }
     }
     
     if ("ambtr" %in% organisms.list)
     {
       tips_to_keep.ambtr <- grep(pattern = "amborella_trichopoda",tree$tip.label)
       if (length(tips_to_keep.ambtr) != 0)
       {
         ambtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.ambtr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.ambtr] <- ambtr.v
       }
     }
     
     if ("nymco" %in% organisms.list)
     {
       tips_to_keep.nymco <- grep(pattern = "nymphaea_colorata",tree$tip.label)
       if (length(tips_to_keep.nymco) != 0)
       {
         nymco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.nymco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.nymco] <- nymco.v
       }
     }
     
     if ("cinka" %in% organisms.list)
     {
       tips_to_keep.cinka <- grep(pattern = "cinnamomum_kanehirae",tree$tip.label)
       if (length(tips_to_keep.cinka) != 0)
       {
         cinka.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cinka]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cinka] <- cinka.v
       }
     }
     
     
     if ("horvu" %in% organisms.list)
     {
       tips_to_keep.horvu <- grep(pattern = "hordeum_vulgare",tree$tip.label)
       if (length(tips_to_keep.horvu) != 0)
       {
         horvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.horvu]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.horvu] <- horvu.v
       }
     }
     
     if ("bradi" %in% organisms.list)
     {
       tips_to_keep.bradi <- grep(pattern = "brachypodium_distachyon",tree$tip.label)
       if (length(tips_to_keep.bradi) != 0)
       {
         bradi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.bradi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.bradi] <- bradi.v
       }
     }
     
     if ("setit" %in% organisms.list)
     {
       tips_to_keep.setit <- grep(pattern = "setaria_italica",tree$tip.label)
       if (length(tips_to_keep.setit) != 0)
       {
         setit.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.setit]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.setit] <- setit.v
       }
     }
     
     if ("panha" %in% organisms.list)
     {
       tips_to_keep.panha <- grep(pattern = "panicum_hallii",tree$tip.label)
       if (length(tips_to_keep.panha) != 0)
       {
         panha.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.panha]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.panha] <- panha.v
       }
     }
     
     if ("missi" %in% organisms.list)
     {
       tips_to_keep.missi <- grep(pattern = "miscanthus_sinensis",tree$tip.label)
       if (length(tips_to_keep.missi) != 0)
       {
         missi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.missi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.missi] <- missi.v
       }
     }
     
     if ("oroth" %in% organisms.list)
     {
       tips_to_keep.oroth <- grep(pattern = "oropetium_thomaeum",tree$tip.label)
       if (length(tips_to_keep.oroth) != 0)
       {
         oroth.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oroth]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.oroth] <- oroth.v
       }
     }
     
     if ("eleco" %in% organisms.list)
     {
       tips_to_keep.eleco <- grep(pattern = "eleusine_coracana",tree$tip.label)
       if (length(tips_to_keep.eleco) != 0)
       {
         eleco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eleco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eleco] <- eleco.v
       }
     }
     
     if ("anaco" %in% organisms.list)
     {
       tips_to_keep.anaco <- grep(pattern = "ananas_comosus",tree$tip.label)
       if (length(tips_to_keep.anaco) != 0)
       {
         anaco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.anaco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.anaco] <- anaco.v
       }
     }
     
     if ("musac" %in% organisms.list)
     {
       tips_to_keep.musac <- grep(pattern = "musa_acuminata",tree$tip.label)
       if (length(tips_to_keep.musac) != 0)
       {
         musac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.musac]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.musac] <- musac.v
       }
     }
     
     if ("dioal" %in% organisms.list)
     {
       tips_to_keep.dioal <- grep(pattern = "dioscorea_alata",tree$tip.label)
       if (length(tips_to_keep.dioal) != 0)
       {
         dioal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dioal]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dioal] <- dioal.v
       }
     }
     
     if ("spipo" %in% organisms.list)
     {
       tips_to_keep.spipo <- grep(pattern = "spirodela_polyrhiza",tree$tip.label)
       if (length(tips_to_keep.spipo) != 0)
       {
         spipo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.spipo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.spipo] <- spipo.v
       }
     }
     
     if ("aspof" %in% organisms.list)
     {
       tips_to_keep.aspof <- grep(pattern = "asparagus_officinalis",tree$tip.label)
       if (length(tips_to_keep.aspof) != 0)
       {
         aspof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aspof]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aspof] <- aspof.v
       }
     }
     
     if ("zosma" %in% organisms.list)
     {
       tips_to_keep.zosma <- grep(pattern = "zostera_marina",tree$tip.label)
       if (length(tips_to_keep.zosma) != 0)
       {
         zosma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.zosma]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.zosma] <- zosma.v
       }
     }
     
     if ("araly" %in% organisms.list)
     {
       tips_to_keep.araly <- grep(pattern = "arabidopsis_lyrata",tree$tip.label)
       if (length(tips_to_keep.araly) != 0)
       {
         araly.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.araly]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.araly] <- araly.v
       }
     }
     
     if ("capgr" %in% organisms.list)
     {
       tips_to_keep.capgr <- grep(pattern = "capsella_grandiflora",tree$tip.label)
       if (length(tips_to_keep.capgr) != 0)
       {
         capgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.capgr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.capgr] <- capgr.v
       }
     }
     
     if ("brara" %in% organisms.list)
     {
       tips_to_keep.brara <- grep(pattern = "brassica_rapa",tree$tip.label)
       if (length(tips_to_keep.brara) != 0)
       {
         brara.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.brara]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.brara] <- brara.v
       }
     }
     
     if ("braol" %in% organisms.list)
     {
       tips_to_keep.braol <- grep(pattern = "brassica_oleracea",tree$tip.label)
       if (length(tips_to_keep.braol) != 0)
       {
         braol.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.braol]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.braol] <- braol.v
       }
     }
     
     if ("sisir" %in% organisms.list)
     {
       tips_to_keep.sisir <- grep(pattern = "sisymbrium_irio",tree$tip.label)
       if (length(tips_to_keep.sisir) != 0)
       {
         sisir.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sisir]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sisir] <- sisir.v
       }
     }
     
     if ("schpa" %in% organisms.list)
     {
       tips_to_keep.schpa <- grep(pattern = "schrenkiella_parvula",tree$tip.label)
       if (length(tips_to_keep.schpa) != 0)
       {
         schpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.schpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.schpa] <- schpa.v
       }
     }
     
     if ("eutsa" %in% organisms.list)
     {
       tips_to_keep.eutsa <- grep(pattern = "eutrema_salsugineum",tree$tip.label)
       if (length(tips_to_keep.eutsa) != 0)
       {
         eutsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eutsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eutsa] <- eutsa.v
       }
     }
     
     if ("thlar" %in% organisms.list)
     {
       tips_to_keep.thlar <- grep(pattern = "thlaspi_arvense",tree$tip.label)
       if (length(tips_to_keep.thlar) != 0)
       {
         thlar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.thlar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.thlar] <- thlar.v
       }
     }
     
     if ("boest" %in% organisms.list)
     {
       tips_to_keep.boest <- grep(pattern = "boechera_stricta",tree$tip.label)
       if (length(tips_to_keep.boest) != 0)
       {
         boest.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.boest]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.boest] <- boest.v
       }
     }
     
     if ("carpa" %in% organisms.list)
     {
       tips_to_keep.carpa <- grep(pattern = "carica_papaya",tree$tip.label)
       if (length(tips_to_keep.carpa) != 0)
       {
         carpa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.carpa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.carpa] <- carpa.v
       }
     }
     
     if ("theca" %in% organisms.list)
     {
       tips_to_keep.theca <- grep(pattern = "theobroma_cacao",tree$tip.label)
       if (length(tips_to_keep.theca) != 0)
       {
         theca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.theca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.theca] <- theca.v
       }
     }
     
     if ("goshi" %in% organisms.list)
     {
       tips_to_keep.goshi <- grep(pattern = "gossypium_hirsutum",tree$tip.label)
       if (length(tips_to_keep.goshi) != 0)
       {
         goshi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.goshi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.goshi] <- goshi.v
       }
     }
     
     if ("gosra" %in% organisms.list)
     {
       tips_to_keep.gosra <- grep(pattern = "gossypium_raimondii",tree$tip.label)
       if (length(tips_to_keep.gosra) != 0)
       {
         gosra.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.gosra]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.gosra] <- gosra.v
       }
     }
     
     if ("citsi" %in% organisms.list)
     {
       tips_to_keep.citsi <- grep(pattern = "citrus_sinensis",tree$tip.label)
       if (length(tips_to_keep.citsi) != 0)
       {
         citsi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.citsi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.citsi] <- citsi.v
       }
     }
     
     if ("pontr" %in% organisms.list)
     {
       tips_to_keep.pontr <- grep(pattern = "poncirus_trifoliata",tree$tip.label)
       if (length(tips_to_keep.pontr) != 0)
       {
         pontr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.pontr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.pontr] <- pontr.v
       }
     }
     
     if ("eucgr" %in% organisms.list)
     {
       tips_to_keep.eucgr <- grep(pattern = "eucalyptus_grandis",tree$tip.label)
       if (length(tips_to_keep.eucgr) != 0)
       {
         eucgr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.eucgr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.eucgr] <- eucgr.v
       }
     }
     
     if ("maldo" %in% organisms.list)
     {
       tips_to_keep.maldo <- grep(pattern = "malus_domestica",tree$tip.label)
       if (length(tips_to_keep.maldo) != 0)
       {
         maldo.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.maldo]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.maldo] <- maldo.v
       }
     }
     
     if ("prupe" %in% organisms.list)
     {
       tips_to_keep.prupe <- grep(pattern = "prunus_persica",tree$tip.label)
       if (length(tips_to_keep.prupe) != 0)
       {
         prupe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.prupe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.prupe] <- prupe.v
       }
     }
     
     if ("frave" %in% organisms.list)
     {
       tips_to_keep.frave <- grep(pattern = "fragaria_vesca",tree$tip.label)
       if (length(tips_to_keep.frave) != 0)
       {
         frave.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.frave]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.frave] <- frave.v
       }
     }
     
     if ("corav" %in% organisms.list)
     {
       tips_to_keep.corav <- grep(pattern = "corylus_avellana",tree$tip.label)
       if (length(tips_to_keep.corav) != 0)
       {
         corav.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corav]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.corav] <- corav.v
       }
     }
     
     if ("caril" %in% organisms.list)
     {
       tips_to_keep.caril <- grep(pattern = "carya_illinoinensis",tree$tip.label)
       if (length(tips_to_keep.caril) != 0)
       {
         caril.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.caril]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.caril] <- caril.v
       }
     }
     
     if ("queru" %in% organisms.list)
     {
       tips_to_keep.queru <- grep(pattern = "quercus_rubra",tree$tip.label)
       if (length(tips_to_keep.queru) != 0)
       {
         queru.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.queru]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.queru] <- queru.v
       }
     }
     
     if ("cucsa" %in% organisms.list)
     {
       tips_to_keep.cucsa <- grep(pattern = "cucumis_sativus",tree$tip.label)
       if (length(tips_to_keep.cucsa) != 0)
       {
         cucsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cucsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cucsa] <- cucsa.v
       }
     }
     
     if ("glyma" %in% organisms.list)
     {
       tips_to_keep.glyma <- grep(pattern = "glycine_max",tree$tip.label)
       if (length(tips_to_keep.glyma) != 0)
       {
         glyma.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.glyma]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.glyma] <- glyma.v
       }
     }
     
     if ("medtr" %in% organisms.list)
     {
       tips_to_keep.medtr <- grep(pattern = "medicago_truncatula",tree$tip.label)
       if (length(tips_to_keep.medtr) != 0)
       {
         medtr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.medtr]), "_"), 
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.medtr] <- medtr.v
       }
     }
     
     if ("tripr" %in% organisms.list)
     {
       tips_to_keep.tripr <- grep(pattern = "trifolium_pratense",tree$tip.label)
       if (length(tips_to_keep.tripr) != 0)
       {
         tripr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.tripr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.tripr] <- tripr.v
       }
     }
     
     if ("cicar" %in% organisms.list)
     {
       tips_to_keep.cicar <- grep(pattern = "cicer_arietinum",tree$tip.label)
       if (length(tips_to_keep.cicar) != 0)
       {
         cicar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cicar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cicar] <- cicar.v
       }
     }
     
     if ("lotja" %in% organisms.list)
     {
       tips_to_keep.lotja <- grep(pattern = "lotus_japonicus",tree$tip.label)
       if (length(tips_to_keep.lotja) != 0)
       {
         lotja.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lotja]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lotja] <- lotja.v
       }
     }
     
     if ("phaac" %in% organisms.list)
     {
       tips_to_keep.phaac <- grep(pattern = "phaseolus_acutifolius",tree$tip.label)
       if (length(tips_to_keep.phaac) != 0)
       {
         phaac.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.phaac]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.phaac] <- phaac.v
       }
     }
     
     if ("vigun" %in% organisms.list)
     {
       tips_to_keep.vigun <- grep(pattern = "vigna_unguiculata",tree$tip.label)
       if (length(tips_to_keep.vigun) != 0)
       {
         vigun.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vigun]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vigun] <- vigun.v
       }
     }
     
     if ("lupal" %in% organisms.list)
     {
       tips_to_keep.lupal <- grep(pattern = "lupinus_albus",tree$tip.label)
       if (length(tips_to_keep.lupal) != 0)
       {
         lupal.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lupal]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lupal] <- lupal.v
       }
     }
     
     if ("lencu" %in% organisms.list)
     {
       tips_to_keep.lencu <- grep(pattern = "lens_culinaris",tree$tip.label)
       if (length(tips_to_keep.lencu) != 0)
       {
         lencu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lencu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lencu] <- lencu.v
       }
     }
     
     if ("arahy" %in% organisms.list)
     {
       tips_to_keep.arahy <- grep(pattern = "arachis_hypogaea",tree$tip.label)
       if (length(tips_to_keep.arahy) != 0)
       {
         arahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.arahy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.arahy] <- arahy.v
       }
     }
     
     if ("poptr" %in% organisms.list)
     {
       tips_to_keep.poptr <- grep(pattern = "populus_trichocarpa",tree$tip.label)
       if (length(tips_to_keep.poptr) != 0)
       {
         poptr.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.poptr]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.poptr] <- poptr.v
       }
     }
     
     if ("manes" %in% organisms.list)
     {
       tips_to_keep.manes <- grep(pattern = "manihot_esculenta",tree$tip.label)
       if (length(tips_to_keep.manes) != 0)
       {
         manes.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.manes]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.manes] <- manes.v
       }
     }
     
     if ("salpu" %in% organisms.list)
     {
       tips_to_keep.salpu <- grep(pattern = "salix_purpurea",tree$tip.label)
       if (length(tips_to_keep.salpu) != 0)
       {
         salpu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.salpu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.salpu] <- salpu.v
       }
     }
     
     if ("linus" %in% organisms.list)
     {
       tips_to_keep.linus <- grep(pattern = "linum_usitatissimum",tree$tip.label)
       if (length(tips_to_keep.linus) != 0)
       {
         linus.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.linus]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.linus] <- linus.v
       }
     }
     
     if ("corci" %in% organisms.list)
     {
       tips_to_keep.corci <- grep(pattern = "corymbia_citriodora",tree$tip.label)
       if (length(tips_to_keep.corci) != 0)
       {
         corci.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.corci]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.corci] <- corci.v
       }
     }
     
     if ("vitvi" %in% organisms.list)
     {
       tips_to_keep.vitvi <- grep(pattern = "vitis_vinifera",tree$tip.label)
       if (length(tips_to_keep.vitvi) != 0)
       {
         vitvi.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vitvi]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vitvi] <- vitvi.v
       }
     }
     
     if ("kalfe" %in% organisms.list)
     {
       tips_to_keep.kalfe <- grep(pattern = "kalanchoe_fedtschenkoi",tree$tip.label)
       if (length(tips_to_keep.kalfe) != 0)
       {
         kalfe.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.kalfe]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.kalfe] <- kalfe.v
       }
     }
     
     if ("vacda" %in% organisms.list)
     {
       tips_to_keep.vacda <- grep(pattern = "vaccinium_darrowii",tree$tip.label)
       if (length(tips_to_keep.vacda) != 0)
       {
         vacda.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.vacda]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.vacda] <- vacda.v
       }
     }
     
     if ("oleeu" %in% organisms.list)
     {
       tips_to_keep.oleeu <- grep(pattern = "olea_europaea",tree$tip.label)
       if (length(tips_to_keep.oleeu) != 0)
       {
         oleeu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.oleeu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.oleeu] <- oleeu.v
       }
     }
     
     if ("mimgu" %in% organisms.list)
     {
       tips_to_keep.mimgu <- grep(pattern = "mimulus_guttatus",tree$tip.label)
       if (length(tips_to_keep.mimgu) != 0)
       {
         mimgu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.mimgu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.mimgu] <- mimgu.v
       }
     }
     
     if ("soltu" %in% organisms.list)
     {
       tips_to_keep.soltu <- grep(pattern = "solanum_tuberosum",tree$tip.label)
       if (length(tips_to_keep.soltu) != 0)
       {
         soltu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.soltu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.soltu] <- soltu.v
       }
     }
     
     if ("cofar" %in% organisms.list)
     {
       tips_to_keep.cofar <- grep(pattern = "coffea_arabica",tree$tip.label)
       if (length(tips_to_keep.cofar) != 0)
       {
         cofar.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.cofar]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.cofar] <- cofar.v
       }
     }
     
     if ("lacsa" %in% organisms.list)
     {
       tips_to_keep.lacsa <- grep(pattern = "lactuca_sativa",tree$tip.label)
       if (length(tips_to_keep.lacsa) != 0)
       {
         lacsa.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.lacsa]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.lacsa] <- lacsa.v
       }
     }
     
     if ("dauca" %in% organisms.list)
     {
       tips_to_keep.dauca <- grep(pattern = "daucus_carota",tree$tip.label)
       if (length(tips_to_keep.dauca) != 0)
       {
         dauca.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.dauca]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.dauca] <- dauca.v
       }
     }
     
     if ("betvu" %in% organisms.list)
     {
       tips_to_keep.betvu <- grep(pattern = "beta_vulgaris",tree$tip.label)
       if (length(tips_to_keep.betvu) != 0)
       {
         betvu.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.betvu]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.betvu] <- betvu.v
       }
     }
     
     if ("amahy" %in% organisms.list)
     {
       tips_to_keep.amahy <- grep(pattern = "amaranthus_hypochondriacus",tree$tip.label)
       if (length(tips_to_keep.amahy) != 0)
       {
         amahy.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.amahy]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.amahy] <- amahy.v
       }
     }
     
     if ("chequ" %in% organisms.list)
     {
       tips_to_keep.chequ <- grep(pattern = "chenopodium_quinoa",tree$tip.label)
       if (length(tips_to_keep.chequ) != 0)
       {
         chequ.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.chequ]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.chequ] <- chequ.v
       }
     }
     
     if ("sapof" %in% organisms.list)
     {
       tips_to_keep.sapof <- grep(pattern = "saponaria_officinalis",tree$tip.label)
       if (length(tips_to_keep.sapof) != 0)
       {
         sapof.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.sapof]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.sapof] <- sapof.v
       }
     }
     
     if ("aquco" %in% organisms.list)
     {
       tips_to_keep.aquco <- grep(pattern = "aquilegia_coerulea",tree$tip.label)
       if (length(tips_to_keep.aquco) != 0)
       {
         aquco.v <- sapply(strsplit(as.character(tree$tip.label[tips_to_keep.aquco]), "_"),
                           function(x) paste0(x[-c(1,2)], collapse="_"))
         tree$tip.label[tips_to_keep.aquco] <- aquco.v
       }
     }
     
     return(tree)
   }) %>% bindEvent(input$run_button5)
   
   # Generate reduced tree when the corresponding button is activated
   tree_reduced5 <- reactive({
     
     tree <- tree_adj5()
     # Define tips to keep (selected organisms and query) and generate the reduced tree
     tips_to_keep.global <- c(tips_to_keep.praco5(), tips_to_keep.ot5(), tips_to_keep.ostlu5(), tips_to_keep.ostme5(), tips_to_keep.bp5(),
                              tips_to_keep.mi5(), tips_to_keep.micco5(), tips_to_keep.cymte5(), tips_to_keep.cr5(), tips_to_keep.chlin5(),
                              tips_to_keep.haema5(), tips_to_keep.vc5(), tips_to_keep.ds5(), tips_to_keep.gonpe5(), tips_to_keep.tetso5(),
                              tips_to_keep.cz5(), tips_to_keep.rs5(), tips_to_keep.sceobli5(), tips_to_keep.desar5(), tips_to_keep.monmi5(),
                              tips_to_keep.caule5(), tips_to_keep.ostqu5(), tips_to_keep.brysp5(), tips_to_keep.cocco5(),
                              tips_to_keep.ellbi5(), tips_to_keep.myrbi5(), tips_to_keep.apalo5(), tips_to_keep.astgl5(),
                              tips_to_keep.tresp5(), tips_to_keep.picso5(), tips_to_keep.chlso5(), tips_to_keep.auxpr5(),
                              tips_to_keep.helsp5(), tips_to_keep.um5(), tips_to_keep.tetst5(), tips_to_keep.pedsp5(),
                              tips_to_keep.chlpr5(), tips_to_keep.picsp5(), tips_to_keep.ca5(), tips_to_keep.mv5(), tips_to_keep.mesvb5(),
                              tips_to_keep.kn5(), tips_to_keep.chara5(), tips_to_keep.me5(), tips_to_keep.meskr5(), tips_to_keep.sp5(),
                              tips_to_keep.zygci5(), tips_to_keep.zygcb5(), tips_to_keep.zygcy5(),
                              tips_to_keep.cp5(), tips_to_keep.pp5(), tips_to_keep.plesc5(), tips_to_keep.funhy5(),
                              tips_to_keep.smag5(), tips_to_keep.sphfa5(), tips_to_keep.takle5(), tips_to_keep.mp5(),
                              tips_to_keep.marpa5(), tips_to_keep.ricfl5(), tips_to_keep.ricso5(), tips_to_keep.aa5(),
                              tips_to_keep.antpu5(), tips_to_keep.antfu5(), tips_to_keep.megfl5(), tips_to_keep.phach5(),
                              tips_to_keep.phyph5(), tips_to_keep.parpe5(), tips_to_keep.notor5(), tips_to_keep.phaca5(),
                              tips_to_keep.phasp5(), tips_to_keep.leidu5(), tips_to_keep.sm5(), tips_to_keep.isosi5(), 
                              tips_to_keep.dipco5(), tips_to_keep.hupas5(), tips_to_keep.lyccl5(),
                              tips_to_keep.cri5(), tips_to_keep.adica5(), tips_to_keep.alssp5(), tips_to_keep.sc5(), 
                              tips_to_keep.af5(), tips_to_keep.marve5(), tips_to_keep.tp5(), tips_to_keep.seqse5(), 
                              tips_to_keep.seqgi5(), tips_to_keep.taxch5(), tips_to_keep.abial5(), tips_to_keep.picab5(),
                              tips_to_keep.gnemo5(), tips_to_keep.welmi5(), tips_to_keep.cyc5(), tips_to_keep.ginbi5(),
                              tips_to_keep.ambtr5(), tips_to_keep.nymco5(), tips_to_keep.cinka5(), tips_to_keep.aegi5(),
                              tips_to_keep.ta5(), tips_to_keep.horvu5(), tips_to_keep.bradi5(), tips_to_keep.os5(),
                              tips_to_keep.setit5(), tips_to_keep.sb5(), tips_to_keep.zm5(), tips_to_keep.panha5(),
                              tips_to_keep.missi5(), tips_to_keep.oroth5(), tips_to_keep.eleco5(), tips_to_keep.anaco5(),
                              tips_to_keep.musac5(), tips_to_keep.dioal5(), tips_to_keep.spipo5(), tips_to_keep.aspof5(),
                              tips_to_keep.zosma5(), tips_to_keep.at5(), tips_to_keep.araly5(), tips_to_keep.capgr5(),
                              tips_to_keep.brara5(), tips_to_keep.braol5(), tips_to_keep.sisir5(), tips_to_keep.schpa5(),
                              tips_to_keep.eutsa5(), tips_to_keep.thlar5(), tips_to_keep.boest5(), tips_to_keep.carpa5(),
                              tips_to_keep.theca5(), tips_to_keep.goshi5(), tips_to_keep.gosra5(), tips_to_keep.citsi5(),
                              tips_to_keep.pontr5(), tips_to_keep.eucgr5(), tips_to_keep.maldo5(), tips_to_keep.prupe5(),
                              tips_to_keep.frave5(), tips_to_keep.corav5(), tips_to_keep.caril5(), tips_to_keep.queru5(),
                              tips_to_keep.cucsa5(), tips_to_keep.glyma5(), tips_to_keep.medtr5(), tips_to_keep.tripr5(),
                              tips_to_keep.cicar5(), tips_to_keep.lotja5(), tips_to_keep.phaac5(), tips_to_keep.vigun5(),
                              tips_to_keep.lupal5(), tips_to_keep.lencu5(), tips_to_keep.arahy5(), tips_to_keep.poptr5(),
                              tips_to_keep.manes5(), tips_to_keep.salpu5(), tips_to_keep.linus5(), tips_to_keep.corci5(),
                              tips_to_keep.vitvi5(), tips_to_keep.kalfe5(), tips_to_keep.vacda5(), tips_to_keep.oleeu5(),
                              tips_to_keep.mimgu5(), tips_to_keep.sl5(), tips_to_keep.soltu5(), tips_to_keep.cofar5(),
                              tips_to_keep.lacsa5(), tips_to_keep.dauca5(), tips_to_keep.betvu5(), tips_to_keep.amahy5(),
                              tips_to_keep.chequ5(), tips_to_keep.sapof5(), tips_to_keep.aquco5(), tips_to_keep.query5())
     
     # Error message if trying to build tree with less than two tips
     {
     if (length(tips_to_keep.global) < 2)
     {
       shinyjs::hideElement(id = 'loading.tree5')
       
       if (UI_exist_tree5)
       {
         removeUI(
           selector = "div:has(>> #treeTips5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>>> #presentorg5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_image5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTree5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadNewick5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadTreeSeqs5)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_tree5 <<- F
       output$error_tree5 <- renderUI({renderText({print("Unable to construct
      tree with a single tip, please select more organisms.")})})
       validate(need(length(tips_to_keep.global) > 1, " "))
     }
     }
     
     tips_to_drop <- setdiff(1:length(tree$tip.label), tips_to_keep.global)
     tree_reduced <- drop.tip(tree, tips_to_drop)
     
     return(tree_reduced)
   }) %>% bindEvent(input$run_button5)
   
   organims_reduced5 <- reactive({
     
     tree_reduced <- tree_reduced5()
     shoot_msa <- shoot_msa5()
     
     len.mp <- length(tips_to_keep.mp5())
     len.ot <- length(tips_to_keep.ot5())
     len.at <- length(tips_to_keep.at5())
     len.cp <- length(tips_to_keep.cp5())
     len.cr <- length(tips_to_keep.cr5())
     len.cz <- length(tips_to_keep.cz5())
     len.kn <- length(tips_to_keep.kn5())
     len.me <- length(tips_to_keep.me5())
     len.mi <- length(tips_to_keep.mi5())
     len.pp <- length(tips_to_keep.pp5())
     len.sl <- length(tips_to_keep.sl5())
     len.sm <- length(tips_to_keep.sm5())
     len.sp <- length(tips_to_keep.sp5())
     len.ta <- length(tips_to_keep.ta5())
     len.vc <- length(tips_to_keep.vc5())
     len.bp <- length(tips_to_keep.bp5())
     len.cri <- length(tips_to_keep.cri5())
     len.ds <- length(tips_to_keep.ds5())
     len.os <- length(tips_to_keep.os5())
     len.smag <- length(tips_to_keep.smag5())
     len.tp <- length(tips_to_keep.tp5())
     len.aa <- length(tips_to_keep.aa5())
     len.um <- length(tips_to_keep.um5())
     len.rs <- length(tips_to_keep.rs5())
     len.cyc <- length(tips_to_keep.cyc5())
     len.ca <- length(tips_to_keep.ca5())
     len.mv <- length(tips_to_keep.mv5())
     len.af <- length(tips_to_keep.af5())
     len.sc <- length(tips_to_keep.sc5())
     len.aegi <- length(tips_to_keep.aegi5())
     len.sb <- length(tips_to_keep.sb5())
     len.chara <- length(tips_to_keep.chara5())
     len.sceobli <- length(tips_to_keep.sceobli5())
     len.cocco <- length(tips_to_keep.cocco5())
     len.haema <- length(tips_to_keep.haema5())
     len.zea <- length(tips_to_keep.zm5())
     len.praco <- length(tips_to_keep.praco5())
     len.ostlu <- length(tips_to_keep.ostlu5())
     len.ostme <- length(tips_to_keep.ostme5())
     len.micco <- length(tips_to_keep.micco5())
     len.cymte <- length(tips_to_keep.cymte5())
     len.chlin <- length(tips_to_keep.chlin5())
     len.gonpe <- length(tips_to_keep.gonpe5())
     len.tetso <- length(tips_to_keep.tetso5())
     len.desar <- length(tips_to_keep.desar5())
     len.monmi <- length(tips_to_keep.monmi5())
     len.caule <- length(tips_to_keep.caule5())
     len.ostqu <- length(tips_to_keep.ostqu5())
     len.brysp <- length(tips_to_keep.brysp5())
     len.ellbi <- length(tips_to_keep.ellbi5())
     len.myrbi <- length(tips_to_keep.myrbi5())
     len.apalo <- length(tips_to_keep.apalo5())
     len.astgl <- length(tips_to_keep.astgl5())
     len.tresp <- length(tips_to_keep.tresp5())
     len.picso <- length(tips_to_keep.picso5())
     len.chlso <- length(tips_to_keep.chlso5())
     len.auxpr <- length(tips_to_keep.auxpr5())
     len.helsp <- length(tips_to_keep.helsp5())
     len.tetst <- length(tips_to_keep.tetst5())
     len.pedsp <- length(tips_to_keep.pedsp5())
     len.chlpr <- length(tips_to_keep.chlpr5())
     len.picsp <- length(tips_to_keep.picsp5())
     len.mesvb <- length(tips_to_keep.mesvb5())
     len.meskr <- length(tips_to_keep.meskr5())
     len.zygci <- length(tips_to_keep.zygci5())
     len.zygcb <- length(tips_to_keep.zygcb5())
     len.zygcy <- length(tips_to_keep.zygcy5())
     len.plesc <- length(tips_to_keep.plesc5())
     len.funhy <- length(tips_to_keep.funhy5())
     len.sphfa <- length(tips_to_keep.sphfa5())
     len.takle <- length(tips_to_keep.takle5())
     len.marpa <- length(tips_to_keep.marpa5())
     len.ricfl <- length(tips_to_keep.ricfl5())
     len.ricso <- length(tips_to_keep.ricso5())
     len.antpu <- length(tips_to_keep.antpu5())
     len.antfu <- length(tips_to_keep.antfu5())
     len.megfl <- length(tips_to_keep.megfl5())
     len.phach <- length(tips_to_keep.phach5())
     len.phyph <- length(tips_to_keep.phyph5())
     len.parpe <- length(tips_to_keep.parpe5())
     len.notor <- length(tips_to_keep.notor5())
     len.phaca <- length(tips_to_keep.phaca5())
     len.phasp <- length(tips_to_keep.phasp5())
     len.leidu <- length(tips_to_keep.leidu5())
     len.isosi <- length(tips_to_keep.isosi5())
     len.dipco <- length(tips_to_keep.dipco5())
     len.hupas <- length(tips_to_keep.hupas5())
     len.lyccl <- length(tips_to_keep.lyccl5())
     len.adica <- length(tips_to_keep.adica5())
     len.alssp <- length(tips_to_keep.alssp5())
     len.marve <- length(tips_to_keep.marve5())
     len.seqse <- length(tips_to_keep.seqse5())
     len.seqgi <- length(tips_to_keep.seqgi5())
     len.taxch <- length(tips_to_keep.taxch5())
     len.abial <- length(tips_to_keep.abial5())
     len.picab <- length(tips_to_keep.picab5())
     len.gnemo <- length(tips_to_keep.gnemo5())
     len.welmi <- length(tips_to_keep.welmi5())
     len.ginbi <- length(tips_to_keep.ginbi5())
     len.ambtr <- length(tips_to_keep.ambtr5())
     len.nymco <- length(tips_to_keep.nymco5())
     len.cinka <- length(tips_to_keep.cinka5())
     len.horvu <- length(tips_to_keep.horvu5())
     len.bradi <- length(tips_to_keep.bradi5())
     len.setit <- length(tips_to_keep.setit5())
     len.panha <- length(tips_to_keep.panha5())
     len.missi <- length(tips_to_keep.missi5())
     len.oroth <- length(tips_to_keep.oroth5())
     len.eleco <- length(tips_to_keep.eleco5())
     len.anaco <- length(tips_to_keep.anaco5())
     len.musac <- length(tips_to_keep.musac5())
     len.dioal <- length(tips_to_keep.dioal5())
     len.spipo <- length(tips_to_keep.spipo5())
     len.aspof <- length(tips_to_keep.aspof5())
     len.zosma <- length(tips_to_keep.zosma5())
     len.araly <- length(tips_to_keep.araly5())
     len.capgr <- length(tips_to_keep.capgr5())
     len.brara <- length(tips_to_keep.brara5())
     len.braol <- length(tips_to_keep.braol5())
     len.sisir <- length(tips_to_keep.sisir5())
     len.schpa <- length(tips_to_keep.schpa5())
     len.eutsa <- length(tips_to_keep.eutsa5())
     len.thlar <- length(tips_to_keep.thlar5())
     len.boest <- length(tips_to_keep.boest5())
     len.carpa <- length(tips_to_keep.carpa5())
     len.theca <- length(tips_to_keep.theca5())
     len.goshi <- length(tips_to_keep.goshi5())
     len.gosra <- length(tips_to_keep.gosra5())
     len.citsi <- length(tips_to_keep.citsi5())
     len.pontr <- length(tips_to_keep.pontr5())
     len.eucgr <- length(tips_to_keep.eucgr5())
     len.maldo <- length(tips_to_keep.maldo5())
     len.prupe <- length(tips_to_keep.prupe5())
     len.frave <- length(tips_to_keep.frave5())
     len.corav <- length(tips_to_keep.corav5())
     len.caril <- length(tips_to_keep.caril5())
     len.queru <- length(tips_to_keep.queru5())
     len.cucsa <- length(tips_to_keep.cucsa5())
     len.glyma <- length(tips_to_keep.glyma5())
     len.medtr <- length(tips_to_keep.medtr5())
     len.tripr <- length(tips_to_keep.tripr5())
     len.cicar <- length(tips_to_keep.cicar5())
     len.lotja <- length(tips_to_keep.lotja5())
     len.phaac <- length(tips_to_keep.phaac5())
     len.vigun <- length(tips_to_keep.vigun5())
     len.lupal <- length(tips_to_keep.lupal5())
     len.lencu <- length(tips_to_keep.lencu5())
     len.arahy <- length(tips_to_keep.arahy5())
     len.poptr <- length(tips_to_keep.poptr5())
     len.manes <- length(tips_to_keep.manes5())
     len.salpu <- length(tips_to_keep.salpu5())
     len.linus <- length(tips_to_keep.linus5())
     len.corci <- length(tips_to_keep.corci5())
     len.vitvi <- length(tips_to_keep.vitvi5())
     len.kalfe <- length(tips_to_keep.kalfe5())
     len.vacda <- length(tips_to_keep.vacda5())
     len.oleeu <- length(tips_to_keep.oleeu5())
     len.mimgu <- length(tips_to_keep.mimgu5())
     len.soltu <- length(tips_to_keep.soltu5())
     len.cofar <- length(tips_to_keep.cofar5())
     len.lacsa <- length(tips_to_keep.lacsa5())
     len.dauca <- length(tips_to_keep.dauca5())
     len.betvu <- length(tips_to_keep.betvu5())
     len.amahy <- length(tips_to_keep.amahy5())
     len.chequ <- length(tips_to_keep.chequ5())
     len.sapof <- length(tips_to_keep.sapof5())
     len.aquco <- length(tips_to_keep.aquco5())
     
     
     organims_reduced <- c(rep("Prasinoderma coloniale", len.praco),rep("Ostreococcus tauri", len.ot),
                           rep("Ostreococcus lucimarinus", len.ostlu),rep("Ostreococcus mediterraneus", len.ostme),
                           rep("Bathycoccus prasinos", len.bp),rep("Micromonas pusilla", len.mi),
                           rep("Micromonas commoda", len.micco),rep("Cymbomonas tetramitiformis", len.cymte),
                           rep("Chlamydomonas reinhardtii", len.cr),rep("Chlamydomonas incerta", len.chlin),
                           rep("Haematococcus lacustris", len.haema),rep("Volvox carteri", len.vc),
                           rep("Dunaliella salina", len.ds),rep("Gonium pectorale", len.gonpe),
                           rep("Tetrabaena socialis", len.tetso),rep("Chromochloris zofingiensis", len.cz),
                           rep("Raphidocelis subcapitata", len.rs),rep("Scenedesmus obliquus", len.sceobli),
                           rep("Desmodesmus armatus", len.desar),rep("Monoraphidium minutum", len.monmi),
                           rep("Caulerpa lentillifera", len.caule),rep("Ostreobium quekettii", len.ostqu),
                           rep("Bryopsis sp.", len.brysp),rep("Coccomyxa subellipsoidea", len.cocco),
                           rep("Elliptochloris bilobata", len.ellbi),rep("Myrmecia bisecta", len.myrbi),
                           rep("Apatococcus lobatus", len.apalo),rep("Asterochloris glomerata", len.astgl),
                           rep("Trebouxia sp.", len.tresp),rep("Picochlorum soloecismus", len.picso),
                           rep("Chlorella sorokiniana", len.chlso),rep("Auxenochlorella protothecoides", len.auxpr),
                           rep("Helicosporidium sp.", len.helsp),rep("Ulva mutabilis", len.um),
                           rep("Tetraselmis striata", len.tetst),rep("Pedinophyceae sp.", len.pedsp),
                           rep("Chloropicon primus", len.chlpr),rep("Picocystis sp.", len.picsp),
                           rep("Chlorokybus atmophyticus", len.ca),rep("Mesostigma viride CCAC 1140", len.mv),
                           rep("Mesostigma viride NIES 596", len.mesvb),rep("Klebsormidium nitens", len.kn),
                           rep("Chara braunii", len.chara),rep("Mesotaenium endlicherianum", len.me),
                           rep("Mesotaenium kramstae", len.meskr),rep("Spirogloea muscicola", len.sp),
                           rep("Zygnema circumcarinatum SAG", len.zygci),rep("Zygnema circumcarinatum UTEX 1559", len.zygcb),
                           rep("Zygnema cylindricum", len.zygcy),rep("Ceratodon purpureus", len.cp),
                           rep("Physcomitrium patens", len.pp),rep("Pleurozium schreberi", len.plesc),
                           rep("Funaria hygrometrica", len.funhy),rep("Sphagnum magellanicum", len.smag),
                           rep("Sphagnum fallax", len.sphfa),rep("Takakia lepidozioides", len.takle),
                           rep("Marchantia polymorpha", len.mp),rep("Marchantia paleacea", len.marpa),
                           rep("Riccia fluitans", len.ricfl),rep("Riccia sorocarpa", len.ricso),
                           rep("Anthoceros agrestis", len.aa),rep("Anthoceros punctatus", len.antpu),
                           rep("Anthoceros fusiformis", len.antfu),rep("Megaceros flagellaris", len.megfl),
                           rep("Phaeomegaceros chiloensis", len.phach),rep("Phymatoceros phymatodes", len.phyph),
                           rep("Paraphymatoceros pearsonii", len.parpe),rep("Notothylas orbicularis", len.notor),
                           rep("Phaeoceros carolinianus", len.phaca),rep("Phaeoceros sp.", len.phasp),
                           rep("Leiosporoceros dussii", len.leidu),rep("Selaginella moellendorffii", len.sm),
                           rep("Isoetes sinensis", len.isosi),rep("Diphasiastrum complanatum", len.dipco),
                           rep("Huperzia asiatica", len.hupas),rep("Lycopodium clavatum", len.lyccl),
                           rep("Ceratopteris richardii", len.cri),rep("Adiantum capillus", len.adica),
                           rep("Alsophila spinulosa", len.alssp),rep("Salvinia cucullata", len.sc),
                           rep("Azolla filiculoides", len.af),rep("Marsilea vestita", len.marve),
                           rep("Thuja plicata", len.tp),rep("Sequoia sempervirens", len.seqse),
                           rep("Sequoiadendron giganteum", len.seqgi),rep("Taxus chinensis", len.taxch),
                           rep("Abies alba", len.abial),rep("Picea abies", len.picab),rep("Gnetum montanum", len.gnemo),
                           rep("Welwitschia mirabilis", len.welmi),rep("Cycas panzhihuaensis", len.cyc),
                           rep("Ginkgo biloba", len.ginbi),rep("Amborella trichopoda", len.ambtr),
                           rep("Nymphaea colorata", len.nymco),rep("Cinnamomum kanehirae", len.cinka),
                           rep("Aegilops tauschii", len.aegi),rep("Triticum aestivum", len.ta),
                           rep("Hordeum vulgare", len.horvu),rep("Brachypodium distachyon", len.bradi),
                           rep("Oryza sativa", len.os),rep("Setaria italica", len.setit),rep("Sorghum bicolor", len.sb),
                           rep("Zea mays", len.zea),rep("Panicum hallii", len.panha),rep("Miscanthus sinensis", len.missi),
                           rep("Oropetium thomaeum", len.oroth),rep("Eleusine coracana", len.eleco),
                           rep("Ananas comosus", len.anaco),rep("Musa acuminata", len.musac),rep("Dioscorea alata", len.dioal),
                           rep("Spirodela polyrhiza", len.spipo),rep("Asparagus officinalis", len.aspof),
                           rep("Zostera marina", len.zosma),rep("Arabidopsis thaliana", len.at),
                           rep("Arabidopsis lyrata", len.araly),rep("Capsella grandiflora", len.capgr),
                           rep("Brassica rapa", len.brara),rep("Brassica oleracea", len.braol),
                           rep("Sisymbrium irio", len.sisir),rep("Schrenkiella parvula", len.schpa),
                           rep("Eutrema salsugineum", len.eutsa),rep("Thlaspi arvense", len.thlar),
                           rep("Boechera stricta", len.boest),rep("Carica papaya", len.carpa),rep("Theobroma cacao", len.theca),
                           rep("Gossypium hirsutum", len.goshi),rep("Gossypium raimondii", len.gosra),
                           rep("Citrus sinensis", len.citsi),rep("Poncirus trifoliata", len.pontr),
                           rep("Eucalyptus grandis", len.eucgr),rep("Malus domestica", len.maldo),
                           rep("Prunus persica", len.prupe),rep("Fragaria vesca", len.frave),rep("Corylus avellana", len.corav),
                           rep("Carya illinoinensis", len.caril),rep("Quercus rubra", len.queru),
                           rep("Cucumis sativus", len.cucsa),rep("Glycine max", len.glyma),rep("Medicago truncatula", len.medtr),
                           rep("Trifolium pratense", len.tripr),rep("Cicer arietinum", len.cicar),
                           rep("Lotus japonicus", len.lotja),rep("Phaseolus acutifolius", len.phaac),
                           rep("Vigna unguiculata", len.vigun),rep("Lupinus albus", len.lupal),rep("Lens culinaris", len.lencu),
                           rep("Arachis hypogaea", len.arahy),rep("Populus trichocarpa", len.poptr),
                           rep("Manihot esculenta", len.manes),rep("Salix purpurea", len.salpu),
                           rep("Linum usitatissimum", len.linus),rep("Corymbia citriodora", len.corci),
                           rep("Vitis vinifera", len.vitvi),rep("Kalanchoe fedtschenkoi", len.kalfe),
                           rep("Vaccinium darrowii", len.vacda),rep("Olea europaea", len.oleeu),
                           rep("Mimulus guttatus", len.mimgu),rep("Solanum lycopersicum", len.sl),
                           rep("Solanum tuberosum", len.soltu),rep("Coffea arabica", len.cofar),
                           rep("Lactuca sativa", len.lacsa),rep("Daucus carota", len.dauca),rep("Beta vulgaris", len.betvu),
                           rep("Amaranthus hypochondriacus", len.amahy),rep("Chenopodium quinoa", len.chequ),
                           rep("Saponaria officinalis", len.sapof),rep("Aquilegia coerulea", len.aquco))
     return(organims_reduced)
   }) #%>% bindEvent(input$run_button5)
   
   tree_plot5 <- reactive({
     
     # Define previous variables
     tree_reduced <- tree_reduced5()
     gene.name.tree <- "query_prot"
     tree <- tree_adj5()
     
     tips_to_keep.mp <- tips_to_keep.mp5()
     tips_to_keep.ot <- tips_to_keep.ot5()
     tips_to_keep.at <- tips_to_keep.at5()
     tips_to_keep.cp <- tips_to_keep.cp5()
     tips_to_keep.cr <- tips_to_keep.cr5()
     tips_to_keep.cz <- tips_to_keep.cz5()
     tips_to_keep.kn <- tips_to_keep.kn5()
     tips_to_keep.me <- tips_to_keep.me5()
     tips_to_keep.mi <- tips_to_keep.mi5()
     tips_to_keep.pp <- tips_to_keep.pp5()
     tips_to_keep.sl <- tips_to_keep.sl5()
     tips_to_keep.sm <- tips_to_keep.sm5()
     tips_to_keep.sp <- tips_to_keep.sp5()
     tips_to_keep.ta <- tips_to_keep.ta5()
     tips_to_keep.vc <- tips_to_keep.vc5()
     tips_to_keep.bp <- tips_to_keep.bp5()
     tips_to_keep.cri <- tips_to_keep.cri5()
     tips_to_keep.ds <- tips_to_keep.ds5()
     tips_to_keep.os <- tips_to_keep.os5()
     tips_to_keep.smag <- tips_to_keep.smag5()
     tips_to_keep.tp <- tips_to_keep.tp5()
     tips_to_keep.aa <- tips_to_keep.aa5()
     tips_to_keep.um <- tips_to_keep.um5()
     tips_to_keep.rs <- tips_to_keep.rs5()
     tips_to_keep.cyc <- tips_to_keep.cyc5()
     tips_to_keep.ca <- tips_to_keep.ca5()
     tips_to_keep.mv <- tips_to_keep.mv5()
     tips_to_keep.af <- tips_to_keep.af5()
     tips_to_keep.sc <- tips_to_keep.sc5()
     tips_to_keep.aegi <- tips_to_keep.aegi5()
     tips_to_keep.sb <- tips_to_keep.sb5()
     tips_to_keep.chara <- tips_to_keep.chara5()
     tips_to_keep.sceobli <- tips_to_keep.sceobli5()
     tips_to_keep.cocco <- tips_to_keep.cocco5()
     tips_to_keep.haema <- tips_to_keep.haema5()
     tips_to_keep.zm <- tips_to_keep.zm5()
     tips_to_keep.praco <- tips_to_keep.praco5()
     tips_to_keep.ostlu <- tips_to_keep.ostlu5()
     tips_to_keep.ostme <- tips_to_keep.ostme5()
     tips_to_keep.micco <- tips_to_keep.micco5()
     tips_to_keep.cymte <- tips_to_keep.cymte5()
     tips_to_keep.chlin <- tips_to_keep.chlin5()
     tips_to_keep.gonpe <- tips_to_keep.gonpe5()
     tips_to_keep.tetso <- tips_to_keep.tetso5()
     tips_to_keep.desar <- tips_to_keep.desar5()
     tips_to_keep.monmi <- tips_to_keep.monmi5()
     tips_to_keep.caule <- tips_to_keep.caule5()
     tips_to_keep.ostqu <- tips_to_keep.ostqu5()
     tips_to_keep.brysp <- tips_to_keep.brysp5()
     tips_to_keep.ellbi <- tips_to_keep.ellbi5()
     tips_to_keep.myrbi <- tips_to_keep.myrbi5()
     tips_to_keep.apalo <- tips_to_keep.apalo5()
     tips_to_keep.astgl <- tips_to_keep.astgl5()
     tips_to_keep.tresp <- tips_to_keep.tresp5()
     tips_to_keep.picso <- tips_to_keep.picso5()
     tips_to_keep.chlso <- tips_to_keep.chlso5()
     tips_to_keep.auxpr <- tips_to_keep.auxpr5()
     tips_to_keep.helsp <- tips_to_keep.helsp5()
     tips_to_keep.tetst <- tips_to_keep.tetst5()
     tips_to_keep.pedsp <- tips_to_keep.pedsp5()
     tips_to_keep.chlpr <- tips_to_keep.chlpr5()
     tips_to_keep.picsp <- tips_to_keep.picsp5()
     tips_to_keep.mesvb <- tips_to_keep.mesvb5()
     tips_to_keep.meskr <- tips_to_keep.meskr5()
     tips_to_keep.zygci <- tips_to_keep.zygci5()
     tips_to_keep.zygcb <- tips_to_keep.zygcb5()
     tips_to_keep.zygcy <- tips_to_keep.zygcy5()
     tips_to_keep.plesc <- tips_to_keep.plesc5()
     tips_to_keep.funhy <- tips_to_keep.funhy5()
     tips_to_keep.sphfa <- tips_to_keep.sphfa5()
     tips_to_keep.takle <- tips_to_keep.takle5()
     tips_to_keep.marpa <- tips_to_keep.marpa5()
     tips_to_keep.ricfl <- tips_to_keep.ricfl5()
     tips_to_keep.ricso <- tips_to_keep.ricso5()
     tips_to_keep.antpu <- tips_to_keep.antpu5()
     tips_to_keep.antfu <- tips_to_keep.antfu5()
     tips_to_keep.megfl <- tips_to_keep.megfl5()
     tips_to_keep.phach <- tips_to_keep.phach5()
     tips_to_keep.phyph <- tips_to_keep.phyph5()
     tips_to_keep.parpe <- tips_to_keep.parpe5()
     tips_to_keep.notor <- tips_to_keep.notor5()
     tips_to_keep.phaca <- tips_to_keep.phaca5()
     tips_to_keep.phasp <- tips_to_keep.phasp5()
     tips_to_keep.leidu <- tips_to_keep.leidu5()
     tips_to_keep.isosi <- tips_to_keep.isosi5()
     tips_to_keep.dipco <- tips_to_keep.dipco5()
     tips_to_keep.hupas <- tips_to_keep.hupas5()
     tips_to_keep.lyccl <- tips_to_keep.lyccl5()
     tips_to_keep.adica <- tips_to_keep.adica5()
     tips_to_keep.alssp <- tips_to_keep.alssp5()
     tips_to_keep.marve <- tips_to_keep.marve5()
     tips_to_keep.seqse <- tips_to_keep.seqse5()
     tips_to_keep.seqgi <- tips_to_keep.seqgi5()
     tips_to_keep.taxch <- tips_to_keep.taxch5()
     tips_to_keep.abial <- tips_to_keep.abial5()
     tips_to_keep.picab <- tips_to_keep.picab5()
     tips_to_keep.gnemo <- tips_to_keep.gnemo5()
     tips_to_keep.welmi <- tips_to_keep.welmi5()
     tips_to_keep.ginbi <- tips_to_keep.ginbi5()
     tips_to_keep.ambtr <- tips_to_keep.ambtr5()
     tips_to_keep.nymco <- tips_to_keep.nymco5()
     tips_to_keep.cinka <- tips_to_keep.cinka5()
     tips_to_keep.horvu <- tips_to_keep.horvu5()
     tips_to_keep.bradi <- tips_to_keep.bradi5()
     tips_to_keep.setit <- tips_to_keep.setit5()
     tips_to_keep.panha <- tips_to_keep.panha5()
     tips_to_keep.missi <- tips_to_keep.missi5()
     tips_to_keep.oroth <- tips_to_keep.oroth5()
     tips_to_keep.eleco <- tips_to_keep.eleco5()
     tips_to_keep.anaco <- tips_to_keep.anaco5()
     tips_to_keep.musac <- tips_to_keep.musac5()
     tips_to_keep.dioal <- tips_to_keep.dioal5()
     tips_to_keep.spipo <- tips_to_keep.spipo5()
     tips_to_keep.aspof <- tips_to_keep.aspof5()
     tips_to_keep.zosma <- tips_to_keep.zosma5()
     tips_to_keep.araly <- tips_to_keep.araly5()
     tips_to_keep.capgr <- tips_to_keep.capgr5()
     tips_to_keep.brara <- tips_to_keep.brara5()
     tips_to_keep.braol <- tips_to_keep.braol5()
     tips_to_keep.sisir <- tips_to_keep.sisir5()
     tips_to_keep.schpa <- tips_to_keep.schpa5()
     tips_to_keep.eutsa <- tips_to_keep.eutsa5()
     tips_to_keep.thlar <- tips_to_keep.thlar5()
     tips_to_keep.boest <- tips_to_keep.boest5()
     tips_to_keep.carpa <- tips_to_keep.carpa5()
     tips_to_keep.theca <- tips_to_keep.theca5()
     tips_to_keep.goshi <- tips_to_keep.goshi5()
     tips_to_keep.gosra <- tips_to_keep.gosra5()
     tips_to_keep.citsi <- tips_to_keep.citsi5()
     tips_to_keep.pontr <- tips_to_keep.pontr5()
     tips_to_keep.eucgr <- tips_to_keep.eucgr5()
     tips_to_keep.maldo <- tips_to_keep.maldo5()
     tips_to_keep.prupe <- tips_to_keep.prupe5()
     tips_to_keep.frave <- tips_to_keep.frave5()
     tips_to_keep.corav <- tips_to_keep.corav5()
     tips_to_keep.caril <- tips_to_keep.caril5()
     tips_to_keep.queru <- tips_to_keep.queru5()
     tips_to_keep.cucsa <- tips_to_keep.cucsa5()
     tips_to_keep.glyma <- tips_to_keep.glyma5()
     tips_to_keep.medtr <- tips_to_keep.medtr5()
     tips_to_keep.tripr <- tips_to_keep.tripr5()
     tips_to_keep.cicar <- tips_to_keep.cicar5()
     tips_to_keep.lotja <- tips_to_keep.lotja5()
     tips_to_keep.phaac <- tips_to_keep.phaac5()
     tips_to_keep.vigun <- tips_to_keep.vigun5()
     tips_to_keep.lupal <- tips_to_keep.lupal5()
     tips_to_keep.lencu <- tips_to_keep.lencu5()
     tips_to_keep.arahy <- tips_to_keep.arahy5()
     tips_to_keep.poptr <- tips_to_keep.poptr5()
     tips_to_keep.manes <- tips_to_keep.manes5()
     tips_to_keep.salpu <- tips_to_keep.salpu5()
     tips_to_keep.linus <- tips_to_keep.linus5()
     tips_to_keep.corci <- tips_to_keep.corci5()
     tips_to_keep.vitvi <- tips_to_keep.vitvi5()
     tips_to_keep.kalfe <- tips_to_keep.kalfe5()
     tips_to_keep.vacda <- tips_to_keep.vacda5()
     tips_to_keep.oleeu <- tips_to_keep.oleeu5()
     tips_to_keep.mimgu <- tips_to_keep.mimgu5()
     tips_to_keep.soltu <- tips_to_keep.soltu5()
     tips_to_keep.cofar <- tips_to_keep.cofar5()
     tips_to_keep.lacsa <- tips_to_keep.lacsa5()
     tips_to_keep.dauca <- tips_to_keep.dauca5()
     tips_to_keep.betvu <- tips_to_keep.betvu5()
     tips_to_keep.amahy <- tips_to_keep.amahy5()
     tips_to_keep.chequ <- tips_to_keep.chequ5()
     tips_to_keep.sapof <- tips_to_keep.sapof5()
     tips_to_keep.aquco <- tips_to_keep.aquco5()
     
     if (length(tree_reduced$tip.label) < 2)
     {
       cat("")
     }
     else 
     {
       # Highlight the target gene
       high.gene5 <<- tree_reduced$tip.label[grep(pattern = gene.name.tree, tree_reduced$tip.label)]
       
       
       # Color asignment per species
       col.factor <- c()
       org.factor <- c()
       
       library(glue)
       library(ggtree)
       library(ggplot2)
       
       for (i in 1:length(tree_reduced$tip.label))
       {
         if (tree_reduced$tip.label[i] %in% high.gene5)
         {
           col.factor <- c(col.factor,"#CD0000")
           org.factor <- c(org.factor,"Query sequence")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
         {
           col.factor <- c(col.factor,"#006400")
           org.factor <- c(org.factor,"Marchantia")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
         {
           col.factor <- c(col.factor,"#00008B")
           org.factor <- c(org.factor,"Ostreococcus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
         {
           col.factor <- c(col.factor,"#CD661D")
           org.factor <- c(org.factor,"Arabidopsis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
         {
           col.factor <- c(col.factor,"#458B74")
           org.factor <- c(org.factor,"Ceratodon")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
         {
           col.factor <- c(col.factor,"#8B7355")
           org.factor <- c(org.factor,"Chlamydomonas")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
         {
           col.factor <- c(col.factor,"#458B00")
           org.factor <- c(org.factor,"Chromochloris")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
         {
           col.factor <- c(col.factor,"#CD1076")
           org.factor <- c(org.factor,"Klebsormidium")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
         {
           col.factor <- c(col.factor,"#8B8878")
           org.factor <- c(org.factor,"Mesotaenium")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
         {
           col.factor <- c(col.factor,"#666666")
           org.factor <- c(org.factor,"Micromonas")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
         {
           col.factor <- c(col.factor,"#B8860B")
           org.factor <- c(org.factor,"Physcomitrium")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
         {
           col.factor <- c(col.factor,"#8B008B")
           org.factor <- c(org.factor,"Solanum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
         {
           col.factor <- c(col.factor,"#6E8B3D")
           org.factor <- c(org.factor,"Selaginella")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
         {
           col.factor <- c(col.factor,"#79CDCD")
           org.factor <- c(org.factor,"Spirogloea")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
         {
           col.factor <- c(col.factor,"#CDCD00")
           org.factor <- c(org.factor,"Triticum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
         {
           col.factor <- c(col.factor,"#16317d")
           org.factor <- c(org.factor,"Volvox")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
         {
           col.factor <- c(col.factor,"#007e2f")
           org.factor <- c(org.factor,"Bathycoccus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
         {
           col.factor <- c(col.factor,"#ffcd12")
           org.factor <- c(org.factor,"Ceratopteris")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
         {
           col.factor <- c(col.factor,"#b86092")
           org.factor <- c(org.factor,"Dunaliella")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
         {
           col.factor <- c(col.factor,"#721b3e")
           org.factor <- c(org.factor,"Oryza")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
         {
           col.factor <- c(col.factor,"#00b7a7")
           org.factor <- c(org.factor,"Sphagnum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
         {
           col.factor <- c(col.factor,"#67000d")
           org.factor <- c(org.factor,"Thuja")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
         {
           col.factor <- c(col.factor,"#5b2c6f")
           org.factor <- c(org.factor,"Anthoceros")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
         {
           col.factor <- c(col.factor,"#15e71b")
           org.factor <- c(org.factor,"Ulva")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
         {
           col.factor <- c(col.factor,"#e67e22")
           org.factor <- c(org.factor,"Raphidocelis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
         {
           col.factor <- c(col.factor,"#873600")
           org.factor <- c(org.factor,"Cycas")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pu])
         {
           col.factor <- c(col.factor,"#dc1c0f")
           org.factor <- c(org.factor,"Porphyra")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pt])
         {
           col.factor <- c(col.factor,"#a04000")
           org.factor <- c(org.factor,"Phaeodactylum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ng])
         {
           col.factor <- c(col.factor,"#935116")
           org.factor <- c(org.factor,"Nannochloropsis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyano])
         {
           col.factor <- c(col.factor,"#2874a6")
           org.factor <- c(org.factor,"Cyanophora")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
         {
           col.factor <- c(col.factor,"#0b5345")
           org.factor <- c(org.factor,"Chlorokybus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
         {
           col.factor <- c(col.factor,"#283747")
           org.factor <- c(org.factor,"Mesostigma")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
         {
           col.factor <- c(col.factor,"#145a32")
           org.factor <- c(org.factor,"Azolla")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
         {
           col.factor <- c(col.factor,"#3339e6")
           org.factor <- c(org.factor,"Salvinia")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
         {
           col.factor <- c(col.factor,"#e6338f")
           org.factor <- c(org.factor,"Aegilops")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
         {
           col.factor <- c(col.factor,"#cd016a")
           org.factor <- c(org.factor,"Sorghum")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
         {
           col.factor <- c(col.factor,"#117a65")
           org.factor <- c(org.factor,"Chara")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.guilla])
         {
           col.factor <- c(col.factor,"#424949")
           org.factor <- c(org.factor,"Guillardia")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.crypto])
         {
           col.factor <- c(col.factor,"#515a5a")
           org.factor <- c(org.factor,"Cryptophyceae")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymero])
         {
           col.factor <- c(col.factor,"#641e16")
           org.factor <- c(org.factor,"Cyanidioschyzon")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.galsul])
         {
           col.factor <- c(col.factor,"#633974")
           org.factor <- c(org.factor,"Galdieria")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gracichor])
         {
           col.factor <- c(col.factor,"#a93226")
           org.factor <- c(org.factor,"Gracilariopsis")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
         {
           col.factor <- c(col.factor,"#148f77")
           org.factor <- c(org.factor,"Scenedesmus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
         {
           col.factor <- c(col.factor,"#9c640c")
           org.factor <- c(org.factor,"Coccomyxa")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.saccha])
         {
           col.factor <- c(col.factor,"#6e2c00")
           org.factor <- c(org.factor,"Saccharina")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
         {
           col.factor <- c(col.factor,"#196f3d")
           org.factor <- c(org.factor,"Haematococcus")
         }
         else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
         {
           col.factor <- c(col.factor,"#666909")
           org.factor <- c(org.factor,"Zea")
         }
         
       }
       
       
       
       #Matrix with labels and colors and transform to dplyr format
       data.tree <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                               col = col.factor, org = org.factor)
       
       d2 <- dplyr::mutate(data.tree, lab = data.tree$label,
                           color = data.tree$col,
                           organism = data.tree$org,
                           name = glue("<i style='color:{color}'> {lab} </i>"))
       
       tree_plot <- ggtree(tree_reduced) %<+% d2 + geom_tiplab() + theme(legend.position =) +
             xlim(0, max(tree_reduced$edge.length)*3) + geom_tiplab(aes(label = label, color = organism)) +
             scale_color_manual(values = unique(d2$col), breaks = unique(d2$org)) +
             geom_highlight(mapping=aes(subset = label %in% high.gene5,
                                        node = node,
                                        fill = as.factor(node)), extend = 0.8) + 
             labs(fill = "Node of interest")
           
       #shinyjs::hideElement(id = 'loading.tree5')
       return(tree_plot)
     }}) #%>% bindEvent(input$run_button5)
   
   
   # Identify OG from SHOOT output
   og.name5 <- reactive({
     
     tree_reduced <- tree_reduced5()
     random_name <- random.file5()
     random_system <- paste0(random_name, "_system")

     # Path1: read .sh.ogs.txt.gz file for OG
     og.name.file <- read.table(paste0("pharaoh_folder/", random_system,".fa.assign.txt"), header = F, sep="\t", skip = 1)
     og.name <- paste0("OG", sprintf("%07d", og.name.file$V2))
     return(og.name)

   }) %>% bindEvent(input$run_button5)
   
   # Create sequences file for extended OG
   ortho_seq5 <- reactive({
     
     file.name <- og.name5()
     shoot_sequence <- shoot_sequence5()
     random_name <- random.file5()
     
     library(seqinr)
     
     # Load orthogroup sequences file
     ortho.seq.name <- paste("Orthogroup_Sequences",paste(file.name, "fa", sep = "."), sep="/")
     
     # Add query protein sequence
     ortho_seq <- read.fasta(ortho.seq.name, seqtype = "AA")
     ortho_seq_names <- c("query_prot", getName(ortho_seq))
     ortho_seq_seqs <- c(list(shoot_sequence), getSequence(ortho_seq))
     write.fasta(ortho_seq_seqs, names =  ortho_seq_names, file.out = paste0(random_name, "_tmp.fa"))
     ortho_seq_ext <- read.fasta(paste0(random_name, "_tmp.fa"), seqtype = "AA")
     file.remove(paste0(random_name, "_tmp.fa"))
     
     return(ortho_seq_ext)
   })
   
   ### Select orthogroup sequences based on the reduced tree
   ortho_reduced5 <- reactive({
     
     tree_reduced <- tree_reduced5()
     ortho_seq <- ortho_seq5()
     ortho_reduced <- ortho_seq[tree_reduced$tip.label]
     return(ortho_reduced)
   }) %>% bindEvent(input$run_button5)
   
   # Outputs
   observeEvent(isTruthy(ortho_reduced5()), {
     output$error_tree5 <- NULL
   })
   
   # Create boxes
   observeEvent(isTruthy(ortho_reduced5()), {
     
     if (UI_exist_tree5)
     {
       removeUI(
         selector = "div:has(>> #treeTips5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #presentorg5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadTree5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadNewick5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadTreeSeqs5)",
         multiple = TRUE,
         immediate = TRUE
       )
     }
     
     
     insertUI("#box_tree_text5", "afterEnd", ui = {
       box(
         title = "Genes in Orthogroup", status = "primary", solidHeader = TRUE, width = 12,
         collapsible = TRUE,
         verbatimTextOutput("treeTips5")
       )
     }) 
     
     insertUI("#box_tree_pie5", "afterEnd", ui = {
       box(
         title = "Present Organisms", status = "primary", solidHeader = TRUE,
         collapsible = TRUE, width = 12,
         plotlyOutput("presentorg5")
       )
     }) 
     
     insertUI("#box_tree_plot5", "afterEnd", ui = {
       image_height <- 300 + 15*length(tree_reduced5()$tip.label)
       box(width = 12,
           title = "Gene Tree", status = "primary", solidHeader = TRUE,
           collapsible = TRUE, 
           plotOutput("tree_image5", height = image_height, width = 1100)
       )
     })
     
     insertUI("#download_tree5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTree5", 
                                                                          "Download Tree Plot",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#download_newick5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadNewick5", 
                                                                          "Download NEWICK Tree",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#download_tree_seqs5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadTreeSeqs5", 
                                                                          "Download Protein Sequences",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_tree5 <<- TRUE
     shinyjs::hideElement(id = 'loading.tree5')
   })
   
   # Fill boxes with output
   output$treeTips5 <- renderPrint({
     print(tree_reduced5()$tip.label)
   }, width = 400) # %>% bindEvent(input$run_button5)
   
   
   # Render pie chart
   output$presentorg5 <- renderPlotly({
     
     {library(ggplot2)
       library(dplyr)
       
       data <- data.frame(table(organims_reduced5()))
       colnames(data) <- c("group", "value")
       
       # Compute the position of labels
       data <- data %>%
         arrange(desc(group)) %>%
         mutate(prop = value / sum(data$value) *100) %>%
         mutate(ypos = cumsum(prop)- 0.5*prop )
       
       # Create plot
       
       plotly::plot_ly(data=data,values=~prop,labels=~factor(group),
                       marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                              floor(nrow(data)/9)+1)),
                       type="pie",showlegend = F, text= ~group,
                       textinfo = "none", hoverinfo = "text")} 
     
   })
   
   # Render tree image
   output$tree_image5 <- renderImage({
     image_height <- 300 + 15*length(tree_reduced5()$tip.label)
     png("tree5.png", height = image_height, width = 1100)
     plot(tree_plot5())
     dev.off()
     
     list(src = "tree5.png",
          contentType="image/png", width=1100,height=image_height)
   }, deleteFile = T)
   
   # Download results
   output$downloadTree5 <- downloadHandler(
     filename= function() {
       paste("tree", ".png", sep="")
     },
     content= function(file) {
       image_height <- (300 + 11*length(tree_reduced5()$tip.label))*3
       image_width <- (200 + 400*max(tree_reduced5()$edge.length))*3
       png(file, height = image_height, width = image_width, res = (70 + 0.1*length(tree_reduced5()$tip.label))*3)
       plot(tree_plot5())
       dev.off()
     })
   
   
   # Create and download tree in newick format
   output$downloadNewick5 <- downloadHandler(
     filename= function() {
       paste("tree_newick", ".txt", sep="")
     },
     content= function(file) {
       write.tree(tree_reduced5(), file)
     })
   
   #  # Create and download sequences for genes in tree
   output$downloadTreeSeqs5 <- downloadHandler(
     filename= function() {
       paste("tree_seqs", ".fa", sep="")
     },
     content= function(file) {
       seqinr::write.fasta(sequences = seqinr::getSequence(ortho_reduced5()),
                           names = seqinr::getName(ortho_reduced5()), file.out = file)
     })
   
   ####################### PHYLOWIDGET ############################
   # Remove previous outputs when updated by a new search
   observeEvent(input$run_button5, {
     if (UI_exist_phylo5)
     {
       removeUI(
         selector = "div:has(>>> #phylo_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_phylo5 <<- F
     }
     
   })
   
   
   phylo_tree5 <- reactive({
     
     library(ape)
     tree_phylo <- tree_reduced5()
     
     # Normalize tree depth
     root_id <- length(tree_phylo$tip.label)+1
     norm_factor <- max(dist.nodes(tree_phylo)[root_id,])
     tree_phylo$edge.length <- tree_phylo$edge.length/norm_factor
     
     return(tree_phylo)
     
   }) %>% bindEvent(input$phylo_start5)
   
   observeEvent(isTruthy(phylo_tree5()),{
     
     if(UI_exist_phylo5)
     {
       removeUI(
         selector = "div:has(>>> #phylo_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_phylo5 <<- F
     }
     
     insertUI("#box_phylo5", "afterEnd", ui = {
       phylo_tree <- phylo_tree5()
       phylo_height <- length(phylo_tree$tip.label) *14 + 220
       box(width = 12,
           title = "Interactive Tree", status = "primary", solidHeader = TRUE, height = phylo_height + 100,
           collapsible = TRUE,
           tags$div(id = "phylo_pocket5", style = paste0("width: 1300px; height: ",  phylo_height + 50, "px"),
                    phylowidgetOutput("phylo_plot5", height = paste0(phylo_height,"px"), width = "98%"))
       )
     })
     
     
     UI_exist_phylo5 <<- T
     
   })
   
   output$phylo_plot5 <- renderPhylowidget({
     
     phylo_tree <- phylo_tree5()
     phylowidget(phylo_tree)
   })
   
   
   
   #########################  PFAM  ###############################
   
   observeEvent(input$run_button5, {
     removeUI(
       selector = "div:has(>> #selected_pfamsI5)",
       multiple = TRUE,
       immediate = TRUE
     )
   })
   
   observeEvent(input$pfam_start5, {
     insertUI("#selected_pfams5", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_pfamsI5","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced5()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({"query_prot"}))
     })
   })
   
   observeEvent(input$run_button5, {
     removeUI("#pfam_selection5")
   })
   
   observeEvent(input$pfam_start5, {
     insertUI("#pfam_selectionI5", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("pfam_selection5", "Show Pfam Domains", size = "sm",
                                style = "float", color = "royal")
     })
   })
   
   
   observeEvent(input$run_button5, {
     if (UI_exist_pfam5)
     {
       removeUI(
         selector = "div:has(>> #output_pfam_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pfam5 <<- F
     }
   })
   
   
   total_table_pfam5 <- reactive({
     shinyjs::showElement(id = 'loading.pfam.pf5')
     ortho_reduced <- ortho_reduced5()
     sel_genes <- as.vector(input$selected_pfamsI5)
     
     if (length(sel_genes) < 1)
     {
       shinyjs::hideElement(id = 'loading.pfam.pf5')
       removeUI(
         selector = "div:has(>> #output_pfam_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       UI_exist_pfam5 <<- F
       output$error_pfam5 <- renderUI({renderText({print("Please select at least one gene.")})})
       validate(need(length(sel_genes) > 0, "Please select at least one gene."))
     }
     
     output$error_pfam5 <- NULL
     #library(bio3d)
     library(RCurl)
     library(drawProteins)
     library(ggplot2)
     library(jsonlite)
     
     # Get the sequences as a vector of strings
     
     
     # Create data frame with proper columns
     total_table_pfam <- data.frame(type=NA,
                                    description=NA,
                                    begin=NA, end=NA,
                                    length=NA,
                                    accession=NA, entryName=NA,
                                    taxid=NA, order=NA)
     
     
     # Fill data frame with the information about domains obtained with hmmer
     for (i in 1:length(sel_genes))
     {
       ortho_comp <- ortho_reduced[[sel_genes[i]]]
       ortho_str <- seqinr::getSequence(ortho_comp, as.string = T)
       ortho_cha <- unlist(ortho_str)
       
       
       # Curl POST request
       url <- "https://www.ebi.ac.uk/Tools/hmmer/api/v1/search/hmmscan"
       
       data <- list(
         database = "pfam",
         input = ortho_cha
       )
       
       json_data <- toJSON(data, auto_unbox = T)
       
       response <- postForm(url, 
                            .opts = list(
                              postfields = json_data,
                              httpheader = c(
                                "Content-Type" = "application/json",
                                "Accept" = "application/json"
                              )
                            ))
       
       
       new_id <- strsplit(as.character(response), '"')[[1]][4]
       
       
       # Results URL
       url_res <- paste0("https://www.ebi.ac.uk/Tools/hmmer/api/v1/result/", new_id)
       
       # Iterate until job is finished, for that, check if SUCCESS is present in GET result
       nap_again <- T
       while(nap_again)
       {
         get_response <- getURL(url_res,
                                httpheader = c(
                                  "Accept" = "application/json"
                                ))
         
         if (grepl("SUCCESS", get_response))
         {
           nap_again <- F
           
           # Once computation is over, parse JSON response
           json_resp <- jsonlite::fromJSON(url_res, simplifyDataFrame = TRUE)
           json_hits <- data.frame(json_resp$result$hits)
           
           hits_acc <- json_hits$acc
           hits_desc <- json_hits$metadata$description
           hits_domains <- json_hits$domains
           
           df_acc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_acc, y = hits_domains, USE.NAMES = F))
           df_desc <- unlist(mapply({function(x,y) rep(x,length(y$iali))}, x = hits_desc, y = hits_domains, USE.NAMES = F))
           df_begin <- unlist(sapply(hits_domains, function(x) x$iali))
           df_end <- unlist(sapply(hits_domains, function(x) x$jali))
           
           # Create PFAM table
           pfam_table <- data.frame(type=c("CHAIN", rep("DOMAIN", length(df_acc))),
                                    description=c("Protein chain",df_acc),
                                    begin=c(1, df_begin), end=c(nchar(ortho_cha),df_end),
                                    length=c(nchar(ortho_cha)-1, df_end - df_begin),
                                    accession=sel_genes[i], entryName=sel_genes[i],
                                    taxid=c("Chain", df_desc), order=i)
           
           total_table_pfam <- rbind(total_table_pfam, pfam_table)
           
         }
         else if (grepl("PENDING", get_response))
         {
           Sys.sleep(10)
         }
         # If there is a problem with server
         else
         {
           nap_again <- F
           pfam_table <- data.frame(type="CHAIN",
                                    description="Protein chain",
                                    begin=1, end=nchar(ortho_cha),
                                    length=nchar(ortho_cha)-1,
                                    accession=sel_genes[i], entryName=sel_genes[i],
                                    taxid="Chain", order=i)
           total_table_pfam <- rbind(total_table_pfam, pfam_table)
         }
       }
     }
     
     total_table_pfam <- total_table_pfam[-1,]
     total_table_pfam <- total_table_pfam[!duplicated(total_table_pfam),]
     # Remove protein chain results
     total_table_pfam <- subset(total_table_pfam, !(type=="DOMAIN" & description=="Protein chain"))
     
     return(total_table_pfam)
     
   }) %>% bindEvent(input$pfam_selection5)
   
   pfplot5 <- reactive({
     
     total_table_pfam <- total_table_pfam5()
     # Now we can plot domains information as chains
     pfplot <- draw_canvas(total_table_pfam)
     pfplot <- draw_chains(pfplot, total_table_pfam)
     pfplot <- draw_domains(pfplot, total_table_pfam, label_domains = F)
     pfplot <- pfplot + theme_bw(base_size = 20) + # white background
       theme(panel.grid.minor=element_blank(),
             panel.grid.major=element_blank()) +
       theme(axis.ticks = element_blank(),
             axis.text.y = element_blank()) +
       theme(panel.border = element_blank())
     #pfplot <- pfplot + labs(title = "Pfam domains")
     pfplot <- pfplot + theme(legend.position="top") + labs(fill="")
     
   }) %>% bindEvent(input$pfam_selection5)
   
   
   # Outputs
   
   observeEvent(isTruthy(pfplot5()), {
     
     if (UI_exist_pfam5)
     {
       removeUI(
         selector = "div:has(>> #output_pfam_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #pfam_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadPFAMTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_pfam5", "afterEnd", ui = {
       
       box(
         title = "PFAM Table", status = "primary", solidHeader = TRUE, width = 12,
         collapsible = TRUE,
         dataTableOutput(outputId = "output_pfam_table5"))
     }) 
     
     insertUI("#box_pfplot5", "afterEnd", ui = {
       total_table_pfam <- total_table_pfam5()
       box_pfplot_height <- 150 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
       box(
         title = "Domains Localization", status = "primary", solidHeader = TRUE, width = 12, #height = box_pfplot_height,
         collapsible = TRUE,
         plotOutput("pfam_plot5", height = box_pfplot_height))
     }) 
     
     insertUI("#pfam_down_button5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "pfam_download5", "Download PFAM figure",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#download_ui_for_pfam_table5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadPFAMTable5", "Download PFAM Table",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_pfam5 <<- TRUE
     shinyjs::hideElement(id = 'loading.pfam.pf5')
   })
   
   output$output_pfam_table5 <- renderDataTable({
     total_table_pfam <- total_table_pfam5()
     out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
     out_pf_table$description <- sapply(out_pf_table$description, function(x) pfam.link(x))
     colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
     return(out_pf_table)
   }, escape=FALSE, options = list(pageLength = 5))
   
   output$pfam_plot5 <- renderImage({
     total_table_pfam <- total_table_pfam5()
     pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
     pfam_width <- 1000
     pfplot <- pfplot5()
     png("pharaoh_folder/pfam.png",  width = pfam_width, height = pfam_height)
     plot(pfplot)
     dev.off()
     list(src = "pharaoh_folder/pfam.png",
          contentType="image/png")
   }, deleteFile = T
   )
   
   # Download results
   
   output$pfam_download5 <- downloadHandler(
     filename= function() {
       paste("pfam", ".png", sep="")
     },
     content= function(file) {
       total_table_pfam <- total_table_pfam5()
       pfam_height <- 50 + 700*length(total_table_pfam$order[nrow(total_table_pfam)])
       pfam_width <- 1150
       pfplot <- pfplot5()
       
       png(file, height = pfam_height, width = pfam_width)
       plot(pfplot)
       dev.off()
     })
   
   output$downloadPFAMTable5<- downloadHandler(
     filename= function() {
       paste("pfam_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_pfam <- total_table_pfam5()
       out_pf_table <- subset(total_table_pfam[,c(1:6,8)], total_table_pfam$type != "CHAIN")
       colnames(out_pf_table) <- c(colnames(total_table_pfam)[1:6],"biological description")
       write.table(x = out_pf_table, quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   
   ####################### CAFE #################################
   
   # Remove previous outputs when updated by a new search
   observeEvent(input$run_button5, {
     if (UI_exist_cafe5)
     {
       removeUI(
         selector = "div:has(>> #cafe_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_mrca5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCAFEPlot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_cafe5 <<- F
     }
     
     if (UI_exist_error_cafe5)
     {
       removeUI(
         selector = "div:has(>> #cafe_error_message5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_error_cafe5 <<- F
     }
   })
   
   ### CAFE parser and tree generator
   cafe_table5 <- reactive({
     
     shinyjs::showElement(id = 'loading.cafe5')
     
     # Import OG name
     og.cafe <- og.name5()
     
     # Define path to COUNT file
     cafe_comp_tree_file <- "pharaoh_folder/family_table_count.tsv"
     
     # Extract COUNT vector for the current OG
     library(data.table)
     library(ape)
     genecount <- fread(cafe_comp_tree_file, sep="\t", header = T)
     genecountdf <- as.data.frame(genecount)
     rownames(genecountdf) <- genecountdf$Orthogroup
     new_cols <- gsub(" ", "_", colnames(genecountdf))
     colnames(genecountdf) <- new_cols
     
     cafe.vector <- genecountdf[og.cafe,]
     
     if (sum(is.na(cafe.vector)) != 0)
     {
       shinyjs::hideElement(id = 'loading.cafe5')
       if (UI_exist_cafe5)
       {
         removeUI(
           selector = "div:has(>> #cafe_plot5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #cafe_mrca5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #cafe_download5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadCAFEPlot5)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       
       UI_exist_cafe5 <<- F
       
       if(UI_exist_error_cafe5)
       {
         removeUI(
           selector = "div:has(>> #cafe_error_message5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
       }
       insertUI("#error_cafe5", "afterEnd", ui = {
         box(width = 12,
             title = "Ancestral State Reconstruction", status = "info", solidHeader = TRUE,
             collapsible = TRUE,
             textOutput("cafe_error_message5"))
       })
       
       output$cafe_error_message5 <- renderText({print("No expansions/contraction detected for this orthogroup,
                                                        or infeasible computation due to large size and variance across
                                                        species.")})
       UI_exist_error_cafe5 <<- T
       
       validate(need(sum(is.na(cafe.vector)) == 0 , ""))
     }
     
     return(genecountdf)
   }) %>% bindEvent(input$cafe_start5)
   
   mrca.tree5 <- reactive({
     
     og.cafe <- og.name5()
     genecountdf <- cafe_table5()
     
     # Create phylogenomic tree with internal nodes names
     mrca.tree <- read.tree("pharaoh_folder/tree_with_internal_nodes.txt")
     
     return(mrca.tree)
     
   }) %>% bindEvent(input$cafe_start5)
   
   cafe_tree5 <- reactive({
     og.cafe <- og.name5()
     genecountdf <- cafe_table5()
     cafe.tree <- mrca.tree5()
     
     # Add counts to nodes
     tip.count <- as.numeric(genecountdf[og.cafe,2:172])
     names(tip.count) <- colnames(genecountdf)[2:172]
     
     node.count <- as.numeric(genecountdf[og.cafe,173:342])
     names(node.count) <- colnames(genecountdf)[173:342]
     
     cafe.tree$tip.label[match(names(tip.count), cafe.tree$tip.label)] <- tip.count
     cafe.tree$node.label[match(names(node.count), cafe.tree$node.label)] <- node.count
     
     return(cafe.tree)
     
   }) %>% bindEvent(input$cafe_start5)
   
   evo_plot_data5 <- reactive({
     
     og.cafe <- og.name5()
     genecountdf <- cafe_table5()
     cafe.tree <- mrca.tree5()
     
     # Add counts to nodes
     tip.count <- as.numeric(genecountdf[og.cafe,2:172])
     names(tip.count) <- colnames(genecountdf)[2:172]
     
     node.count <- as.numeric(genecountdf[og.cafe,173:342])
     names(node.count) <- colnames(genecountdf)[173:342]
     
     # Identify parental node to determine if a general expansion or contraction 
     # occured at the level of each node
     edge_table <- as.data.frame(cafe.tree$edge)
     rownames(edge_table) <- paste("edge", 1:nrow(edge_table), sep = "")
     colnames(edge_table) <- c("parent", "child")
     
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) <
          as.numeric(tip.count[x])) "Expansion"
       else if (as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) >
                as.numeric(tip.count[x])) "Contraction"
       else "No change"
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x) # 170 is root, no parent
       if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) <
          as.numeric(node.count[x])) "Expansion"
       else if(as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) >
               as.numeric(node.count[x])) "Contraction"
       else "No change"
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # A gain occurs when a node experiences an expansion and its parental exhibits a value of 0
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if (exp_cont_tip[x] == "Expansion" &
           as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match(x, edge_table$child)]-Ntip(cafe.tree)]]) == 0
       ) "Gain"
       else exp_cont_tip[x]
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x)
       if (exp_cont_node[match(x,names(exp_cont_node))] == "Expansion" &
           as.numeric(node.count[cafe.tree$node.label[edge_table$parent[match((match(x, cafe.tree$node.label) + Ntip(cafe.tree)), edge_table$child)]-Ntip(cafe.tree)]]) == 0
       ) "Gain"
       else exp_cont_node[x]
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # A loss occurs when a node experiences a contraction and takes the value 0
     exp_cont_tip <- sapply(1:length(tip.count), function(x)
       if (exp_cont_tip[x] == "Contraction" &
           tip.count[x] == 0
       ) "Loss"
       else exp_cont_tip[x]
     )
     names(exp_cont_tip) <- names(tip.count)
     
     exp_cont_node <- sapply(names(node.count)[-170], function(x)
       if (exp_cont_node[match(x,names(exp_cont_node))] == "Contraction" &
           node.count[x] == 0
       ) "Loss"
       else exp_cont_node[x]
     )
     names(exp_cont_node) <- names(node.count)[-170]
     
     # Create complete sorted vectors
     exp_cont_node["root"] <- "No change"
     node_ordered <- exp_cont_node[order(match(names(exp_cont_node),cafe.tree$node.label))]
     tip_ordered <- exp_cont_tip[order(match(names(exp_cont_tip),cafe.tree$tip.label))]
     
     change_vector <- c(tip_ordered, node_ordered)
     
     # Merge tips and nodes reconstruction
     node_ordered_values <- node.count[order(match(names(node.count),cafe.tree$node.label))]
     tip_ordered_values <- tip.count[order(match(names(tip.count),cafe.tree$tip.label))]
     cafe.count <- c(tip_ordered_values, node_ordered_values)
     
     # Create a timeline for a given OG
     tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
     tree.ancestor <- read.tree(tree.name)
     tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
     tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
     tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
     
     mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
     evo.paths <- c()
     for (i in 1:length(unique(tips.orgs)))
     {
       evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
     }
     
     evo.paths <- unique(evo.paths)
     evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
     
     # Associate gray and 0 to reconstruction for nodes not in allowed paths
     change_vector[setdiff(1:length(change_vector), evo.paths)] <- 
       sapply(change_vector[setdiff(1:length(change_vector), evo.paths)], function(x) if(x != "Loss") "OG not present" 
              else x)  
     
     change_vector <- unlist(change_vector)
     
     cafe.count[setdiff(1:length(change_vector), evo.paths)] <- 0
     
     unique(change_vector)
     color_cafe <- sapply(change_vector, function(x) if (x == "No change") "black" 
                          else if (x == "Expansion") "red" else if (x == "Contraction") "blue"
                          else if (x == "Gain") "green3" else if (x == "Loss") "purple"
                          else "gray", USE.NAMES = F)
     
     # Create tree representation
     cafe.table.tips <- data.frame(node = 1:length(cafe.tree$tip.label), label = cafe.tree$tip.label,
                                   col = color_cafe[1:length(cafe.tree$tip.label)], reconst = change_vector[1:length(cafe.tree$tip.label)],
                                   dup_number = cafe.count[1:length(cafe.tree$tip.label)])
     
     cafe.table.nodes <- data.frame(node = (Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label)), label = cafe.tree$node.label,
                                    col = color_cafe[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                    reconst = change_vector[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))],
                                    dup_number = cafe.count[(Ntip(cafe.tree)+1):(Ntip(cafe.tree)+length(cafe.tree$node.label))])
     
     cafe.table.node.comp <- rbind(cafe.table.tips, cafe.table.nodes)
     
     d <- dplyr::mutate(cafe.table.node.comp)
     d$text <- d$label
     
     return(d)
     
   }) %>% bindEvent(input$cafe_start5)
   
   evo_plot5 <- reactive({
     
     d <- evo_plot_data5() 
     mrca.tree <- mrca.tree5()
     
     library(ggtree)
     library(ggplot2)
     
     evo_plot <- ggtree(mrca.tree) %<+% d + aes(colour = I(d$col)) +
       geom_tiplab(aes(label=gsub("_", " ", tools::toTitleCase(d$label))), offset = 30) +
       theme(legend.position = "none") +
       xlim(0, max(mrca.tree$edge.length)*1.85) +
       geom_nodepoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                      alpha = .75) +
       scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
       geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                     alpha = .75)
     
     return(evo_plot)
     
   }) %>% bindEvent(input$cafe_start5)
   
   evo_plotly5 <- reactive({
     
     d <- evo_plot_data5() 
     cafe.tree <- mrca.tree5()
     
     library(ggtree)
     library(ggplot2)
     
     # Remove internal nodes from direct visualization, retaining names in panels
     d$label[172:341] <- NA
     
     evo_plotly <- ggtree(cafe.tree) %<+% d + aes(colour = I(d$col),text=paste0("</br> Duplications: ",dup_number,
                                                                                "</br> Name: ",gsub("_", " ", tools::toTitleCase(d$text)))) + 
       geom_text(aes(x =  1970, label=gsub("_", " ", tools::toTitleCase(d$label)))) + 
       theme(legend.position = "none") +
       xlim(0, max(cafe.tree$edge.length)*1.25) +
       geom_point(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                  alpha = .75) +
       scale_color_manual(values = unique(d$col), breaks = unique(d$col)) +
       geom_tippoint(aes(color=d$col, size = as.numeric(gsub("[*]", "", d$dup_number))*4/max(as.numeric(gsub("[*]", "", d$dup_number)))),
                     alpha = .75)
     
     p <- ggplotly(evo_plotly, tooltip = "text", width = 1400, height = 2200) 
     p <- p %>%
       plotly::style(textposition = "right",xanchor="right")
     
     return(p)
     
   }) %>% bindEvent(input$cafe_start5)
   
   evo.paths.id5 <- reactive({
     
     # Create phylogenomic tree with internal nodes names
     og.cafe <- og.name5()
     cafe.tree <- mrca.tree5()
     
     # Create timeline
     tree.name <- paste("Gene_Trees",paste(og.cafe, "tree.txt", sep = "_"), sep="/")
     tree.ancestor <- read.tree(tree.name)
     tips.orgs1 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[1]])
     tips.orgs2 <- sapply(strsplit(as.character(tree.ancestor$tip.label), "_"), function(x) x[[2]])
     tips.orgs <- paste(tips.orgs1, tips.orgs2, sep = "_")
     
     mrca.id <- getMRCA(cafe.tree,unique(tips.orgs))
     evo.paths <- c()
     for (i in 1:length(unique(tips.orgs)))
     {
       evo.paths <- c(evo.paths, nodepath(cafe.tree, mrca.id, which(unique(tips.orgs)[i] == cafe.tree$tip.label)))
     }
     
     evo.paths <- unique(evo.paths)
     evo.paths.id <- sapply(evo.paths, function(x) if (x <= Ntip(cafe.tree)) cafe.tree$tip.label[x] else cafe.tree$node.label[x - Ntip(cafe.tree)])
     
     #print(paste0("Most recent common ancestor for this orthogroup is ", evo.paths.id[1]))
     
     return(evo.paths.id)
     
   }) %>% bindEvent(input$cafe_start5)
   
   # Outputs
   
   # Remove previous boxes if they exist and create new ones
   observeEvent(isTruthy(evo_plot5()), {
     
     if (UI_exist_cafe5)
     {
       removeUI(
         selector = "div:has(>> #cafe_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_mrca5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #cafe_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCAFEPlot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     if (UI_exist_error_cafe5)
     {
       removeUI(
         selector = "div:has(>> #cafe_error_message5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_cafe5", "afterEnd", ui = {
       box(width = 12,
           title = "Ancestral State Reconstruction", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           plotlyOutput("cafe_plot5", height = ifelse(model.selected5(), "2300px", "2300px")))
     })
     
     insertUI("#box_mrca5", "afterEnd", ui = {
       box(width = 8,
           title = "Most Recent Common Ancestor", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           textOutput("cafe_mrca5")
       )
     })
     
     insertUI("#cafe_down_button5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "cafe_download5", "Download NEWICK tree",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#download_ui_for_cafe_plot5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadCAFEPlot5", "Download Ancestral State Plot",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_cafe5 <<- TRUE
     shinyjs::hideElement(id = 'loading.cafe5')
   })
   
   # Fill outputs
   
   output$cafe_plot5 <- renderPlotly({
     evo_plotly5()
   })
   
   output$cafe_mrca5 <- renderText({
     print(paste0("Most recent common ancestor for this orthogroup is the
                   ancestor of the clade: ", evo.paths.id5()[1]))
   })
   
   # Download tab's results
   
   output$cafe_download5 <- downloadHandler(
     filename= function() {
       paste("ancestral_newick", ".txt", sep="")
     },
     content= function(file) {
       cafe_tree <- cafe_tree5()
       
       write.tree(cafe_tree, file)
     })
   
   output$downloadCAFEPlot5<- downloadHandler(
     filename= function() {
       paste("ancestral_plot", ".png", sep="")
     },
     content= function(file) {
       evo_plot <- evo_plot5()
       
       png(file, width = 1400, height = 2800, res = 100)
       plot(evo_plot)
       dev.off()
     })
   
   
   ####################### MSA #################################
   
   observeEvent(input$run_button5, {
     removeUI(
       selector = "div:has(>> #selected_msaI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(>> #msa_methodI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#msa_selection5")
     
     if (UI_exist_msa5)
     {
       removeUI(
         selector = "div:has(>>> #msa_print5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_fa5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_msa5 <<- F
     }
     
   })
   
   observeEvent(input$msa_start5, {
     insertUI("#selected_msa5", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_msaI5","Select the desired genes from the tree to align",
                                 choices=isolate({tree_reduced5()$tip.label}), options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({"query_prot"}))
       
     })
     
     insertUI("#msa_method5", "afterEnd", ui = {
       shinyWidgets::pickerInput(inputId = "msa_methodI5", label = "Choose alignment method", 
                                 choices = c("ClustalOmega", "MAFFT"), selected = "ClustalOmega")
       
     })
     
     insertUI("#msa_selectionI5", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("msa_selection5", "Align Sequences", size = "sm",
                                style = "float", color = "royal")
     })
     
   })
   
   alignseqs5 <- reactive({
     
     library(msa)
     shinyjs::showElement(id = 'loading.msa5')
     
     selected_genes <- as.vector(input$selected_msaI5)
     selected_method <- as.character(input$msa_methodI5)
     file.name <- og.name5()
     
     if (length(selected_genes) < 2)
     {
       shinyjs::hideElement(id = 'loading.msa5')
       
       if (UI_exist_msa5)
       {
         removeUI(
           selector = "div:has(>>> #msa_print5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #msa_download_plot5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #msa_download_fa5)",
           multiple = TRUE,
           immediate = TRUE
         )
       }
       UI_exist_msa5 <<- F
       output$error_msa5 <- renderUI({renderText({print("Please select at least two genes.")})})
       validate(need(length(selected_genes) > 1, "Please select at least two genes."))
     }
     
     output$error_msa5 <- NULL
     
     # If de novo alignment is selected
     {
       if(selected_method == "ClustalOmega")
       {
         # Define path to orthogroup sequences file
         # ortho.seq.name <- ifelse(model.selected5(),
         #                          paste("Global_Orthogroup_Sequences", paste(file.name, "fa", sep = "."), sep="/"),
         #                          paste("Green_Orthogroup_Sequences", paste(file.name, "fa", sep = "."), sep="/"))
         # Load sequences from OG and query
         
         
         # Read orthogroup sequences file and select the genes for alignment
         random.file <- random.file5()
         ortho_reduced <- ortho_reduced5()
         seqinr::write.fasta(sequences = seqinr::getSequence(ortho_reduced),
                             names = seqinr::getName(ortho_reduced), 
                             file.out = paste0(random.file, "_for_msa.fa"))
         
         mySequences1 <- Biostrings::readAAStringSet(paste0(random.file, "_for_msa.fa"))
         mysubseqs <- mySequences1[selected_genes]
         
         # Remove auxiliar fasta and create alignment
         file.remove(paste0(random.file, "_for_msa.fa"))
         alignseqs <- msa(mysubseqs, verbose = F, method = "ClustalOmega")
       }
       
       # If MAFFT alignment is selected
       else
       {
         mySequences1 <- shoot_msa5()
         #mySequences1 <- seqinr::read.fasta(ortho.seq.name, seqtype = "AA")
         mysubseqs <- mySequences1[selected_genes]
         mysubnames <- seqinr::getName(mySequences1)
         
         # Identify indexes associated with reduced names
         indexes_msa <- sapply(selected_genes, function(x) grep(mysubnames, pattern = x))
         
         # Retrieve those sequences from alignment keeping gaps
         mysubseqs <- mySequences1[indexes_msa]
         names(mysubseqs) <- names(indexes_msa)
         
         # Remove columns with gaps and remove empty spaces in last positions
         seqs_mysubseqs <- seqinr::getSequence(mysubseqs)
         last <- seqs_mysubseqs[[length(seqs_mysubseqs)]]
         last <- last[which(last != " ")]
         seqs_mysubseqs[[length(seqs_mysubseqs)]] <- last
         seqs_mysubseqs <- remove_gaps(seqs_mysubseqs)
         names(seqs_mysubseqs) <- names(mysubseqs)
         
         mysubseqs2 <- unlist(lapply(seqs_mysubseqs, function(x) paste(x, collapse="")))
         
         alignseqs <- Biostrings::AAMultipleAlignment(mysubseqs2, use.names = T)
         
       }
     }
     
     detach("package:msa", unload=TRUE)
     
     return(alignseqs)
     
   }) %>% bindEvent(input$msa_selection5)
   
   # Create boxes for outputs
   observeEvent(isTruthy(alignseqs5()), {
     
     if (UI_exist_msa5)
     {
       removeUI(
         selector = "div:has(>>> #msa_print5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #msa_download_fa5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
    
     insertUI("#box_msa5", "afterEnd", ui = {
       selected_msa <- isolate({input$selected_msaI5})
       msa_height <- ifelse(length(selected_msa) > 14, 550, 400 + 5*length(selected_msa))
       box(width = 12,
           title = "MSA Explorer", status = "primary", solidHeader = TRUE, height = msa_height,
           collapsible = TRUE,
           tags$div(id = "msa_pocket5", style = "width: 1300px; height: 400px",
                    msaROutput("msa_print5"))
           
       )
     })
     
     insertUI("#msa_down_plot5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_plot5", "Download Colored MSA",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#msa_down_fasta5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "msa_download_fa5", "Download MSA FASTA",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_msa5 <<- TRUE
   })
   
   
   # Fill Output
   output$msa_print5 <- renderMsaR({
     alignseqs <- alignseqs5()
     msaout <- msa::msaConvert(alignseqs, "ape::AAbin")
     msaR(msaout, menu=T, overviewbox = F,  colorscheme = "clustal")
   })
   
   # Prepare variables for pdf construction
   observeEvent(isTruthy(alignseqs5()), {
     alignseqs <- alignseqs5()
     
     library(ggmsa)
     class(alignseqs) <- "AAMultipleAlignment"
     
     for(i in 1:(ncol(alignseqs)%/%100 +1)){
       assign(paste("msapseq", i, sep = ""), ggmsa(alignseqs, 1+(100*(i-1)), i*100, seq_name = TRUE, char_width = 0.5) +
                geom_seqlogo(color = "Chemistry_AA"), envir = as.environment(1), pos=1)
     }
     shinyjs::hideElement(id = 'loading.msa5')
   })
   
   # Download tab's results
   # Download colored MSA in pdf
   output$msa_download_plot5 <- downloadHandler(
     filename= function() {
       paste("msa", ".pdf", sep="")
     },
     content= function(file) {
       selected_msa <- input$selected_msaI5
       alignseqs <- alignseqs5()
       pdf(file, height = 2+length(selected_msa)*0.25, width = 16)
       {
         for(i in 1:(ncol(alignseqs)%/%100 +1)){
           print(mget(paste0("msapseq", i), envir = as.environment(1)))
         }
         dev.off()
       }
     })
   
   # Download MSA in FASTA format
   output$msa_download_fa5<- downloadHandler(
     filename= function() {
       paste("msa", ".fa", sep="")
     },
     content= function(file) {
       alignseqs <- alignseqs5()
       writeXStringSet(as(unmasked(alignseqs), "XStringSet"), file)
     })
   
   
   ####################### GO #################################
   
   observeEvent(input$run_button5, {
     removeUI(
       selector = "div:has(>> #selected_gosI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(>> #selected_gos_modeI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#gos_selectionI5")
     
     if (UI_exist_go5)
     {
       removeUI(
         selector = "div:has(>> #output_gos_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_treeplot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadGOSTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_gos_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_go5 <<- F
     }
     
   })
   
   
   observeEvent(input$go_start5, {
     insertUI("#selected_gos5", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_gosI5","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced5()$tip.label[which(tree_reduced5()$tip.label != "query_prot")]}), 
                                 options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({tree_reduced5()$tip.label[1]}))
       
     })
     
     insertUI("#selected_gos_mode5", "afterEnd", ui = {
       
       selectInput(inputId = "selected_gos_modeI5",
                   choices=c("Biological Processes" = "bp",
                             "Molecular Functions" = "mf",
                             "Cellular Components" = "cc"),
                   label = "Select the gene ontology to use",
                   multiple = F, selected = c("bp"))
       
     })
     
     insertUI("#gos_selection5", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("gos_selectionI5", "Show GO terms", size = "sm",
                                style = "float", color = "royal")
     })
     
   })
   
   total_table_gos5 <- reactive({
     
     shinyjs::showElement(id = 'loading.go5')
     gos_anot <- read.csv("pharaoh_folder/final_anot_table.tsv", sep="\t", header = T)
     sel.genes.go <- as.vector(input$selected_gosI5)
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI5}))
     
     total_table_gos <- subset(gos_anot, gos_anot$name %in% sel.genes.go)
     
     # Show an error if no terms are identified in the input
     if (nrow(total_table_gos) == 0) 
     {
       shinyjs::hideElement(id = 'loading.go5')
       if (UI_exist_go5)
       {
         removeUI(
           selector = "div:has(>> #output_gos_table5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_plot5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_treeplot5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadGOSTable5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_download5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_gos_download5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_go5 <<- F
       }
       output$error_gos5 <- renderUI({
         renderPrint({cat("0 GO terms identified.")})
       })
       
       validate(need(nrow(total_table_gos) != 0, " "))
     }
     
     
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     terms_sel <- paste("terms", selected_gos_mode, sep="_")
     total_table_gos <- total_table_gos[,c("organism", "id", "name", gos_sel, terms_sel)]
     
     # Once removed the two GO categories not selected, remove rows with blank cells
     total_table_gos_clean <- total_table_gos[ total_table_gos[[gos_sel]] != "" , ]
     
     # Show an error if no terms are identified after the previous operation
     if (nrow(total_table_gos_clean) == 0) 
     {
       shinyjs::hideElement(id = 'loading.go5')
       if (UI_exist_go5)
       {
         removeUI(
           selector = "div:has(>> #output_gos_table5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_plot5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_treeplot5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadGOSTable5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #gos_download5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #tree_gos_download5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_go5 <<- F
       }
       
       output$error_gos5 <- renderUI({
         renderPrint({cat("0 GO terms identified.")})
       })
       
       validate(need(nrow(total_table_gos_clean) != 0, " "))
     }
     
     output$error_gos5 <- NULL
     
     return(total_table_gos_clean)
     
   }) %>% bindEvent(input$gos_selectionI5)
   
   enr_table5 <- reactive({
     
     total_table_gos <- total_table_gos5()
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI5}))
     
     # Create the plot
     
     # Create a list of GO terms vector of each gene
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     gos_list <- apply(total_table_gos,MARGIN=1,FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
     names(gos_list) <- total_table_gos$name
     
     # Count GOs for each gene and create a matrix of gene-GO pairs
     count_gos_in_genes <- sapply(gos_list, FUN = function(x) length(x))
     comp_data <- data.frame(gene = rep(names(count_gos_in_genes), count_gos_in_genes), gos = as.character(unlist(gos_list)))
     
     # Collapse genes that share a same GO
     gene.v <- c()
     for (i in 1:length(unique(comp_data$gos)))
     {
       new.table <- subset(comp_data, gos == unique(comp_data$gos)[i])
       new.cha <- as.character(new.table$gene)
       gene.v <- c(gene.v, paste(new.cha, collapse = "/"))
     }
     
     names(gene.v) <- unique(comp_data$gos)
     
     # Load libraries and create gene chains, count and GO IDs fields (same order)
     library(GO.db)
     library("multienrichjam")
     library(clusterProfiler)
     library(enrichplot)
     library(ggplot2)
     
     count_go <- table(comp_data$gos)
     geneids <- gene.v[names(count_go)]
     count_terms <- mapply(function(x) {Term(x)}, names(count_go), USE.NAMES = F)
     
     # If there  exists any obsolote term
     count_terms <- count_terms[which(!is.na(count_terms))] 
     count_go <- count_go[which(!is.na(count_terms))]
     geneids <- geneids[which(!is.na(count_terms))]
     
     # Create pseudo-enrichment table
     enr_table <- data.frame(ID=names(count_go), Description=count_terms, GeneRatio="90/100", BgRatio="90/10000", 
                             pvalue=0.000005, p.adjust=0.000005, qvalue=0.000005, geneID=geneids, Count = as.vector(count_go))
     
     return(enr_table)
     
   }) %>% bindEvent(input$gos_selectionI5)
   
   ema_gos_plot5 <- reactive({
     
     enr_table <- enr_table5()
     
     # Transform to enrichResult object
     enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                  geneColname = "geneID", pvalueColname = "p.adjust",
                                  descriptionColname = "Description", pvalueCutoff = 0.05)
     
     # Create plot
     ema_gos_plot <- emapplot(pairwise_termsim(enr), showCategory = 15) + theme(legend.position='none')
     return(ema_gos_plot)
   })
   
   tree_gos_plot5 <- reactive({
     
     enr_table <- enr_table5()
     
     # Transform to enrichResult object
     enr <- enrichDF2enrichResult(enrichDF = enr_table, keyColname = "ID",
                                  geneColname = "geneID", pvalueColname = "p.adjust",
                                  descriptionColname = "Description", pvalueCutoff = 0.05)
     {
       if (nrow(enr_table) > 4)
       {
         tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(label_words_n = 3)) +
           theme(legend.position='none')
       }
       else if (nrow(enr_table) > 2)
       {
         tree_gos_plot <- treeplot(pairwise_termsim(enr),showCategory = 15, cluster.params = list(n = 2, label_words_n = 3)) + 
           theme(legend.position='none')
       }
       else
       {
         text <- paste("\n  Unable to create treeplot with less than 3 GO terms \n")
         tree_gos_plot <- ggplot() + 
           annotate("text", x = 4, y = 25, size=8, label = text) + 
           theme_void()
       }
     }
     
     shinyjs::hideElement(id = 'loading.go5')
     return(tree_gos_plot)
   })
   
   # Create boxes for outputs
   observeEvent(isTruthy(tree_gos_plot5()), {
     
     if (UI_exist_go5)
     {
       removeUI(
         selector = "div:has(>> #output_gos_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_treeplot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadGOSTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #gos_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #tree_gos_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_gos_table5", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Table", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_gos_table5")
       )
     })
     
     insertUI("#box_gos_plot5", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Plot", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           plotOutput("gos_plot5", height = 610)
       )
     })
     
     insertUI("#box_gos_treeplot5", "afterEnd", ui = {
       box(width = 12,
           title = "GO Terms Treeplot", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           plotOutput("gos_treeplot5", height = 610)
       )
     })
     
     insertUI("#download_ui_for_gos_table5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadGOSTable5", "Download GO Table",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#gos_down_button5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "gos_download5", "Download GO Plot",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#tree_gos_down_button5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "tree_gos_download5", "Download GO Treeplot",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_go5 <<- TRUE
   })
   
   # Fill outputs
   # Table
   output$output_gos_table5 <- renderDataTable({
     total_table_gos <- total_table_gos5()
     selected_gos_mode <- as.character(isolate({input$selected_gos_modeI5}))
     gos_sel <- paste("gos", selected_gos_mode, sep="_")
     gos_list <- apply(total_table_gos,MARGIN=1,
                       FUN = function(x) trimws(strsplit(as.character(x[gos_sel]), split = "[|]")[[1]]))
     {
       if(!is.list(gos_list))
       {
         gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
         gos_formatted <- paste0(gos_links, collapse = " | ")
         total_table_gos[,gos_sel] <- gos_formatted
         total_table_gos 
       }
       else
       {
         gos_links <- sapply(gos_list, function(x) sapply(x, go.link))
         gos_formatted <- unlist(sapply(gos_links, function(x) paste0(x, collapse = " | ")))
         total_table_gos[,gos_sel] <- gos_formatted
         total_table_gos  
       }
     }
   },escape=FALSE, rownames=F, options =list(pageLength = 5))
   
   # First plot
   output$gos_plot5 <- renderImage({
     ema_gos_plot <- ema_gos_plot5()
     
     png("pharaoh_folder/gosplot.png", width = 590, height = 590, res = 90)
     plot(ema_gos_plot)
     dev.off()
     list(src = "pharaoh_folder/gosplot.png",
          contentType="image/png")
     
   }, deleteFile=T
   )
   
   # Second plot
   output$gos_treeplot5 <- renderImage({
     tree_gos_plot <- tree_gos_plot5()
     
     png("pharaoh_folder/treeplot.png", width = 620, height = 590, res = 80)
     plot(tree_gos_plot)
     dev.off()
     list(src = "pharaoh_folder/treeplot.png",
          contentType="image/png")
     
   }, deleteFile=T
   )
   
   # Download tab's results
   # Download GO table
   output$downloadGOSTable5 <- downloadHandler(
     filename= function() {
       paste("GOS_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_gos <- total_table_gos5()
       
       write.table(x = total_table_gos,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download emaplot
   output$gos_download5 <- downloadHandler(
     filename= function() {
       paste("gos_plot", ".png", sep="")
     },
     content= function(file) {
       ema_gos_plot <- ema_gos_plot5()
       
       png(file, height = 1200, width = 1500, res=140)
       plot(ema_gos_plot)
       dev.off()
     })
   
   # Download treeplot
   output$tree_gos_download5 <- downloadHandler(
     filename= function() {
       paste("gos_treeplot", ".png", sep="")
     },
     content= function(file) {
       tree_gos_plot <- tree_gos_plot5()
       
       png(file, height = 1200, width = 1500, res=140)
       plot(tree_gos_plot)
       dev.off()
     })
   
   
   ###################### KEGG ###########################
   
   observeEvent(input$run_button5, {
     removeUI(
       selector = "div:has(>> #selected_kosI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#kos_selectionI5")
     
     if (UI_exist_kegg5)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg5 <<- F
     }
     
     if (UI_exist_kegg_path5)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI5")
       
       UI_exist_kegg_path5 <<- F
     }
     
     if (UI_exist_pathview5)
     {
       removeUI(
         selector = "div:has(>>> #path_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview5 <<- F
     }
     
     output$error_kos5 <- NULL
     
     
   })
   
   observeEvent(input$kegg_start5, {
     
     if (UI_exist_kegg5)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg5 <<- F
     }
     
     if (UI_exist_kegg_path5)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI5")
       
       UI_exist_kegg_path5 <<- F
     }
     
     if (UI_exist_pathview5)
     {
       removeUI(
         selector = "div:has(>>> #path_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview5 <<- F
     }
     
     output$error_kos5 <- NULL
     
     insertUI("#selected_kos5", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_kosI5","Select the desired genes from the tree",
                                 choices=isolate({tree_reduced5()$tip.label[which(tree_reduced5()$tip.label != "query_prot")]}), 
                                 options = list(`actions-box` = TRUE),
                                 multiple = T, selected = isolate({tree_reduced5()$tip.label[1]}))
       
       
     })
     
     
     insertUI("#kos_selection5", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("kos_selectionI5", "Show KEGG pathways", size = "sm",
                                style = "float", color = "royal")
     })
     
   })
   
   tab_kegg5 <- reactive({
     
     shinyjs::showElement(id = 'loading.ko5')
     # Create KOs set
     kos_anot <- read.csv("pharaoh_folder/ko_table_funtree.tsv", sep="\t", header = T)
     sel.genes.ko <- as.vector(isolate({input$selected_kosI5}))
     
     tab_kegg <- subset(kos_anot, gene %in% sel.genes.ko)
     set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
     
     # Show an error if no terms are identified in the input
     {
       if (length(set_kegg) == 0) 
       {
         shinyjs::hideElement(id = 'loading.ko5')
         output$error_kos5 <- renderUI({
           renderPrint({cat("0 KO terms identified. Please select more genes. If this 
        message persists, it should be interpreted as a lack of KO annotation for this orthogroup")})
         })
         
         if (UI_exist_kegg5)
         {
           removeUI(
             selector = "div:has(>> #output_kos_table5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKOSTable5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           
           UI_exist_kegg5 <<- F
         }
         
         if (UI_exist_kegg_path5)
         {
           removeUI(
             selector = "div:has(>> #output_kegg_table5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #selected_pathsI5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGTable5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI("#paths_buttonI5")
           
           UI_exist_kegg_path5 <<- F
         }
         
         if (UI_exist_pathview5)
         {
           removeUI(
             selector = "div:has(>>> #path_image5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGpathway5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           UI_exist_pathview5 <<- F
         }
         
         validate("No KO terms detected")
       }
     }
     
     
     return(tab_kegg)
     
   }) %>% bindEvent(input$kos_selectionI5)
   
   total_table_kegg5 <- reactive({
     
     tab_kegg <- tab_kegg5()
     set_kegg <- tab_kegg$ko[tab_kegg$ko != ""]
     
     # Load libraries
     library(clusterProfiler)
     library(enrichplot)
     
     # Enrich with pvalue cutoff = 1 to show all paths
     os_enrich <- enrichKEGG(gene         = unique(unlist(strsplit(set_kegg, "/"))),
                             organism     = 'ko',
                             pvalueCutoff = 1)
     
     total_table_kegg <- as.data.frame(kos_enrich)
     
     # Show an error if the KOs are not mapped to any KEGG pathway
     {
       if (nrow(total_table_kegg) == 0) 
       {
         shinyjs::hideElement(id = 'loading.ko5')
         output$error_kos5 <- renderUI({
           renderPrint({cat("No KEGG pathway appears in this OG.")})
         })
         
         
         if (UI_exist_kegg_path5)
         {
           removeUI(
             selector = "div:has(>> #output_kegg_table5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #selected_pathsI5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGTable5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI("#paths_buttonI5")
           
           UI_exist_kegg_path5 <<- F
         }
         
         if (UI_exist_pathview5)
         {
           removeUI(
             selector = "div:has(>>> #path_image5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           removeUI(
             selector = "div:has(>> #downloadKEGGpathway5)",
             multiple = TRUE,
             immediate = TRUE
           )
           
           UI_exist_pathview5 <<- F
         }
         
         validate("No KEGG pathway appears in this OG.")
       }
     }
     
     output$error_kos5 <- NULL
     
     # Filter out pathways that are not present in plants
     kegg_plants <- read.csv("pharaoh_folder/pathways_plant.ids", sep = "\t", header = T)$x
     total_table_kegg <- subset(total_table_kegg, ID %in% kegg_plants)
     
     return(total_table_kegg)
     
   }) %>% bindEvent(input$kos_selectionI5)
   
   total_table_kos5 <- reactive({
     
     tab_kegg <- tab_kegg5()
     
     library(KEGGREST)
     
     # Collapse genes that share KOs
     
     tab_kegg_for_ko <- subset(tab_kegg, ko != "")
     gene.v.ko <- c()
     individual_kos <- unique(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
     
     for (i in 1:length(individual_kos))
     {
       new.table <- subset(tab_kegg_for_ko, grepl(individual_kos[i],ko))
       new.cha <- as.character(new.table$gene)
       gene.v.ko <- c(gene.v.ko, paste(new.cha, collapse = "/"))
     }
     
     names(gene.v.ko) <- individual_kos
     
     # Create gene chains, count and KO IDs fields (same order)
     count_ko <- table(unlist(strsplit(tab_kegg_for_ko$ko, "/")))
     geneids.ko <- gene.v.ko[names(count_ko)]
     count_terms.ko <- mapply(function(x) {keggFind("ko", x)}, names(count_ko), USE.NAMES = F)
     total_table_kos <- data.frame(ko=names(count_ko), name=count_terms.ko, count=as.numeric(count_ko),
                                   genes=geneids.ko)
     
     return(total_table_kos)
     
   }) %>% bindEvent(input$kos_selectionI5)
   
   # Create boxes for outputs
   observeEvent(isTruthy(total_table_kos5()), {
     
     if (UI_exist_kegg5)
     {
       removeUI(
         selector = "div:has(>> #output_kos_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKOSTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       
       UI_exist_kegg5 <<- F
     }
     
     insertUI("#box_kos_table5", "afterEnd", ui = {
       box(width = 12,
           title = "KO Terms Table", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_kos_table5")
       )
     })
     
     insertUI("#download_ui_for_kos_table5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKOSTable5", "Download KO Table",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_kegg5 <<- TRUE
   })
   
   observeEvent(isTruthy(total_table_kegg5()), {
     
     if (UI_exist_kegg_path5)
     {
       removeUI(
         selector = "div:has(>> #output_kegg_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_pathsI5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#paths_buttonI5")
       
       UI_exist_kegg_path5 <<- F
     }
     
     
     insertUI("#box_kegg_table5", "afterEnd", ui = {
       box(width = 12,
           title = "KEGG Pathways Table", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_kegg_table5")
       )
     })
     
     
     
     insertUI("#download_ui_for_kegg_table5", "afterEnd", ui = {
       tags$div(style = "margin-left: 200px;", shinyWidgets::downloadBttn(outputId= "downloadKEGGTable5", "Download Pathways Table",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_kegg_path5 <<- TRUE
     
     # Remove previous results for pathview
     if (UI_exist_pathview5)
     {
       removeUI(
         selector = "div:has(>>> #path_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_pathview5 <<- F
     }
   })
   
   # Fill outputs
   # Render KO table
   output$output_kos_table5 <- renderDataTable({
     total_table_kos <- total_table_kos5()
     total_table_kos$ko <- sapply(total_table_kos$ko, ko.link)
     total_table_kos
   },escape=FALSE, rownames= F, options =list(pageLength = 5))
   
   # Render KEGG table
   output$output_kegg_table5 <- renderDataTable({
     total_table_kegg <- total_table_kegg5()
     total_table_kegg$ID <- sapply(total_table_kegg$ID, kegg.link)
     total_table_kegg[,c("ID", "Description", "geneID")]
   },escape=FALSE, rownames= F, options =list(pageLength = 5))
   
   # Download tab's outputs
   # Download KO table
   output$downloadKOSTable5 <- downloadHandler(
     filename= function() {
       paste("KO_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_kos <- total_table_kos5()
       write.table(x = total_table_kos,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download KEGG table
   output$downloadKEGGTable5 <- downloadHandler(
     filename= function() {
       paste("KEGG_table", ".tsv", sep="")
     },
     content= function(file) {
       total_table_kegg <- total_table_kegg5()
       write.table(x = total_table_kegg[,c("ID", "Description", "geneID")],quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Create pathway selector and button
   observeEvent(input$kos_selectionI5,{
     
     total_table_kos <- total_table_kos5()
     total_table_kegg <- total_table_kegg5()
     
     if(nrow(total_table_kegg) != 0)
     {
       paths.options <- sapply(strsplit(total_table_kegg$ID, split = "ko"), function(x) x[[2]])
       
       
       insertUI("#selected_paths5", "afterEnd", ui = {
         shinyWidgets::pickerInput(inputId = "selected_pathsI5", label = "Select the pathway to plot", 
                                   choices = paths.options, selected = paths.options[1], multiple = F)
         
       })
       
       
       insertUI("#paths_button5", "afterEnd", ui = {
         
         shinyWidgets::actionBttn("paths_buttonI5", "Plot Pathway", size = "sm",
                                  style = "float", color = "royal")
       })
       
       shinyjs::hideElement(id = 'loading.ko5')
     }
   })
   
   observeEvent(input$paths_buttonI5,{
     
     if (UI_exist_pathview5)
     {
       removeUI(
         selector = "div:has(>>> #path_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadKEGGpathway5)",
         multiple = TRUE,
         immediate = TRUE
       )
     }
     
     UI_exist_pathview5 <<- F
     
   })
   
   # Create Image to Render and save path name
   pathway.current.id5 <- reactive({
     pathway.current.id <- input$selected_pathsI5
     total_table_kos <- total_table_kos5()
     
     kos_unique <- unique(total_table_kos$ko)
     gene.pathway <- rep(0, length(kos_unique))
     names(gene.pathway) <-  kos_unique
     gene.pathway[kos_unique] <-1
     
     library(pathview)
     pathview(gene.data = sort(gene.pathway,decreasing = TRUE),kegg.dir = "pharaoh_folder",
              pathway.id = pathway.current.id,
              species = "ko",
              limit = list(gene=max(abs(gene.pathway)), cpd=1),
              gene.idtype ="kegg")
     
     return(pathway.current.id)
     
   }) %>% bindEvent(input$paths_buttonI5)
   
   # Create output box and download button
   observeEvent(isTruthy(pathway.current.id5()),{
     
     insertUI("#box_path_image5", "afterEnd", ui = {
       box(width = 12,
           title = "KEGG Pathway Plot", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           fluidRow(column(1), imageOutput("path_image5", width = "100%", height = "700px"))
       )
     })
     
     insertUI("#path_download_ui5", "afterEnd", ui = {
       tags$div(shinyWidgets::downloadBttn(outputId= "downloadKEGGpathway5", "Download KEGG Pathway Plot",
                                           size = "sm", color = "primary"))
     })
     
     UI_exist_pathview5 <<- T
     
   })
   
   # Fill path image output
   output$path_image5 <- renderImage({
     
     pathway.current.id <- pathway.current.id5()
     list(src = paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."),
          contentType="image/png",width=900,height=700)
   },deleteFile = F)
   
   # Download and remove path image output
   output$downloadKEGGpathway5 <- downloadHandler(
     filename= function() {
       paste("path_plot", ".png", sep="")
     },
     content= function(file) {
       pathway.current.id <- pathway.current.id5()
       file.copy(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."), file)
       file.remove(paste(c(paste0(c("ko",pathway.current.id), collapse=""),"pathview","png"), collapse="."))
     })   
   
   
   
   ############################# LITERATURE ANNOTATION ##########################################
   
   observeEvent(input$run_button5, {
     removeUI(
       selector = "div:has(>> #selected_litI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI(
       selector = "div:has(> #query_litI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#lit_selectionI5")
     
     
     if (UI_exist_lit5)
     {
       removeUI(
         selector = "div:has(>> #output_lit_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadLITTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_lit5 <<- F
     }
     
   })
   
   observeEvent(input$lit_start5, {
     
     insertUI("#query_lit5", "afterEnd", ui = {
       
       textInput(inputId = "query_litI5",value = "", label = "Enter search term", placeholder = "CCA1")
       
     })
     
     
     insertUI("#selected_lit5", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_litI5","Select the search mode",
                                 choices=c("Normal","Exact", "Substring", "Alias"),
                                 multiple = F, selected = "Normal")
       
       
     })
     
     
     insertUI("#lit_selection5", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("lit_selectionI5", "Get biological information and papers", size = "sm",
                                style = "float", color = "royal")
     })
     
   })
   
   pc_result5 <- reactive({
     
     shinyjs::showElement(id = 'loading.lit5')
     pc_search <- as.character(input$query_litI5)
     pc_search <- gsub(" ", "%20", pc_search) 
     pc_modality <- tolower(as.character(input$selected_litI5))
     
     # Get PlantConnectome URL for query
     pc_url <- paste(c("https://connectome.plant.tools", pc_modality, pc_search), collapse = "/")
     
     library(RCurl)
     
     pc_res <- getURL(pc_url)
     
     if (!length(grep("No hits", pc_res)) == 0)
     {
       shinyjs::hideElement(id = 'loading.lit5')
       
       if (UI_exist_lit5)
       {
         removeUI(
           selector = "div:has(>> #output_lit_table5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         removeUI(
           selector = "div:has(>> #downloadLITTable5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_lit5 <<- F
       }
       
       output$error_lit5 <- renderUI({
         renderPrint({cat("No results found for this query")})
       })
       
       validate(" ")
       
     }
     
     output$error_lit5 <- NULL
     
     # Isolate data frame from complete HTML file
     pc_split <- strsplit(as.character(pc_res), split = "<tbody")[[1]][2]
     pc_split <- strsplit(as.character(pc_split), split = "</tbody>")[[1]][1]
     
     pc_clean <- gsub("</tr>", "", pc_split)
     pc_clean <- gsub("[\r\n\t]", "", pc_clean)
     
     pc_vector <- strsplit(pc_clean, split = "<tr>")[[1]][-1]
     pc_vector2 <- sapply(pc_vector, FUN=function(x) strsplit(x, split = "<td> | </td>"))
     pc_table <- sapply(pc_vector2, FUN=function(x) as.character(unlist(x)))
     
     colnames(pc_table) <- NULL
     pc_result <- data.frame(t(pc_table[c(2,4,6,8),]))
     colnames(pc_result) <- c("Source", "Interaction Type", "Target", "Pubmed ID")
     
     return(pc_result)
     
   }) %>% bindEvent(input$lit_selectionI5)
   
   pc_result_show5 <- reactive({
     
     pc_result <- pc_result5()
     
     # Add links to papers
     urls_connect <- sapply(pc_result$`Pubmed ID`, FUN = function(x) paste0(c("<a href=\"",
                                                                              "https://pubmed.ncbi.nlm.nih.gov/",x,"/",
                                                                              "\" target=\"_blank\">", x,
                                                                              "</a>"),
                                                                            collapse=""))
     pc_result_show <- pc_result
     pc_result_show$`Pubmed ID` <- urls_connect
     
     return(pc_result_show)
     
   })
   
   # Create boxes for outputs
   observeEvent(isTruthy(pc_result_show5()), {
     
     if (UI_exist_lit5)
     {
       removeUI(
         selector = "div:has(>> #output_lit_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadLITTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     insertUI("#box_lit_table5", "afterEnd", ui = {
       box(width = 12,
           title = "Literature Table", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_lit_table5")
       )
     })
     
     insertUI("#download_ui_for_lit_table5", "afterEnd", ui = {
       tags$div(style = "margin-left: 400px;", shinyWidgets::downloadBttn(outputId= "downloadLITTable5", "Download Literature Table",
                                                                          size = "sm", color = "primary"))
     })
     
     shinyjs::hideElement(id = 'loading.lit5')
     
     UI_exist_lit5 <<- TRUE
     
   })
   
   # Fill outputs
   # Render table
   output$output_lit_table5 <- renderDataTable({
     pc_result_show5()
   },escape=FALSE, rownames= F, options =list(pageLength = 10))
   
   # Download results
   output$downloadLITTable5 <- downloadHandler(
     filename= function() {
       paste("literature_table", ".tsv", sep="")
     },
     content= function(file) {
       pc_result <- pc_result5()
       write.table(x = pc_result,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })   
   
   
   ######################## STRING ###########################
   
   observeEvent(input$run_button5, {
     removeUI(
       selector = "div:has(>> #selected_stringI5)",
       multiple = TRUE,
       immediate = TRUE
     )
     
     removeUI("#string_selectionI5")
     
     shinyjs::hideElement("error_string5")
     
     
     if (UI_exist_string5)
     {
       removeUI(
         selector = "div:has(>> #output_st_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #output_count_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #count_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadSTRINGTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCOUNTTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #count_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_networkI5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#network_buttonI5")
       
       UI_exist_string5 <<- F
     }
     
     if (UI_exist_network5)
     {
       removeUI(
         selector = "div:has(>>>> #network_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_network5 <<- F
     }
     
   })
   
   # First, create a table that links reduced gene names to its species
   string_sel_table5 <- reactive({
     
     # Define previous variables
     tree_reduced <- tree_reduced5()
     tree <- tree_adj5()
  
     tips_to_keep.mp <- tips_to_keep.mp5()
     tips_to_keep.ot <- tips_to_keep.ot5()
     tips_to_keep.at <- tips_to_keep.at5()
     tips_to_keep.cp <- tips_to_keep.cp5()
     tips_to_keep.cr <- tips_to_keep.cr5()
     tips_to_keep.cz <- tips_to_keep.cz5()
     tips_to_keep.kn <- tips_to_keep.kn5()
     tips_to_keep.me <- tips_to_keep.me5()
     tips_to_keep.mi <- tips_to_keep.mi5()
     tips_to_keep.pp <- tips_to_keep.pp5()
     tips_to_keep.sl <- tips_to_keep.sl5()
     tips_to_keep.sm <- tips_to_keep.sm5()
     tips_to_keep.sp <- tips_to_keep.sp5()
     tips_to_keep.ta <- tips_to_keep.ta5()
     tips_to_keep.vc <- tips_to_keep.vc5()
     tips_to_keep.bp <- tips_to_keep.bp5()
     tips_to_keep.cri <- tips_to_keep.cri5()
     tips_to_keep.ds <- tips_to_keep.ds5()
     tips_to_keep.os <- tips_to_keep.os5()
     tips_to_keep.smag <- tips_to_keep.smag5()
     tips_to_keep.tp <- tips_to_keep.tp5()
     tips_to_keep.aa <- tips_to_keep.aa5()
     tips_to_keep.um <- tips_to_keep.um5()
     tips_to_keep.rs <- tips_to_keep.rs5()
     tips_to_keep.cyc <- tips_to_keep.cyc5()
     tips_to_keep.ca <- tips_to_keep.ca5()
     tips_to_keep.mv <- tips_to_keep.mv5()
     tips_to_keep.af <- tips_to_keep.af5()
     tips_to_keep.sc <- tips_to_keep.sc5()
     tips_to_keep.aegi <- tips_to_keep.aegi5()
     tips_to_keep.sb <- tips_to_keep.sb5()
     tips_to_keep.chara <- tips_to_keep.chara5()
     tips_to_keep.sceobli <- tips_to_keep.sceobli5()
     tips_to_keep.cocco <- tips_to_keep.cocco5()
     tips_to_keep.haema <- tips_to_keep.haema5()
     tips_to_keep.zm <- tips_to_keep.zm5()
     tips_to_keep.praco <- tips_to_keep.praco5()
     tips_to_keep.ostlu <- tips_to_keep.ostlu5()
     tips_to_keep.ostme <- tips_to_keep.ostme5()
     tips_to_keep.micco <- tips_to_keep.micco5()
     tips_to_keep.cymte <- tips_to_keep.cymte5()
     tips_to_keep.chlin <- tips_to_keep.chlin5()
     tips_to_keep.gonpe <- tips_to_keep.gonpe5()
     tips_to_keep.tetso <- tips_to_keep.tetso5()
     tips_to_keep.desar <- tips_to_keep.desar5()
     tips_to_keep.monmi <- tips_to_keep.monmi5()
     tips_to_keep.caule <- tips_to_keep.caule5()
     tips_to_keep.ostqu <- tips_to_keep.ostqu5()
     tips_to_keep.brysp <- tips_to_keep.brysp5()
     tips_to_keep.ellbi <- tips_to_keep.ellbi5()
     tips_to_keep.myrbi <- tips_to_keep.myrbi5()
     tips_to_keep.apalo <- tips_to_keep.apalo5()
     tips_to_keep.astgl <- tips_to_keep.astgl5()
     tips_to_keep.tresp <- tips_to_keep.tresp5()
     tips_to_keep.picso <- tips_to_keep.picso5()
     tips_to_keep.chlso <- tips_to_keep.chlso5()
     tips_to_keep.auxpr <- tips_to_keep.auxpr5()
     tips_to_keep.helsp <- tips_to_keep.helsp5()
     tips_to_keep.tetst <- tips_to_keep.tetst5()
     tips_to_keep.pedsp <- tips_to_keep.pedsp5()
     tips_to_keep.chlpr <- tips_to_keep.chlpr5()
     tips_to_keep.picsp <- tips_to_keep.picsp5()
     tips_to_keep.mesvb <- tips_to_keep.mesvb5()
     tips_to_keep.meskr <- tips_to_keep.meskr5()
     tips_to_keep.zygci <- tips_to_keep.zygci5()
     tips_to_keep.zygcb <- tips_to_keep.zygcb5()
     tips_to_keep.zygcy <- tips_to_keep.zygcy5()
     tips_to_keep.plesc <- tips_to_keep.plesc5()
     tips_to_keep.funhy <- tips_to_keep.funhy5()
     tips_to_keep.sphfa <- tips_to_keep.sphfa5()
     tips_to_keep.takle <- tips_to_keep.takle5()
     tips_to_keep.marpa <- tips_to_keep.marpa5()
     tips_to_keep.ricfl <- tips_to_keep.ricfl5()
     tips_to_keep.ricso <- tips_to_keep.ricso5()
     tips_to_keep.antpu <- tips_to_keep.antpu5()
     tips_to_keep.antfu <- tips_to_keep.antfu5()
     tips_to_keep.megfl <- tips_to_keep.megfl5()
     tips_to_keep.phach <- tips_to_keep.phach5()
     tips_to_keep.phyph <- tips_to_keep.phyph5()
     tips_to_keep.parpe <- tips_to_keep.parpe5()
     tips_to_keep.notor <- tips_to_keep.notor5()
     tips_to_keep.phaca <- tips_to_keep.phaca5()
     tips_to_keep.phasp <- tips_to_keep.phasp5()
     tips_to_keep.leidu <- tips_to_keep.leidu5()
     tips_to_keep.isosi <- tips_to_keep.isosi5()
     tips_to_keep.dipco <- tips_to_keep.dipco5()
     tips_to_keep.hupas <- tips_to_keep.hupas5()
     tips_to_keep.lyccl <- tips_to_keep.lyccl5()
     tips_to_keep.adica <- tips_to_keep.adica5()
     tips_to_keep.alssp <- tips_to_keep.alssp5()
     tips_to_keep.marve <- tips_to_keep.marve5()
     tips_to_keep.seqse <- tips_to_keep.seqse5()
     tips_to_keep.seqgi <- tips_to_keep.seqgi5()
     tips_to_keep.taxch <- tips_to_keep.taxch5()
     tips_to_keep.abial <- tips_to_keep.abial5()
     tips_to_keep.picab <- tips_to_keep.picab5()
     tips_to_keep.gnemo <- tips_to_keep.gnemo5()
     tips_to_keep.welmi <- tips_to_keep.welmi5()
     tips_to_keep.ginbi <- tips_to_keep.ginbi5()
     tips_to_keep.ambtr <- tips_to_keep.ambtr5()
     tips_to_keep.nymco <- tips_to_keep.nymco5()
     tips_to_keep.cinka <- tips_to_keep.cinka5()
     tips_to_keep.horvu <- tips_to_keep.horvu5()
     tips_to_keep.bradi <- tips_to_keep.bradi5()
     tips_to_keep.setit <- tips_to_keep.setit5()
     tips_to_keep.panha <- tips_to_keep.panha5()
     tips_to_keep.missi <- tips_to_keep.missi5()
     tips_to_keep.oroth <- tips_to_keep.oroth5()
     tips_to_keep.eleco <- tips_to_keep.eleco5()
     tips_to_keep.anaco <- tips_to_keep.anaco5()
     tips_to_keep.musac <- tips_to_keep.musac5()
     tips_to_keep.dioal <- tips_to_keep.dioal5()
     tips_to_keep.spipo <- tips_to_keep.spipo5()
     tips_to_keep.aspof <- tips_to_keep.aspof5()
     tips_to_keep.zosma <- tips_to_keep.zosma5()
     tips_to_keep.araly <- tips_to_keep.araly5()
     tips_to_keep.capgr <- tips_to_keep.capgr5()
     tips_to_keep.brara <- tips_to_keep.brara5()
     tips_to_keep.braol <- tips_to_keep.braol5()
     tips_to_keep.sisir <- tips_to_keep.sisir5()
     tips_to_keep.schpa <- tips_to_keep.schpa5()
     tips_to_keep.eutsa <- tips_to_keep.eutsa5()
     tips_to_keep.thlar <- tips_to_keep.thlar5()
     tips_to_keep.boest <- tips_to_keep.boest5()
     tips_to_keep.carpa <- tips_to_keep.carpa5()
     tips_to_keep.theca <- tips_to_keep.theca5()
     tips_to_keep.goshi <- tips_to_keep.goshi5()
     tips_to_keep.gosra <- tips_to_keep.gosra5()
     tips_to_keep.citsi <- tips_to_keep.citsi5()
     tips_to_keep.pontr <- tips_to_keep.pontr5()
     tips_to_keep.eucgr <- tips_to_keep.eucgr5()
     tips_to_keep.maldo <- tips_to_keep.maldo5()
     tips_to_keep.prupe <- tips_to_keep.prupe5()
     tips_to_keep.frave <- tips_to_keep.frave5()
     tips_to_keep.corav <- tips_to_keep.corav5()
     tips_to_keep.caril <- tips_to_keep.caril5()
     tips_to_keep.queru <- tips_to_keep.queru5()
     tips_to_keep.cucsa <- tips_to_keep.cucsa5()
     tips_to_keep.glyma <- tips_to_keep.glyma5()
     tips_to_keep.medtr <- tips_to_keep.medtr5()
     tips_to_keep.tripr <- tips_to_keep.tripr5()
     tips_to_keep.cicar <- tips_to_keep.cicar5()
     tips_to_keep.lotja <- tips_to_keep.lotja5()
     tips_to_keep.phaac <- tips_to_keep.phaac5()
     tips_to_keep.vigun <- tips_to_keep.vigun5()
     tips_to_keep.lupal <- tips_to_keep.lupal5()
     tips_to_keep.lencu <- tips_to_keep.lencu5()
     tips_to_keep.arahy <- tips_to_keep.arahy5()
     tips_to_keep.poptr <- tips_to_keep.poptr5()
     tips_to_keep.manes <- tips_to_keep.manes5()
     tips_to_keep.salpu <- tips_to_keep.salpu5()
     tips_to_keep.linus <- tips_to_keep.linus5()
     tips_to_keep.corci <- tips_to_keep.corci5()
     tips_to_keep.vitvi <- tips_to_keep.vitvi5()
     tips_to_keep.kalfe <- tips_to_keep.kalfe5()
     tips_to_keep.vacda <- tips_to_keep.vacda5()
     tips_to_keep.oleeu <- tips_to_keep.oleeu5()
     tips_to_keep.mimgu <- tips_to_keep.mimgu5()
     tips_to_keep.soltu <- tips_to_keep.soltu5()
     tips_to_keep.cofar <- tips_to_keep.cofar5()
     tips_to_keep.lacsa <- tips_to_keep.lacsa5()
     tips_to_keep.dauca <- tips_to_keep.dauca5()
     tips_to_keep.betvu <- tips_to_keep.betvu5()
     tips_to_keep.amahy <- tips_to_keep.amahy5()
     tips_to_keep.chequ <- tips_to_keep.chequ5()
     tips_to_keep.sapof <- tips_to_keep.sapof5()
     tips_to_keep.aquco <- tips_to_keep.aquco5()
     tips_to_keep.query <- tips_to_keep.query5()
     
     # Table construction
     org.factor <- c()
     
     for (i in 1:length(tree_reduced$tip.label))
     {
       if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mp])
       {
         org.factor <- c(org.factor,"Marchantia polymorpha")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ot])
       {
         org.factor <- c(org.factor,"Ostreococcus tauri")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.at])
       {
         org.factor <- c(org.factor,"Arabidopsis thaliana")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cp])
       {
         org.factor <- c(org.factor,"Ceratodon purpureus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cr])
       {
         org.factor <- c(org.factor,"Chlamydomonas reinhardtii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cz])
       {
         org.factor <- c(org.factor,"Chromochloris zofingiensis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kn])
       {
         org.factor <- c(org.factor,"Klebsormidium nitens")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.me])
       {
         org.factor <- c(org.factor,"Mesotaenium endlicherianum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mi])
       {
         org.factor <- c(org.factor,"Micromonas pusilla")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pp])
       {
         org.factor <- c(org.factor,"Physcomitrium patens")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sl])
       {
         org.factor <- c(org.factor,"Solanum lycopersicum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sm])
       {
         org.factor <- c(org.factor,"Selaginella moellendorffii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sp])
       {
         org.factor <- c(org.factor,"Spirogloea muscicola")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ta])
       {
         org.factor <- c(org.factor,"Triticum aestivum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vc])
       {
         org.factor <- c(org.factor,"Volvox carteri")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bp])
       {
         org.factor <- c(org.factor,"Bathycoccus prasinos")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cri])
       {
         org.factor <- c(org.factor,"Ceratopteris richardii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ds])
       {
         org.factor <- c(org.factor,"Dunaliella salina")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.os])
       {
         org.factor <- c(org.factor,"Oryza sativa")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.smag])
       {
         org.factor <- c(org.factor,"Sphagnum magellanicum")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tp])
       {
         org.factor <- c(org.factor,"Thuja plicata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aa])
       {
         org.factor <- c(org.factor,"Anthoceros agrestis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.um])
       {
         org.factor <- c(org.factor,"Ulva mutabilis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.rs])
       {
         org.factor <- c(org.factor,"Raphidocelis subcapitata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cyc])
       {
         org.factor <- c(org.factor,"Cycas panzhihuaensis")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ca])
       {
         org.factor <- c(org.factor,"Chlorokybus atmophyticus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mv])
       {
         org.factor <- c(org.factor,"Mesostigma viride CCAC 1140")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.af])
       {
         org.factor <- c(org.factor,"Azolla filiculoides")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sc])
       {
         org.factor <- c(org.factor,"Salvinia cucullata")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aegi])
       {
         org.factor <- c(org.factor,"Aegilops tauschii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sb])
       {
         org.factor <- c(org.factor,"Sorghum bicolor")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chara])
       {
         org.factor <- c(org.factor,"Chara braunii")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sceobli])
       {
         org.factor <- c(org.factor,"Scenedesmus obliquus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cocco])
       {
         org.factor <- c(org.factor,"Coccomyxa subellipsoidea")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.haema])
       {
         org.factor <- c(org.factor,"Haematococcus lacustris")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zm])
       {
         org.factor <- c(org.factor,"Zea mays")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.praco])
       {
         org.factor <- c(org.factor,"Prasinoderma coloniale")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostlu])
       {
         org.factor <- c(org.factor,"Ostreococcus lucimarinus")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostme])
       {
         org.factor <- c(org.factor,"Ostreococcus mediterraneus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.micco])
       {
         org.factor <- c(org.factor,"Micromonas commoda")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cymte])
       {
         org.factor <- c(org.factor,"Cymbomonas tetramitiformis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlin])
       {
         org.factor <- c(org.factor,"Chlamydomonas incerta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gonpe])
       {
         org.factor <- c(org.factor,"Gonium pectorale")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetso])
       {
         org.factor <- c(org.factor,"Tetrabaena socialis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.desar])
       {
         org.factor <- c(org.factor,"Desmodesmus armatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.monmi])
       {
         org.factor <- c(org.factor,"Monoraphidium minutum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caule])
       {
         org.factor <- c(org.factor,"Caulerpa lentillifera")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ostqu])
       {
         org.factor <- c(org.factor,"Ostreobium quekettii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brysp])
       {
         org.factor <- c(org.factor,"Bryopsis sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ellbi])
       {
         org.factor <- c(org.factor,"Elliptochloris bilobata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.myrbi])
       {
         org.factor <- c(org.factor,"Myrmecia bisecta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.apalo])
       {
         org.factor <- c(org.factor,"Apatococcus lobatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.astgl])
       {
         org.factor <- c(org.factor,"Asterochloris glomerata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tresp])
       {
         org.factor <- c(org.factor,"Trebouxia sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picso])
       {
         org.factor <- c(org.factor,"Picochlorum soloecismus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlso])
       {
         org.factor <- c(org.factor,"Chlorella sorokiniana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.auxpr])
       {
         org.factor <- c(org.factor,"Auxenochlorella protothecoides")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.helsp])
       {
         org.factor <- c(org.factor,"Helicosporidium sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tetst])
       {
         org.factor <- c(org.factor,"Tetraselmis striata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pedsp])
       {
         org.factor <- c(org.factor,"Pedinophyceae sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chlpr])
       {
         org.factor <- c(org.factor,"Chloropicon primus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picsp])
       {
         org.factor <- c(org.factor,"Picocystis sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mesvb])
       {
         org.factor <- c(org.factor,"Mesostigma viride NIES 296")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.meskr])
       {
         org.factor <- c(org.factor,"Mesotaenium kramstae")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygci])
       {
         org.factor <- c(org.factor,"Zygnema circumcarinatum SAG")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcb])
       {
         org.factor <- c(org.factor,"Zygnema circumcarinatum UTEX 1559")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zygcy])
       {
         org.factor <- c(org.factor,"Zygnema cylindricum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.plesc])
       {
         org.factor <- c(org.factor,"Pleurozium schreberi")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.funhy])
       {
         org.factor <- c(org.factor,"Funaria hygrometrica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sphfa])
       {
         org.factor <- c(org.factor,"Sphagnum fallax")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.takle])
       {
         org.factor <- c(org.factor,"Takakia lepidozioides")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marpa])
       {
         org.factor <- c(org.factor,"Marchantia paleacea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricfl])
       {
         org.factor <- c(org.factor,"Riccia fluitans")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ricso])
       {
         org.factor <- c(org.factor,"Riccia sorocarpa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antpu])
       {
         org.factor <- c(org.factor,"Anthoceros punctatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.antfu])
       {
         org.factor <- c(org.factor,"Anthoceros fusiformis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.megfl])
       {
         org.factor <- c(org.factor,"Megaceros flagellaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phach])
       {
         org.factor <- c(org.factor,"Phaeomegaceros chiloensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phyph])
       {
         org.factor <- c(org.factor,"Phymatoceros phymatodes")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.parpe])
       {
         org.factor <- c(org.factor,"Paraphymatoceros pearsonii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.notor])
       {
         org.factor <- c(org.factor,"Notothylas orbicularis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaca])
       {
         org.factor <- c(org.factor,"Phaeoceros carolinianus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phasp])
       {
         org.factor <- c(org.factor,"Phaeoceros sp.")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.leidu])
       {
         org.factor <- c(org.factor,"Leiosporoceros dussii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.isosi])
       {
         org.factor <- c(org.factor,"Isoetes sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dipco])
       {
         org.factor <- c(org.factor,"Diphasiastrum complanatum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.hupas])
       {
         org.factor <- c(org.factor,"Huperzia asiatica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lyccl])
       {
         org.factor <- c(org.factor,"Lycopodium clavatum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.adica])
       {
         org.factor <- c(org.factor,"Adiantum capillus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.alssp])
       {
         org.factor <- c(org.factor,"Alsophila spinulosa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.marve])
       {
         org.factor <- c(org.factor,"Marsilea vestita")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqse])
       {
         org.factor <- c(org.factor,"Sequoia sempervirens")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.seqgi])
       {
         org.factor <- c(org.factor,"Sequoiadendron giganteum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.taxch])
       {
         org.factor <- c(org.factor,"Taxus chinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.abial])
       {
         org.factor <- c(org.factor,"Abies alba")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.picab])
       {
         org.factor <- c(org.factor,"Picea abies")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gnemo])
       {
         org.factor <- c(org.factor,"Gnetum montanum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.welmi])
       {
         org.factor <- c(org.factor,"Welwitschia mirabilis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ginbi])
       {
         org.factor <- c(org.factor,"Ginkgo biloba")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.ambtr])
       {
         org.factor <- c(org.factor,"Amborella trichopoda")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.nymco])
       {
         org.factor <- c(org.factor,"Nymphaea colorata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cinka])
       {
         org.factor <- c(org.factor,"Cinnamomum kanehirae")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.horvu])
       {
         org.factor <- c(org.factor,"Hordeum vulgare")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.bradi])
       {
         org.factor <- c(org.factor,"Brachypodium distachyon")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.setit])
       {
         org.factor <- c(org.factor,"Setaria italica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.panha])
       {
         org.factor <- c(org.factor,"Panicum hallii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.missi])
       {
         org.factor <- c(org.factor,"Miscanthus sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oroth])
       {
         org.factor <- c(org.factor,"Oropetium thomaeum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eleco])
       {
         org.factor <- c(org.factor,"Eleusine coracana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.anaco])
       {
         org.factor <- c(org.factor,"Ananas comosus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.musac])
       {
         org.factor <- c(org.factor,"Musa acuminata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dioal])
       {
         org.factor <- c(org.factor,"Dioscorea alata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.spipo])
       {
         org.factor <- c(org.factor,"Spirodela polyrhiza")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aspof])
       {
         org.factor <- c(org.factor,"Asparagus officinalis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.zosma])
       {
         org.factor <- c(org.factor,"Zostera marina")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.araly])
       {
         org.factor <- c(org.factor,"Arabidopsis lyrata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.capgr])
       {
         org.factor <- c(org.factor,"Capsella grandiflora")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.brara])
       {
         org.factor <- c(org.factor,"Brassica rapa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.braol])
       {
         org.factor <- c(org.factor,"Brassica oleracea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sisir])
       {
         org.factor <- c(org.factor,"Sisymbrium irio")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.schpa])
       {
         org.factor <- c(org.factor,"Schrenkiella parvula")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eutsa])
       {
         org.factor <- c(org.factor,"Eutrema salsugineum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.thlar])
       {
         org.factor <- c(org.factor,"Thlaspi arvense")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.boest])
       {
         org.factor <- c(org.factor,"Boechera stricta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.carpa])
       {
         org.factor <- c(org.factor,"Carica papaya")
       }
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.theca])
       {
         org.factor <- c(org.factor,"Theobroma cacao")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.goshi])
       {
         org.factor <- c(org.factor,"Gossypium hirsutum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.gosra])
       {
         org.factor <- c(org.factor,"Gossypium raimondii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.citsi])
       {
         org.factor <- c(org.factor,"Citrus sinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.pontr])
       {
         org.factor <- c(org.factor,"Poncirus trifoliata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.eucgr])
       {
         org.factor <- c(org.factor,"Eucalyptus grandis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.maldo])
       {
         org.factor <- c(org.factor,"Malus domestica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.prupe])
       {
         org.factor <- c(org.factor,"Prunus persica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.frave])
       {
         org.factor <- c(org.factor,"Fragaria vesca")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corav])
       {
         org.factor <- c(org.factor,"Corylus avellana")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.caril])
       {
         org.factor <- c(org.factor,"Carya illinoinensis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.queru])
       {
         org.factor <- c(org.factor,"Quercus rubra")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cucsa])
       {
         org.factor <- c(org.factor,"Cucumis sativus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.glyma])
       {
         org.factor <- c(org.factor,"Glycine max")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.medtr])
       {
         org.factor <- c(org.factor,"Medicago truncatula")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.tripr])
       {
         org.factor <- c(org.factor,"Trifolium pratense")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cicar])
       {
         org.factor <- c(org.factor,"Cicer arietinum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lotja])
       {
         org.factor <- c(org.factor,"Lotus japonicus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.phaac])
       {
         org.factor <- c(org.factor,"Phaseolus acutifolius")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vigun])
       {
         org.factor <- c(org.factor,"Vigna unguiculata")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lupal])
       {
         org.factor <- c(org.factor,"Lupinus albus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lencu])
       {
         org.factor <- c(org.factor,"Lens culinaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.arahy])
       {
         org.factor <- c(org.factor,"Arachis hypogaea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.poptr])
       {
         org.factor <- c(org.factor,"Populus trichocarpa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.manes])
       {
         org.factor <- c(org.factor,"Manihot esculenta")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.salpu])
       {
         org.factor <- c(org.factor,"Salix purpurea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.linus])
       {
         org.factor <- c(org.factor,"Linum usitatissimum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.corci])
       {
         org.factor <- c(org.factor,"Corymbia citriodora")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vitvi])
       {
         org.factor <- c(org.factor,"Vitis vinifera")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.kalfe])
       {
         org.factor <- c(org.factor,"Kalanchoe fedtschenkoi")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.vacda])
       {
         org.factor <- c(org.factor,"Vaccinium darrowii")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.oleeu])
       {
         org.factor <- c(org.factor,"Olea europaea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.mimgu])
       {
         org.factor <- c(org.factor,"Mimulus guttatus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.soltu])
       {
         org.factor <- c(org.factor,"Solanum tuberosum")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.cofar])
       {
         org.factor <- c(org.factor,"Coffea arabica")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.lacsa])
       {
         org.factor <- c(org.factor,"Lactuca sativa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.dauca])
       {
         org.factor <- c(org.factor,"Daucus carota")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.betvu])
       {
         org.factor <- c(org.factor,"Beta vulgaris")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.amahy])
       {
         org.factor <- c(org.factor,"Amaranthus hypochondriacus")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.chequ])
       {
         org.factor <- c(org.factor,"Chenopodium quinoa")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.sapof])
       {
         org.factor <- c(org.factor,"Saponaria officinalis")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.aquco])
       {
         org.factor <- c(org.factor,"Aquilegia coerulea")
       }
       
       else if (tree_reduced$tip.label[i] %in% tree$tip.label[tips_to_keep.query])
       {
         org.factor <- c(org.factor,"Query")
       }
       
     }
     
     
     #Matrix with labels and colors and transform to dplyr format
     string.sel.table <- data.frame(node = 1:length(tree_reduced$tip.label), label = tree_reduced$tip.label,
                                    org = org.factor)
     
     return(string.sel.table)
     
   }) %>% bindEvent(input$string_start5)
   
   # Now, selection is allowed for genes of species with STRING support
   observeEvent(input$string_start5, {
     
     string_sel_table <- string_sel_table5()
     allow_string_species <- c("Aegilops tauschii", "Arabidopsis thaliana", "Bathycoccus prasinos", "Chara braunii", 
                               "Chlamydomonas reinhardtii","Coccomyxa subellipsoidea", "Klebsormidium nitens", 
                               "Micromonas commoda", "Oryza sativa", "Micromonas pusilla", "Ostreococcus lucimarinus",
                               "Ostreococcus tauri", "Physcomitrium patens", "Raphidocelis subcapitata",
                               "Scenedesmus obliquus", "Selaginella moellendorffii", "Solanum lycopersicum",
                               "Sorghum bicolor", "Triticum aestivum", "Volvox carteri")
     
     st_genes <- subset(string_sel_table, string_sel_table$org %in% allow_string_species)$label
     
     if (length(st_genes) == 0)
     {
       shinyjs::showElement("error_string5")
       output$error_string5 <- renderUI({renderText({print("No results for this
      analysis due to lack of genes of STRING-supported species in the selection.")})})
       validate(" ")
     }
     
     output$error_string5 <- NULL
     
     insertUI("#selected_string5", "afterEnd", ui = {
       
       shinyWidgets::pickerInput("selected_stringI5","Select the desired genes from the tree",
                                 choices=st_genes, options = list(`actions-box` = TRUE),
                                 multiple = T)
       
     })
     
     
     insertUI("#string_selection5", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("string_selectionI5", "Show STRING Interactions", size = "sm",
                                style = "float", color = "royal")
     })
     
   })
   
   phys_table5 <- reactive({
     
     shinyjs::showElement(id = 'loading.string5')
     query_genes <- input$selected_stringI5
     string_sel_table <- string_sel_table5()
     
     # Load complete STRING annotation table
     library(data.table)
     
     # Load iteratively from split files using data.table format
     # data_phys <- fread("pharaoh_folder/string_physical/string_physical_1.tsv")
     # 
     # for (x in list.files("pharaoh_folder/string_physical/")[-1])
     # {
     #   data_phys <- rbind(data_phys, 
     #                      fread(paste0("pharaoh_folder/string_physical/", x)))
     # }
     
     # Load data only for selected organisms
     species_selected <- gsub("[.]", "", gsub(" ", "_", tolower(unique(string_sel_table$org))))
     
     # Get a List of all files named with a key word, say all `.csv` files
     filenames_string <- list.files("pharaoh_folder/string_physical/")
     files_selected <- filenames_string[unlist(sapply(species_selected, function(x) grep(x, filenames_string)))]
     
     # If in the future any of these species is added to STRING:
     # { 
     #   if (seq_org == "Zygnema circumcarinatum SAG")
     #   {
     #     org_search <- "zygnema_circumcarinatum"
     #   }
     #   else if (seq_org == "Zygnema circumcarinatum UTEX 1559")
     #   {
     #     org_search <- "zygnema_circumcarinatum1559"
     #   }
     #   else if (seq_org == "Mesostigma viride CCAC 1140")
     #   {
     #     org_search <- "mesostigma_viride1140"
     #   }
     #   else if (seq_org == "Mesostigma viride NIES 296")
     #   {
     #     org_search <- "mesostigma_viride296"
     #   }
     #   else
     #   {
     #    org_search <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(seq_org))))
     #   }
     # }
     
     # Load and bind all data sets
     data_phys <- rbindlist(lapply(paste0("pharaoh_folder/string_physical/", files_selected),fread))
     
     # Subset by query genes using data.table for speed
     #string_res <- subset(data_phys, data_phys$prot_query %in% query_genes)
     string_res <- data_phys[prot_query %in% query_genes,]
     string_res <- as.data.frame(string_res)
     
     # Assign OG ID to each target
     ortho_data_file <- "Gene_Trees/Orthogroups.tsv"
     
     ortho_data <- as.data.frame(fread(ortho_data_file))
     
     ortho_char <- apply(ortho_data, MARGIN = 1, function(x) paste(x, collapse = ","))
     ortho.numbers <- sapply(string_res$prot_interaction, function(x) grep(x, ortho_char), USE.NAMES = F)
     
     # If a pattern is found in several names, i.e., is a subpattern of several genes,
     # search for the exact match
     if(class(ortho.numbers) == "list")
     {
       # If a gene isn't associated to an OG
       index.none <- which(sapply(ortho.numbers, function(x) length(x) == 0))
       
       if (length(index.none) != 0)
       {
         # We create another row for OG table to associate those genes
         ortho_data <- rbind(ortho_data, "No OG")
         ortho.numbers[index.none] <- nrow(ortho_data)
       }
       
       # If a gene has more than one match due to subpatterns
       index.wrong <- which(sapply(ortho.numbers, function(x) length(x) > 1))
       {
         if (length(index.wrong) == 0)
         {
           ortho.numbers <- unlist(ortho.numbers, use.names = F)
         }
         else
         {
           for (i in index.wrong)
           {
             for (j in ortho.numbers[[i]])
             {
               ortho_char_split <- strsplit(ortho_char[j],split = ",")
               ortho_char_split_clean <- sapply(ortho_char_split, function(x) gsub(" ", "", x))
               if (string_res$prot_interaction[i] %in% ortho_char_split_clean)
               {
                 ortho.numbers[i] <- j
                 ortho.numbers <- unlist(ortho.numbers, use.names = F)
               }
             }
           }
         }
       }
     }
     
     
     
     # Create the final table
     ortho.string.names <- ortho_data$Orthogroup[ortho.numbers]
     phys_table <- data.frame(string_res, orthogroup = ortho.string.names)
     
     return(phys_table)
   }) %>% bindEvent(input$string_selectionI5)
   
   # Create count table to identify enriched OGs in STRING result
   string_counts5 <- reactive({
     
     phys_table <- phys_table5()
     string_counts <- sort(table(phys_table$orthogroup), decreasing = T)
     
     return(string_counts)
     
   }) %>% bindEvent(input$string_selectionI5)
   
   string_count_plot5 <- reactive({
     
     library(ggplot2)
     library(dplyr)
     
     data_count <- as.data.frame(string_counts5())
     colnames(data_count) <- c("orthogroup", "value")
     
     # Compute the position of labels
     data_count <- data_count %>%
       arrange(desc(orthogroup)) %>%
       mutate(prop = value / sum(data_count$value) *100) %>%
       mutate(ypos = cumsum(prop)- 0.5*prop )
     
     # Create plot
     count_plot <- ggplot(data_count, aes(x="", y=prop, fill=orthogroup)) +
       geom_bar(stat="identity", width=1, color="white") +
       coord_polar("y", start=0) +
       theme_void() +
       theme(legend.position="none") +
       #geom_text(aes(y = ypos, label = orthogroup), color = "white", size=6) +
       scale_fill_manual(values = rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"), 
                                      floor(nrow(data_count)/9)+1))
     
     return(count_plot)
     
   }) %>% bindEvent(input$string_selectionI5)
   
   string_count_plotly5 <- reactive({
     
     library(ggplot2)
     library(dplyr)
     
     data_count <- as.data.frame(string_counts5())
     colnames(data_count) <- c("orthogroup", "value")
     
     # Compute the position of labels
     data_count <- data_count %>%
       arrange(desc(orthogroup)) %>%
       mutate(prop = value / sum(data_count$value) *100) %>%
       mutate(ypos = cumsum(prop)- 0.5*prop )
     
     # Create plot
     count_plotly <- plotly::plot_ly(data=data_count,values=~prop,labels=~factor(orthogroup),
                                     marker=list(colors=rep(RColorBrewer::brewer.pal(n = 9, name = "Set1"),
                                                            floor(nrow(data_count)/9)+1)),
                                     type="pie",showlegend = F, text= ~paste0("</br> ", orthogroup,
                                                                              "</br> ",prop, "%"),
                                     textinfo = "none", hoverinfo = "text") 
     
     
     
     return(count_plotly)
     
   }) %>% bindEvent(input$string_selectionI5)
   
   # Create boxes for outputs
   observeEvent(isTruthy(string_count_plot5()), {
     
     if (UI_exist_string5)
     {
       removeUI(
         selector = "div:has(>> #output_st_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #output_count_table5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>>> #count_plot5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadSTRINGTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #downloadCOUNTTable5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #count_download5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI(
         selector = "div:has(>> #selected_networkI5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       removeUI("#network_buttonI5")
       
     }
     
     insertUI("#box_st_table5", "afterEnd", ui = {
       box(width = 12,
           title = "STRING Interactions Table", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_st_table5")
       )
     })
     
     insertUI("#box_count_table5", "afterEnd", ui = {
       box(width = 12,
           title = "Interacting Orthogroups Table", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           dataTableOutput("output_count_table5")
       )
     })
     
     insertUI("#box_count_plot5", "afterEnd", ui = {
       box(width = 12,
           title = "Interacting Orthogroups Plot", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           plotlyOutput("count_plot5")
       )
     })
     
     
     insertUI("#download_ui_for_st_table5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadSTRINGTable5", "Download STRING Table",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#download_ui_for_count_table5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "downloadCOUNTTable5", "Download OG Count Table",
                                                                          size = "sm", color = "primary"))
     })
     
     insertUI("#count_down_button5", "afterEnd", ui = {
       tags$div(style = "margin-left: 100px;", shinyWidgets::downloadBttn(outputId= "count_download5", "Download OG Count Plot",
                                                                          size = "sm", color = "primary"))
     })
     
     UI_exist_string5 <<- TRUE
     
     # Remove previous results for STRING network
     if (UI_exist_network5)
     {
       removeUI(
         selector = "div:has(>>>> #network_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
       UI_exist_network5 <<- F
     }
     
   })
   
   # Fill outputs
   # Render STRING table
   output$output_st_table5 <- renderDataTable({
     phys_table <- phys_table5()
     datatable(phys_table, escape=FALSE, rownames= F, options =list(pageLength = 10)) %>%
       formatStyle(
         'type',
         color = styleEqual(
           c("Direct interaction", "Interolog"), c('green', '#CA931B')
         )
       )
   }) 
   
   
   # Render OG count table
   output$output_count_table5 <- renderDataTable({
     string_counts <- as.data.frame(string_counts5())
     colnames(string_counts) <- c("orthogroup", "count")
     string_counts
   },escape=FALSE, rownames= F, options =list(pageLength = 7))
   
   # Render OG count pie chart
   output$count_plot5 <- renderPlotly({
     string_count_plotly5()
   })
   
   
   # Download tab's outputs
   # Download STRING table
   output$downloadSTRINGTable5 <- downloadHandler(
     filename= function() {
       paste("string_table", ".tsv", sep="")
     },
     content= function(file) {
       phys_table <- phys_table5()
       write.table(x = phys_table,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download count table
   output$downloadCOUNTTable5 <- downloadHandler(
     filename= function() {
       paste("string_count_table", ".tsv", sep="")
     },
     content= function(file) {
       string_counts <- string_counts5()
       write.table(x = string_counts,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   # Download count plot
   output$count_download5 <- downloadHandler(
     filename= function() {
       paste("string_count", ".png", sep="")
     },
     content= function(file) {
       string_count_plot <- string_count_plot5()
       
       png(file, height = 450, width = 450)
       plot(string_count_plot)
       dev.off()
     })
   
   # Create gene selector and button for STRING network representation
   observeEvent(input$string_selectionI5,{
     
     phys_table <- phys_table5()
     network_genes <- unique(phys_table$prot_query)
     
     # Error message if no genes are allowed for selection should have  been reported
     # earlier
     
     insertUI("#selected_network5", "afterEnd", ui = {
       
       shinyWidgets::pickerInput(inputId = "selected_networkI5", label = "Select the gene whose network you want to plot", 
                                 choices = network_genes, selected = network_genes[1],
                                 options = list(`actions-box` = TRUE), multiple = T)
       
     })
     
     
     insertUI("#network_button5", "afterEnd", ui = {
       
       shinyWidgets::actionBttn("network_buttonI5", "Plot STRING Network", size = "sm",
                                style = "float", color = "royal")
     })
     
     shinyjs::hideElement(id = 'loading.string5')
     
   })
   
   mapped_string5 <- reactive({
     
     # Load genes (PharaohFUN IDs) and convert to STRING IDs
     library(data.table)
     
     network_genes <- input$selected_networkI5
     map_table <- fread("pharaoh_folder/string_map.tsv")
     
     # Fast subset using data.table
     map_network <- map_table[pharaohfun_id %in% network_genes,]
     
     # Error if no genes from selection have an associated high fidelity STRING ID
     if (nrow(map_network) == 0)
     {
       
       if (UI_exist_network5)
       {
         removeUI(
           selector = "div:has(>>>> #network_image5)",
           multiple = TRUE,
           immediate = TRUE
         )
         
         UI_exist_network5 <<- F
       }
       
       output$error_network5 <- renderUI({renderText({print("It's not possible to map any
                               genes from selection to STRING IDs, please select different ones.")})})
       validate( " ")
       
     }
     
     output$error_network5 <- NULL
     
     # Get only STRING IDS and paste in a format interpretable by JS function
     string_ids <- as.data.frame(map_network)$string_id
     
     
     mapped_string <- paste0(string_ids, collapse = "%0d")
     return(mapped_string)
     
   }) %>% bindEvent(input$network_buttonI5)
   
   # Create boxes
   observeEvent(isTruthy(mapped_string5()),{
     
     mapped_string <- mapped_string5()
     
     if (UI_exist_network5)
     {
       removeUI(
         selector = "div:has(>>>> #network_image5)",
         multiple = TRUE,
         immediate = TRUE
       )
       
     }
     
     url_interactive <- paste0("https://string-db.org/cgi/network?identifiers=", 
                               mapped_string, 
                               "&add_color_nodes=25&network_flavor=confidence&show_query_node_labels=1")
     
     insertUI("#box_output_network5", "afterEnd", ui = {
       box(width = 12,
           title = "Image", status = "primary", solidHeader = TRUE,
           collapsible = TRUE,
           fluidRow(column(1), 
                    column(8, htmlOutput("network_image5")), 
                    column(3, div( style = "margin-top: 300px;", 
                                   shinyWidgets::actionBttn("network_link5", "Interactive Network", size = "md", 
                                                            icon = icon("circle-nodes"), style = "float", color = "royal", 
                                                            onclick=paste0("window.open('", url_interactive,"','_blank')")))
                           
                    ))
           
       )
     })
     
     UI_exist_network5 <<- T
     
   })
   
   # Fill network box
   
   output$network_image5 <- renderText({
     
     mapped_string <- mapped_string5()
     src_map <- paste0("https://string-db.org/api/image/network?identifiers=", mapped_string, 
                       "&add_color_nodes=25&network_flavor=confidence")
     
     c('<img src="',src_map,'"width="675" height="625">')
   })
   
   
# End of SHOOT Search
   
   ##################### WHOLE DATASET ##########################
   
   # Set global variables for tracking changes in output
   #UI_exist_data6 <<- F
   
   # To avoid autoupdating some inputs, define variables with its values
   model.selected6 <- reactive({
     model.selected <- F
     return(model.selected)
   })%>% bindEvent(input$run_button6)
   
   # Load organisms selection based on the model selected
   selected_organisms6 <- reactive({
     selected_organisms <- c(input$prasino_check_6, input$mami_check_6,
                             input$chloro_check_6, input$strepto_check_6,
                             input$bryo_check_6, input$lyco_check_6,
                             input$gymno_check_6, input$magno_check_6,
                             input$monocots_check_6, input$dicots_first_check_6,
                             input$dicots_second_check_6)
     return(selected_organisms)
     
   }) %>% bindEvent(input$run_button6)
   
   selected_values_org6 <- reactive(organisms_values[selected_organisms6()]) %>% bindEvent(input$run_button6)
   
   # Reactive for creating results table when Run button is clicked 
   whole_data_table6 <- reactive({
     
     # Create gene tables with reduced species for each OG (first word because of
     # posible errors in second word)
     selected_organisms <- selected_organisms6()
     org_sel <- as.character(sapply(selected_organisms, function(x) gsub(" ", "_", tolower(x))))
     org_first <- sapply(strsplit(org_sel, split = "_"), function(x) x[[1]])
     
     ortho.file <- "Gene_Trees/Orthogroups.tsv"
     
     table_ogs <- fread(ortho.file)
     org_index <- sapply(org_first, function(x) grep(x, colnames(table_ogs)), USE.NAMES = F)
     org_index <- as.numeric(c(1, org_index)) # 1 to include Orthogroups column
     table_ogs_red <- table_ogs[,c(org_index), with=F]
     
     return(table_ogs_red)
     
   }) %>% bindEvent(input$run_button6)
   
   
   # Text output
   output$text_dataset6 <- renderUI({renderText({
     
   selected_organisms <- selected_organisms6()
   table_ogs_red <- whole_data_table6()
   org_sel <- as.character(sapply(selected_organisms, function(x) gsub(" ", "_", tolower(x))))
   org_first <- sapply(strsplit(org_sel, split = "_"), function(x) x[[1]])
   
   print(paste0("Created table for the species: ", paste0(c(org_first), collapse = "/")))
      
         })
   })
   
   # Download count table
   output$downloadDataset6 <- downloadHandler(
     filename= function() {
       paste("dataset_table", ".tsv", sep="")
     },
     content= function(file) {
       whole_data_table <- whole_data_table6()
       write.table(x = whole_data_table,quote = F,sep = "\t",
                   file=file,row.names=FALSE,col.names=TRUE)
     })
   
   ############################ DOWNLOAD GENOMES  ########################
   
   gen_fasta7 <- reactive({
     organism_down <- input$organism_down_7
     
     { 
       if (organism_down == "Zygnema circumcarinatum SAG")
       {
         gen_search <- "zygnema_circumcarinatum"
       }
       else if (organism_down == "Zygnema circumcarinatum UTEX 1559")
       {
         gen_search <- "zygnema_circumcarinatum1559"
       }
       else if (organism_down == "Mesostigma viride CCAC 1140")
       {
         gen_search <- "mesostigma_viride1140"
       }
       else if (organism_down == "Mesostigma viride NIES 296")
       {
         gen_search <- "mesostigma_viride296"
       }
       else
       {
         gen_search <- gsub("[.]", "", gsub(" ", "_", tolower(as.character(organism_down))))
       }
     }
     gen_fasta <- paste("pharaohfun_proteomes", paste0(gen_search, ".fa", sep=""), sep = "/")
   })
   
   output$downloadGenomes7 <- downloadHandler(
     filename <- function() {
       strsplit(gen_fasta7(),split = "[/]")[[1]][2]
     },
     
     content <- function(file) {
       file.copy(gen_fasta7(), file)
     },
     contentType = "text/*"
   )
   
# End of the whole server function
     }




shinyApp(ui, server)
